--- a/src/main/java/io/papermc/paper/datacomponent/DataComponentAdapters.java
+++ b/src/main/java/io/papermc/paper/datacomponent/DataComponentAdapters.java
@@ -195,6 +_,15 @@
         register(DataComponents.SHEEP_COLOR, nms -> DyeColor.getByWoolData((byte) nms.getId()), api -> net.minecraft.world.item.DyeColor.byId(api.getWoolData()));
         register(DataComponents.SHULKER_COLOR, nms -> DyeColor.getByWoolData((byte) nms.getId()), api -> net.minecraft.world.item.DyeColor.byId(api.getWoolData()));
 
+        // KioCG start - customize default components
+        registerUntyped(DataComponents.CANNOT_PLACE);
+        registerUntyped(DataComponents.CANNOT_USE);
+        registerUntyped(DataComponents.SLOT_LOCK);
+        registerIdentity(DataComponents.CONTAINER_ROWS);
+        register(DataComponents.EXTENDED_BEFORE_LORE, PaperItemLore::new);
+        register(DataComponents.EXTENDED_AFTER_LORE, PaperItemLore::new);
+        // KioCG end - customize default components
+
         for (final Map.Entry<ResourceKey<DataComponentType<?>>, DataComponentType<?>> componentType : BuiltInRegistries.DATA_COMPONENT_TYPE.entrySet()) {
             if (!ADAPTERS.containsKey(componentType.getKey())) {
                 registerUnimplemented(componentType.getValue());
