--- a/net/minecraft/world/Containers.java
+++ b/net/minecraft/world/Containers.java
@@ -37,7 +_,10 @@
         double d5 = Math.floor(z) + level.random.nextDouble() * d1 + d2;
 
         while (!stack.isEmpty()) {
-            ItemEntity itemEntity = new ItemEntity(level, d3, d4, d5, stack.split(level.random.nextInt(21) + 10));
+            // KioCG start - allow overstacked items: don't split items
+            ItemEntity itemEntity = new ItemEntity(level, d3, d4, d5, stack.copy());
+            stack.setCount(0);
+            // KioCG end - allow overstacked items: don't split items
             float f = 0.05F;
             itemEntity.setDeltaMovement(
                 level.random.triangle(0.0, 0.11485000171139836),
