--- a/net/minecraft/world/effect/WeavingMobEffect.java
+++ b/net/minecraft/world/effect/WeavingMobEffect.java
@@ -24,7 +_,7 @@
 
     @Override
     public void onMobRemoved(ServerLevel level, LivingEntity entity, int amplifier, Entity.RemovalReason reason) {
-        if (reason == Entity.RemovalReason.KILLED && (entity instanceof Player || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))) {
+        if (reason == Entity.RemovalReason.KILLED && (false || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1))) { // KioCG - better mob griefing setting
             this.spawnCobwebsRandomlyAround(level, entity.getRandom(), entity.blockPosition(), entity); // Paper - Fire EntityChangeBlockEvent in more places
         }
     }
