--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -79,6 +_,12 @@
 
     @Override
     protected void onTake(Player player, ItemStack stack) {
+        // KioCG start - AnvilResultTakenEvent
+        org.bukkit.inventory.ItemStack oldFirstItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(this.inputSlots.getItem(0));
+        org.bukkit.inventory.ItemStack oldSecondItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(this.inputSlots.getItem(1));
+        org.bukkit.inventory.ItemStack oldResult = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack);
+        // KioCG end - AnvilResultTakenEvent
+
         if (!player.hasInfiniteMaterials()) {
             player.giveExperienceLevels(-this.cost.get());
         }
@@ -97,9 +_,10 @@
 
         this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
         this.inputSlots.setItem(0, ItemStack.EMPTY);
+        new com.kiocg.event.inventory.AnvilResultTakenEvent(player.getBukkitEntity(), getBukkitView(), oldResult, oldFirstItem, oldSecondItem).callEvent(); // KioCG - AnvilResultTakenEvent
         this.access.execute((level, blockPos) -> {
             BlockState blockState = level.getBlockState(blockPos);
-            if (!player.hasInfiniteMaterials() && blockState.is(BlockTags.ANVIL) && player.getRandom().nextFloat() < 0.12F) {
+            if (false && !player.hasInfiniteMaterials() && blockState.is(BlockTags.ANVIL) && player.getRandom().nextFloat() < 0.12F) { // KioCG - adjust anvil: 铁砧无法损坏
                 BlockState blockState1 = AnvilBlock.damage(blockState);
                 // Paper start - AnvilDamageEvent
                 com.destroystokyo.paper.event.block.AnvilDamagedEvent event = new com.destroystokyo.paper.event.block.AnvilDamagedEvent(getBukkitView(), blockState1 != null ? org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(blockState1) : null);
@@ -129,16 +_,27 @@
         ItemStack item = this.inputSlots.getItem(0);
         this.onlyRenaming = false;
         this.cost.set(1);
-        int i = 0;
-        long l = 0L;
-        int i1 = 0;
+        int i = 0; // 铁砧基础消耗
+        long l = 0L; // 铁砧累计惩罚
+        int i1 = 0; // 重命名消耗
+        boolean allowZeroLevel = false; // KioCG - adjust anvil
         if (!item.isEmpty() && EnchantmentHelper.canStoreEnchantments(item)) {
             ItemStack itemStack = item.copy();
-            ItemStack item1 = this.inputSlots.getItem(1);
+            ItemStack item1 = this.inputSlots.getItem(1); ItemStack itemAtSlot2 = item1; // KioCG - OBFHELPER
             ItemEnchantments.Mutable mutable = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(itemStack));
             l += (long)item.getOrDefault(DataComponents.REPAIR_COST, 0).intValue() + item1.getOrDefault(DataComponents.REPAIR_COST, 0).intValue();
+            l = Math.min(l, io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.anvil.maxCumulativeCost); // KioCG - adjust anvil
             this.repairItemCountCost = 0;
             if (!item1.isEmpty()) {
+                // KioCG start - vanilla recipe excludes model items
+                if (!this.allowModelItem(itemAtSlot2)) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.EMPTY); // CraftBukkit
+                    this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
+                    sendAllDataToRemote();
+                    return;
+                }
+                // KioCG end - vanilla recipe excludes model items
+
                 boolean hasStoredEnchantments = item1.has(DataComponents.STORED_ENCHANTMENTS);
                 if (itemStack.isDamageableItem() && item.isValidRepairItem(item1)) {
                     int min = Math.min(itemStack.getDamageValue(), itemStack.getMaxDamage() / 4);
@@ -153,6 +_,12 @@
                         int i3 = itemStack.getDamageValue() - min;
                         itemStack.setDamageValue(i3);
                         i++;
+                        // KioCG start - adjust anvil
+                        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.anvil.repairNoLevelCost) {
+                            allowZeroLevel = true;
+                            i--;
+                        }
+                        // KioCG end - adjust anvil
                         min = Math.min(itemStack.getDamageValue(), itemStack.getMaxDamage() / 4);
                     }
 
@@ -235,21 +_,33 @@
                 if (!this.itemName.equals(item.getHoverName().getString())) {
                     i1 = 1;
                     i += i1;
-                    itemStack.set(DataComponents.CUSTOM_NAME, Component.literal(this.itemName));
+                    // KioCG start - adjust anvil
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.anvil.renameNoLevelCost) {
+                        allowZeroLevel = true;
+                        i -= i1;
+                    }
+                    // KioCG end - adjust anvil
+                    itemStack.set(DataComponents.CUSTOM_NAME, java.util.Objects.requireNonNullElseGet(renameColors(), () -> Component.literal(this.itemName))); // KioCG - adjust anvil
                 }
             } else if (item.has(DataComponents.CUSTOM_NAME)) {
                 i1 = 1;
                 i += i1;
+                // KioCG start - adjust anvil
+                if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.anvil.renameNoLevelCost) {
+                    allowZeroLevel = true;
+                    i -= i1;
+                }
+                // KioCG end - adjust anvil
                 itemStack.remove(DataComponents.CUSTOM_NAME);
             }
 
             int i6 = i <= 0 ? 0 : (int)Mth.clamp(l + i, 0L, 2147483647L);
             this.cost.set(i6);
-            if (i <= 0) {
+            if (i < 0 || i == 0 && !allowZeroLevel) { // KioCG - adjust anvil
                 itemStack = ItemStack.EMPTY;
             }
 
-            if (i1 == i && i1 > 0) {
+            if ((i1 == i || io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.anvil.renameNoLevelCost && i == 0) && i1 > 0) { // KioCG - adjust anvil
                 // CraftBukkit start
                 if (this.cost.get() >= this.maximumRepairCost) {
                     this.cost.set(this.maximumRepairCost - 1);
@@ -270,6 +_,7 @@
                 }
 
                 if (i1 != i || i1 == 0) {
+                    if (this.cost.get() != 0) // KioCG - adjust anvil: 不消耗等级不要增加铁砧惩罚
                     minxx = calculateIncreasedRepairCost(minxx);
                 }
 
@@ -290,11 +_,52 @@
         return (int)Math.min(oldRepairCost * 2L + 1L, 2147483647L);
     }
 
+    // KioCG start - adjust anvil
+    @Nullable
+    private Component renameColors() {
+        if (this.player == null) return null;
+
+        org.bukkit.craftbukkit.entity.CraftHumanEntity player = this.player.getBukkitEntity();
+        String name = this.itemName;
+        boolean removeItalics = false;
+        if (player.hasPermission("kiocg.anvil.remove_italics")) {
+            if (name.startsWith("&r")) {
+                name = name.substring(2);
+                removeItalics = true;
+            } else if (name.startsWith("<reset>")) {
+                name = name.substring(7);
+                removeItalics = true;
+            }
+        }
+
+        net.kyori.adventure.text.Component component;
+        if (player.hasPermission("kiocg.anvil.minimessage")) {
+            component = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(name);
+        } else {
+            // 不能允许大写字母, 会被转义为正体字
+            if (player.hasPermission("kiocg.anvil.color")) {
+                name = name.replaceAll("&([0-9a-fr])", "\u00a7$1");
+                name = name.replaceAll("&#([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])", "\u00a7x\u00a7$1\u00a7$2\u00a7$3\u00a7$4\u00a7$5\u00a7$6");
+            }
+            if (player.hasPermission("kiocg.anvil.format")) {
+                name = name.replaceAll("&([k-or])", "\u00a7$1");
+            }
+
+            component = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(name);
+        }
+
+        if (removeItalics) {
+            component = component.decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false);
+        }
+        return io.papermc.paper.adventure.PaperAdventure.asVanilla(component);
+    }
+    // KioCG end - adjust anvil
+
     public boolean setItemName(String itemName) {
         String string = validateName(itemName);
         if (string != null && !string.equals(this.itemName)) {
             this.itemName = string;
-            if (this.getSlot(2).hasItem()) {
+            if (false && this.getSlot(2).hasItem()) { // KioCG - 不需要设置物品名, 因为将在createResult()中设置
                 ItemStack item = this.getSlot(2).getItem();
                 if (StringUtil.isBlank(string)) {
                     item.remove(DataComponents.CUSTOM_NAME);
