--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -26,11 +_,18 @@
     }
 
     public PlayerEnderChestContainer(Player owner) {
-        super(27);
+        super(6 * 9); // KioCG - expand container rows: 始终加载最大末影箱
         this.owner = owner;
         // CraftBukkit end
     }
 
+    // KioCG start - expand container rows
+    @Override
+    public int getContainerSize() {
+        return this.owner.enderchestRows * 9;
+    }
+    // KioCG end - expand container rows
+
     public void setActiveChest(EnderChestBlockEntity enderChestBlockEntity) {
         this.activeChest = enderChestBlockEntity;
     }
@@ -40,19 +_,19 @@
     }
 
     public void fromSlots(ValueInput.TypedInputList<ItemStackWithSlot> input) {
-        for (int i = 0; i < this.getContainerSize(); i++) {
+        for (int i = 0; i < 6 * 9; i++) { // KioCG - expand container rows: 始终加载最大末影箱
             this.setItem(i, ItemStack.EMPTY);
         }
 
         for (ItemStackWithSlot itemStackWithSlot : input) {
-            if (itemStackWithSlot.isValidInContainer(this.getContainerSize())) {
+            if (itemStackWithSlot.isValidInContainer(6 * 9)) { // KioCG - expand container rows: 始终加载最大末影箱
                 this.setItem(itemStackWithSlot.slot(), itemStackWithSlot.stack());
             }
         }
     }
 
     public void storeAsSlots(ValueOutput.TypedOutputList<ItemStackWithSlot> output) {
-        for (int i = 0; i < this.getContainerSize(); i++) {
+        for (int i = 0; i < 6 * 9; i++) { // KioCG - expand container rows: 始终加载最大末影箱
             ItemStack item = this.getItem(i);
             if (!item.isEmpty()) {
                 output.add(new ItemStackWithSlot(i, item));
