--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -25,11 +_,18 @@
     }
 
     public PlayerEnderChestContainer(Player owner) {
-        super(27);
+        super(6 * 9); // KioCG - expand container rows: 始终加载最大末影箱
         this.owner = owner;
         // CraftBukkit end
     }
 
+    // KioCG start - expand container rows
+    @Override
+    public int getContainerSize() {
+        return this.owner.enderchestRows * 9;
+    }
+    // KioCG end - expand container rows
+
     public void setActiveChest(EnderChestBlockEntity enderChestBlockEntity) {
         this.activeChest = enderChestBlockEntity;
     }
@@ -40,14 +_,14 @@
 
     @Override
     public void fromTag(ListTag tag, HolderLookup.Provider levelRegistry) {
-        for (int i = 0; i < this.getContainerSize(); i++) {
+        for (int i = 0; i < 6 * 9; i++) { // KioCG - expand container rows: 始终加载最大末影箱
             this.setItem(i, ItemStack.EMPTY);
         }
 
         for (int i = 0; i < tag.size(); i++) {
             CompoundTag compound = tag.getCompound(i);
             int i1 = compound.getByte("Slot") & 255;
-            if (i1 >= 0 && i1 < this.getContainerSize()) {
+            if (i1 >= 0 && i1 < 6 * 9) { // KioCG - expand container rows: 始终加载最大末影箱
                 this.setItem(i1, ItemStack.parse(levelRegistry, compound).orElse(ItemStack.EMPTY));
             }
         }
@@ -57,7 +_,7 @@
     public ListTag createTag(HolderLookup.Provider levelRegistry) {
         ListTag listTag = new ListTag();
 
-        for (int i = 0; i < this.getContainerSize(); i++) {
+        for (int i = 0; i < 6 * 9; i++) { // KioCG - expand container rows: 始终加载最大末影箱
             ItemStack item = this.getItem(i);
             if (!item.isEmpty()) {
                 CompoundTag compoundTag = new CompoundTag();
