--- a/net/minecraft/world/inventory/Slot.java
+++ b/net/minecraft/world/inventory/Slot.java
@@ -42,6 +_,7 @@
     }
 
     public boolean mayPlace(ItemStack stack) {
+        if (container instanceof net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity) return stack.canFitInsideContainerItems(); // KioCG - expand container rows: 修复潜影盒嵌套 // KioCG - prevent stacking containers
         return true;
     }
 
