--- a/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/net/minecraft/world/inventory/StonecutterMenu.java
@@ -119,6 +_,12 @@
         return this.recipesForInput;
     }
 
+    // KioCG start - more stonecutter api
+    public void setVisibleRecipes(SelectableRecipe.SingleInputSet<StonecutterRecipe> recipes) {
+        this.recipesForInput = recipes;
+    }
+    // KioCG end - more stonecutter api
+
     public int getNumberOfVisibleRecipes() {
         return this.recipesForInput.size();
     }
@@ -180,7 +_,7 @@
     @Override
     public void slotsChanged(Container inventory) {
         ItemStack item = this.inputSlot.getItem();
-        if (!item.is(this.input.getItem())) {
+        if (!ItemStack.isSameItemSameComponents(item, this.input)) { // KioCG - 更准确的判断
             this.input = item.copy();
             this.setupRecipeList(item);
         }
@@ -192,11 +_,32 @@
         this.resultSlot.set(ItemStack.EMPTY);
         if (!stack.isEmpty()) {
             this.recipesForInput = this.level.recipeAccess().stonecutterRecipes().selectByInput(stack);
+            // KioCG start - vanilla recipe excludes model items
+            if (!this.allowModelItem(stack) && !this.recipesForInput.isEmpty()) {
+                this.recipesForInput = new SelectableRecipe.SingleInputSet<>(this.recipesForInput.entries().stream().filter(entry -> entry.recipe().recipe().filter(recipeHolder -> recipeHolder.id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)).isEmpty()).toList());
+            }
+            // KioCG end - vanilla recipe excludes model items
+
+            if (((org.bukkit.craftbukkit.entity.CraftHumanEntity) this.player).getHandle() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) refreshRecipe(serverPlayer, true); // KioCG - fix stonecutter recipe list display
         } else {
             this.recipesForInput = SelectableRecipe.SingleInputSet.empty();
         }
     }
 
+    // KioCG start - fix stonecutter recipe list display
+    public void refreshRecipe(net.minecraft.server.level.ServerPlayer serverPlayer, boolean visibleRecipes) {
+        net.minecraft.world.item.crafting.RecipeManager recipeManager = net.minecraft.server.MinecraftServer.getServer().getRecipeManager();
+        net.minecraft.network.protocol.game.ClientboundUpdateRecipesPacket clientboundUpdateRecipesPacket = new net.minecraft.network.protocol.game.ClientboundUpdateRecipesPacket(
+                recipeManager.getSynchronizedItemProperties(), visibleRecipes ? this.getVisibleRecipes() : recipeManager.getSynchronizedStonecutterRecipes()
+        );
+        serverPlayer.connection.send(clientboundUpdateRecipesPacket);
+
+        // Force client to update recipe list
+        serverPlayer.containerSynchronizer.sendSlotChange(this, INPUT_SLOT, new ItemStack(net.minecraft.world.item.Items.AIR));
+        serverPlayer.containerSynchronizer.sendSlotChange(this, INPUT_SLOT, this.inputSlot.getItem());
+    }
+    // KioCG end - fix stonecutter recipe list display
+
     void setupResultSlot(int id) {
         Optional<RecipeHolder<StonecutterRecipe>> optional;
         if (!this.recipesForInput.isEmpty() && this.isValidRecipeIndex(id)) {
@@ -286,5 +_,7 @@
         super.removed(player);
         this.resultContainer.removeItemNoUpdate(1);
         this.access.execute((level, pos) -> this.clearContainer(player, this.container));
+
+        if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer) refreshRecipe(serverPlayer, false); // KioCG - fix stonecutter recipe list display
     }
 }
