--- a/net/minecraft/core/component/DataComponents.java
+++ b/net/minecraft/core/component/DataComponents.java
@@ -231,13 +_,13 @@
         "debug_stick_state", builder -> builder.persistent(DebugStickState.CODEC).cacheEncoding()
     );
     public static final DataComponentType<CustomData> ENTITY_DATA = register(
-        "entity_data", builder -> builder.persistent(CustomData.CODEC_WITH_ID).networkSynchronized(CustomData.STREAM_CODEC)
+        "entity_data", builder -> builder.persistent(CustomData.CODEC_WITH_ID).networkSynchronized(io.papermc.paper.util.sanitizer.OversizedItemComponentSanitizer.CUSTOM_DATA) // KioCG - sanitize custom data
     );
     public static final DataComponentType<CustomData> BUCKET_ENTITY_DATA = register(
         "bucket_entity_data", builder -> builder.persistent(CustomData.CODEC).networkSynchronized(CustomData.STREAM_CODEC)
     );
     public static final DataComponentType<CustomData> BLOCK_ENTITY_DATA = register(
-        "block_entity_data", builder -> builder.persistent(CustomData.CODEC_WITH_ID).networkSynchronized(CustomData.STREAM_CODEC)
+        "block_entity_data", builder -> builder.persistent(CustomData.CODEC_WITH_ID).networkSynchronized(io.papermc.paper.util.sanitizer.OversizedItemComponentSanitizer.CUSTOM_DATA) // KioCG - sanitize custom data
     );
     public static final DataComponentType<InstrumentComponent> INSTRUMENT = register(
         "instrument", builder -> builder.persistent(InstrumentComponent.CODEC).networkSynchronized(InstrumentComponent.STREAM_CODEC).cacheEncoding()
@@ -393,4 +_,12 @@
     private static <T> DataComponentType<T> register(String name, UnaryOperator<DataComponentType.Builder<T>> builder) {
         return Registry.register(BuiltInRegistries.DATA_COMPONENT_TYPE, name, builder.apply(DataComponentType.builder()).build());
     }
+
+    // KioCG start - customize default components: 确保最后一个注册以免影响数据包
+    public static final DataComponentType<ItemLore> EXTENDED_BEFORE_LORE = register("extended_before_lore", builder -> builder.persistent(ItemLore.CODEC));
+    public static final DataComponentType<ItemLore> EXTENDED_AFTER_LORE = register("extended_after_lore", builder -> builder.persistent(ItemLore.CODEC));
+    public static final DataComponentType<CustomData> STACKABLE_CUSTOM_DATA = register("stackable_custom_data", builder -> builder.persistent(CustomData.CODEC));
+
+    public static final java.util.Set<DataComponentType<?>> UNABLE_SEND_TO_CLIENT = java.util.Set.of(EXTENDED_BEFORE_LORE, EXTENDED_AFTER_LORE, STACKABLE_CUSTOM_DATA);
+    // KioCG end - customize default components: 确保最后一个注册以免影响数据包
 }
