--- a/src/main/java/io/papermc/paper/util/OversizedItemComponentSanitizer.java
+++ b/src/main/java/io/papermc/paper/util/OversizedItemComponentSanitizer.java
@@ -18,6 +_,7 @@
     /*
     These represent codecs that are meant to help get rid of possibly big items by ALWAYS hiding this data.
      */
+    public static final StreamCodec<? super RegistryFriendlyByteBuf, net.minecraft.world.item.component.CustomData> CUSTOM_DATA = codec(net.minecraft.world.item.component.CustomData.STREAM_CODEC, customData -> net.minecraft.world.item.component.CustomData.EMPTY); // KioCG - sanitize custom data
     public static final StreamCodec<RegistryFriendlyByteBuf, ChargedProjectiles> CHARGED_PROJECTILES = codec(ChargedProjectiles.STREAM_CODEC, OversizedItemComponentSanitizer::sanitizeChargedProjectiles);
     public static final StreamCodec<RegistryFriendlyByteBuf, ItemContainerContents> CONTAINER = codec(ItemContainerContents.STREAM_CODEC, contents -> ItemContainerContents.EMPTY);
     public static final StreamCodec<RegistryFriendlyByteBuf, BundleContents> BUNDLE_CONTENTS = new StreamCodec<>() {
