--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -385,4 +_,11 @@
     public void clearReputations() {
         getHandle().getGossips().gossips.clear();
     }
+
+    // KioCG start - lobotomized villager
+    @Override
+    public boolean isLobotomized() {
+        return getHandle().isLobotomized();
+    }
+    // KioCG end - lobotomized villager
 }
