--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -222,6 +_,7 @@
     private double health = 20;
     private boolean scaledHealth = false;
     private double healthScale = 20;
+    public int realAir; // KioCG - the long dark
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
     public org.bukkit.event.player.PlayerResourcePackStatusEvent.Status resourcePackStatus; // Paper - more resource pack API
@@ -249,6 +_,29 @@
         }
     }
 
+    // KioCG start - expand container rows
+    @Override
+    public void recalculatePermissions() {
+        super.recalculatePermissions();
+
+        if (this.hasPermission("kiocg.enderchestrows.6")) {
+            this.getHandle().enderchestRows = 6;
+        } else if (this.hasPermission("kiocg.enderchestrows.5")) {
+            this.getHandle().enderchestRows = 5;
+        } else if (this.hasPermission("kiocg.enderchestrows.4")) {
+            this.getHandle().enderchestRows = 4;
+        } else if (this.hasPermission("kiocg.enderchestrows.3")) {
+            this.getHandle().enderchestRows = 3;
+        } else if (this.hasPermission("kiocg.enderchestrows.2")) {
+            this.getHandle().enderchestRows = 2;
+        } else if (this.hasPermission("kiocg.enderchestrows.1")) {
+            this.getHandle().enderchestRows = 1;
+        } else {
+            this.getHandle().enderchestRows = 3;
+        }
+    }
+    // KioCG end - expand container rows
+
     @Override
     public boolean isOp() {
         return this.server.getHandle().isOp(this.getProfile());
@@ -2247,7 +_,7 @@
         playerList.sendAllPlayerInfo(handle);
 
         // Resend their XP and effects because the respawn packet resets it
-        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.experienceProgress, handle.totalExperience, handle.experienceLevel));
+        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.coldData.getColdProgress(), handle.totalExperience, handle.experienceLevel)); // KioCG - the long dark
         for (net.minecraft.world.effect.MobEffectInstance mobEffect : handle.getActiveEffects()) {
             connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket(handle.getId(), mobEffect, false));
         }
@@ -3504,6 +_,54 @@
         return FeatureHooks.isChunkSent(this.getHandle(), chunkKey);
     }
     // Paper end
+
+    // KioCG start - player in structure api
+    @Override
+    public boolean isInStructure() {
+        return getHandle().inStructure;
+    }
+    // KioCG end - player in structure api
+
+    // KioCG start - the long dark
+    @Override
+    public int getColdValue() {
+        return this.getHandle().coldData.getColdValue();
+    }
+    @Override
+    public void setColdValue(int value) {
+        this.getHandle().coldData.setColdValue(value);
+    }
+    @Override
+    public void addColdValue(int add) {
+        this.getHandle().coldData.addColdValue(add);
+    }
+
+    @Override
+    public int getThirstValue() {
+        return this.getHandle().thirstData.getThirstValue();
+    }
+    @Override
+    public void setThirstValue(int value) {
+        this.getHandle().thirstData.setThirstValue(value);
+    }
+    @Override
+    public void addThirstValue(int add) {
+        this.getHandle().thirstData.addThirstValue(add);
+    }
+
+    @Override
+    public float getThirstRegain() {
+        return this.getHandle().thirstData.getThirstRegain();
+    }
+    @Override
+    public void setThirstRegain(float value) {
+        this.getHandle().thirstData.setThirstRegain(value);
+    }
+    @Override
+    public void addThirstRegain(float add) {
+        this.getHandle().thirstData.addThirstRegain(add);
+    }
+    // KioCG end - the long dark
 
     public Player.Spigot spigot()
     {
