--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java
@@ -52,6 +_,16 @@
         return this.getHandle().isTame();
     }
 
+    // KioCG start - dont kill my pet
+    public boolean isFollowOwner() {
+        return this.getHandle().isFollowOwner();
+    }
+
+    public void setFollowOwner(boolean followOwner) {
+        this.getHandle().setFollowOwner(followOwner);
+    }
+    // KioCG end - dont kill my pet
+
     @Override
     public void setOwner(AnimalTamer tamer) {
         if (tamer != null) {
