--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -73,6 +_,7 @@
     private final EntityType entityType;
     private EntityDamageEvent lastDamageEvent;
     private final CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(CraftEntity.DATA_TYPE_REGISTRY);
+    public final java.util.Map<String, Object> temporaryMeta = new java.util.HashMap<>(); // KioCG - temporary meta
     protected net.kyori.adventure.pointer.Pointers adventure$pointers; // Paper - implement pointers
     // Paper start - Folia shedulers
     public final io.papermc.paper.threadedregions.EntityScheduler taskScheduler = new io.papermc.paper.threadedregions.EntityScheduler(this);
@@ -915,6 +_,13 @@
     public CraftPersistentDataContainer getPersistentDataContainer() {
         return this.persistentDataContainer;
     }
+
+    // KioCG start - temporary meta
+    @Override
+    public java.util.Map<String, Object> getTemporaryMeta() {
+        return this.temporaryMeta;
+    }
+    // KioCG end - temporary meta
 
     @Override
     public Pose getPose() {
