--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -431,7 +_,7 @@
 
     @Override
     public boolean requiresCustomPersistence() {
-        return super.requiresCustomPersistence() || this.getCarriedBlock() != null;
+        return super.requiresCustomPersistence()/* || this.getCarriedBlock() != null*/; // KioCG - optimize entity behavior
     }
 
     static class EndermanFreezeWhenLookedAt extends Goal {
@@ -545,6 +_,13 @@
             this.pendingTarget = getServerLevel(this.enderman).getNearestPlayer(this.startAggroTargetConditions.range(this.getFollowDistance()), this.enderman);
             return this.pendingTarget != null;
         }
+
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return 10;
+        }
+        // KioCG end - optimize entity behavior
 
         @Override
         public void start() {
