--- a/net/minecraft/world/entity/monster/Ravager.java
+++ b/net/minecraft/world/entity/monster/Ravager.java
@@ -141,7 +_,7 @@
 
             if (this.level() instanceof ServerLevel serverLevel
                 && this.horizontalCollision
-                && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) { // KioCG - better mob griefing setting
                 boolean flag = false;
                 AABB aabb = this.getBoundingBox().inflate(0.2);
 
@@ -227,7 +_,7 @@
     private void roar() {
         if (this.isAlive()) {
             if (this.level() instanceof ServerLevel serverLevel) {
-                Predicate<Entity> predicate = serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+                Predicate<Entity> predicate = false // KioCG - better mob griefing setting
                     ? ROAR_TARGET_WITH_GRIEFING
                     : ROAR_TARGET_WITHOUT_GRIEFING;
 
