--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -135,7 +_,7 @@
     @Override
     protected void dropCustomDeathLoot(ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
         super.dropCustomDeathLoot(level, damageSource, recentlyHit);
-        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull()) {
+        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull(this)) { // KioCG - drop mobs skull chance
             creeper.increaseDroppedSkulls();
             this.spawnAtLocation(level, Items.SKELETON_SKULL);
         }
