--- a/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -73,7 +_,7 @@
     @Override
     protected void dropCustomDeathLoot(ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
         super.dropCustomDeathLoot(level, damageSource, recentlyHit);
-        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull()) {
+        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull(this)) { // KioCG - drop mobs skull chance
             creeper.increaseDroppedSkulls();
             this.spawnAtLocation(level, Items.WITHER_SKELETON_SKULL);
         }
@@ -96,6 +_,11 @@
         SpawnGroupData spawnGroupData1 = super.finalizeSpawn(level, difficulty, spawnReason, spawnGroupData);
         this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(4.0);
         this.reassessWeaponGoal();
+        // KioCG start - anti mob farm
+        if (level.getBlockState(this.blockPosition()).is(net.minecraft.world.level.block.Blocks.WITHER_ROSE)) {
+            level.removeBlock(this.blockPosition(), false);
+        }
+        // KioCG end - anti mob farm
         return spawnGroupData1;
     }
 
