--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -109,14 +_,24 @@
     }
 
     protected void addBehaviourGoals() {
+        // KioCG start - monster strengthens
+        if (this.level().paperConfig().kiocgConfig.entity.monster.babyZombieJockey) {
+            this.goalSelector.addGoal(3, new net.minecraft.world.entity.ai.goal.LeapAtTargetGoal(this, 0.4F) {
+                @Override
+                public boolean canUse() {
+                    return Zombie.this.isBaby() && super.canUse();
+                }
+            });
+        }
+        // KioCG end - monster strengthens
         this.goalSelector.addGoal(2, new ZombieAttackGoal(this, 1.0, false));
-        this.goalSelector.addGoal(6, new MoveThroughVillageGoal(this, 1.0, true, 4, this::canBreakDoors));
+        // this.goalSelector.addGoal(6, new MoveThroughVillageGoal(this, 1.0, true, 4, this::canBreakDoors)); // KioCG - optimize entity behavior
         this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.targetSelector.addGoal(1, new HurtByTargetGoal(this).setAlertOthers(ZombifiedPiglin.class));
         this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, true));
         if (this.level().spigotConfig.zombieAggressiveTowardsVillager) this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false)); // Spigot
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, IronGolem.class, true));
-        this.targetSelector.addGoal(5, new NearestAttackableTargetGoal<>(this, Turtle.class, 10, true, false, Turtle.BABY_ON_LAND_SELECTOR));
+        // this.targetSelector.addGoal(5, new NearestAttackableTargetGoal<>(this, Turtle.class, 10, true, false, Turtle.BABY_ON_LAND_SELECTOR)); // KioCG - optimize entity behavior
     }
 
     public static AttributeSupplier.Builder createAttributes() {
@@ -181,7 +_,10 @@
     }
 
     @Override
-    public void setBaby(boolean childZombie) {
+    // KioCG start - entity baby api
+    public void setBaby1(boolean childZombie) {
+        if (childZombie) removeScaleAttributes();
+    // KioCG end - entity baby api
         this.getEntityData().set(DATA_BABY_ID, childZombie);
         if (this.level() != null && !this.level().isClientSide) {
             AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
@@ -376,6 +_,33 @@
         }
     }
 
+    // KioCG start - monster strengthens
+    public void die(DamageSource damageSource) {
+        if (!this.isRemoved() && !this.dead) {
+            if (damageSource.is(net.minecraft.world.damagesource.DamageTypes.PLAYER_ATTACK) && this.random.nextFloat() < (float) this.level().paperConfig().kiocgConfig.entity.monster.zombiesSkinnedChance) {
+                Skeleton converted = this.convertTo(EntityType.SKELETON, ConversionParams.single(this, true, true), skeleton -> {
+                    if (!this.isSilent()) {
+                        this.level().levelEvent(null, net.minecraft.world.level.block.LevelEvent.SOUND_SKELETON_TO_STRAY, this.blockPosition(), 0);
+                    }
+
+                    skeleton.setTarget(this.getTarget(), null, false);
+                    skeleton.getNavigation().moveTo(this.getNavigation().getPath(), 1.2D);
+                    skeleton.setAggressive(true); // 僵尸始终抬起手
+                    skeleton.setRemainingFireTicks(this.getRemainingFireTicks());
+                }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.PEELING, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PEELING);
+                if (converted != null) {
+                    this.forceDrops = true;
+                    this.dropFromLootTable(this.serverLevel(), damageSource, this.lastHurtByPlayerTime > 0);
+                    this.forceDrops = false;
+                    return;
+                }
+            }
+        }
+
+        super.die(damageSource);
+    }
+    // KioCG end - monster strengthens
+
     @Override
     public boolean doHurtTarget(ServerLevel level, Entity source) {
         boolean flag = super.doHurtTarget(level, source);
@@ -609,7 +_,7 @@
     @Override
     protected void dropCustomDeathLoot(ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
         super.dropCustomDeathLoot(level, damageSource, recentlyHit);
-        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull()) {
+        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull(this)) { // KioCG - drop mobs skull chance
             ItemStack skull = this.getSkull();
             if (!skull.isEmpty()) {
                 creeper.increaseDroppedSkulls();
