--- a/net/minecraft/world/item/component/Consumable.java
+++ b/net/minecraft/world/item/component/Consumable.java
@@ -82,6 +_,15 @@
             CriteriaTriggers.CONSUME_ITEM.trigger(serverPlayer, stack);
         }
 
+        // kioCG start - the long dark
+        if (entity instanceof Player player) {
+            stack.getOrDefault(DataComponents.CUSTOM_DATA, CustomData.EMPTY).getUnsafe()
+                 .getCompound("KioCG.FoodExtended")
+                 .map(tag -> tag.getIntOr("moisture", 0))
+                 .ifPresent(moisture -> player.thirstData.addThirstValue(moisture));
+        }
+        // kioCG end - the long dark
+
         stack.getAllOfType(ConsumableListener.class).forEach(consumableListener -> consumableListener.onConsume(level, entity, stack, this));
         if (!level.isClientSide) {
             // CraftBukkit start
