--- a/net/minecraft/world/item/component/Consumables.java
+++ b/net/minecraft/world/item/component/Consumables.java
@@ -65,6 +_,17 @@
     public static final Consumable MILK_BUCKET = defaultDrink().onConsume(ClearAllStatusEffectsConsumeEffect.INSTANCE).build();
     public static final Consumable CHORUS_FRUIT = defaultFood().onConsume(new TeleportRandomlyConsumeEffect()).build();
 
+    // KioCG start - more food
+    public static final Consumable POISONOUS_MUSHROOM = defaultFood()
+            .onConsume(new ApplyStatusEffectsConsumeEffect(new MobEffectInstance(MobEffects.POISON, 100, 0), 0.6F))
+            .build();
+    public static final Consumable SLIME_BALL = defaultFood()
+            .consumeSeconds(3.2F)
+            .sound(net.minecraft.core.Holder.direct(SoundEvents.HONEY_BLOCK_SLIDE))
+            .onConsume(new ApplyStatusEffectsConsumeEffect(new MobEffectInstance(MobEffects.SLOWNESS, 200, 1)))
+            .build();
+    // KioCG end - more food
+
     public static Consumable.Builder defaultFood() {
         return Consumable.builder().consumeSeconds(1.6F).animation(ItemUseAnimation.EAT).sound(SoundEvents.GENERIC_EAT).hasConsumeParticles(true);
     }
