--- a/net/minecraft/world/item/component/UseRemainder.java
+++ b/net/minecraft/world/item/component/UseRemainder.java
@@ -27,6 +_,21 @@
         }
     }
 
+    // KioCG start - food freshness
+    @org.jetbrains.annotations.Nullable
+    public static ItemStack mayRotInto(ItemStack stack) {
+        if (stack.isBroken()) {
+            UseRemainder rotInto = stack.get(net.minecraft.core.component.DataComponents.ROT_INTO);
+            if (rotInto != null) {
+                return rotInto.convertInto.copyWithCount(stack.getCount());
+            } else {
+                return new ItemStack(net.minecraft.world.item.Items.ROTTEN_FLESH, stack.getCount());
+            }
+        }
+        return null;
+    }
+    // KioCG end - food freshness
+
     @Override
     public boolean equals(Object other) {
         if (this == other) {
