--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -59,11 +_,11 @@
     public PathNavigation(Mob mob, Level level) {
         this.mob = mob;
         this.level = level;
-        this.pathFinder = this.createPathFinder(Mth.floor(mob.getAttributeBaseValue(Attributes.FOLLOW_RANGE) * 16.0));
+        this.pathFinder = this.createPathFinder(Mth.floor(mob.getAttributeBaseValue(Attributes.FOLLOW_RANGE) * 4.0)); // KioCG - optimize entity behavior
     }
 
     public void updatePathfinderMaxVisitedNodes() {
-        int floor = Mth.floor(this.getMaxPathLength() * 16.0F);
+        int floor = Mth.floor(this.getMaxPathLength() * 4.0F); // KioCG - optimize entity behavior
         this.pathFinder.setMaxVisitedNodes(floor);
     }
 
@@ -123,6 +_,23 @@
         return this.createPath(positions, 8, false, distance);
     }
 
+    // KioCG start - optimize entity behavior
+    @Nullable
+    public Path createPathLightweight(BlockPos pos, int accuracy) {
+        BlockPos blockPos = this.mob.blockPosition();
+        int i = Math.abs(blockPos.getX() - pos.getX());
+        int i1 = Math.abs(blockPos.getY() - pos.getY());
+        int i2 = Math.abs(blockPos.getZ() - pos.getZ());
+        int distManhattan = i + i1 + i2;
+        int distChessboard = Math.max(Math.max(i, i1), i2);
+        int maxVisitedNodes = this.pathFinder.maxVisitedNodes;
+        this.pathFinder.maxVisitedNodes = distManhattan + 4;
+        Path path = this.createPath(ImmutableSet.of(pos), null, 2, false, accuracy, distChessboard);
+        this.pathFinder.maxVisitedNodes = maxVisitedNodes;
+        return path;
+    }
+    // KioCG end - optimize entity behavior
+
     @Nullable
     public Path createPath(BlockPos pos, int accuracy) {
         // Paper start - EntityPathfindEvent
@@ -200,6 +_,7 @@
                 this.targetPos = path.getTarget();
                 this.reachRange = accuracy;
                 this.resetStuckTimeout();
+                this.prevEntity = new java.lang.ref.WeakReference<>(target); // KioCG - optimize entity behavior
             }
 
             return path;
@@ -212,20 +_,87 @@
     // Paper end - Perf: Optimise pathfinding
 
     public boolean moveTo(double x, double y, double z, double speed) {
-        return this.moveTo(this.createPath(x, y, z, 1), speed);
-    }
+        return this.moveTo(x, y, z, 1, speed); // KioCG - optimize entity behavior
+    }
+
+    // KioCG start - optimize entity behavior
+    public boolean moveToLightweight(double x, double y, double z, double speed) {
+        BlockPos blockPos = this.mob.blockPosition();
+        int i = Math.abs(blockPos.getX() - Mth.floor(x));
+        int i1 = Math.abs(blockPos.getY() - Mth.floor(y));
+        int i2 = Math.abs(blockPos.getZ() - Mth.floor(z));
+        int distManhattan = i + i1 + i2;
+        int distChessboard = Math.max(Math.max(i, i1), i2);
+        int maxVisitedNodes = this.pathFinder.maxVisitedNodes;
+        this.pathFinder.maxVisitedNodes = distManhattan + 4;
+        boolean moveTo = this.moveTo(x, y, z, 1, speed, 2, distChessboard);
+        this.pathFinder.maxVisitedNodes = maxVisitedNodes;
+        return moveTo;
+    }
+    // KioCG end - optimize entity behavior
 
     public boolean moveTo(double x, double y, double z, int accuracy, double speed) {
-        return this.moveTo(this.createPath(x, y, z, accuracy), speed);
-    }
-
+        // KioCG start - optimize entity behavior
+        return this.moveTo(x, y, z, accuracy, speed, 8, this.getMaxPathLength());
+    }
+    public boolean moveTo(double x, double y, double z, int accuracy, double speed, int regionOffset, float followRange) {
+        if (this.pathfindFailures > 10 && this.path == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
+            return false;
+        }
+
+        Set<BlockPos> blockPos = ImmutableSet.of(BlockPos.containing(x, y, z));
+        Path path = this.createPath(blockPos, null, regionOffset, false, accuracy, followRange);
+        if (this.moveTo(path, speed)) {
+            this.lastFailure = 0;
+            this.pathfindFailures = 0;
+            return true;
+        } else {
+            this.pathfindFailures++;
+            this.lastFailure = net.minecraft.server.MinecraftServer.currentTick;
+            return false;
+        }
+        // KioCG end - optimize entity behavior
+    }
+
+    // KioCG start - optimize entity behavior
+    public boolean moveToLightweight(Entity entity, double speed) {
+        BlockPos blockPos = this.mob.blockPosition();
+        BlockPos blockPos1 = entity.blockPosition();
+        int i = Math.abs(blockPos.getX() - blockPos1.getX());
+        int i1 = Math.abs(blockPos.getY() - blockPos1.getY());
+        int i2 = Math.abs(blockPos.getZ() - blockPos1.getZ());
+        int distManhattan = i + i1 + i2;
+        int distChessboard = Math.max(Math.max(i, i1), i2);
+        int maxVisitedNodes = this.pathFinder.maxVisitedNodes;
+        this.pathFinder.maxVisitedNodes = distManhattan + 4;
+        boolean moveTo = this.moveTo(entity, speed, 2, distChessboard);
+        this.pathFinder.maxVisitedNodes = maxVisitedNodes;
+        return moveTo;
+    }
+    // KioCG end - optimize entity behavior
+
+    // KioCG start - optimize entity behavior
+    private java.lang.ref.WeakReference<Entity> prevEntity = new java.lang.ref.WeakReference<>(null);
+    private int tickPause = 0;
+    // KioCG end - optimize entity behavior
     public boolean moveTo(Entity entity, double speed) {
+        // KioCG start - optimize entity behavior
+        return this.moveTo(entity, speed, 16, this.getMaxPathLength());
+    }
+    public boolean moveTo(Entity entity, double speed, int regionOffset, float followRange) {
+        if (prevEntity.refersTo(entity) && entity.tickCount <= this.tickPause) {
+            return true;
+        }
+        this.tickPause = entity.tickCount + (this.mob.numCollisions > 0 ? (8 + this.mob.numCollisions * 2) : 4);
         // Paper start - Perf: Optimise pathfinding
         if (this.pathfindFailures > 10 && this.path == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
             return false;
         }
         // Paper end - Perf: Optimise pathfinding
-        Path path = this.createPath(entity, 1);
+        // KioCG start - optimize entity behavior
+        Set<BlockPos> blockPos = ImmutableSet.of(entity.blockPosition());
+        Path path = this.createPath(blockPos, entity, regionOffset, true, 1, followRange);
+        // KioCG end - optimize entity behavior
         // Paper start - Perf: Optimise pathfinding
         if (path != null && this.moveTo(path, speed)) {
             this.lastFailure = 0;
@@ -409,7 +_,7 @@
     protected abstract boolean canUpdatePath();
 
     protected void trimPath() {
-        if (this.path != null) {
+        if (false && this.path != null) { // KioCG - optimize entity behavior: 这段代码意义不明
             for (int i = 0; i < this.path.getNodeCount(); i++) {
                 Node node = this.path.getNode(i);
                 Node node1 = i + 1 < this.path.getNodeCount() ? this.path.getNode(i + 1) : null;
