--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1094,6 +_,7 @@
         this.safeShutdown(waitForServer, false);
     }
     public void safeShutdown(boolean waitForServer, boolean isRestarting) {
+        com.kiocg.task.BossBarTask.stopAll(); // KioCG - bossbar task
         this.isRestarting = isRestarting;
         this.hasLoggedStop = true; // Paper - Debugging
         if (isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
@@ -1245,7 +_,7 @@
                 lastTick = currentTime;
                 this.nextTickTimeNanos += l;
 
-                try (Profiler.Scope scope = Profiler.use(this.createProfiler())) {
+                try { // KioCG - Remove Mojang Profiler
                     ProfilerFiller profilerFiller = Profiler.get();
                     profilerFiller.push("tick");
                     this.tickFrame.start();
@@ -1271,7 +_,7 @@
                     profilerFiller.pop();
                     this.logFullTickTime();
                 } finally {
-                    this.endMetricsRecordingTick();
+                    // KioCG - Remove Mojang Profiler
                 }
 
                 this.isReady = true;
@@ -1523,6 +_,17 @@
             }
         }
 
+        // KioCG start - chunk hot
+        if (this.tickCount % 20 == 0) {
+            for (ServerLevel serverLevel : this.getAllLevels()) {
+                for (net.minecraft.world.level.chunk.LevelChunk chunk : serverLevel.moonrise$getTickingChunks()) {
+                    chunk.getChunkHot().newTick();
+                }
+            }
+            com.kiocg.world.ChunkHot.nextTick();
+        }
+        // KioCG end - chunk hot
+
         this.server.spark.tickStart(); // Paper - spark
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
         this.tickCount++;
@@ -2235,9 +_,11 @@
                     if (Thread.currentThread() != this.serverThread) return; // Paper
                     // Paper start - we don't need to save everything, just advancements
                     // this.getPlayerList().saveAll();
-                    for (final ServerPlayer player : this.getPlayerList().getPlayers()) {
-                        player.getAdvancements().save();
-                    }
+                    // KioCG start - fix advancements save: 在PlayerAdvancements#reload()里保存
+                    // for (final ServerPlayer player : this.getPlayerList().getPlayers()) {
+                    //     player.getAdvancements().save();
+                    // }
+                    // KioCG end - fix advancements save: 在PlayerAdvancements#reload()里保存
                     // Paper end - we don't need to save everything, just advancements
                     this.getPlayerList().reloadResources();
                     this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
