--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -53,6 +_,7 @@
     private AdvancementTree tree;
     private final Map<AdvancementHolder, AdvancementProgress> progress = new LinkedHashMap<>();
     private final Set<AdvancementHolder> visible = new HashSet<>();
+    private final com.google.common.collect.Multimap<AdvancementHolder, String> registered = com.google.common.collect.HashMultimap.create(); // KioCG - linear achievement system
     private final Set<AdvancementHolder> progressChanged = new HashSet<>();
     private final Set<AdvancementNode> rootsToUpdate = new HashSet<>();
     private ServerPlayer player;
@@ -86,6 +_,7 @@
         this.stopListening();
         this.progress.clear();
         this.visible.clear();
+        this.registered.clear(); // KioCG - linear achievement system
         this.rootsToUpdate.clear();
         this.progressChanged.clear();
         this.isFirstPacket = true;
@@ -96,10 +_,33 @@
 
     private void registerListeners(ServerAdvancementManager manager) {
         for (AdvancementHolder advancementHolder : manager.getAllAdvancements()) {
+            if (this.needRegisterListeners(advancementHolder)) // KioCG - linear achievement system
             this.registerListeners(advancementHolder);
         }
     }
 
+    // KioCG start - linear achievement system
+    private boolean needRegisterListeners(AdvancementHolder advancement) {
+        if (advancement.value().display().isEmpty()) {
+            return true;
+        } else if (advancement.value().isRoot()) {
+            return true;
+        } else if (this.visible.contains(advancement)) {
+            return true;
+        } else {
+            String path = advancement.id().getPath();
+            for (String registerPath : io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.advancementRegisterPath) {
+                if (registerPath.endsWith("/")) {
+                    return path.startsWith(registerPath);
+                } else {
+                    return path.equals(registerPath);
+                }
+            }
+        }
+        return false;
+    }
+    // KioCG end - linear achievement system
+
     private void checkForAutomaticTriggers(ServerAdvancementManager manager) {
         for (AdvancementHolder advancementHolder : manager.getAllAdvancements()) {
             Advancement advancement = advancementHolder.value();
@@ -124,7 +_,7 @@
         }
 
         this.checkForAutomaticTriggers(manager);
-        this.registerListeners(manager);
+        // this.registerListeners(manager); // KioCG - linear achievement system: 推迟注册
     }
 
     public void save() {
@@ -213,7 +_,7 @@
         AdvancementProgress orStartProgress = this.getOrStartProgress(advancement);
         boolean isDone = orStartProgress.isDone();
         if (orStartProgress.revokeProgress(criterionKey)) {
-            this.registerListeners(advancement);
+            // this.registerListeners(advancement); // KioCG - linear achievement system: 推迟注册
             this.progressChanged.add(advancement);
             flag = true;
         }
@@ -238,6 +_,7 @@
             for (Entry<String, Criterion<?>> entry : advancement.value().criteria().entrySet()) {
                 CriterionProgress criterion = orStartProgress.getCriterion(entry.getKey());
                 if (criterion != null && !criterion.isDone()) {
+                    if (!this.registered.containsEntry(advancement, entry.getKey())) // KioCG - linear achievement system
                     this.registerListener(advancement, entry.getKey(), entry.getValue());
                 }
             }
@@ -245,6 +_,7 @@
     }
 
     private <T extends CriterionTriggerInstance> void registerListener(AdvancementHolder advancement, String criterionKey, Criterion<T> criterion) {
+        this.registered.put(advancement, criterionKey); // KioCG - linear achievement system
         criterion.trigger().addPlayerListener(this, new CriterionTrigger.Listener<>(criterion.triggerInstance(), advancement, criterionKey));
     }
 
@@ -260,6 +_,7 @@
     }
 
     private <T extends CriterionTriggerInstance> void removeListener(AdvancementHolder advancement, String criterionKey, Criterion<T> criterion) {
+        this.registered.remove(advancement, criterionKey); // KioCG - linear achievement system
         criterion.trigger().removePlayerListener(this, new CriterionTrigger.Listener<>(criterion.triggerInstance(), advancement, criterionKey));
     }
 
@@ -287,6 +_,7 @@
             }
         }
 
+        if (this.isFirstPacket) this.registerListeners(MinecraftServer.getServer().getAdvancements()); // KioCG - linear achievement system
         this.isFirstPacket = false;
     }
 
@@ -324,12 +_,14 @@
                 AdvancementHolder advancementHolder = node.holder();
                 if (visible) {
                     if (this.visible.add(advancementHolder)) {
+                        this.registerListeners(advancementHolder); // KioCG - linear achievement system
                         advancementOutput.add(advancementHolder);
                         if (this.progress.containsKey(advancementHolder)) {
                             this.progressChanged.add(advancementHolder);
                         }
                     }
                 } else if (this.visible.remove(advancementHolder)) {
+                    if (!this.needRegisterListeners(advancementHolder)) this.unregisterListeners(advancementHolder); // KioCG - linear achievement system
                     idOutput.add(advancementHolder.id());
                 }
             }
