--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -125,6 +_,7 @@
     public int numberOfRestocksToday = 0;
     private long lastRestockCheckDayTime;
     private boolean assignProfessionWhenSpawned = false;
+    private com.kiocg.entity.Lobotomized lobotomized = new com.kiocg.entity.Lobotomized(); public boolean isLobotomized() { return lobotomized.isLobotomized; } // KioCG - lobotomized villager
     private static final ImmutableList<MemoryModuleType<?>> MEMORY_TYPES = ImmutableList.of(
         MemoryModuleType.HOME,
         MemoryModuleType.JOB_SITE,
@@ -160,12 +_,12 @@
     private static final ImmutableList<SensorType<? extends Sensor<? super Villager>>> SENSOR_TYPES = ImmutableList.of(
         SensorType.NEAREST_LIVING_ENTITIES,
         SensorType.NEAREST_PLAYERS,
-        SensorType.NEAREST_ITEMS,
-        SensorType.NEAREST_BED,
+        // SensorType.NEAREST_ITEMS, // KioCG - optimize villager
+        // SensorType.NEAREST_BED, // KioCG - optimize villager
         SensorType.HURT_BY,
         SensorType.VILLAGER_HOSTILES,
         SensorType.VILLAGER_BABIES,
-        SensorType.SECONDARY_POIS,
+        // SensorType.SECONDARY_POIS, // KioCG - optimize villager
         SensorType.GOLEM_DETECTED
     );
     public static final Map<MemoryModuleType<GlobalPos>, BiPredicate<Villager, Holder<PoiType>>> POI_MEMORIES = ImmutableMap.of(
@@ -227,11 +_,13 @@
             villagerBrain.addActivity(Activity.PLAY, VillagerGoalPackages.getPlayPackage(0.5F));
         } else {
             villagerBrain.setSchedule(Schedule.VILLAGER_DEFAULT);
-            villagerBrain.addActivityWithConditions(
-                Activity.WORK,
-                VillagerGoalPackages.getWorkPackage(holder, 0.5F),
-                ImmutableSet.of(Pair.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT))
-            );
+            // KioCG start - optimize villager
+            // villagerBrain.addActivityWithConditions(
+            //     Activity.WORK,
+            //     VillagerGoalPackages.getWorkPackage(holder, 0.5F),
+            //     ImmutableSet.of(Pair.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT))
+            // );
+            // KioCG end - optimize villager
         }
 
         villagerBrain.addActivity(Activity.CORE, VillagerGoalPackages.getCorePackage(holder, 0.5F));
@@ -292,22 +_,32 @@
         // Paper start - EAR 2
         this.customServerAiStep(level, false);
     }
-    protected void customServerAiStep(ServerLevel level, final boolean inactive) {
+    protected void customServerAiStep(ServerLevel level, boolean inactive) { // KioCG - lobotomized villager: not final
         // Paper end - EAR 2
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("villagerBrain");
+        // KioCG start - lobotomized villager
+        if (this.level().isBrightOutside() && (this.tickCount + this.getId()) % 100 == 0 && this.shouldRestock()) {
+            this.playWorkSound();
+            this.restock();
+        }
+        if (this.lobotomized.checkLobotomized(this)) {
+            inactive = true; // treat as inactive if lobotomized
+        }
+        // KioCG end - lobotomized villager
         if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
         profilerFiller.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
         }
 
-        if (!this.isTrading() && this.updateMerchantTimer > 0) {
+        if (this.updateMerchantTimer > 0) { // KioCG - villager random profession: 交易过程中允许升级
             this.updateMerchantTimer--;
             if (this.updateMerchantTimer <= 0) {
                 if (this.increaseProfessionLevelOnUpdate) {
                     this.increaseMerchantCareer();
                     this.increaseProfessionLevelOnUpdate = false;
+                    this.playWorkSound(); // KioCG - villager random profession
                 }
 
                 this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
@@ -327,9 +_,21 @@
             }
         }
 
-        if (this.getVillagerData().profession().is(VillagerProfession.NONE) && this.isTrading()) {
-            this.stopTrading();
+        // KioCG start - villager random profession
+        if (this.getVillagerData().profession().is(VillagerProfession.NONE)) {
+            if (this.isTrading()) {
+                this.stopTrading();
+            }
+
+            if (this.level().paperConfig().kiocgConfig.entity.animal.villagerRandomProfession && !this.isBaby()) {
+                List<ResourceKey<VillagerProfession>> professions = new java.util.ArrayList<>(BuiltInRegistries.VILLAGER_PROFESSION.registryKeySet());
+                professions.remove(VillagerProfession.NONE);
+                professions.remove(VillagerProfession.NITWIT);
+                ResourceKey<VillagerProfession> profession = professions.get(this.random.nextInt(professions.size()));
+                this.setVillagerData(this.getVillagerData().withProfession(BuiltInRegistries.VILLAGER_PROFESSION.getOrThrow(profession)));
+            }
         }
+        // KioCG end - villager random profession
 
         if (inactive) return; // Paper - EAR 2
         super.customServerAiStep(level);
@@ -383,6 +_,7 @@
     }
 
     private void startTrading(Player player) {
+        if (this.lobotomized.isLobotomized) this.lookAt(net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.FEET, player.position()); // KioCG - lobotomized villager
         this.updateSpecialPrices(player);
         this.setTradingPlayer(player);
         this.openTradingScreen(player, this.getDisplayName(), this.getVillagerData().level());
@@ -465,21 +_,22 @@
     public boolean shouldRestock() {
         long l = this.lastRestockGameTime + 12000L;
         long gameTime = this.level().getGameTime();
-        boolean flag = gameTime > l;
+        boolean flag = gameTime > l && false; // KioCG - villager restock every day: 去除半天刷新一次补货? 意义不明
         long dayTime = this.level().getDayTime();
         if (this.lastRestockCheckDayTime > 0L) {
             long l1 = this.lastRestockCheckDayTime / 24000L;
             long l2 = dayTime / 24000L;
+            // KioCG - 注释: 天数不同则补货
             flag |= l2 > l1;
         }
 
         this.lastRestockCheckDayTime = dayTime;
         if (flag) {
             this.lastRestockGameTime = gameTime;
-            this.resetNumberOfRestocks();
+            this.numberOfRestocksToday = 1; // KioCG - villager restock every day: 补货后会变成2, 用于兼容插件
         }
 
-        return this.allowedToRestock() && this.needsToRestock();
+        return flag; // KioCG - villager restock every day
     }
 
     private void catchUpDemand() {
@@ -555,6 +_,7 @@
         if (this.assignProfessionWhenSpawned) {
             compound.putBoolean("AssignProfessionWhenSpawned", true);
         }
+        compound.putBoolean("KioCG.Lobotomized", this.lobotomized.isLobotomized); // KioCG - lobotomized villager
     }
 
     @Override
@@ -573,6 +_,7 @@
 
         this.numberOfRestocksToday = compound.getIntOr("RestocksToday", 0);
         this.assignProfessionWhenSpawned = compound.getBooleanOr("AssignProfessionWhenSpawned", false);
+        this.lobotomized.isLobotomized = compound.getBooleanOr("KioCG.Lobotomized", false); // KioCG - lobotomized villager
     }
 
     @Override
@@ -697,6 +_,7 @@
 
     @Override
     public boolean canBreed() {
+        if (!this.level().paperConfig().kiocgConfig.entity.animal.villagerCanBreed) return false; // KioCG - breeding animals
         return this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0;
     }
 
@@ -891,6 +_,7 @@
                 if (itemListings != null) {
                     MerchantOffers offers = this.getOffers();
                     this.addOffersFromItemListings(offers, itemListings, amount); // Paper - More vanilla friendly methods to update trades
+                    this.resendOffersToTradingPlayer(); // KioCG - villager random profession: 交易过程中允许升级
                     return true; // Paper - More vanilla friendly methods to update trades
                 }
             }
