From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KioProject <1105919949@qq.com>
Date: Sun, 6 Apr 2025 22:11:52 +0800
Subject: [PATCH] CauldronBlockEntity


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
index 56453454cbd4b9e9270fc833f8ab38d5fa7a3763..75c6af7b9f1fdd30c47274210a6402eb5b55712f 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
@@ -180,6 +180,7 @@ public final class CraftBlockStates {
         register(BlockEntityType.TRIAL_SPAWNER, CraftTrialSpawner.class, CraftTrialSpawner::new);
         register(BlockEntityType.VAULT, CraftVault.class, CraftVault::new);
         // Paper end
+        register(BlockEntityType.CAULDRON, CraftCauldron.class, CraftCauldron::new); // KioCG - CauldronBlockEntity: 确保最后一个注册
     }
 
     private static void register(Material blockType, BlockStateFactory<?> factory) {
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java b/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java
new file mode 100644
index 0000000000000000000000000000000000000000..eb917559a11891ba1d13e1b4ce5d4c4cdd199e08
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java
@@ -0,0 +1,34 @@
+package org.bukkit.craftbukkit.block;
+
+import net.minecraft.world.level.block.entity.CauldronBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.Cauldron;
+
+// KioCG start - CauldronBlockEntity: all class
+public class CraftCauldron extends CraftBlockEntityState<CauldronBlockEntity> implements Cauldron {
+
+    public CraftCauldron(World world, CauldronBlockEntity tileEntity) {
+        super(world, tileEntity);
+    }
+
+    protected CraftCauldron(CraftCauldron state, Location location) {
+        super(state, location);
+    }
+
+    @Override
+    public boolean isBoiling() {
+        return getSnapshot().isBoiling();
+    }
+
+    @Override
+    public CraftCauldron copy() {
+        return new CraftCauldron(this, null);
+    }
+
+    @Override
+    public CraftCauldron copy(Location location) {
+        return new CraftCauldron(this, location);
+    }
+}
+// KioCG end - CauldronBlockEntity: all class
