From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KioProject <1105919949@qq.com>
Date: Sun, 6 Apr 2025 22:11:52 +0800
Subject: [PATCH] Custom block entity


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
index 3c9131ac3cfb4f758e76830ca48813feb321e4c6..2e541a7fbc3012570dab40573c54d58d9246007f 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
@@ -134,6 +134,10 @@ public final class CraftBlockStates {
         register(BlockEntityType.TRIAL_SPAWNER, CraftTrialSpawner.class, CraftTrialSpawner::new);
         register(BlockEntityType.VAULT, CraftVault.class, CraftVault::new);
         // End generate - CraftBlockEntityStates
+        // KioCG start - custom block entity: 确保最后一个注册
+        register(BlockEntityType.CAULDRON, CraftCauldron.class, CraftCauldron::new);
+        register(BlockEntityType.FLOWER_POT, CraftFlowerPot.class, CraftFlowerPot::new);
+        // KioCG end - custom block entity: 确保最后一个注册
     }
 
     private static void register(Material blockType, BlockStateFactory<?> factory) {
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java b/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java
new file mode 100644
index 0000000000000000000000000000000000000000..4c237f3d5fe57c3751ce6fa26c594e31e3a64786
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCauldron.java
@@ -0,0 +1,29 @@
+package org.bukkit.craftbukkit.block;
+
+import net.minecraft.world.level.block.entity.CauldronBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.Cauldron;
+
+// KioCG start - custom block entity: all class
+public class CraftCauldron extends CraftBlockEntityState<CauldronBlockEntity> implements Cauldron {
+
+    public CraftCauldron(World world, CauldronBlockEntity tileEntity) {
+        super(world, tileEntity);
+    }
+
+    protected CraftCauldron(CraftCauldron state, Location location) {
+        super(state, location);
+    }
+
+    @Override
+    public CraftCauldron copy() {
+        return new CraftCauldron(this, null);
+    }
+
+    @Override
+    public CraftCauldron copy(Location location) {
+        return new CraftCauldron(this, location);
+    }
+}
+// KioCG end - custom block entity: all class
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftFlowerPot.java b/src/main/java/org/bukkit/craftbukkit/block/CraftFlowerPot.java
new file mode 100644
index 0000000000000000000000000000000000000000..7ccff50e3b7bf5f696fbda48268c01ec19ffc51c
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftFlowerPot.java
@@ -0,0 +1,29 @@
+package org.bukkit.craftbukkit.block;
+
+import net.minecraft.world.level.block.entity.FlowerPotBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.FlowerPot;
+
+// KioCG start - custom block entity: all class
+public class CraftFlowerPot extends CraftBlockEntityState<FlowerPotBlockEntity> implements FlowerPot {
+
+    public CraftFlowerPot(World world, FlowerPotBlockEntity tileEntity) {
+        super(world, tileEntity);
+    }
+
+    protected CraftFlowerPot(CraftFlowerPot state, Location location) {
+        super(state, location);
+    }
+
+    @Override
+    public CraftFlowerPot copy() {
+        return new CraftFlowerPot(this, null);
+    }
+
+    @Override
+    public CraftFlowerPot copy(Location location) {
+        return new CraftFlowerPot(this, location);
+    }
+}
+// KioCG end - custom block entity: all class
