--- a/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -23,6 +_,7 @@
     private static final boolean DEFAULT_ENABLED = true;
     private boolean enabled = true;
     private boolean consumedItemThisFrame = false;
+    @javax.annotation.Nullable public java.util.Set<net.minecraft.world.item.Item> itemFilter; // KioCG - hopper filter
 
     public MinecartHopper(EntityType<? extends MinecartHopper> entityType, Level level) {
         super(entityType, level);
@@ -60,6 +_,14 @@
         this.enabled = enabled;
     }
 
+    // KioCG start - optimize block entity
+    private static final net.minecraft.world.phys.AABB SUCK_AABB = net.minecraft.world.level.block.Block.column(16.0, 0.0, 32.0).toAabbs().get(0);
+    @Override
+    public net.minecraft.world.phys.AABB getSuckAabb() {
+        return SUCK_AABB;
+    }
+    // KioCG end - optimize block entity
+
     @Override
     public double getLevelX() {
         return this.getX();
@@ -101,11 +_,21 @@
         }
     }
 
+    // KioCG start - hopper filter
+    @Override
+    public void setCustomName(@javax.annotation.Nullable net.minecraft.network.chat.Component name) {
+        super.setCustomName(name);
+        this.itemFilter = this.hasCustomName() ? HopperBlockEntity.parseItemFilter(org.bukkit.craftbukkit.util.CraftChatMessage.fromComponent(this.getCustomName())) : null;
+    }
+    // KioCG end - hopper filter
+
     public boolean suckInItems() {
+        if (this.inventoryFull()) return false; // KioCG - optimize block entity
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             this.immunize(); // Paper
             return true;
         } else {
+            if (true) return false; // KioCG start - optimize block entity: 在SuckAabb中搜索周围物品
             for (ItemEntity itemEntity : this.level()
                 .getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(0.25, 0.0, 0.25), EntitySelector.ENTITY_STILL_ALIVE)) {
                 if (HopperBlockEntity.addItem(this, itemEntity)) {
@@ -117,6 +_,18 @@
             return false;
         }
     }
+
+    // KioCG start - optimize block entity
+    private boolean inventoryFull() {
+        for (ItemStack itemStack : this.getItemStacks()) {
+            if (itemStack.isEmpty() || itemStack.getCount() < itemStack.getMaxStackSize()) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+    // KioCG end - optimize block entity
 
     @Override
     public Item getDropItem() {
