--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -148,14 +_,16 @@
             }
 
             final int worldHeight = world.getHeight();
-            ActivationRange.maxBB = player.getBoundingBox().inflate(maxRange, worldHeight, maxRange);
-            ActivationType.MISC.boundingBox = player.getBoundingBox().inflate(miscActivationRange, worldHeight, miscActivationRange);
-            ActivationType.RAIDER.boundingBox = player.getBoundingBox().inflate(raiderActivationRange, worldHeight, raiderActivationRange);
-            ActivationType.ANIMAL.boundingBox = player.getBoundingBox().inflate(animalActivationRange, worldHeight, animalActivationRange);
-            ActivationType.MONSTER.boundingBox = player.getBoundingBox().inflate(monsterActivationRange, worldHeight, monsterActivationRange);
-            ActivationType.WATER.boundingBox = player.getBoundingBox().inflate(waterActivationRange, worldHeight, waterActivationRange);
-            ActivationType.FLYING_MONSTER.boundingBox = player.getBoundingBox().inflate(flyingActivationRange, worldHeight, flyingActivationRange);
-            ActivationType.VILLAGER.boundingBox = player.getBoundingBox().inflate(villagerActivationRange, worldHeight, villagerActivationRange);
+            // KioCG start - optimize entity activation range
+            ActivationRange.maxBB = getActivationRange(player, maxRange, flyingActivationRange);
+            ActivationType.MISC.boundingBox = getActivationRange(player, miscActivationRange, 0);
+            ActivationType.RAIDER.boundingBox = getActivationRange(player, raiderActivationRange, 0);
+            ActivationType.ANIMAL.boundingBox = getActivationRange(player, animalActivationRange, 0);
+            ActivationType.MONSTER.boundingBox = getActivationRange(player, monsterActivationRange, 0);
+            ActivationType.WATER.boundingBox = getActivationRange(player, waterActivationRange, 0);
+            ActivationType.FLYING_MONSTER.boundingBox = getActivationRange(player, flyingActivationRange, flyingActivationRange);
+            ActivationType.VILLAGER.boundingBox = getActivationRange(player, villagerActivationRange, 0);
+            // KioCG end - optimize entity activation range
 
             final List<Entity> entities = world.getEntities((Entity) null, ActivationRange.maxBB, e -> true);
             final boolean tickMarkers = world.paperConfig().entities.markers.tick;
@@ -169,6 +_,32 @@
         }
     }
 
+    // KioCG start - optimize entity activation range
+    public static AABB getActivationRange(Player player, double xzRange, int extendSky) {
+        Level world = player.level();
+        if (!world.dimensionType().hasCeiling()) {
+            int oceanFloor = world.getHeight(net.minecraft.world.level.levelgen.Heightmap.Types.OCEAN_FLOOR, player.blockPosition());
+            if (player.getY() >= oceanFloor) {
+                return inflate(player.getBoundingBox(), xzRange, oceanFloor - 16, player.getY() + 16 + extendSky);
+            } else {
+                return inflate(player.getBoundingBox(), xzRange, player.getY() - 16, player.getY() + 16);
+            }
+        } else {
+            return player.getBoundingBox().inflate(xzRange, 48, xzRange);
+        }
+    }
+
+    public static AABB inflate(AABB aabb, double xz, double minY, double maxY) {
+        double d = aabb.minX - xz;
+        double d1 = minY;
+        double d2 = aabb.minZ - xz;
+        double d3 = aabb.maxX + xz;
+        double d4 = maxY;
+        double d5 = aabb.maxZ + xz;
+        return new AABB(d, d1, d2, d3, d4, d5);
+    }
+    // KioCG end - optimize entity activation range
+
     /**
      * Tries to activate an entity.
      *
@@ -209,11 +_,17 @@
         if ((entity.activationType != ActivationType.WATER && entity.isInWater() && entity.isPushedByFluid())) {
             return 100;
         }
+        if (entity instanceof ItemEntity) // KioCG - Check items only. https://github.com/PaperMC/Paper/pull/7644
         if (!entity.onGround() || entity.getDeltaMovement().horizontalDistanceSqr() > 9.999999747378752E-6D) {
             return 100;
         }
         if (!(entity instanceof final AbstractArrow arrow)) {
             if ((!entity.onGround() && !isEntityThatFlies(entity))) {
+                // KioCG start - radical optimization
+                if (!(entity instanceof net.minecraft.world.entity.animal.WaterAnimal) && !(entity instanceof net.minecraft.world.entity.animal.axolotl.Axolotl) && !(entity instanceof net.minecraft.world.entity.animal.Turtle)
+                    && !(entity instanceof net.minecraft.world.entity.animal.FlyingAnimal) && !(entity instanceof net.minecraft.world.entity.animal.allay.Allay)
+                    && !(entity instanceof net.minecraft.world.entity.ambient.Bat))
+                // KioCG end - radical optimization
                 return 10;
             }
         } else if (!arrow.isInGround()) {
@@ -257,7 +_,7 @@
                 return 1;
             }
             if (entity instanceof final Animal animal) {
-                if (animal.isBaby() || animal.isInLove()) {
+                if (animal.isInLove()) { // KioCG - breeding animals: 不再需要豁免baby
                     return 5;
                 }
                 if (entity instanceof final Sheep sheep && sheep.isSheared()) {
