--- a/src/main/java/org/bukkit/craftbukkit/inventory/RecipeIterator.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/RecipeIterator.java
@@ -7,13 +_,37 @@
 import net.minecraft.world.item.crafting.RecipeType;
 import org.bukkit.inventory.Recipe;
 
-public class RecipeIterator implements Iterator<Recipe> {
-    private final Iterator<Map.Entry<RecipeType<?>, RecipeHolder<?>>> recipes;
+public class RecipeIterator<T extends Recipe> implements Iterator<T> { // KioCG - extended recipe iterator
+    private final Iterator<RecipeHolder<?>> recipes; // KioCG - extended recipe iterator
     private RecipeHolder<?> currentRecipe;
 
     public RecipeIterator() {
-        this.recipes = MinecraftServer.getServer().getRecipeManager().recipes.byType.entries().iterator();
-    }
+        this.recipes = MinecraftServer.getServer().getRecipeManager().recipes.byType.values().iterator(); // KioCG - extended recipe iterator
+    }
+
+    // KioCG start - extended recipe iterator
+    public RecipeIterator(org.bukkit.inventory.RecipeType<T> recipeType) {
+        RecipeType<?> vanilla;
+        if (recipeType == org.bukkit.inventory.RecipeType.CRAFTING) {
+            vanilla = RecipeType.CRAFTING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.SMELTING) {
+            vanilla = RecipeType.SMELTING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.BLASTING) {
+            vanilla = RecipeType.BLASTING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.SMOKING) {
+            vanilla = RecipeType.SMOKING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.CAMPFIRE_COOKING) {
+            vanilla = RecipeType.CAMPFIRE_COOKING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.STONECUTTING) {
+            vanilla = RecipeType.STONECUTTING;
+        } else if (recipeType == org.bukkit.inventory.RecipeType.SMITHING) {
+            vanilla = RecipeType.SMITHING;
+        } else {
+            throw new IllegalStateException("unknown recipe type");
+        }
+        this.recipes = MinecraftServer.getServer().getRecipeManager().recipes.byType.get(vanilla).iterator();
+    }
+    // KioCG end - extended recipe iterator
 
     @Override
     public boolean hasNext() {
@@ -21,16 +_,18 @@
     }
 
     @Override
-    public Recipe next() {
-        this.currentRecipe = this.recipes.next().getValue();
-        return this.currentRecipe.toBukkitRecipe();
+    // KioCG start - extended recipe iterator
+    public T next() {
+        this.currentRecipe = this.recipes.next();
+        return (T) this.currentRecipe.toBukkitRecipe();
     }
+    // KioCG end - extended recipe iterator
 
     @Override
     public void remove() {
         MinecraftServer.getServer().getRecipeManager().recipes.byKey.remove(this.currentRecipe.id());
         this.recipes.remove();
         MinecraftServer.getServer().getRecipeManager().finalizeRecipeLoading();
-        MinecraftServer.getServer().getPlayerList().reloadRecipes();
+        // MinecraftServer.getServer().getPlayerList().reloadRecipes(); // KioCG - more api: 数据包已在内部重新发送
     }
 }
