--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -472,6 +_,36 @@
         return event;
     }
 
+    // KioCG start - PlayerBottleEvent
+    public static com.kiocg.event.player.PlayerBottleEmptyEvent callPlayerBottleEmptyEvent(Level world, net.minecraft.world.entity.player.Player player, BlockPos clicked, ItemStack itemInHand, ItemStack result, InteractionHand hand) {
+        return (com.kiocg.event.player.PlayerBottleEmptyEvent) CraftEventFactory.getPlayerBottleEvent(false, world, player, clicked, itemInHand, result, hand);
+    }
+
+    public static com.kiocg.event.player.PlayerBottleFillEvent callPlayerBottleFillEvent(Level world, net.minecraft.world.entity.player.Player player, BlockPos clicked, ItemStack itemInHand, ItemStack result, InteractionHand hand) {
+        return (com.kiocg.event.player.PlayerBottleFillEvent) CraftEventFactory.getPlayerBottleEvent(true, world, player, clicked, itemInHand, result, hand);
+    }
+
+    private static PlayerEvent getPlayerBottleEvent(boolean isFilling, Level world, net.minecraft.world.entity.player.Player player, BlockPos clicked, ItemStack bottle, ItemStack result, InteractionHand hand) {
+        Player cplayer = (Player) player.getBukkitEntity();
+        org.bukkit.inventory.ItemStack bottleItem = CraftItemStack.asBukkitCopy(bottle);
+        CraftItemStack resultItem = CraftItemStack.asCraftMirror(result);
+
+        Block clickedBlock = CraftBlock.at(world, clicked);
+        EquipmentSlot handSlot = CraftEquipmentSlot.getHand(hand);
+
+        PlayerEvent event;
+        if (isFilling) {
+            event = new com.kiocg.event.player.PlayerBottleFillEvent(cplayer, clickedBlock, bottleItem, resultItem, handSlot);
+        } else {
+            event = new com.kiocg.event.player.PlayerBottleEmptyEvent(cplayer, clickedBlock, bottleItem, resultItem, handSlot);
+        }
+
+        event.callEvent();
+
+        return event;
+    }
+    // KioCG end - PlayerBottleEvent
+
     public static PlayerInteractEvent callPlayerInteractEvent(net.minecraft.world.entity.player.Player player, Action action, ItemStack item, InteractionHand hand) {
         if (action != Action.LEFT_CLICK_AIR && action != Action.RIGHT_CLICK_AIR) {
             throw new AssertionError(String.format("%s performing %s with %s", player, action, item));
@@ -528,6 +_,41 @@
         return event;
     }
 
+    // KioCG start - PlayerItemUseEvent
+    public static com.kiocg.event.player.PlayerItemUseEvent callPlayerItemUseEvent(net.minecraft.world.entity.player.Player who, Action action, BlockPos position, Direction direction, ItemStack itemstack, InteractionHand hand, Vec3 targetPos, PlayerInteractEvent interactEvent) {
+        Player player = (who == null) ? null : (Player) who.getBukkitEntity();
+        CraftItemStack itemInHand = CraftItemStack.asCraftMirror(itemstack);
+
+        Vector clickedPos = null;
+        if (position != null && targetPos != null) {
+            clickedPos = CraftVector.toBukkit(targetPos.subtract(Vec3.atLowerCornerOf(position)));
+        }
+
+        CraftWorld craftWorld = (CraftWorld) player.getWorld();
+        CraftServer craftServer = (CraftServer) player.getServer();
+
+        Block blockClicked = null;
+        if (position != null) {
+            blockClicked = craftWorld.getBlockAt(position.getX(), position.getY(), position.getZ());
+        } else {
+            switch (action) {
+                case LEFT_CLICK_BLOCK:
+                    action = Action.LEFT_CLICK_AIR;
+                    break;
+                case RIGHT_CLICK_BLOCK:
+                    action = Action.RIGHT_CLICK_AIR;
+                    break;
+            }
+        }
+        BlockFace blockFace = CraftBlock.notchToBlockFace(direction);
+
+        com.kiocg.event.player.PlayerItemUseEvent event = new com.kiocg.event.player.PlayerItemUseEvent(player, action, itemInHand, blockClicked, blockFace, ((hand == InteractionHand.OFF_HAND) ? EquipmentSlot.OFF_HAND : EquipmentSlot.HAND), clickedPos, interactEvent);
+        craftServer.getPluginManager().callEvent(event);
+
+        return event;
+    }
+    // KioCG end - PlayerItemUseEvent
+
     public static EntityTransformEvent callEntityTransformEvent(net.minecraft.world.entity.LivingEntity original, net.minecraft.world.entity.LivingEntity coverted, EntityTransformEvent.TransformReason transformReason) {
         return CraftEventFactory.callEntityTransformEvent(original, Collections.singletonList(coverted), transformReason);
     }
@@ -890,10 +_,16 @@
     }
 
     public static PlayerDeathEvent callPlayerDeathEvent(ServerPlayer victim, DamageSource damageSource, List<Entity.DefaultDrop> drops, net.kyori.adventure.text.Component deathMessage, boolean showDeathMessages, boolean keepInventory) {
+        // KioCG start - death drop protect
+        return callPlayerDeathEvent(victim, damageSource, drops, List.of(), deathMessage, showDeathMessages, keepInventory);
+    }
+    public static PlayerDeathEvent callPlayerDeathEvent(ServerPlayer victim, DamageSource damageSource, List<Entity.DefaultDrop> drops, List<org.bukkit.inventory.ItemStack> keeps, net.kyori.adventure.text.Component deathMessage, boolean showDeathMessages, boolean keepInventory) {
+        // KioCG end - death drop protect
         CraftPlayer entity = victim.getBukkitEntity();
         CraftDamageSource bukkitDamageSource = new CraftDamageSource(damageSource);
         PlayerDeathEvent event = new PlayerDeathEvent(entity, bukkitDamageSource, new io.papermc.paper.util.TransformingRandomAccessList<>(drops, Entity.DefaultDrop::stack, FROM_FUNCTION), victim.getExpReward(victim.level(), damageSource.getEntity()), 0, deathMessage, showDeathMessages);
         event.setKeepInventory(keepInventory);
+        event.getItemsToKeep().addAll(keeps); // KioCG - death drop protect
         event.setKeepLevel(victim.keepLevel); // SPIGOT-2222: pre-set keepLevel
         populateFields(victim, event); // Paper - make cancellable
         Bukkit.getServer().getPluginManager().callEvent(event);
@@ -1295,6 +_,26 @@
 
         return com.mojang.datafixers.util.Pair.of(event.titleOverride(), container); // Paper - Add titleOverride to InventoryOpenEvent
     }
+
+    // KioCG start - InventoryExchangeItemEvent
+    public static boolean callInventoryExchangeItemEvent(InventoryView view, org.bukkit.inventory.Inventory sourceInventory, java.util.function.Supplier<org.bukkit.inventory.ItemStack> itemSupplier, org.bukkit.inventory.Inventory destinationInventory) {
+        if (sourceInventory == destinationInventory) {
+            return true;
+        }
+
+        org.bukkit.inventory.ItemStack itemStack = itemSupplier.get();
+        if (itemStack == null || itemStack.isEmpty()) {
+            return true;
+        }
+
+        com.kiocg.event.inventory.InventoryExchangeItemEvent event = new com.kiocg.event.inventory.InventoryExchangeItemEvent(view, sourceInventory, itemStack, destinationInventory);
+        // KioCG start - expand inventory holder
+        com.kiocg.inventory.LockedInventoryHolder.onInventoryExchangeItem(event);
+        com.kiocg.inventory.LootableInventoryHolder.onInventoryExchangeItem(event);
+        // KioCG end - expand inventory holder
+        return event.callEvent();
+    }
+    // KioCG end - InventoryExchangeItemEvent
 
     public static ItemStack callPreCraftEvent(CraftingContainer matrix, Container resultInventory, ItemStack result, InventoryView lastCraftView, boolean isRepair) {
         CraftInventoryCrafting inventory = new CraftInventoryCrafting(matrix, resultInventory);
