--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -182,6 +_,19 @@
 
     protected void onExplosionHit(BlockState state, ServerLevel level, BlockPos pos, Explosion explosion, BiConsumer<ItemStack, BlockPos> dropConsumer) {
         if (!state.isAir() && explosion.getBlockInteraction() != Explosion.BlockInteraction.TRIGGER_BLOCK && state.isDestroyable()) { // Paper - Protect Bedrock and End Portal/Frames from being destroyed
+            // KioCG start - creeper damage block
+            if (explosion.getDirectSourceEntity() instanceof net.minecraft.world.entity.monster.Creeper creeper) {
+                net.minecraft.server.level.BlockDestructionProgress blockDestructionProgress = level.destroyingBlocks.get(pos);
+                int progress = blockDestructionProgress.getProgress() + 5;
+                if (progress <= 10) {
+                    level.destroyBlockProgress(creeper.getId(), pos, progress);
+                    return;
+                }
+
+                level.destroyBlockProgress(creeper.getId(), pos, -1);
+            }
+            // KioCG end - creeper damage block
+
             Block block = state.getBlock();
             boolean flag = explosion.getIndirectSourceEntity() instanceof Player;
             if (block.dropFromExplosion(explosion)) {
@@ -313,6 +_,25 @@
         return true;
     }
 
+    // KioCG start - extended block placement
+    protected void playPlaceSound(BlockState state, Level level, BlockPos pos, Player player, @Nullable InteractionHand swingHand) {
+        boolean canBuild = level.checkEntityCollision(state, player, CollisionContext.of(player), pos, true); // Paper - Cancel hit for vanished players
+        if (canBuild) {
+            ((net.minecraft.server.level.ServerPlayer) player).connection.send(
+                    new net.minecraft.network.protocol.game.ClientboundSoundPacket(
+                            net.minecraft.core.registries.BuiltInRegistries.SOUND_EVENT.wrapAsHolder(soundType.getPlaceSound()),
+                            net.minecraft.sounds.SoundSource.BLOCKS,
+                            pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5,
+                            (soundType.getVolume() + 1.0F) / 2.0F,
+                            soundType.getPitch() * 0.8F,
+                            player.random.nextLong()));
+            if (swingHand != null) {
+                player.swing(swingHand, true);
+            }
+        }
+    }
+    // KioCG end - extended block placement
+
     protected float getShadeBrightness(BlockState state, BlockGetter level, BlockPos pos) {
         return state.isCollisionShapeFullBlock(level, pos) ? 0.2F : 1.0F;
     }
@@ -344,6 +_,12 @@
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
     }
 
+    // KioCG start - crops grow every day
+    protected void dayTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        this.randomTick(state, level, pos, random);
+    }
+    // KioCG end - crops grow every day
+
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
     }
 
@@ -370,6 +_,11 @@
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
     }
 
+    // KioCG start - optimize block entity: item search hopper
+    protected void itemSearchHopper(BlockState state, Level level, BlockPos pos, net.minecraft.world.entity.item.ItemEntity itemEntity, InsideBlockEffectApplier effectApplier) {
+    }
+    // KioCG end - optimize block entity: item search hopper
+
     protected int getDirectSignal(BlockState state, BlockGetter level, BlockPos pos, Direction direction) {
         return 0;
     }
@@ -393,6 +_,12 @@
         return this.isRandomlyTicking;
     }
 
+    // KioCG start - crops grow every day
+    protected boolean isDailyTicking(BlockState state) {
+        return false;
+    }
+    // KioCG end - crops grow every day
+
     protected SoundType getSoundType(BlockState state) {
         return this.soundType;
     }
@@ -446,6 +_,7 @@
         private BlockBehaviour.BlockStateBase.Cache cache;
         private FluidState fluidState = Fluids.EMPTY.defaultFluidState();
         private boolean isRandomlyTicking;
+        private boolean isDailyTicking; // KioCG - crops grow every day
         private boolean solidRender;
         private VoxelShape occlusionShape;
         private VoxelShape[] occlusionShapesByFace;
@@ -578,6 +_,7 @@
         public void initCache() {
             this.fluidState = this.owner.getFluidState(this.asState());
             this.isRandomlyTicking = this.owner.isRandomlyTicking(this.asState());
+            this.isDailyTicking = this.owner.isDailyTicking(this.asState()); // KioCG - crops grow every day
             if (!this.getBlock().hasDynamicShape()) {
                 this.cache = new BlockBehaviour.BlockStateBase.Cache(this.asState());
             }
@@ -883,10 +_,22 @@
             this.getBlock().randomTick(this.asState(), level, pos, random);
         }
 
+        // KioCG start - crops grow every day
+        public void dayTick(ServerLevel level, BlockPos pos, RandomSource random) {
+            this.getBlock().dayTick(this.asState(), level, pos, random);
+        }
+        // KioCG end - crops grow every day
+
         public void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
             this.getBlock().entityInside(this.asState(), level, pos, entity, effectApplier);
         }
 
+        // KioCG start - optimize block entity: item search hopper
+        public void itemSearchHopper(Level level, BlockPos pos, net.minecraft.world.entity.item.ItemEntity itemEntity, InsideBlockEffectApplier effectApplier) {
+            this.getBlock().itemSearchHopper(this.asState(), level, pos, itemEntity, effectApplier);
+        }
+        // KioCG end - optimize block entity: item search hopper
+
         public void spawnAfterBreak(ServerLevel level, BlockPos pos, ItemStack stack, boolean dropExperience) {
             this.getBlock().spawnAfterBreak(this.asState(), level, pos, stack, dropExperience);
             if (dropExperience) {this.getBlock().popExperience(level, pos, this.getBlock().getExpDrop(this.asState(), level, pos, stack, true));} // Paper - Properly handle xp dropping
@@ -897,6 +_,12 @@
         }
 
         public InteractionResult useItemOn(ItemStack stack, Level level, Player player, InteractionHand hand, BlockHitResult hitResult) {
+            // KioCG start - customize default components
+            if (stack.has(net.minecraft.core.component.DataComponents.CANNOT_USE)) {
+                return InteractionResult.TRY_WITH_EMPTY_HAND;
+            }
+            // KioCG end - customize default components
+
             return this.getBlock().useItemOn(stack, this.asState(), level, hitResult.getBlockPos(), player, hand, hitResult);
         }
 
@@ -1001,6 +_,12 @@
         public final boolean isRandomlyTicking() { // Paper - Perf: Final for inlining
             return this.isRandomlyTicking;
         }
+
+        // KioCG start - crops grow every day
+        public final boolean isDailyTicking() { // Paper - Perf: Final for inlining
+            return this.isDailyTicking;
+        }
+        // KioCG end - crops grow every day
 
         public long getSeed(BlockPos pos) {
             return this.getBlock().getSeed(this.asState(), pos);
