--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -313,6 +_,25 @@
         return true;
     }
 
+    // KioCG start - extended block placement
+    protected void playPlaceSound(BlockState state, Level level, BlockPos pos, Player player, @Nullable InteractionHand swingHand) {
+        boolean canBuild = level.checkEntityCollision(state, player, CollisionContext.of(player), pos, true); // Paper - Cancel hit for vanished players
+        if (canBuild) {
+            ((net.minecraft.server.level.ServerPlayer) player).connection.send(
+                    new net.minecraft.network.protocol.game.ClientboundSoundPacket(
+                            net.minecraft.core.registries.BuiltInRegistries.SOUND_EVENT.wrapAsHolder(soundType.getPlaceSound()),
+                            net.minecraft.sounds.SoundSource.BLOCKS,
+                            pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5,
+                            (soundType.getVolume() + 1.0F) / 2.0F,
+                            soundType.getPitch() * 0.8F,
+                            player.random.nextLong()));
+            if (swingHand != null) {
+                player.swing(swingHand, true);
+            }
+        }
+    }
+    // KioCG end - extended block placement
+
     protected float getShadeBrightness(BlockState state, BlockGetter level, BlockPos pos) {
         return state.isCollisionShapeFullBlock(level, pos) ? 0.2F : 1.0F;
     }
@@ -370,6 +_,11 @@
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
     }
 
+    // KioCG start - optimize block entity: item search hopper
+    protected void itemOnTop(BlockState state, Level level, BlockPos pos, net.minecraft.world.entity.item.ItemEntity itemEntity, InsideBlockEffectApplier effectApplier) {
+    }
+    // KioCG end - optimize block entity: item search hopper
+
     protected int getDirectSignal(BlockState state, BlockGetter level, BlockPos pos, Direction direction) {
         return 0;
     }
@@ -887,6 +_,12 @@
             this.getBlock().entityInside(this.asState(), level, pos, entity, effectApplier);
         }
 
+        // KioCG start - optimize block entity: item search hopper
+        public void itemOnTop(Level level, BlockPos pos, net.minecraft.world.entity.item.ItemEntity itemEntity, InsideBlockEffectApplier effectApplier) {
+            this.getBlock().itemOnTop(this.asState(), level, pos, itemEntity, effectApplier);
+        }
+        // KioCG end - optimize block entity: item search hopper
+
         public void spawnAfterBreak(ServerLevel level, BlockPos pos, ItemStack stack, boolean dropExperience) {
             this.getBlock().spawnAfterBreak(this.asState(), level, pos, stack, dropExperience);
             if (dropExperience) {this.getBlock().popExperience(level, pos, this.getBlock().getExpDrop(this.asState(), level, pos, stack, true));} // Paper - Properly handle xp dropping
@@ -897,6 +_,12 @@
         }
 
         public InteractionResult useItemOn(ItemStack stack, Level level, Player player, InteractionHand hand, BlockHitResult hitResult) {
+            // KioCG start - customize default components
+            if (stack.has(net.minecraft.core.component.DataComponents.CANNOT_USE)) {
+                return InteractionResult.TRY_WITH_EMPTY_HAND;
+            }
+            // KioCG end - customize default components
+
             return this.getBlock().useItemOn(stack, this.asState(), level, hitResult.getBlockPos(), player, hand, hitResult);
         }
 
