--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -387,7 +_,7 @@
     public boolean wantsToPickUp(ServerLevel level, ItemStack stack) {
         ItemStack itemInHand = this.getItemInHand(InteractionHand.MAIN_HAND);
         return !itemInHand.isEmpty()
-            && level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            && true // KioCG - better mob griefing setting
             && this.inventory.canAddItem(stack)
             && this.allayConsidersItemEqual(itemInHand, stack);
     }
