--- a/net/minecraft/world/entity/animal/allay/AllayAi.java
+++ b/net/minecraft/world/entity/animal/allay/AllayAi.java
@@ -76,7 +_,7 @@
         brain.addActivityWithConditions(
             Activity.IDLE,
             ImmutableList.of(
-                Pair.of(0, GoToWantedItem.create(allay -> true, 1.75F, true, 32)),
+                Pair.of(0, GoToWantedItem.create(allay -> true, 1.75F, true, 8)), // KioCG - optimize entity behavior: 降低搜索范围
                 Pair.of(1, new GoAndGiveItemsToTarget<>(AllayAi::getItemDepositPosition, 2.25F, 20)),
                 Pair.of(2, StayCloseToTarget.create(AllayAi::getItemDepositPosition, Predicate.not(AllayAi::hasWantedItem), 4, 16, 2.25F)),
                 Pair.of(3, SetEntityLookTargetSometimes.create(6.0F, UniformInt.of(30, 60))),
@@ -146,7 +_,7 @@
             if (memory.isPresent()) {
                 if (serverLevel.getEntity(memory.get()) instanceof ServerPlayer serverPlayer
                     && (serverPlayer.gameMode.isSurvival() || serverPlayer.gameMode.isCreative())
-                    && serverPlayer.closerThan(entity, 64.0)) {
+                    && serverPlayer.closerThan(entity, 16.0)) { // KioCG - optimize entity behavior: 降低搜索范围
                     return Optional.of(serverPlayer);
                 }
 
