--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -380,6 +_,7 @@
             && this.server.getPlayerIdleTimeout() > 0
             && Util.getMillis() - this.player.getLastActionTime() > this.server.getPlayerIdleTimeout() * 1000L * 60L && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits
             this.player.resetLastActionTime(); // CraftBukkit - SPIGOT-854
+            if (this.player.getNearbyChunkHot() >= this.player.level().paperConfig().kiocgConfig.chunkHot.phaseMedium || this.player.getNearbyChunkHot() >= this.player.level().paperConfig().kiocgConfig.chunkHot.phaseGood && MinecraftServer.getServer().tps5.getAverage() < 18.0) // KioCG - chunk hot
             this.disconnect(Component.translatable("multiplayer.disconnect.idling"), org.bukkit.event.player.PlayerKickEvent.Cause.IDLING); // Paper - kick event cause
         }
         // Paper start - Prevent causing expired keys from impacting new joins
@@ -1821,6 +_,7 @@
         }
 
         this.player.teleportSetPosition(posMoveRotation, relatives);
+        this.player.gameMode.checkFlying(); // KioCG - keep player abilities
         this.awaitingPositionFromClient = this.player.position();
         // CraftBukkit start - update last location
         this.lastPosX = this.awaitingPositionFromClient.x;
@@ -2064,31 +_,37 @@
                 BlockHitResult hitResult = this.player.level().clip(new net.minecraft.world.level.ClipContext(from, to, net.minecraft.world.level.ClipContext.Block.OUTLINE, net.minecraft.world.level.ClipContext.Fluid.NONE, this.player));
 
                 boolean cancelled;
+                org.bukkit.event.player.PlayerInteractEvent interactEvent; // KioCG - PlayerItemUseEvent
                 if (hitResult == null || hitResult.getType() != HitResult.Type.BLOCK) {
                     org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(this.player, Action.RIGHT_CLICK_AIR, itemInHand, hand);
                     cancelled = event.useItemInHand() == Event.Result.DENY;
+                    interactEvent = event; // KioCG - PlayerItemUseEvent
                 } else {
                     if (this.player.gameMode.firedInteract && this.player.gameMode.interactPosition.equals(hitResult.getBlockPos()) && this.player.gameMode.interactHand == hand && ItemStack.isSameItemSameComponents(this.player.gameMode.interactItemStack, itemInHand)) {
                         cancelled = this.player.gameMode.interactResult;
+                        interactEvent = null; // KioCG - PlayerItemUseEvent
                     } else {
                         org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(this.player, Action.RIGHT_CLICK_BLOCK, hitResult.getBlockPos(), hitResult.getDirection(), itemInHand, true, hand, hitResult.getLocation());
                         cancelled = event.useItemInHand() == Event.Result.DENY;
+                        interactEvent = event; // KioCG - PlayerItemUseEvent
                     }
                     this.player.gameMode.firedInteract = false;
                 }
 
-                if (cancelled) {
-                    this.player.resyncUsingItem(this.player); // Paper - Properly cancel usable items
-                    this.player.containerMenu.sendAllDataToRemote(); // SPIGOT-2524
-                    return;
-                }
+                // KioCG start - PlayerItemUseEvent: 移动到useItem()
+                // if (cancelled) {
+                //     this.player.resyncUsingItem(this.player); // Paper - Properly cancel usable items
+                //     this.player.containerMenu.sendAllDataToRemote(); // SPIGOT-2524
+                //     return;
+                // }
+                // KioCG end - PlayerItemUseEvent: 移动到useItem()
                 itemInHand = this.player.getItemInHand(hand); // Update in case it was changed in the event
                 if (itemInHand.isEmpty()) {
                     return;
                 }
                 // CraftBukkit end
 
-                if (this.player.gameMode.useItem(this.player, serverLevel, itemInHand, hand) instanceof InteractionResult.Success success
+                if (this.player.gameMode.useItem(this.player, serverLevel, itemInHand, hand, hitResult, cancelled, interactEvent) instanceof InteractionResult.Success success // KioCG - PlayerItemUseEvent
                     && success.swingSource() == InteractionResult.SwingSource.SERVER) {
                     this.player.swing(hand, true);
                 }
@@ -2747,6 +_,11 @@
                                     boolean triggerLeashUpdate = itemInHand.is(net.minecraft.world.item.Items.LEAD) && target instanceof net.minecraft.world.entity.Leashable;
                                     Item origItem = ServerGamePacketListenerImpl.this.player.getInventory().getSelectedItem().isEmpty() ? null : ServerGamePacketListenerImpl.this.player.getInventory().getSelectedItem().getItem();
 
+                                    // KioCG start - customize default components
+                                    if (itemInHand.has(DataComponents.CANNOT_USE)) {
+                                        event.setCancelled(true);
+                                    }
+                                    // KioCG end - customize default components
                                     ServerGamePacketListenerImpl.this.cserver.getPluginManager().callEvent(event);
 
                                     // Entity in bucket - SPIGOT-4048 and SPIGOT-6859a
@@ -2972,7 +_,7 @@
                                                     action = packet.buttonNum() == 0 ? InventoryAction.PICKUP_ALL : InventoryAction.PICKUP_HALF;
                                                 }
                                             } else if (slot.mayPlace(cursor)) {
-                                                if (ItemStack.isSameItemSameComponents(clickedItem, cursor)) {
+                                                if (ItemStack.isSameItemSameComponentsForFoodFreshness(clickedItem, cursor)) { // KioCG - food freshness
                                                     int toPlace = packet.buttonNum() == 0 ? cursor.getCount() : 1;
                                                     toPlace = Math.min(toPlace, clickedItem.getMaxStackSize() - clickedItem.getCount());
                                                     toPlace = Math.min(toPlace, slot.container.getMaxStackSize() - clickedItem.getCount());
@@ -3008,7 +_,7 @@
                                                         action = InventoryAction.SWAP_WITH_CURSOR;
                                                     }
                                                 }
-                                            } else if (ItemStack.isSameItemSameComponents(cursor, clickedItem)) {
+                                            } else if (ItemStack.isSameItemSameComponentsForFoodFreshness(cursor, clickedItem)) { // KioCG - food freshness
                                                 if (clickedItem.getCount() >= 0) {
                                                     if (clickedItem.getCount() + cursor.getCount() <= cursor.getMaxStackSize()) {
                                                         // As of 1.5, this is result slots only
@@ -3148,6 +_,15 @@
                     }
 
                     if (packet.clickType() != net.minecraft.world.inventory.ClickType.QUICK_CRAFT) {
+                        // KioCG start - customize default components
+                        if (slotNum >= 0) {
+                            Slot slot = this.player.containerMenu.getSlot(slotNum);
+                            if (slot != null && slot.hasItem() && slot.getItem().has(DataComponents.SLOT_LOCK) && !this.player.canUseGameMasterBlocks()) {
+                                cancelled = true;
+                            }
+                        }
+                        // KioCG end - customize default components
+
                         if (click == ClickType.NUMBER_KEY) {
                             event = new InventoryClickEvent(inventory, type, slotNum, click, action, packet.buttonNum());
                         } else {
