--- a/src/main/java/org/bukkit/block/Chest.java
+++ b/src/main/java/org/bukkit/block/Chest.java
@@ -9,7 +_,7 @@
 /**
  * Represents a captured state of a chest.
  */
-public interface Chest extends Container, LootableBlockInventory, Lidded { // Paper
+public interface Chest extends RowsContainer, LootableBlockInventory, Lidded { // Paper // KioCG - expand container rows
 
     /**
      * Gets the inventory of the chest block represented by this block state.
@@ -37,4 +_,13 @@
      */
     boolean isBlocked();
     // Paper end - More Chest Block API
+
+    // KioCG start - more api
+    @org.jspecify.annotations.Nullable
+    default Block getConnectedChest() {
+        return getConnectedChest(true);
+    }
+    @org.jspecify.annotations.Nullable
+    Block getConnectedChest(boolean verify);
+    // KioCG end - more api
 }
