--- a/src/main/java/io/papermc/paper/datacomponent/DataComponentTypes.java
+++ b/src/main/java/io/papermc/paper/datacomponent/DataComponentTypes.java
@@ -397,4 +_,12 @@
 
     private DataComponentTypes() {
     }
+
+    // KioCG start - customize default components
+    public static final DataComponentType.NonValued CANNOT_PLACE = unvalued("cannot_place");
+    public static final DataComponentType.NonValued CANNOT_USE = unvalued("cannot_use");
+    public static final DataComponentType.Valued<@IntRange(from = 1, to = 6) Integer> CONTAINER_ROWS = valued("container_rows");
+    public static final DataComponentType.Valued<ItemLore> EXTENDED_BEFORE_LORE = valued("extended_before_lore");
+    public static final DataComponentType.Valued<ItemLore> EXTENDED_AFTER_LORE = valued("extended_after_lore");
+    // KioCG end - customize default components
 }
