--- a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -45,6 +_,13 @@
         return this.potentialTarget != null && !(this.potentialTarget instanceof Player player1 && (player1.isSpectator() || player1.isCreative()));
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    protected int canUseDelay() {
+        return 20;
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public void start() {
         this.golem.setTarget(this.potentialTarget, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE); // CraftBukkit - reason
