--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
@@ -17,9 +_,16 @@
         this.setFlags(EnumSet.of(Goal.Flag.TARGET));
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    public boolean registerWhen() {
+        return this.tameAnimal.isTame() && !this.tameAnimal.isOrderedToSit();
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canUse() {
-        if (this.tameAnimal.isTame() && !this.tameAnimal.isOrderedToSit()) {
+        if (true || this.tameAnimal.isTame() && !this.tameAnimal.isOrderedToSit()) { // KioCG - optimize entity behavior
             LivingEntity owner = this.tameAnimal.getOwner();
             if (owner == null) {
                 return false;
