--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -342,14 +_,31 @@
         }
     }
 
-    public void tick() {
-        for (int i = 0; i < this.items.size(); i++) {
-            ItemStack item = this.getItem(i);
+    // KioCG start - food freshness
+    public void tick(boolean tickFood) {
+        ItemStack handItem = this.items.get(this.selected);
+        java.util.function.Function<ItemStack, Boolean> function = item -> {
+            item.inventoryTick(this.player.level(), this.player, item == handItem ? EquipmentSlot.MAINHAND : null);
+            return tickFood && item.tickFood(this.player.level().getGameTime());
+        };
+        tick(this.items, function);
+    }
+
+    public static void tick(NonNullList<ItemStack> itemStacks, java.util.function.Function<ItemStack, Boolean> function) {
+        for (int i = 0; i < itemStacks.size(); i++) {
+            ItemStack item = itemStacks.get(i);
             if (!item.isEmpty()) {
-                item.inventoryTick(this.player.level(), this.player, i == this.selected ? EquipmentSlot.MAINHAND : null);
+                if (function.apply(item)) {
+                    ItemStack rotInto = item.mayRotInto();
+                    if (rotInto != null) {
+                        item = rotInto;
+                        itemStacks.set(i, item);
+                    }
+                }
             }
         }
     }
+    // KioCG end - food freshness
 
     public boolean add(ItemStack stack) {
         return this.add(-1, stack);
@@ -360,7 +_,7 @@
             return false;
         } else {
             try {
-                if (stack.isDamaged()) {
+                if (!stack.isStackable()) { // KioCG - food freshness
                     if (slot == -1) {
                         slot = this.getFreeSlot();
                     }
