--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -181,7 +_,9 @@
     protected PlayerEnderChestContainer enderChestInventory = new PlayerEnderChestContainer(this); // CraftBukkit - add "this" to constructor
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
+    @Nullable public org.bukkit.inventory.Inventory carriedFrom = null; // KioCG - InventoryExchangeItemEvent
     protected FoodData foodData = new FoodData();
+    public com.kiocg.entity.player.ThirstData thirstData = new com.kiocg.entity.player.ThirstData(this); // KioCG - player thirsty
     protected int jumpTriggerTime;
     private boolean clientLoaded = false;
     protected int clientLoadedTimeoutTimer = 60;
@@ -219,6 +_,7 @@
     private int currentImpulseContextResetGraceTime = 0;
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
+    public int enderchestRows; // KioCG - expand container rows
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -296,6 +_,8 @@
             this.takeXpDelay--;
         }
 
+        this.campfireHealed = false; // KioCG - campfire healing
+
         if (this.isSleeping()) {
             this.sleepCounter++;
             // Paper start - Add PlayerDeepSleepEvent
@@ -303,13 +_,47 @@
                 if (!new io.papermc.paper.event.player.PlayerDeepSleepEvent((org.bukkit.entity.Player) getBukkitEntity()).callEvent()) {
                     this.sleepCounter = Integer.MIN_VALUE;
                 }
+
+                // KioCG start - sweet dreams
+                if (!this.level().dimensionType().bedWorks()) {
+                    this.getSleepingPos().ifPresent(pos -> {
+                        final BlockState blockState = this.level().getBlockState(pos);
+                        if (blockState.getBlock() instanceof net.minecraft.world.level.block.BedBlock bedBlock) {
+                            bedBlock.explodeBed(blockState, this.level(), pos);
+                        }
+                    });
+                }
+                // KioCG end - sweet dreams
             }
             // Paper end - Add PlayerDeepSleepEvent
             if (this.sleepCounter > 100) {
                 this.sleepCounter = 100;
+
+                // KioCG start - sleep progress
+                if (this instanceof ServerPlayer serverPlayer) {
+                    Stat<ResourceLocation> stat = Stats.CUSTOM.get(Stats.TIME_SINCE_REST);
+                    int timeSinceRest = serverPlayer.getStats().getValue(stat);
+                    if (timeSinceRest > 0) {
+                        int decrement;
+                        int playerInsomniaStartTicks = serverPlayer.level().paperConfig().entities.behavior.playerInsomniaStartTicks;
+                        if (timeSinceRest > playerInsomniaStartTicks) {
+                            decrement = -Math.min(timeSinceRest - playerInsomniaStartTicks, timeSinceRest);
+                        } else {
+                            decrement = -Mth.clamp(playerInsomniaStartTicks / 10 / 20, 24000 / 10 / 20, timeSinceRest);
+                        }
+                        serverPlayer.awardStat(stat, decrement);
+
+                        timeSinceRest = serverPlayer.getStats().getValue(stat);
+                        serverPlayer.sendSleepMessage(timeSinceRest, playerInsomniaStartTicks);
+                        if (timeSinceRest == 0) {
+
+                        }
+                    }
+                }
+                // KioCG end - sleep progress
             }
 
-            if (!this.level().isClientSide && this.level().isBrightOutside()) {
+            if (false && !this.level().isClientSide && this.level().isBrightOutside()) { // KioCG - sleep any time
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
@@ -329,6 +_,7 @@
         this.moveCloak();
         if (this instanceof ServerPlayer serverPlayer) {
             this.foodData.tick(serverPlayer);
+            this.thirstData.tick(); // KioCG - player thirsty
             this.awardStat(Stats.PLAY_TIME);
             this.awardStat(Stats.TOTAL_WORLD_TIME);
             if (this.isAlive()) {
@@ -341,6 +_,17 @@
 
             if (!this.isSleeping()) {
                 this.awardStat(Stats.TIME_SINCE_REST);
+
+                // KioCG start - sleep progress: 睡眠提醒
+                if (this.tickCount % (20 * 60 * 10) == 0) {
+                    Stat<ResourceLocation> stat = Stats.CUSTOM.get(Stats.TIME_SINCE_REST);
+                    int timeSinceRest = serverPlayer.getStats().getValue(stat);
+                    int playerInsomniaStartTicks = serverPlayer.level().paperConfig().entities.behavior.playerInsomniaStartTicks;
+                    if (timeSinceRest >= playerInsomniaStartTicks) {
+                        serverPlayer.sendSleepMessage(playerInsomniaStartTicks, playerInsomniaStartTicks);
+                    }
+                }
+                // KioCG end - sleep progress: 睡眠提醒
             }
         }
 
@@ -822,6 +_,7 @@
 
         this.setScore(input.getIntOr("Score", 0));
         this.foodData.readAdditionalSaveData(input);
+        this.thirstData.readAdditionalSaveData(input); // KioCG - player thirsty
         input.read("abilities", Abilities.Packed.CODEC).ifPresent(this.abilities::apply);
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.abilities.getWalkingSpeed());
         this.enderChestInventory.fromSlots(input.listOrEmpty("EnderItems", ItemStackWithSlot.CODEC));
@@ -846,6 +_,7 @@
         output.putInt("XpSeed", this.enchantmentSeed);
         output.putInt("Score", this.getScore());
         this.foodData.addAdditionalSaveData(output);
+        this.thirstData.addAdditionalSaveData(output); // KioCG - player thirsty
         output.store("abilities", Abilities.Packed.CODEC, this.abilities.pack());
         this.enderChestInventory.storeAsSlots(output.list("EnderItems", ItemStackWithSlot.CODEC));
         if (!this.getShoulderEntityLeft().isEmpty()) {
@@ -915,6 +_,17 @@
         }
     }
 
+    // KioCG start - campfire healing
+    private boolean campfireHealed;
+    public void campfireHealing() {
+        if (!campfireHealed && this.isHurt()) {
+            campfireHealed = true;
+            this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.SATIATED); // CraftBukkit - added RegainReason
+            this.causeFoodExhaustion(this.level().spigotConfig.regenExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+        }
+    }
+    // KioCG end - campfire healing
+
     @Override
     protected void blockUsingItem(ServerLevel level, LivingEntity entity) {
         super.blockUsingItem(level, entity);
@@ -1703,9 +_,13 @@
 
     public void giveExperiencePoints(int xpPoints) {
         this.increaseScore(xpPoints);
-        this.experienceProgress = this.experienceProgress + (float)xpPoints / this.getXpNeededForNextLevel();
+        // this.experienceProgress = this.experienceProgress + (float)xpPoints / this.getXpNeededForNextLevel(); // KioCG - level is money
         this.totalExperience = Mth.clamp(this.totalExperience + xpPoints, 0, Integer.MAX_VALUE);
 
+        // KioCG start - level is money
+        this.giveExperienceLevels(xpPoints);
+
+        /*
         while (this.experienceProgress < 0.0F) {
             float f = this.experienceProgress * this.getXpNeededForNextLevel();
             if (this.experienceLevel > 0) {
@@ -1722,6 +_,8 @@
             this.giveExperienceLevels(1);
             this.experienceProgress = this.experienceProgress / this.getXpNeededForNextLevel();
         }
+         */
+        // KioCG end - level is money
     }
 
     public int getEnchantmentSeed() {
@@ -1747,7 +_,7 @@
             this.totalExperience = 0;
         }
 
-        if (levels > 0 && this.experienceLevel % 5 == 0 && this.lastLevelUpTime < this.tickCount - 100.0F) {
+        if (false && levels > 0 && this.experienceLevel % 5 == 0 && this.lastLevelUpTime < this.tickCount - 100.0F) { // KioCG - level is money
             float f = this.experienceLevel > 30 ? 1.0F : this.experienceLevel / 30.0F;
             this.level().playSound(null, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_LEVELUP, this.getSoundSource(), f * 0.75F, 1.0F);
             this.lastLevelUpTime = this.tickCount;
@@ -1811,6 +_,14 @@
     }
 
     public boolean mayUseItemAt(BlockPos pos, Direction facing, ItemStack stack) {
+        // KioCG start - customize default components
+        if (stack.has(DataComponents.CANNOT_PLACE)) {
+            this.containerMenu.forceHeldSlot(InteractionHand.MAIN_HAND);
+            this.containerMenu.forceHeldSlot(InteractionHand.OFF_HAND);
+            return false;
+        }
+        // KioCG end - customize default components
+
         if (this.abilities.mayBuild) {
             return true;
         } else {
@@ -1822,7 +_,7 @@
 
     @Override
     protected int getBaseExperienceReward(ServerLevel level) {
-        return !level.getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) && !this.isSpectator() ? Math.min(this.experienceLevel * 7, 100) : 0;
+        return !level.getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) && !this.isSpectator() ? this.experienceLevel * 7 : 0; // KioCG - death drop protect
     }
 
     @Override
@@ -2323,6 +_,13 @@
     public boolean onClimbable() {
         return !this.abilities.flying && super.onClimbable();
     }
+
+    // KioCG start - chunk hot
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end - chunk hot
 
     public String debugInfo() {
         return MoreObjects.toStringHelper(this)
