--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -163,6 +_,8 @@
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
     protected FoodData foodData = new FoodData();
+    public com.kiocg.entity.player.ColdData coldData = new com.kiocg.entity.player.ColdData(); // KioCG - the long dark
+    public com.kiocg.entity.player.ThirstData thirstData = new com.kiocg.entity.player.ThirstData(); // KioCG - the long dark
     protected int jumpTriggerTime;
     private boolean clientLoaded = false;
     protected int clientLoadedTimeoutTimer = 60;
@@ -200,6 +_,7 @@
     private int currentImpulseContextResetGraceTime;
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
+    public int enderchestRows; // KioCG - expand container rows
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -246,7 +_,8 @@
             .add(Attributes.SUBMERGED_MINING_SPEED)
             .add(Attributes.SNEAKING_SPEED)
             .add(Attributes.MINING_EFFICIENCY)
-            .add(Attributes.SWEEPING_DAMAGE_RATIO);
+            .add(Attributes.SWEEPING_DAMAGE_RATIO)
+            .add(Attributes.SPAWN_REINFORCEMENTS_CHANCE); // KioCG - the long dark: 属性用于玩家温度
     }
 
     @Override
@@ -271,6 +_,8 @@
             this.takeXpDelay--;
         }
 
+        this.campfireHealed = false; // KioCG - campfire healing
+
         if (this.isSleeping()) {
             this.sleepCounter++;
             // Paper start - Add PlayerDeepSleepEvent
@@ -282,9 +_,32 @@
             // Paper end - Add PlayerDeepSleepEvent
             if (this.sleepCounter > 100) {
                 this.sleepCounter = 100;
+
+                // KioCG start - sleep progress
+                if (this instanceof ServerPlayer serverPlayer) {
+                    Stat<ResourceLocation> stat = Stats.CUSTOM.get(Stats.TIME_SINCE_REST);
+                    int timeSinceRest = serverPlayer.getStats().getValue(stat);
+                    if (timeSinceRest > 0) {
+                        int decrement;
+                        int playerInsomniaStartTicks = serverPlayer.level().paperConfig().entities.behavior.playerInsomniaStartTicks;
+                        if (timeSinceRest > playerInsomniaStartTicks) {
+                            decrement = -Math.min(timeSinceRest - playerInsomniaStartTicks, timeSinceRest);
+                        } else {
+                            decrement = -Mth.clamp(playerInsomniaStartTicks / 10 / 20, 24000 / 10 / 20, timeSinceRest);
+                        }
+                        serverPlayer.awardStat(stat, decrement);
+
+                        timeSinceRest = serverPlayer.getStats().getValue(stat);
+                        serverPlayer.sendSleepMessage(timeSinceRest, playerInsomniaStartTicks);
+                        if (timeSinceRest == 0) {
+
+                        }
+                    }
+                }
+                // KioCG end - sleep progress
             }
 
-            if (!this.level().isClientSide && this.level().isDay()) {
+            if (false && !this.level().isClientSide && this.level().isDay()) { // KioCG - sleep any time
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
@@ -304,6 +_,8 @@
         this.moveCloak();
         if (this instanceof ServerPlayer serverPlayer) {
             this.foodData.tick(serverPlayer);
+            this.coldData.tick(serverPlayer); // KioCG - the long dark
+            this.thirstData.tick(serverPlayer); // KioCG - the long dark
             this.awardStat(Stats.PLAY_TIME);
             this.awardStat(Stats.TOTAL_WORLD_TIME);
             if (this.isAlive()) {
@@ -828,6 +_,10 @@
 
         this.setScore(compound.getInt("Score"));
         this.foodData.readAdditionalSaveData(compound);
+        // KioCG start - the long dark
+        this.coldData.readAdditionalSaveData(compound);
+        this.thirstData.readAdditionalSaveData(compound);
+        // KioCG end - the long dark
         this.abilities.loadSaveData(compound);
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.abilities.getWalkingSpeed());
         if (compound.contains("EnderItems", 9)) {
@@ -870,6 +_,10 @@
         compound.putInt("XpSeed", this.enchantmentSeed);
         compound.putInt("Score", this.getScore());
         this.foodData.addAdditionalSaveData(compound);
+        // KioCG start - the long dark
+        this.coldData.addAdditionalSaveData(compound);
+        this.thirstData.addAdditionalSaveData(compound);
+        // KioCG end - the long dark
         this.abilities.addSaveData(compound);
         compound.put("EnderItems", this.enderChestInventory.createTag(this.registryAccess()));
         if (!this.getShoulderEntityLeft().isEmpty()) {
@@ -944,6 +_,28 @@
         }
     }
 
+    // KioCG start - campfire healing
+    private boolean campfireHealed;
+    public void campfireHealing() {
+        if (!campfireHealed && this.isHurt()) {
+            campfireHealed = true;
+            final float health = this.getHealth();
+            this.heal(1.0F);
+            this.causeFoodExhaustion(this.level().spigotConfig.regenExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+            if (this.getHealth() > health) {
+                this.serverLevel().sendParticlesSource(
+                        (net.minecraft.server.level.ServerPlayer) this,
+                        net.minecraft.core.particles.ParticleTypes.HEART,
+                        false, false,
+                        this.getX(), this.getY(), this.getZ(),
+                        5,
+                        0.3, 0.3, 0.3,
+                        1.0);
+            }
+        }
+    }
+    // KioCG end - campfire healing
+
     @Override
     protected void blockUsingShield(LivingEntity entity) {
         super.blockUsingShield(entity);
@@ -2415,6 +_,13 @@
     public boolean onClimbable() {
         return !this.abilities.flying && super.onClimbable();
     }
+
+    // KioCG start - chunk hot
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end - chunk hot
 
     public static enum BedSleepingProblem {
         NOT_POSSIBLE_HERE,
