--- a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
@@ -37,6 +_,7 @@
  * An implementation will handle the creation and application for ItemMeta.
  * This class should not be implemented by a plugin in a live environment.
  */
+@Deprecated // KioCG
 public interface ItemMeta extends Cloneable, ConfigurationSerializable, PersistentDataHolder {
 
     // Paper start
