--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -60,6 +_,7 @@
     private static final boolean DEFAULT_SHEARED = false;
     private int eatAnimationTick;
     private EatBlockGoal eatBlockGoal;
+    public int woolTime = this.random.nextInt(200); // KioCG - breeding animals
 
     private static int createSheepColor(DyeColor color) {
         if (color == DyeColor.WHITE) {
@@ -115,6 +_,14 @@
             this.eatAnimationTick = Math.max(0, this.eatAnimationTick - 1);
         }
 
+        // KioCG start - breeding animals
+        if (this.produceToday == 0 && --this.woolTime <= 0) {
+            this.produce();
+            this.ate();
+            this.woolTime = this.random.nextInt(200);
+        }
+        // KioCG end - breeding animals
+
         super.aiStep();
     }
 
@@ -222,6 +_,7 @@
                 }
             }
         );
+        if (this.canProduce()) { this.produce(); } else // KioCG - breeding animals
         this.setSheared(true);
     }
 
@@ -332,7 +_,7 @@
         super.ate();
         this.setSheared(false);
         if (this.isBaby()) {
-            this.ageUp(60);
+            // this.ageUp(60); // KioCG - breeding animals
         }
     }
 
