--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -17,7 +_,11 @@
     public final java.net.SocketAddress socketAddress;
     private final org.bukkit.craftbukkit.command.CraftRemoteConsoleCommandSender remoteConsole = new org.bukkit.craftbukkit.command.CraftRemoteConsoleCommandSender(this);
 
-    public RconConsoleSource(MinecraftServer server, java.net.SocketAddress socketAddress) {
+    // KioCG start - rcon
+    private final net.minecraft.server.rcon.thread.RconClient rconClient;
+    public RconConsoleSource(MinecraftServer server, java.net.SocketAddress socketAddress, net.minecraft.server.rcon.thread.RconClient rconClient) {
+        this.rconClient = rconClient;
+    // KioCG end - rcon
         this.socketAddress = socketAddress;
         // CraftBukkit end
         this.server = server;
@@ -40,7 +_,12 @@
 
     // CraftBukkit start - Send a String
     public void sendMessage(String message) {
-        this.buffer.append(message);
+        // KioCG start - rcon
+        try {
+            this.rconClient.sendCmdResponse(1, message);
+        } catch (Exception ex) {
+        }
+        // KioCG end - rcon
     }
 
     @Override
@@ -51,7 +_,7 @@
 
     @Override
     public void sendSystemMessage(Component component) {
-        this.buffer.append(component.getString());
+        this.sendMessage(component.getString()); // KioCG - rcon
     }
 
     @Override
