--- a/net/minecraft/world/entity/animal/Chicken.java
+++ b/net/minecraft/world/entity/animal/Chicken.java
@@ -68,7 +_,7 @@
 
     public Chicken(EntityType<? extends Chicken> entityType, Level level) {
         super(entityType, level);
-        this.eggTime = this.random.nextInt(6000) + 6000;
+        this.eggTime = this.random.nextInt(200); // KioCG - breeding animals
         this.setPathfindingMalus(PathType.WATER, 0.0F);
     }
 
@@ -111,7 +_,7 @@
         }
 
         this.flap = this.flap + this.flapping * 2.0F;
-        if (this.level() instanceof ServerLevel serverLevel && this.isAlive() && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) {
+        if (this.level() instanceof ServerLevel serverLevel && this.isAlive() && !this.isBaby() && !this.isChickenJockey() && this.canProduce() && --this.eggTime <= 0) { // KioCG - breeding animals
             this.forceDrops = true; // CraftBukkit
             if (this.dropFromGiftLootTable(serverLevel, BuiltInLootTables.CHICKEN_LAY, this::spawnAtLocation)) {
                 this.playSound(SoundEvents.CHICKEN_EGG, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
@@ -119,7 +_,8 @@
             }
             this.forceDrops = false; // CraftBukkit
 
-            this.eggTime = this.random.nextInt(6000) + 6000;
+            this.produce(); // KioCG - breeding animals
+            this.eggTime = this.random.nextInt(200); // KioCG - breeding animals
         }
     }
 
