--- a/net/minecraft/world/entity/animal/Dolphin.java
+++ b/net/minecraft/world/entity/animal/Dolphin.java
@@ -332,8 +_,9 @@
             }
 
             if (this.isBaby()) {
+                if (!this.canFeed(player)) return InteractionResult.PASS; // KioCG - breeding animals
                 itemInHand.consume(1, player);
-                this.ageUp(getSpeedUpSecondsWhenFeeding(-this.age), true);
+                this.ageUp(); // KioCG - breeding animals
             } else {
                 this.setGotFish(true);
                 itemInHand.consume(1, player);
@@ -501,6 +_,13 @@
             return this.player != null && this.player.isSwimming() && this.dolphin.getTarget() != this.player;
         }
 
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return 20;
+        }
+        // KioCG end - optimize entity behavior
+
         @Override
         public boolean canContinueToUse() {
             return this.player != null && this.player.isSwimming() && this.dolphin.distanceToSqr(this.player) < 256.0;
@@ -545,6 +_,13 @@
                 return !entitiesOfClass.isEmpty() || !Dolphin.this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty();
             }
         }
+
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return 20;
+        }
+        // KioCG end - optimize entity behavior
 
         @Override
         public void start() {
