--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -916,8 +_,10 @@
                     CriteriaTriggers.BRED_ANIMALS.trigger(serverPlayer, this.animal, this.partner, fox);
                 }
 
-                this.animal.setAge(6000);
-                this.partner.setAge(6000);
+                // KioCG start - breeding animals: 不需要繁殖冷却
+                // this.animal.setAge(6000);
+                // this.partner.setAge(6000);
+                // KioCG end - breeding animals: 不需要繁殖冷却
                 this.animal.resetLove();
                 this.partner.resetLove();
                 serverLevel.addFreshEntityWithPassengers(fox, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
@@ -972,7 +_,7 @@
         }
 
         protected void onReachedTarget() {
-            if (getServerLevel(Fox.this.level()).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (getServerLevel(Fox.this.level()).getGameRules().getInt(GameRules.RULE_RANDOMTICKING) > 0) { // KioCG - better mob griefing setting
                 BlockState blockState = Fox.this.level().getBlockState(this.blockPos);
                 if (blockState.is(Blocks.SWEET_BERRY_BUSH)) {
                     this.pickSweetBerries(blockState);
@@ -1167,7 +_,7 @@
                     } else {
                         boolean isPathClear = Fox.isPathClear(Fox.this, target);
                         if (!isPathClear) {
-                            Fox.this.getNavigation().createPath(target, 0);
+                            // Fox.this.getNavigation().createPath(target, 0); // KioCG - optimize entity behavior: 未使用?
                             Fox.this.setIsCrouching(false);
                             Fox.this.setIsInterested(false);
                         }
@@ -1180,6 +_,13 @@
             }
         }
 
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return 10;
+        }
+        // KioCG end - optimize entity behavior
+
         @Override
         public boolean canContinueToUse() {
             LivingEntity target = Fox.this.getTarget();
@@ -1238,7 +_,7 @@
                 }
             }
 
-            if (target != null && Fox.this.distanceTo(target) <= 2.0F) {
+            if (target != null && Fox.this.distanceToSqr(target) <= 2.0F * 2.0F) { // KioCG - optimize distanceTo
                 Fox.this.doHurtTarget(getServerLevel(Fox.this.level()), target);
             } else if (Fox.this.getXRot() > 0.0F
                 && Fox.this.onGround()
@@ -1279,7 +_,7 @@
                 .getEntitiesOfClass(ItemEntity.class, Fox.this.getBoundingBox().inflate(8.0, 8.0, 8.0), Fox.ALLOWED_ITEMS);
             ItemStack itemBySlot = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
             if (itemBySlot.isEmpty() && !entitiesOfClass.isEmpty()) {
-                Fox.this.getNavigation().moveTo(entitiesOfClass.get(0), 1.2F);
+                Fox.this.getNavigation().moveToLightweight(entitiesOfClass.get(0), 1.2F); // KioCG - optimize entity behavior
             }
         }
 
@@ -1288,7 +_,7 @@
             List<ItemEntity> entitiesOfClass = Fox.this.level()
                 .getEntitiesOfClass(ItemEntity.class, Fox.this.getBoundingBox().inflate(8.0, 8.0, 8.0), Fox.ALLOWED_ITEMS);
             if (!entitiesOfClass.isEmpty()) {
-                Fox.this.getNavigation().moveTo(entitiesOfClass.get(0), 1.2F);
+                Fox.this.getNavigation().moveToLightweight(entitiesOfClass.get(0), 1.2F); // KioCG - optimize entity behavior
             }
         }
     }
@@ -1427,6 +_,13 @@
             return Fox.this.xxa == 0.0F && Fox.this.yya == 0.0F && Fox.this.zza == 0.0F && (this.canSleep() || Fox.this.isSleeping());
         }
 
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return this.countdown > 0 ? 1 : 20;
+        }
+        // KioCG end - optimize entity behavior
+
         @Override
         public boolean canContinueToUse() {
             return this.canSleep();
@@ -1510,7 +_,7 @@
                     Fox.this.setIsCrouching(true);
                     Fox.this.getNavigation().stop();
                 } else {
-                    Fox.this.getNavigation().moveTo(target, 1.5);
+                    Fox.this.getNavigation().moveToLightweight(target, 1.5); // KioCG - optimize entity behavior
                 }
             }
         }
