--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -625,9 +_,9 @@
                 this.gotBamboo = true;
             }
 
-            if (this.isBaby()) {
+            if (this.isBaby() && this.canFeed()) { // KioCG - breeding animals
                 this.usePlayerItem(player, hand, itemInHand);
-                this.ageUp((int)(-this.getAge() / 20 * 0.1F), true);
+                this.ageUp(); // KioCG - breeding animals
             } else if (!this.level().isClientSide && this.getAge() == 0 && this.canFallInLove()) {
                 final ItemStack breedCopy = itemInHand.copy(); // Paper - Fix EntityBreedEvent copying
                 this.usePlayerItem(player, hand, itemInHand);
