--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -119,23 +_,32 @@
     protected void registerGoals() {
         this.goalSelector.addGoal(1, new FloatGoal(this));
         this.goalSelector.addGoal(1, new TamableAnimal.TamableAnimalPanicGoal(1.5, DamageTypeTags.PANIC_ENVIRONMENTAL_CAUSES));
-        this.goalSelector.addGoal(2, new SitWhenOrderedToGoal(this));
-        this.goalSelector.addGoal(3, new Wolf.WolfAvoidEntityGoal<>(this, Llama.class, 24.0F, 1.5, 1.5));
-        this.goalSelector.addGoal(4, new LeapAtTargetGoal(this, 0.4F));
-        this.goalSelector.addGoal(5, new MeleeAttackGoal(this, 1.0, true));
-        this.goalSelector.addGoal(6, new FollowOwnerGoal(this, 1.0, 10.0F, 2.0F));
-        this.goalSelector.addGoal(7, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(8, new WaterAvoidingRandomStrollGoal(this, 1.0));
+        // KioCG start - optimize entity behavior
+        this.goalRegister.addGoal(2, new SitWhenOrderedToGoal(this), wolf -> ((Wolf) wolf).isOrderedToSit());
+        this.goalRegister.addGoal(3, new Wolf.WolfAvoidEntityGoal<>(this, Llama.class, 24.0F, 1.5, 1.5), wolf -> !((Wolf) wolf).isTame());
+        this.goalRegister.addGoal(4, new LeapAtTargetGoal(this, 0.4F), wolf -> !((Wolf) wolf).isOrderedToSit());
+        this.goalRegister.addGoal(5, new MeleeAttackGoal(this, 1.0, true), wolf -> !((Wolf) wolf).isOrderedToSit());
+        this.goalRegister.addGoal(6, new FollowOwnerGoal(this, 1.0, 10.0F, 2.0F), wolf -> ((Wolf) wolf).isTame() && !((Wolf) wolf).isOrderedToSit() && ((Wolf) wolf).isFollowOwner());
+        this.goalRegister.addGoal(7, new BreedGoal(this, 1.0), wolf -> ((Wolf) wolf).isTame());
+        this.goalSelector.addGoal(8, new WaterAvoidingRandomStrollGoal(this, 1.0) {
+            @Override
+            public boolean canUse() {
+                return !((TamableAnimal) this.mob).isInSittingPose() && super.canUse();
+            }
+        });
+        // KioCG end - optimize entity behavior
         this.goalSelector.addGoal(9, new BegGoal(this, 8.0F));
         this.goalSelector.addGoal(10, new LookAtPlayerGoal(this, Player.class, 8.0F));
         this.goalSelector.addGoal(10, new RandomLookAroundGoal(this));
-        this.targetSelector.addGoal(1, new OwnerHurtByTargetGoal(this));
-        this.targetSelector.addGoal(2, new OwnerHurtTargetGoal(this));
+        // KioCG start - optimize entity behavior
+        this.goalRegister.addGoal(1, new OwnerHurtByTargetGoal(this), wolf -> ((Wolf) wolf).isTame() && !((Wolf) wolf).isOrderedToSit());
+        this.goalRegister.addGoal(2, new OwnerHurtTargetGoal(this), wolf -> ((Wolf) wolf).isTame() && !((Wolf) wolf).isOrderedToSit());
         this.targetSelector.addGoal(3, new HurtByTargetGoal(this).setAlertOthers());
-        this.targetSelector.addGoal(4, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
-        this.targetSelector.addGoal(5, new NonTameRandomTargetGoal<>(this, Animal.class, false, PREY_SELECTOR));
-        this.targetSelector.addGoal(6, new NonTameRandomTargetGoal<>(this, Turtle.class, false, Turtle.BABY_ON_LAND_SELECTOR));
-        this.targetSelector.addGoal(7, new NearestAttackableTargetGoal<>(this, AbstractSkeleton.class, false));
+        this.goalRegister.addGoal(4, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt), wolf -> !((Wolf) wolf).isOrderedToSit());
+        this.goalRegister.addGoal(5, new NonTameRandomTargetGoal<>(this, Animal.class, false, PREY_SELECTOR), wolf -> !((Wolf) wolf).isTame());
+        this.goalRegister.addGoal(6, new NonTameRandomTargetGoal<>(this, Turtle.class, false, Turtle.BABY_ON_LAND_SELECTOR), wolf -> !((Wolf) wolf).isTame());
+        this.goalRegister.addGoal(7, new NearestAttackableTargetGoal<>(this, AbstractSkeleton.class, false), wolf -> !((Wolf) wolf).isOrderedToSit());
+        // KioCG end - optimize entity behavior
         this.targetSelector.addGoal(8, new ResetUniversalAngerTargetGoal<>(this, true));
     }
 
@@ -415,6 +_,18 @@
                 return InteractionResult.SUCCESS;
             }
 
+            // KioCG start - dont kill my pet
+            if (itemInHand.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) {
+                this.usePlayerItem(player, hand, itemInHand);
+                this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                if (player.isCreative() || !this.isInvulnerable()) {
+                    this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                }
+
+                return InteractionResult.SUCCESS;
+            }
+            // KioCG end - dont kill my pet
+
             if (!(item instanceof DyeItem dyeItem && this.isOwnedBy(player))) {
                 if (this.isEquippableInSlot(itemInHand, EquipmentSlot.BODY) && !this.isWearingBodyArmor() && this.isOwnedBy(player) && !this.isBaby()) {
                     this.setBodyArmorItem(itemInHand.copyWithCount(1));
@@ -650,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return super.canUse() && this.toAvoid instanceof Llama && !this.wolf.isTame() && this.avoidLlama((Llama)this.toAvoid);
+            return super.canUse() && this.toAvoid instanceof Llama && this.avoidLlama((Llama)this.toAvoid); // KioCG - optimize entity behavior
         }
 
         private boolean avoidLlama(Llama llama) {
