--- a/net/minecraft/network/HashedStack.java
+++ b/net/minecraft/network/HashedStack.java
@@ -50,7 +_,7 @@
         public boolean matches(ItemStack stack, HashedPatchMap.HashGenerator hashGenerator) {
             return this.count == stack.getCount()
                 && this.item.equals(stack.getItemHolder())
-                && this.components.matches(stack.getComponentsPatch(), hashGenerator);
+                && this.components.matches(stack.getRemoteComponentsPatch(), hashGenerator); // KioCG - customize default components
         }
     }
 }
