--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -455,7 +_,7 @@
             f = 4.0F;
             i = 60;
             i1 = 5;
-            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
+            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && this.canFallInLove()) { // KioCG - breeding animals
                 flag = true;
                 this.setInLove(player, stack.copy()); // Paper - Fix EntityBreedEvent copying
             }
@@ -463,7 +_,7 @@
             f = 10.0F;
             i = 240;
             i1 = 10;
-            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
+            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && this.canFallInLove()) { // KioCG - breeding animals
                 flag = true;
                 this.setInLove(player, stack.copy()); // Paper - Fix EntityBreedEvent copying
             }
@@ -478,6 +_,7 @@
             this.level().addParticle(ParticleTypes.HAPPY_VILLAGER, this.getRandomX(1.0), this.getRandomY() + 0.5, this.getRandomZ(1.0), 0.0, 0.0, 0.0);
             if (!this.level().isClientSide) {
                 this.ageUp(i);
+                // this.ageUp(); // KioCG - breeding animals: 保留额外成长
                 flag = true;
             }
         }
@@ -672,6 +_,19 @@
                     this.equipBodyArmor(player, itemInHand);
                     return InteractionResult.SUCCESS;
                 }
+
+                // KioCG start - modern horse armor
+                if (this.isWearingBodyArmor()
+                    && this.getBodyArmorItem().isDamaged()
+                    && this.getBodyArmorItem().isValidRepairItem(itemInHand)) {
+                    itemInHand.shrink(1);
+                    this.playSound(SoundEvents.WOLF_ARMOR_REPAIR);
+                    ItemStack bodyArmorItem = this.getBodyArmorItem();
+                    int i = (int)(bodyArmorItem.getMaxDamage() * 0.125F);
+                    bodyArmorItem.setDamageValue(Math.max(0, bodyArmorItem.getDamageValue() - i));
+                    return InteractionResult.SUCCESS;
+                }
+                // KioCG end - modern horse armor
             }
 
             this.doPlayerRide(player);
