--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1452,4 +_,13 @@
      */
     boolean canUseEquipmentSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
     // Paper end - Expose canUseSlot
+
+    // KioCG start - entity baby api
+    public void setNormalSimulate();
+    public void setBabySimulate();
+    public void setHugeSimulate();
+    public boolean isBabySimulate();
+    public boolean isBabyOrBabySimulate();
+    public boolean isHugeSimulate();
+    // KioCG end - entity baby api
 }
