--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3895,6 +_,14 @@
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // KioCG start - add items vanilla-like
+    default void giveItem(@NotNull ItemStack... items) throws IllegalArgumentException {
+        this.getInventory().addItem(items).values().forEach(leftover -> this.getOpenInventory().setItem(-999, leftover));
+    }
+    // KioCG end - add items vanilla-like
+
+    boolean isInStructure(); // KioCG - player in structure api
+
     @NotNull
     @Override
     Spigot spigot();
