--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3853,6 +_,28 @@
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // KioCG start - add items vanilla-like
+    default void giveItem(ItemStack... items) throws IllegalArgumentException {
+        this.getInventory().addItem(items).values().forEach(leftover -> this.getOpenInventory().setItem(-999, leftover));
+    }
+    // KioCG end - add items vanilla-like
+
+    boolean isInStructure(); // KioCG - player in structure api
+
+    // KioCG start - the long dark
+    int getColdValue();
+    void setColdValue(int value);
+    void addColdValue(int add);
+
+    int getThirstValue();
+    void setThirstValue(int value);
+    void addThirstValue(int add);
+
+    float getThirstRegain();
+    void setThirstRegain(float value);
+    void addThirstRegain(float add);
+    // KioCG end - the long dark
+
     @Override
     Spigot spigot();
     // Spigot end
