--- a/src/main/java/org/bukkit/entity/Wolf.java
+++ b/src/main/java/org/bukkit/entity/Wolf.java
@@ -104,6 +_,11 @@
      */
     void setSoundVariant(@NotNull SoundVariant soundVariant);
 
+    // KioCG start - dont kill my pet
+    public boolean isFollowOwner();
+    public void setFollowOwner(boolean followOwner);
+    // KioCG end - dont kill my pet
+
     /**
      * Represents the variant of a wolf.
      */
