--- a/net/minecraft/server/rcon/thread/RconThread.java
+++ b/net/minecraft/server/rcon/thread/RconThread.java
@@ -65,7 +_,7 @@
         int i = properties.rconPort;
         if (0 < i && 65535 >= i) {
             String string = properties.rconPassword;
-            if (string.isEmpty()) {
+            if (false && string.isEmpty()) { // KioCG - rcon: allow empty/any password
                 LOGGER.warn("No rcon password set in server.properties, rcon disabled!");
                 return null;
             } else {
