--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -24,6 +_,12 @@
     protected int forcedAge = 0;
     protected int forcedAgeTimer;
     public boolean ageLocked; // CraftBukkit
+    // KioCG start - breeding animals
+    private int survivalDays;
+    private int feedingDays;
+    public int feedingToday;
+    public int produceToday;
+    // KioCG end - breeding animals
 
     protected AgeableMob(EntityType<? extends AgeableMob> entityType, Level level) {
         super(entityType, level);
@@ -70,6 +_,8 @@
     }
 
     public void ageUp(int amount, boolean forced) {
+        if (forced) { ageUp(); return; } // KioCG - breeding animals: 额外的保障
+
         if (this.ageLocked) return; // Paper - Honor ageLock
         int age = this.getAge();
         int previousAge = age;
@@ -96,6 +_,19 @@
         this.ageUp(amount, false);
     }
 
+    // KioCG start - breeding animals
+    public void ageUp() {
+        if (!this.canFeed()) {
+            this.hurtServer(this.serverLevel(), this.damageSources().generic(), 2.0F);
+            return;
+        }
+
+        this.heal(2.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
+        ageUp(240);
+        this.feedingFeed();
+    }
+    // KioCG end - breeding animals
+
     public void setAge(int age) {
         int age1 = this.getAge();
         this.age = age;
@@ -105,12 +_,42 @@
         }
     }
 
+    // KioCG start - breeding animals
+    public boolean canFeed() {
+        return this.feedingToday < (this.isHugeSimulate() ? 3 : 1);
+    }
+
+    public void feedingFeed() {
+        if (this.feedingToday++ == 0 && !this.isBaby()) {
+            if (++this.feedingDays >= this.level().paperConfig().kiocgConfig.entity.animal.feedingHugeStartDays) {
+                if (!this.isHugeSimulate() && this.random.nextFloat() < (float) this.level().paperConfig().kiocgConfig.entity.animal.feedingHugeChance) {
+                    this.setHugeSimulate();
+                }
+            }
+        }
+    }
+
+    public boolean canProduce() {
+        return this.produceToday < (this.isHugeSimulate() ? 3 : 1);
+    }
+
+    public void produce() {
+        ++produceToday;
+    }
+    // KioCG end - breeding animals
+
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
         super.addAdditionalSaveData(output);
         output.putInt("Age", this.getAge());
         output.putInt("ForcedAge", this.forcedAge);
         output.putBoolean("AgeLocked", this.ageLocked); // CraftBukkit
+        // KioCG start - breeding animals
+        output.putInt("KioCG.SurvivalDays", this.survivalDays);
+        output.putInt("KioCG.FeedingDays", this.feedingDays);
+        output.putInt("KioCG.FeedingToday", this.feedingToday);
+        output.putInt("KioCG.ProduceToday", this.produceToday);
+        // KioCG end - breeding animals
     }
 
     @Override
@@ -119,6 +_,12 @@
         this.setAge(input.getIntOr("Age", 0));
         this.forcedAge = input.getIntOr("ForcedAge", 0);
         this.ageLocked = input.getBooleanOr("AgeLocked", false); // CraftBukkit
+        // KioCG start - breeding animals
+        this.survivalDays = input.getIntOr("KioCG.SurvivalDays", 0);
+        this.feedingDays = input.getIntOr("KioCG.FeedingDays", 0);
+        this.feedingToday = input.getIntOr("KioCG.FeedingToday", 0);
+        this.produceToday = input.getIntOr("KioCG.ProduceToday", 0);
+        // KioCG end - breeding animals
     }
 
     @Override
@@ -139,7 +_,7 @@
         } else {
             int age = this.getAge();
             if (age < 0) {
-                this.setAge(++age);
+                // this.setAge(++age); // KioCG - breeding animals: 阻止小动物自然生长
             } else if (age > 0) {
                 this.setAge(--age);
             }
@@ -161,10 +_,19 @@
         } else if (this.isAlive()) {
             int age = this.getAge();
             if (age < 0) {
-                this.setAge(++age);
+                // this.setAge(++age); // KioCG - breeding animals: 阻止小动物自然生长
             } else if (age > 0) {
                 this.setAge(--age);
             }
+
+            // KioCG start - breeding animals
+            final int days = this.level().getDays();
+            if (days != this.survivalDays) {
+                this.survivalDays = days;
+                this.feedingToday = 0;
+                this.produceToday = 0;
+            }
+            // KioCG end - breeding animals
         }
     }
 
@@ -180,7 +_,10 @@
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    // KioCG start - entity baby api
+    public void setBaby1(boolean baby) {
+        if (baby) removeScaleAttributes();
+    // KioCG end - entity baby api
         this.setAge(baby ? -24000 : 0);
     }
 
