--- a/net/minecraft/world/entity/ConversionType.java
+++ b/net/minecraft/world/entity/ConversionType.java
@@ -98,6 +_,8 @@
             newMob.setBaby(true);
         }
 
+        if (!newMob.isBaby()) newMob.transformScaleSimulate(oldMob); // KioCG - entity baby api
+
         if (oldMob instanceof AgeableMob ageableMob && newMob instanceof AgeableMob ageableMob1) {
             ageableMob1.setAge(ageableMob.getAge());
             ageableMob1.forcedAge = ageableMob.forcedAge;
@@ -113,6 +_,15 @@
         if (conversionParams.preserveCanPickUpLoot()) {
             newMob.setCanPickUpLoot(oldMob.canPickUpLoot());
         }
+
+        // KioCG start - fix converted entity data
+        newMob.aware = oldMob.aware;
+        newMob.persist = oldMob.persist;
+        newMob.fromNetherPortal = oldMob.fromNetherPortal;
+        newMob.spawnedViaMobSpawner = oldMob.spawnedViaMobSpawner;
+        newMob.copyOrigin(oldMob);
+        // KioCG end - fix converted entity data
+        newMob.spawningAnimation = false; // KioCG - mob spawning or dying animation
 
         newMob.setLeftHanded(oldMob.isLeftHanded());
         newMob.setNoAi(oldMob.isNoAi());
