--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -360,6 +_,12 @@
     // Paper start
     public @Nullable Vec3 origin;
     public @Nullable UUID originWorld;
+    // KioCG start - fix converted entity data
+    public void copyOrigin(Entity entity) {
+        this.origin = entity.origin;
+        this.originWorld = entity.originWorld;
+    }
+    // KioCG end - fix converted entity data
     // Paper end
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean fixedPose = false; // Paper - Expand Pose API
@@ -2245,7 +_,11 @@
                         this.push(-d, 0.0, -d1);
                     }
 
-                    if (!entity.isVehicle() && entity.isPushable()) {
+                    // KioCG start - item entity collision: 物品不要挤压生物
+                    boolean isItem1 = this instanceof ItemEntity;
+                    boolean isItem2 = entity instanceof ItemEntity;
+                    if (isItem1 == isItem2 && !entity.isVehicle() && entity.isPushable()) {
+                    // KioCG end - item entity collision: 物品不要挤压生物
                         entity.push(d, 0.0, d1);
                     }
                 }
@@ -3606,6 +_,7 @@
     }
 
     public int getAirSupply() {
+        if (this instanceof ServerPlayer player) return player.getBukkitEntity().realAir; // KioCG - the long dark
         return this.entityData.get(DATA_AIR_SUPPLY_ID);
     }
 
@@ -3620,6 +_,7 @@
             this.entityData.markDirty(Entity.DATA_AIR_SUPPLY_ID);
             return;
         }
+        if (this instanceof ServerPlayer player) { player.getBukkitEntity().realAir = event.getAmount(); } else // KioCG - the long dark
         this.entityData.set(Entity.DATA_AIR_SUPPLY_ID, event.getAmount());
         // CraftBukkit end
     }
@@ -5200,6 +_,12 @@
         return this.level;
     }
 
+    // KioCG start
+    public ServerLevel serverLevel() {
+        return (ServerLevel)this.level();
+    }
+    // KioCG end
+
     public void setLevel(Level level) {
         this.level = level;
     }
@@ -5365,4 +_,10 @@
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    // KioCG start - chunk hot
+    public boolean shouldTickHot() {
+        return this.tickCount > 20 && this.isAlive();
+    }
+    // KioCG end - chunk hot
 }
