--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -349,6 +_,13 @@
         );
         if (randomItemWith.isPresent()) {
             ItemStack itemStack = randomItemWith.get().itemStack();
+            // KioCG start - adjust enchantment
+            int level = itemStack.getEnchantmentLevel(this.registryAccess(), net.minecraft.world.item.enchantment.Enchantments.MENDING);
+            if (player.random.nextInt(10) >= level) {
+                return value;
+            }
+            // KioCG end - adjust enchantment
+
             int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(player.serverLevel(), itemStack, value);
             int min = Math.min(i, itemStack.getDamageValue());
             // CraftBukkit start
