--- a/net/minecraft/world/entity/InsideBlockEffectType.java
+++ b/net/minecraft/world/entity/InsideBlockEffectType.java
@@ -11,8 +_,16 @@
         }
     }),
     FIRE_IGNITE(BaseFireBlock::fireIgnite),
-    LAVA_IGNITE((entity, pos) -> entity.lavaIgnite(pos)), // Paper - track lava contact
-    EXTINGUISH(Entity::clearFire);
+    // KioCG start - the long dark
+    LAVA_IGNITE((entity, pos) -> {
+        entity.lavaIgnite(pos); // Paper - track lava contact
+        if (entity instanceof net.minecraft.world.entity.player.Player player) player.coldData.isInLava = true;
+    }),
+    EXTINGUISH(entity -> {
+        entity.clearFire();
+        if (entity instanceof net.minecraft.world.entity.player.Player player) player.coldData.isInWater = true;
+    });
+    // KioCG end - the long dark
 
     private final Applier effect; // Paper - track position inside effect was triggered on
 
