--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -168,6 +_,7 @@
             }
 
             Entity leashHolder = entity.getLeashHolder();
+            if (!(entity instanceof LeashFenceKnotEntity && leashHolder instanceof LeashFenceKnotEntity)) // KioCG - leash fence
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double d = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
@@ -328,6 +_,15 @@
         return new Vec3(0.0, entity.getEyeHeight(), entity.getBbWidth() * 0.4F);
     }
 
+    // KioCG start - leash fence
+    default Vec3 getLeashPosition(float partialTick) {
+        Entity entity = (Entity)this;
+        float h = entity.getPreciseBodyRotation(partialTick) * (float) (Math.PI / 180.0);
+        Vec3 vec3 = this.getLeashOffset(partialTick).yRot(-h);
+        return entity.getPosition(partialTick).add(vec3);
+    }
+    // KioCG end - leash fence
+
     default void setLeashedTo(Entity leashHolder, boolean broadcastPacket) {
         if (this != leashHolder) {
             setLeashedTo((Entity & Leashable)this, leashHolder, broadcastPacket);
@@ -342,6 +_,7 @@
         } else {
             Entity entity1 = leashData.leashHolder;
             leashData.setLeashHolder(leashHolder);
+            entity.setLeashData(leashData); // KioCG - leash fence
             if (entity1 != null && entity1 != leashHolder) {
                 entity1.notifyLeasheeRemoved(entity);
             }
@@ -371,6 +_,7 @@
                 Entity var3 = entity.level().getEntity(leashData.delayedLeashHolderId);
                 if (var3 instanceof Entity) {
                     leashData.setLeashHolder(var3);
+                    entity.setLeashData(leashData); // KioCG - leash fence
                 }
             }
 
