--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -301,6 +_,12 @@
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
+    // KioCG start - entity baby api
+    private ScaleSimulate scaleSimulate = ScaleSimulate.NORMAL;
+    private final ResourceLocation SCALE_MODIFIER_ID = ResourceLocation.withDefaultNamespace("scale_simulate");
+    private final AttributeModifier SCALE_MODIFIER_BABY = new AttributeModifier(SCALE_MODIFIER_ID, this.level().paperConfig().kiocgConfig.entity.sizeScalingBaby.getOrDefault(this.getType(), 0.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    private final AttributeModifier SCALE_MODIFIER_HUGE = new AttributeModifier(SCALE_MODIFIER_ID, this.level().paperConfig().kiocgConfig.entity.sizeScalingHuge.getOrDefault(this.getType(), 1.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    // KioCG end - entity baby api
 
     @Override
     public float getBukkitYaw() {
@@ -572,7 +_,7 @@
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+        if (true || !this.getBlockStateOnLegacy().isAir()) { // KioCG - the long dark
             int ticksFrozen = this.getTicksFrozen();
             if (ticksFrozen > 0) {
                 AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
@@ -594,6 +_,70 @@
         return false;
     }
 
+    // KioCG start - entity baby api
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> {
+                if (isBaby()) return; // Vanilla baby
+
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_BABY);
+            }
+            case HUGE -> {
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_HUGE);
+            }
+            case NORMAL -> {
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.removeModifier(this.SCALE_MODIFIER_ID);
+            }
+        }
+    }
+
+    public void setBabySimulate() {
+        this.setScaleSimulate(ScaleSimulate.BABY);
+    }
+
+    public void setHugeSimulate() {
+        this.setScaleSimulate(ScaleSimulate.HUGE);
+    }
+
+    public boolean isBabySimulate() {
+        return this.scaleSimulate == ScaleSimulate.BABY;
+    }
+
+    public boolean isBabyOrBabySimulate() {
+        return this.isBabySimulate() || this.isBaby();
+    }
+
+    public boolean isHugeSimulate() {
+        return this.scaleSimulate == ScaleSimulate.HUGE;
+    }
+
+    public void removeScaleAttributes() {
+        this.setScaleSimulate(ScaleSimulate.NORMAL);
+    }
+
+    public void transformScaleSimulate(LivingEntity original) {
+        if (original.isBabyOrBabySimulate()) {
+            this.setBabySimulate();
+        } else if (original.isHugeSimulate()) {
+            this.setHugeSimulate();
+        } else {
+            this.removeScaleAttributes();
+        }
+    }
+
+    public enum ScaleSimulate {
+        NORMAL,
+        BABY,
+        HUGE
+    }
+    // KioCG end - entity baby api
+
     public float getAgeScale() {
         return this.isBaby() ? 0.5F : 1.0F;
     }
@@ -620,11 +_,11 @@
     }
 
     public boolean shouldDropExperience() {
-        return !this.isBaby();
+        return !this.isBabyOrBabySimulate(); // KioCG - entity baby api
     }
 
     protected boolean shouldDropLoot() {
-        return !this.isBaby();
+        return !this.isBabyOrBabySimulate(); // KioCG - entity baby api
     }
 
     protected int decreaseAirSupply(int currentAir) {
@@ -799,6 +_,12 @@
         });
         DataResult<Tag> dataResult = this.brain.serializeStart(NbtOps.INSTANCE);
         dataResult.resultOrPartial(LOGGER::error).ifPresent(brain -> compound.put("Brain", brain));
+
+        // KioCG start - entity baby api
+        if (this.scaleSimulate != ScaleSimulate.NORMAL) {
+            compound.putString("KioCG.ScaleSimulate", this.scaleSimulate.toString());
+        }
+        // KioCG end - entity baby api
     }
 
     @Override
@@ -882,6 +_,17 @@
         if (compound.contains("Brain", 10)) {
             this.brain = this.makeBrain(new Dynamic<>(NbtOps.INSTANCE, compound.get("Brain")));
         }
+
+        // KioCG start - entity baby api
+        if (compound.contains("KioCG.ScaleSimulate")) {
+            String ss = compound.getString("KioCG.ScaleSimulate");
+            try {
+                this.setScaleSimulate(ScaleSimulate.valueOf(ss));
+            } catch (Exception ignored) {
+                LOGGER.error("Unknown scale simulate " + ss + " for " + this);
+            }
+        }
+        // KioCG end - entity baby api
     }
 
     // CraftBukkit start
@@ -1307,6 +_,7 @@
 
     public void heal(float healAmount, EntityRegainHealthEvent.RegainReason regainReason, boolean isFastRegen) {
         // Paper end - Forward
+        if (!(this instanceof net.minecraft.world.entity.player.Player player) || !player.thirstData.isThirsty(player)) { // KioCG - the long dark
         float health = this.getHealth();
         if (health > 0.0F) {
             EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), healAmount, regainReason, isFastRegen); // Paper
@@ -1317,9 +_,11 @@
 
             if (!event.isCancelled()) {
                 this.setHealth((float) (this.getHealth() + event.getAmount()));
+                if (this instanceof net.minecraft.world.entity.player.Player player) player.thirstData.addThirstRegain((float) (Math.max(this.getHealth() - health, 0) * this.level().paperConfig().kiocgConfig.theLongDark.thirstValue.thirstRegainMultiplierHealth)); // KioCG - the long dark
             }
             // CraftBukkit end
         }
+        } // KioCG - the long dark
     }
 
     public float getHealth() {
@@ -1755,11 +_,17 @@
         }
     }
 
+    // KioCG start - dont kill my pet
+    public boolean nearDeath() {
+        return false;
+    }
+    // KioCG end - dont kill my pet
+
     protected void createWitherRose(@Nullable LivingEntity entitySource) {
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (this.dead && entitySource instanceof WitherBoss) { // Paper
-                if (serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                if (serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) { // KioCG - better mob griefing setting
                     BlockPos blockPos = this.blockPosition();
                     BlockState blockState = Blocks.WITHER_ROSE.defaultBlockState();
                     if (this.level().getBlockState(blockPos).isAir() && blockState.canSurvive(this.level(), blockPos)) {
@@ -2382,7 +_,15 @@
                 }
                 // CraftBukkit end
                 this.getCombatTracker().recordDamage(damageSource, amount);
-                this.setHealth(this.getHealth() - amount);
+                // KioCG start - dont kill my pet
+                float health = this.getHealth() - amount;
+                if (health <= 0.0F && originalDamage != Float.MAX_VALUE && !damageSource.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this instanceof TamableAnimal pet && pet.isTame()) {
+                        health = Float.MIN_VALUE;
+                    }
+                }
+                this.setHealth(health);
+                // KioCG end - dont kill my pet
                 // CraftBukkit start
                 if (!human) {
                     this.setAbsorptionAmount(this.getAbsorptionAmount() - amount);
@@ -2743,7 +_,7 @@
     }
 
     public float getVoicePitch() {
-        return this.isBaby()
+        return this.isBabyOrBabySimulate() // KioCG - entity baby api
             ? (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.5F
             : (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F;
     }
@@ -3260,6 +_,8 @@
                 ItemStack itemBySlot = entry.getValue();
                 if (!itemBySlot.isEmpty() && !itemBySlot.isBroken()) {
                     itemBySlot.forEachModifier(equipmentSlot1, (attribute, modifier) -> {
+                        if (this instanceof net.minecraft.world.entity.monster.Zombie && attribute == Attributes.SPAWN_REINFORCEMENTS_CHANCE) return; // KioCG - the long dark: 属性用于玩家温度
+
                         AttributeInstance instance = this.attributes.getInstance(attribute);
                         if (instance != null) {
                             instance.removeModifier(modifier.id());
@@ -3456,7 +_,7 @@
         profilerFiller.push("freezing");
         if (!this.level().isClientSide && !this.isDeadOrDying() && !this.freezeLocked) { // Paper - Freeze Tick Lock API
             int ticksFrozen = this.getTicksFrozen();
-            if (this.isInPowderSnow && this.canFreeze()) {
+            if (this.isInPowderSnow && this.canFreeze() || this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen(player)) { // KioCG - the long dark
                 this.setTicksFrozen(Math.min(this.getTicksRequiredToFreeze(), ticksFrozen + 1));
             } else {
                 this.setTicksFrozen(Math.max(0, ticksFrozen - 2));
@@ -3476,11 +_,12 @@
             this.checkAutoSpinAttack(boundingBox, this.getBoundingBox());
         }
 
-        this.pushEntities();
+        boolean moved = this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot(); // KioCG - optimize entity collide: 提取变量
+        this.pushEntities(moved); // KioCG - optimize entity collide
         profilerFiller.pop();
         // Paper start - Add EntityMoveEvent
         if (((ServerLevel) this.level()).hasEntityMoveEvent && !(this instanceof Player)) {
-            if (this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot()) {
+            if (moved) { // KioCG - optimize entity collide: 提取变量
                 Location from = new Location(this.level().getWorld(), this.xo, this.yo, this.zo, this.yRotO, this.xRotO);
                 Location to = new Location(this.level().getWorld(), this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
                 io.papermc.paper.event.entity.EntityMoveEvent event = new io.papermc.paper.event.entity.EntityMoveEvent(this.getBukkitLivingEntity(), from, to.clone());
@@ -3541,7 +_,10 @@
     protected void serverAiStep() {
     }
 
-    protected void pushEntities() {
+    // KioCG start - optimize entity collide
+    protected void pushEntities(boolean moved) {
+        if (!moved) return;
+        // KioCG end - optimize entity collide
         if (!(this.level() instanceof ServerLevel serverLevel)) {
             this.level().getEntities(EntityTypeTest.forClass(Player.class), this.getBoundingBox(), EntitySelector.pushableBy(this)).forEach(this::doPush);
         } else {
