--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -185,7 +_,10 @@
     protected static final int LIVING_ENTITY_FLAG_OFF_HAND = 2;
     public static final int LIVING_ENTITY_FLAG_SPIN_ATTACK = 4;
     protected static final EntityDataAccessor<Byte> DATA_LIVING_ENTITY_FLAGS = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.BYTE);
-    public static final EntityDataAccessor<Float> DATA_HEALTH_ID = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.FLOAT);
+    // KioCG start - hide entity attribute
+    public static final EntityDataAccessor<Float> DATA_HEALTH_ID1 = SynchedEntityData.defineId(LivingEntity.class, EntityDataSerializers.FLOAT);
+    public float realHealth;
+    // KioCG end - hide entity attribute
     private static final EntityDataAccessor<List<ParticleOptions>> DATA_EFFECT_PARTICLES = SynchedEntityData.defineId(
         LivingEntity.class, EntityDataSerializers.PARTICLES
     );
@@ -287,13 +_,23 @@
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
     public int shieldBlockingDelay = this.level().paperConfig().misc.shieldBlockingDelay; // Paper - Make shield blocking delay configurable
     // CraftBukkit end
+    // KioCG start - entity baby api
+    private ScaleSimulate scaleSimulate = ScaleSimulate.NORMAL;
+    private static final ResourceLocation SCALE_MODIFIER_ID = ResourceLocation.withDefaultNamespace("scale_simulate");
+    private final AttributeModifier SCALE_MODIFIER_BABY = new AttributeModifier(SCALE_MODIFIER_ID, this.level().paperConfig().kiocgConfig.entity.sizeScalingBaby.getOrDefault(this.getType(), 0.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    private final AttributeModifier SCALE_MODIFIER_HUGE = new AttributeModifier(SCALE_MODIFIER_ID, this.level().paperConfig().kiocgConfig.entity.sizeScalingHuge.getOrDefault(this.getType(), 1.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    // KioCG end - entity baby api
+    private boolean damagedKnockback; // KioCG - fix damage knockback
 
     protected LivingEntity(EntityType<? extends LivingEntity> entityType, Level level) {
         super(entityType, level);
         this.attributes = new AttributeMap(DefaultAttributes.getSupplier(entityType));
         this.craftAttributes = new org.bukkit.craftbukkit.attribute.CraftAttributeMap(this.attributes); // CraftBukkit
         // CraftBukkit - this.setHealth(this.getMaxHealth()) inlined and simplified to skip the instanceof check for Player, as getBukkitEntity() is not initialized in constructor
-        this.entityData.set(LivingEntity.DATA_HEALTH_ID, this.getMaxHealth());
+        // KioCG start - hide entity attribute
+        this.entityData.set(LivingEntity.DATA_HEALTH_ID1, (float) net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH.value().getDefaultValue());
+        this.realHealth = this.getMaxHealth();
+        // KioCG end - hide entity attribute
         this.equipment = this.createEquipment();
         this.blocksBuilding = true;
         this.reapplyPosition();
@@ -337,7 +_,7 @@
         builder.define(DATA_EFFECT_AMBIENCE_ID, false);
         builder.define(DATA_ARROW_COUNT_ID, 0);
         builder.define(DATA_STINGER_COUNT_ID, 0);
-        builder.define(DATA_HEALTH_ID, 1.0F);
+        builder.define(DATA_HEALTH_ID1, 1.0F); // KioCG - hide entity attribute
         builder.define(SLEEPING_POS_ID, Optional.empty());
     }
 
@@ -449,6 +_,7 @@
                 }
             }
 
+            if (this instanceof ServerPlayer serverPlayer) serverPlayer.thirstData.underWater = false; // KioCG - the long dark
             if (this.isEyeInFluid(FluidTags.WATER)
                 && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
                 boolean flag1 = !this.canBreatheUnderwater()
@@ -461,8 +_,19 @@
                         serverLevel1.broadcastEntityEvent(this, (byte)67);
                         this.hurtServer(serverLevel1, this.damageSources().drown(), 2.0F);
                     }
+                    // KioCG start - the long dark
+                    if (this instanceof ServerPlayer serverPlayer) {
+                        serverPlayer.thirstData.underWater = true;
+                        serverPlayer.sendAirMessage(Math.max(this.getAirSupply(), 0), this.getMaxAirSupply());
+                    }
+                    // KioCG end - the long dark
                 } else if (this.getAirSupply() < this.getMaxAirSupply()) {
                     this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+                    // KioCG start - the long dark
+                    if (this instanceof ServerPlayer serverPlayer) {
+                        serverPlayer.sendAirMessage(Math.max(this.getAirSupply(), 0), this.getMaxAirSupply());
+                    }
+                    // KioCG end - the long dark
                 }
 
                 if (this.isPassenger() && this.getVehicle() != null && this.getVehicle().dismountsUnderwater()) {
@@ -470,6 +_,11 @@
                 }
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
                 this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+                // KioCG start - the long dark
+                if (this instanceof ServerPlayer serverPlayer) {
+                    serverPlayer.sendAirMessage(Math.max(this.getAirSupply(), 0), this.getMaxAirSupply());
+                }
+                // KioCG end - the long dark
             }
 
             BlockPos blockPos = this.blockPosition();
@@ -537,7 +_,7 @@
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+        if (true || !this.getBlockStateOnLegacy().isAir()) { // KioCG - the long dark
             int ticksFrozen = this.getTicksFrozen();
             if (ticksFrozen > 0) {
                 AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
@@ -559,6 +_,70 @@
         return false;
     }
 
+    // KioCG start - entity baby api
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> {
+                if (isBaby()) return; // Vanilla baby
+
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_BABY);
+            }
+            case HUGE -> {
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_HUGE);
+            }
+            case NORMAL -> {
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.removeModifier(SCALE_MODIFIER_ID);
+            }
+        }
+    }
+
+    public void setBabySimulate() {
+        this.setScaleSimulate(ScaleSimulate.BABY);
+    }
+
+    public void setHugeSimulate() {
+        this.setScaleSimulate(ScaleSimulate.HUGE);
+    }
+
+    public boolean isBabySimulate() {
+        return this.scaleSimulate == ScaleSimulate.BABY;
+    }
+
+    public boolean isBabyOrBabySimulate() {
+        return this.isBabySimulate() || this.isBaby();
+    }
+
+    public boolean isHugeSimulate() {
+        return this.scaleSimulate == ScaleSimulate.HUGE;
+    }
+
+    public void removeScaleAttributes() {
+        this.setScaleSimulate(ScaleSimulate.NORMAL);
+    }
+
+    public void transformScaleSimulate(LivingEntity original) {
+        if (original.isBabyOrBabySimulate()) {
+            this.setBabySimulate();
+        } else if (original.isHugeSimulate()) {
+            this.setHugeSimulate();
+        } else {
+            this.removeScaleAttributes();
+        }
+    }
+
+    public enum ScaleSimulate {
+        NORMAL,
+        BABY,
+        HUGE
+    }
+    // KioCG end - entity baby api
+
     public float getAgeScale() {
         return this.isBaby() ? 0.5F : 1.0F;
     }
@@ -578,18 +_,21 @@
 
     protected void tickDeath() {
         this.deathTime++;
-        if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
+        // KioCG start - mob spawning or dying animation
+        int maxDeathTime = (this instanceof Mob mob && mob.dyingAnimation) ? 30 : 20;
+        if (this.deathTime >= maxDeathTime && !this.level().isClientSide() && !this.isRemoved()) {
+        // KioCG end - mob spawning or dying animation
             this.level().broadcastEntityEvent(this, (byte)60);
             this.remove(Entity.RemovalReason.KILLED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
         }
     }
 
     public boolean shouldDropExperience() {
-        return !this.isBaby();
+        return !this.isBabyOrBabySimulate(); // KioCG - entity baby api
     }
 
     protected boolean shouldDropLoot() {
-        return !this.isBaby();
+        return !this.isBabyOrBabySimulate(); // KioCG - entity baby api
     }
 
     protected int decreaseAirSupply(int currentAir) {
@@ -796,6 +_,12 @@
         if (this.locatorBarIcon.hasData()) {
             output.store("locator_bar_icon", Waypoint.Icon.CODEC, this.locatorBarIcon);
         }
+
+        // KioCG start - entity baby api
+        if (this.scaleSimulate != ScaleSimulate.NORMAL) {
+            output.putString("KioCG.ScaleSimulate", this.scaleSimulate.toString());
+        }
+        // KioCG end - entity baby api
     }
 
     // Paper start - Extend dropItem API
@@ -911,6 +_,16 @@
         this.lastHurtByMobTimestamp = input.getIntOr("ticks_since_last_hurt_by_mob", 0) + this.tickCount;
         this.equipment.setAll(input.read("equipment", EntityEquipment.CODEC).orElseGet(EntityEquipment::new));
         this.locatorBarIcon = input.read("locator_bar_icon", Waypoint.Icon.CODEC).orElseGet(Waypoint.Icon::new);
+
+        // KioCG start - entity baby api
+        input.getString("KioCG.ScaleSimulate").ifPresent(scaleSimulate -> {
+            try {
+                this.setScaleSimulate(ScaleSimulate.valueOf(scaleSimulate));
+            } catch (Exception ignored) {
+                LOGGER.error("Unknown scale simulate " + scaleSimulate + " for " + this);
+            }
+        });
+        // KioCG end - entity baby api
     }
 
     // CraftBukkit start
@@ -1352,6 +_,7 @@
 
     public void heal(float healAmount, EntityRegainHealthEvent.RegainReason regainReason, boolean isFastRegen) {
         // Paper end - Forward
+        if (!(this instanceof net.minecraft.world.entity.player.Player player) || !player.thirstData.isThirsty(player)) { // KioCG - the long dark
         float health = this.getHealth();
         if (health > 0.0F) {
             EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), healAmount, regainReason, isFastRegen); // Paper
@@ -1362,9 +_,11 @@
 
             if (!event.isCancelled()) {
                 this.setHealth((float) (this.getHealth() + event.getAmount()));
+                if (this instanceof net.minecraft.world.entity.player.Player player) player.thirstData.addThirstRegain((float) (Math.max(this.getHealth() - health, 0) * this.level().paperConfig().kiocgConfig.theLongDark.thirstValue.thirstRegainMultiplierHealth)); // KioCG - the long dark
             }
             // CraftBukkit end
         }
+        } // KioCG - the long dark
     }
 
     public float getHealth() {
@@ -1373,7 +_,7 @@
             return (float) serverPlayer.getBukkitEntity().getHealth();
         }
         // CraftBukkit end
-        return this.entityData.get(DATA_HEALTH_ID);
+        return this.realHealth; // KioCG - hide entity attribute
     }
 
     public void setHealth(float health) {
@@ -1397,7 +_,7 @@
             return;
         }
         // CraftBukkit end
-        this.entityData.set(DATA_HEALTH_ID, Mth.clamp(health, 0.0F, this.getMaxHealth()));
+        this.realHealth = Mth.clamp(health, 0.0F, this.getMaxHealth()); // KioCG - hide entity attribute
     }
 
     public boolean isDeadOrDying() {
@@ -1477,11 +_,12 @@
                 // CraftBukkit end
                 this.hurtDuration = 10;
                 this.hurtTime = this.hurtDuration;
+                this.damagedKnockback = false; // KioCG - fix damage knockback
             }
 
             this.resolveMobResponsibleForDamage(damageSource);
             this.resolvePlayerResponsibleForDamage(damageSource);
-            if (flag1) {
+            if (flag1 || !this.damagedKnockback) { // KioCG - fix damage knockback
                 BlocksAttacks blocksAttacks = this.getUseItem().get(DataComponents.BLOCKS_ATTACKS);
                 if (flag && blocksAttacks != null) {
                     blocksAttacks.onBlocked(level, this);
@@ -1873,11 +_,17 @@
         }
     }
 
+    // KioCG start - dont kill my pet
+    public boolean nearDeath() {
+        return false;
+    }
+    // KioCG end - dont kill my pet
+
     protected void createWitherRose(@Nullable LivingEntity entitySource) {
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (this.dead && entitySource instanceof WitherBoss) { // Paper
-                if (serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                if (serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) { // KioCG - better mob griefing setting
                     BlockPos blockPos = this.blockPosition();
                     BlockState blockState = Blocks.WITHER_ROSE.defaultBlockState();
                     if (this.level().getBlockState(blockPos).isAir() && blockState.canSurvive(this.level(), blockPos)) {
@@ -2062,9 +_,14 @@
                 return;
             }
 
+            // KioCG start - fix damage knockback
+            if (!event.getKnockback().isZero()) {
+            this.damagedKnockback = true;
+            // KioCG end - fix damage knockback
             this.hasImpulse = true;
             this.setDeltaMovement(deltaMovement.add(event.getKnockback().getX(), event.getKnockback().getY(), event.getKnockback().getZ())); // Paper - knockback events
             // CraftBukkit end
+            } // KioCG - fix damage knockback
         }
     }
 
@@ -2471,7 +_,15 @@
                 }
                 // CraftBukkit end
                 this.getCombatTracker().recordDamage(damageSource, amount);
-                this.setHealth(this.getHealth() - amount);
+                // KioCG start - dont kill my pet
+                float health = this.getHealth() - amount;
+                if (health <= 0.0F && originalDamage != Float.MAX_VALUE && !damageSource.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this instanceof TamableAnimal pet && pet.isTame()) {
+                        health = Float.MIN_VALUE;
+                    }
+                }
+                this.setHealth(health);
+                // KioCG end - dont kill my pet
                 // CraftBukkit start
                 if (!human) {
                     this.setAbsorptionAmount(this.getAbsorptionAmount() - amount);
@@ -2833,7 +_,7 @@
     }
 
     public float getVoicePitch() {
-        return this.isBaby()
+        return this.isBabyOrBabySimulate() // KioCG - entity baby api
             ? (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.5F
             : (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F;
     }
@@ -3375,6 +_,8 @@
                 ItemStack itemBySlot = entry.getValue();
                 if (!itemBySlot.isEmpty() && !itemBySlot.isBroken()) {
                     itemBySlot.forEachModifier(equipmentSlot1, (holder, attributeModifier) -> {
+                        if (this instanceof net.minecraft.world.entity.monster.Zombie && holder == Attributes.SPAWN_REINFORCEMENTS_CHANCE) return; // KioCG - the long dark: 属性用于玩家温度
+
                         AttributeInstance instance = this.attributes.getInstance(holder);
                         if (instance != null) {
                             instance.removeModifier(attributeModifier.id());
@@ -3550,7 +_,11 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             profilerFiller.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // KioCG start - the long dark
+                if (this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen(player)) {
+                    player.setTicksFrozen(Math.min(player.getTicksRequiredToFreeze(), player.getTicksFrozen() + 1));
+                } else this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // KioCG end - the long dark
             }
 
             this.removeFrost();
@@ -3568,11 +_,12 @@
             this.checkAutoSpinAttack(boundingBox, this.getBoundingBox());
         }
 
-        this.pushEntities();
+        boolean moved = this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot(); // KioCG - optimize entity collide: 提取变量
+        this.pushEntities(moved); // KioCG - optimize entity collide
         profilerFiller.pop();
         // Paper start - Add EntityMoveEvent
         if (((ServerLevel) this.level()).hasEntityMoveEvent && !(this instanceof Player)) {
-            if (this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot()) {
+            if (moved) { // KioCG - optimize entity collide: 提取变量
                 Location from = new Location(this.level().getWorld(), this.xo, this.yo, this.zo, this.yRotO, this.xRotO);
                 Location to = new Location(this.level().getWorld(), this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
                 io.papermc.paper.event.entity.EntityMoveEvent event = new io.papermc.paper.event.entity.EntityMoveEvent(this.getBukkitLivingEntity(), from, to.clone());
@@ -3645,7 +_,10 @@
     protected void serverAiStep() {
     }
 
-    protected void pushEntities() {
+    // KioCG start - optimize entity collide
+    protected void pushEntities(boolean moved) {
+        if (!moved) return;
+        // KioCG end - optimize entity collide
         // Paper start - don't run getEntities if we're not going to use its result
         if (!this.isPushable()) {
             return;
