--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -125,6 +_,8 @@
     public GoalSelector goalSelector;
     public @Nullable net.minecraft.world.entity.ai.goal.FloatGoal goalFloat; // Paper - Allow nerfed mobs to jump and float
     public GoalSelector targetSelector;
+    public com.kiocg.entity.GoalRegister goalRegister = new com.kiocg.entity.GoalRegister(); // KioCG - optimize entity behavior
+    public boolean tickRunningGoalsOnly; // KioCG - optimize entity behavior
     @Nullable
     private LivingEntity target;
     private final Sensing sensing;
@@ -140,6 +_,8 @@
     private int homeRadius = -1;
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
+    @Nullable private net.minecraft.server.level.ServerBossEvent bossbar; // KioCG - entity bossbar api
+    private int ticksSinceLastPlayerScan = 21; // KioCG - entity bossbar api
 
     protected Mob(EntityType<? extends Mob> entityType, Level level) {
         super(entityType, level);
@@ -286,6 +_,7 @@
             }
         }
         this.target = target;
+        this.goalRegister.reassess(); // KioCG - optimize entity behavior
         return true;
         // CraftBukkit end
     }
@@ -319,6 +_,12 @@
 
     @Override
     public void baseTick() {
+        // KioCG start - optimize entity behavior
+        if (this.firstTick) {
+            this.goalRegister.reassess();
+        }
+        // KioCG end - optimize entity behavior
+
         super.baseTick();
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("mobBaseTick");
@@ -383,6 +_,16 @@
         if (!this.level().isClientSide && this.tickCount % 5 == 0) {
             this.updateControlFlags();
         }
+
+        // KioCG start - entity bossbar api
+        if (this.bossbar != null) {
+            this.bossbar.setProgress(this.getHealth() / this.getMaxHealth());
+            if (++this.ticksSinceLastPlayerScan >= 20) {
+                updatePlayers();
+                this.ticksSinceLastPlayerScan = 0;
+            }
+        }
+        // KioCG end - entity bossbar api
     }
 
     protected void updateControlFlags() {
@@ -433,6 +_,12 @@
             output.putString("Paper.DespawnInPeacefulOverride", this.despawnInPeacefulOverride.name());
         }
         // Paper end - allow changing despawnInPeaceful
+
+        // KioCG start - entity bossbar api
+        if (this.bossbar != null) {
+            output.store("KioCG.Bossbar", net.minecraft.nbt.CompoundTag.CODEC, this.bossbar.save(this.registryAccess()));
+        }
+        // KioCG end - entity bossbar api
     }
 
     @Override
@@ -461,6 +_,12 @@
         this.setNoAi(input.getBooleanOr("NoAI", false));
         this.aware = input.getBooleanOr("Bukkit.Aware", true); // CraftBukkit
         this.despawnInPeacefulOverride = input.read("Paper.DespawnInPeacefulOverride", io.papermc.paper.util.PaperCodecs.TRI_STATE_CODEC).orElse(net.kyori.adventure.util.TriState.NOT_SET); // Paper - allow changing despawnInPeaceful
+
+        // KioCG start - entity bossbar api
+        input.read("KioCG.Bossbar", net.minecraft.nbt.CompoundTag.CODEC).ifPresent(bossbar -> {
+            this.bossbar = net.minecraft.server.level.ServerBossEvent.load(bossbar, this.registryAccess());
+        });
+        // KioCG end - entity bossbar api
     }
 
     @Override
@@ -599,6 +_,7 @@
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot slot, ItemStack stack) {
         this.setItemSlot(slot, stack);
         this.setGuaranteedDrop(slot);
+        if (this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) // KioCG - drop picked up items when despawn
         this.persistenceRequired = true;
     }
 
@@ -711,10 +_,13 @@
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldActuallyDespawnInPeaceful()) { //Paper - allow changing despawnInPeaceful
+            this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity nearestPlayer = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
             if (nearestPlayer != null) {
+                if (net.minecraft.server.level.ChunkMap.euclideanDistanceSquared(this.chunkPosition(), nearestPlayer.position()) > 16384.0D) return; // KioCG - fix mobs disappearing when player disconnects
+
                 // Paper start - Configurable despawn distances
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
                 final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
@@ -725,11 +_,13 @@
                 final double distanceSquared = dxSqr + dzSqr + dySqr;
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                    this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                  }
 
                 if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                        this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     }
                 } else {
@@ -742,6 +_,16 @@
         }
     }
 
+    // KioCG start - drop picked up items when despawn
+    private void dropPickedUpItems() {
+        this.forceDrops = true; // CraftBukkit
+        this.dropPreservedEquipment(serverLevel(), itemStack -> {
+            return !EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP);
+        });
+        this.forceDrops = false; // CraftBukkit
+    }
+    // KioCG end - drop picked up items when despawn
+
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
@@ -759,7 +_,7 @@
         this.sensing.tick();
         profilerFiller.pop();
         int i = this.tickCount + this.getId();
-        if (i % 2 != 0 && this.tickCount > 1) {
+        if (tickRunningGoalsOnly = (i % 2 != 0 && this.tickCount > 1)) { // KioCG - optimize entity behavior
             profilerFiller.push("targetSelector");
             this.targetSelector.tickRunningGoals(false);
             profilerFiller.pop();
@@ -1479,8 +_,23 @@
         return (this.entityData.get(DATA_MOB_FLAGS_ID) & 4) != 0;
     }
 
-    public void setBaby(boolean baby) {
-    }
+    // KioCG start - entity baby api
+    public final void setBaby(boolean baby) { // final防止重写, 便于版本更新检查
+        this.setBaby1(baby);
+    }
+
+    public void setBaby1(boolean baby) {
+    }
+
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> setBaby(true);
+            case HUGE -> setBaby(false);
+        }
+
+        super.setScaleSimulate(scale);
+    }
+    // KioCG end - entity baby api
 
     @Override
     public HumanoidArm getMainArm() {
@@ -1569,6 +_,51 @@
         }
     }
 
+    // KioCG start - entity bossbar api
+    @Nullable
+    public net.minecraft.server.level.ServerBossEvent getBossBar() {
+        return this.bossbar;
+    }
+
+    public void setBossBar(@Nullable net.minecraft.server.level.ServerBossEvent bossbar) {
+        if (this.bossbar != null) {
+            this.bossbar.removeAllPlayers();
+        }
+        this.bossbar = bossbar;
+        updatePlayers();
+    }
+
+    private void updatePlayers() {
+        java.util.Set<net.minecraft.server.level.ServerPlayer> set = com.google.common.collect.Sets.newHashSet();
+
+        net.minecraft.server.level.ChunkMap.TrackedEntity entityTracker = this.serverLevel().getChunkSource().chunkMap.entityMap.get(this.getId());
+        if (entityTracker != null) {
+            for (net.minecraft.server.network.ServerPlayerConnection connection : entityTracker.seenBy) {
+                net.minecraft.server.level.ServerPlayer serverPlayer = connection.getPlayer();
+                if (this.position().distanceToSqr(serverPlayer.position()) <= 16 * 16) {
+                    this.bossbar.addPlayer(serverPlayer);
+                    set.add(serverPlayer);
+                }
+            }
+        }
+
+        java.util.Set<net.minecraft.server.level.ServerPlayer> set1 = com.google.common.collect.Sets.newHashSet(this.bossbar.getPlayers());
+        set1.removeAll(set);
+
+        for (net.minecraft.server.level.ServerPlayer serverPlayer1 : set1) {
+            this.bossbar.removePlayer(serverPlayer1);
+        }
+    }
+
+    @Override
+    public void stopSeenByPlayer(net.minecraft.server.level.ServerPlayer player) {
+        super.stopSeenByPlayer(player);
+        if (this.bossbar != null) {
+            this.bossbar.removePlayer(player);
+        }
+    }
+    // KioCG end - entity bossbar api
+
     @VisibleForTesting
     public void removeFreeWill() {
         this.removeAllGoals(goal -> true);
@@ -1601,8 +_,9 @@
     @Override
     protected void onAttributeUpdated(Holder<Attribute> attribute) {
         super.onAttributeUpdated(attribute);
-        if (attribute.is(Attributes.FOLLOW_RANGE) || attribute.is(Attributes.TEMPT_RANGE)) {
+        if (attribute.is(Attributes.FOLLOW_RANGE) /*|| attribute.is(Attributes.TEMPT_RANGE)*/) { // KioCG - optimize entity behavior
             this.getNavigation().updatePathfinderMaxVisitedNodes();
+            ((net.minecraft.world.entity.ai.Brain<Mob>) this.getBrain()).updateSensors(this); // KioCG - optimize entity behavior: 在属性修改时更新
         }
     }
 }
