--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -145,6 +_,13 @@
     private BlockPos restrictCenter = BlockPos.ZERO;
     private float restrictRadius = -1.0F;
     public boolean aware = true; // CraftBukkit
+    @Nullable private net.minecraft.server.level.ServerBossEvent bossbar; // KioCG - entity bossbar api
+    private int ticksSinceLastPlayerScan = 21; // KioCG - entity bossbar api
+    // KioCG start - mob spawning or dying animation
+    private static final ResourceLocation ANIMATION_SCALE_MODIFIER_ID = ResourceLocation.withDefaultNamespace("animation_scale");
+    public boolean spawningAnimation = this.level().paperConfig().kiocgConfig.entity.mobSpawningAnimation;
+    public boolean dyingAnimation = this.level().paperConfig().kiocgConfig.entity.mobDyingAnimation;
+    // KioCG end - mob spawning or dying animation
 
     protected Mob(EntityType<? extends Mob> entityType, Level level) {
         super(entityType, level);
@@ -399,7 +_,37 @@
         if (!this.level().isClientSide && this.tickCount % 5 == 0) {
             this.updateControlFlags();
         }
-    }
+
+        // KioCG start - entity bossbar api
+        if (this.bossbar != null) {
+            this.bossbar.setProgress(this.getHealth() / this.getMaxHealth());
+            if (++this.ticksSinceLastPlayerScan >= 20) {
+                updatePlayers();
+                this.ticksSinceLastPlayerScan = 0;
+            }
+        }
+        // KioCG end - entity bossbar api
+
+        // KioCG start - mob spawning or dying animation
+        if (this.spawningAnimation && this.tickCount <= 10) {
+            AttributeInstance scale = this.getAttribute(Attributes.SCALE);
+            if (scale.hasModifier(ANIMATION_SCALE_MODIFIER_ID)) {
+                scale.addOrUpdateTransientModifier(new AttributeModifier(ANIMATION_SCALE_MODIFIER_ID, (this.tickCount / 10.0) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL));
+            }
+        }
+        // KioCG end - mob spawning or dying animation
+    }
+
+    // KioCG start - mob spawning or dying animation
+    @Override
+    protected void tickDeath() {
+        super.tickDeath();
+        if (this.dyingAnimation && this.deathTime > 20 && this.deathTime <= 30) {
+            int reventDeathTime = 30 - this.deathTime;
+            this.getAttribute(Attributes.SCALE).addOrUpdateTransientModifier(new AttributeModifier(ANIMATION_SCALE_MODIFIER_ID, (reventDeathTime / 10.0) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL));
+        }
+    }
+    // KioCG end - mob spawning or dying animation
 
     protected void updateControlFlags() {
         boolean flag = !(this.getControllingPassenger() instanceof Mob);
@@ -486,6 +_,12 @@
             compound.putBoolean("NoAI", this.isNoAi());
         }
         compound.putBoolean("Bukkit.Aware", this.aware); // CraftBukkit
+
+        // KioCG start - entity bossbar api
+        if (this.bossbar != null) {
+            compound.put("KioCG.Bossbar", this.bossbar.save(this.registryAccess()));
+        }
+        // KioCG end - entity bossbar api
     }
 
     @Override
@@ -568,6 +_,12 @@
             this.aware = compound.getBoolean("Bukkit.Aware");
         }
         // CraftBukkit end
+
+        // KioCG start - entity bossbar api
+        if (compound.contains("KioCG.Bossbar")) {
+            this.bossbar = net.minecraft.server.level.ServerBossEvent.load(compound.getCompound("KioCG.Bossbar"), this.registryAccess());
+        }
+        // KioCG end - entity bossbar api
     }
 
     @Override
@@ -620,7 +_,7 @@
             && this.canPickUpLoot()
             && this.isAlive()
             && !this.dead
-            && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            && true) { // KioCG - better mob griefing setting
             Vec3i pickupReach = this.getPickupReach();
 
             for (ItemEntity itemEntity : this.level()
@@ -700,6 +_,7 @@
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot slot, ItemStack stack) {
         this.setItemSlot(slot, stack);
         this.setGuaranteedDrop(slot);
+        if (this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) // KioCG - drop picked up items when despawn
         this.persistenceRequired = true;
     }
 
@@ -811,6 +_,7 @@
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
+            this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity nearestPlayer = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
@@ -825,11 +_,13 @@
                 final double distanceSquared = dxSqr + dzSqr + dySqr;
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                    this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                  }
 
                 if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                        this.dropPickedUpItems(); // KioCG - drop picked up items when despawn
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     }
                 } else {
@@ -842,6 +_,16 @@
         }
     }
 
+    // KioCG start - drop picked up items when despawn
+    private void dropPickedUpItems() {
+        this.forceDrops = true; // CraftBukkit
+        this.dropPreservedEquipment(serverLevel(), itemStack -> {
+            return !EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP);
+        });
+        this.forceDrops = false; // CraftBukkit
+    }
+    // KioCG end - drop picked up items when despawn
+
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
@@ -1282,6 +_,7 @@
             );
         }
 
+        this.getAttribute(Attributes.SCALE).addOrUpdateTransientModifier(new AttributeModifier(ANIMATION_SCALE_MODIFIER_ID, -1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL)); // KioCG - mob spawning or dying animation
         this.setLeftHanded(random.nextFloat() < 0.05F);
         return spawnGroupData;
     }
@@ -1562,8 +_,23 @@
         return (this.entityData.get(DATA_MOB_FLAGS_ID) & 4) != 0;
     }
 
-    public void setBaby(boolean baby) {
-    }
+    // KioCG start - entity baby api
+    public final void setBaby(boolean baby) { // final防止重写, 便于版本更新检查
+        this.setBaby1(baby);
+    }
+
+    public void setBaby1(boolean baby) {
+    }
+
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> setBaby(true);
+            case HUGE -> setBaby(false);
+        }
+
+        super.setScaleSimulate(scale);
+    }
+    // KioCG end - entity baby api
 
     @Override
     public HumanoidArm getMainArm() {
@@ -1686,6 +_,58 @@
             this.getNavigation().updatePathfinderMaxVisitedNodes();
         }
     }
+
+    // KioCG start - entity bossbar api
+    @Nullable
+    public net.minecraft.server.level.ServerBossEvent getBossBar() {
+        return this.bossbar;
+    }
+
+    public void setBossBar(@Nullable net.minecraft.server.level.ServerBossEvent bossbar) {
+        if (this.bossbar != null) {
+            this.bossbar.removeAllPlayers();
+        }
+        this.bossbar = bossbar;
+        updatePlayers();
+    }
+
+    private void updatePlayers() {
+        java.util.Set<net.minecraft.server.level.ServerPlayer> set = com.google.common.collect.Sets.newHashSet();
+
+        net.minecraft.server.level.ChunkMap.TrackedEntity entityTracker = this.serverLevel().getChunkSource().chunkMap.entityMap.get(this.getId());
+        if (entityTracker != null) {
+            for (net.minecraft.server.network.ServerPlayerConnection connection : entityTracker.seenBy) {
+                net.minecraft.server.level.ServerPlayer serverPlayer = connection.getPlayer();
+                if (this.position().distanceToSqr(serverPlayer.position()) <= 16 * 16) {
+                    this.bossbar.addPlayer(serverPlayer);
+                    set.add(serverPlayer);
+                }
+            }
+        }
+
+        java.util.Set<net.minecraft.server.level.ServerPlayer> set1 = com.google.common.collect.Sets.newHashSet(this.bossbar.getPlayers());
+        set1.removeAll(set);
+
+        for (net.minecraft.server.level.ServerPlayer serverPlayer1 : set1) {
+            this.bossbar.removePlayer(serverPlayer1);
+        }
+    }
+
+    @Override
+    public void stopSeenByPlayer(net.minecraft.server.level.ServerPlayer player) {
+        super.stopSeenByPlayer(player);
+        if (this.bossbar != null) {
+            this.bossbar.removePlayer(player);
+        }
+    }
+    // KioCG end - entity bossbar api
+
+    // KioCG start - chunk hot
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && (this.isPersistenceRequired() || this.requiresCustomPersistence() || !this.removeWhenFarAway(0.0));
+    }
+    // KioCG end - chunk hot
 
     @VisibleForTesting
     public float[] getHandDropChances() {
