--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -60,11 +_,15 @@
         if (this.shouldStayCloseToLeashHolder() && !this.isPanicking()) {
             this.goalSelector.enableControlFlag(Goal.Flag.MOVE);
             float f = 2.0F;
-            float f1 = this.distanceTo(entity);
+            // KioCG start - optimize distanceTo
+            float f1 = (float) this.distanceToSqr(entity);
+            if (f1 <= 2.0F * 2.0F) return;
+            f1 = net.minecraft.util.Mth.sqrt(f1);
+            // KioCG end - optimize distanceTo
             Vec3 vec3 = new Vec3(entity.getX() - this.getX(), entity.getY() - this.getY(), entity.getZ() - this.getZ())
                 .normalize()
                 .scale(Math.max(f1 - 2.0F, 0.0F));
-            this.getNavigation().moveTo(this.getX() + vec3.x, this.getY() + vec3.y, this.getZ() + vec3.z, this.followLeashSpeed());
+            this.getNavigation().moveToLightweight(this.getX() + vec3.x, this.getY() + vec3.y, this.getZ() + vec3.z, this.followLeashSpeed()); // KioCG - optimize entity behavior
         }
     }
 
