--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -57,7 +_,7 @@
     @Override
     public void closeRangeLeashBehaviour(Entity entity) {
         super.closeRangeLeashBehaviour(entity);
-        if (this.shouldStayCloseToLeashHolder() && !this.isPanicking()) {
+        if (this.shouldStayCloseToLeashHolder() && !this.isPanicking() && !(entity instanceof net.minecraft.world.entity.decoration.LeashFenceKnotEntity)) { // KioCG - TODO 临时修复https://github.com/PaperMC/Paper/issues/8587
             this.goalSelector.enableControlFlag(Goal.Flag.MOVE);
             float f = 2.0F;
             float f1 = this.distanceTo(entity);
