--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -37,6 +_,8 @@
         TamableAnimal.class, EntityDataSerializers.OPTIONAL_LIVING_ENTITY_REFERENCE
     );
     private boolean orderedToSit = false;
+    private int ticketTicks; // KioCG - dont kill my pet
+    private boolean followOwner = true; // KioCG - pet follow owner api
 
     protected TamableAnimal(EntityType<? extends TamableAnimal> entityType, Level level) {
         super(entityType, level);
@@ -58,6 +_,7 @@
         }
 
         compound.putBoolean("Sitting", this.orderedToSit);
+        compound.putBoolean("KioCG.FollowOwner", this.followOwner); // KioCG - pet follow owner api
     }
 
     @Override
@@ -78,6 +_,7 @@
 
         this.orderedToSit = compound.getBooleanOr("Sitting", false);
         this.setInSittingPose(this.orderedToSit, false); // Paper - Add EntityToggleSitEvent
+        this.followOwner = compound.getBooleanOr("KioCG.FollowOwner", true); // KioCG - pet follow owner api
     }
 
     @Override
@@ -146,6 +_,7 @@
         if (applyTamingSideEffects) {
             this.applyTamingSideEffects();
         }
+        this.goalRegister.reassess(); // KioCG - optimize entity behavior
     }
 
     protected void applyTamingSideEffects() {
@@ -256,13 +_,86 @@
         super.die(cause);
     }
 
+    // KioCG start - dont kill my pet
+    @Override
+    public boolean nearDeath() {
+        return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this.isTame() && this.getHealth() == Float.MIN_VALUE;
+    }
+
+    @Override
+    public boolean canBeSeenAsEnemy() {
+        return !this.nearDeath() && super.canBeSeenAsEnemy();
+    }
+
+    @Override
+    public boolean hurtServer(ServerLevel level, DamageSource damageSource, float amount) {
+        return (!this.nearDeath() || amount == Float.MAX_VALUE || damageSource.is(net.minecraft.tags.DamageTypeTags.BYPASSES_INVULNERABILITY))
+               && super.hurtServer(level, damageSource, amount);
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+
+        if (this.nearDeath()) {
+            this.addNearDeathParticles();
+            if (!this.isOrderedToSit()) {
+                this.jumping = false;
+                this.navigation.stop();
+                this.setTarget(null);
+                this.setOrderedToSit(true);
+            }
+        }
+        this.keepChunkLoaded();
+    }
+
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        this.keepChunkLoaded();
+    }
+
+    public void addNearDeathParticles() {
+        if (this.tickCount % 5 == 0) {
+            this.serverLevel().sendParticles(ParticleTypes.SOUL,
+                                             this.getX(), this.getEyeY(), this.getZ(),
+                                             1,
+                                             this.getBbWidth() / 2, this.getBbHeight() / 2, this.getBbWidth() / 2,
+                                             0.0);
+        }
+    }
+
+    private void keepChunkLoaded() {
+        if (!io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petFollowKeepChunkLoaded) return;
+        if (this.isTame() && !this.unableToMoveToOwner()) {
+            if (this.tickCount >= ticketTicks) {
+                ticketTicks = this.tickCount + (int) net.minecraft.server.level.TicketType.PET_FOLLOW.timeout() - 1;
+                this.serverLevel().getChunkSource().addTicketWithRadius(net.minecraft.server.level.TicketType.PET_FOLLOW, this.chunkPosition(), 2);
+            }
+        }
+    }
+    // KioCG end - dont kill my pet
+
     public boolean isOrderedToSit() {
         return this.orderedToSit;
     }
 
     public void setOrderedToSit(boolean orderedToSit) {
+        if (!orderedToSit && this.nearDeath()) return; // KioCG - dont kill my pet
         this.orderedToSit = orderedToSit;
-    }
+        this.goalRegister.reassess(); // KioCG - optimize entity behavior
+    }
+
+    // KioCG start - pet follow owner api
+    public boolean isFollowOwner() {
+        return this.followOwner;
+    }
+
+    public void setFollowOwner(boolean followOwner) {
+        this.followOwner = followOwner;
+        this.goalRegister.reassess(); // KioCG - optimize entity behavior
+    }
+    // KioCG end - pet follow owner api
 
     public void tryToTeleportToOwner() {
         LivingEntity owner = this.getOwner();
@@ -273,15 +_,18 @@
 
     public boolean shouldTryTeleportToOwner() {
         LivingEntity owner = this.getOwner();
-        return owner != null && this.distanceToSqr(this.getOwner()) >= 144.0;
+        return owner != null && this.distanceToSqr(this.getOwner()) >= io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petTeleportDistanceSqr; // KioCG - pet teleport distance
     }
 
     private void teleportToAroundBlockPos(BlockPos pos) {
-        for (int i = 0; i < 10; i++) {
-            int randomInt = this.random.nextIntBetweenInclusive(-3, 3);
-            int randomInt1 = this.random.nextIntBetweenInclusive(-3, 3);
+        // KioCG start - pet teleport distance
+        if (!this.getOwner().serverLevel().isLoaded(pos)) return; // 防止区块未加载时传送到空中
+        /*for (int i = 0; i < 10; i++)*/ {
+            int randomInt = this.random.nextIntBetweenInclusive(-8, 8);
+            int randomInt1 = this.random.nextIntBetweenInclusive(-8, 8);
             if (Math.abs(randomInt) >= 2 || Math.abs(randomInt1) >= 2) {
-                int randomInt2 = this.random.nextIntBetweenInclusive(-1, 1);
+                int randomInt2 = this.random.nextIntBetweenInclusive(-2, 2);
+        // KioCG end - pet teleport distance
                 if (this.maybeTeleportTo(pos.getX() + randomInt, pos.getY() + randomInt2, pos.getZ() + randomInt1)) {
                     return;
                 }
@@ -322,6 +_,7 @@
     }
 
     public final boolean unableToMoveToOwner() {
+        if (!this.followOwner) return true; // KioCG - pet follow owner api
         return this.isOrderedToSit() || this.isPassenger() || this.mayBeLeashed() || this.getOwner() != null && this.getOwner().isSpectator();
     }
 
@@ -337,6 +_,13 @@
         public TamableAnimalPanicGoal(final double speedModifier) {
             super(TamableAnimal.this, speedModifier);
         }
+
+        // KioCG start - dont kill my pet
+        @Override
+        public boolean shouldPanic() {
+            return !this.mob.nearDeath() && super.shouldPanic();
+        }
+        // KioCG end - dont kill my pet
 
         @Override
         public void tick() {
