--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -213,7 +_,7 @@
 
                 furnace.litTimeRemaining = furnaceBurnEvent.getBurnTime(); // CraftBukkit - respect event output
                 furnace.litTotalTime = furnace.litTimeRemaining;
-                if (furnace.isLit() && furnaceBurnEvent.isBurning()) { // CraftBukkit - respect event output
+                if (false && furnace.isLit() && furnaceBurnEvent.isBurning()) { // CraftBukkit - respect event output // KioCG - change container behavior: 燃烧完成后再扣除燃料
                     flag = true;
                     if (flag2 && furnaceBurnEvent.willConsumeFuel()) { // Paper - add consumeFuel to FurnaceBurnEvent
                         Item item = itemStack.getItem();
@@ -247,10 +_,33 @@
                 if (furnace.cookingTimer >= furnace.cookingTotalTime) { // Paper - cook speed multiplier API
                     furnace.cookingTimer = 0;
                     furnace.cookingTotalTime = getTotalCookTime(level, furnace, furnace.recipeType, furnace.cookSpeedMultiplier); // Paper - cook speed multiplier API
-                    if (burn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize, level, furnace.worldPosition)) { // CraftBukkit
+                    // KioCG start - change container behavior
+                    boolean burned = false;
+                    while (burn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize, level, furnace.worldPosition)) { // CraftBukkit
+                        burned = true;
+                    // KioCG end - change container behavior
                         furnace.setRecipeUsed(recipeHolder);
                     }
 
+                    // KioCG start - change container behavior
+                    if (burned) {
+                        furnace.getRecipesToAwardAndPopExperience(level, Vec3.atCenterOf(pos));
+                        furnace.recipesUsed.clear();
+                        furnace.litTimeRemaining = 0;
+
+                        // 燃烧完成后再扣除燃料
+                        if (!itemStack.isEmpty()) {
+                            Item item = itemStack.getItem();
+                            itemStack.shrink(1);
+
+                            ItemStack craftingRemainder = item.getCraftingRemainder();
+                            if (!craftingRemainder.isEmpty()) {
+                                net.minecraft.world.level.block.Block.popResource(level, pos, craftingRemainder);
+                            }
+                        }
+                    }
+                    // KioCG end - change container behavior
+
                     flag = true;
                 }
             } else {
@@ -269,7 +_,31 @@
         if (flag) {
             setChanged(level, pos, state);
         }
-    }
+        else if (!furnace.isLit() && furnace.cookingTimer <= 0) furnace.updateBlockEntityTicker(); // KioCG - optimize block entity
+    }
+
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(net.minecraft.world.level.Level level, BlockPos pos, BlockState state) {
+        if (this.isLit() || this.cookingTimer > 0) {
+            return true;
+        } else {
+            ItemStack itemStack = this.items.get(1);
+            ItemStack itemStack1 = this.items.get(0);
+            boolean flag1 = !itemStack1.isEmpty();
+            boolean flag2 = !itemStack.isEmpty();
+            if (flag2 && flag1) {
+                SingleRecipeInput singleRecipeInput = new SingleRecipeInput(itemStack1);
+                RecipeHolder<? extends AbstractCookingRecipe> recipeHolder = this.quickCheck.getRecipeFor(singleRecipeInput, (ServerLevel) level).orElse(null);
+                int maxStackSize = this.getMaxStackSize();
+                if (canBurn(level.registryAccess(), recipeHolder, singleRecipeInput, this.items, maxStackSize)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // KioCG end - optimize block entity
 
     private static boolean canBurn(
         RegistryAccess registryAccess,
@@ -283,6 +_,13 @@
             if (itemStack.isEmpty()) {
                 return false;
             } else {
+                // KioCG start - vanilla recipe excludes model items
+                if (recipe.id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, items.get(SLOT_INPUT))) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
                 ItemStack itemStack1 = items.get(2);
                 return itemStack1.isEmpty()
                     || ItemStack.isSameItemSameComponents(itemStack1, itemStack)
@@ -325,6 +_,10 @@
             itemStack1 = result = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(apiResult);
 
             if (!result.isEmpty()) {
+                // KioCG start - change container behavior
+                net.minecraft.world.level.block.Block.popResource(level, blockPos, result.copy());
+            } else if (false) {
+                // KioCG end - change container behavior
                 if (existingResults.isEmpty()) {
                     items.set(2, result.copy());
                 } else if (org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(existingResults).isSimilar(apiResult)) {
@@ -344,7 +_,10 @@
             // CraftBukkit end
 
             if (itemStack.is(Blocks.WET_SPONGE.asItem()) && !items.get(1).isEmpty() && items.get(1).is(Items.BUCKET)) {
-                items.set(1, new ItemStack(Items.WATER_BUCKET));
+                // KioCG start - change container behavior
+                items.set(1, ItemStack.EMPTY);
+                net.minecraft.world.level.block.Block.popResource(level, blockPos, new ItemStack(Items.WATER_BUCKET));
+                // KioCG end - change container behavior
             }
 
             itemStack.shrink(1);
@@ -366,12 +_,14 @@
             ? furnace.quickCheck.getRecipeFor(singleRecipeInput, level).map(recipeHolder -> recipeHolder.value().cookingTime()).orElse(200)
             /* passing a null level here is safe. world is only used for map extending recipes which won't happen here */
             : (net.minecraft.server.MinecraftServer.getServer().getRecipeManager().getRecipeFor(recipeType, singleRecipeInput, level).map(recipeHolder -> recipeHolder.value().cookingTime()).orElse(200));
+        if (furnace instanceof BlastFurnaceBlockEntity || furnace instanceof SmokerBlockEntity) cookTime *= 2; // KioCG - change container behavior: 不要加速高炉和烟熏炉
         return (int) Math.ceil (cookTime / cookSpeedMultiplier);
         // Paper end - cook speed multiplier API
     }
 
     @Override
     public int[] getSlotsForFace(Direction side) {
+        if (true) return new int[0]; // KioCG - 阻止漏斗交互
         if (side == Direction.DOWN) {
             return SLOTS_FOR_DOWN;
         } else {
@@ -415,6 +_,7 @@
             this.cookingTimer = 0;
             this.setChanged();
         }
+        this.updateBlockEntityTicker(); // KioCG - optimize block entity
     }
 
     @Override
@@ -425,7 +_,7 @@
             return true;
         } else {
             ItemStack itemStack = this.items.get(1);
-            return this.level.fuelValues().isFuel(stack) || stack.is(Items.BUCKET) && !itemStack.is(Items.BUCKET);
+            return this.level.fuelValues().isFuel(stack) || stack.is(Items.BUCKET) && !itemStack.is(Items.BUCKET) && net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack); // KioCG - vanilla recipe excludes model items
         }
     }
 
