--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -146,6 +_,7 @@
     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
         super.loadAdditional(tag, registries);
         this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
+        this.shouldTick = true; // KioCG - optimize block entity
         ContainerHelper.loadAllItems(tag, this.items, registries);
         this.cookingTimer = tag.getShort("cooking_time_spent");
         this.cookingTotalTime = tag.getShort("cooking_total_time");
@@ -183,11 +_,19 @@
         tag.put("RecipesUsed", compoundTag);
     }
 
+    private boolean shouldTick = true; // KioCG - optimize block entity
     public static void serverTick(ServerLevel level, BlockPos pos, BlockState state, AbstractFurnaceBlockEntity furnace) {
+        // KioCG start - optimize block entity
+        if (!furnace.shouldTick && !furnace.isLit()) {
+            return;
+        }
+        furnace.shouldTick = false;
+        // KioCG end - optimize block entity
         boolean isLit = furnace.isLit();
         boolean flag = false;
         if (furnace.isLit()) {
             furnace.litTimeRemaining--;
+            furnace.shouldTick = true; // KioCG - optimize block entity
         }
 
         ItemStack itemStack = furnace.items.get(1);
@@ -263,6 +_,7 @@
             }
         } else if (!furnace.isLit() && furnace.cookingTimer > 0) {
             furnace.cookingTimer = Mth.clamp(furnace.cookingTimer - 2, 0, furnace.cookingTotalTime);
+            furnace.shouldTick = true; // KioCG - optimize block entity: 燃烧冷却中
         }
 
         if (isLit != furnace.isLit()) {
@@ -288,6 +_,13 @@
             if (itemStack.isEmpty()) {
                 return false;
             } else {
+                // KioCG start - vanilla recipe excludes model items
+                if (recipe.id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, items.get(0))) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
                 ItemStack itemStack1 = items.get(2);
                 return itemStack1.isEmpty()
                     || ItemStack.isSameItemSameComponents(itemStack1, itemStack)
@@ -420,6 +_,7 @@
             this.cookingTimer = 0;
             this.setChanged();
         }
+        this.shouldTick = true; // KioCG - optimize block entity
     }
 
     @Override
@@ -430,7 +_,7 @@
             return true;
         } else {
             ItemStack itemStack = this.items.get(1);
-            return this.level.fuelValues().isFuel(stack) || stack.is(Items.BUCKET) && !itemStack.is(Items.BUCKET);
+            return this.level.fuelValues().isFuel(stack) || stack.is(Items.BUCKET) && !itemStack.is(Items.BUCKET) && net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack); // KioCG - vanilla recipe excludes model items
         }
     }
 
