--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -83,6 +_,7 @@
     @Nullable
     public BlockPos savedFlowerPos;
     public int maxBees = MAX_OCCUPANTS; // CraftBukkit - allow setting max amount of bees a hive can hold
+    private int honeyDays; // KioCG - beehive refill every night
 
     public BeehiveBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.BEEHIVE, pos, blockState);
@@ -341,16 +_,69 @@
 
     public static void serverTick(Level level, BlockPos pos, BlockState state, BeehiveBlockEntity beehive) {
         tickOccupants(level, pos, state, beehive.stored, beehive.savedFlowerPos);
-        if (!beehive.stored.isEmpty() && level.getRandom().nextDouble() < 0.005) {
+        if (/*!beehive.stored.isEmpty() &&*/ level.getRandom().nextDouble() < 0.005 && calculateNewHoneyLevel(level, pos, 0, BeehiveBlock.MAX_HONEY_LEVELS) > 0) { // KioCG - beehive refill every night: 蜂巢始终工作!
             double d = pos.getX() + 0.5;
             double d1 = pos.getY();
             double d2 = pos.getZ() + 0.5;
             level.playSound(null, d, d1, d2, SoundEvents.BEEHIVE_WORK, SoundSource.BLOCKS, 1.0F, 1.0F);
         }
 
+        // KioCG start - beehive refill every night
+        if (state.is(BlockTags.BEEHIVES, blockStateBase -> blockStateBase.hasProperty(BeehiveBlock.HONEY_LEVEL))) {
+            int days = level.getDays();
+            if (days != beehive.honeyDays) {
+                int daysSpent = days - beehive.honeyDays;
+                beehive.honeyDays = days;
+
+                // 只在傍晚产生蜂蜜, 因此如果现在是白天则需要减少一天
+                if (level.isBrightOutside()) {
+                    --daysSpent;
+                    --beehive.honeyDays;
+                }
+
+                if (daysSpent > 0) {
+                    int honeyLevel = getHoneyLevel(state);
+                    if (honeyLevel < BeehiveBlock.MAX_HONEY_LEVELS) {
+                        int newHoneyLevel = calculateNewHoneyLevel(level, pos, honeyLevel, daysSpent);
+                        if (newHoneyLevel != honeyLevel) {
+                            level.setBlockAndUpdate(pos, state.setValue(BeehiveBlock.HONEY_LEVEL, newHoneyLevel));
+                        }
+                    }
+                }
+            }
+        }
+        // KioCG end - beehive refill every night
+
         DebugPackets.sendHiveInfo(level, pos, state, beehive);
     }
 
+    // KioCG start - beehive refill every night
+    private static int calculateNewHoneyLevel(Level level, BlockPos pos, int honeyLevel, int added) {
+        int xOff = pos.getX();
+        int yOff = pos.getY();
+        int zOff = pos.getZ();
+        for (int dz = -4; dz <= 4; ++dz) {
+            int z = dz + zOff;
+            for (int dx = -4; dx <= 4; ++dx) {
+                int x = xOff + dx;
+                for (int dy = -2; dy <= 0; ++dy) {
+                    int y = dy + yOff;
+                    net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(x >> 4, z >> 4);
+                    if (chunk == null) break; // 同一Y轴不需要再循环
+                    BlockState blockStateFinal = chunk.getBlockStateFinal(x, y, z);
+                    if (Bee.attractsBees(blockStateFinal)) {
+                        honeyLevel = honeyLevel + added;
+                        if (honeyLevel >= BeehiveBlock.MAX_HONEY_LEVELS) {
+                            return BeehiveBlock.MAX_HONEY_LEVELS;
+                        }
+                    }
+                }
+            }
+        }
+        return honeyLevel;
+    }
+    // KioCG end - beehive refill every night
+
     @Override
     protected void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
@@ -358,6 +_,7 @@
         input.read("bees", BeehiveBlockEntity.Occupant.LIST_CODEC).orElse(List.of()).forEach(this::storeBee);
         this.savedFlowerPos = input.read("flower_pos", BlockPos.CODEC).orElse(null);
         this.maxBees = input.getIntOr("Bukkit.MaxEntities", MAX_OCCUPANTS); // Paper - persist max bukkit occupants
+        this.honeyDays = input.getIntOr("KioCG.HoneyDays", 0); // KioCG - beehive refill every night
     }
 
     @Override
@@ -366,6 +_,7 @@
         output.store("bees", BeehiveBlockEntity.Occupant.LIST_CODEC, this.getBees());
         output.storeNullable("flower_pos", BlockPos.CODEC, this.savedFlowerPos);
         output.putInt("Bukkit.MaxEntities", this.maxBees); // Paper - persist max bukkit occupants
+        output.putInt("KioCG.HoneyDays", this.honeyDays); // KioCG - beehive refill every night
     }
 
     @Override
