--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -37,6 +_,7 @@
     // CraftBukkit start - data containers
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer;
+    public final java.util.Map<String, Object> temporaryMeta = new java.util.HashMap<>(); // KioCG - temporary meta
     // CraftBukkit end
     private static final Codec<BlockEntityType<?>> TYPE_CODEC = BuiltInRegistries.BLOCK_ENTITY_TYPE.byNameCodec();
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -66,6 +_,18 @@
         return this.type.isValid(state);
     }
 
+    // KioCG start - optimize block entity
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        return true;
+    }
+
+    public void updateBlockEntityTicker() {
+        if (this.hasLevel()) {
+            this.getLevel().getChunkAt(this.getBlockPos()).updateBlockEntityTicker(this);
+        }
+    }
+    // KioCG end - optimize block entity
+
     public static BlockPos getPosFromTag(ChunkPos chunkPos, CompoundTag tag) {
         int intOr = tag.getIntOr("x", 0);
         int intOr1 = tag.getIntOr("y", 0);
@@ -311,6 +_,7 @@
                 return dataComponentMap.getOrDefault(component, defaultValue);
             }
         });
+        this.updateBlockEntityTicker(); // KioCG - optimize block entity
         DataComponentPatch dataComponentPatch = patch.forget(set::contains);
         this.components = dataComponentPatch.split().added();
         // CraftBukkit start
