--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -26,6 +_,7 @@
 
 public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer {
     private static final int INGREDIENT_SLOT = 3;
+    private static final int[] INGREDIENT_SLOTS = {INGREDIENT_SLOT, 10, 11, 12, 13, 14, 15, 16, 17}; // KioCG - change container behavior
     private static final int FUEL_SLOT = 4;
     private static final int[] SLOTS_FOR_UP = new int[]{3};
     private static final int[] SLOTS_FOR_DOWN = new int[]{0, 1, 2, 3};
@@ -36,7 +_,7 @@
     public static final int NUM_DATA_VALUES = 2;
     private static final short DEFAULT_BREW_TIME = 0;
     private static final byte DEFAULT_FUEL = 0;
-    private NonNullList<ItemStack> items = NonNullList.withSize(5, ItemStack.EMPTY);
+    private NonNullList<ItemStack> items = NonNullList.withSize(18, ItemStack.EMPTY); // KioCG - change container behavior
     public int brewTime;
     public int recipeBrewTime = 400; // Paper - Add recipeBrewTime
     private boolean[] lastPotionCount;
@@ -135,8 +_,10 @@
     }
 
     public static void serverTick(Level level, BlockPos pos, BlockState state, BrewingStandBlockEntity blockEntity) {
+        boolean changed = false; // KioCG - optimize block entity
         ItemStack itemStack = blockEntity.items.get(4);
-        if (blockEntity.fuel <= 0 && itemStack.is(ItemTags.BREWING_FUEL)) {
+        blockEntity.fuel = 20; // KioCG - change container behavior: 总是最大燃料值
+        if (blockEntity.fuel <= 0 && PotionBrewing.isFuel(itemStack)) { // KioCG - vanilla recipe excludes model items: 提取方法
             // CraftBukkit start
             org.bukkit.event.inventory.BrewingStandFuelEvent event = new org.bukkit.event.inventory.BrewingStandFuelEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -153,6 +_,7 @@
             }
             // CraftBukkit end
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         }
 
         boolean isBrewable = isBrewable(level.potionBrewing(), blockEntity.items);
@@ -167,7 +_,22 @@
                 blockEntity.brewTime = 0;
             }
 
+            // KioCG start - change container behavior: 效果动画
+            if (level.getGameTime() % 5L == 0L) {
+                double x = (double) pos.getX() + 0.5;
+                double y = (double) pos.getY() + 0.85;
+                double z = (double) pos.getZ() + 0.5;
+                ((net.minecraft.server.level.ServerLevel) level).sendParticles(
+                        net.minecraft.core.particles.ColorParticleOption.create(net.minecraft.core.particles.ParticleTypes.ENTITY_EFFECT, level.random.nextFloat(), level.random.nextFloat(), level.random.nextFloat()),
+                        x, y, z,
+                        1,
+                        0.0, 0.0, 0.0,
+                        0.0);
+            }
+            // KioCG end - change container behavior: 效果动画
+
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         } else if (isBrewable && blockEntity.fuel > 0) {
             blockEntity.fuel--;
             // CraftBukkit start
@@ -180,8 +_,15 @@
             // CraftBukkit end
             blockEntity.ingredient = itemStack1.getItem();
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         }
 
+        // KioCG start - optimize block entity
+        if (!changed) blockEntity.updateBlockEntityTicker();
+        BrewingStandBlockEntity.refreshBottleDisplay(level, pos, state, blockEntity);
+    }
+    public static void refreshBottleDisplay(Level level, BlockPos pos, BlockState state, BrewingStandBlockEntity blockEntity) {
+        // KioCG end - optimize block entity
         boolean[] potionBits = blockEntity.getPotionBits();
         if (!Arrays.equals(potionBits, blockEntity.lastPotionCount)) {
             blockEntity.lastPotionCount = potionBits;
@@ -198,12 +_,42 @@
         }
     }
 
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        BrewingStandBlockEntity.refreshBottleDisplay(level, pos, state, this); // 没有很好的地方放置这个...
+
+        this.fuel = 20; // KioCG - change container behavior: 总是最大燃料值
+        if (this.brewTime > 0) {
+            return true;
+        } else if (this.fuel <= 0 && PotionBrewing.isFuel(this.items.get(FUEL_SLOT))) {
+            return true;
+        } else if (isBrewable(level.potionBrewing(), this.items) && this.fuel > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void setItem(int slot, ItemStack stack) {
+        super.setItem(slot, stack);
+        this.updateBlockEntityTicker();
+    }
+    // KioCG end - optimize block entity
+
     private boolean[] getPotionBits() {
         boolean[] flags = new boolean[3];
 
         for (int i = 0; i < 3; i++) {
             if (!this.items.get(i).isEmpty()) {
-                flags[i] = true;
+                // KioCG start - fix client display
+                flags[switch (i) {
+                    case 0 -> 2;
+                    case 1 -> 0;
+                    case 2 -> 1;
+                    default -> throw new IllegalStateException("Unexpected value: " + i);
+                }] = true;
+                // KioCG end - fix client display
             }
         }
 
@@ -239,6 +_,8 @@
             brewResults.add(i, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(potionBrewing.mix(itemStack, items.get(i))));
         }
 
+        ItemStack craftingRemainder = itemStack.getItem().getCraftingRemainder(); // KioCG - fix MC-259583: Glass bottle is lost after brewing the last Dragon's Breath
+
         if (owner != null) {
             org.bukkit.event.inventory.BrewEvent event = new org.bukkit.event.inventory.BrewEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -250,9 +_,19 @@
                 return;
             }
 
+            // KioCG start - change container behavior
+            itemStack.shrink(1);
+            boolean autoBrewing = tryFillNextIngredient(items);
+            // KioCG end - change container behavior
+
             for (int i = 0; i < 3; i++) {
                 if (i < brewResults.size()) {
                     items.set(i, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(brewResults.get(i)));
+                    // KioCG start - change container behavior
+                    if (!autoBrewing) {
+                        net.minecraft.world.level.block.Block.popResource(level, pos, items.set(i, ItemStack.EMPTY));
+                    }
+                    // KioCG end - change container behavior
                 } else {
                     items.set(i, ItemStack.EMPTY);
                 }
@@ -260,20 +_,37 @@
         }
         // CraftBukkit end
 
-        itemStack.shrink(1);
-        ItemStack craftingRemainder = itemStack.getItem().getCraftingRemainder();
+        // itemStack.shrink(1); // KioCG - change container behavior: 上移
+        // ItemStack craftingRemainder = itemStack.getItem().getCraftingRemainder(); // KioCG - fix MC-259583: 上移
         if (!craftingRemainder.isEmpty()) {
-            if (itemStack.isEmpty()) {
+            if (false && itemStack.isEmpty()) { // KioCG - change container behavior: 始终掉落玻璃瓶
                 itemStack = craftingRemainder;
             } else {
                 Containers.dropItemStack(level, pos.getX(), pos.getY(), pos.getZ(), craftingRemainder);
             }
         }
 
-        items.set(3, itemStack);
+        // items.set(3, itemStack); // KioCG - change container behavior
         level.levelEvent(1035, pos, 0);
     }
 
+    // KioCG start - change container behavior
+    private static boolean tryFillNextIngredient(NonNullList<ItemStack> items) {
+        if (items.get(INGREDIENT_SLOT).isEmpty()) {
+            int index = 0;
+            for (int i = 1; i < INGREDIENT_SLOTS.length; ++i) {
+                final ItemStack itemStack = items.get(INGREDIENT_SLOTS[i]);
+                if (!itemStack.isEmpty()) {
+                    items.set(INGREDIENT_SLOTS[index++], itemStack);
+                    items.set(INGREDIENT_SLOTS[i], ItemStack.EMPTY);
+                }
+            }
+            return index != 0;
+        }
+        return false;
+    }
+    // KioCG end - change container behavior
+
     @Override
     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
         super.loadAdditional(tag, registries);
@@ -302,14 +_,15 @@
             return potionBrewing.isIngredient(stack);
         } else {
             return index == 4
-                ? stack.is(ItemTags.BREWING_FUEL)
-                : (stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE) || potionBrewing.isCustomInput(stack)) // Paper - Custom Potion Mixes
+                ? PotionBrewing.isFuel(stack) // KioCG - vanilla recipe excludes model items: 提取方法
+                : potionBrewing.isPotion(stack) // KioCG - vanilla recipe excludes model items: 提取方法
                     && this.getItem(index).isEmpty();
         }
     }
 
     @Override
     public int[] getSlotsForFace(Direction side) {
+        if (true) return new int[0]; // KioCG - 阻止漏斗交互
         if (side == Direction.UP) {
             return SLOTS_FOR_UP;
         } else {
