--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -135,8 +_,9 @@
     }
 
     public static void serverTick(Level level, BlockPos pos, BlockState state, BrewingStandBlockEntity blockEntity) {
+        boolean changed = false; // KioCG - optimize block entity
         ItemStack itemStack = blockEntity.items.get(4);
-        if (blockEntity.fuel <= 0 && itemStack.is(ItemTags.BREWING_FUEL)) {
+        if (blockEntity.fuel <= 0 && PotionBrewing.isFuel(itemStack)) { // KioCG - vanilla recipe excludes model items: 提取方法
             // CraftBukkit start
             org.bukkit.event.inventory.BrewingStandFuelEvent event = new org.bukkit.event.inventory.BrewingStandFuelEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -153,6 +_,7 @@
             }
             // CraftBukkit end
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         }
 
         boolean isBrewable = isBrewable(level.potionBrewing(), blockEntity.items);
@@ -167,7 +_,22 @@
                 blockEntity.brewTime = 0;
             }
 
+            // KioCG start - change container behavior: 效果动画
+            if (level.getGameTime() % 5L == 0L) {
+                double x = (double) pos.getX() + 0.5;
+                double y = (double) pos.getY() + 0.85;
+                double z = (double) pos.getZ() + 0.5;
+                ((net.minecraft.server.level.ServerLevel) level).sendParticles(
+                        net.minecraft.core.particles.ColorParticleOption.create(net.minecraft.core.particles.ParticleTypes.ENTITY_EFFECT, level.random.nextFloat(), level.random.nextFloat(), level.random.nextFloat()),
+                        x, y, z,
+                        1,
+                        0.0, 0.0, 0.0,
+                        0.0);
+            }
+            // KioCG end - change container behavior: 效果动画
+
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         } else if (isBrewable && blockEntity.fuel > 0) {
             blockEntity.fuel--;
             // CraftBukkit start
@@ -180,8 +_,15 @@
             // CraftBukkit end
             blockEntity.ingredient = itemStack1.getItem();
             setChanged(level, pos, state);
+            changed = true; // KioCG - optimize block entity
         }
 
+        // KioCG start - optimize block entity
+        if (!changed) blockEntity.updateBlockEntityTicker();
+        else BrewingStandBlockEntity.refreshBottleDisplay(level, pos, state, blockEntity);
+    }
+    public static void refreshBottleDisplay(Level level, BlockPos pos, BlockState state, BrewingStandBlockEntity blockEntity) {
+        // KioCG end - optimize block entity
         boolean[] potionBits = blockEntity.getPotionBits();
         if (!Arrays.equals(potionBits, blockEntity.lastPotionCount)) {
             blockEntity.lastPotionCount = potionBits;
@@ -198,12 +_,41 @@
         }
     }
 
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        BrewingStandBlockEntity.refreshBottleDisplay(level, pos, state, this); // 没有很好的地方放置这个...
+
+        if (this.brewTime > 0) {
+            return true;
+        } else if (this.fuel <= 0 && PotionBrewing.isFuel(this.items.get(FUEL_SLOT))) {
+            return true;
+        } else if (isBrewable(level.potionBrewing(), this.items) && this.fuel > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void setItem(int slot, ItemStack stack) {
+        super.setItem(slot, stack);
+        this.updateBlockEntityTicker();
+    }
+    // KioCG end - optimize block entity
+
     private boolean[] getPotionBits() {
         boolean[] flags = new boolean[3];
 
         for (int i = 0; i < 3; i++) {
             if (!this.items.get(i).isEmpty()) {
-                flags[i] = true;
+                // KioCG start - fix client display
+                flags[switch (i) {
+                    case 0 -> 2;
+                    case 1 -> 0;
+                    case 2 -> 1;
+                    default -> throw new IllegalStateException("Unexpected value: " + i);
+                }] = true;
+                // KioCG end - fix client display
             }
         }
 
@@ -302,14 +_,15 @@
             return potionBrewing.isIngredient(stack);
         } else {
             return index == 4
-                ? stack.is(ItemTags.BREWING_FUEL)
-                : (stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE) || potionBrewing.isCustomInput(stack)) // Paper - Custom Potion Mixes
+                ? PotionBrewing.isFuel(stack) // KioCG - vanilla recipe excludes model items: 提取方法
+                : potionBrewing.isPotion(stack) // KioCG - vanilla recipe excludes model items: 提取方法
                     && this.getItem(index).isEmpty();
         }
     }
 
     @Override
     public int[] getSlotsForFace(Direction side) {
+        if (true) return new int[0]; // KioCG - 阻止漏斗交互
         if (side == Direction.UP) {
             return SLOTS_FOR_UP;
         } else {
