--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -129,7 +_,7 @@
 
     public static void serverTick(Level level, BlockPos pos, BlockState state, BrewingStandBlockEntity blockEntity) {
         ItemStack itemStack = blockEntity.items.get(4);
-        if (blockEntity.fuel <= 0 && itemStack.is(ItemTags.BREWING_FUEL)) {
+        if (blockEntity.fuel <= 0 && PotionBrewing.isFuel(itemStack)) { // KioCG - vanilla recipe excludes model items: 提取方法
             // CraftBukkit start
             org.bukkit.event.inventory.BrewingStandFuelEvent event = new org.bukkit.event.inventory.BrewingStandFuelEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -296,8 +_,8 @@
             return potionBrewing.isIngredient(stack);
         } else {
             return index == 4
-                ? stack.is(ItemTags.BREWING_FUEL)
-                : (stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE) || potionBrewing.isCustomInput(stack)) // Paper - Custom Potion Mixes
+                ? PotionBrewing.isFuel(stack) // KioCG - vanilla recipe excludes model items: 提取方法
+                : potionBrewing.isPotion(stack) // KioCG - vanilla recipe excludes model items: 提取方法
                     && this.getItem(index).isEmpty();
         }
     }
