--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -42,6 +_,7 @@
         super(BlockEntityType.CAMPFIRE, pos, blockState);
     }
 
+    public boolean[] shouldTick = {true, true, true, true}; // KioCG - optimize block entity
     public static void cookTick(
         ServerLevel level,
         BlockPos pos,
@@ -52,16 +_,26 @@
         boolean flag = false;
 
         for (int i = 0; i < campfire.items.size(); i++) {
+            // KioCG start - optimize block entity
+            if (!campfire.shouldTick[i]) {
+                return;
+            }
+            campfire.shouldTick[i] = false;
+            // KioCG end - optimize block entity
+
             ItemStack itemStack = campfire.items.get(i);
             if (!itemStack.isEmpty()) {
                 flag = true;
                 if (!campfire.stopCooking[i]) { // Paper - Add more Campfire API
                 campfire.cookingProgress[i]++;
+                campfire.shouldTick[i] = true; // KioCG - optimize block entity
                 } // Paper - Add more Campfire API
                 if (campfire.cookingProgress[i] >= campfire.cookingTime[i]) {
+                    campfire.shouldTick[i] = false; // KioCG - optimize block entity
                     SingleRecipeInput singleRecipeInput = new SingleRecipeInput(itemStack);
                     // Paper start - add recipe to cook events
                     final var optionalCookingRecipe = check.getRecipeFor(singleRecipeInput, level);
+                    if (optionalCookingRecipe.isEmpty()) continue; // KioCG - change container behavior
                     ItemStack itemStack1 = optionalCookingRecipe
                         .map(recipe -> recipe.value().assemble(singleRecipeInput, level.registryAccess()))
                         .orElse(itemStack);
@@ -77,14 +_,23 @@
                             result,
                             (org.bukkit.inventory.CookingRecipe<?>) optionalCookingRecipe.map(RecipeHolder::toBukkitRecipe).orElse(null) // Paper -Add recipe to cook events
                         );
+                        // KioCG start - vanilla recipe excludes model items
+                        if (optionalCookingRecipe.isPresent() && optionalCookingRecipe.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                            if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, itemStack)) {
+                                blockCookEvent.setCancelled(true);
+                            }
+                        }
+                        // KioCG end - vanilla recipe excludes model items
 
                         if (!blockCookEvent.callEvent()) {
-                            return;
+                            continue; // KioCG - fix loop when cancelled
                         }
 
                         result = blockCookEvent.getResult();
                         itemStack1 = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(result);
                         // CraftBukkit end
+                        // KioCG start - change container behavior
+                        /*
                         // Paper start - Fix item locations dropped from campfires
                         double deviation = 0.05F * RandomSource.GAUSSIAN_SPREAD_FACTOR;
                         while (!itemStack1.isEmpty()) {
@@ -93,7 +_,18 @@
                             level.addFreshEntity(droppedItem);
                         }
                         // Paper end - Fix item locations dropped from campfires
+                         */
+                        // KioCG end - change container behavior
+
                         campfire.items.set(i, ItemStack.EMPTY);
+                        // KioCG start - change container behavior
+                        if (!campfire.placeFood(level, null, itemStack1, i)) {
+                            campfire.items.set(i, itemStack1);
+                            campfire.cookingTime[i] = 0;
+                            campfire.cookingProgress[i] = 0;
+                        }
+                        // KioCG end - change container behavior
+
                         level.sendBlockUpdated(pos, state, state, 3);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
                     }
@@ -112,7 +_,7 @@
         for (int i = 0; i < blockEntity.items.size(); i++) {
             if (blockEntity.cookingProgress[i] > 0) {
                 flag = true;
-                blockEntity.cookingProgress[i] = Mth.clamp(blockEntity.cookingProgress[i] - 2, 0, blockEntity.cookingTime[i]);
+                blockEntity.cookingProgress[i] = 0; // KioCG - optimize block entity
             }
         }
 
@@ -154,6 +_,7 @@
     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
         super.loadAdditional(tag, registries);
         this.items.clear();
+        this.shouldTick = new boolean[]{true, true, true, true}; // KioCG - optimize block entity
         ContainerHelper.loadAllItems(tag, this.items, registries);
         if (tag.contains("CookingTimes", 11)) {
             int[] intArray = tag.getIntArray("CookingTimes");
@@ -205,7 +_,12 @@
     }
 
     public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) {
-        for (int i = 0; i < this.items.size(); i++) {
+        // KioCG start - change container behavior
+        return placeFood(level, entity, stack, 0);
+    }
+    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack, int startSlot) {
+        for (int i = startSlot; i < this.items.size(); i++) {
+        // KioCG end - change container behavior
             ItemStack itemStack = this.items.get(i);
             if (itemStack.isEmpty()) {
                 Optional<RecipeHolder<CampfireCookingRecipe>> recipeFor = level.recipeAccess()
@@ -213,6 +_,13 @@
                 if (recipeFor.isEmpty()) {
                     return false;
                 }
+                // KioCG start - vanilla recipe excludes model items
+                if (recipeFor.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack)) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
 
                 // CraftBukkit start
                 org.bukkit.event.block.CampfireStartEvent event = new org.bukkit.event.block.CampfireStartEvent(
@@ -225,6 +_,7 @@
                 // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, stack.consumeAndReturn(1, entity));
+                this.shouldTick[i] = true; // KioCG - optimize block entity
                 level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(entity, this.getBlockState()));
                 this.markUpdated();
                 return true;
@@ -247,6 +_,7 @@
     public void dowse() {
         if (this.level != null) {
             this.markUpdated();
+            cooldownTick(this.level, this.getBlockPos(), this.getBlockState(), this); // KioCG - optimize block entity
         }
     }
 
