--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -37,11 +_,33 @@
     public final int[] cookingProgress = new int[4];
     public final int[] cookingTime = new int[4];
     public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
+    private int cauldronTemperature = 0; // KioCG - campfire boiling water
+
+    // KioCG start - campfire healing
+    private int healingTime;
+    private static final FuelValues fuelValues;
+
+    static {
+        int smeltingTime = 200;
+        fuelValues = new FuelValues.Builder(net.minecraft.server.MinecraftServer.getServer().registryAccess(), net.minecraft.server.MinecraftServer.getServer().getWorldData().enabledFeatures())
+                .add(net.minecraft.tags.ItemTags.LOGS, smeltingTime * 3 / 2)
+                .add(net.minecraft.tags.ItemTags.BAMBOO_BLOCKS, smeltingTime * 3 / 2)
+                .add(net.minecraft.tags.ItemTags.PLANKS, smeltingTime * 3 / 2)
+                .add(net.minecraft.world.level.block.Blocks.BAMBOO_MOSAIC, smeltingTime * 3 / 2)
+                .add(net.minecraft.world.item.Items.STICK, smeltingTime / 2)
+                .add(net.minecraft.world.level.block.Blocks.BAMBOO, smeltingTime / 4)
+                .add(net.minecraft.world.level.block.Blocks.DEAD_BUSH, smeltingTime / 2)
+                .add(net.minecraft.world.level.block.Blocks.MANGROVE_ROOTS, smeltingTime * 3 / 2)
+                .remove(net.minecraft.tags.ItemTags.NON_FLAMMABLE_WOOD)
+                .build();
+    }
+    // KioCG end - campfire healing
 
     public CampfireBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.CAMPFIRE, pos, blockState);
     }
 
+    public boolean[] shouldTick = {true, true, true, true}; // KioCG - optimize block entity
     public static void cookTick(
         ServerLevel level,
         BlockPos pos,
@@ -52,16 +_,26 @@
         boolean flag = false;
 
         for (int i = 0; i < campfire.items.size(); i++) {
+            // KioCG start - optimize block entity
+            if (!campfire.shouldTick[i]) {
+                return;
+            }
+            campfire.shouldTick[i] = false;
+            // KioCG end - optimize block entity
+
             ItemStack itemStack = campfire.items.get(i);
             if (!itemStack.isEmpty()) {
                 flag = true;
                 if (!campfire.stopCooking[i]) { // Paper - Add more Campfire API
                 campfire.cookingProgress[i]++;
+                campfire.shouldTick[i] = true; // KioCG - optimize block entity
                 } // Paper - Add more Campfire API
                 if (campfire.cookingProgress[i] >= campfire.cookingTime[i]) {
+                    campfire.shouldTick[i] = false; // KioCG - optimize block entity
                     SingleRecipeInput singleRecipeInput = new SingleRecipeInput(itemStack);
                     // Paper start - add recipe to cook events
                     final var optionalCookingRecipe = check.getRecipeFor(singleRecipeInput, level);
+                    if (optionalCookingRecipe.isEmpty()) continue; // KioCG - change container behavior
                     ItemStack itemStack1 = optionalCookingRecipe
                         .map(recipe -> recipe.value().assemble(singleRecipeInput, level.registryAccess()))
                         .orElse(itemStack);
@@ -77,14 +_,23 @@
                             result,
                             (org.bukkit.inventory.CookingRecipe<?>) optionalCookingRecipe.map(RecipeHolder::toBukkitRecipe).orElse(null) // Paper -Add recipe to cook events
                         );
+                        // KioCG start - vanilla recipe excludes model items
+                        if (optionalCookingRecipe.isPresent() && optionalCookingRecipe.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                            if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, itemStack)) {
+                                blockCookEvent.setCancelled(true);
+                            }
+                        }
+                        // KioCG end - vanilla recipe excludes model items
 
                         if (!blockCookEvent.callEvent()) {
-                            return;
+                            continue; // KioCG - fix loop when cancelled
                         }
 
                         result = blockCookEvent.getResult();
                         itemStack1 = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(result);
                         // CraftBukkit end
+                        // KioCG start - change container behavior
+                        /*
                         // Paper start - Fix item locations dropped from campfires
                         double deviation = 0.05F * RandomSource.GAUSSIAN_SPREAD_FACTOR;
                         while (!itemStack1.isEmpty()) {
@@ -93,7 +_,18 @@
                             level.addFreshEntity(droppedItem);
                         }
                         // Paper end - Fix item locations dropped from campfires
+                         */
+                        // KioCG end - change container behavior
+
                         campfire.items.set(i, ItemStack.EMPTY);
+                        // KioCG start - change container behavior
+                        if (!campfire.placeFood(level, null, itemStack1, i)) {
+                            campfire.items.set(i, itemStack1);
+                            campfire.cookingTime[i] = 0;
+                            campfire.cookingProgress[i] = 0;
+                        }
+                        // KioCG end - change container behavior
+
                         level.sendBlockUpdated(pos, state, state, 3);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
                     }
@@ -101,10 +_,82 @@
             }
         }
 
+        // KioCG start - campfire healing
+        if (campfire.healingTime > 0) {
+            --campfire.healingTime;
+            if (level.getGameTime() % 20L == 0L) {
+                CampfireBlockEntity.healingNearbyPlayers(level, pos);
+            }
+            if (level.getGameTime() % 5L == 0L) {
+                double x = (double) pos.getX() + 0.5;
+                double y = (double) pos.getY() + 1.0;
+                double z = (double) pos.getZ() + 0.5;
+                level.sendParticles(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.REGENERATION).getParticleOptions(),
+                                    x, y, z,
+                                    2,
+                                    0.25, 0.1, 0.25,
+                                    1.0);
+            }
+        }
+        // KioCG end - campfire healing
+
+        // KioCG start - campfire boiling water
+        BlockPos above = pos.above();
+        BlockState aboveState = level.getBlockState(above);
+        if (aboveState.is(net.minecraft.world.level.block.Blocks.WATER_CAULDRON)) {
+            if (!campfire.isCauldronBoiling()) {
+                int cauldronLevel = aboveState.getValue(net.minecraft.world.level.block.LayeredCauldronBlock.LEVEL);
+                campfire.cauldronTemperature = Math.min(campfire.cauldronTemperature + switch (cauldronLevel) {
+                    case 1 -> 3;
+                    case 2 -> 2;
+                    case 3 -> 1;
+                    default -> throw new IllegalStateException("Unexpected value: " + cauldronLevel);
+                }, level.paperConfig().kiocgConfig.theLongDark.campfireBoilingWaterTime);
+                flag = true;
+            } else {
+                if (level.getRandom().nextDouble() < 0.1) {
+                    level.playSound(null, above, net.minecraft.sounds.SoundEvents.POINTED_DRIPSTONE_DRIP_WATER_INTO_CAULDRON, net.minecraft.sounds.SoundSource.BLOCKS, 1.0F, 0.1F);
+                }
+
+                double x = (double) above.getX() + 0.5;
+                double y = (double) above.getY() + 1.0;
+                double z = (double) above.getZ() + 0.5;
+                level.sendParticles(ParticleTypes.FALLING_WATER, x, y, z, 10, 0.25, 0.1, 0.25, 1.0);
+            }
+        } else if (campfire.cauldronTemperature > 0) {
+            campfire.cauldronTemperature = 0;
+            flag = true;
+        }
+        // KioCG end - campfire boiling water
+
         if (flag) {
             setChanged(level, pos, state);
         }
     }
+
+    // KioCG start - campfire healing
+    private static void healingNearbyPlayers(Level level, BlockPos pos) {
+        net.minecraft.world.phys.AABB aabb = (new net.minecraft.world.phys.AABB(pos)).inflate(10, 5, 10);
+        java.util.List<net.minecraft.world.entity.player.Player> players = level.getEntitiesOfClass(net.minecraft.world.entity.player.Player.class, aabb);
+        for (net.minecraft.world.entity.player.Player player : players) {
+            player.campfireHealing();
+        }
+    }
+
+    public static void entityInside(Level level, BlockPos pos, BlockState state, net.minecraft.world.entity.Entity entity, CampfireBlockEntity blockEntity) {
+        if (entity instanceof net.minecraft.world.entity.item.ItemEntity itemEntity) {
+            ItemStack item = itemEntity.getItem();
+            int burnDuration = CampfireBlockEntity.fuelValues.burnDuration(item);
+            if (burnDuration > 0) {
+                blockEntity.healingTime = Math.min(blockEntity.healingTime + burnDuration * item.getCount(), 20 * 60);
+                itemEntity.discard();
+
+                RandomSource random = level.getRandom();
+                level.playSound(null, pos, net.minecraft.sounds.SoundEvents.FIRECHARGE_USE, net.minecraft.sounds.SoundSource.BLOCKS, 1.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F);
+            }
+        }
+    }
+    // KioCG end - campfire healing
 
     public static void cooldownTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
         boolean flag = false;
@@ -112,14 +_,35 @@
         for (int i = 0; i < blockEntity.items.size(); i++) {
             if (blockEntity.cookingProgress[i] > 0) {
                 flag = true;
-                blockEntity.cookingProgress[i] = Mth.clamp(blockEntity.cookingProgress[i] - 2, 0, blockEntity.cookingTime[i]);
+                blockEntity.cookingProgress[i] = 0; // KioCG - optimize block entity
             }
         }
 
+        // KioCG start - campfire boiling water
+        if (blockEntity.cauldronTemperature > 0) {
+            flag = true;
+            // blockEntity.cauldronTemperature = Math.max(0, blockEntity.cauldronTemperature - 2);
+            blockEntity.cauldronTemperature = 0; // KioCG - optimize block entity
+        }
+        // KioCG end - campfire boiling water
+
         if (flag) {
             setChanged(level, pos, state);
         }
     }
+
+    // KioCG start - campfire boiling water
+    public boolean isCauldronBoiling() {
+        return this.cauldronTemperature >= this.level.paperConfig().kiocgConfig.theLongDark.campfireBoilingWaterTime;
+    }
+
+    public void cauldronLevelChanged(int oldLevel, int newLevel) {
+        if (newLevel > oldLevel) {
+            this.cauldronTemperature = this.cauldronTemperature * oldLevel / newLevel;
+            this.setChanged();
+        }
+    }
+    // KioCG end - campfire boiling water
 
     public static void particleTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
         RandomSource randomSource = level.random;
@@ -154,6 +_,7 @@
     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
         super.loadAdditional(tag, registries);
         this.items.clear();
+        this.shouldTick = new boolean[]{true, true, true, true}; // KioCG - optimize block entity
         ContainerHelper.loadAllItems(tag, this.items, registries);
         if (tag.contains("CookingTimes", 11)) {
             int[] intArray = tag.getIntArray("CookingTimes");
@@ -175,6 +_,8 @@
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, abyte.length));
         }
         // Paper end - Add more Campfire API
+
+        this.cauldronTemperature = tag.getInt("KioCG.CauldronTemperature"); // KioCG - campfire boiling water
     }
 
     @Override
@@ -190,6 +_,8 @@
         }
         tag.putByteArray("Paper.StopCooking", cookingState);
         // Paper end - Add more Campfire API
+
+        tag.putInt("KioCG.CauldronTemperature", this.cauldronTemperature); // KioCG - campfire boiling water
     }
 
     @Override
@@ -205,7 +_,12 @@
     }
 
     public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) {
-        for (int i = 0; i < this.items.size(); i++) {
+        // KioCG start - change container behavior
+        return placeFood(level, entity, stack, 0);
+    }
+    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack, int startSlot) {
+        for (int i = startSlot; i < this.items.size(); i++) {
+        // KioCG end - change container behavior
             ItemStack itemStack = this.items.get(i);
             if (itemStack.isEmpty()) {
                 Optional<RecipeHolder<CampfireCookingRecipe>> recipeFor = level.recipeAccess()
@@ -213,6 +_,13 @@
                 if (recipeFor.isEmpty()) {
                     return false;
                 }
+                // KioCG start - vanilla recipe excludes model items
+                if (recipeFor.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack)) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
 
                 // CraftBukkit start
                 org.bukkit.event.block.CampfireStartEvent event = new org.bukkit.event.block.CampfireStartEvent(
@@ -225,6 +_,7 @@
                 // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, stack.consumeAndReturn(1, entity));
+                this.shouldTick[i] = true; // KioCG - optimize block entity
                 level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(entity, this.getBlockState()));
                 this.markUpdated();
                 return true;
@@ -247,6 +_,7 @@
     public void dowse() {
         if (this.level != null) {
             this.markUpdated();
+            cooldownTick(this.level, this.getBlockPos(), this.getBlockState(), this); // KioCG - optimize block entity
         }
     }
 
