--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -47,6 +_,26 @@
     public final int[] cookingTime = new int[4];
     public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
 
+    // KioCG start - campfire healing
+    private int fiercelyTime;
+    private static final FuelValues fuelValues;
+
+    static {
+        int smeltingTime = 200;
+        fuelValues = new FuelValues.Builder(net.minecraft.server.MinecraftServer.getServer().registryAccess(), net.minecraft.server.MinecraftServer.getServer().getWorldData().enabledFeatures())
+                .add(net.minecraft.tags.ItemTags.LOGS, smeltingTime * 3 / 2)
+                .add(net.minecraft.tags.ItemTags.BAMBOO_BLOCKS, smeltingTime * 3 / 2)
+                .add(net.minecraft.tags.ItemTags.PLANKS, smeltingTime * 3 / 2)
+                .add(net.minecraft.world.level.block.Blocks.BAMBOO_MOSAIC, smeltingTime * 3 / 2)
+                .add(net.minecraft.world.item.Items.STICK, smeltingTime / 2)
+                .add(net.minecraft.world.level.block.Blocks.BAMBOO, smeltingTime / 4)
+                .add(net.minecraft.world.level.block.Blocks.DEAD_BUSH, smeltingTime / 2)
+                .add(net.minecraft.world.level.block.Blocks.MANGROVE_ROOTS, smeltingTime * 3 / 2)
+                .remove(net.minecraft.tags.ItemTags.NON_FLAMMABLE_WOOD)
+                .build();
+    }
+    // KioCG end - campfire healing
+
     public CampfireBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.CAMPFIRE, pos, blockState);
     }
@@ -63,14 +_,20 @@
         for (int i = 0; i < campfire.items.size(); i++) {
             ItemStack itemStack = campfire.items.get(i);
             if (!itemStack.isEmpty()) {
-                flag = true;
+                // flag = true; // KioCG start - optimize block entity
                 if (!campfire.stopCooking[i]) { // Paper - Add more Campfire API
-                campfire.cookingProgress[i]++;
+                    // KioCG start - optimize block entity
+                    if (campfire.cookingProgress[i] < campfire.cookingTime[i]) {
+                        campfire.cookingProgress[i] += campfire.fiercelyTime > 0 ? 2 : 1; // KioCG start - campfire healing: 加速食物烹饪
+                        flag = true;
+                    }
+                    // KioCG end - optimize block entity
                 } // Paper - Add more Campfire API
                 if (campfire.cookingProgress[i] >= campfire.cookingTime[i]) {
                     SingleRecipeInput singleRecipeInput = new SingleRecipeInput(itemStack);
                     // Paper start - add recipe to cook events
                     final var optionalCookingRecipe = check.getRecipeFor(singleRecipeInput, level);
+                    if (optionalCookingRecipe.isEmpty()) continue; // KioCG - change container behavior
                     ItemStack itemStack1 = optionalCookingRecipe
                         .map(recipe -> recipe.value().assemble(singleRecipeInput, level.registryAccess()))
                         .orElse(itemStack);
@@ -86,14 +_,23 @@
                             result,
                             (org.bukkit.inventory.CookingRecipe<?>) optionalCookingRecipe.map(RecipeHolder::toBukkitRecipe).orElse(null) // Paper -Add recipe to cook events
                         );
+                        // KioCG start - vanilla recipe excludes model items
+                        if (optionalCookingRecipe.isPresent() && optionalCookingRecipe.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                            if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, itemStack)) {
+                                blockCookEvent.setCancelled(true);
+                            }
+                        }
+                        // KioCG end - vanilla recipe excludes model items
 
                         if (!blockCookEvent.callEvent()) {
-                            return;
+                            continue; // KioCG - fix loop when cancelled
                         }
 
                         result = blockCookEvent.getResult();
                         itemStack1 = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(result);
                         // CraftBukkit end
+                        // KioCG start - change container behavior
+                        /*
                         // Paper start - Fix item locations dropped from campfires
                         double deviation = 0.05F * RandomSource.GAUSSIAN_SPREAD_FACTOR;
                         while (!itemStack1.isEmpty()) {
@@ -102,7 +_,18 @@
                             level.addFreshEntity(droppedItem);
                         }
                         // Paper end - Fix item locations dropped from campfires
+                         */
+                        // KioCG end - change container behavior
+
                         campfire.items.set(i, ItemStack.EMPTY);
+                        // KioCG start - change container behavior
+                        if (!campfire.placeFood(level, null, itemStack1, i)) {
+                            campfire.items.set(i, itemStack1);
+                            campfire.cookingTime[i] = 0;
+                            campfire.cookingProgress[i] = 0;
+                        }
+                        // KioCG end - change container behavior
+
                         level.sendBlockUpdated(pos, state, state, 3);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
                     }
@@ -110,10 +_,91 @@
             }
         }
 
+        // KioCG start - campfire healing
+        if (campfire.fiercelyTime > 0) {
+            --campfire.fiercelyTime;
+            if (level.getGameTime() % 20L == 0L) {
+                CampfireBlockEntity.healingNearbyPlayers(level, pos);
+            }
+            if (level.getGameTime() % 5L == 0L) {
+                double x = (double) pos.getX() + 0.5;
+                double y = (double) pos.getY() + 0.5;
+                double z = (double) pos.getZ() + 0.5;
+                double percent = campfire.fiercelyTime / (20.0 * 60.0);
+                int count = (int) (percent * 30);
+                double offset = percent * 0.25;
+                level.sendParticles(ParticleTypes.FLAME,
+                                    x, y, z,
+                                    count,
+                                    offset, offset * 2.0, offset,
+                                    0.0);
+                if (level.random.nextFloat() < percent) {
+                    level.playSound(null, x, y, z,
+                                    net.minecraft.sounds.SoundEvents.CAMPFIRE_CRACKLE,
+                                    net.minecraft.sounds.SoundSource.BLOCKS,
+                                    0.5F + level.random.nextFloat(),
+                                    level.random.nextFloat() * 0.7F + 0.6F);
+                }
+            }
+        }
+        // KioCG end - campfire healing
+
         if (flag) {
             setChanged(level, pos, state);
         }
-    }
+        else campfire.updateBlockEntityTicker(); // KioCG - optimize block entity
+    }
+
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        if (state.getValue(CampfireBlock.LIT)) {
+            if (this.fiercelyTime > 0) return true; // KioCG - campfire healing
+            for (int i = 0; i < this.items.size(); i++) {
+                ItemStack itemStack = this.items.get(i);
+                if (!itemStack.isEmpty()) {
+                    if (!this.stopCooking[i]) {
+                        if (this.cookingProgress[i] < this.cookingTime[i]) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        } else {
+            for (int progress : this.cookingProgress) {
+                if (progress > 0) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // KioCG end - optimize block entity
+
+    // KioCG start - campfire healing
+    private static void healingNearbyPlayers(Level level, BlockPos pos) {
+        net.minecraft.world.phys.AABB aabb = (new net.minecraft.world.phys.AABB(pos)).inflate(10, 5, 10);
+        java.util.List<net.minecraft.world.entity.player.Player> players = level.getEntitiesOfClass(net.minecraft.world.entity.player.Player.class, aabb);
+        for (net.minecraft.world.entity.player.Player player : players) {
+            player.campfireHealing();
+        }
+    }
+
+    public static void entityInside(Level level, BlockPos pos, BlockState state, net.minecraft.world.entity.Entity entity, CampfireBlockEntity blockEntity) {
+        if (entity instanceof net.minecraft.world.entity.item.ItemEntity itemEntity) {
+            ItemStack item = itemEntity.getItem();
+            int burnDuration = CampfireBlockEntity.fuelValues.burnDuration(item);
+            if (burnDuration > 0) {
+                blockEntity.fiercelyTime = Math.min(blockEntity.fiercelyTime + burnDuration * item.getCount(), 20 * 60);
+                blockEntity.updateBlockEntityTicker(); // KioCG - optimize block entity
+                itemEntity.discard();
+
+                RandomSource random = level.getRandom();
+                level.playSound(null, pos, net.minecraft.sounds.SoundEvents.FIRECHARGE_USE, net.minecraft.sounds.SoundSource.BLOCKS, 1.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F);
+            }
+        }
+    }
+    // KioCG end - campfire healing
 
     public static void cooldownTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
         boolean flag = false;
@@ -124,10 +_,12 @@
                 blockEntity.cookingProgress[i] = Mth.clamp(blockEntity.cookingProgress[i] - 2, 0, blockEntity.cookingTime[i]);
             }
         }
+        blockEntity.fiercelyTime = 0; // KioCG - campfire healing
 
         if (flag) {
             setChanged(level, pos, state);
         }
+        else blockEntity.updateBlockEntityTicker(); // KioCG - optimize block entity
     }
 
     public static void particleTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
@@ -217,8 +_,13 @@
         return var4;
     }
 
-    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) {
-        for (int i = 0; i < this.items.size(); i++) {
+    public final boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) { // KiCG - final防止重写
+        // KioCG start - change container behavior
+        return placeFood(level, entity, stack, 0);
+    }
+    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack, int startSlot) {
+        for (int i = startSlot; i < this.items.size(); i++) {
+        // KioCG end - change container behavior
             ItemStack itemStack = this.items.get(i);
             if (itemStack.isEmpty()) {
                 Optional<RecipeHolder<CampfireCookingRecipe>> recipeFor = level.recipeAccess()
@@ -226,6 +_,13 @@
                 if (recipeFor.isEmpty()) {
                     return false;
                 }
+                // KioCG start - vanilla recipe excludes model items
+                if (recipeFor.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack)) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
 
                 // CraftBukkit start
                 org.bukkit.event.block.CampfireStartEvent event = new org.bukkit.event.block.CampfireStartEvent(
@@ -238,6 +_,7 @@
                 // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, stack.consumeAndReturn(1, entity));
+                this.updateBlockEntityTicker(); // KioCG - optimize block entity
                 level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(entity, this.getBlockState()));
                 this.markUpdated();
                 return true;
