--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -63,14 +_,20 @@
         for (int i = 0; i < campfire.items.size(); i++) {
             ItemStack itemStack = campfire.items.get(i);
             if (!itemStack.isEmpty()) {
-                flag = true;
+                // flag = true; // KioCG start - optimize block entity
                 if (!campfire.stopCooking[i]) { // Paper - Add more Campfire API
-                campfire.cookingProgress[i]++;
+                    // KioCG start - optimize block entity
+                    if (campfire.cookingProgress[i] < campfire.cookingTime[i]) {
+                        campfire.cookingProgress[i]++;
+                        flag = true;
+                    }
+                    // KioCG end - optimize block entity
                 } // Paper - Add more Campfire API
                 if (campfire.cookingProgress[i] >= campfire.cookingTime[i]) {
                     SingleRecipeInput singleRecipeInput = new SingleRecipeInput(itemStack);
                     // Paper start - add recipe to cook events
                     final var optionalCookingRecipe = check.getRecipeFor(singleRecipeInput, level);
+                    if (optionalCookingRecipe.isEmpty()) continue; // KioCG - change container behavior
                     ItemStack itemStack1 = optionalCookingRecipe
                         .map(recipe -> recipe.value().assemble(singleRecipeInput, level.registryAccess()))
                         .orElse(itemStack);
@@ -86,14 +_,23 @@
                             result,
                             (org.bukkit.inventory.CookingRecipe<?>) optionalCookingRecipe.map(RecipeHolder::toBukkitRecipe).orElse(null) // Paper -Add recipe to cook events
                         );
+                        // KioCG start - vanilla recipe excludes model items
+                        if (optionalCookingRecipe.isPresent() && optionalCookingRecipe.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                            if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, itemStack)) {
+                                blockCookEvent.setCancelled(true);
+                            }
+                        }
+                        // KioCG end - vanilla recipe excludes model items
 
                         if (!blockCookEvent.callEvent()) {
-                            return;
+                            continue; // KioCG - fix loop when cancelled
                         }
 
                         result = blockCookEvent.getResult();
                         itemStack1 = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(result);
                         // CraftBukkit end
+                        // KioCG start - change container behavior
+                        /*
                         // Paper start - Fix item locations dropped from campfires
                         double deviation = 0.05F * RandomSource.GAUSSIAN_SPREAD_FACTOR;
                         while (!itemStack1.isEmpty()) {
@@ -102,7 +_,18 @@
                             level.addFreshEntity(droppedItem);
                         }
                         // Paper end - Fix item locations dropped from campfires
+                         */
+                        // KioCG end - change container behavior
+
                         campfire.items.set(i, ItemStack.EMPTY);
+                        // KioCG start - change container behavior
+                        if (!campfire.placeFood(level, null, itemStack1, i)) {
+                            campfire.items.set(i, itemStack1);
+                            campfire.cookingTime[i] = 0;
+                            campfire.cookingProgress[i] = 0;
+                        }
+                        // KioCG end - change container behavior
+
                         level.sendBlockUpdated(pos, state, state, 3);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
                     }
@@ -113,7 +_,33 @@
         if (flag) {
             setChanged(level, pos, state);
         }
-    }
+        else campfire.updateBlockEntityTicker(); // KioCG - optimize block entity
+    }
+
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        if (state.getValue(CampfireBlock.LIT)) {
+            for (int i = 0; i < this.items.size(); i++) {
+                ItemStack itemStack = this.items.get(i);
+                if (!itemStack.isEmpty()) {
+                    if (!this.stopCooking[i]) {
+                        if (this.cookingProgress[i] < this.cookingTime[i]) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        } else {
+            for (int progress : this.cookingProgress) {
+                if (progress > 0) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // KioCG end - optimize block entity
 
     public static void cooldownTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
         boolean flag = false;
@@ -128,6 +_,7 @@
         if (flag) {
             setChanged(level, pos, state);
         }
+        else blockEntity.updateBlockEntityTicker(); // KioCG - optimize block entity
     }
 
     public static void particleTick(Level level, BlockPos pos, BlockState state, CampfireBlockEntity blockEntity) {
@@ -217,8 +_,13 @@
         return var4;
     }
 
-    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) {
-        for (int i = 0; i < this.items.size(); i++) {
+    public final boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack) { // KiCG - final防止重写
+        // KioCG start - change container behavior
+        return placeFood(level, entity, stack, 0);
+    }
+    public boolean placeFood(ServerLevel level, @Nullable LivingEntity entity, ItemStack stack, int startSlot) {
+        for (int i = startSlot; i < this.items.size(); i++) {
+        // KioCG end - change container behavior
             ItemStack itemStack = this.items.get(i);
             if (itemStack.isEmpty()) {
                 Optional<RecipeHolder<CampfireCookingRecipe>> recipeFor = level.recipeAccess()
@@ -226,6 +_,13 @@
                 if (recipeFor.isEmpty()) {
                     return false;
                 }
+                // KioCG start - vanilla recipe excludes model items
+                if (recipeFor.get().id().location().getNamespace().equals(org.bukkit.NamespacedKey.MINECRAFT)) {
+                    if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack)) {
+                        return false;
+                    }
+                }
+                // KioCG end - vanilla recipe excludes model items
 
                 // CraftBukkit start
                 org.bukkit.event.block.CampfireStartEvent event = new org.bukkit.event.block.CampfireStartEvent(
@@ -238,6 +_,7 @@
                 // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, stack.consumeAndReturn(1, entity));
+                this.updateBlockEntityTicker(); // KioCG - optimize block entity
                 level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(entity, this.getBlockState()));
                 this.markUpdated();
                 return true;
