--- a/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
@@ -105,6 +_,47 @@
         }
     }
 
+    // KioCG start - infinite decorated pot
+    // public static void suckInItems(net.minecraft.world.level.Level level, BlockPos pos, BlockState state, DecoratedPotBlockEntity decoratedPot) {
+    //     for (net.minecraft.world.entity.item.ItemEntity itemEntity : getItemsAtAndAbove(level, pos)) {
+    //         if (!level.paperConfig().kiocgConfig.hopperPullsEntityOwnedItem && itemEntity.target != null) continue; // KioCG - hopper filter
+    //         if (!level.paperConfig().kiocgConfig.hopperPullsInfinitePickupItem && itemEntity.pickupDelay == Short.MAX_VALUE) continue; // KioCG - hopper filter
+    //         if (itemEntity.canDecoratedPotSuck && net.minecraft.world.level.block.DecoratedPotBlock.putItem(itemEntity.getItem(), level, pos, null, decoratedPot).consumesAction()) {
+    //             itemEntity.pickupDelay++;
+    //             decoratedPot.setSuckInMovement(itemEntity);
+    //             return;
+    //         }
+    //     }
+    // }
+
+    private int suckInTime = 0;
+    public static void itemSearchHopper(net.minecraft.world.level.Level level, BlockPos pos, BlockState state, net.minecraft.world.entity.item.ItemEntity itemEntity, DecoratedPotBlockEntity blockEntity) {
+        if (blockEntity.suckInTime != net.minecraft.server.MinecraftServer.currentTick && !itemEntity.getItem().isEmpty()
+            && itemEntity.getBoundingBox().move(-pos.getX(), -pos.getY(), -pos.getZ()).intersects(SUCK_AABB)) {
+            if (itemEntity.canDecoratedPotSuck && net.minecraft.world.level.block.DecoratedPotBlock.putItem(itemEntity.getItem(), level, pos, null, blockEntity).consumesAction()) {
+                itemEntity.pickupDelay++;
+                itemEntity.searchingHopper = true;
+                blockEntity.setSuckInMovement(itemEntity);
+                blockEntity.suckInTime = net.minecraft.server.MinecraftServer.currentTick;
+            }
+        }
+    }
+
+    private void setSuckInMovement(net.minecraft.world.entity.item.ItemEntity itemEntity) {
+        net.minecraft.world.phys.Vec3 deltaMovement = itemEntity.getDeltaMovement();
+        if (deltaMovement.horizontalDistanceSqr() > 1.0E-5F) {
+            itemEntity.setDeltaMovement(deltaMovement.x * 0.5F, deltaMovement.y, deltaMovement.z * 0.5F);
+            itemEntity.hurtMarked = true;
+        }
+    }
+
+    private static final net.minecraft.world.phys.AABB SUCK_AABB = net.minecraft.world.level.block.Block.column(8.0, 16.0, 24.0).toAabbs().get(0);
+    public static List<net.minecraft.world.entity.item.ItemEntity> getItemsAtAndAbove(net.minecraft.world.level.Level level, BlockPos pos) {
+        net.minecraft.world.phys.AABB aabb = SUCK_AABB.move(pos.getX(), pos.getY(), pos.getZ());
+        return level.getEntitiesOfClass(net.minecraft.world.entity.item.ItemEntity.class, aabb, net.minecraft.world.entity.EntitySelector.ENTITY_STILL_ALIVE);
+    }
+    // KioCG end - infinite decorated pot
+
     @Override
     public ClientboundBlockEntityDataPacket getUpdatePacket() {
         return ClientboundBlockEntityDataPacket.create(this);
@@ -198,6 +_,7 @@
     @Override
     public void setTheItem(ItemStack item) {
         this.unpackLootTable(null);
+        if (this.hasLevel()) item.tickFoodFreeze(this.getLevel().getGameTime()); // KioCG - food freshness
         this.item = item;
     }
 
