--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -24,6 +_,11 @@
     }
 
     public boolean isFuel(ItemStack stack) {
+        // KioCG start - vanilla recipe excludes model items
+        if (!net.minecraft.world.inventory.AbstractContainerMenu.allowModelItem(net.minecraft.world.inventory.MenuType.FURNACE, stack)) {
+            return false;
+        }
+        // KioCG end - vanilla recipe excludes model items
         return this.values.containsKey(stack.getItem());
     }
 
@@ -41,6 +_,8 @@
 
     public static FuelValues vanillaBurnTimes(HolderLookup.Provider registries, FeatureFlagSet enabledFeatures, int smeltingTime) {
         return new FuelValues.Builder(registries, enabledFeatures)
+            // KioCG start - change container behavior
+            /*
             .add(Items.LAVA_BUCKET, smeltingTime * 100)
             .add(Blocks.COAL_BLOCK, smeltingTime * 8 * 10)
             .add(Items.BLAZE_ROD, smeltingTime * 12)
@@ -100,6 +_,10 @@
             .add(Blocks.AZALEA, smeltingTime / 2)
             .add(Blocks.FLOWERING_AZALEA, smeltingTime / 2)
             .add(Blocks.MANGROVE_ROOTS, smeltingTime * 3 / 2)
+             */
+            .add(Items.COAL, smeltingTime * 8)
+            .add(Items.CHARCOAL, smeltingTime * 8)
+            // KioCG end
             .remove(ItemTags.NON_FLAMMABLE_WOOD)
             .build();
     }
