--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -33,10 +_,66 @@
     public static final int HOPPER_CONTAINER_SIZE = 5;
     private static final int[][] CACHED_SLOTS = new int[54][];
     private static final int NO_COOLDOWN_TIME = -1;
-    private NonNullList<ItemStack> items = NonNullList.withSize(5, ItemStack.EMPTY);
+    // KioCG start - optimize block entity
+    private HopperNonNullList<ItemStack> items = withSize(5, ItemStack.EMPTY);
+    public <E> HopperNonNullList<E> withSize(int size, E defaultValue) {
+        org.apache.commons.lang3.Validate.notNull(defaultValue);
+        Object[] objects = new Object[size];
+        java.util.Arrays.fill(objects, defaultValue);
+        return new HopperNonNullList<>(java.util.Arrays.asList((E[])objects), defaultValue);
+    }
+    public class HopperNonNullList<E> extends NonNullList<E> {
+        protected HopperNonNullList(List<E> list, @Nullable E defaultValue) {
+            super(list, defaultValue);
+        }
+
+        @Override
+        public E set(int index, E value) {
+            E prev = super.set(index, value);
+            if (!HopperBlockEntity.this.tickerTicking && !HopperBlockEntity.this.preFirstTick) {
+                if (HopperBlockEntity.this.needEjectItems && !((ItemStack) value).isEmpty()) {
+                    HopperBlockEntity.this.updateBlockEntityTicker();
+                }
+                if (HopperBlockEntity.this.needSuckInItems && ((ItemStack) value).getCount() != ((ItemStack) value).getMaxStackSize()) {
+                    HopperBlockEntity.this.updateBlockEntityTicker();
+                }
+            }
+            return prev;
+        }
+
+        @Override
+        public void add(int index, E value) {
+            super.add(index, value);
+            if (!HopperBlockEntity.this.tickerTicking && !HopperBlockEntity.this.preFirstTick) {
+                if (HopperBlockEntity.this.needEjectItems && !((ItemStack) value).isEmpty()) {
+                    HopperBlockEntity.this.updateBlockEntityTicker();
+                }
+            }
+        }
+
+        @Override
+        public E remove(int index) {
+            E prev = super.remove(index);
+            if (!HopperBlockEntity.this.tickerTicking && !HopperBlockEntity.this.preFirstTick) {
+                if (HopperBlockEntity.this.needSuckInItems) {
+                    HopperBlockEntity.this.updateBlockEntityTicker();
+                }
+            }
+            return prev;
+        }
+    }
+    // KioCG end - optimize block entity
+
     public int cooldownTime = -1;
     private long tickedGameTime;
     private Direction facing;
+    // KioCG start - optimize block entity
+    private boolean needSuckInItems = true;
+    private boolean needEjectItems = true;
+    private boolean needDoEntityInside = false;
+    private boolean needDropItems = false;
+    // KioCG end - optimize block entity
+    @Nullable public java.util.Set<net.minecraft.world.item.Item> itemFilter; // KioCG - hopper filter
 
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
@@ -80,12 +_,13 @@
     @Override
     protected void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
-        this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
+        this.items = withSize(this.getContainerSize(), ItemStack.EMPTY); // KioCG - optimize block entity
         if (!this.tryLoadLootTable(input)) {
             ContainerHelper.loadAllItems(input, this.items);
         }
 
         this.cooldownTime = input.getIntOr("TransferCooldown", -1);
+        this.itemFilter = this.hasCustomName() ? HopperBlockEntity.parseItemFilter(org.bukkit.craftbukkit.util.CraftChatMessage.fromComponent(this.getCustomName())) : null; // KioCG - hopper filter
     }
 
     @Override
@@ -106,7 +_,11 @@
     @Override
     public ItemStack removeItem(int index, int count) {
         this.unpackLootTable(null);
-        return ContainerHelper.removeItem(this.getItems(), index, count);
+        // KioCG start - optimize block entity: 触发更新
+        ItemStack itemStack = ContainerHelper.removeItem(this.getItems(), index, count);
+        this.getItems().set(index, this.getItems().get(index));
+        return itemStack;
+        // KioCG end - optimize block entity: 触发更新
     }
 
     @Override
@@ -134,7 +_,7 @@
             blockEntity.setCooldown(0);
             // Spigot start
             boolean result = tryMoveItems(level, pos, state, blockEntity, () -> {
-                return suckInItems(level, blockEntity);
+                return blockEntity.needSuckInItems && suckInItems(level, blockEntity); // KioCG - optimize block entity
             });
             if (!result && blockEntity.level.spigotConfig.hopperCheck > 1) {
                 blockEntity.setCooldown(blockEntity.level.spigotConfig.hopperCheck);
@@ -143,6 +_,29 @@
         }
     }
 
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        if (!state.getValue(HopperBlock.ENABLED)) return false;
+
+        BlockPos abovePos = pos.above();
+        BlockState aboveState = level.getBlockState(abovePos);
+        BlockPos facingPos = pos.relative(this.facing);
+        BlockState facingState = level.getBlockState(facingPos);
+        this.needDoEntityInside = !hasBlockContainer(level, abovePos, aboveState);
+        // 不吸上方的漏斗来进一步优化
+        this.needSuckInItems = (!this.needDoEntityInside && !aboveState.is(net.minecraft.world.level.block.Blocks.HOPPER)) || aboveState.is(BlockTags.RAILS);
+        this.needDropItems = facingState.getCollisionShape(level, pos).isEmpty();
+        this.needEjectItems = this.needDropItems || facingState.is(BlockTags.RAILS) || hasBlockContainer(level, facingPos, facingState);
+        final int fullState = getFullState(this); // Paper - Perf: Optimize Hoppers
+        return this.needSuckInItems && fullState != HOPPER_IS_FULL || this.needEjectItems && fullState != HOPPER_EMPTY;
+    }
+
+    private static boolean hasBlockContainer(Level level, BlockPos pos, BlockState state) {
+        return state.getBlock() instanceof WorldlyContainerHolder || state.hasBlockEntity() && level.getBlockEntity(pos) instanceof Container;
+    }
+    // KioCG end - optimize block entity
+
     // Paper start - Perf: Optimize Hoppers
     private static final int HOPPER_EMPTY = 0;
     private static final int HOPPER_HAS_ITEMS = 1;
@@ -188,12 +_,20 @@
                 boolean flag = false;
                 final int fullState = getFullState(blockEntity); // Paper - Perf: Optimize Hoppers
                 if (fullState != HOPPER_EMPTY) { // Paper - Perf: Optimize Hoppers
-                    flag = ejectItems(level, pos, blockEntity);
+                    flag = blockEntity.needEjectItems && ejectItems(level, pos, blockEntity); // KioCG - optimize block entity
                 }
 
                 if (fullState != HOPPER_IS_FULL || flag) { // Paper - Perf: Optimize Hoppers
                     flag |= validator.getAsBoolean(); // Paper - note: this is not a validator, it's what adds/sucks in items
-                }
+                } else { blockEntity.setCooldown(level.spigotConfig.hopperTransfer); // KioCG - optimize block entity
+                }
+
+                // KioCG start - optimize block entity
+                if (!blockEntity.needSuckInItems && fullState == HOPPER_EMPTY
+                    || !blockEntity.needEjectItems && fullState == HOPPER_IS_FULL) {
+                    blockEntity.updateBlockEntityTicker();
+                }
+                // KioCG end - optimize block entity
 
                 if (flag) {
                     blockEntity.setCooldown(level.spigotConfig.hopperTransfer); // Spigot
@@ -208,7 +_,7 @@
 
     private boolean inventoryFull() {
         for (ItemStack itemStack : this.items) {
-            if (itemStack.isEmpty() || itemStack.getCount() != itemStack.getMaxStackSize()) {
+            if (itemStack.isEmpty() || itemStack.getCount() < itemStack.getMaxStackSize()) { // KioCG - optimize block entity: != -> <
                 return false;
             }
         }
@@ -221,12 +_,14 @@
     private static boolean skipPullModeEventFire;
     private static boolean skipPushModeEventFire;
 
-    private static boolean hopperPush(final Level level, final Container destination, final Direction direction, final HopperBlockEntity hopper) {
+    private static boolean hopperPush(final Level level, final @Nullable Container destination, final Direction direction, final HopperBlockEntity hopper) { // KioCG - hopper drop items: allow nullable
         skipPushModeEventFire = skipHopperEvents;
         boolean foundItem = false;
         for (int i = 0; i < hopper.getContainerSize(); ++i) {
             final ItemStack item = hopper.getItem(i);
             if (!item.isEmpty()) {
+                if (!HopperBlockEntity.suckInFilter(hopper, item) || !HopperBlockEntity.suckInFilter(destination, item)) continue; // KioCG - hopper filter
+
                 foundItem = true;
                 ItemStack origItemStack = item;
                 ItemStack movedItem = origItemStack;
@@ -237,7 +_,7 @@
 
                 // We only need to fire the event once to give protection plugins a chance to cancel this event
                 // Because nothing uses getItem, every event call should end up the same result.
-                if (!skipPushModeEventFire) {
+                if (!skipPushModeEventFire && destination != null) { // KioCG - hopper drop items: allow nullable
                     movedItem = callPushMoveEvent(destination, movedItem, hopper);
                     if (movedItem == null) { // cancelled
                         origItemStack.setCount(originalItemCount);
@@ -245,7 +_,34 @@
                     }
                 }
 
-                final ItemStack remainingItem = addItem(hopper, destination, movedItem, direction);
+                // KioCG start - hopper drop items
+                final ItemStack remainingItem;
+                if (destination != null) {
+                    remainingItem = addItem(hopper, destination, movedItem, direction);
+                } else {
+                    BlockPos toCopy = hopper.worldPosition;
+                    // 当destination == null时, direction为漏斗朝向
+                    net.minecraft.world.phys.Vec3 vec3 = switch (direction) {
+                        case DOWN ->
+                                new net.minecraft.world.phys.Vec3(toCopy.getX() + 0.5, toCopy.getY() - net.minecraft.world.entity.EntityType.ITEM.getHeight() - 0.125, toCopy.getZ() + 0.5);
+                        case NORTH ->
+                                new net.minecraft.world.phys.Vec3(toCopy.getX() + 0.5, toCopy.getY() + (1.0 / 16.0 * 6.0) - (net.minecraft.world.entity.EntityType.ITEM.getHeight() / 2.0) - 0.125, toCopy.getZ() - (net.minecraft.world.entity.EntityType.ITEM.getWidth() / 2.0));
+                        case SOUTH ->
+                                new net.minecraft.world.phys.Vec3(toCopy.getX() + 0.5, toCopy.getY() + (1.0 / 16.0 * 6.0) - (net.minecraft.world.entity.EntityType.ITEM.getHeight() / 2.0) - 0.125, toCopy.getZ() + 1 + (net.minecraft.world.entity.EntityType.ITEM.getWidth() / 2.0));
+                        case WEST ->
+                                new net.minecraft.world.phys.Vec3(toCopy.getX() - (net.minecraft.world.entity.EntityType.ITEM.getWidth() / 2.0), toCopy.getY() + (1.0 / 16.0 * 6.0) - (net.minecraft.world.entity.EntityType.ITEM.getHeight() / 2.0) - 0.125, toCopy.getZ() + 0.5);
+                        case EAST ->
+                                new net.minecraft.world.phys.Vec3(toCopy.getX() + 1 + (net.minecraft.world.entity.EntityType.ITEM.getWidth() / 2.0), toCopy.getY() + (1.0 / 16.0 * 6.0) - (net.minecraft.world.entity.EntityType.ITEM.getHeight() / 2.0) - 0.125, toCopy.getZ() + 0.5);
+                        default -> throw new IllegalStateException();
+                    };
+                    ItemEntity itemEntity = new ItemEntity(level, vec3.x(), vec3.y(), vec3.z(), movedItem);
+                    itemEntity.setDeltaMovement(net.minecraft.world.phys.Vec3.ZERO);
+                    itemEntity.setDefaultPickUpDelay();
+                    level.addFreshEntity(itemEntity);
+
+                    remainingItem = ItemStack.EMPTY;
+                }
+                // KioCG end - hopper drop items
                 final int remainingItemCount = remainingItem.getCount();
                 if (remainingItemCount != movedItemCount) {
                     origItemStack = origItemStack.copy(true);
@@ -254,6 +_,7 @@
                         origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
                     }
                     hopper.setItem(i, origItemStack);
+                    if (destination != null) // KioCG - hopper drop items: allow nullable
                     destination.setChanged();
                     return true;
                 }
@@ -267,6 +_,8 @@
     }
 
     private static boolean hopperPull(final Level level, final Hopper hopper, final Container container, ItemStack origItemStack, final int i) {
+        if (!HopperBlockEntity.suckInFilter(hopper, origItemStack) || !HopperBlockEntity.suckInFilter(container, origItemStack)) return false; // KioCG - hopper filter
+
         ItemStack movedItem = origItemStack;
         final int originalItemCount = origItemStack.getCount();
         final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
@@ -417,12 +_,19 @@
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(Level level, BlockPos pos, HopperBlockEntity blockEntity) {
+        if (blockEntity.needDropItems) return hopperPush(level, null, blockEntity.facing, blockEntity); // KioCG - hopper drop items
+
         Container attachedContainer = getAttachedContainer(level, pos, blockEntity);
         if (attachedContainer == null) {
             return false;
         } else {
             Direction opposite = blockEntity.facing.getOpposite();
             if (isFullContainer(attachedContainer, opposite)) {
+                // KioCG start - optimize block entity
+                if (level.paperConfig().hopper.cooldownWhenFull) { // Inventory was full - cooldown
+                    blockEntity.setCooldown(level.spigotConfig.hopperTransfer);
+                }
+                // KioCG end - optimize block entity
                 return false;
             } else {
                 // Paper start - Perf: Optimize Hoppers
@@ -527,6 +_,64 @@
         return true;
     }
 
+    // KioCG start - hopper filter
+    @Override
+    protected void applyImplicitComponents(net.minecraft.core.component.DataComponentGetter componentGetter) {
+        super.applyImplicitComponents(componentGetter);
+        this.itemFilter = this.hasCustomName() ? HopperBlockEntity.parseItemFilter(org.bukkit.craftbukkit.util.CraftChatMessage.fromComponent(this.getCustomName())) : null;
+    }
+
+    @Nullable
+    public static java.util.Set<net.minecraft.world.item.Item> parseItemFilter(String string) {
+        java.util.Set<net.minecraft.world.item.Item> filter = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>();
+
+        for (String split : string.split(",")) {
+            if (split.startsWith("#")) {
+                split = split.substring(1);
+                if (split.isEmpty() || !net.minecraft.resources.ResourceLocation.isValidPath(split)) {
+                    return null;
+                }
+
+                net.minecraft.tags.TagKey<net.minecraft.world.item.Item> itemTagKey = net.minecraft.tags.TagKey.create(net.minecraft.core.registries.Registries.ITEM, net.minecraft.resources.ResourceLocation.withDefaultNamespace(split));
+                java.util.Optional<net.minecraft.core.HolderSet.Named<net.minecraft.world.item.Item>> itemTag = net.minecraft.core.registries.BuiltInRegistries.ITEM.get(itemTagKey);
+                if (itemTag.isEmpty()) {
+                    return null;
+                }
+
+                for (net.minecraft.core.Holder<net.minecraft.world.item.Item> itemHolder : itemTag.get()) {
+                    filter.add(itemHolder.value());
+                }
+            } else {
+                if (split.isEmpty() || !net.minecraft.resources.ResourceLocation.isValidPath(split)) {
+                    return null;
+                }
+
+                final net.minecraft.world.item.Item item = net.minecraft.core.registries.BuiltInRegistries.ITEM.getValue(net.minecraft.resources.ResourceLocation.withDefaultNamespace(split));
+                if (item == net.minecraft.world.item.Items.AIR) {
+                    return null;
+                }
+
+                filter.add(item);
+            }
+        }
+
+        return !filter.isEmpty() ? filter : null;
+    }
+
+    public static boolean suckInFilter(@Nullable Container hopper, ItemStack itemstack) {
+        if (hopper instanceof HopperBlockEntity blockEntity) {
+            if (blockEntity.itemFilter != null) {
+                return blockEntity.itemFilter.contains(itemstack.getItem());
+            }
+        } else if (hopper instanceof net.minecraft.world.entity.vehicle.MinecartHopper minecartHopper) {
+            if (minecartHopper.itemFilter != null) {
+                return minecartHopper.itemFilter.contains(itemstack.getItem());
+            }
+        }
+        return true;
+    }
+    // KioCG end - hopper filter
+
     public static boolean suckInItems(Level level, Hopper hopper) {
         BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockState = level.getBlockState(blockPos);
@@ -543,7 +_,9 @@
 
             return false;
         } else {
+            if (hopper.isGridAligned()) return false; // KioCG - optimize block entity: item search hopper
             boolean flag = hopper.isGridAligned() && blockState.isCollisionShapeFullBlock(level, blockPos) && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
+            // flag = flag || hopper.isGridAligned() && level.paperConfig().kiocgConfig.hopperRailsOptimize && blockState.is(BlockTags.RAILS); // KioCG - hopper rails optimize: 不要吸取铁轨上的物品
             if (!flag) {
                 for (ItemEntity itemEntity : getItemsAtAndAbove(level, hopper)) {
                     if (addItem(hopper, itemEntity)) {
@@ -617,6 +_,10 @@
     }
 
     public static boolean addItem(Container container, ItemEntity item) {
+        if (!HopperBlockEntity.suckInFilter(container, item.getItem())) return false; // KioCG - hopper filter
+        if (!item.level().paperConfig().kiocgConfig.block.hopperPullsEntityOwnedItem && item.target != null) return false; // KioCG - hopper filter
+        if (!item.level().paperConfig().kiocgConfig.block.hopperPullsInfinitePickupItem && item.pickupDelay == Short.MAX_VALUE) return false; // KioCG - hopper filter
+
         boolean flag = false;
         // CraftBukkit start
         if (org.bukkit.event.inventory.InventoryPickupItemEvent.getHandlerList().getRegisteredListeners().length > 0) { // Paper - optimize hoppers
@@ -628,20 +_,33 @@
         }
         // CraftBukkit end
         } // Paper - Perf: Optimize Hoppers
-        ItemStack itemStack = item.getItem().copy();
+        // KioCG start - allow overstacked items: fix hopper
+        final ItemStack origItemStack = item.getItem();
+        ItemStack itemStack = origItemStack.split(origItemStack.getMaxStackSize());
+        // KioCG end - allow overstacked items: fix hopper
+        itemStack.hopperEnableFreshness(item.level().getGameTime()); // KioCG - food freshness
         ItemStack itemStack1 = addItem(null, container, itemStack, null);
         if (itemStack1.isEmpty()) {
             flag = true;
+            if (origItemStack.isEmpty()) { // KioCG - allow overstacked items: fix hopper
             item.setItem(ItemStack.EMPTY);
             item.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
+            } // KioCG - allow overstacked items: fix hopper
         } else {
-            item.setItem(itemStack1);
+            origItemStack.grow(itemStack1.getCount()); // KioCG - allow overstacked items: fix hopper
         }
 
         return flag;
     }
 
     public static ItemStack addItem(@Nullable Container source, Container destination, ItemStack stack, @Nullable Direction direction) {
+        // KioCG start - hopper filter
+        if (source instanceof BaseContainerBlockEntity lockable1 && !lockable1.getLevel().paperConfig().kiocgConfig.block.hopperPullsLockedContainer && lockable1.lockKey != net.minecraft.world.LockCode.NO_LOCK
+            || destination instanceof BaseContainerBlockEntity lockable2 && !lockable2.getLevel().paperConfig().kiocgConfig.block.hopperPullsLockedContainer && lockable2.lockKey != net.minecraft.world.LockCode.NO_LOCK) {
+            return stack;
+        }
+        // KioCG end - hopper filter
+
         if (destination instanceof WorldlyContainer worldlyContainer && direction != null) {
             int[] slotsForFace = worldlyContainer.getSlotsForFace(direction);
 
@@ -737,6 +_,8 @@
         BlockPos searchPosition = pos.relative(blockEntity.facing);
         Container inventory = getContainerAt(level, searchPosition);
 
+        if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory; // KioCG - optimize block entity
+
         org.bukkit.craftbukkit.block.CraftBlock hopper = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
         org.bukkit.craftbukkit.block.CraftBlock searchBlock = org.bukkit.craftbukkit.block.CraftBlock.at(level, searchPosition);
         return HopperBlockEntity.runHopperInventorySearchEvent(
@@ -753,6 +_,8 @@
         // CraftBukkit start
         final Container inventory = HopperBlockEntity.getContainerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0D, hopper.getLevelZ());
 
+        if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory; // KioCG - optimize block entity
+
         final BlockPos hopperPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY(), hopper.getLevelZ());
         org.bukkit.craftbukkit.block.CraftBlock hopperBlock = org.bukkit.craftbukkit.block.CraftBlock.at(level, hopperPos);
         org.bukkit.craftbukkit.block.CraftBlock containerBlock = org.bukkit.craftbukkit.block.CraftBlock.at(level, hopperPos.above());
@@ -772,7 +_,11 @@
 
     @Nullable
     public static Container getContainerAt(Level level, BlockPos pos) {
-        return getContainerAt(level, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, true); // Paper - Optimize hoppers
+        // KioCG start - optimize block entity: 正确阻止区块加载
+        BlockState blockState = !level.spigotConfig.hopperCanLoadChunks ? level.getBlockStateIfLoaded(pos) : level.getBlockState(pos);
+        if (blockState == null) return null;
+        return getContainerAt(level, pos, blockState, pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, !level.paperConfig().kiocgConfig.block.hopperRailsOptimize); // Paper - Optimize hoppers // KioCG - hopper rails optimize: 不用额外判断阻塞
+        // KioCG end - optimize block entity: 正确阻止区块加载
     }
 
     @Nullable
@@ -785,6 +_,7 @@
         // Paper end - Perf: Optimize Hoppers
         Container blockContainer = getBlockContainer(level, pos, state);
         if (blockContainer == null && (!optimizeEntities || !level.paperConfig().hopper.ignoreOccludingBlocks || !state.getBukkitMaterial().isOccluding())) { // Paper - Perf: Optimize Hoppers
+            if (level.paperConfig().kiocgConfig.block.hopperRailsOptimize && state.is(BlockTags.RAILS)) // KioCG - hopper rails optimize: 不要吸取铁轨外的矿车
             blockContainer = getEntityContainer(level, x, y, z);
         }
 
@@ -793,11 +_,13 @@
 
     @Nullable
     private static Container getBlockContainer(Level level, BlockPos pos, BlockState state) {
-        if (!level.spigotConfig.hopperCanLoadChunks && !level.hasChunkAt(pos)) return null; // Spigot
+        // if (!level.spigotConfig.hopperCanLoadChunks && !level.hasChunkAt(pos)) return null; // Spigot // KioCG - optimize block entity: 正确阻止区块加载, getBlockState已经加载区块...
         Block block = state.getBlock();
         if (block instanceof WorldlyContainerHolder) {
             return ((WorldlyContainerHolder)block).getContainer(state, level, pos);
         } else if (state.hasBlockEntity() && level.getBlockEntity(pos) instanceof Container container) {
+            if (state.is(net.minecraft.world.level.block.Blocks.HOPPER)) return null; // KioCG - optimize block entity: 不吸上方的漏斗来进一步优化
+
             if (container instanceof ChestBlockEntity && block instanceof ChestBlock) {
                 container = ChestBlock.getContainer((ChestBlock)block, state, level, pos, true);
             }
@@ -859,14 +_,24 @@
 
     @Override
     protected void setItems(NonNullList<ItemStack> items) {
-        this.items = items;
+        this.items = new HopperNonNullList<>(new java.util.ArrayList<>(items), ItemStack.EMPTY); // KioCG - optimize block entity
     }
 
+    private int suckInTime = 0; // KioCG - optimize block entity: item search hopper
     public static void entityInside(Level level, BlockPos pos, BlockState state, Entity entity, HopperBlockEntity blockEntity) {
+        if (!blockEntity.needDoEntityInside || !state.getValue(HopperBlock.ENABLED)) return; // KioCG - optimize block entity: item search hopper
         if (entity instanceof ItemEntity itemEntity
             && !itemEntity.getItem().isEmpty()
             && entity.getBoundingBox().move(-pos.getX(), -pos.getY(), -pos.getZ()).intersects(blockEntity.getSuckAabb())) {
-            tryMoveItems(level, pos, state, blockEntity, () -> addItem(blockEntity, itemEntity));
+            // if (level.paperConfig().kiocgConfig.hopperRailsOptimize && level.getBlockState(pos.relative(Direction.UP)).is(BlockTags.RAILS)) return; // KioCG - hopper rails optimize: 不要吸取铁轨上的物品
+            // KioCG start - optimize block entity: item search hopper
+            if (blockEntity.suckInTime != net.minecraft.server.MinecraftServer.currentTick) {
+                if (!blockEntity.inventoryFull()) {
+                    addItem(blockEntity, itemEntity);
+                }
+                blockEntity.suckInTime = net.minecraft.server.MinecraftServer.currentTick;
+            }
+            // KioCG end - optimize block entity: item search hopper
         }
     }
 
