--- a/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java
@@ -24,4 +_,11 @@
             level.updateNeighborsAt(pos.below(), block, orientation);
         }
     }
+
+    // KioCG start - expand container rows
+    @Override
+    protected int getDefaultRows() {
+        return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.container.defaultTrappedChestRows;
+    }
+    // KioCG end - expand container rows
 }
