--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -125,23 +_,32 @@
     protected void registerGoals() {
         this.goalSelector.addGoal(1, new FloatGoal(this));
         this.goalSelector.addGoal(1, new TamableAnimal.TamableAnimalPanicGoal(1.5, DamageTypeTags.PANIC_ENVIRONMENTAL_CAUSES));
-        this.goalSelector.addGoal(2, new SitWhenOrderedToGoal(this));
-        this.goalSelector.addGoal(3, new Wolf.WolfAvoidEntityGoal<>(this, Llama.class, 24.0F, 1.5, 1.5));
-        this.goalSelector.addGoal(4, new LeapAtTargetGoal(this, 0.4F));
-        this.goalSelector.addGoal(5, new MeleeAttackGoal(this, 1.0, true));
-        this.goalSelector.addGoal(6, new FollowOwnerGoal(this, 1.0, 10.0F, 2.0F));
-        this.goalSelector.addGoal(7, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(8, new WaterAvoidingRandomStrollGoal(this, 1.0));
+        // KioCG start - optimize entity behavior
+        this.goalRegister.addGoal(this.goalSelector, 2, new SitWhenOrderedToGoal(this));
+        this.goalRegister.addGoal(this.goalSelector, 3, new Wolf.WolfAvoidEntityGoal<>(this, Llama.class, 24.0F, 1.5, 1.5));
+        this.goalRegister.addGoal(this.goalSelector, 4, new LeapAtTargetGoal(this, 0.4F));
+        this.goalRegister.addGoal(this.goalSelector, 5, new MeleeAttackGoal(this, 1.0, true));
+        this.goalRegister.addGoal(this.goalSelector, 6, new FollowOwnerGoal(this, 1.0, 10.0F, 2.0F));
+        this.goalRegister.addGoal(this.goalSelector, 7, new BreedGoal(this, 1.0), this::isTame);
+        this.goalSelector.addGoal(8, new WaterAvoidingRandomStrollGoal(this, 1.0) {
+            @Override
+            public boolean canUse() {
+                return !((TamableAnimal) this.mob).isInSittingPose() && super.canUse();
+            }
+        });
+        // KioCG end - optimize entity behavior
         this.goalSelector.addGoal(9, new BegGoal(this, 8.0F));
         this.goalSelector.addGoal(10, new LookAtPlayerGoal(this, Player.class, 8.0F));
         this.goalSelector.addGoal(10, new RandomLookAroundGoal(this));
-        this.targetSelector.addGoal(1, new OwnerHurtByTargetGoal(this));
-        this.targetSelector.addGoal(2, new OwnerHurtTargetGoal(this));
+        // KioCG start - optimize entity behavior
+        this.goalRegister.addGoal(this.targetSelector, 1, new OwnerHurtByTargetGoal(this));
+        this.goalRegister.addGoal(this.targetSelector, 2, new OwnerHurtTargetGoal(this));
         this.targetSelector.addGoal(3, new HurtByTargetGoal(this).setAlertOthers());
-        this.targetSelector.addGoal(4, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
-        this.targetSelector.addGoal(5, new NonTameRandomTargetGoal<>(this, Animal.class, false, PREY_SELECTOR));
-        this.targetSelector.addGoal(6, new NonTameRandomTargetGoal<>(this, Turtle.class, false, Turtle.BABY_ON_LAND_SELECTOR));
-        this.targetSelector.addGoal(7, new NearestAttackableTargetGoal<>(this, AbstractSkeleton.class, false));
+        this.goalRegister.addGoal(this.targetSelector, 4, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt), () -> !isOrderedToSit());
+        this.goalRegister.addGoal(this.targetSelector, 5, new NonTameRandomTargetGoal<>(this, Animal.class, false, PREY_SELECTOR));
+        this.goalRegister.addGoal(this.targetSelector, 6, new NonTameRandomTargetGoal<>(this, Turtle.class, false, Turtle.BABY_ON_LAND_SELECTOR));
+        this.goalRegister.addGoal(this.targetSelector, 7, new NearestAttackableTargetGoal<>(this, AbstractSkeleton.class, false), () -> !isOrderedToSit());
+        // KioCG end - optimize entity behavior
         this.targetSelector.addGoal(8, new ResetUniversalAngerTargetGoal<>(this, true));
     }
 
@@ -463,12 +_,28 @@
         Item item = itemInHand.getItem();
         if (this.isTame()) {
             if (this.isFood(itemInHand) && this.getHealth() < this.getMaxHealth()) {
+                this.playEatingSound(); // KioCG - dont kill my pet
+
                 this.usePlayerItem(player, hand, itemInHand);
                 FoodProperties foodProperties = itemInHand.get(DataComponents.FOOD);
                 float f = foodProperties != null ? foodProperties.nutrition() : 1.0F;
                 this.heal(2.0F * f, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
+                return InteractionResult.SUCCESS_SERVER; // KioCG - dont kill my pet
+            }
+
+            // KioCG start - dont kill my pet
+            if (itemInHand.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) {
+                this.playEatingSound();
+
+                this.usePlayerItem(player, hand, itemInHand);
+                this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                if (player.isCreative() || !this.isInvulnerable()) {
+                    this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                }
+
                 return InteractionResult.SUCCESS;
             }
+            // KioCG end - dont kill my pet
 
             if (!(item instanceof DyeItem dyeItem && this.isOwnedBy(player))) {
                 if (this.isEquippableInSlot(itemInHand, EquipmentSlot.BODY) && !this.isWearingBodyArmor() && this.isOwnedBy(player) && !this.isBaby()) {
@@ -516,6 +_,8 @@
                 return InteractionResult.SUCCESS;
             }
         } else if (!this.level().isClientSide && itemInHand.is(Items.BONE) && !this.isAngry()) {
+            this.playEatingSound(); // KioCG - dont kill my pet
+
             itemInHand.consume(1, player);
             this.tryToTame(player);
             return InteractionResult.SUCCESS_SERVER;
@@ -524,6 +_,13 @@
         return super.mobInteract(player, hand);
     }
 
+    // KioCG start - dont kill my pet
+    @Override
+    protected void playEatingSound() {
+        this.playSound(SoundEvents.CAT_EAT, 1.0F, 0.5F);
+    }
+    // KioCG end - dont kill my pet
+
     private void tryToTame(Player player) {
         if (this.random.nextInt(3) == 0 && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTameEvent(this, player).isCancelled()) { // CraftBukkit - added event call and isCancelled check.
             this.tame(player);
@@ -688,9 +_,16 @@
             this.wolf = wolf;
         }
 
+        // KioCG start - optimize entity behavior
+        @Override
+        public boolean registerWhen() {
+            return !this.wolf.isTame();
+        }
+        // KioCG end - optimize entity behavior
+
         @Override
         public boolean canUse() {
-            return super.canUse() && this.toAvoid instanceof Llama && !this.wolf.isTame() && this.avoidLlama((Llama)this.toAvoid);
+            return super.canUse() && this.toAvoid instanceof Llama && this.avoidLlama((Llama)this.toAvoid); // KioCG - optimize entity behavior
         }
 
         private boolean avoidLlama(Llama llama) {
