--- a/net/minecraft/server/level/BlockDestructionProgress.java
+++ b/net/minecraft/server/level/BlockDestructionProgress.java
@@ -3,6 +_,13 @@
 import net.minecraft.core.BlockPos;
 
 public class BlockDestructionProgress implements Comparable<BlockDestructionProgress> {
+    // KioCG start - keep block destruction
+    private static int blockDestroyId = Integer.MAX_VALUE;
+    public static int getNextBlockDestroyId() {
+        return blockDestroyId--;
+    }
+    // KioCG end - keep block destruction
+
     private final int id;
     private final BlockPos pos;
     private int progress;
