--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -721,7 +_,7 @@
         }
 
         int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
-        if (this.sleepStatus.areEnoughSleeping(_int) && this.sleepStatus.areEnoughDeepSleeping(_int, this.players)) {
+        if (false && this.sleepStatus.areEnoughSleeping(_int) && this.sleepStatus.areEnoughDeepSleeping(_int, this.players)) { // KioCG - sleep any time
             // Paper start - create time skip event - move up calculations
             final long newDayTime = this.levelData.getDayTime() + 24000L;
             org.bukkit.event.world.TimeSkipEvent event = new org.bukkit.event.world.TimeSkipEvent(
@@ -749,6 +_,7 @@
         this.updateSkyBrightness();
         if (runsNormally) {
             this.tickTime();
+            this.gameEventDispatcher.tick(this.getGameTime()); // KioCG - optimize game event
         }
 
         profilerFiller.push("tickPending");
@@ -857,7 +_,12 @@
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> player.stopSleepInBed(false, false));
+        // KioCG start - sleep progress
+        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> {
+            player.stopSleepInBed(false, false);
+            player.resetStat(net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.TIME_SINCE_REST));
+        });
+        // KioCG end - sleep progress
     }
 
     // Paper start - optimise random ticking
@@ -1055,6 +_,7 @@
     }
 
     private void announceSleepStatus() {
+        if (true) return; // KioCG - sleep any time
         if (this.canSleepThroughNights()) {
             if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
@@ -1272,11 +_,17 @@
         entity.totalEntityAge++; // Paper - age-like counter for all entities
         profilerFiller.push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(entity.getType()).toString());
         profilerFiller.incrementCounter("tickNonPassenger");
-        final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
+        final boolean isActive = entity.defaultActivationState || io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2 // KioCG - optimize entity activation
+        // KioCG start - chunk hot
+        LevelChunk levelChunk = entity.shouldTickHot() ? this.getChunkIfLoaded(entity.blockPosition()) : null;
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking();
+        try {
+        // KioCG end - chunk hot
         if (isActive) { // Paper - EAR 2
         entity.tick();
         entity.postTick(); // CraftBukkit
         } else {entity.inactiveTick();} // Paper - EAR 2
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().endTickingAndCount(); } // KioCG - chunk hot
         profilerFiller.pop();
 
         for (Entity entity1 : entity.getPassengers()) {
@@ -1301,6 +_,11 @@
             ProfilerFiller profilerFiller = Profiler.get();
             profilerFiller.push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(passengerEntity.getType()).toString());
             profilerFiller.incrementCounter("tickPassenger");
+            // KioCG start - chunk hot
+            LevelChunk levelChunk = passengerEntity.shouldTickHot() ? this.getChunkIfLoaded(passengerEntity.blockPosition()) : null;
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking();
+            try {
+            // KioCG end - chunk hot
             // Paper start - EAR 2
             if (isActive) {
             passengerEntity.rideTick();
@@ -1312,6 +_,7 @@
                 ridingEntity.positionRider(passengerEntity);
             }
             // Paper end - EAR 2
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().endTickingAndCount(); } // KioCG - chunk hot
             profilerFiller.pop();
 
             for (Entity entity : passengerEntity.getPassengers()) {
@@ -2691,6 +_,7 @@
 
             entity.updateDynamicGameEventListener(DynamicGameEventListener::add);
             entity.inWorld = true; // CraftBukkit - Mark entity as in world
+            entity.getBukkitEntity().virtualEntity = null; // KioCG - send virtual entity api
             entity.valid = true; // CraftBukkit
             ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server
             // Paper start - Entity origin API
