--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -393,6 +_,7 @@
     public com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent playerNaturallySpawnedEvent; // Paper - PlayerNaturallySpawnCreaturesEvent
     public @Nullable String clientBrandName = null; // Paper - Brand support
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
+    public boolean inStructure; // KioCG - player in structure api
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
@@ -919,11 +_,16 @@
 
             if (this.totalExperience != this.lastSentExp) {
                 this.lastSentExp = this.totalExperience;
-                this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
+                // this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel)); // KioCG - the long dark
             }
 
             if (this.tickCount % 20 == 0) {
                 CriteriaTriggers.LOCATION.trigger(this);
+                // KioCG start - player in structure api
+                if (this.serverLevel().isLoaded(this.blockPosition())) {
+                    this.inStructure = this.serverLevel().structureManager().hasAnyStructureAtExact(this.blockPosition());
+                }
+                // KioCG end - player in structure api
             }
 
             // CraftBukkit start - initialize oldLevel, fire PlayerLevelChangeEvent, and tick client-sided world border
@@ -1065,11 +_,31 @@
             return;
         }
         List<DefaultDrop> loot = new java.util.ArrayList<>(this.getInventory().getContainerSize()); // Paper - Restore vanilla drops behavior
+        List<org.bukkit.inventory.ItemStack> deathKeeps = new java.util.ArrayList<>(this.getInventory().getContainerSize()); // KioCG - death drop protect
         boolean keepInventory = this.serverLevel().getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) || this.isSpectator();
         if (!keepInventory) {
-            for (ItemStack item : this.getInventory().getContents()) {
+            // KioCG start - death drop protect
+            List<ItemStack> contents = this.getInventory().getContents();
+            for (int i = 0; i < contents.size(); ++i) {
+                ItemStack item = contents.get(i);
+            // KioCG end - death drop protect
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, net.minecraft.world.item.enchantment.EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG start - death drop protect
+                    if (this.level().paperConfig().kiocgConfig.deathKeepHotbarAndArmorItems) {
+                        if (i < Inventory.SELECTION_SIZE || i >= Inventory.INVENTORY_SIZE) {
+                            deathKeeps.add(item.getBukkitStack());
+                            continue;
+                        }
+                    }
+
+                    loot.add(new DefaultDrop(item, stack -> {
+                        ItemEntity droped = this.drop(stack, true, false, false);
+                        if (droped != null) {
+                            if (this.level().paperConfig().kiocgConfig.deathItemsOnlyPickupByOneself) droped.setTarget(this.getUUID());
+                            droped.age = -32768 + 1; // -32768代表无限生命周期
+                        }
+                    })); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG end - death drop protect
                 }
             }
         }
@@ -1085,7 +_,7 @@
 
         String deathmessage = defaultMessage.getString();
         this.keepLevel = keepInventory; // SPIGOT-2222: pre-set keepLevel
-        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, cause, loot, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), keepInventory); // Paper - Adventure
+        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, cause, loot, deathKeeps, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), keepInventory); // Paper - Adventure // KioCG - death drop protect
         // Paper start - cancellable death event
         if (event.isCancelled()) {
             // make compatible with plugins that might have already set the health in an event listener
@@ -1155,6 +_,8 @@
         this.setCamera(this); // Remove spectated target
         // CraftBukkit end
 
+        if (this.useItem.has(DataComponents.CONSUMABLE)) this.stopUsingItem(); // KioCG - fix MC-133218
+
         this.level().getCraftServer().getScoreboardManager().forAllObjectives(ObjectiveCriteria.DEATH_COUNT, this, ScoreAccess::increment); // CraftBukkit - Get our scores instead
         LivingEntity killCredit = this.getKillCredit();
         if (killCredit != null) {
@@ -2049,6 +_,8 @@
             // this.getAttributes().assignPermanentModifiers(that.getAttributes()); // CraftBukkit
             this.setHealth(that.getHealth());
             this.foodData = that.foodData;
+            this.coldData = that.coldData; // KioCG - the long dark
+            this.thirstData = that.thirstData; // KioCG - the long dark
 
             for (MobEffectInstance mobEffectInstance : that.getActiveEffects()) {
                 // this.addEffect(new MobEffectInstance(mobEffectInstance)); // CraftBukkit
@@ -3051,6 +_,8 @@
         this.setRemainingFireTicks(0);
         this.fallDistance = 0;
         this.foodData = new net.minecraft.world.food.FoodData();
+        this.coldData = new com.kiocg.player.ColdData(); // KioCG - the long dark
+        this.thirstData = new com.kiocg.player.ThirstData(); // KioCG - the long dark
         this.experienceLevel = this.newLevel;
         this.totalExperience = this.newTotalExp;
         this.experienceProgress = 0;
