--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -721,7 +_,7 @@
         }
 
         int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
-        if (this.sleepStatus.areEnoughSleeping(_int) && this.sleepStatus.areEnoughDeepSleeping(_int, this.players)) {
+        if (false && this.sleepStatus.areEnoughSleeping(_int) && this.sleepStatus.areEnoughDeepSleeping(_int, this.players)) { // KioCG - sleep any time
             // Paper start - create time skip event - move up calculations
             final long newDayTime = this.levelData.getDayTime() + 24000L;
             org.bukkit.event.world.TimeSkipEvent event = new org.bukkit.event.world.TimeSkipEvent(
@@ -749,6 +_,21 @@
         this.updateSkyBrightness();
         if (runsNormally) {
             this.tickTime();
+
+            // KioCG start - keep block destruction
+            // 客户端会清理400刻之后的方块破坏纹理
+            if (this.lagCompensationTick % 20 == 0) {
+                java.util.Iterator<BlockDestructionProgress> iterator = this.destroyingBlocks.values().iterator();
+
+                while (iterator.hasNext()) {
+                    BlockDestructionProgress blockDestructionProgress = iterator.next();
+                    int i = blockDestructionProgress.getUpdatedRenderTick();
+                    if (this.lagCompensationTick - i > 400) {
+                        iterator.remove();
+                    }
+                }
+            }
+            // KioCG end - keep block destruction
         }
 
         profilerFiller.push("tickPending");
@@ -857,13 +_,18 @@
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> player.stopSleepInBed(false, false));
+        // KioCG start - sleep progress
+        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> {
+            player.stopSleepInBed(false, false);
+            player.resetStat(net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.TIME_SINCE_REST));
+        });
+        // KioCG end - sleep progress
     }
 
     // Paper start - optimise random ticking
     private final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom simpleRandom = new ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom(net.minecraft.world.level.levelgen.RandomSupport.generateUniqueSeed());
 
-    private void optimiseRandomTick(final LevelChunk chunk, final int tickSpeed) {
+    private void optimiseRandomTick(final LevelChunk chunk, final int tickSpeed, final int daysSpent) { // KioCG - crops grow every day
         final LevelChunkSection[] sections = chunk.getSections();
         final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection((ServerLevel)(Object)this);
         final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom simpleRandom = this.simpleRandom;
@@ -908,6 +_,38 @@
             }
         }
 
+        // KioCG start - crops grow every day
+        if (daysSpent > 0)
+        for (int i = 0; i < daysSpent; ++i) {
+        // 放在外层循环, 防止交界处的多格作物生长两次
+        List<Pair<BlockState, BlockPos>> dayTickList = new ArrayList<>();
+        for (int sectionIndex = 0, sectionsLen = sections.length; sectionIndex < sectionsLen; sectionIndex++) {
+            final int offsetY = (sectionIndex + minSection) << 4;
+            final LevelChunkSection section = sections[sectionIndex];
+            final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> states = section.states;
+            if (!section.isDailyTickingBlocks()) {
+                continue;
+            }
+
+            final ca.spottedleaf.moonrise.common.list.ShortList tickList = section.dailyTickingBlocks;
+
+            final int tickingBlocks = tickList.size();
+            for (int index = 0; index < tickingBlocks; ++index) {
+                final int location = (int)tickList.getRaw(index) & 0xFFFF;
+                final BlockState state = states.get(location);
+
+                // do not use a mutable pos, as some random tick implementations store the input without calling immutable()!
+                final BlockPos pos = new BlockPos((location & 15) | offsetX, ((location >>> (4 + 4)) & 15) | offsetY, ((location >>> 4) & 15) | offsetZ);
+
+                dayTickList.add(Pair.of(state, pos));
+            }
+        }
+        for (Pair<BlockState, BlockPos> ticking : dayTickList) {
+            ticking.getFirst().dayTick((ServerLevel)(Object)this, ticking.getSecond(), simpleRandom);
+        }
+        }
+        // KioCG end - crops grow every day
+
         return;
     }
     // Paper end - optimise random ticking
@@ -928,9 +_,35 @@
         }
         } // Paper - Option to disable ice and snow
 
+        // KioCG start - crops grow every day
+        int daysSpent = 0;
+        int days = this.getDays();
+        if (chunk.chunkDays.getDays() != days) {
+            daysSpent = Mth.clamp(days - chunk.chunkDays.getDays(), 0, 7);
+            chunk.chunkDays.setDays(days);
+        }
+        // KioCG end - crops grow every day
+
+        // KioCG start - mobs disappear every day
+        if (chunk.chunkDays.getNights() != days && (int)(this.getDayTime() % 24000L) > 13800) {
+            chunk.chunkDays.setNights(days);
+
+            final int chunkX = chunk.getPos().x;
+            final int chunkZ = chunk.getPos().z;
+            ca.spottedleaf.moonrise.patches.chunk_system.level.entity.ChunkEntitySlices slices = this.moonrise$getEntityLookup().getChunk(chunkX, chunkZ);
+            if (slices != null) {
+                slices.getAllEntities().forEach(entity -> {
+                    if (entity instanceof Mob mob) {
+                        mob.checkDespawn(true);
+                    }
+                });
+            }
+        }
+        // KioCG end - mobs disappear every day
+
         profilerFiller.popPush("tickBlocks");
         if (randomTickSpeed > 0) {
-            this.optimiseRandomTick(chunk, randomTickSpeed); // Paper - optimise random ticking
+            this.optimiseRandomTick(chunk, randomTickSpeed, daysSpent); // Paper - optimise random ticking // KioCG - crops grow every day
         }
 
         profilerFiller.pop();
@@ -1055,6 +_,7 @@
     }
 
     private void announceSleepStatus() {
+        if (true) return; // KioCG - sleep any time
         if (this.canSleepThroughNights()) {
             if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
@@ -1272,11 +_,17 @@
         entity.totalEntityAge++; // Paper - age-like counter for all entities
         profilerFiller.push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(entity.getType()).toString());
         profilerFiller.incrementCounter("tickNonPassenger");
-        final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
+        final boolean isActive = entity.defaultActivationState || io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2 // KioCG - optimize entity activation
+        // KioCG start - chunk hot
+        LevelChunk levelChunk = entity.shouldTickHot() ? this.getChunkIfLoaded(entity.blockPosition()) : null;
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking();
+        try {
+        // KioCG end - chunk hot
         if (isActive) { // Paper - EAR 2
         entity.tick();
         entity.postTick(); // CraftBukkit
         } else {entity.inactiveTick();} // Paper - EAR 2
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().endTickingAndCount(); } // KioCG - chunk hot
         profilerFiller.pop();
 
         for (Entity entity1 : entity.getPassengers()) {
@@ -1301,6 +_,11 @@
             ProfilerFiller profilerFiller = Profiler.get();
             profilerFiller.push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(passengerEntity.getType()).toString());
             profilerFiller.incrementCounter("tickPassenger");
+            // KioCG start - chunk hot
+            LevelChunk levelChunk = passengerEntity.shouldTickHot() ? this.getChunkIfLoaded(passengerEntity.blockPosition()) : null;
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking();
+            try {
+            // KioCG end - chunk hot
             // Paper start - EAR 2
             if (isActive) {
             passengerEntity.rideTick();
@@ -1312,6 +_,7 @@
                 ridingEntity.positionRider(passengerEntity);
             }
             // Paper end - EAR 2
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().endTickingAndCount(); } // KioCG - chunk hot
             profilerFiller.pop();
 
             for (Entity entity : passengerEntity.getPassengers()) {
@@ -1609,8 +_,28 @@
     }
     // CraftBukkit end
 
+    // KioCG start - keep block destruction
+    public void abortDestroyBlock(int breakerId, BlockPos pos, int destroyTick) {
+        if (this.destroyingBlocks.containsKey(pos) && destroyTick < 20 * 1.5) {
+            this.destroyBlockProgress(breakerId, pos, -1);
+        }
+    }
+    // KioCG end - keep block destruction
+
     @Override
     public void destroyBlockProgress(int breakerId, BlockPos pos, int progress) {
+        // KioCG start - keep block destruction
+        BlockDestructionProgress blockDestructionProgress;
+        if (progress > -1) {
+            blockDestructionProgress = this.destroyingBlocks.computeIfAbsent(pos, k -> new BlockDestructionProgress(BlockDestructionProgress.getNextBlockDestroyId(), pos));
+            blockDestructionProgress.setProgress(progress);
+            blockDestructionProgress.updateTick((int) this.getLagCompensationTick());
+        } else {
+            blockDestructionProgress = this.destroyingBlocks.remove(pos);
+        }
+        int fakeBreakerId = blockDestructionProgress.getId();
+        // KioCG end - keep block destruction
+
         // CraftBukkit start
         Player breakerPlayer = null;
         Entity entity = this.getEntity(breakerId);
@@ -1628,7 +_,7 @@
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
         for (ServerPlayer serverPlayer : this.server.getPlayerList().getPlayers()) {
-            if (serverPlayer != null && serverPlayer.level() == this && serverPlayer.getId() != breakerId) {
+            if (serverPlayer != null && serverPlayer.level() == this) { // KioCG - keep block destruction
                 double d = pos.getX() - serverPlayer.getX();
                 double d1 = pos.getY() - serverPlayer.getY();
                 double d2 = pos.getZ() - serverPlayer.getZ();
@@ -1638,7 +_,7 @@
                 }
                 // CraftBukkit end
                 if (d * d + d1 * d1 + d2 * d2 < 1024.0) {
-                    serverPlayer.connection.send(new ClientboundBlockDestructionPacket(breakerId, pos, progress));
+                    serverPlayer.connection.send(new ClientboundBlockDestructionPacket(fakeBreakerId, pos, Math.min(progress, 9))); // KioCG - keep block destruction: 破坏程度为10客户端不显示
                 }
             }
         }
@@ -1877,7 +_,7 @@
         Explosion.BlockInteraction blockInteraction = switch (explosionInteraction) {
             case NONE -> Explosion.BlockInteraction.KEEP;
             case BLOCK -> this.getDestroyType(GameRules.RULE_BLOCK_EXPLOSION_DROP_DECAY);
-            case MOB -> this.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            case MOB -> this.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1) // KioCG - better mob griefing setting
                 ? this.getDestroyType(GameRules.RULE_MOB_EXPLOSION_DROP_DECAY)
                 : Explosion.BlockInteraction.KEEP;
             case TNT -> this.getDestroyType(GameRules.RULE_TNT_EXPLOSION_DROP_DECAY);
@@ -2691,6 +_,7 @@
 
             entity.updateDynamicGameEventListener(DynamicGameEventListener::add);
             entity.inWorld = true; // CraftBukkit - Mark entity as in world
+            entity.getBukkitEntity().virtualEntity = null; // KioCG - send virtual entity api
             entity.valid = true; // CraftBukkit
             ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server
             // Paper start - Entity origin API
