--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -726,6 +_,7 @@
             profilerFiller.pop();
         }
 
+        if (!this.paperConfig().kiocgConfig.sleepAnyTime) { // KioCG - sleep any time
         int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
         if (this.sleepStatus.areEnoughSleeping(_int) && this.sleepStatus.areEnoughDeepSleeping(_int, this.players)) {
             // Paper start - create time skip event - move up calculations
@@ -751,6 +_,7 @@
                 this.resetWeatherCycle();
             }
         }
+        } // KioCG - sleep any time
 
         this.updateSkyBrightness();
         if (runsNormally) {
@@ -914,6 +_,44 @@
             }
         }
 
+        // KioCG start - crops grow every day
+        int days = this.getDays();
+        if (chunk.chunkDays != days) {
+            int daysSpent = Mth.clamp(days - chunk.chunkDays, 0, 7);
+            chunk.chunkDays = days;
+
+        for (int sectionIndex = 0, sectionsLen = sections.length; sectionIndex < sectionsLen; sectionIndex++) {
+            final int offsetY = (sectionIndex + minSection) << 4;
+            final LevelChunkSection section = sections[sectionIndex];
+            final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> states = section.states;
+            if (!section.isDailyTickingBlocks()) {
+                continue;
+            }
+
+            final ca.spottedleaf.moonrise.common.list.ShortList tickList = section.dailyTickingBlocks;
+
+            for (int i = 0; i < daysSpent; ++i) {
+                final int tickingBlocks = tickList.size();
+                if (tickingBlocks == 0) break;
+
+                List<Integer> dayTickList = new ArrayList<>(tickingBlocks);
+                for (int index = 0; index < tickingBlocks; ++index) {
+                    final int location = (int)tickList.getRaw(index) & 0xFFFF;
+                    dayTickList.add(location);
+                }
+                for (Integer location : dayTickList) {
+                    final BlockState state = states.get(location);
+
+                    // do not use a mutable pos, as some random tick implementations store the input without calling immutable()!
+                    final BlockPos pos = new BlockPos((location & 15) | offsetX, ((location >>> (4 + 4)) & 15) | offsetY, ((location >>> 4) & 15) | offsetZ);
+
+                    state.dayTick((ServerLevel)(Object)this, pos, simpleRandom);
+                }
+            }
+        }
+        }
+        // KioCG end - crops grow every day
+
         return;
     }
     // Paper end - optimise random ticking
@@ -1052,6 +_,7 @@
     }
 
     private void announceSleepStatus() {
+        if (this.paperConfig().kiocgConfig.sleepAnyTime) return; // KioCG - sleep any time
         if (this.canSleepThroughNights()) {
             if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
@@ -1833,7 +_,7 @@
         Explosion.BlockInteraction blockInteraction = switch (explosionInteraction) {
             case NONE -> Explosion.BlockInteraction.KEEP;
             case BLOCK -> this.getDestroyType(GameRules.RULE_BLOCK_EXPLOSION_DROP_DECAY);
-            case MOB -> this.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            case MOB -> this.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1) // KioCG - better mob griefing setting
                 ? this.getDestroyType(GameRules.RULE_MOB_EXPLOSION_DROP_DECAY)
                 : Explosion.BlockInteraction.KEEP;
             case TNT -> this.getDestroyType(GameRules.RULE_TNT_EXPLOSION_DROP_DECAY);
