--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -393,6 +_,8 @@
     public com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent playerNaturallySpawnedEvent; // Paper - PlayerNaturallySpawnCreaturesEvent
     public @Nullable String clientBrandName = null; // Paper - Brand support
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
+    public boolean inStructure; // KioCG - player in structure api
+    private long nearbyChunkHot = 0; public long getNearbyChunkHot() { return this.nearbyChunkHot; } // KioCG - chunk hot
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
@@ -782,6 +_,7 @@
 
     @Override
     public void onInsideBlock(BlockState state) {
+        if (!state.isAir()) // KioCG - optimize trigger advancements
         CriteriaTriggers.ENTER_BLOCK.trigger(this, state);
     }
 
@@ -829,6 +_,7 @@
         }
 
         CriteriaTriggers.TICK.trigger(this);
+        if (this.tickCount % (20 * 60) == 0) CriteriaTriggers.MINUTE.trigger(this); // KioCG - optimize trigger advancements
         if (this.levitationStartPos != null) {
             CriteriaTriggers.LEVITATION.trigger(this, this.levitationStartPos, this.tickCount - this.levitationStartTime);
         }
@@ -837,7 +_,54 @@
         this.trackEnteredOrExitedLavaOnVehicle();
         this.updatePlayerAttributes();
         this.advancements.flushDirty(this);
-    }
+
+        // KioCG start - chunk hot
+        if (this.server.getTickCount() % 20 == 1) {
+            this.wasTooHot = false;
+            this.nearbyChunkHot = this.refreshNearbyChunkHot();
+            if (this.nearbyChunkHot >= this.level().paperConfig().kiocgConfig.chunkHot.phaseGood) {
+                if (this.nearbyChunkHot >= this.level().paperConfig().kiocgConfig.chunkHot.phaseMedium) {
+                    this.wasTooHot = true;
+                    if (!this.hasEffect(MobEffects.FIRE_RESISTANCE)) {
+                        this.hurtServer(serverLevel(), this.damageSources().generic(), 2.0F);
+                    }
+                    this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(this.level().paperConfig().kiocgConfig.chunkHot.lowMessage)), true);
+                } else {
+                    this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(this.level().paperConfig().kiocgConfig.chunkHot.mediumMessage)), true);
+                }
+            }
+        }
+        // KioCG end - chunk hot
+    }
+
+    // KioCG start - chunk hot
+    private boolean wasTooHot = false;
+    @Override
+    public void setSharedFlagOnFire(boolean isOnFire) {
+        if (wasTooHot) {
+            this.setSharedFlag(FLAG_ONFIRE, true);
+            return;
+        }
+
+        super.setSharedFlagOnFire(isOnFire);
+    }
+
+    private long refreshNearbyChunkHot() {
+        int x = SectionPos.blockToSectionCoord(this.getBlockX());
+        int z = SectionPos.blockToSectionCoord(this.getBlockZ());
+        net.minecraft.world.level.Level level = this.level();
+        long total = 0L;
+        for (int i = x - 4; i <= x + 4; ++i) {
+            for (int j = z - 4; j <= z + 4; ++j) {
+                net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(i, j);
+                if (chunk != null) {
+                    total += chunk.getChunkHot().getAverage();
+                }
+            }
+        }
+        return total;
+    }
+    // KioCG end - chunk hot
 
     private void updatePlayerAttributes() {
         AttributeInstance attribute = this.getAttribute(Attributes.BLOCK_INTERACTION_RANGE);
@@ -919,11 +_,16 @@
 
             if (this.totalExperience != this.lastSentExp) {
                 this.lastSentExp = this.totalExperience;
-                this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
+                // this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel)); // KioCG - the long dark
             }
 
             if (this.tickCount % 20 == 0) {
                 CriteriaTriggers.LOCATION.trigger(this);
+                // KioCG start - player in structure api
+                if (this.serverLevel().isLoaded(this.blockPosition())) {
+                    this.inStructure = this.serverLevel().structureManager().hasAnyStructureAtExact(this.blockPosition());
+                }
+                // KioCG end - player in structure api
             }
 
             // CraftBukkit start - initialize oldLevel, fire PlayerLevelChangeEvent, and tick client-sided world border
@@ -1065,11 +_,31 @@
             return;
         }
         List<DefaultDrop> loot = new java.util.ArrayList<>(this.getInventory().getContainerSize()); // Paper - Restore vanilla drops behavior
+        List<org.bukkit.inventory.ItemStack> deathKeeps = new java.util.ArrayList<>(this.getInventory().getContainerSize()); // KioCG - death drop protect
         boolean keepInventory = this.serverLevel().getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) || this.isSpectator();
         if (!keepInventory) {
-            for (ItemStack item : this.getInventory().getContents()) {
+            // KioCG start - death drop protect
+            List<ItemStack> contents = this.getInventory().getContents();
+            for (int i = 0; i < contents.size(); ++i) {
+                ItemStack item = contents.get(i);
+            // KioCG end - death drop protect
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, net.minecraft.world.item.enchantment.EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG start - death drop protect
+                    if (this.level().paperConfig().kiocgConfig.deathKeepHotbarAndArmorItems) {
+                        if (i < Inventory.SELECTION_SIZE || i >= Inventory.INVENTORY_SIZE) {
+                            deathKeeps.add(item.getBukkitStack());
+                            continue;
+                        }
+                    }
+
+                    loot.add(new DefaultDrop(item, stack -> {
+                        ItemEntity droped = this.drop(stack, true, false, false);
+                        if (droped != null) {
+                            if (this.level().paperConfig().kiocgConfig.deathItemsOnlyPickupByOneself) droped.setTarget(this.getUUID());
+                            droped.age = -32768 + 1; // -32768代表无限生命周期
+                        }
+                    })); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG end - death drop protect
                 }
             }
         }
@@ -1085,7 +_,7 @@
 
         String deathmessage = defaultMessage.getString();
         this.keepLevel = keepInventory; // SPIGOT-2222: pre-set keepLevel
-        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, cause, loot, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), keepInventory); // Paper - Adventure
+        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, cause, loot, deathKeeps, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), keepInventory); // Paper - Adventure // KioCG - death drop protect
         // Paper start - cancellable death event
         if (event.isCancelled()) {
             // make compatible with plugins that might have already set the health in an event listener
@@ -1155,6 +_,8 @@
         this.setCamera(this); // Remove spectated target
         // CraftBukkit end
 
+        if (this.useItem.has(DataComponents.CONSUMABLE)) this.stopUsingItem(); // KioCG - fix MC-133218
+
         this.level().getCraftServer().getScoreboardManager().forAllObjectives(ObjectiveCriteria.DEATH_COUNT, this, ScoreAccess::increment); // CraftBukkit - Get our scores instead
         LivingEntity killCredit = this.getKillCredit();
         if (killCredit != null) {
@@ -1551,7 +_,7 @@
             return Either.left(Player.BedSleepingProblem.OBSTRUCTED);
         } else {
             this.setRespawnPosition(this.level().dimension(), at, this.getYRot(), false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED); // Paper - Add PlayerSetSpawnEvent
-            if (this.level().isDay()) {
+            if (false && this.level().isDay()) { // KioCG - sleep any time
                 return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
             } else {
                 if (!this.isCreative()) {
@@ -1600,7 +_,7 @@
                     this.awardStat(Stats.SLEEP_IN_BED);
                     CriteriaTriggers.SLEPT_IN_BED.trigger(this);
                 });
-                if (!this.serverLevel().canSleepThroughNights()) {
+                if (false && !this.serverLevel().canSleepThroughNights()) { // KioCG - sleep any time
                     this.displayClientMessage(Component.translatable("sleep.not_possible"), true);
                 }
 
@@ -1612,10 +_,27 @@
 
     @Override
     public void startSleeping(BlockPos pos) {
-        this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_REST));
+        // this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)); // KioCG - sleep progress
         super.startSleeping(pos);
     }
 
+    // KioCG start - sleep progress
+    public void sendSleepMessage(int value, int maxValue) {
+        String content = io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.sleepProgressContent;
+        float progress = (float) (maxValue - value) / maxValue;
+        int split = (int) (content.length() * progress);
+        String substring1 = content.substring(0, split);
+        String substring2 = content.substring(split);
+        net.kyori.adventure.text.Component result = net.kyori.adventure.text.Component.empty()
+            .append(io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.sleepProgressLeft)
+            .append(net.kyori.adventure.text.Component.text(substring1))
+            .append(io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.sleepProgressClock)
+            .append(net.kyori.adventure.text.Component.text(substring2))
+            .append(io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.sleepProgressRight);
+        this.getBukkitEntity().sendActionBar(result);
+    }
+    // KioCG end - sleep progress
+
     private boolean bedInRange(BlockPos pos, Direction direction) {
         return this.isReachableBedBlock(pos) || this.isReachableBedBlock(pos.relative(direction.getOpposite()));
     }
@@ -1701,9 +_,9 @@
     }
 
     @Override
-    protected void pushEntities() {
+    protected void pushEntities(boolean moved) { // KioCG start - optimize entity collide
         if (this.level().tickRateManager().runsNormally()) {
-            super.pushEntities();
+            super.pushEntities(true); // KioCG start - optimize entity collide
         }
     }
 
@@ -2049,6 +_,8 @@
             // this.getAttributes().assignPermanentModifiers(that.getAttributes()); // CraftBukkit
             this.setHealth(that.getHealth());
             this.foodData = that.foodData;
+            this.coldData = that.coldData; // KioCG - the long dark
+            this.thirstData = that.thirstData; // KioCG - the long dark
 
             for (MobEffectInstance mobEffectInstance : that.getActiveEffects()) {
                 // this.addEffect(new MobEffectInstance(mobEffectInstance)); // CraftBukkit
@@ -2546,7 +_,7 @@
             // Paper end - Add PlayerSetSpawnEvent
 
             if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper - Add PlayerSetSpawnEvent
-                this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getNotification())); // Paper - Add PlayerSetSpawnEvent
+                this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getNotification()), true); // Paper - Add PlayerSetSpawnEvent // KioCG - sleep any time: 使用操作栏提示
             }
 
             this.respawnPosition = position;
@@ -3051,6 +_,8 @@
         this.setRemainingFireTicks(0);
         this.fallDistance = 0;
         this.foodData = new net.minecraft.world.food.FoodData();
+        this.coldData = new com.kiocg.entity.player.ColdData(); // KioCG - the long dark
+        this.thirstData = new com.kiocg.entity.player.ThirstData(); // KioCG - the long dark
         this.experienceLevel = this.newLevel;
         this.totalExperience = this.newTotalExp;
         this.experienceProgress = 0;
