--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -73,7 +_,7 @@
                 return event; // Paper - Expand PlayerGameModeChangeEvent
             }
             // CraftBukkit end
-            this.setGameModeForPlayer(gameModeForPlayer, this.gameModeForPlayer); // Paper - Fix MC-259571
+            this.setGameModeForPlayer(gameModeForPlayer, this.gameModeForPlayer, true); // Paper - Fix MC-259571 // KioCG - keep player abilities
             this.player.onUpdateAbilities();
             this.level
                 .getServer()
@@ -88,12 +_,33 @@
         }
     }
 
-    protected void setGameModeForPlayer(GameType gameModeForPlayer, @Nullable GameType previousGameModeForPlayer) {
+    protected void setGameModeForPlayer(GameType gameModeForPlayer, @Nullable GameType previousGameModeForPlayer, boolean updateAbilities) { // KioCG - keep player abilities
         this.previousGameModeForPlayer = previousGameModeForPlayer;
         this.gameModeForPlayer = gameModeForPlayer;
+        if (updateAbilities) // KioCG - keep player abilities
         gameModeForPlayer.updatePlayerAbilities(this.player.getAbilities());
     }
 
+    // KioCG start - keep player abilities: should keep flying
+    public void checkFlying() {
+        if (player.getAbilities().flying && this.gameModeForPlayer != GameType.SPECTATOR) {
+            net.minecraft.world.phys.AABB boundingBox = player.getBoundingBox();
+            boundingBox = new net.minecraft.world.phys.AABB(
+                    boundingBox.minX,
+                    boundingBox.minY - 1.0E-5,
+                    boundingBox.minZ,
+                    boundingBox.maxX,
+                    boundingBox.minY,
+                    boundingBox.maxZ
+            );
+            if (!player.level().noCollision(player, boundingBox)) {
+                player.getAbilities().flying = false;
+                player.onUpdateAbilities();
+            }
+        }
+    }
+    // KioCG end - keep player abilities: should keep flying
+
     public GameType getGameModeForPlayer() {
         return this.gameModeForPlayer;
     }
@@ -138,6 +_,15 @@
                 this.lastSentState = -1;
                 this.isDestroyingBlock = false;
             } else {
+                // KioCG start - keep block destruction
+                if (this.level.destroyingBlocks.get(this.destroyPos).getProgress() >= 10) {
+                    this.level.destroyBlockProgress(this.player.getId(), this.destroyPos, -1);
+                    this.lastSentState = -1;
+                    this.isDestroyingBlock = false;
+                    this.destroyBlock(this.destroyPos);
+                    this.player.connection.send(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(net.minecraft.world.level.block.LevelEvent.PARTICLES_DESTROY_BLOCK, this.destroyPos, net.minecraft.world.level.block.Block.getId(blockState), false));
+                } else
+                // KioCG end - keep block destruction
                 this.incrementDestroyProgress(blockState, this.destroyPos, this.destroyProgressStart);
             }
         }
@@ -246,9 +_,16 @@
                         this.debugLogging(pos, false, sequence, "abort destroying since another started (client insta mine, server disagreed)");
                     }
 
+                    // KioCG start - keep block destruction
+                    int progress = this.level.destroyingBlocks.get(pos).getProgress();
+                    if (progress > 0) {
+                        this.destroyProgressStart -= (int) (progress / 10.0F / f);
+                    }
+                    // KioCG end - keep block destruction
+
                     this.isDestroyingBlock = true;
                     this.destroyPos = pos.immutable();
-                    int i = (int)(f * 10.0F);
+                    int i = (int)(f * 10.0F) + Math.max(progress, 0); // KioCG - keep block destruction
                     this.level.destroyBlockProgress(this.player.getId(), pos, i);
                     this.debugLogging(pos, true, sequence, "actual start of destroying");
                     this.lastSentState = i;
@@ -283,14 +_,14 @@
                     ServerPlayerGameMode.LOGGER.debug("Mismatch in destroy block pos: {} {}", this.destroyPos, pos); // CraftBukkit - SPIGOT-5457 sent by client when interact event cancelled
                     BlockState type = this.level.getBlockStateIfLoaded(this.destroyPos); // Don't load unloaded chunks for stale records here
                     if (type != null) {
-                    this.level.destroyBlockProgress(this.player.getId(), this.destroyPos, -1);
+                    this.level.abortDestroyBlock(this.player.getId(), this.destroyPos, -1); // KioCG - keep block destruction
                     this.debugLogging(pos, true, sequence, "aborted mismatched destroying");
                     }
                     this.destroyPos = BlockPos.ZERO;
                     // Paper end - Don't allow digging into unloaded chunks
                 }
 
-                this.level.destroyBlockProgress(this.player.getId(), pos, -1);
+                this.level.abortDestroyBlock(this.player.getId(), pos, -1); // KioCG - keep block destruction
                 this.debugLogging(pos, true, sequence, "aborted destroying");
 
                 org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDamageAbortEvent(this.player, pos, this.player.getInventory().getSelectedItem()); // CraftBukkit
@@ -417,11 +_,49 @@
     }
 
     public InteractionResult useItem(ServerPlayer player, Level level, ItemStack stack, InteractionHand hand) {
+        // KioCG start - PlayerItemUseEvent
+        return this.useItem(player, level, stack, hand, null, false, null);
+    }
+    public InteractionResult useItem(ServerPlayer player, Level level, ItemStack stack, InteractionHand hand, net.minecraft.world.phys.BlockHitResult hitResult, boolean interactCancelled, org.bukkit.event.player.PlayerInteractEvent interactEvent) {
+        // KioCG end - PlayerItemUseEvent
         if (this.gameModeForPlayer == GameType.SPECTATOR) {
             return InteractionResult.PASS;
         } else if (player.getCooldowns().isOnCooldown(stack)) {
             return InteractionResult.PASS;
         } else {
+            // KioCG start - PlayerItemUseEvent: 复制PlayerInteractEvent的行为
+            boolean cancelled;
+            if (hitResult == null || hitResult.getType() != net.minecraft.world.phys.HitResult.Type.BLOCK) {
+                com.kiocg.event.player.PlayerItemUseEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemUseEvent(this.player, org.bukkit.event.block.Action.RIGHT_CLICK_AIR, null, null, stack, hand, null, interactEvent);
+                cancelled = event.isCancelled();
+            } else {
+                if (this.player.gameMode.firedUseItem && this.player.gameMode.useItemPosition.equals(hitResult.getBlockPos()) && this.player.gameMode.useItemHand == hand && ItemStack.isSameItemSameComponents(this.player.gameMode.useItemItemStack, stack)) {
+                    cancelled = this.player.gameMode.useItemResult;
+                } else {
+                    com.kiocg.event.player.PlayerItemUseEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemUseEvent(this.player, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, hitResult.getBlockPos(), hitResult.getDirection(), stack, hand, hitResult.getLocation(), interactEvent);
+                    cancelled = event.isCancelled();
+                }
+                this.player.gameMode.firedUseItem = false;
+            }
+
+            if (cancelled || interactCancelled) {
+                this.player.resyncUsingItem(this.player); // Paper - Properly cancel usable items
+                this.player.containerMenu.sendAllDataToRemote(); // SPIGOT-2524
+                return InteractionResult.FAIL;
+            }
+            stack = this.player.getItemInHand(hand); // Update in case it was changed in the event
+            if (stack.isEmpty()) {
+                return InteractionResult.FAIL;
+            }
+            // KioCG end - PlayerItemUseEvent: 复制PlayerInteractEvent的行为
+
+            // KioCG start - customize default components
+            if (stack.has(net.minecraft.core.component.DataComponents.CANNOT_USE)) {
+                player.resyncUsingItem(player); // Paper - Properly cancel usable items
+                player.containerMenu.sendAllDataToRemote(); // SPIGOT-2524
+                return InteractionResult.FAIL;
+            }
+            // KioCG end - customize default components
             int count = stack.getCount();
             int damageValue = stack.getDamageValue();
             InteractionResult interactionResult = stack.use(level, player, hand);
@@ -460,6 +_,13 @@
     public BlockPos interactPosition;
     public InteractionHand interactHand;
     public ItemStack interactItemStack;
+    // KioCG start - PlayerItemUseEvent
+    public boolean useItemResult = false;
+    public boolean firedUseItem = false;
+    public BlockPos useItemPosition;
+    public InteractionHand useItemHand;
+    public ItemStack useItemItemStack;
+    // KioCG end - PlayerItemUseEvent
     public InteractionResult useItemOn(ServerPlayer player, Level level, ItemStack stack, InteractionHand hand, BlockHitResult hitResult) {
         BlockPos blockPos = hitResult.getBlockPos();
         BlockState blockState = level.getBlockState(blockPos);
@@ -491,6 +_,21 @@
             }
             player.getBukkitEntity().updateInventory(); // SPIGOT-2867
             this.player.resyncUsingItem(this.player); // Paper - Properly cancel usable items
+
+            // KioCG start - PlayerItemUseEvent
+            if (!stack.isEmpty()) {
+                com.kiocg.event.player.PlayerItemUseEvent event1 = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemUseEvent(player, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, blockPos, hitResult.getDirection(), stack, hand, hitResult.getLocation(), event);
+                this.firedUseItem = true;
+                this.useItemResult = event1.isCancelled();
+                this.useItemPosition = blockPos.immutable();
+                this.useItemHand = hand;
+                this.useItemItemStack = this.interactItemStack; // 不要重复克隆物品
+                if (event1.isCancelled()) {
+                    return InteractionResult.SUCCESS;
+                }
+            }
+            // KioCG end - PlayerItemUseEvent
+
             return (event.useItemInHand() != org.bukkit.event.Event.Result.ALLOW) ? InteractionResult.SUCCESS : InteractionResult.PASS;
         } else if (this.gameModeForPlayer == GameType.SPECTATOR) {
             MenuProvider menuProvider = blockState.getMenuProvider(level, blockPos);
@@ -518,6 +_,25 @@
                     }
                 }
             }
+
+            // KioCG start - PlayerItemUseEvent
+            if (!stack.isEmpty()) {
+                com.kiocg.event.player.PlayerItemUseEvent event1 = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemUseEvent(player, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, blockPos, hitResult.getDirection(), stack, hand, hitResult.getLocation(), event);
+                this.firedUseItem = true;
+                this.useItemResult = event1.isCancelled();
+                this.useItemPosition = blockPos.immutable();
+                this.useItemHand = hand;
+                this.useItemItemStack = this.interactItemStack; // 不要重复克隆物品
+                if (event1.isCancelled()) {
+                    // Paper start - Properly cancel usable items; Cancel only if cancelled + if the interact result is different from default response
+                    if (!cancelledItem) {
+                        this.player.resyncUsingItem(this.player);
+                    }
+                    // Paper end - Properly cancel usable items
+                    return InteractionResult.PASS;
+                }
+            }
+            // KioCG end - PlayerItemUseEvent
 
             if (!stack.isEmpty() && !this.interactResult) { // add !interactResult SPIGOT-764
                 UseOnContext useOnContext = new UseOnContext(player, hand, hitResult);
