--- a/net/minecraft/world/entity/decoration/BlockAttachedEntity.java
+++ b/net/minecraft/world/entity/decoration/BlockAttachedEntity.java
@@ -86,7 +_,7 @@
     public boolean hurtServer(ServerLevel level, DamageSource damageSource, float amount) {
         if (this.isInvulnerableToBase(damageSource)) {
             return false;
-        } else if (!level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && damageSource.getEntity() instanceof Mob) {
+        } else if (true && damageSource.getEntity() instanceof Mob) { // KioCG - better mob griefing setting
             return false;
         } else {
             if (!this.isRemoved()) {
