--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -25,9 +_,119 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
-public class LeashFenceKnotEntity extends BlockAttachedEntity {
+public class LeashFenceKnotEntity extends BlockAttachedEntity implements Leashable { // KioCG - leash fence
     public static final double OFFSET_Y = 0.375;
 
+    // KioCG start - leash fence
+    @Nullable
+    private Leashable.LeashData leashData;
+    @Nullable
+    private FakeLeashTracker fakeLeashTracker = null;
+
+    private class FakeLeashTracker {
+        private final java.util.Set<org.bukkit.entity.Player> closer = java.util.Collections.newSetFromMap(new java.util.WeakHashMap<>());
+        private final java.util.Set<org.bukkit.entity.Player> farAway = java.util.Collections.newSetFromMap(new java.util.WeakHashMap<>());
+        private final net.minecraft.world.entity.Mob entity;
+        private final org.bukkit.Location location;
+        private final double scale;
+        private final double ropeLeashOffset;
+        private final double entityLeashOffset;
+
+        private FakeLeashTracker(Level level, BlockPos blockPos) {
+            this.location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockPos, level.getWorld()).add(0.5, OFFSET_Y, 0.5);
+            this.entity = createFakeLeashEntity(this.location);
+            this.ropeLeashOffset = LeashFenceKnotEntity.this.getRopeHoldOffset().get(net.minecraft.core.Direction.Axis.Y);
+            this.entityLeashOffset = this.entity.getLeashOffset().get(net.minecraft.core.Direction.Axis.Y);
+            this.scale = this.calculateScale(this.ropeLeashOffset, this.entityLeashOffset, 0);
+            this.setScale(this.scale);
+        }
+
+        private void updateTracker(boolean force) {
+            this.sendUpdate(LeashFenceKnotEntity.this.getBukkitEntity().getTrackedBy(), force);
+        }
+
+        private void sendUpdate(java.util.Collection<org.bukkit.entity.Player> players, boolean force) {
+            if (!players.isEmpty()) {
+                if (force) {
+                    this.closer.removeAll(players);
+                    this.farAway.removeAll(players);
+                }
+
+                java.util.List<org.bukkit.entity.Player> updateCloser = new java.util.ArrayList<>();
+                java.util.List<org.bukkit.entity.Player> updateFarAway = new java.util.ArrayList<>();
+                for (org.bukkit.entity.Player player : players) {
+                    if (LeashFenceKnotEntity.this.closerThan(((org.bukkit.craftbukkit.entity.CraftPlayer) player).getHandle(), 16)) {
+                        if (this.closer.add(player)) {
+                            this.farAway.remove(player);
+                            updateCloser.add(player);
+                        }
+                    } else {
+                        if (this.farAway.add(player)) {
+                            this.closer.remove(player);
+                            updateFarAway.add(player);
+                        }
+                    }
+                }
+
+                sendUpdateCloser(updateCloser, force);
+                sendUpdateFarAway(updateFarAway, force);
+            }
+        }
+
+        private void sendUpdateCloser(java.util.Collection<org.bukkit.entity.Player> players, boolean force) {
+            if (!players.isEmpty()) {
+                if (force) {
+                    this.entity.getBukkitEntity().sendChanges(players); // 只有这样才能更新视角
+                }
+                this.entity.getBukkitEntity().sendSpawn(players, this.location);
+            }
+        }
+
+        private void sendUpdateFarAway(java.util.Collection<org.bukkit.entity.Player> players, boolean force) {
+            float entityMoveDown = 0.5F;
+
+            if (!players.isEmpty()) {
+                Vec3 leashPosition = this.entity.getLeashPosition(0);
+
+                this.setScale(this.calculateScale(this.ropeLeashOffset, this.entityLeashOffset, entityMoveDown));
+                Vec3 displacement = this.entity.getLeashPosition(0).subtract(leashPosition);
+                if (force) {
+                    this.entity.getBukkitEntity().sendChanges(players); // 只有这样才能更新视角
+                }
+                this.entity.getBukkitEntity().sendSpawn(players, this.location.clone().subtract(org.bukkit.craftbukkit.util.CraftVector.toBukkit(displacement).setY(entityMoveDown)));
+
+                this.setScale(this.scale);
+            }
+        }
+
+        private void sendRemove(java.util.Collection<org.bukkit.entity.Player> players) {
+            if (!players.isEmpty()) {
+                this.entity.getBukkitEntity().sendRemove(players);
+                this.closer.removeAll(players);
+                this.farAway.removeAll(players);
+            }
+        }
+
+        private void setScale(double scale) {
+            this.entity.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.SCALE).setBaseValue(scale);
+            this.entity.refreshDimensions();
+        }
+
+        private net.minecraft.world.entity.Mob createFakeLeashEntity(org.bukkit.Location location) {
+            org.bukkit.entity.Mob fakeEntity = location.getWorld().createEntity(location, org.bukkit.entity.Rabbit.class);
+            fakeEntity.setAI(false);
+            fakeEntity.setNoPhysics(true);
+            fakeEntity.setInvisible(true);
+            fakeEntity.setInvulnerable(true);
+            return ((org.bukkit.craftbukkit.entity.CraftMob) fakeEntity).getHandle();
+        }
+
+        private double calculateScale(double targetOffset, double entityOffset, double entityMoveDown) {
+            return (targetOffset + entityMoveDown) / entityOffset;
+        }
+    }
+    // KioCG end - leash fence
+
     public LeashFenceKnotEntity(EntityType<? extends LeashFenceKnotEntity> entityType, Level level) {
         super(entityType, level);
     }
@@ -56,15 +_,85 @@
 
     @Override
     public void dropItem(ServerLevel level, @Nullable Entity entity) {
+        // KioCG start - leash fence
+        if (this.isLeashed()) {
+            if (!(entity instanceof Player player && player.hasInfiniteMaterials())) {
+                this.dropLeash();
+            } else {
+                this.removeLeash();
+            }
+        }
+        // KioCG end - leash fence
         this.playSound(SoundEvents.LEAD_UNTIED, 1.0F, 1.0F);
     }
 
+    // KioCG start - leash fence
+    @Nullable
+    @Override
+    public Leashable.LeashData getLeashData() {
+        return this.leashData;
+    }
+
+    @Override
+    public void setLeashData(@Nullable Leashable.LeashData leashData) {
+        this.leashData = leashData;
+        if (leashData != null) {
+            if (leashData.leashHolder != null) {
+                if (this.fakeLeashTracker == null) {
+                    this.fakeLeashTracker = new FakeLeashTracker(this.level(), this.blockPosition());
+                }
+
+                this.fakeLeashTracker.entity.lookAt(net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.FEET, leashData.leashHolder.position());
+                this.fakeLeashTracker.entity.setLeashedTo(leashData.leashHolder, false);
+                this.fakeLeashTracker.updateTracker(true);
+            }
+        } else {
+            if (this.fakeLeashTracker != null) {
+                this.fakeLeashTracker.sendRemove(this.getBukkitEntity().getTrackedBy());
+                this.fakeLeashTracker = null;
+            }
+        }
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        Leashable.tickLeash(this.serverLevel(), this);
+        if (this.fakeLeashTracker != null && this.tickCount % 10 == 0) {
+            this.fakeLeashTracker.updateTracker(false);
+        }
+    }
+
+    @Override
+    public void onLeashRemoved() {
+        this.notifyLeasheeRemoved(null);
+    }
+
+    @Override
+    public void startSeenByPlayer(net.minecraft.server.level.ServerPlayer player) {
+        super.startSeenByPlayer(player);
+        if (this.fakeLeashTracker != null) {
+            this.fakeLeashTracker.sendUpdate(java.util.List.of(player.getBukkitEntity()), false);
+        }
+    }
+
+    @Override
+    public void stopSeenByPlayer(net.minecraft.server.level.ServerPlayer player) {
+        super.stopSeenByPlayer(player);
+        if (this.fakeLeashTracker != null) {
+            this.fakeLeashTracker.sendRemove(java.util.List.of(player.getBukkitEntity()));
+        }
+    }
+    // KioCG end - leash fence
+
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
+        this.writeLeashData(output, this.leashData); // KioCG - leash fence
     }
 
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
+        this.readLeashData(input); // KioCG - leash fence
     }
 
     @Override
@@ -110,7 +_,7 @@
 
     @Override
     public void notifyLeasheeRemoved(Leashable leashHolder) {
-        if (Leashable.leashableLeashedTo(this).isEmpty()) {
+        if (!this.isLeashed() && Leashable.leashableLeashedTo(this).isEmpty()) { // KioCG - leash fence
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
         }
     }
@@ -149,8 +_,14 @@
 
     @Override
     public Vec3 getRopeHoldPosition(float partialTicks) {
-        return this.getPosition(partialTicks).add(0.0, 0.2, 0.0);
-    }
+        return this.getPosition(partialTicks).add(getRopeHoldOffset()); // KioCG - leash fence
+    }
+
+    // KioCG start - leash fence
+    public Vec3 getRopeHoldOffset() {
+        return new Vec3(0.0, 0.2, 0.0);
+    }
+    // KioCG end - leash fence
 
     @Override
     public ItemStack getPickResult() {
