--- a/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -76,7 +_,7 @@
         "sneaking_speed", new RangedAttribute("attribute.name.sneaking_speed", 0.3, 0.0, 1.0).setSyncable(true)
     );
     public static final Holder<Attribute> SPAWN_REINFORCEMENTS_CHANCE = register(
-        "spawn_reinforcements", new RangedAttribute("attribute.name.spawn_reinforcements", 0.0, 0.0, 1.0)
+        "spawn_reinforcements", new RangedAttribute("attribute.name.spawn_reinforcements", 0.0, -1024.0D, 1024.0D) // KioCG - the long dark: 属性用于玩家温度
     );
     public static final Holder<Attribute> STEP_HEIGHT = register(
         "step_height", new RangedAttribute("attribute.name.step_height", 0.6, 0.0, 10.0).setSyncable(true)
