From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KioProject <1105919949@qq.com>
Date: Sun, 6 Apr 2025 22:11:28 +0800
Subject: [PATCH] Custom block entity


diff --git a/src/main/java/org/bukkit/block/Cauldron.java b/src/main/java/org/bukkit/block/Cauldron.java
new file mode 100644
index 0000000000000000000000000000000000000000..8d131b20cc283d9e64bf24d48715951d5925d8c9
--- /dev/null
+++ b/src/main/java/org/bukkit/block/Cauldron.java
@@ -0,0 +1,11 @@
+package org.bukkit.block;
+
+/**
+ * Represents a captured state of a cauldron.
+ */
+// KioCG start - custom block entity: all class
+public interface Cauldron extends TileState {
+
+    boolean isBoiling();
+}
+// KioCG end - custom block entity: all class
diff --git a/src/main/java/org/bukkit/block/FlowerPot.java b/src/main/java/org/bukkit/block/FlowerPot.java
new file mode 100644
index 0000000000000000000000000000000000000000..efbb7ddf0620d419e83cbd7866e34a0a12573ebf
--- /dev/null
+++ b/src/main/java/org/bukkit/block/FlowerPot.java
@@ -0,0 +1,10 @@
+package org.bukkit.block;
+
+/**
+ * Represents a captured state of a flower pot.
+ */
+// KioCG start - custom block entity: all class
+public interface FlowerPot extends TileState {
+
+}
+// KioCG end - custom block entity: all class
