--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -99,7 +_,7 @@
 
     private int getAndValidateGrowth(String crop) {
         int modifier = this.getInt("growth." + crop.toLowerCase(java.util.Locale.ENGLISH) + "-modifier", 100);
-        if (modifier == 0) {
+        if (false && modifier == 0) { // KioCG - crops grow every day
             this.log("Cannot set " + crop + " growth to zero, defaulting to 100");
             modifier = 100;
         }
