--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -171,6 +_,16 @@
         }
     }
 
+    // KioCG start - optimize entity activation: 从Arrow移动
+    @Override
+    public void inactiveTick() {
+        if (this.isInGround()) {
+            this.life++;
+        }
+        super.inactiveTick();
+    }
+    // KioCG end - optimize entity activation: 从Arrow移动
+
     @Override
     public void tick() {
         boolean flag = !this.isNoPhysics();
@@ -461,6 +_,14 @@
         if (this.isOnFire() && !flag) {
             // CraftBukkit start
             org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5.0F);
+            // KioCG start - adjust enchantment
+            if (this.getWeaponItem() != null) {
+                final int flameLevel = this.getWeaponItem().getEnchantmentLevel(this.registryAccess(), net.minecraft.world.item.enchantment.Enchantments.FLAME);
+                if (flameLevel > 1) {
+                    combustEvent.setDuration(combustEvent.getDuration() * flameLevel);
+                }
+            }
+            // KioCG end - adjust enchantment
             if (combustEvent.callEvent()) {
                 entity.igniteForSeconds(combustEvent.getDuration(), false);
             }
