--- a/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -24,8 +_,8 @@
     public SmallFireball(Level level, LivingEntity owner, Vec3 movement) {
         super(EntityType.SMALL_FIREBALL, owner, movement, level);
         // CraftBukkit start
-        if (this.getOwner() != null && this.getOwner() instanceof Mob) {
-            this.isIncendiary = (level instanceof ServerLevel serverLevel) && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+        if (true || this.getOwner() != null && this.getOwner() instanceof Mob) { // KioCG - better mob griefing setting
+            this.isIncendiary = (level instanceof ServerLevel serverLevel) && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1); // KioCG - better mob griefing setting
         }
         // CraftBukkit end
     }
