--- a/net/minecraft/world/item/BottleItem.java
+++ b/net/minecraft/world/item/BottleItem.java
@@ -36,6 +_,13 @@
         ItemStack itemInHand = player.getItemInHand(hand);
         if (!entitiesOfClass.isEmpty()) {
             AreaEffectCloud areaEffectCloud = entitiesOfClass.get(0);
+            // KioCG start - PlayerBottleEvent
+            com.kiocg.event.player.PlayerBottleFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBottleFillEvent(level, player, areaEffectCloud.blockPosition(), itemInHand, new ItemStack(Items.DRAGON_BREATH), hand);
+            if (event.isCancelled()) {
+                player.containerMenu.sendAllDataToRemote();
+                return InteractionResult.PASS;
+            }
+            // KioCG end - PlayerBottleEvent
             areaEffectCloud.setRadius(areaEffectCloud.getRadius() - 0.5F);
             level.playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL_DRAGONBREATH, SoundSource.NEUTRAL, 1.0F, 1.0F);
             level.gameEvent(player, GameEvent.FLUID_PICKUP, player.position());
@@ -43,7 +_,7 @@
                 CriteriaTriggers.PLAYER_INTERACTED_WITH_ENTITY.trigger(serverPlayer, itemInHand, areaEffectCloud);
             }
 
-            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemInHand, player, new ItemStack(Items.DRAGON_BREATH)));
+            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemInHand, player, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem()))); // KioCG - PlayerBottleEvent
         } else {
             BlockHitResult playerPovHitResult = getPlayerPOVHitResult(level, player, ClipContext.Fluid.SOURCE_ONLY);
             if (playerPovHitResult.getType() == HitResult.Type.MISS) {
@@ -56,10 +_,17 @@
                     }
 
                     if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
+                        // KioCG start - PlayerBottleEvent
+                        com.kiocg.event.player.PlayerBottleFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBottleFillEvent(level, player, blockPos, itemInHand, PotionContents.createItemStack(Items.POTION, Potions.WATER), hand);
+                        if (event.isCancelled()) {
+                            player.containerMenu.sendAllDataToRemote();
+                            return InteractionResult.PASS;
+                        }
+                        // KioCG end - PlayerBottleEvent
                         level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.NEUTRAL, 1.0F, 1.0F);
                         level.gameEvent(player, GameEvent.FLUID_PICKUP, blockPos);
                         return InteractionResult.SUCCESS
-                            .heldItemTransformedTo(this.turnBottleIntoItem(itemInHand, player, PotionContents.createItemStack(Items.POTION, Potions.WATER)));
+                            .heldItemTransformedTo(this.turnBottleIntoItem(itemInHand, player, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem()))); // KioCG - PlayerBottleEvent
                     }
                 }
 
