--- a/net/minecraft/world/item/EggItem.java
+++ b/net/minecraft/world/item/EggItem.java
@@ -22,6 +_,12 @@
 
     @Override
     public InteractionResult use(Level level, Player player, InteractionHand hand) {
+        // KioCG start - more food
+        if (true) {
+            player.containerMenu.forceHeldSlot(hand);
+            return super.use(level, player, hand);
+        }
+        // KioCG end - more food
         ItemStack itemInHand = player.getItemInHand(hand);
         // Paper start
         final Projectile.Delayed<ThrownEgg> thrownEgg = Projectile.spawnProjectileFromRotationDelayed(ThrownEgg::new, (ServerLevel) level, itemInHand, player, 0.0F, EggItem.PROJECTILE_SHOOT_POWER, 1.0F);
