--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -104,17 +_,54 @@
         Component.translatable("item.op_warning.line3").withStyle(ChatFormatting.RED)
     );
     private static final Component UNBREAKABLE_TOOLTIP = Component.translatable("item.unbreakable").withStyle(ChatFormatting.BLUE);
+
+    public static final ThreadLocal<Boolean> REMOTE_HASH = ThreadLocal.withInitial(() -> false); // KioCG - customize default components
     public static final MapCodec<ItemStack> MAP_CODEC = MapCodec.recursive(
         "ItemStack",
         codec -> RecordCodecBuilder.mapCodec(
             instance -> instance.group(
                     Item.CODEC.fieldOf("id").forGetter(ItemStack::getItemHolder),
-                    ExtraCodecs.intRange(1, 99).fieldOf("count").orElse(1).forGetter(ItemStack::getCount),
-                    DataComponentPatch.CODEC.optionalFieldOf("components", DataComponentPatch.EMPTY).forGetter(stack -> stack.components.asPatch())
-                )
-                .apply(instance, ItemStack::new)
-        )
-    );
+                    ExtraCodecs.POSITIVE_INT.fieldOf("count").orElse(1).forGetter(ItemStack::getCount), // KioCG - allow overstacked items
+                    // KioCG start - customize default components
+                    DataComponentPatch.CODEC.optionalFieldOf("components", DataComponentPatch.EMPTY).forGetter(stack -> {
+                        if (REMOTE_HASH.get()) {
+                            return stack.getRemoteComponentsPatch();
+                        } else {
+                            return stack.components.asPatch();
+                        }
+                    })
+                    // KioCG end - customize default components
+                )
+                .apply(instance, ItemStack::new)
+        )
+    );
+    // KioCG start - customize default components
+    public static final MapCodec<ItemStack> REMOTE_MAP_CODEC = MapCodec.recursive(
+        "ItemStack",
+        codec -> RecordCodecBuilder.mapCodec(
+            instance -> instance.group(
+                    Item.CODEC.fieldOf("id").forGetter(ItemStack::getItemHolder),
+                    ExtraCodecs.POSITIVE_INT.fieldOf("count").orElse(1).forGetter(ItemStack::getCount), // KioCG - allow overstacked items
+                    DataComponentPatch.CODEC.optionalFieldOf("components", DataComponentPatch.EMPTY).forGetter(stack -> {
+                        // 流编解码以获得嵌套物品(收纳袋等)的内部物品的components.
+                        // 仍然有一些bug, 比如底层Item新增的组件将无法正常取得,
+                        // 例如, 食物的MAX_DAMAGE.
+                        // 这只会影响到消息链中的ShowItem, 因此修复它不是高优先级的.
+                        //TODO 需要在正确的地方设置ItemStack#REMOTE_HASH以使内部物品能够编码正确的RemoteComponentsPatch.
+                        io.netty.buffer.ByteBuf byteBuf = io.netty.buffer.ByteBufAllocator.DEFAULT.buffer();
+                        try {
+                            RegistryFriendlyByteBuf registryFriendlyByteBuf = new RegistryFriendlyByteBuf(byteBuf, net.minecraft.server.MinecraftServer.getServer().registryAccess());
+                            DataComponentPatch.STREAM_CODEC.encode(registryFriendlyByteBuf, stack.getRemoteComponentsPatch());
+                            return DataComponentPatch.STREAM_CODEC.decode(registryFriendlyByteBuf);
+                        } finally {
+                            byteBuf.release();
+                        }
+                    })
+                )
+                .apply(instance, ItemStack::new)
+        )
+    );
+    // KioCG end - customize default components
     public static final Codec<ItemStack> CODEC = Codec.lazyInitialized(MAP_CODEC::codec);
     public static final Codec<ItemStack> SINGLE_ITEM_CODEC = Codec.lazyInitialized(
         () -> RecordCodecBuilder.create(
@@ -192,6 +_,20 @@
                 } else {
                     Holder<Item> holder = Item.STREAM_CODEC.decode(buffer);
                     DataComponentPatch dataComponentPatch = codec.decode(buffer);
+                    // KioCG start - customize default components: 修复创造模式物品
+                    if (codec == DataComponentPatch.DELIMITED_STREAM_CODEC) {
+                        @Nullable Optional<? extends CustomData> optional = dataComponentPatch.get(DataComponents.CUSTOM_DATA);
+                        if (optional != null && optional.isPresent()) {
+                            Optional<DataComponentPatch> originalComponents = optional.flatMap(customData -> {
+                                RegistryOps<Tag> registryOps = net.minecraft.server.MinecraftServer.getServer().registryAccess().createSerializationContext(NbtOps.INSTANCE);
+                                return customData.getUnsafe().read("KioCG.OriginalComponents", DataComponentPatch.CODEC, registryOps);
+                            });
+                            if (originalComponents.isPresent()) {
+                                dataComponentPatch = originalComponents.get();
+                            }
+                        }
+                    }
+                    // KioCG end - customize default components: 修复创造模式物品
                     return new ItemStack(holder, varInt, dataComponentPatch);
                 }
             }
@@ -207,7 +_,7 @@
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
                     try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
-                    codec.encode(buffer, value.components.asPatch());
+                    codec.encode(buffer, value.getRemoteComponentsPatch()); // KioCG - customize default components
                     } finally {
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
                     }
@@ -254,6 +_,54 @@
         return !this.isEmpty() ? this.components.asPatch() : DataComponentPatch.EMPTY;
     }
 
+    // KioCG start - customize default components
+    public DataComponentPatch getRemoteComponentsPatch() {
+        if (!this.isEmpty()) {
+            DataComponentPatch.Builder builder = DataComponentPatch.builder();
+            builder.copy(this.getItem().componentsPatchRemote());
+
+            boolean changed = false;
+            DataComponentPatch patch = this.components.asPatch();
+            for (java.util.Map.Entry<DataComponentType<?>, Optional<?>> entry : patch.entrySet()) {
+                if (!DataComponents.UNABLE_SEND_TO_CLIENT.contains(entry.getKey())) {
+                    entry.getValue().ifPresentOrElse(value -> builder.set((DataComponentType) entry.getKey(), value), () -> builder.remove(entry.getKey()));
+                } else {
+                    changed = true;
+                }
+            }
+
+            changed |= saveAdditional(builder);
+            if (changed) {
+                net.minecraft.nbt.CompoundTag compoundTag = new net.minecraft.nbt.CompoundTag();
+                RegistryOps<Tag> registryOps = net.minecraft.server.MinecraftServer.getServer().registryAccess().createSerializationContext(NbtOps.INSTANCE);
+                compoundTag.store("KioCG.OriginalComponents", DataComponentPatch.CODEC, registryOps, patch);
+                builder.set(DataComponents.CUSTOM_DATA, CustomData.of(compoundTag));
+            }
+
+            return builder.build();
+        }
+        return DataComponentPatch.EMPTY;
+    }
+
+    private boolean saveAdditional(DataComponentPatch.Builder builder) {
+        boolean changed = false;
+
+        net.minecraft.world.item.component.ItemLore extendedBeforeLore = this.getOrDefault(DataComponents.EXTENDED_BEFORE_LORE, net.minecraft.world.item.component.ItemLore.EMPTY);
+        net.minecraft.world.item.component.ItemLore extendedAfterLore = this.getOrDefault(DataComponents.EXTENDED_AFTER_LORE, net.minecraft.world.item.component.ItemLore.EMPTY);
+        if (!extendedBeforeLore.lines().isEmpty() || !extendedAfterLore.lines().isEmpty()) {
+            net.minecraft.world.item.component.ItemLore lore = this.getOrDefault(DataComponents.LORE, net.minecraft.world.item.component.ItemLore.EMPTY);
+            List<Component> lores = new java.util.ArrayList<>();
+            lores.addAll(extendedBeforeLore.lines());
+            lores.addAll(lore.lines());
+            lores.addAll(extendedAfterLore.lines());
+            builder.set(DataComponents.LORE, new net.minecraft.world.item.component.ItemLore(lores));
+            changed = true;
+        }
+
+        return changed;
+    }
+    // KioCG end - customize default components
+
     public DataComponentMap immutableComponents() {
         return !this.isEmpty() ? this.components.toImmutableMap() : DataComponentMap.EMPTY;
     }
@@ -295,7 +_,7 @@
     }
 
     public static DataResult<Unit> validateComponents(DataComponentMap components) {
-        if (components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) {
+        if (false && components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) { // KioCG - food freshness
             return DataResult.error(() -> "Item cannot be both damageable and stackable");
         } else {
             ItemContainerContents itemContainerContents = components.getOrDefault(DataComponents.CONTAINER, ItemContainerContents.EMPTY);
@@ -380,6 +_,16 @@
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), clickedPos, false))) {
             return InteractionResult.PASS;
         } else {
+            // KioCG start - customize default components
+            if (this.has(DataComponents.CANNOT_PLACE)) {
+                if (player != null) {
+                    player.containerMenu.sendAllDataToRemote();
+                    player.resyncUsingItem((ServerPlayer) player);
+                }
+                return InteractionResult.FAIL;
+            }
+            // KioCG end - customize default components
+
             Item item = this.getItem();
             // CraftBukkit start - handle all block place event logic here
             DataComponentPatch previousPatch = this.components.asPatch();
@@ -606,7 +_,7 @@
     }
 
     public boolean isStackable() {
-        return this.getMaxStackSize() > 1 && (!this.isDamageableItem() || !this.isDamaged());
+        return this.getMaxStackSize() > 1 && (true || (!this.isDamageableItem() || !this.isDamaged())); // KioCG - food freshness
     }
 
     public boolean isDamageableItem() {
@@ -892,8 +_,167 @@
     }
 
     public static boolean isSameItemSameComponents(ItemStack stack, ItemStack other) {
-        return stack.is(other.getItem()) && (stack.isEmpty() && other.isEmpty() || Objects.equals(stack.components, other.components));
-    }
+        return stack.is(other.getItem()) && (stack.isEmpty() && other.isEmpty() || isSamePatchedComponents(stack.components, other.components)); // KioCG - customize default components
+    }
+
+    // KioCG start - customize default components
+    private static boolean isSamePatchedComponents(PatchedDataComponentMap stack, PatchedDataComponentMap other) {
+        return isSamePatchedComponents(stack, other, new java.util.HashSet<>());
+    }
+    private static boolean isSamePatchedComponents(PatchedDataComponentMap stack, PatchedDataComponentMap other, java.util.Set<DataComponentType<?>> excludeTypes) {
+        if (stack == other) return true;
+
+        excludeTypes.add(DataComponents.STACKABLE_CUSTOM_DATA);
+        return stack.asPatch().forget(excludeTypes).equals(other.asPatch().forget(excludeTypes));
+    }
+    // KioCG end - customize default components
+
+    // KioCG start - food freshness
+    public static boolean isSameItemSameComponentsForFoodFreshness(ItemStack stack, ItemStack other) {
+        if (!isSameItem(stack, other)) return false;
+        if (stack.isEmpty() && other.isEmpty()) return true;
+
+        if (stack.hasFreshness()) {
+            java.util.Set<DataComponentType<?>> set = new java.util.HashSet<>();
+            set.add(DataComponents.DAMAGE);
+            return isSamePatchedComponents(stack.components, other.components, set);
+        } else {
+            return isSamePatchedComponents(stack.components, other.components);
+        }
+    }
+
+    private boolean hasFreshness() {
+        net.minecraft.world.food.FoodProperties foodProperties = this.get(DataComponents.FOOD);
+        return foodProperties != null && foodProperties.freshness() > 0;
+    }
+
+    public boolean deepTick(java.util.function.Function<ItemStack, Boolean> function, int maxDepth) {
+        if (!this.isEmpty() && maxDepth >= 0) {
+            boolean changed = false;
+
+            net.minecraft.world.item.component.BundleContents bundleContents = this.get(DataComponents.BUNDLE_CONTENTS);
+            if (bundleContents != null) {
+                boolean changed1 = false;
+                List<ItemStack> result = new java.util.ArrayList<>(bundleContents.size());
+                for (ItemStack itemStack : bundleContents.items()) {
+                    // 如果不进行克隆, 保存时可能丢失数据
+                    result.add(itemStack = itemStack.copy());
+                    if (itemStack.deepTick(function, maxDepth - 1)) {
+                        changed1 = true;
+                    }
+                }
+                if (changed1) {
+                    this.set(DataComponents.BUNDLE_CONTENTS, new net.minecraft.world.item.component.BundleContents(List.copyOf(result)));
+                    changed = true;
+                }
+            }
+
+            ItemContainerContents container = this.get(DataComponents.CONTAINER);
+            if (container != null) {
+                boolean changed1 = false;
+                List<ItemStack> result = new java.util.ArrayList<>(container.items.size());
+                for (ItemStack itemStack : container.items) {
+                    // 如果不进行克隆, 保存时可能丢失数据
+                    result.add(itemStack = itemStack.copy());
+                    if (itemStack.deepTick(function, maxDepth - 1)) {
+                        changed1 = true;
+                    }
+                }
+                if (changed1) {
+                    this.set(DataComponents.CONTAINER, ItemContainerContents.fromItems(List.copyOf(result)));
+                    changed = true;
+                }
+            }
+
+            changed |= function.apply(this);
+            return changed;
+        }
+        return false;
+    }
+
+    public boolean tickFood(long tickTime) {
+        return this.tickFood(tickTime, true);
+    }
+
+    public boolean tickFood(long tickTime, boolean applyDamage) {
+        return deepTick(itemStack -> itemStack.enableFreshness(tickTime, applyDamage), 1);
+    }
+
+    public boolean tickFoodDeep(long tickTime, boolean applyDamage) {
+        return deepTick(itemStack -> itemStack.enableFreshness(tickTime, applyDamage), 64);
+    }
+
+    public boolean tickFoodFreeze(long tickTime) {
+        return this.tickFoodFreeze(tickTime, true);
+    }
+
+    public boolean tickFoodFreeze(long tickTime, boolean applyDamage) {
+        return deepTick(itemStack -> itemStack.disableFreshness(tickTime, applyDamage), 1);
+    }
+
+    public boolean tickFoodFreezeDeep(long tickTime, boolean applyDamage) {
+        return deepTick(itemStack -> itemStack.disableFreshness(tickTime, applyDamage), 64);
+    }
+
+    private boolean enableFreshness(long tickTime, boolean applyDamage) {
+        if (hasFreshness() && !this.isBroken()) {
+            CustomData stackableCustomData = this.getOrDefault(DataComponents.STACKABLE_CUSTOM_DATA, CustomData.EMPTY);
+            Optional<Long> foodTickTime = stackableCustomData.getUnsafe().getLong("FoodTickTime");
+            if (foodTickTime.isEmpty() || foodTickTime.get() != tickTime) {
+                if (applyDamage) {
+                    tickFreshness(tickTime, foodTickTime.orElse(tickTime - 1));
+                }
+
+                stackableCustomData = stackableCustomData.update(tag -> tag.putLong("FoodTickTime", tickTime));
+                this.set(DataComponents.STACKABLE_CUSTOM_DATA, stackableCustomData);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean hopperEnableFreshness(long tickTime) {
+        if (hasFreshness() && !this.isBroken()) {
+            CustomData stackableCustomData = this.getOrDefault(DataComponents.STACKABLE_CUSTOM_DATA, CustomData.EMPTY);
+            if (!stackableCustomData.getUnsafe().contains("FoodTickTime")) {
+                stackableCustomData = stackableCustomData.update(tag -> tag.putLong("FoodTickTime", tickTime));
+                this.set(DataComponents.STACKABLE_CUSTOM_DATA, stackableCustomData);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean disableFreshness(long tickTime, boolean applyDamage) {
+        CustomData stackableCustomData = this.getOrDefault(DataComponents.STACKABLE_CUSTOM_DATA, CustomData.EMPTY);
+        Optional<Long> foodTickTime = stackableCustomData.getUnsafe().getLong("FoodTickTime");
+        if (foodTickTime.isPresent()) {
+            if (applyDamage && hasFreshness() && !this.isBroken()) {
+                tickFreshness(tickTime, foodTickTime.orElse(tickTime - 1));
+            }
+
+            stackableCustomData = stackableCustomData.update(tag -> tag.remove("FoodTickTime"));
+            this.set(DataComponents.STACKABLE_CUSTOM_DATA, stackableCustomData);
+            return true;
+        }
+        return false;
+    }
+
+    private void tickFreshness(long tickTime, long prevTickTime) {
+        if (prevTickTime < tickTime) {
+            int spend = spendFreshness(prevTickTime + 1, tickTime, io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.foodFreshnessTickInterval);
+            if (spend != 0) {
+                this.setDamageValue(this.getDamageValue() + spend);
+            }
+        }
+    }
+
+    private static int spendFreshness(long start, long end, int interval) {
+        int startPoint = (int) Math.ceilDiv(start, interval);
+        int endPoint = (int) Math.floorDiv(end, interval);
+        return endPoint - startPoint + 1;
+    }
+    // KioCG end - food freshness
 
     public static MapCodec<ItemStack> lenientOptionalFieldOf(String fieldName) {
         return CODEC.lenientOptionalFieldOf(fieldName)
@@ -1276,6 +_,12 @@
         EnchantmentHelper.updateEnchantments(this, enchantments -> enchantments.upgrade(enchantment, level));
     }
 
+    // KioCG start - allow unsafe enchant
+    public void setEnchant(Holder<Enchantment> enchantment, int level) {
+        EnchantmentHelper.updateEnchantments(this, enchantments -> enchantments.set(enchantment, level));
+    }
+    // KioCG end - allow unsafe enchant
+
     public boolean isEnchanted() {
         return !this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).isEmpty();
     }
@@ -1284,6 +_,14 @@
         return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
     }
 
+    // KioCG start - adjust enchantment
+    public int getEnchantmentLevel(net.minecraft.core.RegistryAccess registryManager, net.minecraft.resources.ResourceKey<Enchantment> enchantmentKey) {
+        return registryManager.lookupOrThrow(net.minecraft.core.registries.Registries.ENCHANTMENT).get(enchantmentKey)
+                              .map(reference -> getEnchantments().getLevel(reference))
+                              .orElse(0);
+    }
+    // KioCG end - adjust enchantment
+
     public boolean isFramed() {
         return this.entityRepresentation instanceof ItemFrame;
     }
@@ -1415,6 +_,15 @@
 
     public boolean isValidRepairItem(ItemStack item) {
         Repairable repairable = this.get(DataComponents.REPAIRABLE);
+        // KioCG start - more repair item
+        if (repairable == null) {
+            if (this.is(Items.SHEARS) && item.is(Items.IRON_INGOT)) return true;
+            if (this.is(Items.BOW) && item.is(Items.STRING)) return true;
+            if (this.is(Items.FISHING_ROD) && item.is(Items.STRING)) return true;
+            if (this.is(Items.FLINT_AND_STEEL) && item.is(Items.FLINT)) return true;
+            if (this.is(Items.BRUSH) && item.is(Items.FEATHER)) return true;
+        }
+        // KioCG end - more repair item
         return repairable != null && repairable.isValidRepairItem(item);
     }
 
