--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -109,7 +_,7 @@
         codec -> RecordCodecBuilder.mapCodec(
             instance -> instance.group(
                     Item.CODEC.fieldOf("id").forGetter(ItemStack::getItemHolder),
-                    ExtraCodecs.intRange(1, 99).fieldOf("count").orElse(1).forGetter(ItemStack::getCount),
+                    ExtraCodecs.POSITIVE_INT.fieldOf("count").orElse(1).forGetter(ItemStack::getCount), // KioCG - allow overstacked items
                     DataComponentPatch.CODEC.optionalFieldOf("components", DataComponentPatch.EMPTY).forGetter(stack -> stack.components.asPatch())
                 )
                 .apply(instance, ItemStack::new)
@@ -207,7 +_,7 @@
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
                     try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
-                    codec.encode(buffer, value.components.asPatch());
+                    codec.encode(buffer, value.getRemoteComponentsPatch()); // KioCG - customize default components
                     } finally {
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
                     }
@@ -254,6 +_,18 @@
         return !this.isEmpty() ? this.components.asPatch() : DataComponentPatch.EMPTY;
     }
 
+    // KioCG start - customize default components
+    public DataComponentPatch getRemoteComponentsPatch() {
+        if (!this.isEmpty()) {
+            DataComponentPatch.Builder builder = DataComponentPatch.builder();
+            builder.copy(this.getItem().componentsPatchKioCG());
+            builder.copy(this.components.asPatch());
+            return builder.build();
+        }
+        return DataComponentPatch.EMPTY;
+    }
+    // KioCG end - customize default components
+
     public DataComponentMap immutableComponents() {
         return !this.isEmpty() ? this.components.toImmutableMap() : DataComponentMap.EMPTY;
     }
@@ -1277,6 +_,12 @@
         EnchantmentHelper.updateEnchantments(this, enchantments -> enchantments.upgrade(enchantment, level));
     }
 
+    // KioCG start - allow unsafe enchant
+    public void setEnchant(Holder<Enchantment> enchantment, int level) {
+        EnchantmentHelper.updateEnchantments(this, enchantments -> enchantments.set(enchantment, level));
+    }
+    // KioCG end - allow unsafe enchant
+
     public boolean isEnchanted() {
         return !this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).isEmpty();
     }
@@ -1285,6 +_,14 @@
         return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
     }
 
+    // KioCG start - adjust enchantment
+    public int getEnchantmentLevel(net.minecraft.core.RegistryAccess registryManager, net.minecraft.resources.ResourceKey<Enchantment> enchantmentKey) {
+        return registryManager.lookupOrThrow(net.minecraft.core.registries.Registries.ENCHANTMENT).get(enchantmentKey)
+                              .map(reference -> getEnchantments().getLevel(reference))
+                              .orElse(0);
+    }
+    // KioCG end - adjust enchantment
+
     public boolean isFramed() {
         return this.entityRepresentation instanceof ItemFrame;
     }
@@ -1416,6 +_,15 @@
 
     public boolean isValidRepairItem(ItemStack item) {
         Repairable repairable = this.get(DataComponents.REPAIRABLE);
+        // KioCG start - more repair item
+        if (repairable == null) {
+            if (this.is(Items.SHEARS) && item.is(Items.IRON_INGOT)) return true;
+            if (this.is(Items.BOW) && item.is(Items.STRING)) return true;
+            if (this.is(Items.FISHING_ROD) && item.is(Items.STRING)) return true;
+            if (this.is(Items.FLINT_AND_STEEL) && item.is(Items.FLINT)) return true;
+            if (this.is(Items.BRUSH) && item.is(Items.FEATHER)) return true;
+        }
+        // KioCG end - more repair item
         return repairable != null && repairable.isValidRepairItem(item);
     }
 
