--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -100,6 +_,7 @@
     protected static final int APPROXIMATELY_INFINITE_USE_DURATION = 72000;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
     private final DataComponentMap components;
+    private final net.minecraft.core.component.DataComponentPatch componentsPatchKioCG; // KioCG - customize default components
     @Nullable
     private final Item craftingRemainingItem;
     protected final String descriptionId;
@@ -121,6 +_,7 @@
     public Item(Item.Properties properties) {
         this.descriptionId = properties.effectiveDescriptionId();
         this.components = properties.buildAndValidateComponents(Component.translatable(this.descriptionId), properties.effectiveModel());
+        this.componentsPatchKioCG = properties.afterBuildComponentsPatchKioCG(); // KioCG - customize default components
         this.craftingRemainingItem = properties.craftingRemainingItem;
         this.requiredFeatures = properties.requiredFeatures;
         if (SharedConstants.IS_RUNNING_IN_IDE) {
@@ -140,6 +_,13 @@
         return this.components;
     }
 
+    // KioCG start - customize default components
+    // 只有需要发给客户端的DataComponentPatch
+    public net.minecraft.core.component.DataComponentPatch componentsPatchRemote() {
+        return this.componentsPatchKioCG;
+    }
+    // KioCG end - customize default components
+
     public int getDefaultMaxStackSize() {
         return this.components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
     }
@@ -356,6 +_,13 @@
         private static final DependantName<Item, String> BLOCK_DESCRIPTION_ID = key -> Util.makeDescriptionId("block", key.location());
         private static final DependantName<Item, String> ITEM_DESCRIPTION_ID = key -> Util.makeDescriptionId("item", key.location());
         private final DataComponentMap.Builder components = DataComponentMap.builder().addAll(DataComponents.COMMON_ITEM_COMPONENTS);
+        // KioCG start - customize default components
+        private boolean initCustomize = false;
+        private final net.minecraft.core.component.DataComponentPatch.Builder componentsPatchKioCG = net.minecraft.core.component.DataComponentPatch.builder();
+        private final java.util.List<Component> beforeLores = new java.util.ArrayList<>();
+        private final java.util.List<Component> afterLores = new java.util.ArrayList<>();
+        private final net.minecraft.nbt.CompoundTag customDataTag = new net.minecraft.nbt.CompoundTag();
+        // KioCG end - customize default components
         @Nullable
         Item craftingRemainingItem;
         FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
@@ -368,9 +_,35 @@
             return this.food(food, Consumables.DEFAULT_FOOD);
         }
 
-        public Item.Properties food(FoodProperties food, Consumable consumable) {
-            return this.component(DataComponents.FOOD, food).component(DataComponents.CONSUMABLE, consumable);
-        }
+        // KioCG start - more food
+        public Item.Properties food(FoodProperties food, @Nullable Consumable consumable) {
+            if (food.freshness() > 0) {
+                this.componentKioCG(DataComponents.MAX_DAMAGE, food.freshness());
+            }
+            this.foodLore(food.nutrition(), food.moisture());
+
+            if (consumable == null) {
+                return this.component(DataComponents.FOOD, food);
+            } else {
+                return this.component(DataComponents.FOOD, food).component(DataComponents.CONSUMABLE, consumable);
+            }
+        }
+
+        public Item.Properties foodLore(int nutrition, int moisture) {
+            StringBuilder builder = new StringBuilder();
+            if (nutrition != 0) {
+                builder.append("\uD83C\uDF57").append(nutrition);
+            }
+            if (moisture != 0) {
+                builder.append(" ").append("\uD83D\uDCA7").append(moisture);
+            }
+            String string = builder.toString();
+            if (!string.isEmpty()) {
+                this.loreAfter(net.kyori.adventure.text.Component.text(string, net.kyori.adventure.text.format.NamedTextColor.GRAY));
+            }
+            return this;
+        }
+        // KioCG end - more food
 
         public Item.Properties usingConvertsTo(Item usingConvertsTo) {
             return this.component(DataComponents.USE_REMAINDER, new UseRemainder(new ItemStack(usingConvertsTo)));
@@ -531,16 +_,60 @@
             return this.model.get(Objects.requireNonNull(this.id, "Item id not set"));
         }
 
-        public <T> Item.Properties component(DataComponentType<T> component, T value) {
-            this.components.set(component, value);
-            return this;
-        }
+        // KioCG start - customize default components
+        public <T> Item.Properties component(DataComponentType<T> component, @Nullable T value) {
+            if (this.initCustomize) return this.componentKioCG(component, value);
+            // KioCG end - customize default components
+            this.components.set(component, value);
+            return this;
+        }
+
+        // KioCG start - customize default components
+        public Item.Properties initCustomize() {
+            this.initCustomize = true;
+            return this;
+        }
+
+        public <T> Item.Properties componentKioCG(DataComponentType<T> component, @Nullable T value) {
+            this.components.set(component, value);
+            if (!DataComponents.UNABLE_SEND_TO_CLIENT.contains(component)) {
+                if (value != null) {
+                    this.componentsPatchKioCG.set(component, value);
+                } else {
+                    this.componentsPatchKioCG.remove(component);
+                }
+            }
+            return this;
+        }
+
+        public Item.Properties loreBefore(net.kyori.adventure.text.Component component) {
+            this.beforeLores.add(io.papermc.paper.adventure.PaperAdventure.asVanilla(component.decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false)));
+            return this;
+        }
+
+        public Item.Properties loreAfter(net.kyori.adventure.text.Component component) {
+            this.afterLores.add(io.papermc.paper.adventure.PaperAdventure.asVanilla(component.decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false)));
+            return this;
+        }
+        // KioCG end - customize default components
 
         public Item.Properties attributes(ItemAttributeModifiers attributes) {
             return this.component(DataComponents.ATTRIBUTE_MODIFIERS, attributes);
         }
 
         DataComponentMap buildAndValidateComponents(Component itemName, ResourceLocation itemModel) {
+            // KioCG start - customize default components
+            if (!this.beforeLores.isEmpty()) {
+                this.componentKioCG(DataComponents.EXTENDED_BEFORE_LORE, new net.minecraft.world.item.component.ItemLore(this.beforeLores));
+            }
+            if (!this.afterLores.isEmpty()) {
+                this.componentKioCG(DataComponents.EXTENDED_AFTER_LORE, new net.minecraft.world.item.component.ItemLore(this.afterLores));
+            }
+            if (!this.customDataTag.isEmpty()) {
+                this.componentKioCG(DataComponents.CUSTOM_DATA, net.minecraft.world.item.component.CustomData.of(this.customDataTag));
+            }
+            // KioCG end - customize default components
+
             DataComponentMap dataComponentMap = this.components.set(DataComponents.ITEM_NAME, itemName).set(DataComponents.ITEM_MODEL, itemModel).build();
             if (dataComponentMap.has(DataComponents.DAMAGE) && dataComponentMap.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) {
                 throw new IllegalStateException("Item cannot have both durability and be stackable");
@@ -548,6 +_,12 @@
                 return dataComponentMap;
             }
         }
+
+        // KioCG start - customize default components
+        net.minecraft.core.component.DataComponentPatch afterBuildComponentsPatchKioCG() {
+            return this.componentsPatchKioCG.build();
+        }
+        // KioCG end - customize default components
     }
 
     public interface TooltipContext {
