--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -165,8 +_,10 @@
             if (breedOffspring == null) {
                 return Optional.empty();
             } else {
-                breedOffspring.setBaby(true);
-                if (!breedOffspring.isBaby()) {
+                // KioCG start - entity baby api
+                breedOffspring.setBabySimulate();
+                if (!breedOffspring.isBabyOrBabySimulate()) {
+                // KioCG end - entity baby api
                     return Optional.empty();
                 } else {
                     breedOffspring.snapTo(pos.x(), pos.y(), pos.z(), 0.0F, 0.0F);
