--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -80,6 +_,7 @@
     int minSectionY,
     long lastUpdateTime,
     long inhabitedTime,
+    com.kiocg.world.ChunkDays chunkDays, // KioCG - crops grow every day
     ChunkStatus chunkStatus,
     @Nullable BlendingData.Packed blendingData,
     @Nullable BelowZeroRetrogen belowZeroRetrogen,
@@ -250,12 +_,18 @@
                 }
             }
 
+            // KioCG start - crops grow every day
+            int days = tag.getIntOr("ChunkDays", serverLevel.getDays());
+            int nights = tag.getIntOr("ChunkNights", serverLevel.getNights()); // KioCG - mobs disappear every day
+            com.kiocg.world.ChunkDays chunkDays = new com.kiocg.world.ChunkDays(days, nights);
+            // KioCG end - crops grow every day
             return new SerializableChunkData(
                 registry,
                 chunkPos,
                 levelHeightAccessor.getMinSectionY(),
                 longOr,
                 longOr1,
+                chunkDays, // KioCG - crops grow every day
                 chunkStatus,
                 packed,
                 belowZeroRetrogen,
@@ -400,6 +_,7 @@
             }
         }
 
+        chunkAccess.chunkDays = this.chunkDays; // KioCG - crops grow every day
         // CraftBukkit start - load chunk persistent data from nbt - SPIGOT-6814: Already load PDC here to account for 1.17 to 1.18 chunk upgrading.
         if (this.persistentDataContainer instanceof CompoundTag compoundTag) {
             chunkAccess.persistentDataContainer.putAll(compoundTag);
@@ -562,6 +_,7 @@
                 chunk.getMinSectionY(),
                 level.getGameTime(),
                 chunk.getInhabitedTime(),
+                chunk.chunkDays, // KioCG - crops grow every day
                 chunk.getPersistedStatus(),
                 Optionull.map(chunk.getBlendingData(), BlendingData::pack),
                 chunk.getBelowZeroRetrogen(),
@@ -587,6 +_,8 @@
         compoundTag.putInt("zPos", this.chunkPos.z);
         compoundTag.putLong("LastUpdate", this.lastUpdateTime); // Paper - Diff on change
         compoundTag.putLong("InhabitedTime", this.inhabitedTime);
+        compoundTag.putInt("ChunkDays", this.chunkDays.getDays()); // KioCG - crops grow every day
+        compoundTag.putInt("ChunkNights", this.chunkDays.getNights()); // KioCG - mobs disappear every day
         compoundTag.putString("Status", BuiltInRegistries.CHUNK_STATUS.getKey(this.chunkStatus).toString());
         compoundTag.storeNullable("blending_data", BlendingData.Packed.CODEC, this.blendingData);
         compoundTag.storeNullable("below_zero_retrogen", BelowZeroRetrogen.CODEC, this.belowZeroRetrogen);
