--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -78,6 +_,7 @@
     int minSectionY,
     long lastUpdateTime,
     long inhabitedTime,
+    int chunkDays, // KioCG - crops grow every day
     ChunkStatus chunkStatus,
     @Nullable BlendingData.Packed blendingData,
     @Nullable BelowZeroRetrogen belowZeroRetrogen,
@@ -254,6 +_,7 @@
                 levelHeightAccessor.getMinSectionY(),
                 longOr,
                 longOr1,
+                tag.getIntOr("ChunkDays", serverLevel.getDays()), // KioCG - crops grow every day
                 chunkStatus,
                 packed,
                 belowZeroRetrogen,
@@ -398,6 +_,7 @@
             }
         }
 
+        chunkAccess.chunkDays = this.chunkDays; // KioCG - crops grow every day
         // CraftBukkit start - load chunk persistent data from nbt - SPIGOT-6814: Already load PDC here to account for 1.17 to 1.18 chunk upgrading.
         if (this.persistentDataContainer instanceof CompoundTag compoundTag) {
             chunkAccess.persistentDataContainer.putAll(compoundTag);
@@ -560,6 +_,7 @@
                 chunk.getMinSectionY(),
                 level.getGameTime(),
                 chunk.getInhabitedTime(),
+                chunk.chunkDays, // KioCG - crops grow every day
                 chunk.getPersistedStatus(),
                 Optionull.map(chunk.getBlendingData(), BlendingData::pack),
                 chunk.getBelowZeroRetrogen(),
@@ -585,6 +_,7 @@
         compoundTag.putInt("zPos", this.chunkPos.z);
         compoundTag.putLong("LastUpdate", this.lastUpdateTime); // Paper - Diff on change
         compoundTag.putLong("InhabitedTime", this.inhabitedTime);
+        compoundTag.putInt("ChunkDays", this.chunkDays); // KioCG - crops grow every day
         compoundTag.putString("Status", BuiltInRegistries.CHUNK_STATUS.getKey(this.chunkStatus).toString());
         compoundTag.storeNullable("blending_data", BlendingData.Packed.CODEC, this.blendingData);
         compoundTag.storeNullable("below_zero_retrogen", BelowZeroRetrogen.CODEC, this.belowZeroRetrogen);
