--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -34,6 +_,11 @@
     public void save(Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
         try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
+            // KioCG start - food freshness
+            java.util.function.Function<net.minecraft.world.item.ItemStack, Boolean> tickFood = item -> item.tickFoodDeep(player.level().getGameTime(), true);
+            net.minecraft.world.entity.player.Inventory.tick(player.getInventory().getNonEquipmentItems(), tickFood);
+            net.minecraft.world.entity.player.Inventory.tick(player.getEnderChestInventory().getItems(), tickFood);
+            // KioCG end - food freshness
             TagValueOutput tagValueOutput = TagValueOutput.createWithContext(scopedCollector, player.registryAccess());
             player.saveWithoutId(tagValueOutput);
             Path path = this.playerDir.toPath();
@@ -104,6 +_,11 @@
 
             ValueInput valueInput = TagValueInput.create(problemReporter, player.registryAccess(), tag);
             player.load(valueInput); // From below
+            // KioCG start - food freshness
+            java.util.function.Consumer<net.minecraft.world.item.ItemStack> tickFood = item -> item.tickFoodDeep(player.level().getGameTime(), false);
+            player.getInventory().getNonEquipmentItems().forEach(tickFood);
+            player.getEnderChestInventory().getItems().forEach(tickFood);
+            // KioCG end - food freshness
             return valueInput;
         });
     }
