--- a/src/main/java/org/bukkit/block/data/BlockData.java
+++ b/src/main/java/org/bukkit/block/data/BlockData.java
@@ -279,6 +_,11 @@
     @NotNull
     BlockState createBlockState();
 
+    // KioCG start - more api
+    @NotNull
+    BlockState createBlockState(Location location);
+    // KioCG end - more api
+
     // Paper start - destroy speed API
     /**
      * Gets the speed at which this block will be destroyed by a given {@link ItemStack}
