--- a/net/minecraft/world/entity/ai/goal/BreakDoorGoal.java
+++ b/net/minecraft/world/entity/ai/goal/BreakDoorGoal.java
@@ -30,7 +_,7 @@
     @Override
     public boolean canUse() {
         return super.canUse()
-            && getServerLevel(this.mob).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            && true // KioCG - better mob griefing setting
             && this.isValidDifficulty(this.mob.level().getDifficulty())
             && !this.isOpen();
     }
@@ -39,6 +_,12 @@
     public void start() {
         super.start();
         this.breakTime = 0;
+        // KioCG start - keep block destruction
+        final int progress = this.mob.level().destroyingBlocks.get(this.doorPos).getProgress();
+        if (progress > 0) {
+            this.breakTime = (int) (progress / 10.0F * this.getDoorBreakTime());
+        }
+        // KioCG end - keep block destruction
     }
 
     @Override
@@ -52,7 +_,7 @@
     @Override
     public void stop() {
         super.stop();
-        this.mob.level().destroyBlockProgress(this.mob.getId(), this.doorPos, -1);
+        this.mob.serverLevel().abortDestroyBlock(this.mob.getId(), this.doorPos, -1); // KioCG - keep block destruction
     }
 
     @Override
@@ -80,7 +_,7 @@
             }
             // CraftBukkit end
             final net.minecraft.world.level.block.state.BlockState oldState = this.mob.level().getBlockState(this.doorPos); // Paper - fix MC-263999
-            this.mob.level().removeBlock(this.doorPos, false);
+            this.setOpen(true); // this.mob.level().removeBlock(this.doorPos, false); // KioCG - better mob griefing setting
             this.mob.level().levelEvent(1021, this.doorPos, 0);
             this.mob.level().levelEvent(2001, this.doorPos, Block.getId(oldState)); // Paper - fix MC-263999
         }
