--- a/net/minecraft/world/entity/ai/goal/BreedGoal.java
+++ b/net/minecraft/world/entity/ai/goal/BreedGoal.java
@@ -39,6 +_,13 @@
         }
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    protected int canUseDelay() {
+        return 10;
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canContinueToUse() {
         return this.partner.isAlive() && this.partner.isInLove() && this.loveTime < 60 && !this.partner.isPanicking();
@@ -53,7 +_,7 @@
     @Override
     public void tick() {
         this.animal.getLookControl().setLookAt(this.partner, 10.0F, this.animal.getMaxHeadXRot());
-        this.animal.getNavigation().moveTo(this.partner, this.speedModifier);
+        this.animal.getNavigation().moveToLightweight(this.partner, this.speedModifier); // KioCG - optimize entity behavior
         this.loveTime++;
         if (this.loveTime >= this.adjustedTickDelay(60) && this.animal.distanceToSqr(this.partner) < 9.0) {
             this.breed();
