--- a/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java
@@ -19,9 +_,16 @@
         this.cat = cat;
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    public boolean registerWhen() {
+        return this.cat.isTame() && !this.cat.isOrderedToSit();
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canUse() {
-        return this.cat.isTame() && !this.cat.isOrderedToSit() && super.canUse();
+        return super.canUse(); // KioCG - optimize entity behavior
     }
 
     @Override
