--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -19,6 +_,7 @@
     private final float stopDistance;
     private final float startDistance;
     private float oldWaterCost;
+    private int ticketTicks; // KioCG - dont kill my pet
 
     public FollowOwnerGoal(TamableAnimal tamable, double speedModifier, float startDistance, float stopDistance) {
         this.tamable = tamable;
@@ -40,6 +_,7 @@
         } else if (this.tamable.unableToMoveToOwner()) {
             return false;
         } else if (this.tamable.distanceToSqr(owner) < this.startDistance * this.startDistance) {
+            keepChunkLoaded(); // KioCG - dont kill my pet
             return false;
         } else {
             this.owner = owner;
@@ -47,6 +_,16 @@
         }
     }
 
+    // KioCG start - dont kill my pet
+    private void keepChunkLoaded() {
+        if (!io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petFollowKeepChunkLoaded) return;
+        if (this.tamable.tickCount >= ticketTicks + 20) {
+            ticketTicks = this.tamable.tickCount;
+            this.tamable.serverLevel().getChunkSource().addRegionTicket(net.minecraft.server.level.TicketType.PET_FOLLOW, this.tamable.chunkPosition(), 1, this.tamable.getId());
+        }
+    }
+    // KioCG end - dont kill my pet
+
     @Override
     public boolean canContinueToUse() {
         return !this.navigation.isDone()
@@ -70,6 +_,7 @@
 
     @Override
     public void tick() {
+        keepChunkLoaded(); // KioCG - dont kill my pet
         boolean shouldTryTeleportToOwner = this.tamable.shouldTryTeleportToOwner();
         if (!shouldTryTeleportToOwner) {
             if (this.tamable.distanceToSqr(this.owner) <= 16 * 16) this.tamable.getLookControl().setLookAt(this.owner, 10.0F, this.tamable.getMaxHeadXRot()); // Paper - Limit pet look distance
