--- a/net/minecraft/world/entity/ai/goal/Goal.java
+++ b/net/minecraft/world/entity/ai/goal/Goal.java
@@ -19,6 +_,16 @@
 
     public abstract boolean canUse();
 
+    // KioCG start - optimize entity behavior
+    public boolean registerWhen() {
+        return true;
+    }
+
+    protected int canUseDelay() {
+        return 1;
+    }
+    // KioCG end - optimize entity behavior
+
     public boolean canContinueToUse() {
         return this.canUse();
     }
