--- a/net/minecraft/world/entity/ai/goal/LeapAtTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LeapAtTargetGoal.java
@@ -16,6 +_,13 @@
         this.setFlags(EnumSet.of(Goal.Flag.JUMP, Goal.Flag.MOVE));
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    public boolean registerWhen() {
+        return this.mob.getTarget() != null;
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canUse() {
         if (this.mob.hasControllingPassenger()) {
