--- a/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
@@ -69,6 +_,13 @@
         }
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    protected int canUseDelay() {
+        return 20;
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canContinueToUse() {
         if (this.llama.inCaravan() && this.llama.getCaravanHead().isAlive() && this.firstIsLeashed(this.llama, 0)) {
@@ -106,7 +_,10 @@
         if (this.llama.inCaravan()) {
             if (!(this.llama.getLeashHolder() instanceof LeashFenceKnotEntity)) {
                 Llama caravanHead = this.llama.getCaravanHead();
-                double d = this.llama.distanceTo(caravanHead);
+                // KioCG start - optimize distanceTo
+                double d = this.llama.distanceToSqr(caravanHead);
+                d = d < 2.0 * 2.0 ? net.minecraft.util.Mth.sqrt((float) d) : 2.0;
+                // KioCG end - optimize distanceTo
                 float f = 2.0F;
                 Vec3 vec3 = new Vec3(caravanHead.getX() - this.llama.getX(), caravanHead.getY() - this.llama.getY(), caravanHead.getZ() - this.llama.getZ())
                     .normalize()
