--- a/net/minecraft/world/entity/ai/goal/PathfindToRaidGoal.java
+++ b/net/minecraft/world/entity/ai/goal/PathfindToRaidGoal.java
@@ -31,6 +_,13 @@
             && !((ServerLevel)this.mob.level()).isVillage(this.mob.blockPosition());
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    protected int canUseDelay() {
+        return 20;
+    }
+    // KioCG end - optimize entity behavior
+
     @Override
     public boolean canContinueToUse() {
         return this.mob.hasActiveRaid()
