--- a/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
@@ -46,7 +_,7 @@
     }
 
     private boolean isValidTarget() {
-        return this.mob.getTarget() != null && this.mob.getTarget().isAlive();
+        return this.mob.getTarget() != null && this.mob.getTarget().isAlive() && !this.mob.getTarget().nearDeath(); // KioCG - dont kill my pet
     }
 
     @Override
