--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -70,6 +_,13 @@
         }
     }
 
+    // KioCG start - optimize entity behavior
+    @Override
+    protected int canUseDelay() {
+        return this.calmDown > 0 ? 1 : 10;
+    }
+    // KioCG end - optimize entity behavior
+
     private boolean shouldFollow(LivingEntity entity) {
         return this.items.test(entity.getMainHandItem()) || this.items.test(entity.getOffhandItem());
     }
@@ -133,7 +_,7 @@
     }
 
     protected void navigateTowards(LivingEntity player) { // Paper
-        this.mob.getNavigation().moveTo(player, this.speedModifier);
+        this.mob.getNavigation().moveToLightweight(player, this.speedModifier); // KioCG - optimize entity behavior
     }
 
     public boolean isRunning() {
