--- a/net/minecraft/world/item/equipment/ArmorMaterial.java
+++ b/net/minecraft/world/item/equipment/ArmorMaterial.java
@@ -39,6 +_,14 @@
             );
         }
 
+        // KioCG start - the long dark
+        if (this == ArmorMaterials.LEATHER) {
+            builder.add(
+                Attributes.SPAWN_REINFORCEMENTS_CHANCE, new AttributeModifier(resourceLocation, 6.0, AttributeModifier.Operation.ADD_VALUE), equipmentSlotGroup
+            );
+        }
+        // KioCG end - the long dark
+
         return builder.build();
     }
 }
