--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -194,7 +_,7 @@
     @Override
     protected void dropCustomDeathLoot(ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
         super.dropCustomDeathLoot(level, damageSource, recentlyHit);
-        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull()) {
+        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull(this)) { // KioCG - drop mobs skull chance
             ItemStack itemStack = new ItemStack(Items.PIGLIN_HEAD);
             creeper.increaseDroppedSkulls();
             this.spawnAtLocation(level, itemStack);
@@ -317,7 +_,10 @@
     }
 
     @Override
-    public void setBaby(boolean childZombie) {
+    // KioCG start - entity baby api
+    public void setBaby1(boolean childZombie) {
+        if (childZombie) removeScaleAttributes();
+    // KioCG end - entity baby api
         this.getEntityData().set(DATA_BABY_ID, childZombie);
         if (!this.level().isClientSide) {
             AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
@@ -447,7 +_,7 @@
 
     @Override
     public boolean wantsToPickUp(ServerLevel level, ItemStack stack) {
-        return level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, stack);
+        return true && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, stack); // KioCG - better mob griefing setting
     }
 
     protected boolean canReplaceCurrentItem(ItemStack candidate) {
