--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -317,7 +_,10 @@
     }
 
     @Override
-    public void setBaby(boolean childZombie) {
+    // KioCG start - entity baby api
+    public void setBaby1(boolean childZombie) {
+        if (childZombie) removeScaleAttributes();
+    // KioCG end - entity baby api
         this.getEntityData().set(DATA_BABY_ID, childZombie);
         if (!this.level().isClientSide) {
             AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
