--- a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -40,34 +_,36 @@
             Pair.of(0, WakeUp.create()),
             Pair.of(0, ReactToBell.create()),
             Pair.of(0, SetRaidStatus.create()),
-            Pair.of(0, ValidateNearbyPoi.create(profession.value().heldJobSite(), MemoryModuleType.JOB_SITE)),
-            Pair.of(0, ValidateNearbyPoi.create(profession.value().acquirableJobSite(), MemoryModuleType.POTENTIAL_JOB_SITE)),
+            // KioCG start - optimize villager
+            // Pair.of(0, ValidateNearbyPoi.create(profession.value().heldJobSite(), MemoryModuleType.JOB_SITE)),
+            // Pair.of(0, ValidateNearbyPoi.create(profession.value().acquirableJobSite(), MemoryModuleType.POTENTIAL_JOB_SITE)),
             Pair.of(1, new MoveToTargetSink()),
-            Pair.of(2, PoiCompetitorScan.create()),
-            Pair.of(3, new LookAndFollowTradingPlayerSink(speedModifier)),
-            Pair.of(5, GoToWantedItem.create(villager -> !villager.isSleeping(), speedModifier, false, 4)), // Paper - Fix MC-157464
-            Pair.of(
-                6,
-                AcquirePoi.create(
-                    profession.value().acquirableJobSite(),
-                    MemoryModuleType.JOB_SITE,
-                    MemoryModuleType.POTENTIAL_JOB_SITE,
-                    true,
-                    Optional.empty(),
-                    (level, pos) -> true
-                )
-            ),
-            Pair.of(7, new GoToPotentialJobSite(speedModifier)),
-            Pair.of(8, YieldJobSite.create(speedModifier)),
-            Pair.of(
-                10,
-                AcquirePoi.create(
-                    poiType -> poiType.is(PoiTypes.HOME), MemoryModuleType.HOME, false, Optional.of((byte)14), VillagerGoalPackages::validateBedPoi
-                )
-            ),
-            Pair.of(10, AcquirePoi.create(poiType -> poiType.is(PoiTypes.MEETING), MemoryModuleType.MEETING_POINT, true, Optional.of((byte)14))),
-            Pair.of(10, AssignProfessionFromJobSite.create()),
-            Pair.of(10, ResetProfession.create())
+            // Pair.of(2, PoiCompetitorScan.create()),
+            Pair.of(3, new LookAndFollowTradingPlayerSink(speedModifier))
+            // Pair.of(5, GoToWantedItem.create(villager -> !villager.isSleeping(), speedModifier, false, 4)), // Paper - Fix MC-157464
+            // Pair.of(
+            //     6,
+            //     AcquirePoi.create(
+            //         profession.value().acquirableJobSite(),
+            //         MemoryModuleType.JOB_SITE,
+            //         MemoryModuleType.POTENTIAL_JOB_SITE,
+            //         true,
+            //         Optional.empty(),
+            //         (level, pos) -> true
+            //     )
+            // ),
+            // Pair.of(7, new GoToPotentialJobSite(speedModifier)),
+            // Pair.of(8, YieldJobSite.create(speedModifier)),
+            // Pair.of(
+            //     10,
+            //     AcquirePoi.create(
+            //         poiType -> poiType.is(PoiTypes.HOME), MemoryModuleType.HOME, false, Optional.of((byte)14), VillagerGoalPackages::validateBedPoi
+            //     )
+            // ),
+            // Pair.of(10, AcquirePoi.create(poiType -> poiType.is(PoiTypes.MEETING), MemoryModuleType.MEETING_POINT, true, Optional.of((byte)14))),
+            // Pair.of(10, AssignProfessionFromJobSite.create()),
+            // Pair.of(10, ResetProfession.create())
+            // KioCG end - optimize villager
         );
     }
 
@@ -76,6 +_,8 @@
         return blockState.is(BlockTags.BEDS) && !blockState.getValue(BedBlock.OCCUPIED);
     }
 
+    // KioCG start - optimize villager
+    /*
     public static ImmutableList<Pair<Integer, ? extends BehaviorControl<? super Villager>>> getWorkPackage(
         Holder<VillagerProfession> profession, float speedModifier
     ) {
@@ -108,6 +_,8 @@
             Pair.of(99, UpdateActivityFromSchedule.create())
         );
     }
+     */
+    // KioCG end - optimize villager
 
     public static ImmutableList<Pair<Integer, ? extends BehaviorControl<? super Villager>>> getPlayPackage(float speedModifier) {
         return ImmutableList.of(
@@ -151,6 +_,14 @@
                     )
                 )
             ),
+            // KioCG start - optimize villager: 只在需要时运行
+            Pair.of(
+                10,
+                AcquirePoi.create(
+                    poiType -> poiType.is(PoiTypes.HOME), MemoryModuleType.HOME, false, Optional.of((byte)14), VillagerGoalPackages::validateBedPoi
+                )
+            ),
+            // KioCG end - optimize villager: 只在需要时运行
             getMinimalLookBehavior(),
             Pair.of(99, UpdateActivityFromSchedule.create())
         );
@@ -181,6 +_,7 @@
                     ImmutableList.of(Pair.of(new TradeWithVillager(), 1))
                 )
             ),
+            Pair.of(10, AcquirePoi.create(poiType -> poiType.is(PoiTypes.MEETING), MemoryModuleType.MEETING_POINT, true, Optional.of((byte)14))), // KioCG - optimize villager: 只在需要时运行
             getFullLookBehavior(),
             Pair.of(99, UpdateActivityFromSchedule.create())
         );
