--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -667,6 +_,8 @@
     @NotNull
     public List<Player> matchPlayer(@NotNull String name);
 
+    public boolean isValidPlayerName(@NotNull String name); // KioCG - more api
+
     /**
      * Gets the player with the given UUID.
      *
@@ -1003,6 +_,8 @@
     @Contract("null -> false")
     boolean addRecipe(@Nullable Recipe recipe);
 
+    boolean addRecipe(@NotNull List<Recipe> recipe); // KioCG - more api
+
     // Paper start - method to send recipes immediately
     /**
      * Adds a recipe to the crafting manager.
@@ -1162,6 +_,8 @@
     @NotNull
     public Iterator<Recipe> recipeIterator();
 
+    @NotNull <T extends Recipe> Iterator<T> recipeIterator(@NotNull org.bukkit.inventory.RecipeType<T> recipeType); // KioCG - extended recipe iterator
+
     /**
      * Clears the list of crafting recipes.
      */
@@ -1184,6 +_,8 @@
      */
     public boolean removeRecipe(@NotNull NamespacedKey key);
 
+    public boolean removeRecipe(@NotNull List<NamespacedKey> key); // KioCG - more api
+
     // Paper start - method to resend recipes
     /**
      * Remove a recipe from the server.
@@ -2723,4 +_,15 @@
      */
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
+
+    // KioCG start - more api
+    void registerItem(@NotNull NamespacedKey namespacedKey, @NotNull ItemStack itemStack);
+    boolean unregisterItem(@NotNull NamespacedKey namespacedKey);
+    default @Nullable ItemStack getRegisterItem(@NotNull NamespacedKey namespacedKey) {
+        return getRegisterItem(namespacedKey, 1);
+    }
+    @Nullable ItemStack getRegisterItem(@NotNull NamespacedKey namespacedKey, int amount);
+
+    String formatStatistic(@NotNull Statistic statistic, int value);
+    // KioCG end - more api
 }
