--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -666,6 +_,8 @@
     @NotNull
     public List<Player> matchPlayer(@NotNull String name);
 
+    public boolean isValidPlayerName(@NotNull String name); // KioCG - more api
+
     /**
      * Gets the player with the given UUID.
      *
@@ -1161,6 +_,8 @@
      */
     @NotNull
     public Iterator<Recipe> recipeIterator();
+
+    @NotNull <T extends Recipe> Iterator<T> recipeIterator(@NotNull org.bukkit.inventory.RecipeType<T> recipeType); // KioCG - extended recipe iterator
 
     /**
      * Clears the list of crafting recipes.
