--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -94,6 +_,7 @@
 
     @Override
     protected void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
+        if (entity.isInWater()) // KioCG start - the long dark: 修复坐船也视为在水中的问题
         effectApplier.apply(InsideBlockEffectType.EXTINGUISH);
     }
 
