--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -407,4 +_,129 @@
             }
         }
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+        public String feedbackForwardingSenderName = "MOSS";
+        public java.util.List<String> advancementRegisterPath = java.util.List.of("bacap/advancement_legend");
+        public boolean dontKillMyPet = true;
+        public double petTeleportDistanceSqr = 256.0;
+        public boolean petFollowKeepChunkLoaded = true;
+        public Component airProgressHave = Component.text("\uD83E\uDEE7", NamedTextColor.BLUE);
+        public Component airProgressLost = Component.text("\uD83E\uDEE7", NamedTextColor.GRAY);
+        public Component airProgressO2 = Component.text("O₂ ", NamedTextColor.BLUE).decorate(net.kyori.adventure.text.format.TextDecoration.BOLD);
+        public Component sleepProgressClock = Component.text("\uD83D\uDE34");
+        public Component sleepProgressLeft = Component.text("☽", NamedTextColor.BLUE);
+        public Component sleepProgressRight = Component.text("☀", NamedTextColor.YELLOW);
+        public String sleepProgressContent = "                              ";
+        public int foodFreshnessTickInterval = 1800;
+
+        public ModelItems modelItems;
+        public class ModelItems extends ConfigurationPart {
+
+            public VanillaRecipeCanUse vanillaRecipeCanUse;
+            public class VanillaRecipeCanUse extends ConfigurationPart {
+                private java.util.Set<String> anvil = java.util.Set.of();
+                private java.util.Set<String> beacon = java.util.Set.of();
+                private java.util.Set<String> cartographyTable = java.util.Set.of();
+                private java.util.Set<String> crafting = java.util.Set.of();
+                private java.util.Set<String> enchantment = java.util.Set.of();
+                private java.util.Set<String> grindstone = java.util.Set.of();
+                private java.util.Set<String> loom = java.util.Set.of();
+                private java.util.Set<String> smithing = java.util.Set.of();
+                private java.util.Set<String> stonecutter = java.util.Set.of();
+                private java.util.Set<String> furnace = java.util.Set.of();
+                private java.util.Set<String> brewingStand = java.util.Set.of();
+
+                public boolean allowed(net.minecraft.world.inventory.MenuType<?> menuType, net.minecraft.world.item.ItemStack itemStack) {
+                    net.minecraft.world.item.component.CustomModelData customModelData = itemStack.get(net.minecraft.core.component.DataComponents.CUSTOM_MODEL_DATA);
+                    if (customModelData == null) {
+                        return true;
+                    }
+
+                    if (menuType == net.minecraft.world.inventory.MenuType.ANVIL) {
+                        return include(anvil, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.BEACON) {
+                        return include(beacon, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.CARTOGRAPHY_TABLE) {
+                        return include(cartographyTable, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.CRAFTING) {
+                        return include(crafting, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.ENCHANTMENT) {
+                        return include(enchantment, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.GRINDSTONE) {
+                        return include(grindstone, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.LOOM) {
+                        return include(loom, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.SMITHING) {
+                        return include(smithing, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.STONECUTTER) {
+                        return include(stonecutter, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.FURNACE
+                               || menuType == net.minecraft.world.inventory.MenuType.BLAST_FURNACE
+                               || menuType == net.minecraft.world.inventory.MenuType.SMOKER) {
+                        return include(furnace, customModelData);
+                    } else if (menuType == net.minecraft.world.inventory.MenuType.BREWING_STAND) {
+                        return include(brewingStand, customModelData);
+                    } else {
+                        LOGGER.error("无效的面板类型 {}", menuType);
+                        return false;
+                    }
+                }
+
+                private boolean include(java.util.Set<String> set, net.minecraft.world.item.component.CustomModelData customModelData) {
+                    final Boolean aBoolean = customModelData.getBoolean(0);
+                    if (aBoolean != null && set.contains(aBoolean.toString())) {
+                        return true;
+                    }
+
+                    final Float aFloat = customModelData.getFloat(0);
+                    if (aFloat != null && set.contains(String.valueOf(aFloat.intValue()))) {
+                        return true;
+                    }
+
+                    final String string = customModelData.getString(0);
+                    if (string != null && set.contains(string)) {
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+        }
+
+        public Command command;
+        public class Command extends ConfigurationPart {
+            public boolean allowUnsafeEnchant = true;
+
+            public String tpsBarTitle = "<gray>TPS<yellow>:</yellow> <tps> MSPT<yellow>:</yellow> <mspt> Ping<yellow>:</yellow> <ping>ms ACL<yellow>:</yellow> <hotpct>%";
+            public net.kyori.adventure.bossbar.BossBar.Overlay tpsBarProgressOverlay = net.kyori.adventure.bossbar.BossBar.Overlay.NOTCHED_20;
+            public net.kyori.adventure.bossbar.BossBar.Color tpsBarProgressColorGood = net.kyori.adventure.bossbar.BossBar.Color.GREEN;
+            public net.kyori.adventure.bossbar.BossBar.Color tpsBarProgressColorMedium = net.kyori.adventure.bossbar.BossBar.Color.YELLOW;
+            public net.kyori.adventure.bossbar.BossBar.Color tpsBarProgressColorLow = net.kyori.adventure.bossbar.BossBar.Color.RED;
+            public String tpsBarTextColorGood = "<green><text>";
+            public String tpsBarTextColorMedium = "<yellow><text>";
+            public String tpsBarTextColorLow = "<red><text>";
+            public int tpsBarTickInterval = 20;
+        }
+
+        public Anvil anvil;
+        public class Anvil extends ConfigurationPart {
+            public int maxCumulativeCost = 31;
+            public boolean repairNoLevelCost = true;
+            public boolean renameNoLevelCost = true;
+        }
+
+        public Container container;
+        public class Container extends ConfigurationPart {
+            public int defaultChestRows = 2;
+            public int defaultTrappedChestRows = 2;
+            public int defaultBarrelRows = 2;
+            public int defaultShulkerBoxRows = 2;
+            public boolean lootableCompleteReplenish = true;
+        }
+    }
+    // KioCG end
 }
