--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -247,6 +_,7 @@
                 .serializers(serializers -> serializers
                     .register(new TypeToken<Reference2IntMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<Reference2IntMap<?>>(Reference2IntOpenHashMap::new, Integer.TYPE))
                     .register(new TypeToken<Reference2LongMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<Reference2LongMap<?>>(Reference2LongOpenHashMap::new, Long.TYPE))
+                    .register(new TypeToken<it.unimi.dsi.fastutil.objects.Reference2DoubleMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<it.unimi.dsi.fastutil.objects.Reference2DoubleMap<?>>(it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap::new, Double.TYPE)) // KioCG
                     .register(new TypeToken<Reference2ObjectMap<?, ?>>() {}, new FastutilMapSerializer.SomethingToSomething<Reference2ObjectMap<?, ?>>(Reference2ObjectOpenHashMap::new))
                     .register(new TypeToken<Table<?, ?, ?>>() {}, new TableSerializer())
                     .register(DespawnRange.class, DespawnRange.SERIALIZER)
