--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -587,4 +_,43 @@
         	HORIZONTAL_FIRST_OUTWARD, HORIZONTAL_FIRST_INWARD, VERTICAL_FIRST_OUTWARD, VERTICAL_FIRST_INWARD
         }
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+        public boolean deathKeepHotbarAndArmorItems = true;
+        public boolean deathItemsOnlyPickupByOneself = true;
+        public boolean hopperPullsLockedContainer = false;
+        public boolean hopperPullsEntityOwnedItem = false;
+
+        public TheLongDark theLongDark;
+        public class TheLongDark extends ConfigurationPart {
+
+            public ColdValue coldValue;
+            public class ColdValue extends ConfigurationPart {
+                public double ambientTemperatureBase = 0.0;
+                public double ambientSkyDarkenMultiplier = -0.15;
+                public double biomeTemperatureOffset = -0.65;
+                public double biomeTemperatureMultiplier = 30.0;
+
+                public double playerTemperatureBase = 0.0;
+                public double playerLightBlockMultiplier = 5.0;
+                public double isInWaterOrBubble = -35.0;
+                public double isInRain = -15.0;
+                public double isInLava = 300.0;
+                public double isOnFire = 10.0;
+
+                public double finalTemperatureMultiplier = 1.0;
+            }
+
+            public ThirstValue thirstValue;
+            public class ThirstValue extends ConfigurationPart {
+                public double thirstRegainMultiplierHealth = 1.0;
+                public double thirstRegainMultiplierFood = 2.0;
+                public double lowDownfallBiomeThirst = 0.02;
+            }
+        }
+    }
+    // KioCG end
 }
