--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -587,4 +_,15 @@
         	HORIZONTAL_FIRST_OUTWARD, HORIZONTAL_FIRST_INWARD, VERTICAL_FIRST_OUTWARD, VERTICAL_FIRST_INWARD
         }
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+        public boolean deathKeepHotbarAndArmorItems = true;
+        public boolean deathItemsOnlyPickupByOneself = true;
+        public boolean hopperPullsLockedContainer = false;
+        public boolean hopperPullsEntityOwnedItem = false;
+    }
+    // KioCG end
 }
