--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -577,4 +_,98 @@
         	HORIZONTAL_FIRST_OUTWARD, HORIZONTAL_FIRST_INWARD, VERTICAL_FIRST_OUTWARD, VERTICAL_FIRST_INWARD
         }
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+        public boolean deathKeepHotbarAndArmorItems = true;
+        public boolean deathItemsOnlyPickupByOneself = true;
+        public boolean hopperPullsLockedContainer = false;
+        public boolean hopperPullsEntityOwnedItem = false;
+        public boolean hopperPullsInfinitePickupItem = false;
+        public boolean hopperRailsOptimize = true;
+
+        public TheLongDark theLongDark;
+        public class TheLongDark extends ConfigurationPart {
+
+            public ColdValue coldValue;
+            public class ColdValue extends ConfigurationPart {
+                public double ambientTemperatureBase = 0.0;
+                public double ambientSkyDarkenMultiplier = -0.15;
+                public double biomeTemperatureOffset = -0.65;
+                public double biomeTemperatureMultiplier = 30.0;
+
+                public double playerTemperatureBase = 0.0;
+                public double playerLightBlockMultiplier = 5.0;
+                public double isInWaterOrBubble = -35.0;
+                public double isInRain = -15.0;
+                public double isInLava = 300.0;
+                public double isOnFire = 10.0;
+
+                public double finalTemperatureMultiplier = 1.0;
+            }
+
+            public ThirstValue thirstValue;
+            public class ThirstValue extends ConfigurationPart {
+                public double thirstRegainMultiplierHealth = 2.0;
+                public double thirstRegainMultiplierFood = 6.0;
+                public double lowDownfallBiomeMultiplier = 1.5;
+            }
+
+            public int campfireBoilingWaterTime = 600;
+        }
+
+        public Entity entity;
+        public class Entity extends ConfigurationPart {
+            public it.unimi.dsi.fastutil.objects.Reference2DoubleMap<EntityType<?>> sizeScalingBaby = Util.make(new it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
+                map.defaultReturnValue(0.5);
+                map.put(EntityType.SKELETON, 0.65);
+            });
+            public it.unimi.dsi.fastutil.objects.Reference2DoubleMap<EntityType<?>> sizeScalingHuge = Util.make(new it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
+                map.defaultReturnValue(1.5);
+                map.put(EntityType.SKELETON, 1.5);
+            });
+
+            public KiocgConfig.Entity.Monster monster;
+            public class Monster extends ConfigurationPart {
+                public boolean spawnInLightAtNight = true;
+                public int darkEnoughToSprint = 7;
+                public double zombiesSkinnedChance = 0.25;
+                public boolean babyZombieJockey = true;
+                public boolean persistentWhenPickedUpItems = false;
+
+                public it.unimi.dsi.fastutil.objects.Reference2DoubleMap<EntityType<?>> poweredCreeperDropMobsSkullChance = Util.make(new it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
+                    map.defaultReturnValue(1.0);
+                    map.put(EntityType.WITHER_SKELETON, 0.0);
+                });
+            }
+
+            public KiocgConfig.Entity.Animal animal;
+            public class Animal extends ConfigurationPart {
+                public boolean villagerCanBreed = false;
+                public double feedingHugeChance = 0.01;
+                public int lobotomizeCheckInterval = 100;
+            }
+
+            public boolean turretEnable = true;
+            public int turretCooldownDefault = 100;
+            public double turretRangeDefault = 8.0;
+            public double turretDamageDefault = 1.0;
+            public boolean itemEntityExplosionKnockback = true;
+        }
+
+        public int sleepHealingRegenRate = 10;
+        public double saplingsWitheredChance = 0.025;
+        public double flowersAndGrassesSpreadChance = 0.01;
+
+        public ChunkHot chunkHot;
+        public class ChunkHot extends ConfigurationPart {
+            public long phaseGood = 300_000L;
+            public long phaseMedium = 500_000L;
+            public String mediumMessage = "<red><i>这片区域令你感到不适";
+            public String lowMessage = "<dark_red><i><b>古神在你耳边窃窃私语";
+        }
+    }
+    // KioCG end
 }
