--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -155,12 +_,22 @@
                 flag = true;
                 level.gameEvent(player, GameEvent.SHEAR, pos);
             } else if (stack.is(Items.GLASS_BOTTLE)) {
+                // KioCG start - PlayerBottleEvent
+                com.kiocg.event.player.PlayerBottleFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBottleFillEvent(level, player, pos, stack, new ItemStack(Items.HONEY_BOTTLE), hand);
+                if (event.isCancelled()) {
+                    player.containerMenu.sendAllDataToRemote();
+                    return InteractionResult.PASS;
+                }
+                final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+                // KioCG end - PlayerBottleEvent
                 stack.shrink(1);
                 level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 if (stack.isEmpty()) {
-                    player.setItemInHand(hand, new ItemStack(Items.HONEY_BOTTLE));
-                } else if (!player.getInventory().add(new ItemStack(Items.HONEY_BOTTLE))) {
-                    player.drop(new ItemStack(Items.HONEY_BOTTLE), false);
+                    // KioCG start - PlayerBottleEvent
+                    player.setItemInHand(hand, resultItem);
+                } else if (!player.getInventory().add(resultItem)) {
+                    player.drop(resultItem, false);
+                    // KioCG end - PlayerBottleEvent
                 }
 
                 flag = true;
