--- a/net/minecraft/world/level/block/BeetrootBlock.java
+++ b/net/minecraft/world/level/block/BeetrootBlock.java
@@ -46,12 +_,16 @@
         return Items.BEETROOT_SEEDS;
     }
 
+    // KioCG start - crops grow every day
+    /*
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (random.nextInt(3) != 0) {
             super.randomTick(state, level, pos, random);
         }
     }
+    */
+    // KioCG end - crops grow every day
 
     @Override
     protected int getBonemealAgeIncrease(Level level) {
