--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -50,6 +_,18 @@
         }
     }
 
+    // KioCG start - crops grow every day
+    @Override
+    protected boolean isRandomlyTicking(BlockState state) {
+        return false;
+    }
+
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        return true;
+    }
+    // KioCG end - crops grow every day
+
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         BlockPos blockPos = pos.above();
@@ -63,12 +_,18 @@
                 }
             }
 
-            if (ageValue == 8 && this.canSurvive(this.defaultBlockState(), level, pos.above())) {
+            // KioCG start - crops grow every day
+            if (random.nextFloat() < (level.spigotConfig.cactusModifier / 100.0F)) return; // Spigot - SPIGOT-7159: Better modifier
+            if (ageValue == 0 && this.canSurvive(this.defaultBlockState(), level, pos.above())) { // 只在ageValue == 0时生成仙人掌花
+            // KioCG end - crops grow every day
                 double d = i >= level.paperConfig().maxGrowthHeight.cactus ? 0.25 : 0.1; // Paper - Configurable cactus/bamboo/reed growth height
                 if (random.nextDouble() <= d) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockPos, Blocks.CACTUS_FLOWER.defaultBlockState(), 3);
                 }
-            } else if (ageValue == 15 && i < level.paperConfig().maxGrowthHeight.cactus) { // Paper - Configurable cactus/bamboo/reed growth height
+            // KioCG start - crops grow every day
+            }
+            if (level.isEmptyBlock(blockPos) && i < level.paperConfig().maxGrowthHeight.cactus) { // Paper - Configurable cactus/bamboo/reed growth height
+            // KioCG end - crops grow every day
                 // Paper start
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockPos, this.defaultBlockState(), 3)) {
                     return;
