--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -54,6 +_,13 @@
 
     @Override
     protected boolean isRandomlyTicking(BlockState state) {
+        // KioCG start - crops grow every day
+        return false;
+    }
+
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        // KioCG end - crops grow every day
         return state.getValue(AGE) < 5;
     }
 
@@ -144,6 +_,14 @@
     private void placeGrownFlower(Level level, BlockPos pos, int age) {
         // level.setBlock(pos, this.defaultBlockState().setValue(AGE, age), 2); // Paper - already done above in the event call
         level.levelEvent(1033, pos, 0);
+        // KioCG start - crops grow every day: 随机继续生长
+        if (level.random.nextFloat() < 0.8F) {
+            BlockState blockState = level.getBlockState(pos);
+            if (blockState.is(this)) {
+                blockState.randomTick((ServerLevel) level, pos, level.random);
+            }
+        }
+        // KioCG end - crops grow every day: 随机继续生长
     }
 
     private void placeDeadFlower(Level level, BlockPos pos) {
