--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -59,12 +_,19 @@
 
     @Override
     protected boolean isRandomlyTicking(BlockState state) {
+        // KioCG start - crops grow every day
+        return false;
+    }
+
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        // KioCG end - crops grow every day
         return state.getValue(AGE) < 2;
     }
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.random.nextFloat() < (level.spigotConfig.cocoaModifier / (100.0f * 5))) { // Spigot - SPIGOT-7159: Better modifier resolution
+        if (level.random.nextFloat() < (level.spigotConfig.cocoaModifier / 100.0f)) { // Spigot - SPIGOT-7159: Better modifier resolution // KioCG - crops grow every day
             int ageValue = state.getValue(AGE);
             if (ageValue < 2) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state.setValue(CocoaBlock.AGE, ageValue + 1), 2); // CraftBukkit
