--- a/net/minecraft/world/level/block/CrafterBlock.java
+++ b/net/minecraft/world/level/block/CrafterBlock.java
@@ -200,6 +_,8 @@
         Direction direction = state.getValue(ORIENTATION).front();
         Container containerAt = HopperBlockEntity.getContainerAt(level, pos.relative(direction));
         ItemStack itemStack = stack.copy();
+
+        if (containerAt != null && HopperBlockEntity.suckInFilter(containerAt, stack)) { // KioCG - hopper filter
         if (containerAt != null && (containerAt instanceof CrafterBlockEntity || stack.getCount() > containerAt.getMaxStackSize(stack))) {
             // CraftBukkit start - InventoryMoveItemEvent
             org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack);
@@ -255,6 +_,7 @@
                 }
             }
         }
+        } // KioCG - hopper filter
 
         if (!itemStack.isEmpty()) {
             Vec3 vec3 = Vec3.atCenterOf(pos);
