--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -204,10 +_,46 @@
             level.setBlock(pos, state, 10);
             this.playSound(player, level, pos, state.getValue(OPEN));
             level.gameEvent(player, this.isOpen(state) ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
+
+            // KioCG start - double door opening
+            BlockPos relativeDoorPos = getRelativeDoor(state, (ServerLevel) level, pos);
+            if (relativeDoorPos != null) {
+                boolean open = !state.getValue(OPEN);
+                BlockState relativeDoorState = level.getBlockState(relativeDoorPos);
+                if (open == relativeDoorState.getValue(OPEN)) {
+                    setOpenWithoutSound(null, level, relativeDoorState, relativeDoorPos, !open);
+                }
+            }
+            // KioCG end - double door opening
+
             return InteractionResult.SUCCESS;
         }
     }
 
+    // KioCG start - double door opening
+    private @Nullable BlockPos getRelativeDoor(BlockState state, ServerLevel level, BlockPos pos) {
+        Direction facing = state.getValue(DoorBlock.FACING);
+        DoorHingeSide hinge = state.getValue(DoorBlock.HINGE);
+        boolean hingeRight = hinge == DoorHingeSide.RIGHT;
+        Direction relative = switch (facing) {
+            case NORTH -> hingeRight ? Direction.WEST : Direction.EAST;
+            case EAST -> hingeRight ? Direction.NORTH : Direction.SOUTH;
+            case SOUTH -> hingeRight ? Direction.EAST : Direction.WEST;
+            case WEST -> hingeRight ? Direction.SOUTH : Direction.NORTH;
+            default -> throw new IllegalStateException("Unexpected value: " + facing);
+        };
+        BlockPos relativePos = pos.relative(relative);
+        BlockState relativeState = level.getBlockState(relativePos);
+        if (relativeState.is(this)
+            && relativeState.getValue(DoorBlock.FACING) == facing
+            && relativeState.getValue(DoorBlock.HINGE) != hinge
+            && relativeState.getValue(DoorBlock.HALF) == state.getValue(DoorBlock.HALF)) {
+            return relativePos;
+        }
+        return null;
+    }
+    // KioCG end - double door opening
+
     public boolean isOpen(BlockState state) {
         return state.getValue(OPEN);
     }
@@ -219,6 +_,15 @@
             level.gameEvent(entity, _open ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
         }
     }
+
+    // KioCG start - double door opening
+    public void setOpenWithoutSound(@Nullable Entity entity, Level level, BlockState state, BlockPos pos, boolean _open) {
+        if (state.is(this) && state.getValue(OPEN) != _open) {
+            level.setBlockAndUpdate(pos, state.setValue(OPEN, _open)); // 需要更新客户端
+            level.gameEvent(entity, _open ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
+        }
+    }
+    // KioCG end - double door opening
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston) {
