--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -112,9 +_,9 @@
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
         super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (level instanceof ServerLevel serverLevel
-            && level.random.nextFloat() < fallDistance - 0.5F
+            && fallDistance > 1.0F && level.random.nextFloat() < fallDistance - 0.5F // KioCG - 摔落大于1格才能退化耕地
             && entity instanceof LivingEntity
-            && (entity instanceof Player || serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
+            && (false || serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) // KioCG - better mob griefing setting
             && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
                 // CraftBukkit start - Interact soil
                 org.bukkit.event.Cancellable cancellable;
