--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -108,13 +_,20 @@
         }
     }
 
+    // KioCG start - crops grow every day
+    public static boolean isMoisture(ServerLevel level, BlockPos pos, RandomSource random) {
+        BlockState iblockdata = level.getBlockState(pos);
+        return iblockdata.is(Blocks.FARMLAND) && iblockdata.getValue(FarmBlock.MOISTURE) > 0;
+    }
+    // KioCG end - crops grow every day
+
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
         super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (level instanceof ServerLevel serverLevel
-            && level.random.nextFloat() < fallDistance - 0.5
+            && fallDistance > 1.0 && level.random.nextFloat() < fallDistance - 0.5 // KioCG - 摔落大于1格才能退化耕地
             && entity instanceof LivingEntity
-            && (entity instanceof Player || serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
+            && (false || serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) // KioCG - better mob griefing setting
             && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
                 // CraftBukkit start - Interact soil
                 org.bukkit.event.Cancellable cancellable;
