--- a/net/minecraft/world/level/block/GrassBlock.java
+++ b/net/minecraft/world/level/block/GrassBlock.java
@@ -30,8 +_,13 @@
     }
 
     @Override
-    public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state) {
-        return level.getBlockState(pos.above()).isAir();
+    public final boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state) { // KiCG - final防止重写
+        // KioCG start - Perf: optimize flowers and Grasses spreading
+        return isValidBonemealTarget(level.getChunk(pos), level, pos, state);
+    }
+    private boolean isValidBonemealTarget(net.minecraft.world.level.chunk.ChunkAccess chunk, LevelReader level, BlockPos pos, BlockState state) {
+        return chunk.getBlockState(pos.above()).isAir();
+        // KioCG end - Perf: optimize flowers and Grasses spreading
     }
 
     @Override
@@ -40,25 +_,43 @@
     }
 
     @Override
-    public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
+    public final void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) { // KiCG - final防止重写
+        // KioCG start - flowers and grasses spread naturally
+        performBonemeal(level, random, pos, state, 0);
+    }
+    public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state, int start) {
+        // KioCG end - flowers and grasses spread naturally
         BlockPos blockPos = pos.above();
         BlockState blockState = Blocks.SHORT_GRASS.defaultBlockState();
         Optional<Holder.Reference<PlacedFeature>> optional = level.registryAccess()
             .lookupOrThrow(Registries.PLACED_FEATURE)
             .get(VegetationPlacements.GRASS_BONEMEAL);
 
+        // KioCG start - Perf: optimize flowers and Grasses spreading
+        final net.minecraft.world.level.chunk.ChunkAccess cachedBlockChunk = level.getChunkIfLoaded(pos);
+        if (cachedBlockChunk == null) { // Is this needed?
+            return;
+        }
+        // KioCG end - Perf: optimize flowers and Grasses spreading
+
         label51:
-        for (int i = 0; i < 128; i++) {
+        for (int i = start; i < 128; i++) { // KioCG - flowers and grasses spread naturally
             BlockPos blockPos1 = blockPos;
 
             for (int i1 = 0; i1 < i / 16; i1++) {
                 blockPos1 = blockPos1.offset(random.nextInt(3) - 1, (random.nextInt(3) - 1) * random.nextInt(3) / 2, random.nextInt(3) - 1);
-                if (!level.getBlockState(blockPos1.below()).is(this) || level.getBlockState(blockPos1).isCollisionShapeFullBlock(level, blockPos1)) {
+                // KioCG start - Perf: optimize flowers and Grasses spreading
+                final net.minecraft.world.level.chunk.ChunkAccess access = getCachedBlockChunk(cachedBlockChunk, level, blockPos1);
+                if (!access.getBlockState(blockPos1.below()).is(this) || access.getBlockState(blockPos1).isCollisionShapeFullBlock(level, blockPos1)) {
+                // KioCG end - Perf: optimize flowers and Grasses spreading
                     continue label51;
                 }
             }
 
-            BlockState blockState1 = level.getBlockState(blockPos1);
+            // KioCG start - Perf: optimize flowers and Grasses spreading
+            final net.minecraft.world.level.chunk.ChunkAccess access = getCachedBlockChunk(cachedBlockChunk, level, blockPos1);
+            BlockState blockState1 = access.getBlockState(blockPos1);
+            // KioCG end - Perf: optimize flowers and Grasses spreading
             if (blockState1.is(blockState.getBlock()) && random.nextInt(10) == 0) {
                 BonemealableBlock bonemealableBlock = (BonemealableBlock)blockState.getBlock();
                 if (bonemealableBlock.isValidBonemealTarget(level, blockPos1, blockState1)) {
@@ -93,4 +_,47 @@
     public BonemealableBlock.Type getType() {
         return BonemealableBlock.Type.NEIGHBOR_SPREADER;
     }
+
+    // KioCG start - flowers and grasses spread naturally
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        return true;
+    }
+
+    @Override
+    protected void dayTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (random.nextFloat() > (float) level.paperConfig().kiocgConfig.block.flowersAndGrassesSpreadChance) { return; }
+        final net.minecraft.world.level.chunk.ChunkAccess cachedBlockChunk = level.getChunkIfLoaded(pos);
+        if (cachedBlockChunk == null) { // Is this needed?
+            return;
+        }
+
+        if (this.isValidBonemealTarget(cachedBlockChunk, level, pos, state)) {
+            int bushBlockCount = 0;
+            int grassBlockCount = 1;
+
+            for (BlockPos blockPos : BlockPos.betweenClosed(pos.offset(-2, 0, -2), pos.offset(2, 2, 2))) {
+                if (pos.getX() == blockPos.getX() && pos.getY() == blockPos.getY() && pos.getZ() == blockPos.getZ()) {
+                    continue;
+                }
+
+                final net.minecraft.world.level.chunk.ChunkAccess access = getCachedBlockChunk(cachedBlockChunk, level, blockPos);
+                final Block block = access.getBlockState(blockPos).getBlock();
+                if (block instanceof VegetationBlock) {
+                    if (++bushBlockCount >= 4) {
+                        return;
+                    }
+                } else if (block == this) {
+                    ++grassBlockCount;
+                }
+            }
+
+            if (grassBlockCount <= 10 && bushBlockCount >= 1) {
+                return;
+            }
+
+            this.performBonemeal(level, random, pos, state, 96);
+        }
+    }
+    // KioCG end - flowers and grasses spread naturally
 }
