--- a/net/minecraft/world/level/block/LadderBlock.java
+++ b/net/minecraft/world/level/block/LadderBlock.java
@@ -61,11 +_,29 @@
     }
 
     @Override
-    protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+    protected final boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) { // KiCG - final防止重写
+        // KioCG start - ladder survive dangling
+        return this.canSurviveVanilla(state, level, pos) || this.canSurviveDangling(state, level, pos);
+    }
+    private boolean canSurviveDangling(BlockState state, LevelReader level, BlockPos pos) {
+        BlockState aboveState = level.getBlockState(pos.above());
+        return aboveState.is(this) && aboveState.getValue(FACING) == state.getValue(FACING);
+    }
+    private boolean canSurviveVanilla(BlockState state, LevelReader level, BlockPos pos) {
+        // KioCG end - ladder survive dangling
         Direction direction = state.getValue(FACING);
         return this.canAttachTo(level, pos.relative(direction.getOpposite()), direction);
     }
 
+    // KioCG start - ladder survive dangling
+    @Override
+    protected void tick(BlockState state, net.minecraft.server.level.ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!state.canSurvive(level, pos)) {
+            level.destroyBlock(pos, true);
+        }
+    }
+    // KioCG end - ladder survive dangling
+
     @Override
     protected BlockState updateShape(
         BlockState state,
@@ -77,8 +_,11 @@
         BlockState neighborState,
         RandomSource random
     ) {
-        if (direction.getOpposite() == state.getValue(FACING) && !state.canSurvive(level, pos)) {
-            return Blocks.AIR.defaultBlockState();
+        // KioCG start - ladder survive dangling
+        if ((direction.getOpposite() == Direction.DOWN || direction.getOpposite() == state.getValue(FACING)) && !state.canSurvive(level, pos)) {
+            scheduledTickAccess.scheduleTick(pos, this, 1);
+            return super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
+        // KioCG end - ladder survive dangling
         } else {
             if (state.getValue(WATERLOGGED)) {
                 scheduledTickAccess.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(level));
@@ -106,7 +_,13 @@
         for (Direction direction : context.getNearestLookingDirections()) {
             if (direction.getAxis().isHorizontal()) {
                 blockState = blockState.setValue(FACING, direction.getOpposite());
-                if (blockState.canSurvive(level, clickedPos)) {
+                // KioCG start - ladder survive dangling
+                boolean surviveDangling = false;
+                if (this.canSurviveVanilla(blockState, level, clickedPos) || (surviveDangling = this.canSurviveDangling(blockState, level, clickedPos))) {
+                    if (surviveDangling && context.getPlayer() != null) {
+                        this.playPlaceSound(blockState, (net.minecraft.world.level.Level) level, clickedPos, context.getPlayer(), context.getHand());
+                    }
+                // KioCG end - ladder survive dangling
                     return blockState.setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
                 }
             }
