--- a/net/minecraft/world/level/block/LavaCauldronBlock.java
+++ b/net/minecraft/world/level/block/LavaCauldronBlock.java
@@ -37,6 +_,7 @@
         if (this.isEntityInsideContent(state, pos, entity)) {
             entity.lavaIgnite(pos); // Paper - track lava contact
             entity.lavaHurt(pos); // Paper - track lava contact
+            if (entity instanceof net.minecraft.world.entity.player.Player player) player.coldData.isInLava = true; // KioCG - the long dark
         }
     }
 
