--- a/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -72,6 +_,7 @@
             // CraftBukkit end
             }
         }
+        if (entity instanceof net.minecraft.world.entity.player.Player player && this.isEntityInsideContent(state, pos, entity)) player.coldData.isInWater = true; // KioCG - the long dark
     }
 
     // CraftBukkit start
@@ -112,6 +_,15 @@
         if (!event.callEvent()) {
             return false;
         }
+
+        // KioCG start - campfire boiling water
+        if (level.getBlockEntity(pos.below()) instanceof net.minecraft.world.level.block.entity.CampfireBlockEntity campfireBlockEntity) {
+            int oldLevel = level.getBlockState(pos).getValue(LayeredCauldronBlock.LEVEL);
+            int newLevel = newBlock.getValue(LayeredCauldronBlock.LEVEL);
+            campfireBlockEntity.cauldronLevelChanged(oldLevel, newLevel);
+        }
+        // KioCG end - campfire boiling water
+
         newState.update(true);
         if (sendGameEvent) level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newBlock)); // Paper - Call CauldronLevelChangeEvent
         return true;
