--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -44,13 +_,20 @@
 
     @Override
     protected boolean isRandomlyTicking(BlockState state) {
+        // KioCG start - crops grow every day
+        return false;
+    }
+
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        // KioCG end - crops grow every day
         return state.getValue(AGE) < 3;
     }
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         int ageValue = state.getValue(AGE);
-        if (ageValue < 3 && random.nextFloat() < (level.spigotConfig.wartModifier / (100.0F * 10))) { // Spigot - SPIGOT-7159: Better modifier resolution
+        if (ageValue < 3 && random.nextFloat() < (level.spigotConfig.wartModifier / 100.0F)) { // Spigot - SPIGOT-7159: Better modifier resolution // KioCG - crops grow every day
             state = state.setValue(AGE, ageValue + 1);
             org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
         }
