--- a/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -119,7 +_,7 @@
     @Override
     public void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)) { // KioCG - better mob griefing setting
             serverLevel.destroyBlock(pos, true, entity);
         }
     }
@@ -135,14 +_,20 @@
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // KioCG start - crops grow every day
+        return false;
+    }
+
+    @Override
+    protected boolean isDailyTicking(BlockState state) {
+        // KioCG end - crops grow every day
         return state.getValue(HALF) == DoubleBlockHalf.LOWER && !this.isMaxAge(state);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         float growthSpeed = CropBlock.getGrowthSpeed(this, level, pos);
-        boolean flag = random.nextFloat() < (level.spigotConfig.pitcherPlantModifier / (100.0F * (Math.floor(25.0F / growthSpeed) + 1))); // Paper - Fix Spigot growth modifiers
-        if (flag) {
+        if (level.spigotConfig.pitcherPlantModifier > 0 && FarmBlock.isMoisture(level, pos.below(), level.random) && random.nextFloat() < (level.spigotConfig.pitcherPlantModifier / 100.0F)) { // Paper - Fix Spigot growth modifiers // KioCG - crops grow every day
             this.grow(level, state, pos, 1);
         }
     }
