--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -45,7 +_,12 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.getMaxLocalRawBrightness(pos.above()) >= 9 && random.nextFloat() < (level.spigotConfig.saplingModifier / (100.0F * 7))) { // Spigot - SPIGOT-7159: Better modifier resolution
+        // KioCG start - saplings withered
+        if (random.nextFloat() < (level.spigotConfig.saplingModifier / (100.0F * 7))) { // Spigot - SPIGOT-7159: Better modifier resolution
+            if (random.nextFloat() < (float) level.paperConfig().kiocgConfig.block.saplingsWitheredChance) {
+                level.setBlockAndUpdate(pos, Blocks.DEAD_BUSH.defaultBlockState());
+            } else if (level.getMaxLocalRawBrightness(pos.above()) >= 9)
+        // KioCG end - saplings withered
             this.advanceTree(level, pos, state, random);
         }
     }
