--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -63,9 +_,36 @@
 
     @Override
     protected boolean mayPlaceOn(BlockState state, BlockGetter level, BlockPos pos) {
+        if (mayPlaceOnFruit(state)) return true; // KioCG - extended block placement
         return state.is(Blocks.FARMLAND);
     }
 
+    // KioCG start - extended block placement
+    protected boolean mayPlaceOnFruit(BlockState state) {
+        if (this == Blocks.PUMPKIN_STEM) {
+            return state.is(Blocks.PUMPKIN) || state.is(Blocks.CARVED_PUMPKIN) || state.is(Blocks.JACK_O_LANTERN);
+        } else if (this == Blocks.MELON_STEM) {
+            return state.is(Blocks.MELON);
+        }
+        return false;
+    }
+
+    @Override
+    public BlockState getStateForPlacement(net.minecraft.world.item.context.BlockPlaceContext context) {
+        BlockState blockState = this.defaultBlockState();
+        Level level = context.getLevel();
+        BlockPos clickedPos = context.getClickedPos();
+        if (mayPlaceOnFruit(level.getBlockState(clickedPos.below()))) {
+            if (context.getPlayer() != null) {
+                this.playPlaceSound(blockState, level, clickedPos, context.getPlayer(), context.getHand());
+            }
+            return blockState.setValue(AGE, 1);
+        } else {
+            return super.getStateForPlacement(context);
+        }
+    }
+    // KioCG end - extended block placement
+
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (level.getRawBrightness(pos, 0) >= 9) {
@@ -115,6 +_,11 @@
     @Override
     public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         int min = Math.min(7, state.getValue(AGE) + Mth.nextInt(level.random, 2, 5));
+        // KioCG start - extended block placement
+        if (mayPlaceOnFruit(level.getBlockState(pos.below()))) {
+            min = Math.min(MAX_AGE, state.getValue(AGE) + 1);
+        }
+        // KioCG end - extended block placement
         BlockState blockState = state.setValue(AGE, min);
         org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, blockState, 2); // CraftBukkit
         if (min == 7) {
