--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -171,6 +_,24 @@
         }
     }
 
+    // KioCG start - automatic door closing
+    @Override
+    protected boolean isRandomlyTicking(BlockState state) {
+        return state.getValue(OPEN) && !state.getValue(POWERED);
+    }
+
+    @Override
+    protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (state.getValue(OPEN) != state.getValue(POWERED)) {
+            if (random.nextFloat() < 0.05688889F * 0.1F) {
+                boolean open = state.getValue(OPEN);
+                level.setBlockAndUpdate(pos, state.setValue(OPEN, !open));
+                level.gameEvent(null, !open ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
+            }
+        }
+    }
+    // KioCG end - automatic door closing
+
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = this.defaultBlockState();
