--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -192,6 +_,6 @@
         return !(entity instanceof Turtle)
             && !(entity instanceof Bat)
             && entity instanceof LivingEntity
-            && (entity instanceof Player || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING));
+            && (false || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1)); // KioCG - better mob griefing setting
     }
 }
