--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -295,7 +_,7 @@
     private final EntityMetadataStore entityMetadata = new EntityMetadataStore();
     private final PlayerMetadataStore playerMetadata = new PlayerMetadataStore();
     private final WorldMetadataStore worldMetadata = new WorldMetadataStore();
-    private final Object2IntOpenHashMap<SpawnCategory> spawnCategoryLimit = new Object2IntOpenHashMap<>();
+    private final java.util.EnumMap<SpawnCategory, Integer> spawnCategoryLimit = new java.util.EnumMap<>(SpawnCategory.class); // KioCG - EnumMap
     private File container;
     private WarningState warningState = WarningState.DEFAULT;
     public ApiVersion minimumAPI;
@@ -316,6 +_,7 @@
     private final io.papermc.paper.potion.PaperPotionBrewer potionBrewer;
     public final io.papermc.paper.SparksFly spark;
     private final ServerConfiguration serverConfig = new PaperServerConfiguration();
+    private final Map<NamespacedKey, ItemStack> registerItems = new java.util.HashMap<>(); // KioCG - more api
 
     // Paper start - Folia region threading API
     private final io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler regionizedScheduler = new io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler();
@@ -715,6 +_,13 @@
         return (player != null) ? player.getBukkitEntity() : null;
     }
 
+    // KioCG start - more api
+    @Override
+    public boolean isValidPlayerName(String name) {
+        return net.minecraft.util.StringUtil.isReasonablePlayerName(name);
+    }
+    // KioCG end - more api
+
     @Override
     public Player getPlayer(UUID id) {
         Preconditions.checkArgument(id != null, "UUID id cannot be null");
@@ -1531,6 +_,21 @@
         this.playerList.saveAll();
     }
 
+    // KioCG start - more api
+    @Override
+    public boolean addRecipe(List<Recipe> recipe) {
+        boolean success = false;
+        net.minecraft.world.item.crafting.RecipeManager recipeManager = MinecraftServer.getServer().getRecipeManager();
+        recipeManager.updateRecipe = false;
+        for (Recipe recipe1 : recipe) {
+            success |= addRecipe(recipe1);
+        }
+        recipeManager.updateRecipe = true;
+        recipeManager.finalizeRecipeLoading();
+        return success;
+    }
+    // KioCG end - more api
+
     @Override
     public boolean addRecipe(Recipe recipe) {
         // Paper start - API for updating recipes on clients
@@ -1573,7 +_,7 @@
         toAdd.addToCraftingManager();
         // Paper start - API for updating recipes on clients
         if (true || resendRecipes) { // Always needs to be resent now... TODO
-            this.playerList.reloadRecipes();
+            // this.playerList.reloadRecipes(); // KioCG - more api: 数据包已在内部重新发送
         }
         // Paper end - API for updating recipes on clients
         return true;
@@ -1751,6 +_,13 @@
         return new RecipeIterator();
     }
 
+    // KioCG start - extended recipe iterator
+    @Override
+    public <T extends Recipe> Iterator<T> recipeIterator(org.bukkit.inventory.RecipeType<T> recipeType) {
+        return new RecipeIterator<>(recipeType);
+    }
+    // KioCG end - extended recipe iterator
+
     @Override
     public void clearRecipes() {
         this.console.getRecipeManager().clearRecipes();
@@ -1761,6 +_,21 @@
         this.reloadData(); // Not ideal but hard to reload a subset of a resource pack
     }
 
+    // KioCG start - more api
+    @Override
+    public boolean removeRecipe(List<NamespacedKey> key) {
+        boolean success = false;
+        net.minecraft.world.item.crafting.RecipeManager recipeManager = MinecraftServer.getServer().getRecipeManager();
+        recipeManager.updateRecipe = false;
+        for (NamespacedKey key1 : key) {
+            success |= removeRecipe(key1);
+        }
+        recipeManager.updateRecipe = true;
+        recipeManager.finalizeRecipeLoading();
+        return success;
+    }
+    // KioCG end - more api
+
     @Override
     public boolean removeRecipe(NamespacedKey recipeKey) {
         // Paper start - API for updating recipes on clients
@@ -1776,7 +_,7 @@
         final ResourceKey<net.minecraft.world.item.crafting.Recipe<?>> minecraftKey = CraftRecipe.toMinecraft(recipeKey);
         final boolean removed = this.getServer().getRecipeManager().removeRecipe(minecraftKey);
         if (removed/* && resendRecipes*/) { // TODO Always need to resend them rn - deprecate this method?
-            this.playerList.reloadRecipes();
+            // this.playerList.reloadRecipes(); // KioCG - more api: 数据包已在内部重新发送
         }
         return removed;
         // Paper end - resend recipes on successful removal
@@ -3186,4 +_,33 @@
     public void allowPausing(final Plugin plugin, final boolean value) {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
+
+    // KioCG start - more api
+    @Override
+    public void registerItem(@NotNull NamespacedKey namespacedKey, @NotNull ItemStack itemStack) {
+        Preconditions.checkArgument(!itemStack.isEmpty(), "ItemStack cannot be empty");
+        if (this.registerItems.putIfAbsent(namespacedKey, itemStack.asOne()) != null) {
+            throw new IllegalArgumentException("NamespacedKey has already been registered");
+        }
+    }
+
+    @Override
+    public boolean unregisterItem(@NotNull NamespacedKey namespacedKey) {
+        return this.registerItems.remove(namespacedKey) != null;
+    }
+
+    @Override
+    public @Nullable ItemStack getRegisterItem(@NotNull NamespacedKey namespacedKey, int amount) {
+        final ItemStack itemStack = this.registerItems.get(namespacedKey);
+        if (itemStack == null) {
+            return null;
+        }
+        return itemStack.asQuantity(amount);
+    }
+
+    @Override
+    public String formatStatistic(@NotNull org.bukkit.Statistic statistic, int value) {
+        return CraftStatistic.getNMSStatistic(statistic).format(value);
+    }
+    // KioCG end - more api
 }
