From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 22 Mar 2024 16:21:27 +0800
Subject: [PATCH] Player in structure API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index eb34f22e973fe46529eb93f435f5bf7f85091404..4da235dc7dd409bf07760c0b988987eaedc7ca21 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3676,6 +3676,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     String getClientBrandName();
     // Paper end
 
+    boolean isInStructure(); // KioCG
+
     // Paper start - Teleport API
     /**
      * Sets the player's rotation.
