From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 22 Mar 2024 16:21:27 +0800
Subject: [PATCH] Player in structure API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6c327a07bf8a6aa11a2d7dad12b2830acc539484..ea44f1548ef1156e0e2b2f622d234c2d2f73b538 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3628,6 +3628,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     String getClientBrandName();
     // Paper end
 
+    boolean isInStructure(); // KioCG
+
     // Paper start - Teleport API
     /**
      * Sets the player's rotation.
