From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 22 Mar 2024 16:21:27 +0800
Subject: [PATCH] Player in structure API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 7c56182acaf827f4b1a986a61cea8e9960604c98..c3c332d8de70018bbe80555c07b0046287862712 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3681,6 +3681,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     String getClientBrandName();
     // Paper end
 
+    boolean isInStructure(); // KioCG
+
     // Paper start - Teleport API
     /**
      * Sets the player's rotation.
