From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 22 Mar 2024 16:21:27 +0800
Subject: [PATCH] Player in structure API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8086acceacbceb2c5a7228fff005e41a86d37008..47e2ea0c2f6c760f131db13dea37b36e75406df4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3681,6 +3681,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     String getClientBrandName();
     // Paper end
 
+    boolean isInStructure(); // KioCG
+
     // Paper start - Teleport API
     /**
      * Sets the player's rotation.
