From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sat, 16 Mar 2024 11:21:14 -0700
Subject: [PATCH] The Long Dark


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c3c332d8de70018bbe80555c07b0046287862712..e086eec475174d8b804c91d617ed5c9149ec556f 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3857,6 +3857,20 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // KioCG start
+    int getColdValue();
+    void setColdValue(int value);
+    void addColdValue(int add);
+
+    int getThirstValue();
+    void setThirstValue(int value);
+    void addThirstValue(int add);
+
+    double getThirstRegain();
+    void setThirstRegain(double value);
+    void addThirstRegain(double add);
+    // KioCG end
+
     @NotNull
     @Override
     Spigot spigot();
