From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 28 May 2024 16:20:26 +0800
Subject: [PATCH] Entity baby API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index 3cd539e265e1c2a10b231ac16bc4ae3415e60356..ed1eb876753cc625eaaaab61809494196c0eb9a8 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -227,4 +227,9 @@ public interface Mob extends LivingEntity, Lootable {
     public org.bukkit.boss.BossBar getBossbar();
     public void setBossbar(org.bukkit.boss.BossBar bossbar);
     // KioCG end
+
+    // KioCG start
+    public boolean isBaby();
+    public void setBaby(boolean flag);
+    // KioCG end
 }
