From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 28 May 2024 16:20:26 +0800
Subject: [PATCH] Entity baby API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index a1e54e9d14393a6c0ea57cca854071c5396d9717..84dc776b3ade7522a1262c78c04ac2aa9eb487ba 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1458,4 +1458,11 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     boolean canUseEquipmentSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
     // Paper end - Expose canUseSlot
+
+    // KioCG start
+    public boolean isBabySimulate();
+    public void setBabySimulate(boolean flag);
+    public boolean isHugeSimulate();
+    public void setHugeSimulate(boolean flag);
+    // KioCG end
 }
