From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 28 May 2024 16:20:26 +0800
Subject: [PATCH] Entity baby API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index db467c74b17d6019b097ea261f633cf2f5937c63..1283bcaec990a4d8fc0c534db38d1acc388a228c 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -235,4 +235,9 @@ public interface Mob extends LivingEntity, Lootable {
     public org.bukkit.boss.BossBar getBossbar();
     public void setBossbar(org.bukkit.boss.BossBar bossbar);
     // KioCG end
+
+    // KioCG start
+    public boolean isBaby();
+    public void setBaby(boolean flag);
+    // KioCG end
 }
