From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 9 Apr 2024 18:38:54 +0800
Subject: [PATCH] Entity bossbar API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index 256e5645bcfa76e7ede58ae365e69b4a4bed6204..db467c74b17d6019b097ea261f633cf2f5937c63 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -229,4 +229,10 @@ public interface Mob extends LivingEntity, Lootable {
      */
     public int getPossibleExperienceReward();
     // Paper end - mob xp reward API
+
+    // KioCG start
+    @Nullable
+    public org.bukkit.boss.BossBar getBossbar();
+    public void setBossbar(org.bukkit.boss.BossBar bossbar);
+    // KioCG end
 }
