From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 9 Apr 2024 18:38:54 +0800
Subject: [PATCH] Entity bossbar API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index 9a10262a952e330f42187d6e4edab5a225bfba62..3632a9f7f9a5a97893fd19de93a6350d3b7f56bd 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -229,4 +229,10 @@ public interface Mob extends LivingEntity, Lootable, io.papermc.paper.entity.Lea
      */
     public int getPossibleExperienceReward();
     // Paper end - mob xp reward API
+
+    // KioCG start
+    @Nullable
+    public org.bukkit.boss.BossBar getBossbar();
+    public void setBossbar(org.bukkit.boss.BossBar bossbar);
+    // KioCG end
 }
