From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 15 Mar 2024 17:35:18 +0100
Subject: [PATCH] Auto discard rider display entity


diff --git a/src/main/java/org/bukkit/entity/Display.java b/src/main/java/org/bukkit/entity/Display.java
index 2b6f0c627a848f37bc5b3f1732d0e0b82d9f89ad..b53b56b66d93562e95ef13c91232d7c9df98bf8b 100644
--- a/src/main/java/org/bukkit/entity/Display.java
+++ b/src/main/java/org/bukkit/entity/Display.java
@@ -204,6 +204,11 @@ public interface Display extends Entity {
      */
     public void setBrightness(@Nullable Brightness brightness);
 
+    // KioCG start
+    public int getRiderPriority();
+    public void setRiderPriority(int riderPriority);
+    // KioCG end
+
     /**
      * Describes the axes/points around which the entity can pivot.
      */
