From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 15 Mar 2024 17:35:18 +0100
Subject: [PATCH] Auto discard rider display entity


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 725ef320f929d5e3d141c1ed3246d73a7d741f31..7701fd0935299915d568378e259f056d4d342eb2 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1159,4 +1159,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @NotNull String getScoreboardEntryName();
     // Paper end - entity scoreboard name
+
+    // KioCG start
+    default @Nullable <T extends Display> T addRiderDisplay(Class<T> clazz, java.util.function.Consumer<T> function) {
+        return addRiderDisplay(clazz, 0, function);
+    }
+    @Nullable <T extends Display> T addRiderDisplay(Class<T> clazz, int priority, java.util.function.Consumer<T> function);
+
+    default boolean removeRiderDisplay() {
+        return removeRiderDisplay(null);
+    }
+    boolean removeRiderDisplay(@NotNull UUID uuid);
+    // KioCG end
 }
