From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 28 May 2024 16:20:26 +0800
Subject: [PATCH] Entity baby API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index fcdc5fce88720cc926a3953d80b5045113d1516c..a66ca24b23b8584d32210e8284e0b3e603c29dba 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1458,4 +1458,12 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     boolean canUseEquipmentSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
     // Paper end - Expose canUseSlot
+
+    // KioCG start
+    public void setBabySimulate();
+    public boolean isBabySimulate();
+    public void setHugeSimulate();
+    public boolean isHugeSimulate();
+    public void removeScaleAttributes();
+    // KioCG end
 }
