From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 22 Mar 2024 14:03:21 +0800
Subject: [PATCH] End crystal turret


diff --git a/src/main/java/org/bukkit/entity/EnderCrystal.java b/src/main/java/org/bukkit/entity/EnderCrystal.java
index febc7b3f4c9a402d74d2bcf27e89578fc6d7d6a3..c7e307128b9d4243a7c56c9dddb6d7497fc800fa 100644
--- a/src/main/java/org/bukkit/entity/EnderCrystal.java
+++ b/src/main/java/org/bukkit/entity/EnderCrystal.java
@@ -40,4 +40,15 @@ public interface EnderCrystal extends Entity {
      * @throws IllegalArgumentException for differing worlds
      */
     void setBeamTarget(@Nullable Location location);
+
+    // KioCG start
+    int getTurretCooldown();
+    void setTurretCooldown(@Nullable Integer cooldown);
+
+    double getTurretRange();
+    void setTurretRange(@Nullable Double range);
+
+    double getTurretDamage();
+    void setTurretDamage(@Nullable Double damage);
+    // KioCG end
 }
