--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -88,6 +_,7 @@
     private final LevelChunkTicks<Block> blockTicks;
     private final LevelChunkTicks<Fluid> fluidTicks;
     private LevelChunk.UnsavedListener unsavedListener = chunkPos -> {};
+    private final com.kiocg.world.ChunkHot chunkHot = new com.kiocg.world.ChunkHot(); public com.kiocg.world.ChunkHot getChunkHot() { return this.chunkHot; } // KioCG - chunk hot
     // CraftBukkit start
     public boolean mustNotSave;
     public boolean needsDecoration;
@@ -207,6 +_,7 @@
         this.setLightCorrect(chunk.isLightCorrect());
         this.markUnsaved();
         this.needsDecoration = true; // CraftBukkit
+        this.chunkDays = chunk.chunkDays; // KioCG - crops grow every day
         // CraftBukkit start
         this.persistentDataContainer = chunk.persistentDataContainer; // SPIGOT-6814: copy PDC to account for 1.17 to 1.18 chunk upgrading.
         // CraftBukkit end
@@ -876,7 +_,11 @@
 
     public <T extends BlockEntity> void updateBlockEntityTicker(T blockEntity) {
         BlockState blockState = blockEntity.getBlockState();
-        BlockEntityTicker<T> ticker = blockState.getTicker(this.level, (BlockEntityType<T>)blockEntity.getType());
+        // KioCG start - optimize block entity
+        boolean needTicking = blockEntity.preFirstTick || blockEntity.needTicking(this.level, blockEntity.getBlockPos(), blockState);
+        BlockEntityTicker<T> ticker = !needTicking ? null : blockState.getTicker(this.level, (BlockEntityType<T>)blockEntity.getType());
+        blockEntity.tickerTicking = false;
+        // KioCG end - optimize block entity
         if (ticker == null) {
             this.removeBlockEntityTicker(blockEntity.getBlockPos());
         } else {
@@ -887,12 +_,14 @@
                         TickingBlockEntity tickingBlockEntity = this.createTicker(blockEntity, ticker);
                         if (ticker1 != null) {
                             ticker1.rebind(tickingBlockEntity);
+                            blockEntity.tickerTicking = true; // KioCG - optimize block entity
                             return (LevelChunk.RebindableTickingBlockEntityWrapper)ticker1;
                         } else if (this.isInLevel()) {
                             LevelChunk.RebindableTickingBlockEntityWrapper rebindableTickingBlockEntityWrapper = new LevelChunk.RebindableTickingBlockEntityWrapper(
                                 tickingBlockEntity
                             );
                             this.level.addBlockEntityTicker(rebindableTickingBlockEntityWrapper);
+                            blockEntity.tickerTicking = true; // KioCG - optimize block entity
                             return rebindableTickingBlockEntityWrapper;
                         } else {
                             return null;
@@ -924,9 +_,13 @@
                     try {
                         ProfilerFiller profilerFiller = Profiler.get();
                         profilerFiller.push(this::getType);
+                        LevelChunk.this.chunkHot.startTicking(); try { // KioCG - chunk hot
                         BlockState blockState = LevelChunk.this.getBlockState(blockPos);
                         if (this.blockEntity.getType().isValid(blockState)) {
+                            if (!this.blockEntity.preFirstTick || !(this.blockEntity.preFirstTick = false) && this.blockEntity.updateBlockEntityTicker()) { // KioCG - optimize block entity
                             this.ticker.tick(LevelChunk.this.level, this.blockEntity.getBlockPos(), blockState, this.blockEntity);
+                            com.kiocg.command.TickerViewCommand.onBlockEntityTicking(LevelChunk.this.level, this.blockEntity.getBlockPos(), blockState); // KioCG - ticker view
+                            } // KioCG - optimize block entity
                             this.loggedInvalidBlockState = false;
                         // Paper start - Remove the Block Entity if it's invalid
                         } else {
@@ -938,6 +_,7 @@
                             // Paper end - Remove the Block Entity if it's invalid
                         }
 
+                        } finally { LevelChunk.this.chunkHot.endTickingAndCount(); } // KioCG - chunk hot
                         profilerFiller.pop();
                     } catch (Throwable var5) {
                         // Paper start - Prevent block entity and entity crashes
