--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -20,6 +_,7 @@
     public static final int BIOME_CONTAINER_BITS = 2;
     short nonEmptyBlockCount; // Paper - package private
     private short tickingBlockCount;
+    public short dailyTickingBlockCount; // KioCG - crops grow every day
     private short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
     private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
@@ -36,6 +_,7 @@
     private static final short CLIENT_FORCED_SPECIAL_COLLIDING_BLOCKS = (short)9999;
     private short specialCollidingBlocks;
     private final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = new ca.spottedleaf.moonrise.common.list.ShortList();
+    public final ca.spottedleaf.moonrise.common.list.ShortList dailyTickingBlocks = new ca.spottedleaf.moonrise.common.list.ShortList(); // KioCG - crops grow every day
 
     @Override
     public final boolean moonrise$hasSpecialCollidingBlocks() {
@@ -51,6 +_,7 @@
     private LevelChunkSection(LevelChunkSection section) {
         this.nonEmptyBlockCount = section.nonEmptyBlockCount;
         this.tickingBlockCount = section.tickingBlockCount;
+        this.dailyTickingBlockCount = section.dailyTickingBlockCount; // KioCG - crops grow every day
         this.tickingFluidCount = section.tickingFluidCount;
         this.states = section.states.copy();
         this.biomes = section.biomes.copy();
@@ -120,17 +_,38 @@
         final boolean oldTicking = oldState.isRandomlyTicking();
         final boolean newTicking = newState.isRandomlyTicking();
         if (oldTicking != newTicking) {
-            final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
-            final short position = (short)(x | (z << 4) | (y << (4+4)));
+            // KioCG start - crops grow every day
+            tickingBlocks(x, y, z, oldTicking, this.tickingBlocks);
+            // final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
+            // final short position = (short)(x | (z << 4) | (y << (4+4)));
+            //
+            // if (oldTicking) {
+            //     tickingBlocks.remove(position);
+            // } else {
+            //     tickingBlocks.add(position);
+            // }
+        }
 
-            if (oldTicking) {
-                tickingBlocks.remove(position);
-            } else {
-                tickingBlocks.add(position);
-            }
+        final boolean oldDailyTicking = oldState.isDailyTicking();
+        final boolean newDailyTicking = newState.isDailyTicking();
+        if (oldDailyTicking != newDailyTicking) {
+            tickingBlocks(x, y, z, oldDailyTicking, this.dailyTickingBlocks);
+            // KioCG end - crops grow every day
         }
     }
     // Paper end - block counting
+
+    // KioCG start - crops grow every day: 提取方法
+    private void tickingBlocks(int x, int y, int z, boolean oldTicking, ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks) {
+        final short position = (short)(x | (z << 4) | (y << (4+4)));
+
+        if (oldTicking) {
+            tickingBlocks.remove(position);
+        } else {
+            tickingBlocks.add(position);
+        }
+    }
+    // KioCG end - crops grow every day: 提取方法
 
     public BlockState setBlockState(int x, int y, int z, BlockState state, boolean useLocks) {
         BlockState blockState;
@@ -147,6 +_,11 @@
             if (blockState.isRandomlyTicking()) {
                 this.tickingBlockCount--;
             }
+            // KioCG start - crops grow every day
+            if (blockState.isDailyTicking()) {
+                this.dailyTickingBlockCount--;
+            }
+            // KioCG end - crops grow every day
         }
 
         if (!!fluidState.isRandomlyTicking()) { // Paper - block counting
@@ -158,6 +_,11 @@
             if (state.isRandomlyTicking()) {
                 this.tickingBlockCount++;
             }
+            // KioCG start - crops grow every day
+            if (state.isDailyTicking()) {
+                this.dailyTickingBlockCount++;
+            }
+            // KioCG end - crops grow every day
         }
 
         if (!!fluidState1.isRandomlyTicking()) { // Paper - block counting
@@ -181,6 +_,12 @@
         return this.tickingBlockCount > 0;
     }
 
+    // KioCG start - crops grow every day
+    public boolean isDailyTickingBlocks() {
+        return this.dailyTickingBlockCount > 0;
+    }
+    // KioCG end - crops grow every day
+
     public boolean isRandomlyTickingFluids() {
         return this.tickingFluidCount > 0;
     }
@@ -190,9 +_,11 @@
         // reset, then recalculate
         this.nonEmptyBlockCount = (short)0;
         this.tickingBlockCount = (short)0;
+        this.dailyTickingBlockCount = (short)0; // KioCG - crops grow every day
         this.tickingFluidCount = (short)0;
         this.specialCollidingBlocks = (short)0;
         this.tickingBlocks.clear();
+        this.dailyTickingBlocks.clear(); // KioCG - crops grow every day
 
         if (this.maybeHas((final BlockState state) -> !state.isAir())) {
             final PalettedContainer.Data<BlockState> data = this.states.data;
@@ -226,17 +_,25 @@
                 this.nonEmptyBlockCount += (short)paletteCount;
                 if (state.isRandomlyTicking()) {
                     this.tickingBlockCount += (short)paletteCount;
-                    final short[] raw = coordinates.elements();
-                    final int rawLen = raw.length;
-
-                    final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
-
-                    tickingBlocks.setMinCapacity(Math.min((rawLen + tickingBlocks.size()) * 3 / 2, 16*16*16));
-
-                    java.util.Objects.checkFromToIndex(0, paletteCount, raw.length);
-                    for (int i = 0; i < paletteCount; ++i) {
-                        tickingBlocks.add(raw[i]);
-                    }
+                    // KioCG start - crops grow every day
+                    tickingBlocks(paletteCount, coordinates, this.tickingBlocks);
+                    // final short[] raw = coordinates.elements();
+                    // final int rawLen = raw.length;
+                    //
+                    // final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
+                    //
+                    // tickingBlocks.setMinCapacity(Math.min((rawLen + tickingBlocks.size()) * 3 / 2, 16*16*16));
+                    //
+                    // java.util.Objects.checkFromToIndex(0, paletteCount, raw.length);
+                    // for (int i = 0; i < paletteCount; ++i) {
+                    //     tickingBlocks.add(raw[i]);
+                    // }
+                }
+
+                if (state.isDailyTicking()) {
+                    this.dailyTickingBlockCount += (short)paletteCount;
+                    tickingBlocks(paletteCount, coordinates, this.dailyTickingBlocks);
+                    // KioCG end - crops grow every day
                 }
 
                 final FluidState fluid = state.getFluidState();
@@ -251,6 +_,20 @@
         }
         // Paper end - block counting
     }
+
+    // KioCG start - crops grow every day: 提取方法
+    private void tickingBlocks(int paletteCount, it.unimi.dsi.fastutil.shorts.ShortArrayList coordinates, ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks) {
+        final short[] raw = coordinates.elements();
+        final int rawLen = raw.length;
+
+        tickingBlocks.setMinCapacity(Math.min((rawLen + tickingBlocks.size()) * 3 / 2, 16*16*16));
+
+        java.util.Objects.checkFromToIndex(0, paletteCount, raw.length);
+        for (int i = 0; i < paletteCount; ++i) {
+            tickingBlocks.add(raw[i]);
+        }
+    }
+    // KioCG end - crops grow every day: 提取方法
 
     public PalettedContainer<BlockState> getStates() {
         return this.states;
