--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -617,7 +_,7 @@
             protected ItemStack execute(BlockSource blockSource, ItemStack item) {
                 ServerLevel serverLevel = blockSource.level();
                 BlockPos blockPos = blockSource.pos().relative(blockSource.state().getValue(DispenserBlock.FACING));
-                List<Armadillo> entitiesOfClass = serverLevel.getEntitiesOfClass(Armadillo.class, new AABB(blockPos), EntitySelector.NO_SPECTATORS);
+                List<net.minecraft.world.entity.animal.Animal> entitiesOfClass = serverLevel.getEntitiesOfClass(net.minecraft.world.entity.animal.Animal.class, new AABB(blockPos), EntitySelector.NO_SPECTATORS.and(entity -> entity instanceof Armadillo || entity instanceof net.minecraft.world.entity.animal.Turtle)); // KioCG - breeding animals
                 if (entitiesOfClass.isEmpty()) {
                     this.setSuccess(false);
                     return item;
@@ -644,8 +_,12 @@
                         }
                     }
                     // CraftBukkit end
-                    for (Armadillo armadillo : entitiesOfClass) {
-                        if (armadillo.brushOffScute()) {
+                    // KioCG start - breeding animals
+                    for (net.minecraft.world.entity.animal.Animal entity : entitiesOfClass) {
+                        boolean isArmadillo = entity instanceof Armadillo;
+                        boolean isTurtle = !isArmadillo && entity instanceof net.minecraft.world.entity.animal.Turtle;
+                        if (isArmadillo && ((Armadillo) entity).brushOffScute() || isTurtle && ((net.minecraft.world.entity.animal.Turtle) entity).brushOffScute()) {
+                    // KioCG end - breeding animals
                             item.hurtAndBreak(16, serverLevel, null, item1 -> {});
                             return item;
                         }
