From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 5 Aug 2024 12:14:08 +0800
Subject: [PATCH] More API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1b36e94617d4e777c419660936460d5cf8a4b3e8..404d277f34052a82e3c8d4d3ccff8578a3dba4fd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -705,6 +705,13 @@ public final class CraftServer implements Server {
         return (player != null) ? player.getBukkitEntity() : null;
     }
 
+    // KioCG start
+    @Override
+    public boolean isValidPlayerName(String name) {
+        return net.minecraft.util.StringUtil.isReasonablePlayerName(name);
+    }
+    // KioCG end
+
     @Override
     public Player getPlayer(UUID id) {
         Preconditions.checkArgument(id != null, "UUID id cannot be null");
