From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 3 Mar 2024 05:25:55 +0800
Subject: [PATCH] Death drop protection


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 278a2f999a6e4d564582a4b2e103c24f5036774f..d84f248cb0ac053fe31c6f78c741f9e9f74d6c22 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -552,7 +552,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public KiocgConfig kiocgConfig;
 
     public class KiocgConfig extends ConfigurationPart {
-
+        public boolean deathItemsOnlyPickedUpByOneself = false;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index ad1fd0c01526cce26e4898279f915a0c4ee21723..a1e9ff0d1fca859cd34f01e3b62e59a1faba3fa3 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1037,7 +1037,15 @@ public class ServerPlayer extends Player {
         if (!keepInventory) {
             for (ItemStack item : this.getInventory().getContents()) {
                 if (!item.isEmpty() && !EnchantmentHelper.hasVanishingCurse(item)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG start
+                    loot.add(new DefaultDrop(item, stack -> {
+                        ItemEntity droped = this.drop(stack, true, false, false);
+                        if (droped != null) {
+                            if (this.level().paperConfig().kiocgConfig.deathItemsOnlyPickedUpByOneself) droped.setTarget(this.getUUID());
+                            droped.age = -32767;
+                        }
+                    })); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG end
                 }
             }
         }
