From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 3 Mar 2024 05:25:55 +0800
Subject: [PATCH] Death drop protection


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 5251de7f1c475fcbb279e0c9161dbc1b54b92f99..71cfc0f7e502cb07aad74254ea396bbde219a84d 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -553,7 +553,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public KiocgConfig kiocgConfig;
 
     public class KiocgConfig extends ConfigurationPart {
-
+        public boolean deathItemsOnlyPickedUpByOneself = true;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 5a746714b9958abee8319854d58f1b9f1095bc98..d2508509df3f3d6d3eda6ef78d35c1f3ec585fa5 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1028,7 +1028,15 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         if (!keepInventory) {
             for (ItemStack item : this.getInventory().getContents()) {
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG start
+                    loot.add(new DefaultDrop(item, stack -> {
+                        ItemEntity droped = this.drop(stack, true, false, false);
+                        if (droped != null) {
+                            if (this.level().paperConfig().kiocgConfig.deathItemsOnlyPickedUpByOneself) droped.setTarget(this.getUUID());
+                            droped.age = -32767;
+                        }
+                    })); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    // KioCG end
                 }
             }
         }
