From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 5 Mar 2024 10:34:43 +0800
Subject: [PATCH] The Long Dark


diff --git a/src/main/java/com/kiocg/java/NamespacedKeys.java b/src/main/java/com/kiocg/java/NamespacedKeys.java
index e3a9806f9972007b732032fd137741a5d14a7e63..c09d96b28876b6c684813f699908b9be1c5d2fc4 100644
--- a/src/main/java/com/kiocg/java/NamespacedKeys.java
+++ b/src/main/java/com/kiocg/java/NamespacedKeys.java
@@ -5,4 +5,7 @@ import org.bukkit.NamespacedKey;
 public class NamespacedKeys {
     public static final String KIOCG = "kiocg";
 
+    public static NamespacedKey kiocgColdValue = new NamespacedKey(KIOCG, "cold_value");
+    public static NamespacedKey kiocgThirstValue = new NamespacedKey(KIOCG, "thirst_value");
+    public static NamespacedKey kiocgThirstRegain = new NamespacedKey(KIOCG, "thirst_regain");
 }
diff --git a/src/main/java/com/kiocg/java/player/ColdData.java b/src/main/java/com/kiocg/java/player/ColdData.java
new file mode 100644
index 0000000000000000000000000000000000000000..7f837e983256952fa779bd9a3fe06d1147327dc7
--- /dev/null
+++ b/src/main/java/com/kiocg/java/player/ColdData.java
@@ -0,0 +1,95 @@
+package com.kiocg.java.player;
+
+import com.kiocg.java.NamespacedKeys;
+import io.papermc.paper.configuration.WorldConfiguration;
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.protocol.game.ClientboundSetExperiencePacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LightLayer;
+import org.bukkit.persistence.PersistentDataType;
+
+public class ColdData {
+    public static final int MAX_VALUE = 60 * 20 * 10;
+
+    private Player entityhuman;
+    private int tempColdValue;
+    private int coldValueCache;
+
+    public ColdData(Player entityhuman) {
+        org.apache.commons.lang.Validate.notNull(entityhuman);
+        this.entityhuman = entityhuman;
+    }
+
+    public int getColdValue() {
+        return coldValueCache;
+    }
+
+    private int getColdValue0() {
+        return entityhuman.getBukkitEntity().getPersistentDataContainer().getOrDefault(NamespacedKeys.kiocgColdValue, PersistentDataType.INTEGER, MAX_VALUE);
+    }
+
+    public void setColdValue(int value) {
+        entityhuman.getBukkitEntity().getPersistentDataContainer().set(NamespacedKeys.kiocgColdValue, PersistentDataType.INTEGER, value);
+        tempColdValue = 0;
+    }
+
+    public void addColdValue(int add) {
+        tempColdValue += add;
+    }
+
+    public float getColdProgress() {
+        return (float) coldValueCache / MAX_VALUE;
+    }
+
+    public boolean isFrozen() {
+        return coldValueCache <= 0;
+    }
+
+    public void tick() {
+        if (!entityhuman.getAbilities().invulnerable) {
+            final Level world = entityhuman.level();
+            final BlockPos pos = entityhuman.blockPosition();
+            final WorldConfiguration.KiocgConfig.TheLongDark theLongDark = world.paperConfig().kiocgConfig.theLongDark;
+
+            final float temperature = (world.getBiome(pos).value().getTemperature(pos) + theLongDark.coldValue.ambientTemperatureOffset) * theLongDark.coldValue.ambientTemperatureOffsetMultiplier;
+            tempColdValue += theLongDark.coldValue.ambientBase + temperature;
+
+            if (theLongDark.coldValue.lightSimulationMultiplier != 0) {
+                int light = world.getRawBrightness(pos, world.getSkyDarken());
+                tempColdValue += light * theLongDark.coldValue.lightSimulationMultiplier;
+            }
+            if (theLongDark.coldValue.lightBlockMultiplier != 0) {
+                int blockLight = world.getBrightness(LightLayer.BLOCK, pos);
+                tempColdValue += blockLight * theLongDark.coldValue.lightBlockMultiplier;
+            }
+            if (theLongDark.coldValue.lightSkyMultiplier != 0) {
+                int skyLight = world.getBrightness(LightLayer.SKY, pos);
+                tempColdValue += skyLight * theLongDark.coldValue.lightSkyMultiplier;
+            }
+
+            if (theLongDark.coldValue.isInWaterOrBubble != 0 && (entityhuman.isInPowderSnow || entityhuman.isInWaterOrBubble())) {
+                tempColdValue += theLongDark.coldValue.isInWaterOrBubble;
+            } else if (theLongDark.coldValue.isInRain != 0 && entityhuman.isInRain()) {
+                tempColdValue += theLongDark.coldValue.isInRain;
+            }
+
+            if (theLongDark.coldValue.isInLava != 0 && entityhuman.isInLava()) {
+                tempColdValue += theLongDark.coldValue.isInLava;
+            } else if (theLongDark.coldValue.onFireTickDivision != 0) {
+                int remainingFireTicks = entityhuman.getRemainingFireTicks();
+                if (remainingFireTicks > 0) {
+                    tempColdValue += remainingFireTicks / theLongDark.coldValue.onFireTickDivision;
+                }
+            }
+
+            setColdValue(coldValueCache = Mth.clamp(getColdValue0() + tempColdValue, 0, MAX_VALUE));
+            ((ServerPlayer) entityhuman).connection.send(new ClientboundSetExperiencePacket(getColdProgress(), entityhuman.totalExperience, entityhuman.experienceLevel));
+        } else {
+            tempColdValue = 0;
+            coldValueCache = MAX_VALUE;
+        }
+    }
+}
diff --git a/src/main/java/com/kiocg/java/player/ThirstData.java b/src/main/java/com/kiocg/java/player/ThirstData.java
new file mode 100644
index 0000000000000000000000000000000000000000..f61acdb002b66350925ef0f4d022bde55c3038d0
--- /dev/null
+++ b/src/main/java/com/kiocg/java/player/ThirstData.java
@@ -0,0 +1,87 @@
+package com.kiocg.java.player;
+
+import com.kiocg.java.NamespacedKeys;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
+import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import org.bukkit.persistence.PersistentDataType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ThirstData {
+    public static final int MAX_VALUE = 10;
+
+    private Player entityhuman;
+
+    public ThirstData(Player entityhuman) {
+        org.apache.commons.lang.Validate.notNull(entityhuman);
+        this.entityhuman = entityhuman;
+    }
+
+    public int getThirstValue() {
+        return (Integer) entityhuman.getEntityData().get(Entity.DATA_AIR_SUPPLY_ID);
+    }
+
+    private int getThirstValue0() {
+        return entityhuman.getBukkitEntity().getPersistentDataContainer().getOrDefault(NamespacedKeys.kiocgThirstValue, PersistentDataType.INTEGER, MAX_VALUE);
+    }
+
+    public void setThirstValue(int value) {
+        entityhuman.getBukkitEntity().getPersistentDataContainer().set(NamespacedKeys.kiocgThirstValue, PersistentDataType.INTEGER, value);
+    }
+
+    public int getThirstProgress() {
+        return 300 / 10 * (getThirstValue() - 1) + 15;
+    }
+
+    public boolean isThirsty() {
+        return getThirstValue() <= 0;
+    }
+
+    public void tick() {
+        if (!entityhuman.getAbilities().invulnerable) {
+            double regain = getThirstRegain0();
+            int quotient = (int) regain / 20;
+            if (quotient > 0) {
+                setThirstRegain(regain - 20 * quotient);
+            }
+
+            int value = Mth.clamp(getThirstValue0() - quotient, 0, MAX_VALUE);
+            setThirstValue(value);
+            entityhuman.getEntityData().set(Entity.DATA_AIR_SUPPLY_ID, value);
+
+            List<SynchedEntityData.DataValue<?>> list = new ArrayList<>();
+            list.add(SynchedEntityData.DataValue.create(Entity.DATA_AIR_SUPPLY_ID, getThirstProgress()));
+            ((ServerPlayer) entityhuman).connection.send(new ClientboundSetEntityDataPacket(entityhuman.getId(), list));
+        }
+    }
+
+    public double getThirstRegain() {
+        return getThirstRegain0();
+    }
+
+    private double getThirstRegain0() {
+        return entityhuman.getBukkitEntity().getPersistentDataContainer().getOrDefault(NamespacedKeys.kiocgThirstRegain, PersistentDataType.DOUBLE, 0.0);
+    }
+
+    public void setThirstRegain(double value) {
+        entityhuman.getBukkitEntity().getPersistentDataContainer().set(NamespacedKeys.kiocgThirstRegain, PersistentDataType.DOUBLE, value);
+    }
+
+    public void addThirstRegain(double add) {
+        setThirstRegain(getThirstRegain0() + add);
+    }
+
+    public void readAdditionalSaveData(CompoundTag nbt) {
+        setThirstValue(getThirstValue0());
+    }
+
+    public void addAdditionalSaveData(CompoundTag nbt) {
+
+    }
+}
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 63992c84ec75bd77499cfc921ddc7c4491c7447e..0c4ce9a7a53b353e9415403dd06cd8cbf5eaf98d 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -551,6 +551,30 @@ public class WorldConfiguration extends ConfigurationPart {
 
     public class KiocgConfig extends ConfigurationPart {
         public boolean deathItemsOnlyPickedUpByOneself = false;
+
+        public TheLongDark theLongDark;
+        public class TheLongDark extends ConfigurationPart {
+            public double frozenTickDamageMultiplier = 0.01;
+
+            public ColdValue coldValue;
+            public class ColdValue extends ConfigurationPart {
+                public int ambientBase = -10;
+                public float ambientTemperatureOffset = -1.0F;
+                public int ambientTemperatureOffsetMultiplier = 10;
+
+                public int lightSimulationMultiplier = 1;
+                public int lightBlockMultiplier = 2;
+                public int lightSkyMultiplier = 0;
+
+                public int isInWaterOrBubble = -30;
+                public int isInRain = -15;
+                public int isInLava = 50;
+                public int onFireTickDivision = 5;
+            }
+
+            public int campfireExtinguishingOne = -1;
+            public int campfireExtinguishingTwo = -1;
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index e187d1c65446eb53869921110740c8e559308d5e..882feeb2e39882f269473be2d659f3fbbd3f28bd 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -825,7 +825,7 @@ public class ServerPlayer extends Player {
                 this.updateScoreForCriteria(ObjectiveCriteria.LEVEL, Mth.ceil((float) this.lastRecordedLevel));
             }
 
-            if (this.totalExperience != this.lastSentExp) {
+            if (false && this.totalExperience != this.lastSentExp) { // KioCG - cold system
                 this.lastSentExp = this.totalExperience;
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
             }
@@ -2747,6 +2747,8 @@ public class ServerPlayer extends Player {
         this.setRemainingFireTicks(0);
         this.fallDistance = 0;
         this.foodData = new FoodData(this);
+        this.coldData = new com.kiocg.java.player.ColdData(this); // KioCG
+        this.thirstData = new com.kiocg.java.player.ThirstData(this); // KioCG
         this.experienceLevel = this.newLevel;
         this.totalExperience = this.newTotalExp;
         this.experienceProgress = 0;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c5dfa1037dc6ecc18774d55b08dbcc4acb032f7c..a80b29f8d9585e41a1084eecf09f9ffe17a1d095 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -951,7 +951,7 @@ public abstract class PlayerList {
         entityplayer1.connection.teleport(CraftLocation.toBukkit(entityplayer1.position(), worldserver2.getWorld(), entityplayer1.getYRot(), entityplayer1.getXRot())); // CraftBukkit
         entityplayer1.connection.send(new ClientboundSetDefaultSpawnPositionPacket(worldserver1.getSharedSpawnPos(), worldserver1.getSharedSpawnAngle()));
         entityplayer1.connection.send(new ClientboundChangeDifficultyPacket(worlddata.getDifficulty(), worlddata.isDifficultyLocked()));
-        entityplayer1.connection.send(new ClientboundSetExperiencePacket(entityplayer1.experienceProgress, entityplayer1.totalExperience, entityplayer1.experienceLevel));
+        entityplayer1.connection.send(new ClientboundSetExperiencePacket(entityplayer1.coldData.getColdProgress(), entityplayer1.totalExperience, entityplayer1.experienceLevel)); // KioCG - cold system
         this.sendLevelInfo(entityplayer1, worldserver1);
         this.sendPlayerPermissionLevel(entityplayer1);
         if (!entityplayer.connection.isDisconnected()) {
@@ -995,6 +995,8 @@ public abstract class PlayerList {
         // Paper start
         if (isRespawn) {
             cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(entityplayer.getBukkitEntity(), location, isBedSpawn));
+            entityplayer.coldData.setColdValue(com.kiocg.java.player.ColdData.MAX_VALUE); // KioCG
+            entityplayer.thirstData.setThirstValue(com.kiocg.java.player.ThirstData.MAX_VALUE); // KioCG
         }
         // Paper end
 
diff --git a/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java b/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
index b994ae09621934df2cdd6a83a7d8ecb44649fb16..e082df98a69246dd015928e2404fa0faab3ae845 100644
--- a/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
+++ b/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
@@ -19,6 +19,8 @@ class SaturationMobEffect extends InstantenousMobEffect {
         if (!entity.level().isClientSide && entity instanceof Player) {
             Player entityhuman = (Player) entity;
 
+            if (true) { entityhuman.coldData.addColdValue(amplifier + 1); return; } // KioCG
+
             // CraftBukkit start
             int oldFoodLevel = entityhuman.getFoodData().foodLevel;
             org.bukkit.event.entity.FoodLevelChangeEvent event = CraftEventFactory.callFoodLevelChangeEvent(entityhuman, amplifier + 1 + oldFoodLevel);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 1be10c57e374ad4018c08d96cfb69397a2f541d3..ca0a71a8f2cfdc7a067058246110a7b8e7f4ccb6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -355,7 +355,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     private static final int FLAG_INVISIBLE = 5;
     protected static final int FLAG_GLOWING = 6;
     protected static final int FLAG_FALL_FLYING = 7;
-    private static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT);
+    public static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT); // KioCG - private -> public
     private static final EntityDataAccessor<Optional<Component>> DATA_CUSTOM_NAME = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.OPTIONAL_COMPONENT);
     private static final EntityDataAccessor<Boolean> DATA_CUSTOM_NAME_VISIBLE = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Boolean> DATA_SILENT = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
@@ -602,7 +602,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         // Spigot end
         this.entityData = new SynchedEntityData(this);
         this.entityData.define(Entity.DATA_SHARED_FLAGS_ID, (byte) 0);
-        this.entityData.define(Entity.DATA_AIR_SUPPLY_ID, this.getMaxAirSupply());
+        this.entityData.define(Entity.DATA_AIR_SUPPLY_ID, this instanceof Player ? com.kiocg.java.player.ThirstData.MAX_VALUE : this.getMaxAirSupply()); // KioCG
         this.entityData.define(Entity.DATA_CUSTOM_NAME_VISIBLE, false);
         this.entityData.define(Entity.DATA_CUSTOM_NAME, Optional.empty());
         this.entityData.define(Entity.DATA_SILENT, false);
@@ -3369,6 +3369,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     }
 
     public int getAirSupply() {
+        if (this instanceof Player player) return player.realAir; // KioCG
         return (Integer) this.entityData.get(Entity.DATA_AIR_SUPPLY_ID);
     }
 
@@ -3379,10 +3380,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         if (this.valid) {
             event.getEntity().getServer().getPluginManager().callEvent(event);
         }
-        if (event.isCancelled() && this.getAirSupply() != air) {
+        if (event.isCancelled() && this.getAirSupply() != air && !(this instanceof Player)) { // KioCG
             this.entityData.markDirty(Entity.DATA_AIR_SUPPLY_ID);
             return;
         }
+        if (this instanceof Player player) { player.realAir = event.getAmount(); } else // KioCG
         this.entityData.set(Entity.DATA_AIR_SUPPLY_ID, event.getAmount());
         // CraftBukkit end
     }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index bc908b75cb99536df658281ae7f8b4eeedbbedc9..c4c92a404000ca8ef71d45fbc0c801a335cda8cb 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1356,6 +1356,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         // Paper end
         float f1 = this.getHealth();
 
+        if (!(this instanceof net.minecraft.world.entity.player.Player player) || !player.thirstData.isThirsty()) // KioCG
         if (f1 > 0.0F) {
             EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), f, regainReason, isFastRegen); // Paper
             // Suppress during worldgen
@@ -1365,6 +1366,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             if (!event.isCancelled()) {
                 this.setHealth((float) (this.getHealth() + event.getAmount()));
+                if (this instanceof net.minecraft.world.entity.player.Player player) player.thirstData.addThirstRegain(event.getAmount()); // KioCG
             }
             // CraftBukkit end
         }
@@ -3456,17 +3458,23 @@ public abstract class LivingEntity extends Entity implements Attackable {
         if (!this.level().isClientSide && !this.isDeadOrDying() && !this.freezeLocked) { // Paper - Freeze Tick Lock API
             int i = this.getTicksFrozen();
 
-            if (this.isInPowderSnow && this.canFreeze()) {
-                this.setTicksFrozen(Math.min(this.getTicksRequiredToFreeze(), i + 1));
+            if (this.isInPowderSnow && this.canFreeze() || this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen()) { // KioCG
+                this.setTicksFrozen(i + 1); // KioCG - 不要限制最大值
             } else {
-                this.setTicksFrozen(Math.max(0, i - 2));
+                this.setTicksFrozen(Math.min(this.getTicksRequiredToFreeze(), Math.max(0, i - 2))); // KioCG - 还原最大值设定, 防止长时间冻结
             }
         }
 
         this.removeFrost();
         this.tryAddFrost();
         if (!this.level().isClientSide && this.tickCount % 40 == 0 && this.isFullyFrozen() && this.canFreeze()) {
-            this.hurt(this.damageSources().freeze(), 1.0F);
+            // KioCG start
+            float increase = 0.0F;
+            if (this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen()) {
+                increase = (this.getTicksFrozen() - this.getTicksRequiredToFreeze()) * (float) this.level().paperConfig().kiocgConfig.theLongDark.frozenTickDamageMultiplier;
+            }
+            this.hurt(this.damageSources().freeze(), 1.0F + increase);
+            // KioCG end
         }
 
         this.level().getProfiler().pop();
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index ccc1caafb0ada52c7b99b7358253826f5390843e..ea644ac618d8bcef41d3c4b690d247397d2b1e1b 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -152,6 +152,9 @@ public abstract class Player extends LivingEntity {
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
     protected FoodData foodData = new FoodData(this); // CraftBukkit - add "this" to constructor
+    public com.kiocg.java.player.ColdData coldData = new com.kiocg.java.player.ColdData(this); // KioCG
+    public com.kiocg.java.player.ThirstData thirstData = new com.kiocg.java.player.ThirstData(this); // KioCG
+    public int realAir; // KioCG
     protected int jumpTriggerTime;
     public float oBob;
     public float bob;
@@ -278,6 +281,8 @@ public abstract class Player extends LivingEntity {
         this.moveCloak();
         if (!this.level().isClientSide) {
             this.foodData.tick(this);
+            this.coldData.tick(); // KioCG
+            this.thirstData.tick(); // KioCG
             this.awardStat(Stats.PLAY_TIME);
             this.awardStat(Stats.TOTAL_WORLD_TIME);
             if (this.isAlive()) {
@@ -872,6 +877,7 @@ public abstract class Player extends LivingEntity {
 
         this.setScore(nbt.getInt("Score"));
         this.foodData.readAdditionalSaveData(nbt);
+        this.thirstData.readAdditionalSaveData(nbt); // KioCG
         this.abilities.loadSaveData(nbt);
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue((double) this.abilities.getWalkingSpeed());
         if (nbt.contains("EnderItems", 9)) {
@@ -909,6 +915,7 @@ public abstract class Player extends LivingEntity {
         nbt.putInt("XpSeed", this.enchantmentSeed);
         nbt.putInt("Score", this.getScore());
         this.foodData.addAdditionalSaveData(nbt);
+        this.thirstData.addAdditionalSaveData(nbt); // KioCG
         this.abilities.addSaveData(nbt);
         nbt.put("EnderItems", this.enderChestInventory.createTag());
         if (!this.getShoulderEntityLeft().isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
index bc099abfda6555cade0a83eddfe1be9c243fa83b..4dcfbf9b7267017229640d0f7f7370118a942e39 100644
--- a/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
@@ -35,6 +35,7 @@ public class LavaCauldronBlock extends AbstractCauldronBlock {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper
         if (this.isEntityInsideContent(state, pos, entity)) {
             entity.lavaHurt();
+            if (entity instanceof net.minecraft.world.entity.player.Player player) player.coldData.addColdValue(world.paperConfig().kiocgConfig.theLongDark.coldValue.isInLava); // KioCG
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
index 843926beb1edd897724b512ad915894fa4a2da6c..e6505b79bd3a3d4b50a97dd689ae37a30e2550c5 100644
--- a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -77,6 +77,7 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
             // CraftBukkit end
         }
 
+        if (entity instanceof net.minecraft.world.entity.player.Player player && this.isEntityInsideContent(state, pos, entity)) player.coldData.addColdValue(world.paperConfig().kiocgConfig.theLongDark.coldValue.isInWaterOrBubble); // KioCG
     }
 
     @Deprecated // Paper - use #handleEntityOnFireInsideWithEvent
diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index f929c90c470bd7fd0b40d915a3101e7e73e7f4f0..03d25edf000fab0671ef96d8ed885eb5a8c0ccdd 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -43,6 +43,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     public final int[] cookingTime;
     private final RecipeManager.CachedCheck<Container, CampfireCookingRecipe> quickCheck;
     public final boolean[] stopCooking; // Paper
+    private int litTicks = 0; // KioCG
 
     public CampfireBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.CAMPFIRE, pos, state);
@@ -56,6 +57,28 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     public static void cookTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
         boolean flag = false;
 
+        // KioCG start
+        ++campfire.litTicks;
+        int extinguishingTick = state.is(net.minecraft.world.level.block.Blocks.CAMPFIRE) ? world.paperConfig().kiocgConfig.theLongDark.campfireExtinguishingOne : world.paperConfig().kiocgConfig.theLongDark.campfireExtinguishingTwo;
+        if (extinguishingTick >= 0 && campfire.litTicks > extinguishingTick) {
+            if (state.is(net.minecraft.world.level.block.Blocks.CAMPFIRE)) {
+                world.removeBlockEntity(pos); // 移除BlockEntity来保留物品
+                world.setBlock(pos, net.minecraft.world.level.block.Blocks.SOUL_CAMPFIRE.withPropertiesOf(state), 3);
+
+                CampfireBlockEntity soulCampfire = (CampfireBlockEntity) world.getBlockEntity(pos);
+                for (int i = 0; i < campfire.items.size(); i++) {
+                    soulCampfire.items.set(i, campfire.items.get(i));
+                }
+                System.arraycopy(campfire.cookingProgress, 0, soulCampfire.cookingProgress, 0, campfire.cookingProgress.length);
+                System.arraycopy(campfire.cookingTime, 0, soulCampfire.cookingTime, 0, campfire.cookingTime.length);
+                System.arraycopy(campfire.stopCooking, 0, soulCampfire.stopCooking, 0, campfire.stopCooking.length);
+            } else {
+                world.setBlock(pos, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), 3);
+            }
+            return;
+        }
+        // KioCG end
+
         for (int i = 0; i < campfire.items.size(); ++i) {
             ItemStack itemstack = (ItemStack) campfire.items.get(i);
 
@@ -228,6 +251,8 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, abyte.length));
         }
         // Paper end
+
+        if (nbt.contains("KioCG.LitTicks", org.bukkit.craftbukkit.util.CraftMagicNumbers.NBT.TAG_INT)) litTicks = nbt.getInt("KioCG.LitTicks"); // KioCG
     }
 
     @Override
@@ -243,6 +268,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
         }
         nbt.putByteArray("Paper.StopCooking", cookingState);
         // Paper end
+        nbt.putInt("KioCG.LitTicks", litTicks); // KioCG
     }
 
     @Override
@@ -298,6 +324,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
         if (this.level != null) {
             this.markUpdated();
             this.cooldownImmediately(this.getLevel(), this.getBlockPos(), this.getBlockState(), (CampfireBlockEntity) this); // KioCG - 优化未燃烧的运算
+            this.litTicks = 0; // KioCG
         }
 
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index da63b4050be25dcb91d04df8c2fcc643cbb0751d..b714186810b3a8a5f9cd3ef0c5e2ad7de2c98a47 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2120,7 +2120,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         playerList.sendAllPlayerInfo(handle);
 
         // Resend their XP and effects because the respawn packet resets it
-        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.experienceProgress, handle.totalExperience, handle.experienceLevel));
+        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.coldData.getColdProgress(), handle.totalExperience, handle.experienceLevel)); // KioCG - cold system
         for (net.minecraft.world.effect.MobEffectInstance mobEffect : handle.getActiveEffects()) {
             connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket(handle.getId(), mobEffect));
         }
