From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 01:33:14 +0800
Subject: [PATCH] Adjust anvil


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 5864532b75b433488136ac6c46ca4c1bed477bba..6f40cd27e623e4068a637e9d8541ca59aabe1c73 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -407,6 +407,8 @@ public class GlobalConfiguration extends ConfigurationPart {
         public Enchant enchant;
         public class Enchant extends ConfigurationPart {
             public boolean allowUnsafeCommand = true;
+            public int fixedCumulativeCost = 4;
+            public boolean anvilRenameInSurvival = false;
         }
     }
     // KioCG end
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index daa0701dc538b0c97332217bccafc40b0c8824cb..0b03a236978639ce156027e984434df0703148ee 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -274,6 +274,7 @@ public class AnvilMenu extends ItemCombinerMenu {
                 }
             }
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild) // KioCG
             if (this.itemName != null && !Util.isBlank(this.itemName)) {
                 if (!this.itemName.equals(itemstack.getHoverName().getString())) {
                     b1 = 1;
@@ -321,6 +322,7 @@ public class AnvilMenu extends ItemCombinerMenu {
     }
 
     public static int calculateIncreasedRepairCost(int cost) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost; // KioCG
         return cost * 2 + 1;
     }
 
@@ -329,6 +331,7 @@ public class AnvilMenu extends ItemCombinerMenu {
 
         if (s1 != null && !s1.equals(this.itemName)) {
             this.itemName = s1;
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild) // KioCG
             if (this.getSlot(2).hasItem()) {
                 ItemStack itemstack = this.getSlot(2).getItem();
 
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 9ea50a580d6d2162141c37a4416af9b6dc0a93a1..1d50dd46ca8e538c8f4f2ecb3631524f78f04f3a 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -1257,10 +1257,12 @@ public final class ItemStack {
     }
 
     public int getBaseRepairCost() {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost; // KioCG
         return this.hasTag() && this.tag.contains("RepairCost", 3) ? this.tag.getInt("RepairCost") : 0;
     }
 
     public void setRepairCost(int repairCost) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) repairCost = 0; // KioCG
         if (repairCost > 0) {
             this.getOrCreateTag().putInt("RepairCost", repairCost);
         } else {
