From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 01:33:14 +0800
Subject: [PATCH] Adjust anvil


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 1eae15f8dff440aebdc6805e58460c95d852589b..d1d52d34ab7f079ba6689ea371e0b8b90676efce 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -410,6 +410,8 @@ public class GlobalConfiguration extends ConfigurationPart {
         public Enchant enchant;
         public class Enchant extends ConfigurationPart {
             public boolean allowUnsafeCommand = true;
+            public int fixedCumulativeCost = 4;
+            public boolean anvilRenameInSurvival = false;
         }
     }
     // KioCG end
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index f280f22fdfc27a61d659d64e0e38318b62a460af..0266de806479b96685cfeccdbd5d06116db23e71 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -274,6 +274,7 @@ public class AnvilMenu extends ItemCombinerMenu {
                 }
             }
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild) // KioCG
             if (this.itemName != null && !Util.isBlank(this.itemName)) {
                 if (!this.itemName.equals(itemstack.getHoverName().getString())) {
                     b1 = 1;
@@ -321,6 +322,7 @@ public class AnvilMenu extends ItemCombinerMenu {
     }
 
     public static int calculateIncreasedRepairCost(int cost) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost; // KioCG
         return cost * 2 + 1;
     }
 
@@ -329,6 +331,7 @@ public class AnvilMenu extends ItemCombinerMenu {
 
         if (s1 != null && !s1.equals(this.itemName)) {
             this.itemName = s1;
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild) // KioCG
             if (this.getSlot(2).hasItem()) {
                 ItemStack itemstack = this.getSlot(2).getItem();
 
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 0d0e4e657e8392dd1029ea9cc005df0997db35b2..b3dd67794e4a9acd1f9a5a089d405d242bc89e4f 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -1258,10 +1258,12 @@ public final class ItemStack {
     }
 
     public int getBaseRepairCost() {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost; // KioCG
         return this.hasTag() && this.tag.contains("RepairCost", 3) ? this.tag.getInt("RepairCost") : 0;
     }
 
     public void setRepairCost(int repairCost) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) repairCost = 0; // KioCG
         if (repairCost > 0) {
             this.getOrCreateTag().putInt("RepairCost", repairCost);
         } else {
