From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 01:33:14 +0800
Subject: [PATCH] Adjust anvil


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 46a759890f058c57a4569cdb7794e370b4f286f0..7b006a8d6f33d1f1059df35b767b88900cba7e09 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -414,6 +414,8 @@ public class GlobalConfiguration extends ConfigurationPart {
         public Enchant enchant;
         public class Enchant extends ConfigurationPart {
             public boolean allowUnsafeCommand = true;
+            public int fixedCumulativeCost = 4;
+            public boolean anvilRenameInSurvival = false;
         }
     }
     // KioCG end
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index ab70c286fb11024ec50b1683442ea7192424c4fe..81511fca9b07bfea4e50793701f5c2f7bb3f5af2 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -148,6 +148,7 @@ public class AnvilMenu extends ItemCombinerMenu {
             ItemStack itemstack2 = this.inputSlots.getItem(1);
             ItemEnchantments.Mutable itemenchantments_a = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(itemstack1));
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) j += io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost * 2L; else // KioCG
             j += (long) (Integer) itemstack.getOrDefault(DataComponents.REPAIR_COST, 0) + (long) (Integer) itemstack2.getOrDefault(DataComponents.REPAIR_COST, 0);
             this.repairItemCountCost = 0;
             int k;
@@ -261,6 +262,7 @@ public class AnvilMenu extends ItemCombinerMenu {
                 }
             }
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild || itemstack.is(Items.NAME_TAG)) { // KioCG
             if (this.itemName != null && !StringUtil.isBlank(this.itemName)) {
                 if (!this.itemName.equals(itemstack.getHoverName().getString())) {
                     b0 = 1;
@@ -272,6 +274,7 @@ public class AnvilMenu extends ItemCombinerMenu {
                 i += b0;
                 itemstack1.remove(DataComponents.CUSTOM_NAME);
             }
+            } // KioCG
 
             int k2 = (int) Mth.clamp(j + (long) i, 0L, 2147483647L);
 
@@ -323,11 +326,13 @@ public class AnvilMenu extends ItemCombinerMenu {
             if (this.getSlot(2).hasItem()) {
                 ItemStack itemstack = this.getSlot(2).getItem();
 
+                if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.anvilRenameInSurvival || this.player.getAbilities().instabuild || itemstack.is(Items.NAME_TAG)) { // KioCG
                 if (StringUtil.isBlank(s1)) {
                     itemstack.remove(DataComponents.CUSTOM_NAME);
                 } else {
                     itemstack.set(DataComponents.CUSTOM_NAME, Component.literal(s1));
                 }
+                } // KioCG
             }
 
             this.createResult();
