From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 01:33:14 +0800
Subject: [PATCH] Fixed cumulative cost


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index a145720768cab69c58e66f26a923e92188ff4d96..af9a57860a8c41e35b1aabac7a5fc1317b16a63f 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -423,6 +423,7 @@ public class GlobalConfiguration extends ConfigurationPart {
         public Enchant enchant;
         public class Enchant extends ConfigurationPart {
             public boolean allowUnsafeCommand = true;
+            public int fixedCumulativeCost = 4;
         }
     }
     // KioCG end
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index 03ce932a285a4183c4023a84618a4b8460815dbf..af76c89a3d84a806268107f3cf19dd3b2dba6f4f 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -148,6 +148,7 @@ public class AnvilMenu extends ItemCombinerMenu {
             ItemStack itemstack2 = this.inputSlots.getItem(1);
             ItemEnchantments.Mutable itemenchantments_a = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(itemstack1));
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost >= 0) j += io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.enchant.fixedCumulativeCost * 2L; else // KioCG
             j += (long) (Integer) itemstack.getOrDefault(DataComponents.REPAIR_COST, 0) + (long) (Integer) itemstack2.getOrDefault(DataComponents.REPAIR_COST, 0);
             this.repairItemCountCost = 0;
             int k;
