From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 29 Mar 2024 23:28:23 +0800
Subject: [PATCH] Allow potions to stack


diff --git a/src/main/java/net/minecraft/world/item/Items.java b/src/main/java/net/minecraft/world/item/Items.java
index d00b59efb754594cf532f8598f4b6d3b29693232..41972cbefaf52c18ccc0d313f3c2c10f940d6326 100644
--- a/src/main/java/net/minecraft/world/item/Items.java
+++ b/src/main/java/net/minecraft/world/item/Items.java
@@ -1364,7 +1364,7 @@ public class Items {
     public static final Item GOLD_NUGGET = registerItem("gold_nugget", new Item(new Item.Properties()));
     public static final Item NETHER_WART = registerItem("nether_wart", new ItemNameBlockItem(Blocks.NETHER_WART, new Item.Properties()));
     public static final Item POTION = registerItem(
-        "potion", new PotionItem(new Item.Properties().stacksTo(1).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY))
+        "potion", new PotionItem(new Item.Properties().stacksTo(16).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY)) // KioCG
     );
     public static final Item GLASS_BOTTLE = registerItem("glass_bottle", new BottleItem(new Item.Properties()));
     public static final Item SPIDER_EYE = registerItem("spider_eye", new Item(new Item.Properties().food(Foods.SPIDER_EYE)));
@@ -1768,14 +1768,14 @@ public class Items {
     public static final Item BEETROOT_SOUP = registerItem("beetroot_soup", new BowlFoodItem(new Item.Properties().stacksTo(1).food(Foods.BEETROOT_SOUP)));
     public static final Item DRAGON_BREATH = registerItem("dragon_breath", new Item(new Item.Properties().craftRemainder(GLASS_BOTTLE).rarity(Rarity.UNCOMMON)));
     public static final Item SPLASH_POTION = registerItem(
-        "splash_potion", new SplashPotionItem(new Item.Properties().stacksTo(1).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY))
+        "splash_potion", new SplashPotionItem(new Item.Properties().stacksTo(16).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY)) // KioCG
     );
     public static final Item SPECTRAL_ARROW = registerItem("spectral_arrow", new SpectralArrowItem(new Item.Properties()));
     public static final Item TIPPED_ARROW = registerItem(
         "tipped_arrow", new TippedArrowItem(new Item.Properties().component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY))
     );
     public static final Item LINGERING_POTION = registerItem(
-        "lingering_potion", new LingeringPotionItem(new Item.Properties().stacksTo(1).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY))
+        "lingering_potion", new LingeringPotionItem(new Item.Properties().stacksTo(16).component(DataComponents.POTION_CONTENTS, PotionContents.EMPTY)) // KioCG
     );
     public static final Item SHIELD = registerItem(
         "shield", new ShieldItem(new Item.Properties().durability(336).component(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY))
diff --git a/src/main/java/net/minecraft/world/item/PotionItem.java b/src/main/java/net/minecraft/world/item/PotionItem.java
index db038cb423c3b951890c81d833aa77ebcea3304f..e920c5fc2c83952d4e090b3816ad86b70e12af8c 100644
--- a/src/main/java/net/minecraft/world/item/PotionItem.java
+++ b/src/main/java/net/minecraft/world/item/PotionItem.java
@@ -91,7 +91,9 @@ public class PotionItem extends Item {
             }
 
             if (entityhuman != null) {
+                boolean add = // KioCG
                 entityhuman.getInventory().add(new ItemStack(Items.GLASS_BOTTLE));
+                if (!add) entityhuman.drop(new ItemStack(Items.GLASS_BOTTLE), false); // KioCG
             }
         }
 
