From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 9 Apr 2024 12:35:05 +0800
Subject: [PATCH] Entity bossbar API


diff --git a/src/main/java/net/minecraft/server/level/ServerBossEvent.java b/src/main/java/net/minecraft/server/level/ServerBossEvent.java
index 4f91107f9ae42f96c060c310596db9aa869a8dbc..e0cd8255c962927461915ef2a8cc14903f180777 100644
--- a/src/main/java/net/minecraft/server/level/ServerBossEvent.java
+++ b/src/main/java/net/minecraft/server/level/ServerBossEvent.java
@@ -131,4 +131,32 @@ public class ServerBossEvent extends BossEvent {
     public Collection<ServerPlayer> getPlayers() {
         return this.unmodifiablePlayers;
     }
+
+    // KioCG start
+    public net.minecraft.nbt.CompoundTag save(net.minecraft.core.HolderLookup.Provider provider) {
+        net.minecraft.nbt.CompoundTag nbttagcompound = new net.minecraft.nbt.CompoundTag();
+
+        nbttagcompound.putString("Name", Component.Serializer.toJson(this.name, provider));
+        nbttagcompound.putBoolean("Visible", this.isVisible());
+        nbttagcompound.putString("Color", this.getColor().getName());
+        nbttagcompound.putString("Overlay", this.getOverlay().getName());
+        nbttagcompound.putBoolean("DarkenScreen", this.shouldDarkenScreen());
+        nbttagcompound.putBoolean("PlayBossMusic", this.shouldPlayBossMusic());
+        nbttagcompound.putBoolean("CreateWorldFog", this.shouldCreateWorldFog());
+        return nbttagcompound;
+    }
+
+    public static ServerBossEvent load(net.minecraft.nbt.CompoundTag nbt, net.minecraft.core.HolderLookup.Provider provider) {
+        Component displayName1 = Component.Serializer.fromJson(nbt.getString("Name"), provider);
+        BossBarColor color1 = BossBarColor.byName(nbt.getString("Color"));
+        BossBarOverlay overlay1 = BossBarOverlay.byName(nbt.getString("Overlay"));
+        ServerBossEvent bossbar = new ServerBossEvent(displayName1, color1, overlay1);
+
+        bossbar.setVisible(nbt.getBoolean("Visible"));
+        bossbar.setDarkenScreen(nbt.getBoolean("DarkenScreen"));
+        bossbar.setPlayBossMusic(nbt.getBoolean("PlayBossMusic"));
+        bossbar.setCreateWorldFog(nbt.getBoolean("CreateWorldFog"));
+        return bossbar;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 8b612b772ca87c852d0b108c2afd6785c261c9b9..3a595fcdb2c1b2948a3be1b1faf76bef6dcc3103 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -150,6 +150,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
     private Either<UUID, BlockPos> delayedLeashInfo;
     private BlockPos restrictCenter;
     private float restrictRadius;
+    @Nullable private net.minecraft.server.level.ServerBossEvent bossbar; // KioCG
 
     public boolean aware = true; // CraftBukkit
 
@@ -457,6 +458,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
             }
         }
 
+        updateBossbar(); // KioCG
     }
 
     protected void updateControlFlags() {
@@ -584,6 +586,8 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
         }
 
         nbt.putBoolean("Bukkit.Aware", this.aware); // CraftBukkit
+
+        if (this.bossbar != null) nbt.put("KioCG.Bossbar", this.bossbar.save(this.registryAccess())); // KioCG
     }
 
     @Override
@@ -668,6 +672,8 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
             this.aware = nbt.getBoolean("Bukkit.Aware");
         }
         // CraftBukkit end
+
+        if (nbt.contains("KioCG.Bossbar")) this.bossbar = net.minecraft.server.level.ServerBossEvent.load(nbt.getCompound("KioCG.Bossbar"), this.registryAccess()); // KioCG
     }
 
     @Override
@@ -1902,4 +1908,52 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
 
         return itemmonsteregg == null ? null : new ItemStack(itemmonsteregg);
     }
+
+    // KioCG start
+    @Nullable
+    public net.minecraft.server.level.ServerBossEvent getBossBar() {
+        return this.bossbar;
+    }
+
+    public void setBossBar(net.minecraft.server.level.ServerBossEvent bossbar) {
+        removeBossBar();
+        this.bossbar = bossbar;
+    }
+
+    public void removeBossBar() {
+        if (this.bossbar != null) {
+            this.bossbar.removeAllPlayers();
+            this.bossbar = null;
+        }
+    }
+
+    private void updateBossbar() {
+        if (this.bossbar != null) {
+            this.bossbar.setProgress(this.getHealth() / this.getMaxHealth());
+
+            java.util.Set<ServerPlayer> set = com.google.common.collect.Sets.newHashSet();
+            Iterator iterator = this.tracker.seenBy.iterator();
+
+            while (iterator.hasNext()) {
+                ServerPlayer entityplayer = (ServerPlayer) iterator.next();
+
+                if (this.position().distanceToSqr(entityplayer.position()) <= 16 * 16) {
+                    this.bossbar.addPlayer(entityplayer);
+                    set.add(entityplayer);
+                }
+            }
+
+            java.util.Set<ServerPlayer> set1 = com.google.common.collect.Sets.newHashSet(this.bossbar.getPlayers());
+
+            set1.removeAll(set);
+            Iterator iterator1 = set1.iterator();
+
+            while (iterator1.hasNext()) {
+                ServerPlayer entityplayer1 = (ServerPlayer) iterator1.next();
+
+                this.bossbar.removePlayer(entityplayer1);
+            }
+        }
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index e155501fb3a88edf3ddac3aa0aae1b6a5a84962e..a063dfab31f13979b4c6ae3395874736d4e498ce 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -174,4 +174,16 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         return getHandle().getExperienceReward();
     }
     // Paper end
+
+    // KioCG start
+    @Override
+    public org.bukkit.boss.BossBar getBossbar() {
+        return new org.bukkit.craftbukkit.boss.CraftBossBar(getHandle().getBossBar());
+    }
+
+    @Override
+    public void setBossbar(org.bukkit.boss.BossBar bossbar) {
+        getHandle().setBossBar(((org.bukkit.craftbukkit.boss.CraftBossBar) bossbar).getHandle());
+    }
+    // KioCG end
 }
