From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 15 Mar 2024 10:48:31 +0800
Subject: [PATCH] Auto discard rider display entity


diff --git a/src/main/java/net/minecraft/world/entity/Display.java b/src/main/java/net/minecraft/world/entity/Display.java
index 2ecd3ed5b36475125c6f63af2612011fc8fba8a5..827f3ca4da682d7f5af09368e654875f12dc760e 100644
--- a/src/main/java/net/minecraft/world/entity/Display.java
+++ b/src/main/java/net/minecraft/world/entity/Display.java
@@ -82,6 +82,7 @@ public abstract class Display extends Entity {
     private Display.RenderState renderState;
     @Nullable
     private Display.PosRotInterpolationTarget posRotInterpolationTarget;
+    private int riderPriority = 0; public int getRiderPriority() { return this.riderPriority; } public void setRiderPriority(int riderPriority) { this.riderPriority = riderPriority; } // KioCG
 
     public Display(EntityType<?> type, Level world) {
         super(type, world);
@@ -167,6 +168,17 @@ public abstract class Display extends Entity {
 
     }
 
+    // KioCG start
+    @Override
+    public void stopRiding(boolean suppressCancellation) {
+        Entity entity = this.getVehicle();
+        super.stopRiding(suppressCancellation);
+        if (entity != null && this.getVehicle() == null) {
+            this.discard();
+        }
+    }
+    // KioCG end
+
     protected abstract void updateRenderSubState(boolean shouldLerp, float lerpProgress);
 
     @Override
@@ -249,6 +261,7 @@ public abstract class Display extends Entity {
             this.setBrightnessOverride((Brightness)null);
         }
 
+        if (nbt.contains("KioCG.RiderPriority")) this.riderPriority = nbt.getInt("KioCG.RiderPriority"); // KioCG
     }
 
     public void setTransformation(Transformation transformation) {
@@ -281,6 +294,7 @@ public abstract class Display extends Entity {
             });
         }
 
+        nbt.putInt("KioCG.RiderPriority", this.riderPriority); // KioCG
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 5c6424a3800344c7fc5f3a0cbdf36a4b57e0a12c..e1ea2bbe0a35cc261ae5ba7b4e12c9927c67ea3f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4877,6 +4877,42 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         this.setRot(f, f1);
     }
 
+    // KioCG start
+    @Nullable
+    public <T extends Display> T addRiderDisplay(EntityType<T> type, java.util.function.Consumer<T> function) {
+        return addRiderDisplay(type, 0, function);
+    }
+
+    @Nullable
+    public <T extends Display> T addRiderDisplay(EntityType<T> type, int priority, java.util.function.Consumer<T> function) {
+        final Entity firstPassenger = this.getFirstPassenger();
+        if (firstPassenger == null || firstPassenger instanceof Display display && display.getRiderPriority() < priority) {
+            this.ejectPassengers();
+
+            T display = type.create(this.level());
+            if (display != null) {
+                display.moveTo(this.position());
+                display.persist = false;
+                display.setRiderPriority(priority);
+                display.startRiding(this);
+                function.accept(display);
+                this.level().addFreshEntity(display);
+            }
+            return display;
+        }
+        return null;
+    }
+
+    public boolean removeRiderDisplay(@Nullable UUID uuid) {
+        final Entity firstPassenger = this.getFirstPassenger();
+        if (firstPassenger instanceof Display display && (uuid == null || display.getUUID().equals(uuid))) {
+            firstPassenger.discard();
+            return true;
+        }
+        return false;
+    }
+    // KioCG end
+
     public static enum RemovalReason {
 
         KILLED(true, false), DISCARDED(true, false), UNLOADED_TO_CHUNK(false, true), UNLOADED_WITH_PLAYER(false, false), CHANGED_DIMENSION(false, false);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
index 48eeb1d9ba0ad6f895bfe507a6fbe4b9c9530e47..9f188b9e47948c111a89f3923c0d5386811215e1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
@@ -167,4 +167,6 @@ public class CraftDisplay extends CraftEntity implements Display {
             this.getHandle().setBrightnessOverride(null);
         }
     }
+
+    public int getRiderPriority() { return this.getHandle().getRiderPriority(); } public void setRiderPriority(int riderPriority) { this.getHandle().setRiderPriority(riderPriority); } // KioCG
 }
