From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 13 Mar 2024 01:10:57 +0800
Subject: [PATCH] Monster strengthens


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 386a3af68b0d0288a5fabf8728a4ac18077a38a8..72477c0ee5f2577518a6937aaf7ba90e8f6fbd38 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -582,6 +582,14 @@ public class WorldConfiguration extends ConfigurationPart {
             public int campfireExtinguishingOne = -1;
             public int campfireExtinguishingTwo = -1;
         }
+
+        public Monster monster;
+        public class Monster extends ConfigurationPart {
+            public boolean spawnInLightAtNight = true;
+            public int darkEnoughToSprint = 7;
+            public boolean babyZombieJockey = true;
+            public boolean zombiesSkinnedIntoSkeletons = true;
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Monster.java b/src/main/java/net/minecraft/world/entity/monster/Monster.java
index 759839e912c54598b257ad738481364940f88a18..15ebe8467ba5af964b94e2561a90ae807328589f 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Monster.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Monster.java
@@ -43,6 +43,19 @@ public abstract class Monster extends PathfinderMob implements Enemy {
         this.updateSwingTime();
         this.updateNoActionTime();
         super.aiStep();
+
+        // KioCG start
+        final int darkEnoughToSprint = this.level().paperConfig().kiocgConfig.monster.darkEnoughToSprint;
+        if (darkEnoughToSprint >= 0 && this.level().getMaxLocalRawBrightness(this.blockPosition()) <= darkEnoughToSprint) {
+            if (!this.isSprinting()) {
+                this.setSprinting(true);
+            }
+        } else {
+            if (this.isSprinting()) {
+                this.setSprinting(false);
+            }
+        }
+        // KioCG end
     }
 
     protected void updateNoActionTime() {
@@ -88,6 +101,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
     }
 
     public static boolean isDarkEnoughToSpawn(ServerLevelAccessor world, BlockPos pos, RandomSource random) {
+        if (world instanceof net.minecraft.server.level.ServerLevel level && level.paperConfig().kiocgConfig.monster.spawnInLightAtNight && world.getSkyDarken() >= 5) return true; // KioCG
         if (world.getBrightness(LightLayer.SKY, pos) > random.nextInt(32)) {
             return false;
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index 7618364e5373fe17cfe45a5a4ee9ab25e591581c..8ef8e64972fe7361dbf37484c24fc2f9037ad166 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -207,7 +207,7 @@ public class Spider extends Monster {
 
         @Override
         public boolean canUse() {
-            return super.canUse() && !this.mob.isVehicle();
+            return super.canUse() && !(this.mob.getControllingPassenger() instanceof Monster); // KioCG
         }
 
         @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index b8dd08a48ed27341b9f6a19341f10aa7ee842a9e..dd128ed095ecfe6612d284a0d32c1c64fb2284c3 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -116,6 +116,16 @@ public class Zombie extends Monster {
     }
 
     protected void addBehaviourGoals() {
+        // KioCG start
+        if (this.level().paperConfig().kiocgConfig.monster.babyZombieJockey) {
+            this.goalSelector.addGoal(3, new net.minecraft.world.entity.ai.goal.LeapAtTargetGoal(this, 0.4F) {
+                @Override
+                public boolean canUse() {
+                    return Zombie.this.isBaby() && super.canUse();
+                }
+            });
+        }
+        // KioCG end
         this.goalSelector.addGoal(2, new ZombieAttackGoal(this, 1.0D, false));
         // this.goalSelector.addGoal(6, new MoveThroughVillageGoal(this, 1.0D, true, 4, this::canBreakDoors)); // KioCG - 优化实体行为
         this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0D));
@@ -304,6 +314,48 @@ public class Zombie extends Monster {
 
     }
 
+    // KioCG start
+    @Override
+    public void die(DamageSource damageSource) {
+        if (this.level().paperConfig().kiocgConfig.monster.zombiesSkinnedIntoSkeletons) {
+            if (!this.isRemoved() && !this.dead) {
+                Entity entity = damageSource.getEntity();
+                Level world = this.level();
+
+                if (world instanceof ServerLevel) {
+                    org.bukkit.event.entity.EntityDeathEvent deathEvent = this.dropAllDeathLoot(damageSource);
+                    if (deathEvent == null || !deathEvent.isCancelled()) {
+                        if (this.isSleeping()) {
+                            this.stopSleeping();
+                        }
+
+                        this.getCombatTracker().recheckStatus();
+                        if (entity != null) {
+                            entity.killedEntity((ServerLevel) this.level(), this);
+                        }
+                    } else {
+                        this.setHealth((float) deathEvent.getReviveHealth());
+                        return;
+                    }
+
+                    Skeleton skeleton = this.convertTo(EntityType.SKELETON, true, EntityTransformEvent.TransformReason.PEELING, CreatureSpawnEvent.SpawnReason.PEELING);
+                    if (skeleton != null) {
+                        skeleton.setTarget(this.getTarget(), null, true);
+                        skeleton.getNavigation().moveTo(this.getNavigation().getPath(), 1.2D);
+                        skeleton.setAggressive(this.isAggressive());
+                    }
+
+                    if (!this.isSilent()) {
+                        this.level().levelEvent((Player) null, net.minecraft.world.level.block.LevelEvent.SOUND_SKELETON_TO_STRAY, this.blockPosition(), 0);
+                    }
+                }
+            }
+        } else {
+            super.die(damageSource);
+        }
+    }
+    // KioCG end
+
     public boolean isSunSensitive() {
         return this.shouldBurnInDay; // Paper - Add more Zombie API
     }
@@ -603,6 +655,7 @@ public class Zombie extends Monster {
 
     @Override
     protected void dropCustomDeathLoot(DamageSource source, int lootingMultiplier, boolean allowDrops) {
+        if (!this.level().paperConfig().kiocgConfig.monster.zombiesSkinnedIntoSkeletons) // KioCG - 僵尸剥皮: 不要掉落护甲
         super.dropCustomDeathLoot(source, lootingMultiplier, allowDrops);
         Entity entity = source.getEntity();
 
