From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 5 Mar 2024 10:34:43 +0800
Subject: [PATCH] The Long Dark


diff --git a/src/main/java/com/kiocg/player/ColdData.java b/src/main/java/com/kiocg/player/ColdData.java
new file mode 100644
index 0000000000000000000000000000000000000000..35de173aa5cb8eb9859f4f9eebee681602f349b3
--- /dev/null
+++ b/src/main/java/com/kiocg/player/ColdData.java
@@ -0,0 +1,107 @@
+package com.kiocg.player;
+
+import io.papermc.paper.configuration.WorldConfiguration;
+import net.minecraft.core.BlockPos;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.protocol.game.ClientboundSetExperiencePacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LightLayer;
+
+public class ColdData {
+    public static final int MAX_VALUE = 60 * 20 * 10;
+
+    private Player entityhuman;
+    private int coldValue = MAX_VALUE;
+
+    public ColdData(Player entityhuman) {
+        org.apache.commons.lang.Validate.notNull(entityhuman);
+        this.entityhuman = entityhuman;
+    }
+
+    public int getColdValue() {
+        return coldValue;
+    }
+
+    public void setColdValue(int value) {
+        coldValue = value;
+    }
+
+    public void addColdValue(int add) {
+        coldValue = Mth.clamp(coldValue + add, 0, MAX_VALUE);
+    }
+
+    public float getColdProgress() {
+        return (float) coldValue / MAX_VALUE;
+    }
+
+    public boolean isFrozen() {
+        return coldValue <= 0 && !entityhuman.getAbilities().invulnerable;
+    }
+
+    public boolean isInWater;
+    public boolean isInLava;
+
+    public void tick() {
+        if (!entityhuman.getAbilities().invulnerable) {
+            final Level world = entityhuman.level();
+            final BlockPos pos = entityhuman.blockPosition();
+            final WorldConfiguration.KiocgConfig.TheLongDark theLongDark = world.paperConfig().kiocgConfig.theLongDark;
+
+            double ambientTemperature = theLongDark.coldValue.ambientTemperatureBase;
+            if (theLongDark.coldValue.ambientSkyDarkenMultiplier != 0) {
+                ambientTemperature += world.getSkyDarken() * theLongDark.coldValue.ambientSkyDarkenMultiplier * world.getBrightness(LightLayer.SKY, pos);
+            }
+            if (theLongDark.coldValue.biomeTemperatureMultiplier != 0) {
+                ambientTemperature += (world.getBiome(pos).value().getTemperature(pos) + theLongDark.coldValue.biomeTemperatureOffset) * theLongDark.coldValue.biomeTemperatureMultiplier;
+            }
+
+            double playerTemperature = theLongDark.coldValue.playerTemperatureBase + entityhuman.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE); // 属性用于玩家温度
+            if (theLongDark.coldValue.playerLightBlockMultiplier != 0.0) {
+                int blockLight = world.getBrightness(LightLayer.BLOCK, pos);
+                playerTemperature += Math.min(blockLight, 12) * theLongDark.coldValue.playerLightBlockMultiplier;
+            }
+
+            if (theLongDark.coldValue.isInWaterOrBubble != 0 && (this.isInWater || entityhuman.isInPowderSnow || entityhuman.isInWaterOrBubble())) {
+                playerTemperature += theLongDark.coldValue.isInWaterOrBubble;
+            } else if (theLongDark.coldValue.isInRain != 0 && entityhuman.isInRain()) {
+                playerTemperature += theLongDark.coldValue.isInRain;
+            }
+
+            if (theLongDark.coldValue.isInLava != 0 && (this.isInLava || entityhuman.isInLava())) {
+                playerTemperature += theLongDark.coldValue.isInLava;
+            } else if (theLongDark.coldValue.isOnFire != 0) {
+                int remainingFireTicks = entityhuman.getRemainingFireTicks();
+                if (remainingFireTicks > 0) {
+                    playerTemperature += theLongDark.coldValue.isOnFire;
+                }
+            }
+
+            double finalTemperature = (playerTemperature + ambientTemperature) * theLongDark.coldValue.finalTemperatureMultiplier;
+            addColdValue((int) finalTemperature);
+            ((ServerPlayer) entityhuman).connection.send(new ClientboundSetExperiencePacket(getColdProgress(), entityhuman.totalExperience, entityhuman.experienceLevel));
+        }
+
+        isInWater = false;
+        isInLava = false;
+    }
+
+    public void reset() {
+        coldValue = MAX_VALUE;
+        isInWater = false;
+        isInLava = false;
+    }
+
+    public void readAdditionalSaveData(CompoundTag nbt) {
+        if (nbt.contains("KioCG.ColdValue")) {
+            coldValue = nbt.getInt("KioCG.ColdValue");
+        }
+    }
+
+    public void addAdditionalSaveData(CompoundTag nbt) {
+        nbt.putInt("KioCG.ColdValue", coldValue);
+    }
+}
diff --git a/src/main/java/com/kiocg/player/ThirstData.java b/src/main/java/com/kiocg/player/ThirstData.java
new file mode 100644
index 0000000000000000000000000000000000000000..c684e63291e85bdd30bd4ba72bbc006f67be69c5
--- /dev/null
+++ b/src/main/java/com/kiocg/player/ThirstData.java
@@ -0,0 +1,97 @@
+package com.kiocg.player;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.util.Mth;
+import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.effect.MobEffects;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+
+public class ThirstData {
+    public static final int MAX_VALUE = 10;
+
+    private Player entityhuman;
+    private int thirstValue = MAX_VALUE;
+    private double thirstRegain;
+
+    public ThirstData(Player entityhuman) {
+        org.apache.commons.lang.Validate.notNull(entityhuman);
+        this.entityhuman = entityhuman;
+    }
+
+    public int getThirstValue() {
+        return thirstValue;
+    }
+
+    public void setThirstValue(int value) {
+        thirstValue = value;
+    }
+
+    public void addThirstValue(int add) {
+        thirstValue = Mth.clamp(thirstValue + add, 0, MAX_VALUE);
+    }
+
+    public double getThirstRegain() {
+        return thirstRegain;
+    }
+
+    public void setThirstRegain(double value) {
+        thirstRegain = value;
+    }
+
+    public void addThirstRegain(double add) {
+        thirstRegain += add;
+    }
+
+    private int getThirstProgress() {
+        return 300 / 10 * (thirstValue - 1) + 8;
+    }
+
+    public boolean isThirsty() {
+        return thirstValue <= 0 && !entityhuman.getAbilities().invulnerable;
+    }
+
+    public void tick() {
+        if (!entityhuman.getAbilities().invulnerable) {
+            if (entityhuman.level().getBiome(entityhuman.blockPosition()).value().climateSettings.temperature() >= 2.0) {
+                addThirstRegain(entityhuman.level().paperConfig().kiocgConfig.theLongDark.thirstValue.highTemperatureBiomeThirst);
+            }
+
+            int quotient = (int) thirstRegain / 20;
+            if (quotient > 0) {
+                addThirstRegain(-20 * quotient);
+            }
+
+            final int before = thirstValue;
+            addThirstValue(-quotient);
+            if (entityhuman instanceof net.minecraft.server.level.ServerPlayer serverplayer && before > 3 && thirstValue <= 3) {
+                serverplayer.connection.send(new net.minecraft.network.protocol.game.ClientboundSoundEntityPacket(net.minecraft.core.registries.BuiltInRegistries.SOUND_EVENT.wrapAsHolder(net.minecraft.sounds.SoundEvents.HUSK_AMBIENT), net.minecraft.sounds.SoundSource.MASTER, serverplayer, 1.0F, 1.75F, 2L));
+            }
+
+            if (thirstValue <= 3 && entityhuman.tickCount % 10 == 0) {
+                if (!entityhuman.hasEffect(MobEffects.WEAKNESS) || entityhuman.getEffect(MobEffects.WEAKNESS).getAmplifier() == 0 && entityhuman.getEffect(MobEffects.WEAKNESS).endsWithin(20)) {
+                    entityhuman.addEffect(new MobEffectInstance(MobEffects.WEAKNESS, 20, 0));
+                }
+            }
+
+            this.entityhuman.getEntityData().set(Entity.DATA_AIR_SUPPLY_ID, getThirstProgress(), true);
+        }
+    }
+
+    public void reset() {
+        thirstValue = MAX_VALUE;
+        thirstRegain = 0.0;
+    }
+
+    public void readAdditionalSaveData(CompoundTag nbt) {
+        if (nbt.contains("KioCG.ThirstValue")) {
+            thirstValue = nbt.getInt("KioCG.ThirstValue");
+            thirstRegain = nbt.getDouble("KioCG.ThirstRegain");
+        }
+    }
+
+    public void addAdditionalSaveData(CompoundTag nbt) {
+        nbt.putInt("KioCG.ThirstValue", thirstValue);
+        nbt.putDouble("KioCG.ThirstRegain", thirstRegain);
+    }
+}
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 9a90cb9932a8223c4593e28a6394edbd7dbee849..95280590b995174f9ae6136bcfc67aad65006656 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -582,6 +582,35 @@ public class WorldConfiguration extends ConfigurationPart {
     public class KiocgConfig extends ConfigurationPart {
         public boolean deathItemsOnlyPickedUpByOneself = true;
         public boolean removeAutomationGameplay = true;
+
+        public TheLongDark theLongDark;
+        public class TheLongDark extends ConfigurationPart {
+            public double frozenTickDamageMultiplier = 0.01;
+
+            public ColdValue coldValue;
+            public class ColdValue extends ConfigurationPart {
+                public double ambientTemperatureBase = 0.0;
+                public double ambientSkyDarkenMultiplier = -0.15;
+                public double biomeTemperatureOffset = -0.65;
+                public double biomeTemperatureMultiplier = 30.0;
+
+                public double playerTemperatureBase = 0.0;
+                public double playerLightBlockMultiplier = 5.0;
+                public double isInWaterOrBubble = -35.0;
+                public double isInRain = -15.0;
+                public double isInLava = 300.0;
+                public double isOnFire = 10.0;
+
+                public double finalTemperatureMultiplier = 1.0;
+            }
+
+            public ThirstValue thirstValue;
+            public class ThirstValue extends ConfigurationPart {
+                public double thirstRegainMultiplierHealth = 1.0;
+                public double thirstRegainMultiplierFood = 2.0;
+                public double highTemperatureBiomeThirst = 0.02;
+            }
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index efbbf9f9b078b0faaf20c980142a1e5404b934f6..dc6552fb554c831e6ae76a6a86bbae9a225f7f40 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -891,7 +891,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
                 this.updateScoreForCriteria(ObjectiveCriteria.LEVEL, Mth.ceil((float) this.lastRecordedLevel));
             }
 
-            if (this.totalExperience != this.lastSentExp) {
+            if (false && this.totalExperience != this.lastSentExp) { // KioCG - cold system
                 this.lastSentExp = this.totalExperience;
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
             }
@@ -2937,6 +2937,8 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.setRemainingFireTicks(0);
         this.fallDistance = 0;
         this.foodData = new FoodData(this);
+        this.coldData = new com.kiocg.player.ColdData(this); // KioCG
+        this.thirstData = new com.kiocg.player.ThirstData(this); // KioCG
         this.experienceLevel = this.newLevel;
         this.totalExperience = this.newTotalExp;
         this.experienceProgress = 0;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 90d7ad8f5d13f55ddf4a4c7f53bbe78499f088ca..065c8937c6f9fc2530ea5f9e76bd334beb3a5cee 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -887,7 +887,7 @@ public abstract class PlayerList {
         entityplayer1.connection.teleport(CraftLocation.toBukkit(entityplayer1.position(), worldserver1.getWorld(), entityplayer1.getYRot(), entityplayer1.getXRot())); // CraftBukkit
         entityplayer1.connection.send(new ClientboundSetDefaultSpawnPositionPacket(worldserver.getSharedSpawnPos(), worldserver.getSharedSpawnAngle()));
         entityplayer1.connection.send(new ClientboundChangeDifficultyPacket(worlddata.getDifficulty(), worlddata.isDifficultyLocked()));
-        entityplayer1.connection.send(new ClientboundSetExperiencePacket(entityplayer1.experienceProgress, entityplayer1.totalExperience, entityplayer1.experienceLevel));
+        entityplayer1.connection.send(new ClientboundSetExperiencePacket(entityplayer1.coldData.getColdProgress(), entityplayer1.totalExperience, entityplayer1.experienceLevel)); // KioCG - cold system
         this.sendActivePlayerEffects(entityplayer1);
         this.sendLevelInfo(entityplayer1, worldserver);
         this.sendPlayerPermissionLevel(entityplayer1);
@@ -936,6 +936,8 @@ public abstract class PlayerList {
         // Paper start - Add PlayerPostRespawnEvent
         if (isRespawn) {
             cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(entityplayer.getBukkitEntity(), location, isBedSpawn));
+            entityplayer.coldData.reset(); // KioCG
+            entityplayer.thirstData.reset(); // KioCG
         }
         // Paper end - Add PlayerPostRespawnEvent
 
diff --git a/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java b/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
index 7b415dca88f50dc472fe4be96e5ef0996f117913..742177e7d9246c259065216ca0483ddd20a880eb 100644
--- a/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
+++ b/src/main/java/net/minecraft/world/effect/SaturationMobEffect.java
@@ -16,6 +16,11 @@ class SaturationMobEffect extends InstantenousMobEffect {
     @Override
     public boolean applyEffectTick(LivingEntity entity, int amplifier) {
         if (!entity.level().isClientSide && entity instanceof Player entityhuman) {
+            // KioCG start
+            entityhuman.coldData.addColdValue(5 * (amplifier + 1));
+            if (true) return true;
+            // KioCG end
+
             // CraftBukkit start
             int oldFoodLevel = entityhuman.getFoodData().foodLevel;
             org.bukkit.event.entity.FoodLevelChangeEvent event = CraftEventFactory.callFoodLevelChangeEvent(entityhuman, amplifier + 1 + oldFoodLevel);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e0827d8bb3fa17d4f590a5342ff41a514f623e68..b4e4b2f2d77b1cd8bfabec909d5882564e61fb7f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -362,7 +362,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public static final int FLAG_INVISIBLE = 5;
     protected static final int FLAG_GLOWING = 6;
     protected static final int FLAG_FALL_FLYING = 7;
-    private static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT);
+    public static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT); // KioCG - private -> public
     private static final EntityDataAccessor<Optional<Component>> DATA_CUSTOM_NAME = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.OPTIONAL_COMPONENT);
     private static final EntityDataAccessor<Boolean> DATA_CUSTOM_NAME_VISIBLE = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Boolean> DATA_SILENT = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
@@ -3461,6 +3461,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public int getAirSupply() {
+        if (this instanceof ServerPlayer player) return player.getBukkitEntity().realAir; // KioCG
         return (Integer) this.entityData.get(Entity.DATA_AIR_SUPPLY_ID);
     }
 
@@ -3471,10 +3472,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (this.valid) {
             event.getEntity().getServer().getPluginManager().callEvent(event);
         }
-        if (event.isCancelled() && this.getAirSupply() != air) {
+        if (event.isCancelled() && this.getAirSupply() != air && !(this instanceof ServerPlayer)) { // KioCG
             this.entityData.markDirty(Entity.DATA_AIR_SUPPLY_ID);
             return;
         }
+        if (this instanceof ServerPlayer player) { player.getBukkitEntity().realAir = event.getAmount(); } else // KioCG
         this.entityData.set(Entity.DATA_AIR_SUPPLY_ID, event.getAmount());
         // CraftBukkit end
     }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 9c9e2377f44b1f1bbd8bb0e8bc963a80f4f9ee68..9613724b4652227ef7e0806fa6c2ac73e475320d 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1357,6 +1357,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         // Paper end
         float f1 = this.getHealth();
 
+        if (!(this instanceof net.minecraft.world.entity.player.Player player) || !player.thirstData.isThirsty()) { // KioCG
         if (f1 > 0.0F) {
             EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), f, regainReason, isFastRegen); // Paper
             // Suppress during worldgen
@@ -1366,10 +1367,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             if (!event.isCancelled()) {
                 this.setHealth((float) (this.getHealth() + event.getAmount()));
+                if (this instanceof net.minecraft.world.entity.player.Player player) player.thirstData.addThirstRegain(Math.max(this.getHealth() - f1, 0) * this.level().paperConfig().kiocgConfig.theLongDark.thirstValue.thirstRegainMultiplierHealth); // KioCG
             }
             // CraftBukkit end
         }
-
+        } // KioCG
     }
 
     public float getHealth() {
@@ -3319,6 +3321,8 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
                 if (!itemstack3.isEmpty()) {
                     itemstack3.forEachModifier(enumitemslot1, (holder, attributemodifier) -> {
+                        if (this instanceof net.minecraft.world.entity.monster.Zombie && holder == Attributes.SPAWN_REINFORCEMENTS_CHANCE) return; // KioCG - 属性用于玩家温度
+
                         AttributeInstance attributemodifiable = this.attributes.getInstance(holder);
 
                         if (attributemodifiable != null) {
@@ -3530,17 +3534,23 @@ public abstract class LivingEntity extends Entity implements Attackable {
         if (!this.level().isClientSide && !this.isDeadOrDying() && !this.freezeLocked) { // Paper - Freeze Tick Lock API
             int i = this.getTicksFrozen();
 
-            if (this.isInPowderSnow && this.canFreeze()) {
-                this.setTicksFrozen(Math.min(this.getTicksRequiredToFreeze(), i + 1));
+            if (this.isInPowderSnow && this.canFreeze() || this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen()) { // KioCG
+                this.setTicksFrozen(i + 1); // KioCG - 不要限制最大值
             } else {
-                this.setTicksFrozen(Math.max(0, i - 2));
+                this.setTicksFrozen(Math.min(this.getTicksRequiredToFreeze(), Math.max(0, i - 2))); // KioCG - 还原最大值设定, 防止长时间冻结
             }
         }
 
         this.removeFrost();
         this.tryAddFrost();
         if (!this.level().isClientSide && this.tickCount % 40 == 0 && this.isFullyFrozen() && this.canFreeze()) {
-            this.hurt(this.damageSources().freeze(), 1.0F);
+            // KioCG start
+            float increase = 0.0F;
+            if (this instanceof net.minecraft.world.entity.player.Player player && player.coldData.isFrozen()) {
+                increase = (this.getTicksFrozen() - this.getTicksRequiredToFreeze()) * (float) this.level().paperConfig().kiocgConfig.theLongDark.frozenTickDamageMultiplier;
+            }
+            this.hurt(this.damageSources().freeze(), 1.0F + increase);
+            // KioCG end
         }
 
         this.level().getProfiler().pop();
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index 5fd346c520b73b30fabff6094f478d5989d2fca7..db561d2399f4b1ac8f5c442ec63cf2fefdc5f4b6 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -34,7 +34,7 @@ public class Attributes {
     public static final Holder<Attribute> SAFE_FALL_DISTANCE = Attributes.register("generic.safe_fall_distance", (new RangedAttribute("attribute.name.generic.safe_fall_distance", 3.0D, -1024.0D, 1024.0D)).setSyncable(true));
     public static final Holder<Attribute> SCALE = Attributes.register("generic.scale", (new RangedAttribute("attribute.name.generic.scale", 1.0D, 0.0625D, 16.0D)).setSyncable(true).setSentiment(Attribute.Sentiment.NEUTRAL));
     public static final Holder<Attribute> SNEAKING_SPEED = Attributes.register("player.sneaking_speed", (new RangedAttribute("attribute.name.player.sneaking_speed", 0.3D, 0.0D, 1.0D)).setSyncable(true));
-    public static final Holder<Attribute> SPAWN_REINFORCEMENTS_CHANCE = Attributes.register("zombie.spawn_reinforcements", new RangedAttribute("attribute.name.zombie.spawn_reinforcements", 0.0D, 0.0D, 1.0D));
+    public static final Holder<Attribute> SPAWN_REINFORCEMENTS_CHANCE = Attributes.register("zombie.spawn_reinforcements", new RangedAttribute("attribute.name.zombie.spawn_reinforcements", 0.0D, -1024.0D, 1024.0D)); // KioCG - 属性用于玩家温度
     public static final Holder<Attribute> STEP_HEIGHT = Attributes.register("generic.step_height", (new RangedAttribute("attribute.name.generic.step_height", 0.6D, 0.0D, 10.0D)).setSyncable(true));
     public static final Holder<Attribute> SUBMERGED_MINING_SPEED = Attributes.register("player.submerged_mining_speed", (new RangedAttribute("attribute.name.player.submerged_mining_speed", 0.2D, 0.0D, 20.0D)).setSyncable(true));
     public static final Holder<Attribute> SWEEPING_DAMAGE_RATIO = Attributes.register("player.sweeping_damage_ratio", (new RangedAttribute("attribute.name.player.sweeping_damage_ratio", 0.0D, 0.0D, 1.0D)).setSyncable(true));
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 09bcbc0ae36e4e69fee87a7e0c49acf496117a39..97404a624465a369737eff635718f0f65d8e7866 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -162,6 +162,8 @@ public abstract class Player extends LivingEntity {
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
     protected FoodData foodData = new FoodData(this); // CraftBukkit - add "this" to constructor
+    public com.kiocg.player.ColdData coldData = new com.kiocg.player.ColdData(this); // KioCG
+    public com.kiocg.player.ThirstData thirstData = new com.kiocg.player.ThirstData(this); // KioCG
     protected int jumpTriggerTime;
     public float oBob;
     public float bob;
@@ -236,7 +238,7 @@ public abstract class Player extends LivingEntity {
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK).add(Attributes.BLOCK_INTERACTION_RANGE, 4.5D).add(Attributes.ENTITY_INTERACTION_RANGE, 3.0D).add(Attributes.BLOCK_BREAK_SPEED).add(Attributes.SUBMERGED_MINING_SPEED).add(Attributes.SNEAKING_SPEED).add(Attributes.MINING_EFFICIENCY).add(Attributes.SWEEPING_DAMAGE_RATIO);
+        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK).add(Attributes.BLOCK_INTERACTION_RANGE, 4.5D).add(Attributes.ENTITY_INTERACTION_RANGE, 3.0D).add(Attributes.BLOCK_BREAK_SPEED).add(Attributes.SUBMERGED_MINING_SPEED).add(Attributes.SNEAKING_SPEED).add(Attributes.MINING_EFFICIENCY).add(Attributes.SWEEPING_DAMAGE_RATIO).add(Attributes.SPAWN_REINFORCEMENTS_CHANCE); // KioCG - 属性用于玩家温度
     }
 
     @Override
@@ -294,6 +296,8 @@ public abstract class Player extends LivingEntity {
         this.moveCloak();
         if (!this.level().isClientSide) {
             this.foodData.tick(this);
+            this.coldData.tick(); // KioCG
+            this.thirstData.tick(); // KioCG
             this.awardStat(Stats.PLAY_TIME);
             this.awardStat(Stats.TOTAL_WORLD_TIME);
             if (this.isAlive()) {
@@ -888,6 +892,10 @@ public abstract class Player extends LivingEntity {
 
         this.setScore(nbt.getInt("Score"));
         this.foodData.readAdditionalSaveData(nbt);
+        // KioCG start
+        this.coldData.readAdditionalSaveData(nbt);
+        this.thirstData.readAdditionalSaveData(nbt);
+        // KioCG end
         this.abilities.loadSaveData(nbt);
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue((double) this.abilities.getWalkingSpeed());
         if (nbt.contains("EnderItems", 9)) {
@@ -937,6 +945,10 @@ public abstract class Player extends LivingEntity {
         nbt.putInt("XpSeed", this.enchantmentSeed);
         nbt.putInt("Score", this.getScore());
         this.foodData.addAdditionalSaveData(nbt);
+        // KioCG start
+        this.coldData.addAdditionalSaveData(nbt);
+        this.thirstData.addAdditionalSaveData(nbt);
+        // KioCG end
         this.abilities.addSaveData(nbt);
         nbt.put("EnderItems", this.enderChestInventory.createTag(this.registryAccess()));
         if (!this.getShoulderEntityLeft().isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/food/FoodData.java b/src/main/java/net/minecraft/world/food/FoodData.java
index bd3f78e6453cfe18aa3da38176b04d734d83bb4b..b45e4a9793ddcc1e6d090c6a18f99e3b9bc35635 100644
--- a/src/main/java/net/minecraft/world/food/FoodData.java
+++ b/src/main/java/net/minecraft/world/food/FoodData.java
@@ -74,6 +74,7 @@ public class FoodData {
 
                 if (!event.isCancelled()) {
                     this.foodLevel = event.getFoodLevel();
+                    player.thirstData.addThirstRegain(Math.max(this.lastFoodLevel - this.foodLevel, 0) * player.level().paperConfig().kiocgConfig.theLongDark.thirstValue.thirstRegainMultiplierFood); // KioCG
                 }
 
                 ((ServerPlayer) player).connection.send(new ClientboundSetHealthPacket(((ServerPlayer) player).getBukkitEntity().getScaledHealth(), this.foodLevel, this.saturationLevel));
diff --git a/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
index d29a62775913922ffb8e3c58ae0db7e37f77226e..9ac264e0a6342d556a0df3c9081b4ef828ae0451 100644
--- a/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LavaCauldronBlock.java
@@ -35,6 +35,7 @@ public class LavaCauldronBlock extends AbstractCauldronBlock {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (this.isEntityInsideContent(state, pos, entity)) {
             entity.lavaHurt();
+            if (entity instanceof net.minecraft.world.entity.player.Player player) player.coldData.isInLava = true; // KioCG
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
index a5c7c2d24498c66159316a4f92677625975ce5ca..66dabb34eb4dfdd526ee1800d04aeda0cc99fcda 100644
--- a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -78,6 +78,7 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
             // CraftBukkit end
         }
 
+        if (entity instanceof net.minecraft.world.entity.player.Player player && this.isEntityInsideContent(state, pos, entity)) player.coldData.isInWater = true; // KioCG
     }
 
     @io.papermc.paper.annotation.DoNotUse @Deprecated // Paper - fix powdered snow cauldron extinguishing entities; use #handleEntityOnFireInsideWithEvent
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4a4707d65d2d6d03964bb107e9ef8cac5c0f2314..d5386e47cc6e3996f15373dca5b08a4b447da77d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -207,6 +207,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private double health = 20;
     private boolean scaledHealth = false;
     private double healthScale = 20;
+    public int realAir; // KioCG
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
     public org.bukkit.event.player.PlayerResourcePackStatusEvent.Status resourcePackStatus; // Paper - more resource pack API
@@ -2230,7 +2231,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         playerList.sendAllPlayerInfo(handle);
 
         // Resend their XP and effects because the respawn packet resets it
-        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.experienceProgress, handle.totalExperience, handle.experienceLevel));
+        connection.send(new net.minecraft.network.protocol.game.ClientboundSetExperiencePacket(handle.coldData.getColdProgress(), handle.totalExperience, handle.experienceLevel)); // KioCG - cold system
         for (net.minecraft.world.effect.MobEffectInstance mobEffect : handle.getActiveEffects()) {
             connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket(handle.getId(), mobEffect, false));
         }
@@ -3520,6 +3521,47 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end
 
+    // KioCG start
+    @Override
+    public int getColdValue() {
+        return this.getHandle().coldData.getColdValue();
+    }
+    @Override
+    public void setColdValue(int value) {
+        this.getHandle().coldData.setColdValue(value);
+    }
+    @Override
+    public void addColdValue(int add) {
+        this.getHandle().coldData.addColdValue(add);
+    }
+
+    @Override
+    public int getThirstValue() {
+        return this.getHandle().thirstData.getThirstValue();
+    }
+    @Override
+    public void setThirstValue(int value) {
+        this.getHandle().thirstData.setThirstValue(value);
+    }
+    @Override
+    public void addThirstValue(int add) {
+        this.getHandle().thirstData.addThirstValue(add);
+    }
+
+    @Override
+    public double getThirstRegain() {
+        return this.getHandle().thirstData.getThirstRegain();
+    }
+    @Override
+    public void setThirstRegain(double value) {
+        this.getHandle().thirstData.setThirstRegain(value);
+    }
+    @Override
+    public void addThirstRegain(double add) {
+        this.getHandle().thirstData.addThirstRegain(add);
+    }
+    // KioCG end
+
     public Player.Spigot spigot()
     {
         return this.spigot;
