From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 28 May 2024 15:02:18 +0800
Subject: [PATCH] Entity baby API


diff --git a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
index 1029b6de6f36b08bf634b4056ef5701383f6f258..16d8ae74f535614b21366244eb741551fd5f9e59 100644
--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -241,6 +241,7 @@ public class PaperConfigurations extends Configurations<GlobalConfiguration, Wor
                 .serializers(serializers -> serializers
                     .register(new TypeToken<Reference2IntMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<Reference2IntMap<?>>(Reference2IntOpenHashMap::new, Integer.TYPE))
                     .register(new TypeToken<Reference2LongMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<Reference2LongMap<?>>(Reference2LongOpenHashMap::new, Long.TYPE))
+                    .register(new TypeToken<it.unimi.dsi.fastutil.objects.Reference2DoubleMap<?>>() {}, new FastutilMapSerializer.SomethingToPrimitive<it.unimi.dsi.fastutil.objects.Reference2DoubleMap<?>>(it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap::new, Double.TYPE)) // KioCG
                     .register(new TypeToken<Table<?, ?, ?>>() {}, new TableSerializer())
                     .register(DespawnRange.class, DespawnRange.SERIALIZER)
                     .register(StringRepresentableSerializer::isValidFor, new StringRepresentableSerializer())
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 3be0c10c28b5dd5af8fa865692bbc626cf522c78..1467449327831fae6d6500e704881dd88417cc92 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -616,6 +616,18 @@ public class WorldConfiguration extends ConfigurationPart {
 
             public int campfireBoilingWaterTime = 600;
         }
+
+        public Entity entity;
+        public class Entity extends ConfigurationPart {
+            public it.unimi.dsi.fastutil.objects.Reference2DoubleMap<EntityType<?>> sizeScalingBaby = Util.make(new it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
+                map.defaultReturnValue(0.5);
+                map.put(EntityType.SKELETON, 0.65);
+            });
+            public it.unimi.dsi.fastutil.objects.Reference2DoubleMap<EntityType<?>> sizeScalingHuge = Util.make(new it.unimi.dsi.fastutil.objects.Reference2DoubleOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
+                map.defaultReturnValue(1.5);
+                map.put(EntityType.SKELETON, 1.5);
+            });
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/AgeableMob.java b/src/main/java/net/minecraft/world/entity/AgeableMob.java
index 3dc3609d13a7b823d15384d1c385b68eeb933d26..7cbb0cce42c1a63b10f6e1c3f0c006d747b61c10 100644
--- a/src/main/java/net/minecraft/world/entity/AgeableMob.java
+++ b/src/main/java/net/minecraft/world/entity/AgeableMob.java
@@ -196,7 +196,8 @@ public abstract class AgeableMob extends PathfinderMob {
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
+        if (baby) removeScaleAttributes(); // KioCG
         this.setAge(baby ? -24000 : 0);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 91176b3f928de7db98e69bde28e4db144c7c0775..a37ab1103ad00965c3f36d8b49f4fb8fd78e5d07 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -286,6 +286,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
+    // KioCG start
+    private ScaleSimulate scaleSimulate = ScaleSimulate.NORMAL;
+    private final AttributeModifier SCALE_MODIFIER_BABY = new AttributeModifier(ResourceLocation.withDefaultNamespace("scale_simulate"), this.level().paperConfig().kiocgConfig.entity.sizeScalingBaby.getOrDefault(this.getType(), 0.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    private final AttributeModifier SCALE_MODIFIER_HUGE = new AttributeModifier(ResourceLocation.withDefaultNamespace("scale_simulate"), this.level().paperConfig().kiocgConfig.entity.sizeScalingHuge.getOrDefault(this.getType(), 1.5) - 1D, AttributeModifier.Operation.ADD_MULTIPLIED_TOTAL);
+    // KioCG end
 
     @Override
     public float getBukkitYaw() {
@@ -591,6 +596,66 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return false;
     }
 
+    // KioCG start
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> {
+                if (isBaby()) return; // Vanilla baby
+
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_BABY);
+            }
+            case HUGE -> {
+                this.scaleSimulate = scale;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.addOrUpdateTransientModifier(this.SCALE_MODIFIER_HUGE);
+            }
+            default -> {
+                this.scaleSimulate = ScaleSimulate.NORMAL;
+                AttributeInstance attributemodifiable = this.getAttribute(Attributes.SCALE);
+                attributemodifiable.removeModifier(this.SCALE_MODIFIER_BABY.id()); // SCALE_MODIFIER_BABY 和 SCALE_MODIFIER_HUGE 使用相同的ID
+            }
+        }
+    }
+
+    public void setBabySimulate() {
+        this.setScaleSimulate(ScaleSimulate.BABY);
+    }
+
+    public boolean isBabySimulate() {
+        return this.scaleSimulate == ScaleSimulate.BABY || this.isBaby();
+    }
+
+    public void setHugeSimulate() {
+        this.setScaleSimulate(ScaleSimulate.HUGE);
+    }
+
+    public boolean isHugeSimulate() {
+        return this.scaleSimulate == ScaleSimulate.HUGE;
+    }
+
+    public void removeScaleAttributes() {
+        this.setScaleSimulate(ScaleSimulate.NORMAL);
+    }
+
+    public void transformScaleSimulate(LivingEntity original) {
+        if (original.isBabySimulate()) {
+            setBabySimulate();
+        } else if (original.isHugeSimulate()) {
+            setHugeSimulate();
+        } else {
+            removeScaleAttributes();
+        }
+    }
+
+    enum ScaleSimulate {
+        BABY,
+        HUGE,
+        NORMAL
+    }
+    // KioCG end
+
     public float getAgeScale() {
         return this.isBaby() ? 0.5F : 1.0F;
     }
@@ -812,6 +877,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
         dataresult.resultOrPartial(logger::error).ifPresent((nbtbase) -> {
             nbt.put("Brain", nbtbase);
         });
+
+        // KioCG start
+        if (this.scaleSimulate != ScaleSimulate.NORMAL) {
+            nbt.putString("KioCG.ScaleSimulate", this.scaleSimulate.toString());
+        }
+        // KioCG end
     }
 
     @Override
@@ -900,6 +971,16 @@ public abstract class LivingEntity extends Entity implements Attackable {
             this.brain = this.makeBrain(new Dynamic(NbtOps.INSTANCE, nbt.get("Brain")));
         }
 
+        // KioCG start
+        if (nbt.contains("KioCG.ScaleSimulate")) {
+            String ss = nbt.getString("KioCG.ScaleSimulate");
+            try {
+                this.setScaleSimulate(ScaleSimulate.valueOf(ss));
+            } catch (Exception ignored) {
+                LOGGER.error("Unknown scale simulate " + ss + " for " + this);
+            }
+        }
+        // KioCG end
     }
 
     // CraftBukkit start
@@ -2779,7 +2860,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public float getVoicePitch() {
-        return this.isBaby() ? (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.5F : (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F;
+        return this.isBabySimulate() ? (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.5F : (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F; // KioCG
     }
 
     protected boolean isImmobile() {
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 64d4608388f177d7cf4f16b643e6d08a8660000c..4a1cd027990cc61ee29c496cb2a1e505eb417952 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1692,7 +1692,20 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         return ((Byte) this.entityData.get(Mob.DATA_MOB_FLAGS_ID) & 4) != 0;
     }
 
-    public void setBaby(boolean baby) {}
+    public final void setBaby(boolean baby) { this.setBaby1(baby); } // KioCG - final防止直接重写, 便于版本更新检查
+
+    // KioCG start
+    public void setBaby1(boolean baby) {}
+
+    public void setScaleSimulate(ScaleSimulate scale) {
+        switch (scale) {
+            case BABY -> setBaby(true);
+            case HUGE -> setBaby(false);
+        }
+
+        super.setScaleSimulate(scale);
+    }
+    // KioCG end
 
     @Override
     public HumanoidArm getMainArm() {
diff --git a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
index 816977990639ec0559b652fc9666afd5046f0a5d..413abb374ca11479d658b9cd1c94ca9f0d250455 100644
--- a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
@@ -256,7 +256,7 @@ public class Frog extends Animal implements VariantHolder<Holder<FrogVariant>> {
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index fadd341ff398886a4da102eefa1beb95a63bbd6d..dde0da4c808e10dc942fd8f5f2096597fcc3d678 100644
--- a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -430,7 +430,8 @@ public class Sniffer extends Animal {
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
+        if (baby) removeScaleAttributes(); // KioCG
         this.setAge(baby ? -48000 : 0);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
index aa458ede5bd645ebf524238179edb33f41bd683f..563e287a5e0b970f5dcfa3a7f20bf939262e54c4 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
@@ -240,11 +240,17 @@ public class Zoglin extends Monster implements Enemy, HoglinBase {
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
+        if (baby) removeScaleAttributes(); // KioCG
         this.getEntityData().set(DATA_BABY_ID, baby);
         if (!this.level().isClientSide && baby) {
             this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(0.5);
         }
+        // KioCG start
+        if (!this.level().isClientSide && !baby) {
+            this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(ATTACK_DAMAGE);
+        }
+        // KioCG end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 64be06ad25b0b2af3b5b18fc2aa9c1e3b39a328f..01cdca299beed6ef8b9e39cb4eeb4b78ad3c56ba 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -189,7 +189,8 @@ public class Zombie extends Monster {
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
+        if (baby) removeScaleAttributes(); // KioCG
         this.getEntityData().set(Zombie.DATA_BABY_ID, baby);
         if (this.level() != null && !this.level().isClientSide) {
             AttributeInstance attributemodifiable = this.getAttribute(Attributes.MOVEMENT_SPEED);
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index bc58323801ee16fe9b63c21332144ec002a902f2..2ead7a0286a2109656ead0d1ee779b835ddf82c0 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -266,7 +266,8 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     }
 
     @Override
-    public void setBaby(boolean baby) {
+    public void setBaby1(boolean baby) { // KioCG
+        if (baby) removeScaleAttributes(); // KioCG
         this.getEntityData().set(Piglin.DATA_BABY_ID, baby);
         if (!this.level().isClientSide) {
             AttributeInstance attributemodifiable = this.getAttribute(Attributes.MOVEMENT_SPEED);
diff --git a/src/main/java/net/minecraft/world/item/SpawnEggItem.java b/src/main/java/net/minecraft/world/item/SpawnEggItem.java
index 9cea8da84f39bb3f687139ef213ccea358724dee..4e4a7c346359468c0754285a023bbf677411a8d9 100644
--- a/src/main/java/net/minecraft/world/item/SpawnEggItem.java
+++ b/src/main/java/net/minecraft/world/item/SpawnEggItem.java
@@ -174,8 +174,10 @@ public class SpawnEggItem extends Item {
             if (object == null) {
                 return Optional.empty();
             } else {
-                ((Mob) object).setBaby(true);
-                if (!((Mob) object).isBaby()) {
+                // KioCG start
+                ((Mob) object).setBabySimulate();
+                if (false && !((Mob) object).isBaby()) {
+                // KioCG end
                     return Optional.empty();
                 } else {
                     ((Mob) object).moveTo(pos.x(), pos.y(), pos.z(), 0.0F, 0.0F);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 4750df7354ccb5afd0910efe0415f3a2eb19a546..04e6db8e850464f98e1ab985ae15d1026f736d08 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1192,4 +1192,31 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return this.getHandle().canUseSlot(org.bukkit.craftbukkit.CraftEquipmentSlot.getNMS(slot));
     }
     // Paper end - Expose canUseSlot
+
+    // KioCG start
+    @Override
+    public void setBabySimulate() {
+        getHandle().setBabySimulate();
+    }
+
+    @Override
+    public boolean isBabySimulate() {
+        return getHandle().isBabySimulate();
+    }
+
+    @Override
+    public void setHugeSimulate() {
+        getHandle().setHugeSimulate();
+    }
+
+    @Override
+    public boolean isHugeSimulate() {
+        return getHandle().isHugeSimulate();
+    }
+
+    @Override
+    public void removeScaleAttributes() {
+        getHandle().removeScaleAttributes();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f5408c4bc12e87d80d7137f0a7c719417ba68591..6e0d87d695aa7ee301124f163dcc9ed77e5ff915 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -619,6 +619,7 @@ public class CraftEventFactory {
     public static EntityTransformEvent callEntityTransformEvent(net.minecraft.world.entity.LivingEntity original, List<net.minecraft.world.entity.LivingEntity> convertedList, EntityTransformEvent.TransformReason convertType) {
         List<org.bukkit.entity.Entity> list = new ArrayList<>();
         for (net.minecraft.world.entity.LivingEntity entityLiving : convertedList) {
+            entityLiving.transformScaleSimulate(original); // KioCG
             list.add(entityLiving.getBukkitEntity());
         }
 
@@ -1356,6 +1357,7 @@ public class CraftEventFactory {
     }
 
     public static PigZapEvent callPigZapEvent(Entity pig, Entity lightning, Entity pigzombie) {
+        ((net.minecraft.world.entity.LivingEntity) pigzombie).transformScaleSimulate((net.minecraft.world.entity.LivingEntity) pig); // KioCG
         PigZapEvent event = new PigZapEvent((Pig) pig.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), (PigZombie) pigzombie.getBukkitEntity());
         pig.getBukkitEntity().getServer().getPluginManager().callEvent(event);
         return event;
