From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 15 Mar 2024 10:48:31 +0800
Subject: [PATCH] Auto discard rider display entity


diff --git a/src/main/java/net/minecraft/world/entity/Display.java b/src/main/java/net/minecraft/world/entity/Display.java
index a658943669cdfd49f66ba713505d11b33306ed86..0596375a5a4a25982a398f6d6c1d2a3324544fce 100644
--- a/src/main/java/net/minecraft/world/entity/Display.java
+++ b/src/main/java/net/minecraft/world/entity/Display.java
@@ -97,6 +97,7 @@ public abstract class Display extends Entity {
     private Display.RenderState renderState;
     @Nullable
     private Display.PosRotInterpolationTarget posRotInterpolationTarget;
+    private int riderPriority = 0; public int getRiderPriority() { return this.riderPriority; } public void setRiderPriority(int riderPriority) { this.riderPriority = riderPriority; } // KioCG
 
     public Display(EntityType<?> type, Level world) {
         super(type, world);
@@ -180,6 +181,17 @@ public abstract class Display extends Entity {
         }
     }
 
+    // KioCG start
+    @Override
+    public void stopRiding(boolean suppressCancellation) {
+        Entity entity = this.getVehicle();
+        super.stopRiding(suppressCancellation);
+        if (entity != null && this.getVehicle() == null) {
+            this.discard();
+        }
+    }
+    // KioCG end
+
     protected abstract void updateRenderSubState(boolean shouldLerp, float lerpProgress);
 
     @Override
@@ -264,6 +276,8 @@ public abstract class Display extends Entity {
         } else {
             this.setBrightnessOverride(null);
         }
+
+        if (nbt.contains("KioCG.RiderPriority")) this.riderPriority = nbt.getInt("KioCG.RiderPriority"); // KioCG
     }
 
     public void setTransformation(Transformation transformation) {
@@ -291,6 +305,8 @@ public abstract class Display extends Entity {
         if (brightness != null) {
             Brightness.CODEC.encodeStart(NbtOps.INSTANCE, brightness).ifSuccess(brightnessx -> nbt.put("brightness", brightnessx));
         }
+
+        nbt.putInt("KioCG.RiderPriority", this.riderPriority); // KioCG
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 38f3f368a50cc0e1f0638424e4e92b264ee64abd..597e277481b26178b25f287f005715c86ad360ff 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4762,6 +4762,42 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return null;
     }
 
+    // KioCG start
+    @Nullable
+    public <T extends Display> T addRiderDisplay(EntityType<T> type, java.util.function.Consumer<T> function) {
+        return addRiderDisplay(type, 0, function);
+    }
+
+    @Nullable
+    public <T extends Display> T addRiderDisplay(EntityType<T> type, int priority, java.util.function.Consumer<T> function) {
+        final Entity firstPassenger = this.getFirstPassenger();
+        if (firstPassenger == null || firstPassenger instanceof Display display && display.getRiderPriority() < priority) {
+            this.ejectPassengers();
+
+            T display = type.create(this.level());
+            if (display != null) {
+                display.moveTo(this.position());
+                display.persist = false;
+                display.setRiderPriority(priority);
+                display.startRiding(this);
+                function.accept(display);
+                this.level().addFreshEntity(display);
+            }
+            return display;
+        }
+        return null;
+    }
+
+    public boolean removeRiderDisplay(@Nullable UUID uuid) {
+        final Entity firstPassenger = this.getFirstPassenger();
+        if (firstPassenger instanceof Display display && (uuid == null || display.getUUID().equals(uuid))) {
+            firstPassenger.discard();
+            return true;
+        }
+        return false;
+    }
+    // KioCG end
+
     public static enum RemovalReason {
 
         KILLED(true, false), DISCARDED(true, false), UNLOADED_TO_CHUNK(false, true), UNLOADED_WITH_PLAYER(false, false), CHANGED_DIMENSION(false, false);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
index 48eeb1d9ba0ad6f895bfe507a6fbe4b9c9530e47..9f188b9e47948c111a89f3923c0d5386811215e1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftDisplay.java
@@ -167,4 +167,6 @@ public class CraftDisplay extends CraftEntity implements Display {
             this.getHandle().setBrightnessOverride(null);
         }
     }
+
+    public int getRiderPriority() { return this.getHandle().getRiderPriority(); } public void setRiderPriority(int riderPriority) { this.getHandle().setRiderPriority(riderPriority); } // KioCG
 }
