From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 14 Mar 2024 20:52:54 +0800
Subject: [PATCH] Dont kill my pet


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index f5bcc78445ee29132b6947544748eeae16fa0a0f..5f2e951daf8b74267ebd09c29a44fe12f0fb4200 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -416,6 +416,10 @@ public class GlobalConfiguration extends ConfigurationPart {
             public boolean allowUnsafeCommand = true;
             public int fixedCumulativeCost = 4;
         }
+
+        public boolean dontKillMyPet = true;
+        public double petTeleportDistanceSqr = 256.0;
+        public boolean petFollowKeepChunkLoaded = true;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/TicketType.java b/src/main/java/net/minecraft/server/level/TicketType.java
index 658e63ebde81dc14c8ab5850fb246dc0aab25dea..59b1fc323174dcd9c5d3b98724f6eec575635b8c 100644
--- a/src/main/java/net/minecraft/server/level/TicketType.java
+++ b/src/main/java/net/minecraft/server/level/TicketType.java
@@ -28,6 +28,7 @@ public class TicketType<T> {
     public static final TicketType<Unit> PLUGIN = TicketType.create("plugin", (a, b) -> 0); // CraftBukkit
     public static final TicketType<org.bukkit.plugin.Plugin> PLUGIN_TICKET = TicketType.create("plugin_ticket", (plugin1, plugin2) -> plugin1.getClass().getName().compareTo(plugin2.getClass().getName())); // CraftBukkit
     public static final TicketType<Long> CHUNK_RELIGHT = create("light_update", Long::compareTo); // Paper - ensure chunks stay loaded for lighting
+    public static final TicketType<Integer> PET_FOLLOW = TicketType.create("pet_follow", Integer::compareTo, 20 * 3); // KioCG
     // Paper start - rewrite chunk system
     public static final TicketType<Long> CHUNK_LOAD = create("chunk_load", Long::compareTo);
     public static final TicketType<Long> STATUS_UPGRADE = create("status_upgrade", Long::compareTo);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 9a5f63c03cfff839b58fef63fc3a265143bfc7a7..ea576e709f4616d7cbb09695d63e3b6c1a33deb7 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2376,7 +2376,15 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 }
                 // CraftBukkit end
                 this.getCombatTracker().recordDamage(damagesource, f);
-                this.setHealth(this.getHealth() - f);
+                // KioCG start
+                float health = this.getHealth() - f;
+                if (health <= 0.0F && originalDamage != Float.MAX_VALUE && !damagesource.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this instanceof TamableAnimal pet && pet.isTame()) {
+                        health = Float.MIN_VALUE;
+                    }
+                }
+                this.setHealth(health);
+                // KioCG end
                 // CraftBukkit start
                 if (!human) {
                     this.setAbsorptionAmount(this.getAbsorptionAmount() - f);
diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index 45284bb0f081377bc5913876c6877a424285c0bb..85bc846db2bec56ec1ba44e378bde929a252ee0c 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -213,11 +213,50 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         super.die(damageSource);
     }
 
+    // KioCG start
+    public boolean onTheVergeOfDeath() {
+        return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this.isTame() && this.getHealth() == Float.MIN_VALUE;
+    }
+
+    @Override
+    public boolean canBeSeenAsEnemy() {
+        return !this.onTheVergeOfDeath() && super.canBeSeenAsEnemy();
+    }
+
+    @Override
+    public boolean hurt(DamageSource source, float amount) {
+        return (amount == Float.MAX_VALUE || !onTheVergeOfDeath()) && super.hurt(source, amount);
+    }
+
+    @Nullable private UUID helpDisplayUUID;
+
+    public void tryAddHelpDisplay() {
+        final Display.ItemDisplay itemDisplay = this.addRiderDisplay(EntityType.ITEM_DISPLAY, 99, display -> {
+            display.setItemStack(new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.BARRIER));
+            display.setItemTransform(net.minecraft.world.item.ItemDisplayContext.GROUND);
+            display.setBillboardConstraints(Display.BillboardConstraints.VERTICAL);
+            display.setTransformation(new com.mojang.math.Transformation(new org.joml.Vector3f(0.0F, 0.5F, 0.0F), null, null, null));
+        });
+
+        if (itemDisplay != null) {
+            helpDisplayUUID = itemDisplay.getUUID();
+        }
+    }
+
+    public void tryRemoveHelpDisplay() {
+        if (helpDisplayUUID != null) {
+            this.removeRiderDisplay(helpDisplayUUID);
+            helpDisplayUUID = null;
+        }
+    }
+    // KioCG end
+
     public boolean isOrderedToSit() {
         return this.orderedToSit;
     }
 
     public void setOrderedToSit(boolean sitting) {
+        if (!sitting && onTheVergeOfDeath()) return; // KioCG
         this.orderedToSit = sitting;
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java b/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
index fe635e46569c67dac1d3581ee930d1bfa8b4030e..8c6731fa6d693e652031f35aee2c65c3587ec637 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -53,7 +53,7 @@ public class StopAttackingIfTargetInvalid {
                             reason = EntityTargetEvent.TargetReason.TARGET_INVALID;
                         } else if (shouldForgetIfTargetUnreachable && StopAttackingIfTargetInvalid.isTiredOfTryingToReachTarget(entityinsentient, behaviorbuilder_b.tryGet(memoryaccessor1))) {
                             reason = EntityTargetEvent.TargetReason.FORGOT_TARGET;
-                        } else if (!entityliving.isAlive()) {
+                        } else if (!entityliving.isAlive() || entityliving instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.onTheVergeOfDeath()) { // KioCG
                             reason = EntityTargetEvent.TargetReason.TARGET_DIED;
                         } else if (entityliving.level() != entityinsentient.level()) {
                             reason = EntityTargetEvent.TargetReason.TARGET_OTHER_LEVEL;
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
index edd29c2f4d0151d512618115a8fb4b7423171491..88a6311f3ce0b1d30533ad941e7696e1e9309171 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -36,6 +36,7 @@ public class FollowOwnerGoal extends Goal {
     private final float startDistance;
     private float oldWaterCost;
     private final boolean canFly;
+    private int ticketTicks; // KioCG
 
     public FollowOwnerGoal(TamableAnimal tameable, double speed, float minDistance, float maxDistance, boolean leavesAllowed) {
         this.tamable = tameable;
@@ -58,10 +59,12 @@ public class FollowOwnerGoal extends Goal {
         if (entityliving == null) {
             return false;
         } else if (entityliving.isSpectator()) {
+            if (!this.unableToMove()) keepChunkLoaded(); // KioCG
             return false;
         } else if (this.unableToMove()) {
             return false;
         } else if (this.tamable.distanceToSqr((Entity) entityliving) < (double) (this.startDistance * this.startDistance)) {
+            keepChunkLoaded(); // KioCG
             return false;
         } else {
             this.owner = entityliving;
@@ -69,6 +72,16 @@ public class FollowOwnerGoal extends Goal {
         }
     }
 
+    // KioCG start
+    private void keepChunkLoaded() {
+        if (!io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petFollowKeepChunkLoaded) return;
+        if (this.tamable.tickCount >= ticketTicks + 20) {
+            ticketTicks = this.tamable.tickCount;
+            ((net.minecraft.server.level.ServerLevel) this.level).getChunkSource().addRegionTicket(net.minecraft.server.level.TicketType.PET_FOLLOW, new net.minecraft.world.level.ChunkPos(this.tamable.blockPosition()), 2, this.tamable.getId()); // KioCG
+        }
+    }
+    // KioCG end
+
     @Override
     public boolean canContinueToUse() {
         return this.navigation.isDone() ? false : (this.unableToMove() ? false : this.tamable.distanceToSqr((Entity) this.owner) > (double) (this.stopDistance * this.stopDistance));
@@ -94,10 +107,11 @@ public class FollowOwnerGoal extends Goal {
 
     @Override
     public void tick() {
+        keepChunkLoaded(); // KioCG
         if (this.tamable.distanceToSqr(this.owner) <= 16 * 16) this.tamable.getLookControl().setLookAt(this.owner, 10.0F, (float) this.tamable.getMaxHeadXRot()); // Paper - Limit pet look distance
         if (--this.timeToRecalcPath <= 0) {
             this.timeToRecalcPath = this.adjustedTickDelay(10);
-            if (this.tamable.distanceToSqr((Entity) this.owner) >= 144.0D) {
+            if (this.tamable.distanceToSqr((Entity) this.owner) >= io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petTeleportDistanceSqr) { // KioCG
                 this.teleportToOwner();
             } else {
                 this.navigation.moveTo((Entity) this.owner, this.speedModifier);
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
index d59e0044db03ba30d5bab9a8c126b597b7659015..95696b7a81ea5454f6b571151ab1149c2e784b73 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -39,7 +39,7 @@ public class MeleeAttackGoal extends Goal {
             LivingEntity livingEntity = this.mob.getTarget();
             if (livingEntity == null) {
                 return false;
-            } else if (!livingEntity.isAlive()) {
+            } else if (!livingEntity.isAlive() || livingEntity instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.onTheVergeOfDeath()) { // KioCG
                 return false;
             } else {
                 this.path = this.mob.getNavigation().createPath(livingEntity, 0);
@@ -53,7 +53,7 @@ public class MeleeAttackGoal extends Goal {
         LivingEntity livingEntity = this.mob.getTarget();
         if (livingEntity == null) {
             return false;
-        } else if (!livingEntity.isAlive()) {
+        } else if (!livingEntity.isAlive() || livingEntity instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.onTheVergeOfDeath()) { // KioCG
             return false;
         } else {
             return !this.followingTargetEvenIfNotSeen
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
index cdd006fef3c94205fe8378bfa712aae4faacad04..b2449c9c03b211b4c367f94b8744400f41a5a827 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
@@ -46,7 +46,7 @@ public class RangedCrossbowAttackGoal<T extends Monster & RangedAttackMob & Cros
     }
 
     private boolean isValidTarget() {
-        return this.mob.getTarget() != null && this.mob.getTarget().isAlive();
+        return this.mob.getTarget() != null && this.mob.getTarget().isAlive() && (!(this.mob.getTarget() instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal) || !tamableAnimal.onTheVergeOfDeath()); // KioCG
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cat.java b/src/main/java/net/minecraft/world/entity/animal/Cat.java
index b102a370f5fc1f80c199a4e4d1f6ab9a9480fab3..4d83352e5c100864bf9e881e14ee35b7b06f8eb2 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cat.java
@@ -114,7 +114,14 @@ public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVarian
             return itemstack.is(ItemTags.CAT_FOOD);
         }, true);
         this.goalSelector.addGoal(1, new FloatGoal(this));
-        this.goalSelector.addGoal(1, new PanicGoal(this, 1.5D));
+        // KioCG start
+        this.goalSelector.addGoal(1, new PanicGoal(this, 1.5D) {
+            @Override
+            public boolean shouldPanic() {
+                return !((TamableAnimal) this.mob).onTheVergeOfDeath() && super.shouldPanic();
+            }
+        });
+        // KioCG end
         this.goalSelector.addGoal(2, new SitWhenOrderedToGoal(this));
         this.goalSelector.addGoal(3, new Cat.CatRelaxOnOwnerGoal(this));
         this.goalSelector.addGoal(4, this.temptGoal);
@@ -279,6 +286,16 @@ public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVarian
         }
 
         this.handleLieDown();
+        // KioCG start
+        if (this.onTheVergeOfDeath()) {
+            tryAddHelpDisplay();
+            if (!this.isOrderedToSit()) {
+                this.setOrderedToSit(true);
+            }
+        } else {
+            tryRemoveHelpDisplay();
+        }
+        // KioCG end
     }
 
     private void handleLieDown() {
@@ -419,6 +436,18 @@ public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVarian
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
+                    // KioCG start
+                } else if (itemstack.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) {
+                    this.playSound(SoundEvents.CAT_EAT, 1.0F, 1.0F);
+                    super.usePlayerItem(player, hand, itemstack);
+
+                    this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                    if (player.isCreative() || !this.isInvulnerable()) {
+                        this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                    }
+
+                    return InteractionResult.sidedSuccess(this.level().isClientSide());
+                    // KioCG end
                 }
 
                 enuminteractionresult = super.mobInteract(player, hand);
diff --git a/src/main/java/net/minecraft/world/entity/animal/Parrot.java b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
index 9207b5830b66d26e6277664dcebc5beb30841fc5..1fde2e7e91ed41356fc93fe0356ce0f9d041f01b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Parrot.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
@@ -148,7 +148,14 @@ public class Parrot extends ShoulderRidingEntity implements VariantHolder<Parrot
 
     @Override
     protected void registerGoals() {
-        this.goalSelector.addGoal(0, new PanicGoal(this, 1.25D));
+        // KioCG start
+        this.goalSelector.addGoal(0, new PanicGoal(this, 1.25D) {
+            @Override
+            public boolean shouldPanic() {
+                return !((net.minecraft.world.entity.TamableAnimal) this.mob).onTheVergeOfDeath() && super.shouldPanic();
+            }
+        });
+        // KioCG end
         this.goalSelector.addGoal(0, new FloatGoal(this));
         this.goalSelector.addGoal(1, new LookAtPlayerGoal(this, Player.class, 8.0F));
         this.goalSelector.addGoal(2, new SitWhenOrderedToGoal(this));
@@ -201,6 +208,21 @@ public class Parrot extends ShoulderRidingEntity implements VariantHolder<Parrot
         this.calculateFlapping();
     }
 
+    // KioCG start
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.onTheVergeOfDeath()) {
+            tryAddHelpDisplay();
+            if (!this.isOrderedToSit()) {
+                this.setOrderedToSit(true);
+            }
+        } else {
+            tryRemoveHelpDisplay();
+        }
+    }
+    // KioCG end
+
     @Override
     public void setRecordPlayingNearby(BlockPos songPosition, boolean playing) {
         this.jukebox = songPosition;
@@ -271,7 +293,7 @@ public class Parrot extends ShoulderRidingEntity implements VariantHolder<Parrot
             }
 
             return InteractionResult.sidedSuccess(this.level().isClientSide);
-        } else if (!itemstack.is(ItemTags.PARROT_POISONOUS_FOOD)) {
+        } else if (!itemstack.is(ItemTags.PARROT_POISONOUS_FOOD) && !itemstack.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) { // KioCG
             if (!this.isFlying() && this.isTame() && this.isOwnedBy(player)) {
                 if (!this.level().isClientSide) {
                     this.setOrderedToSit(!this.isOrderedToSit());
diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index 5dfa2184aaa2955e54456e6f9f78507daae1e385..a422de00d5a2a2d155ee7a0f46cb0d27859bfebc 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -324,6 +324,18 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
                 }
             }
 
+            // KioCG start
+            if (this.onTheVergeOfDeath()) {
+                tryAddHelpDisplay();
+                if (!this.isOrderedToSit()) {
+                    this.navigation.stop();
+                    this.setTarget((LivingEntity) null);
+                    this.setOrderedToSit(true);
+                }
+            } else {
+                tryRemoveHelpDisplay();
+            }
+            // KioCG end
         }
     }
 
@@ -460,6 +472,16 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
 
                 this.heal(2.0F * f, EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
                 return InteractionResult.sidedSuccess(this.level().isClientSide());
+                // KioCG start
+            } else if (itemstack.is(Items.POISONOUS_POTATO)) {
+                itemstack.consume(1, player);
+                this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                if (player.isCreative() || !this.isInvulnerable()) {
+                    this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                }
+
+                return InteractionResult.sidedSuccess(this.level().isClientSide());
+                // KioCG end
             } else {
                 if (item instanceof DyeItem) {
                     DyeItem itemdye = (DyeItem) item;
@@ -740,6 +762,7 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
 
         @Override
         protected boolean shouldPanic() {
+            if (((TamableAnimal) this.mob).onTheVergeOfDeath()) return false; // KioCG
             return this.mob.isFreezing() || this.mob.isOnFire();
         }
     }
