From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 11:18:03 +0800
Subject: [PATCH] Dirt path speed boost


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 2dc39973841feff83ac12f546414d1db11140b5c..c1f5aff6040124d88c4fcbbc5d2080f838bffe5a 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -594,6 +594,7 @@ public class WorldConfiguration extends ConfigurationPart {
         }
 
         public boolean betterMobGriefingSetting = true;
+        public double dirtPathSpeedBoost = 0.02;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 01f09d50333df16d7f2e76ca7fed964dd6a0f38a..2b4595855214a47052fdd840c666591377e26944 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1624,6 +1624,34 @@ public class ServerPlayer extends Player {
             super.onChangedBlock(pos);
         }
 
+        // KioCG start
+        this.removeDirtPathSpeed();
+        if (this.level().paperConfig().kiocgConfig.dirtPathSpeedBoost > 0) {
+            this.tryAddDirtPathSpeed();
+        }
+    }
+
+    private static final java.util.UUID SPEED_MODIFIER_DIRT_PATH_SPEED_UUID = java.util.UUID.fromString("6e281992-cd8c-e065-a326-353ce8a91273");
+
+    protected void removeDirtPathSpeed() {
+        if (this.shouldRemoveSoulSpeed(this.getBlockStateOnLegacy())) {
+            net.minecraft.world.entity.ai.attributes.AttributeInstance attributemodifiable = this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED);
+            if (attributemodifiable != null) {
+                attributemodifiable.removeModifier(SPEED_MODIFIER_DIRT_PATH_SPEED_UUID);
+            }
+        }
+    }
+
+    protected void tryAddDirtPathSpeed() {
+        if (!this.getBlockStateOnLegacy().isAir()) {
+            if (!this.getAbilities().flying && this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).is(Blocks.DIRT_PATH)) {
+                net.minecraft.world.entity.ai.attributes.AttributeInstance attributemodifiable = this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED);
+                if (attributemodifiable != null) {
+                    attributemodifiable.addTransientModifier(new net.minecraft.world.entity.ai.attributes.AttributeModifier(SPEED_MODIFIER_DIRT_PATH_SPEED_UUID, "Dirt path speed boost", this.level().paperConfig().kiocgConfig.dirtPathSpeedBoost, net.minecraft.world.entity.ai.attributes.AttributeModifier.Operation.ADD_VALUE));
+                }
+            }
+        }
+        // KioCG end
     }
 
     public void doCheckFallDamage(double xDifference, double yDifference, double zDifference, boolean onGround) {
