From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 18 Mar 2024 12:04:58 +0800
Subject: [PATCH] End crystal turret


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 4ed7a939f2af2045ace5575f8542cb17a145fa4c..604bbae43d54a4402b65bd1e0612d1c269bac919 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -589,6 +589,11 @@ public class WorldConfiguration extends ConfigurationPart {
             public int darkEnoughToSprint = 7;
             public boolean babyZombieJockey = true;
             public boolean zombiesSkinnedIntoSkeletons = true;
+
+            public boolean enableCrystalTurret = true;
+            public int turretCooldownDefault = 100;
+            public double turretRangeDefault = 8.0;
+            public double turretDamageDefault = 1.0;
         }
 
         public boolean betterMobGriefingSetting = true;
diff --git a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
index 036640d49a5e891e9a0f767abe33f1f51d6d4cde..7b27c1030934370ce95dd5cab56efc1cd97c54fe 100644
--- a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -31,6 +31,36 @@ public class EndCrystal extends Entity {
     private static final EntityDataAccessor<Boolean> DATA_SHOW_BOTTOM = SynchedEntityData.defineId(EndCrystal.class, EntityDataSerializers.BOOLEAN);
     public int time;
     public boolean generatedByDragonFight = false; // Paper - Fix invulnerable end crystals
+    // KioCG start
+    @Nullable
+    private Integer turretCooldown = null;
+    public int getTurretCooldown() {
+        return java.util.Objects.requireNonNullElse(this.turretCooldown, this.level().paperConfig().kiocgConfig.monster.turretCooldownDefault);
+    }
+    public void setTurretCooldown(@Nullable Integer cooldown) {
+        this.turretCooldown = cooldown;
+    }
+
+    @Nullable
+    private Double turretRange = null;
+    public double getTurretRange() {
+        return java.util.Objects.requireNonNullElse(this.turretRange, this.level().paperConfig().kiocgConfig.monster.turretRangeDefault);
+    }
+    public void setTurretRange(@Nullable Double range) {
+        this.turretRange = range;
+    }
+
+    @Nullable
+    private Double turretDamage = null;
+    public double getTurretDamage() {
+        return java.util.Objects.requireNonNullElse(this.turretDamage, this.level().paperConfig().kiocgConfig.monster.turretDamageDefault);
+    }
+    public void setTurretDamage(@Nullable Double damage) {
+        this.turretDamage = damage;
+    }
+
+    private int beamTicks = 0;
+    // KioCG end
 
     public EndCrystal(EntityType<? extends EndCrystal> type, Level world) {
         super(type, world);
@@ -78,6 +108,31 @@ public class EndCrystal extends Entity {
                 }
             }
             // Paper end - Fix invulnerable end crystals
+            // KioCG start
+            if (this.level().paperConfig().kiocgConfig.monster.enableCrystalTurret && !this.generatedByDragonFight) {
+                if (--beamTicks <= 0) {
+                    setBeamTarget(null);
+                }
+
+                if (getTurretCooldown() <= 0 || this.time % getTurretCooldown() == 0) { // 使炮塔间的发射间隔错开
+                    destroyTarget();
+                }
+            }
+        }
+    }
+
+    private void destroyTarget() {
+        java.util.List<net.minecraft.world.entity.monster.Monster> list1 = this.level().getEntitiesOfClass(net.minecraft.world.entity.monster.Monster.class, this.getBoundingBox().inflate(getTurretRange()), (entityliving1) -> {
+            return entityliving1.isAlive() && entityliving1.hasLineOfSight(this);
+        });
+
+        if (!list1.isEmpty()) {
+            net.minecraft.world.entity.LivingEntity entityliving = list1.get(this.level().random.nextInt(list1.size()));
+            entityliving.hurt(this.level().damageSources().indirectMagic(this, this), (float) getTurretDamage());
+            setBeamTarget(entityliving.blockPosition().atY((int) entityliving.getEyeY()).offset(0, -2, 0));
+            this.level().playSound(null, this.blockPosition(), net.minecraft.sounds.SoundEvents.TRIDENT_THROW, net.minecraft.sounds.SoundSource.BLOCKS, 1.0F, 2.0F);
+            beamTicks = 4;
+            // KioCG end
         }
 
     }
@@ -90,6 +145,18 @@ public class EndCrystal extends Entity {
 
         nbt.putBoolean("ShowBottom", this.showsBottom());
         if (this.generatedByDragonFight) nbt.putBoolean("Paper.GeneratedByDragonFight", this.generatedByDragonFight); // Paper - Fix invulnerable end crystals
+
+        // KioCG start
+        if (this.turretCooldown != null) {
+            nbt.putInt("KioCG.TurretCooldown", this.turretCooldown);
+        }
+        if (this.turretRange != null) {
+            nbt.putDouble("KioCG.TurretRange", this.turretRange);
+        }
+        if (this.turretDamage != null) {
+            nbt.putDouble("KioCG.TurretDamage", this.turretDamage);
+        }
+        // KioCG end
     }
 
     @Override
@@ -103,6 +170,17 @@ public class EndCrystal extends Entity {
         }
         if (nbt.contains("Paper.GeneratedByDragonFight", 1)) this.generatedByDragonFight = nbt.getBoolean("Paper.GeneratedByDragonFight"); // Paper - Fix invulnerable end crystals
 
+        // KioCG start
+        if (nbt.contains("KioCG.TurretCooldown")) {
+            this.turretCooldown = nbt.getInt("KioCG.TurretCooldown");
+        }
+        if (nbt.contains("KioCG.TurretRange")) {
+            this.turretRange = nbt.getDouble("KioCG.TurretRange");
+        }
+        if (nbt.contains("KioCG.TurretDamage")) {
+            this.turretDamage = nbt.getDouble("KioCG.TurretDamage");
+        }
+        // KioCG end
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
index 676dd5331bec75407a74aea2a89e78ab72d69724..a8c797219e8c89c05ac7b22ecf954678d0e7292b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
@@ -39,6 +39,38 @@ public class CraftEnderCrystal extends CraftEntity implements EnderCrystal {
         }
     }
 
+    // KioCG start
+    @Override
+    public int getTurretCooldown() {
+        return this.getHandle().getTurretCooldown();
+    }
+
+    @Override
+    public void setTurretCooldown(@javax.annotation.Nullable Integer cooldown) {
+        this.getHandle().setTurretCooldown(cooldown);
+    }
+
+    @Override
+    public double getTurretRange() {
+        return this.getHandle().getTurretRange();
+    }
+
+    @Override
+    public void setTurretRange(@javax.annotation.Nullable Double range) {
+        this.getHandle().setTurretRange(range);
+    }
+
+    @Override
+    public double getTurretDamage() {
+        return this.getHandle().getTurretDamage();
+    }
+
+    @Override
+    public void setTurretDamage(@javax.annotation.Nullable Double damage) {
+        this.getHandle().setTurretDamage(damage);
+    }
+    // KioCG end
+
     @Override
     public EndCrystal getHandle() {
         return (EndCrystal) this.entity;
