From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 14 Mar 2024 20:52:54 +0800
Subject: [PATCH] Dont kill my pet


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 8bb51312cc6f63b5d85c2e0cd494af89e61850b1..a914bc20ddaf801b29a0e9c6a44a985ce41de8d2 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -442,6 +442,10 @@ public class GlobalConfiguration extends ConfigurationPart {
             public boolean repairNoLevelCost = true;
             public boolean renameNoLevelCost = true;
         }
+
+        public boolean dontKillMyPet = true;
+        public double petTeleportDistanceSqr = 256.0;
+        public boolean petFollowKeepChunkLoaded = true;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/TicketType.java b/src/main/java/net/minecraft/server/level/TicketType.java
index f56e5c0f53f9b52a9247b9be9265b949494fc924..b622a82986efd7a6d7cd5491f3ba2569b37f8dbe 100644
--- a/src/main/java/net/minecraft/server/level/TicketType.java
+++ b/src/main/java/net/minecraft/server/level/TicketType.java
@@ -25,6 +25,7 @@ public class TicketType<T> {
     public static final TicketType<ChunkPos> UNKNOWN = TicketType.create("unknown", Comparator.comparingLong(ChunkPos::toLong), 1);
     public static final TicketType<Unit> PLUGIN = TicketType.create("plugin", (a, b) -> 0); // CraftBukkit
     public static final TicketType<org.bukkit.plugin.Plugin> PLUGIN_TICKET = TicketType.create("plugin_ticket", (plugin1, plugin2) -> plugin1.getClass().getName().compareTo(plugin2.getClass().getName())); // CraftBukkit
+    public static final TicketType<Integer> PET_FOLLOW = TicketType.create("pet_follow", Integer::compareTo, 20 * 3); // KioCG
 
     public static <T> TicketType<T> create(String name, Comparator<T> argumentComparator) {
         return new TicketType<>(name, argumentComparator, 0L);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 0b348c24742559aa5f0250d416e4f2101522f66b..dfdbff414d2ed50b59b8d31f150fe800a7919d1c 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2495,7 +2495,15 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 }
                 // CraftBukkit end
                 this.getCombatTracker().recordDamage(damagesource, f);
-                this.setHealth(this.getHealth() - f);
+                // KioCG start
+                float health = this.getHealth() - f;
+                if (health <= 0.0F && originalDamage != Float.MAX_VALUE && !damagesource.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this instanceof TamableAnimal pet && pet.isTame()) {
+                        health = Float.MIN_VALUE;
+                    }
+                }
+                this.setHealth(health);
+                // KioCG end
                 // CraftBukkit start
                 if (!human) {
                     this.setAbsorptionAmount(this.getAbsorptionAmount() - f);
diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index 9aee0569447d351729c26eedbe24d5defe620162..fc0f5fb362253841fffeb6af192bc970e73b9b10 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -41,6 +41,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
     protected static final EntityDataAccessor<Byte> DATA_FLAGS_ID = SynchedEntityData.defineId(TamableAnimal.class, EntityDataSerializers.BYTE);
     protected static final EntityDataAccessor<Optional<UUID>> DATA_OWNERUUID_ID = SynchedEntityData.defineId(TamableAnimal.class, EntityDataSerializers.OPTIONAL_UUID);
     private boolean orderedToSit;
+    public boolean followOwner = true; // KioCG
 
     protected TamableAnimal(EntityType<? extends TamableAnimal> type, Level world) {
         super(type, world);
@@ -61,6 +62,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         nbt.putBoolean("Sitting", this.orderedToSit);
+        nbt.putBoolean("KioCG.FollowOwner", this.followOwner); // KioCG
     }
 
     @Override
@@ -87,6 +89,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
 
         this.orderedToSit = nbt.getBoolean("Sitting");
         this.setInSittingPose(this.orderedToSit, false); // Paper - Add EntityToggleSitEvent
+        if (nbt.contains("KioCG.FollowOwner")) this.followOwner = nbt.getBoolean("KioCG.FollowOwner"); // KioCG
     }
 
     @Override
@@ -258,11 +261,39 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         super.die(damageSource);
     }
 
+    // KioCG start
+    public boolean nearDeath() {
+        return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.dontKillMyPet && this.isTame() && this.getHealth() == Float.MIN_VALUE;
+    }
+
+    @Override
+    public boolean canBeSeenAsEnemy() {
+        return !this.nearDeath() && super.canBeSeenAsEnemy();
+    }
+
+    @Override
+    public boolean hurt(DamageSource source, float amount) {
+        return (amount == Float.MAX_VALUE || source.is(net.minecraft.tags.DamageTypeTags.BYPASSES_INVULNERABILITY) || !nearDeath()) && super.hurt(source, amount);
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+    }
+
+    public void addNearDeathParticles() {
+        if (this.tickCount % 5 == 0) {
+            ((net.minecraft.server.level.ServerLevel) this.level()).sendParticles(ParticleTypes.SOUL, this.getX(), this.getEyeY(), this.getZ(), 1, this.getBbWidth() / 2, this.getBbHeight() / 2, this.getBbWidth() / 2, 0.0);
+        }
+    }
+    // KioCG end
+
     public boolean isOrderedToSit() {
         return this.orderedToSit;
     }
 
     public void setOrderedToSit(boolean sitting) {
+        if (!sitting && nearDeath()) return; // KioCG
         this.orderedToSit = sitting;
     }
 
@@ -278,7 +309,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
     public boolean shouldTryTeleportToOwner() {
         LivingEntity entityliving = this.getOwner();
 
-        return entityliving != null && this.distanceToSqr((Entity) this.getOwner()) >= 144.0D;
+        return entityliving != null && this.distanceToSqr((Entity) this.getOwner()) >= io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petTeleportDistanceSqr; // KioCG
     }
 
     private void teleportToAroundBlockPos(BlockPos pos) {
@@ -333,6 +364,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
     }
 
     public final boolean unableToMoveToOwner() {
+        if (!this.followOwner) return true; // KioCG
         return this.isOrderedToSit() || this.isPassenger() || this.mayBeLeashed() || this.getOwner() != null && this.getOwner().isSpectator();
     }
 
@@ -350,6 +382,13 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
             super(TamableAnimal.this, d0);
         }
 
+        // KioCG start
+        @Override
+        public boolean shouldPanic() {
+            return !((TamableAnimal) this.mob).nearDeath() && super.shouldPanic();
+        }
+        // KioCG end
+
         @Override
         public void tick() {
             if (!TamableAnimal.this.unableToMoveToOwner() && TamableAnimal.this.shouldTryTeleportToOwner()) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java b/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
index 508d4f391fe563453d7bf6782b3082741c358006..589e925c9c20dc8847201f98e34d15fbae4de22e 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -53,7 +53,7 @@ public class StopAttackingIfTargetInvalid {
                             reason = EntityTargetEvent.TargetReason.TARGET_INVALID;
                         } else if (shouldForgetIfTargetUnreachable && StopAttackingIfTargetInvalid.isTiredOfTryingToReachTarget(entityinsentient, behaviorbuilder_b.tryGet(memoryaccessor1))) {
                             reason = EntityTargetEvent.TargetReason.FORGOT_TARGET;
-                        } else if (!entityliving.isAlive()) {
+                        } else if (!entityliving.isAlive() || entityliving instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.nearDeath()) { // KioCG
                             reason = EntityTargetEvent.TargetReason.TARGET_DIED;
                         } else if (entityliving.level() != entityinsentient.level()) {
                             reason = EntityTargetEvent.TargetReason.TARGET_OTHER_LEVEL;
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
index 15d7be9ed4a973044dd4399db46aaa244730b836..2f98152fdea3349ffe9e15ff6d40011811c36baa 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -19,6 +19,7 @@ public class FollowOwnerGoal extends Goal {
     private final float stopDistance;
     private final float startDistance;
     private float oldWaterCost;
+    private int ticketTicks; // KioCG
 
     public FollowOwnerGoal(TamableAnimal tameable, double speed, float minDistance, float maxDistance) {
         this.tamable = tameable;
@@ -40,6 +41,7 @@ public class FollowOwnerGoal extends Goal {
         } else if (this.tamable.unableToMoveToOwner()) {
             return false;
         } else if (this.tamable.distanceToSqr(livingEntity) < (double)(this.startDistance * this.startDistance)) {
+            keepChunkLoaded(); // KioCG
             return false;
         } else {
             this.owner = livingEntity;
@@ -47,6 +49,16 @@ public class FollowOwnerGoal extends Goal {
         }
     }
 
+    // KioCG start
+    private void keepChunkLoaded() {
+        if (!io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.petFollowKeepChunkLoaded) return;
+        if (this.tamable.tickCount >= ticketTicks + 20) {
+            ticketTicks = this.tamable.tickCount;
+            ((net.minecraft.server.level.ServerLevel) this.tamable.level()).getChunkSource().addRegionTicket(net.minecraft.server.level.TicketType.PET_FOLLOW, new net.minecraft.world.level.ChunkPos(this.tamable.blockPosition()), 2, this.tamable.getId()); // KioCG
+        }
+    }
+    // KioCG end
+
     @Override
     public boolean canContinueToUse() {
         return !this.navigation.isDone()
@@ -70,6 +82,7 @@ public class FollowOwnerGoal extends Goal {
 
     @Override
     public void tick() {
+        keepChunkLoaded(); // KioCG
         boolean bl = this.tamable.shouldTryTeleportToOwner();
         if (!bl) {
             if (this.tamable.distanceToSqr(this.owner) <= 16 * 16) this.tamable.getLookControl().setLookAt(this.owner, 10.0F, (float)this.tamable.getMaxHeadXRot()); // Paper - Limit pet look distance
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
index d59e0044db03ba30d5bab9a8c126b597b7659015..11428cf840d62d709139c755211802c545b68aad 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -39,7 +39,7 @@ public class MeleeAttackGoal extends Goal {
             LivingEntity livingEntity = this.mob.getTarget();
             if (livingEntity == null) {
                 return false;
-            } else if (!livingEntity.isAlive()) {
+            } else if (!livingEntity.isAlive() || livingEntity instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.nearDeath()) { // KioCG
                 return false;
             } else {
                 this.path = this.mob.getNavigation().createPath(livingEntity, 0);
@@ -53,7 +53,7 @@ public class MeleeAttackGoal extends Goal {
         LivingEntity livingEntity = this.mob.getTarget();
         if (livingEntity == null) {
             return false;
-        } else if (!livingEntity.isAlive()) {
+        } else if (!livingEntity.isAlive() || livingEntity instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal && tamableAnimal.nearDeath()) { // KioCG
             return false;
         } else {
             return !this.followingTargetEvenIfNotSeen
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
index 5633c5ee29938c0b61121c921cb82342d01329e3..19bf651cf1cb4bc0171b14b45356afb180d649ec 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
@@ -46,7 +46,7 @@ public class RangedCrossbowAttackGoal<T extends Monster & RangedAttackMob & Cros
     }
 
     private boolean isValidTarget() {
-        return this.mob.getTarget() != null && this.mob.getTarget().isAlive();
+        return this.mob.getTarget() != null && this.mob.getTarget().isAlive() && (!(this.mob.getTarget() instanceof net.minecraft.world.entity.TamableAnimal tamableAnimal) || !tamableAnimal.nearDeath()); // KioCG
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cat.java b/src/main/java/net/minecraft/world/entity/animal/Cat.java
index 797ae2cf1978e224ae5b07ffb889b1786327bf75..5163233f2c0a6dcada30ffec19e7ea5bba9ec41c 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cat.java
@@ -269,6 +269,14 @@ public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVarian
         }
 
         this.handleLieDown();
+        // KioCG start
+        if (this.nearDeath()) {
+            addNearDeathParticles();
+            if (!this.isOrderedToSit()) {
+                this.setOrderedToSit(true);
+            }
+        }
+        // KioCG end
     }
 
     private void handleLieDown() {
@@ -409,6 +417,18 @@ public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVarian
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
+                    // KioCG start
+                } else if (itemstack.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) {
+                    this.playSound(SoundEvents.CAT_EAT, 1.0F, 1.0F);
+
+                    itemstack.consume(1, player);
+                    this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                    if (player.isCreative() || !this.isInvulnerable()) {
+                        this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                    }
+
+                    return InteractionResult.sidedSuccess(this.level().isClientSide());
+                    // KioCG end
                 }
 
                 enuminteractionresult = super.mobInteract(player, hand);
diff --git a/src/main/java/net/minecraft/world/entity/animal/Parrot.java b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
index b87dfc9e103e78a3a576c8ef2d8bc455c75278d7..3169583c4624cc832a458b37daf57605bb235df6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Parrot.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
@@ -201,6 +201,19 @@ public class Parrot extends ShoulderRidingEntity implements VariantHolder<Parrot
         this.calculateFlapping();
     }
 
+    // KioCG start
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.nearDeath()) {
+            addNearDeathParticles();
+            if (!this.isOrderedToSit()) {
+                this.setOrderedToSit(true);
+            }
+        }
+    }
+    // KioCG end
+
     @Override
     public void setRecordPlayingNearby(BlockPos songPosition, boolean playing) {
         this.jukebox = songPosition;
@@ -271,7 +284,7 @@ public class Parrot extends ShoulderRidingEntity implements VariantHolder<Parrot
             }
 
             return InteractionResult.sidedSuccess(this.level().isClientSide);
-        } else if (!itemstack.is(ItemTags.PARROT_POISONOUS_FOOD)) {
+        } else if (!itemstack.is(ItemTags.PARROT_POISONOUS_FOOD) && !itemstack.is(net.minecraft.world.item.Items.POISONOUS_POTATO)) { // KioCG
             if (!this.isFlying() && this.isTame() && this.isOwnedBy(player)) {
                 if (!this.level().isClientSide) {
                     this.setOrderedToSit(!this.isOrderedToSit());
diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index 0c51106537876f704e382a9afbcd2d7e083cb480..b573b24fc014c44b2117d71214151606517a4394 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -335,6 +335,16 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
                 }
             }
 
+            // KioCG start
+            if (this.nearDeath()) {
+                addNearDeathParticles();
+                if (!this.isOrderedToSit()) {
+                    this.navigation.stop();
+                    this.setTarget((LivingEntity) null);
+                    this.setOrderedToSit(true);
+                }
+            }
+            // KioCG end
         }
     }
 
@@ -470,6 +480,16 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
 
                 this.heal(2.0F * f, EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
                 return InteractionResult.sidedSuccess(this.level().isClientSide());
+                // KioCG start
+            } else if (itemstack.is(Items.POISONOUS_POTATO)) {
+                itemstack.consume(1, player);
+                this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
+                if (player.isCreative() || !this.isInvulnerable()) {
+                    this.hurt(this.damageSources().playerAttack(player), Float.MAX_VALUE);
+                }
+
+                return InteractionResult.sidedSuccess(this.level().isClientSide());
+                // KioCG end
             } else {
                 if (item instanceof DyeItem) {
                     DyeItem itemdye = (DyeItem) item;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java
index cedb8e67e208cdf954d052a4f0a100c1c07a962b..22b6bc3aedc79ff8f846009bb562f7c2dd3844ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTameableAnimal.java
@@ -81,6 +81,16 @@ public class CraftTameableAnimal extends CraftAnimals implements Tameable, Creat
         this.getHandle().setOrderedToSit(sitting);
     }
 
+    // KioCG start
+    public boolean isFollowOwner() {
+        return this.getHandle().followOwner;
+    }
+
+    public void setFollowOwner(boolean followOwner) {
+        this.getHandle().followOwner = followOwner;
+    }
+    // KioCG end
+
     @Override
     public String toString() {
         return this.getClass().getSimpleName() + "{owner=" + this.getOwner() + ",tamed=" + this.isTamed() + "}";
