From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Mar 2024 14:39:32 +0800
Subject: [PATCH] Sleep any time


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 5151a50f37dbc15d3180f6a05469f76035983b58..da5f9ee0478eba4cf0ec2c1b1c31582dc9b6af47 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -608,6 +608,9 @@ public class WorldConfiguration extends ConfigurationPart {
                 public boolean babyZombieJockey = true;
             }
         }
+
+        public boolean sleepAnyTime = true;
+        public double sleepHealingRegenRate = 20.0;
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 45c9d6fe59d3167cd67b9faa9f1b0a00995df3f9..e762c3ccda1f3200bb3b9cd5547caaec4add7287 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -683,6 +683,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         int i = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
         long j;
 
+        if (!this.paperConfig().kiocgConfig.sleepAnyTime) { // KioCG
         if (this.sleepStatus.areEnoughSleeping(i) && this.sleepStatus.areEnoughDeepSleeping(i, this.players)) {
             // CraftBukkit start
             j = this.levelData.getDayTime() + 24000L;
@@ -702,6 +703,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
                 this.resetWeatherCycle();
             }
         }
+        } // KioCG
 
         this.updateSkyBrightness();
         if (flag) {
@@ -1046,6 +1048,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     }
 
     private void announceSleepStatus() {
+        if (this.paperConfig().kiocgConfig.sleepAnyTime) return; // KioCG
         if (this.canSleepThroughNights()) {
             if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int i = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 151e259820185947056a8560f93e151a46d2451d..7287a23cff32e3458a626197eb8ef0cf369a98a7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1513,7 +1513,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
                 return Either.left(net.minecraft.world.entity.player.Player.BedSleepingProblem.OBSTRUCTED);
             } else {
                 this.setRespawnPosition(this.level().dimension(), blockposition, this.getYRot(), false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED); // Paper - Add PlayerSetSpawnEvent
-                if (this.level().isDay()) {
+                if (this.level().isDay() && !this.level().paperConfig().kiocgConfig.sleepAnyTime) { // KioCG
                     return Either.left(net.minecraft.world.entity.player.Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
                 } else {
                     if (!this.isCreative()) {
@@ -1563,7 +1563,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
                         CriteriaTriggers.SLEPT_IN_BED.trigger(this);
                     });
 
-                    if (!this.serverLevel().canSleepThroughNights()) {
+                    if (!this.serverLevel().canSleepThroughNights() && !this.level().paperConfig().kiocgConfig.sleepAnyTime) { // KioCG
                         this.displayClientMessage(Component.translatable("sleep.not_possible"), true);
                     }
 
@@ -2543,7 +2543,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
             // Paper end - Add PlayerSetSpawnEvent
 
             if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper - Add PlayerSetSpawnEvent
-                this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification())); // Paper - Add PlayerSetSpawnEvent
+                this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification()), this.level().paperConfig().kiocgConfig.sleepAnyTime); // Paper - Add PlayerSetSpawnEvent // KioCG
             }
 
             this.respawnPosition = pos;
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 97404a624465a369737eff635718f0f65d8e7866..d7decf1a0fbf641dccce7b29b26b5e7d8fbd9735 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -276,7 +276,7 @@ public abstract class Player extends LivingEntity {
                 this.sleepCounter = 100;
             }
 
-            if (!this.level().isClientSide && this.level().isDay()) {
+            if (!this.level().isClientSide && this.level().isDay() && !this.level().paperConfig().kiocgConfig.sleepAnyTime) { // KioCG
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
diff --git a/src/main/java/net/minecraft/world/food/FoodData.java b/src/main/java/net/minecraft/world/food/FoodData.java
index b45e4a9793ddcc1e6d090c6a18f99e3b9bc35635..8e65c2de9d71869b0c3945f16eeb0e5a4f0dff60 100644
--- a/src/main/java/net/minecraft/world/food/FoodData.java
+++ b/src/main/java/net/minecraft/world/food/FoodData.java
@@ -84,6 +84,27 @@ public class FoodData {
 
         boolean flag = player.level().getGameRules().getBoolean(GameRules.RULE_NATURAL_REGENERATION);
 
+        // KioCG start
+        if (player.level().paperConfig().kiocgConfig.sleepHealingRegenRate >= 0 && this.foodLevel > 6 && player.isSleeping() && player.isHurt()) {
+            ++this.tickTimer;
+            if (this.tickTimer >= player.level().paperConfig().kiocgConfig.sleepHealingRegenRate) {
+                final float health = player.getHealth();
+                player.heal(1.0F);
+                if (player.getHealth() > health) {
+                    ((net.minecraft.server.level.ServerLevel) player.level()).sendParticles(
+                            (ServerPlayer) player,
+                            net.minecraft.core.particles.ParticleTypes.HEART,
+                            player.getX(), player.getY(), player.getZ(),
+                            5,
+                            0.3, 0.3, 0.3,
+                            1.0,
+                            false);
+                }
+
+                this.tickTimer = 0;
+            }
+        } else
+        // KioCG end
         if (flag && this.saturationLevel > 0.0F && player.isHurt() && this.foodLevel >= 20) {
             ++this.tickTimer;
             if (this.tickTimer >= this.saturatedRegenRate) { // CraftBukkit
