From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 2 Apr 2024 19:33:24 +0800
Subject: [PATCH] Breeding animals


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 604bbae43d54a4402b65bd1e0612d1c269bac919..00c9f2006078cd5cc7650ec8608ecb32ce9efcbf 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -596,6 +596,11 @@ public class WorldConfiguration extends ConfigurationPart {
             public double turretDamageDefault = 1.0;
         }
 
+        public Animal animal;
+        public class Animal extends ConfigurationPart {
+            public boolean animalCanBreed = false;
+        }
+
         public boolean betterMobGriefingSetting = true;
         public double dirtPathSpeedBoost = 0.02;
         public double saplingsWitheredChance = 0.025;
diff --git a/src/main/java/net/minecraft/world/entity/AgeableMob.java b/src/main/java/net/minecraft/world/entity/AgeableMob.java
index 8dcdc664470fcac00c203b9499ea0a45df7d86ef..1620414239e4585685e73199372390c457b56721 100644
--- a/src/main/java/net/minecraft/world/entity/AgeableMob.java
+++ b/src/main/java/net/minecraft/world/entity/AgeableMob.java
@@ -21,6 +21,19 @@ public abstract class AgeableMob extends PathfinderMob {
     protected int forcedAge;
     protected int forcedAgeTimer;
     public boolean ageLocked; // CraftBukkit
+    // KioCG start
+    private int feedingDay;
+    private int produceDay;
+    private int totalFeedingDays;
+    private int numberOfProducedToday;
+
+    public int getTotalFeedingDays() {
+        return this.totalFeedingDays;
+    }
+    public int getNumberOfProducedToday() {
+        return this.numberOfProducedToday;
+    }
+    // KioCG end
 
     protected AgeableMob(EntityType<? extends AgeableMob> type, Level world) {
         super(type, world);
@@ -131,6 +144,12 @@ public abstract class AgeableMob extends PathfinderMob {
         nbt.putInt("Age", this.getAge());
         nbt.putInt("ForcedAge", this.forcedAge);
         nbt.putBoolean("AgeLocked", this.ageLocked); // CraftBukkit
+        // KioCG start
+        nbt.putInt("KioCG.FeedingDay", this.feedingDay);
+        nbt.putInt("KioCG.ProduceDay", this.produceDay);
+        nbt.putInt("KioCG.TotalFeedingDays", this.totalFeedingDays);
+        nbt.putInt("KioCG.NumberOfProducedToday", this.numberOfProducedToday);
+        // KioCG end
     }
 
     @Override
@@ -139,6 +158,20 @@ public abstract class AgeableMob extends PathfinderMob {
         this.setAge(nbt.getInt("Age"));
         this.forcedAge = nbt.getInt("ForcedAge");
         this.ageLocked = nbt.getBoolean("AgeLocked"); // CraftBukkit
+        // KioCG start
+        if (nbt.contains("KioCG.FeedingDay", net.minecraft.nbt.Tag.TAG_INT)) {
+            this.feedingDay = nbt.getInt("KioCG.FeedingDay");
+        }
+        if (nbt.contains("KioCG.ProduceDay", net.minecraft.nbt.Tag.TAG_INT)) {
+            this.produceDay = nbt.getInt("KioCG.ProduceDay");
+        }
+        if (nbt.contains("KioCG.TotalFeedingDays", net.minecraft.nbt.Tag.TAG_INT)) {
+            this.totalFeedingDays = nbt.getInt("KioCG.TotalFeedingDays");
+        }
+        if (nbt.contains("KioCG.NumberOfProducedToday", net.minecraft.nbt.Tag.TAG_INT)) {
+            this.numberOfProducedToday = nbt.getInt("KioCG.NumberOfProducedToday");
+        }
+        // KioCG end
     }
 
     @Override
@@ -171,6 +204,8 @@ public abstract class AgeableMob extends PathfinderMob {
                 --i;
                 this.setAge(i);
             }
+
+            tickProduce(); // KioCG
         }
 
     }
@@ -204,6 +239,50 @@ public abstract class AgeableMob extends PathfinderMob {
         return (int) ((float) (breedingAge / 20) * 0.1F);
     }
 
+    // KioCG start
+    public void tickProduce() {
+        int days = (int) this.level().getDays();
+
+        if (this.produceDay != 0 && days > this.produceDay) {
+            this.produceDay = 0;
+            ++this.totalFeedingDays;
+            this.numberOfProducedToday = getNumberOfProduced();
+        }
+
+        if (days < this.feedingDay) {
+            this.feedingDay = days;
+            if (this.produceDay != 0) {
+                this.produceDay = this.feedingDay;
+            }
+        }
+    }
+
+    public int getNumberOfProduced() {
+        return 1;
+    }
+
+    public boolean canFeed() {
+        return (int) this.level().getDays() > this.feedingDay;
+    }
+
+    public void feedingFeed() {
+        this.feedingDay = (int) this.level().getDays();
+        this.produceDay = this.feedingDay;
+    }
+
+    public boolean canProduce() {
+        return this.numberOfProducedToday > 0;
+    }
+
+    public void produce() {
+        produce(1);
+    }
+
+    public void produce(int amount) {
+        this.numberOfProducedToday -= amount;
+    }
+    // KioCG end
+
     public static class AgeableMobGroupData implements SpawnGroupData {
 
         private int groupSize;
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java b/src/main/java/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
index 06662325df5b119ad049e04572ff5c5aaf3ef073..8c43b278dcb49276fabdded164a1d87b4cb14fea 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
@@ -25,6 +25,7 @@ public class AnimalMakeLove extends Behavior<Animal> {
 
     @Override
     protected boolean checkExtraStartConditions(ServerLevel world, Animal entity) {
+        if (!entity.level().paperConfig().kiocgConfig.animal.animalCanBreed) return false; // KioCG
         return entity.isInLove() && this.findValidBreedPartner(entity).isPresent();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/BreedGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/BreedGoal.java
index dbd01cb7828cbe65613dc07a2d0107c83c5fdf98..b630356324a150f1b34128e53992649c8d6c43bb 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/BreedGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/BreedGoal.java
@@ -32,6 +32,7 @@ public class BreedGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (!this.animal.level().paperConfig().kiocgConfig.animal.animalCanBreed) return false; // KioCG
         if (!this.animal.isInLove()) {
             return false;
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/animal/Animal.java b/src/main/java/net/minecraft/world/entity/animal/Animal.java
index 081d1e38b7b1f286e138b0981aaa760e58761215..8fe739cb62a738d4e0a496da6f1dc44104d50f77 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Animal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Animal.java
@@ -151,7 +151,7 @@ public abstract class Animal extends AgeableMob {
         if (this.isFood(itemstack)) {
             int i = this.getAge();
 
-            if (!this.level().isClientSide && i == 0 && this.canFallInLove()) {
+            if (!this.level().isClientSide && this.canFeed()) { // KioCG
                 final ItemStack breedCopy = itemstack.copy(); // Paper - Fix EntityBreedEvent copying
                 this.usePlayerItem(player, hand, itemstack);
                 this.setInLove(player, breedCopy); // Paper - Fix EntityBreedEvent copying
@@ -203,6 +203,7 @@ public abstract class Animal extends AgeableMob {
             this.loveCause = player.getUUID();
         }
         // Paper - Fix EntityBreedEvent copying; set breed item in better place
+        this.feedingFeed(); // KioCG
 
         this.level().broadcastEntityEvent(this, (byte) 18);
     }
diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index 5e7e6526e2750d508a870dfbdbb46e520d201796..c303c563bdaaa7cba6b7de11471fe633b2596ea6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -51,7 +51,7 @@ public class Chicken extends Animal {
 
     public Chicken(EntityType<? extends Chicken> type, Level world) {
         super(type, world);
-        this.eggTime = this.random.nextInt(6000) + 6000;
+        this.eggTime = this.random.nextInt(20 * 10); // KioCG
         this.setPathfindingMalus(BlockPathTypes.WATER, 0.0F);
     }
 
@@ -95,17 +95,25 @@ public class Chicken extends Animal {
         }
 
         this.flap += this.flapping * 2.0F;
-        if (!this.level().isClientSide && this.isAlive() && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) {
+        if (!this.level().isClientSide && this.isAlive() && !this.isBaby() && this.canProduce() && !this.isChickenJockey() && --this.eggTime <= 0) { // KioCG
             this.playSound(SoundEvents.CHICKEN_EGG, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
             this.forceDrops = true; // CraftBukkit
-            this.spawnAtLocation((ItemLike) Items.EGG);
+            this.spawnAtLocation((ItemLike) Items.EGG, this.getNumberOfProducedToday()); // KioCG
             this.forceDrops = false; // CraftBukkit
             this.gameEvent(GameEvent.ENTITY_PLACE);
-            this.eggTime = this.random.nextInt(6000) + 6000;
+            this.eggTime = this.random.nextInt(20 * 10); // KioCG
+            this.produce(this.getNumberOfProducedToday()); // KioCG
         }
 
     }
 
+    // KioCG start
+    @Override
+    public int getNumberOfProduced() {
+        return Math.min(1 + this.getTotalFeedingDays() / 10, 5);
+    }
+    // KioCG end
+
     @Override
     protected boolean isFlapping() {
         return this.flyDist > this.nextFlap;
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index 3cdd9f379c7e2d46ea47c9ef55b121c93ec0bb4a..8c49a4c49a7360d9c6a9188f1530d891be5e053c 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -90,7 +90,7 @@ public class Cow extends Animal {
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
 
-        if (itemstack.is(Items.BUCKET) && !this.isBaby()) {
+        if (itemstack.is(Items.BUCKET) && !this.isBaby() && this.canProduce()) { // KioCG
             // CraftBukkit start - Got milk?
             PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerLevel) player.level(), player, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET, hand);
 
@@ -102,6 +102,7 @@ public class Cow extends Animal {
             player.playSound(SoundEvents.COW_MILK, 1.0F, 1.0F);
             ItemStack itemstack1 = ItemUtils.createFilledResult(itemstack, player, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
 
+            this.produce(); // KioCG
             player.setItemInHand(hand, itemstack1);
             return InteractionResult.sidedSuccess(this.level().isClientSide);
         } else {
@@ -109,6 +110,13 @@ public class Cow extends Animal {
         }
     }
 
+    // KioCG start
+    @Override
+    public int getNumberOfProduced() {
+        return Math.min(1 + this.getTotalFeedingDays() / 30, 3);
+    }
+    // KioCG end
+
     @Nullable
     @Override
     public Cow getBreedOffspring(ServerLevel world, AgeableMob entity) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
index 161c128d27f50f145f88142191f1a5c93649ea65..759cb392c342a892526bea4fbd276dff0c1bf172 100644
--- a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
@@ -95,7 +95,7 @@ public class MushroomCow extends Cow implements Shearable, VariantHolder<Mushroo
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
 
-        if (itemstack.is(Items.BOWL) && !this.isBaby()) {
+        if (itemstack.is(Items.BOWL) && !this.isBaby() && this.canProduce()) { // KioCG
             boolean flag = false;
             ItemStack itemstack1;
 
@@ -110,6 +110,7 @@ public class MushroomCow extends Cow implements Shearable, VariantHolder<Mushroo
 
             ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, player, itemstack1, false);
 
+            this.produce(); // KioCG
             player.setItemInHand(hand, itemstack2);
             SoundEvent soundeffect;
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/Panda.java b/src/main/java/net/minecraft/world/entity/animal/Panda.java
index d683c49fdf2d1e5b0f2620641f9c241e82f96825..b8a5158a6d55164d56d2050e515959baf4699a1b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Panda.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Panda.java
@@ -649,7 +649,7 @@ public class Panda extends Animal {
             if (this.isBaby()) {
                 this.usePlayerItem(player, hand, itemstack);
                 this.ageUp((int) ((float) (-this.getAge() / 20) * 0.1F), true);
-            } else if (!this.level().isClientSide && this.getAge() == 0 && this.canFallInLove()) {
+            } else if (!this.level().isClientSide && this.canFeed()) { // KioCG
                 final ItemStack breedCopy = itemstack.copy(); // Paper - Fix EntityBreedEvent copying
                 this.usePlayerItem(player, hand, itemstack);
                 this.setInLove(player, breedCopy); // Paper - Fix EntityBreedEvent copying
diff --git a/src/main/java/net/minecraft/world/entity/animal/Sheep.java b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
index 1d80678f7e8f658e43616f0baf723f096a99122a..82c1adc6b6538894d78c5d46f80d03d03976f396 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Sheep.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
@@ -148,6 +148,16 @@ public class Sheep extends Animal implements Shearable {
         super.aiStep();
     }
 
+    // KioCG start
+    @Override
+    public void tickProduce() {
+        super.tickProduce();
+        if (this.canProduce() && this.isSheared()) {
+            this.ate();
+        }
+    }
+    // KioCG end
+
     public static AttributeSupplier.Builder createAttributes() {
         return Mob.createMobAttributes().add(Attributes.MAX_HEALTH, 8.0D).add(Attributes.MOVEMENT_SPEED, 0.23000000417232513D);
     }
@@ -251,7 +261,7 @@ public class Sheep extends Animal implements Shearable {
         ItemStack itemstack = player.getItemInHand(hand);
 
         if (itemstack.is(Items.SHEARS)) {
-            if (!this.level().isClientSide && this.readyForShearing()) {
+            if (!this.level().isClientSide && this.readyForShearing() && this.canProduce()) { // KioCG
                 // CraftBukkit start
                 // Paper start - custom shear drops
                 java.util.List<ItemStack> drops = this.generateDefaultDrops();
@@ -264,6 +274,7 @@ public class Sheep extends Animal implements Shearable {
                 }
                 // Paper end - custom shear drops
                 // CraftBukkit end
+                this.produce(); // KioCG
                 this.shear(SoundSource.PLAYERS, drops); // Paper
                 this.gameEvent(GameEvent.SHEAR, player);
                 itemstack.hurtAndBreak(1, player, (entityhuman1) -> {
@@ -278,6 +289,13 @@ public class Sheep extends Animal implements Shearable {
         }
     }
 
+    // KioCG start
+    @Override
+    public int getNumberOfProduced() {
+        return Math.min(1 + this.getTotalFeedingDays() / 15, 5);
+    }
+    // KioCG end
+
     @Override
     public void shear(SoundSource shearedSoundCategory) {
         // Paper start - custom shear drops
@@ -298,6 +316,7 @@ public class Sheep extends Animal implements Shearable {
     public void shear(SoundSource shearedSoundCategory, java.util.List<ItemStack> drops) {
         // Paper end - custom shear drops
         this.level().playSound((Player) null, (Entity) this, SoundEvents.SHEEP_SHEAR, shearedSoundCategory, 1.0F, 1.0F);
+        if (!this.canProduce()) // KioCG
         this.setSheared(true);
         int i = 1 + this.random.nextInt(3);
 
@@ -404,6 +423,7 @@ public class Sheep extends Animal implements Shearable {
         if (event.isCancelled()) return;
         // CraftBukkit end
         super.ate();
+        if (this.canProduce()) // KioCG
         this.setSheared(false);
         if (this.isBaby()) {
             this.ageUp(60);
diff --git a/src/main/java/net/minecraft/world/entity/animal/Turtle.java b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
index 2eb099957a3d0bae3339ff4edbab103fb348abed..980c7067d8af011e0c77e9d917f6dd3994a98298 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Turtle.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
@@ -186,10 +186,12 @@ public class Turtle extends Animal {
         return pos.getY() < world.getSeaLevel() + 4 && TurtleEggBlock.onSand(world, pos) && isBrightEnoughToSpawn(world, pos);
     }
 
+    private Turtle.TurtleBreedGoal turtleBreedGoal; // KioCG
+
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(0, new Turtle.TurtlePanicGoal(this, 1.2D));
-        this.goalSelector.addGoal(1, new Turtle.TurtleBreedGoal(this, 1.0D));
+        this.goalSelector.addGoal(1, turtleBreedGoal = new Turtle.TurtleBreedGoal(this, 1.0D)); // KioCG
         this.goalSelector.addGoal(1, new Turtle.TurtleLayEggGoal(this, 1.0D));
         this.goalSelector.addGoal(2, new TemptGoal(this, 1.1D, Turtle.FOOD_ITEMS, false));
         this.goalSelector.addGoal(3, new Turtle.TurtleGoToWaterGoal(this, 1.0D));
@@ -303,6 +305,17 @@ public class Turtle extends Animal {
 
     }
 
+    // KioCG start
+    @Override
+    public void tickProduce() {
+        super.tickProduce();
+        if (this.canProduce()) {
+            this.produce();
+            turtleBreedGoal.breed();
+        }
+    }
+    // KioCG end
+
     @Override
     protected void ageBoundaryReached() {
         super.ageBoundaryReached();
diff --git a/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java b/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
index 84ae19bf9bddd2b6ee4737577d8836d59be028c2..5401b6411be324d1d5efe99b2bfdba02b8036d92 100644
--- a/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
@@ -395,7 +395,7 @@ public class Camel extends AbstractHorse implements PlayerRideableJumping, Saddl
                 this.heal(2.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING); // Paper - Add missing regain reason
             }
 
-            boolean flag1 = this.isTamed() && this.getAge() == 0 && this.canFallInLove();
+            boolean flag1 = this.isTamed() && this.canFeed(); // KioCG
 
             if (flag1) {
                 this.setInLove(player, item.copy()); // Paper - Fix EntityBreedEvent copying
diff --git a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
index 295769d039f2a1e4f48912a60f9dbe267d8992c1..b4f309f7902727591ea7374fcbcb22fe8c0908d6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
@@ -179,6 +179,17 @@ public class Frog extends Animal implements VariantHolder<FrogVariant> {
         super.tick();
     }
 
+    // KioCG start
+    @Override
+    public void tickProduce() {
+        super.tickProduce();
+        if (this.canProduce()) {
+            this.produce();
+            this.spawnChildFromBreeding((ServerLevel) this.level(), this);
+        }
+    }
+    // KioCG end
+
     @Override
     public void onSyncedDataUpdated(EntityDataAccessor<?> data) {
         if (DATA_POSE.equals(data)) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
index 5d247ac38fe8a61603b3d934f3000bcda773142b..0de7394563a27a5c8b745b568635c23efb5a0f3a 100644
--- a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
@@ -225,7 +225,7 @@ public class Goat extends Animal {
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
 
-        if (itemstack.is(Items.BUCKET) && !this.isBaby()) {
+        if (itemstack.is(Items.BUCKET) && !this.isBaby() && this.canProduce()) { // KioCG
             // CraftBukkit start - Got milk?
             PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerLevel) player.level(), player, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET, hand);
 
@@ -236,6 +236,7 @@ public class Goat extends Animal {
             player.playSound(this.getMilkingSound(), 1.0F, 1.0F);
             ItemStack itemstack1 = ItemUtils.createFilledResult(itemstack, player, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
 
+            this.produce(); // KioCG
             player.setItemInHand(hand, itemstack1);
             return InteractionResult.sidedSuccess(this.level().isClientSide);
         } else {
@@ -250,6 +251,13 @@ public class Goat extends Animal {
         }
     }
 
+    // KioCG start
+    @Override
+    public int getNumberOfProduced() {
+        return Math.min(1 + this.getTotalFeedingDays() / 25, 3);
+    }
+    // KioCG end
+
     @Override
     public SpawnGroupData finalizeSpawn(ServerLevelAccessor world, DifficultyInstance difficulty, MobSpawnType spawnReason, @Nullable SpawnGroupData entityData, @Nullable CompoundTag entityNbt) {
         RandomSource randomsource = world.getRandom();
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 815eb15086976b8f9e03bf8182d9ed50aec14720..00d9f8e682b67c23315cb77d950f4f5068ecde4b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -511,7 +511,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
             f = 4.0F;
             short0 = 60;
             b0 = 5;
-            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
+            if (!this.level().isClientSide && this.isTamed() && this.canFeed()) { // KioCG
                 flag = true;
                 this.setInLove(player, item.copy()); // Paper - Fix EntityBreedEvent copying
             }
@@ -519,7 +519,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
             f = 10.0F;
             short0 = 240;
             b0 = 10;
-            if (!this.level().isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
+            if (!this.level().isClientSide && this.isTamed() && this.canFeed()) { // KioCG
                 flag = true;
                 this.setInLove(player, item.copy()); // Paper - Fix EntityBreedEvent copying
             }
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
index 6623674136b0f865d5b3d7a10d3bf05793b82f87..f7e92392b0a578d890269b0ad8562bd28d168720 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
@@ -189,7 +189,7 @@ public class Llama extends AbstractChestedHorse implements VariantHolder<Llama.V
             b0 = 90;
             b1 = 6;
             f = 10.0F;
-            if (this.isTamed() && this.getAge() == 0 && this.canFallInLove()) {
+            if (this.isTamed() && this.canFeed()) { // KioCG
                 flag = true;
                 this.setInLove(player, item.copy()); // Paper - Fix EntityBreedEvent copying
             }
diff --git a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index 0a5b953bd8c0c7f181da4090b950e9e6524b6d61..33e50f1a3a861ef3ecb905c03bbb1b8add992fc7 100644
--- a/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/src/main/java/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -498,6 +498,17 @@ public class Sniffer extends Animal {
         super.customServerAiStep();
     }
 
+    // KioCG start
+    @Override
+    public void tickProduce() {
+        super.tickProduce();
+        if (this.canProduce()) {
+            this.produce();
+            this.spawnChildFromBreeding((ServerLevel) this.level(), this);
+        }
+    }
+    // KioCG end
+
     @Override
     protected void sendDebugPackets() {
         super.sendDebugPackets();
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 24044795d8e0f1fb15a4f2f5401f44897092f2a3..48e31cb123caa3c3f1dce23b9a73e547c2ac86b0 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -746,6 +746,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
 
     @Override
     public boolean canBreed() {
+        if (!this.level().paperConfig().kiocgConfig.animal.animalCanBreed) return false; // KioCG
         return this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0;
     }
 
diff --git a/src/main/java/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java b/src/main/java/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java
index 3fb1e558c3510243c94981211f9a0e5e0ef1895b..3fd790d924a3c9ef90849dbef2b901fa4f5c055e 100644
--- a/src/main/java/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java
+++ b/src/main/java/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java
@@ -54,6 +54,7 @@ public class LootingEnchantFunction extends LootItemConditionalFunction {
     @Override
     public ItemStack run(ItemStack stack, LootContext context) {
         Entity entity = (Entity) context.getParamOrNull(LootContextParams.KILLER_ENTITY);
+        int i0 = 0; // KioCG
 
         if (entity instanceof LivingEntity) {
             int i = EnchantmentHelper.getMobLooting((LivingEntity) entity);
@@ -62,6 +63,13 @@ public class LootingEnchantFunction extends LootItemConditionalFunction {
                 i = context.getParamOrNull(LootContextParams.LOOTING_MOD);
             }
             // CraftBukkit end
+            // KioCG start
+            i0 = i;
+        }
+        {
+            int i = i0;
+            if (context.getParamOrNull(LootContextParams.THIS_ENTITY) instanceof net.minecraft.world.entity.AgeableMob ageableMob) i += ageableMob.getTotalFeedingDays() * 4;
+            // KioCG end
 
             if (i <= 0) { // CraftBukkit - account for possible negative looting values from Bukkit
                 return stack;
diff --git a/src/main/java/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java b/src/main/java/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
index 2c9161e9f798a1e3770afbed26737bf9a7ceb2d5..8c24138f4d2451de92ed025fefe649c006a4704e 100644
--- a/src/main/java/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
+++ b/src/main/java/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
@@ -34,6 +34,7 @@ public record LootItemRandomChanceWithLootingCondition(float percent, float loot
         if (entity instanceof LivingEntity) {
             i = EnchantmentHelper.getMobLooting((LivingEntity) entity);
         }
+        if (loottableinfo.getParamOrNull(LootContextParams.THIS_ENTITY) instanceof net.minecraft.world.entity.AgeableMob ageableMob) i += ageableMob.getTotalFeedingDays() * 4; // KioCG
         // CraftBukkit start - only use lootingModifier if set by Bukkit
         if (loottableinfo.hasParam(LootContextParams.LOOTING_MOD)) {
             i = loottableinfo.getParamOrNull(LootContextParams.LOOTING_MOD);
