From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 18 Mar 2024 12:04:58 +0800
Subject: [PATCH] End crystal turret


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 29f1a8e24e643b6f68aa381d2ad3040fb452d083..683b25cb32d79041816dd0b045660e8aecf58158 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -634,6 +634,11 @@ public class WorldConfiguration extends ConfigurationPart {
                 public double zombiesSkinnedChance = 0.25;
                 public boolean babyZombieJockey = true;
             }
+
+            public boolean turretEnable = true;
+            public int turretCooldownDefault = 100;
+            public double turretRangeDefault = 8.0;
+            public double turretDamageDefault = 1.0;
         }
 
         public boolean sleepAnyTime = true;
diff --git a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
index a33d89fe9ca9e343edab8bb1cc88c54130ddb4a7..dd1ce5b2a3357c27c6a777a4c1b74202625ce55e 100644
--- a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -31,6 +31,38 @@ public class EndCrystal extends Entity {
     private static final EntityDataAccessor<Boolean> DATA_SHOW_BOTTOM = SynchedEntityData.defineId(EndCrystal.class, EntityDataSerializers.BOOLEAN);
     public int time;
     public boolean generatedByDragonFight = false; // Paper - Fix invulnerable end crystals
+    // KioCG start
+    @Nullable
+    private Integer turretCooldown = null;
+    public int getTurretCooldown() {
+        return java.util.Objects.requireNonNullElse(this.turretCooldown, this.level().paperConfig().kiocgConfig.entity.turretCooldownDefault);
+    }
+    public void setTurretCooldown(@Nullable Integer cooldown) {
+        this.turretCooldown = cooldown;
+    }
+
+    @Nullable
+    private Double turretRange = null;
+    public double getTurretRange() {
+        return java.util.Objects.requireNonNullElse(this.turretRange, this.level().paperConfig().kiocgConfig.entity.turretRangeDefault);
+    }
+    public void setTurretRange(@Nullable Double range) {
+        this.turretRange = range;
+    }
+
+    @Nullable
+    private Double turretDamage = null;
+    public double getTurretDamage() {
+        return java.util.Objects.requireNonNullElse(this.turretDamage, this.level().paperConfig().kiocgConfig.entity.turretDamageDefault);
+    }
+    public void setTurretDamage(@Nullable Double damage) {
+        this.turretDamage = damage;
+    }
+
+    private int beamTicks = 0;
+    @Nullable
+    private net.minecraft.world.entity.LivingEntity target;
+    // KioCG end
 
     public EndCrystal(EntityType<? extends EndCrystal> type, Level world) {
         super(type, world);
@@ -80,6 +112,47 @@ public class EndCrystal extends Entity {
                 }
             }
             // Paper end - Fix invulnerable end crystals
+            // KioCG start
+            if (this.level().paperConfig().kiocgConfig.entity.turretEnable && !this.generatedByDragonFight) {
+                if (--beamTicks <= 0) {
+                    setBeamTarget(null);
+                }
+
+                if (getTurretCooldown() < 0) {
+                    this.target = null;
+                } else if (getTurretCooldown() == 0 || this.time % getTurretCooldown() == 0) { // 使炮塔间的发射间隔错开
+                    refreshTarget();
+
+                    if (this.target != null) {
+                        this.target.hurt(this.level().damageSources().indirectMagic(this, this), (float) getTurretDamage());
+                        setBeamTarget(this.target.blockPosition().atY((int) this.target.getEyeY()).offset(0, -2, 0));
+                        this.level().playSound(null, this.blockPosition(), net.minecraft.sounds.SoundEvents.TRIDENT_THROW.value(), net.minecraft.sounds.SoundSource.BLOCKS, 1.0F, 2.0F);
+                        this.beamTicks = 4;
+                    }
+                }
+            }
+        }
+
+    }
+
+    private void refreshTarget() {
+        final net.minecraft.world.phys.AABB aabb = this.getBoundingBox().inflate(getTurretRange(), getTurretRange() / 2.0, getTurretRange());
+
+        if (this.target != null) {
+            if (!this.target.isAlive() || !this.target.getBoundingBox().intersects(aabb) || !this.target.hasLineOfSight(this)) {
+                this.target = null;
+            }
+        }
+
+        if (this.target == null) {
+            java.util.List<net.minecraft.world.entity.monster.Monster> list1 = this.level().getEntitiesOfClass(net.minecraft.world.entity.monster.Monster.class, aabb, (entityliving1) -> {
+                return entityliving1.isAlive() && entityliving1.hasLineOfSight(this);
+            });
+
+            if (!list1.isEmpty()) {
+                this.target = list1.get(this.level().random.nextInt(list1.size()));
+            }
+            // KioCG end
         }
 
     }
@@ -92,6 +165,18 @@ public class EndCrystal extends Entity {
 
         nbt.putBoolean("ShowBottom", this.showsBottom());
         if (this.generatedByDragonFight) nbt.putBoolean("Paper.GeneratedByDragonFight", this.generatedByDragonFight); // Paper - Fix invulnerable end crystals
+
+        // KioCG start
+        if (this.turretCooldown != null) {
+            nbt.putInt("KioCG.TurretCooldown", this.turretCooldown);
+        }
+        if (this.turretRange != null) {
+            nbt.putDouble("KioCG.TurretRange", this.turretRange);
+        }
+        if (this.turretDamage != null) {
+            nbt.putDouble("KioCG.TurretDamage", this.turretDamage);
+        }
+        // KioCG end
     }
 
     @Override
@@ -102,6 +187,17 @@ public class EndCrystal extends Entity {
         }
         if (nbt.contains("Paper.GeneratedByDragonFight", 1)) this.generatedByDragonFight = nbt.getBoolean("Paper.GeneratedByDragonFight"); // Paper - Fix invulnerable end crystals
 
+        // KioCG start
+        if (nbt.contains("KioCG.TurretCooldown")) {
+            this.turretCooldown = nbt.getInt("KioCG.TurretCooldown");
+        }
+        if (nbt.contains("KioCG.TurretRange")) {
+            this.turretRange = nbt.getDouble("KioCG.TurretRange");
+        }
+        if (nbt.contains("KioCG.TurretDamage")) {
+            this.turretDamage = nbt.getDouble("KioCG.TurretDamage");
+        }
+        // KioCG end
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
index 676dd5331bec75407a74aea2a89e78ab72d69724..a8c797219e8c89c05ac7b22ecf954678d0e7292b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
@@ -39,6 +39,38 @@ public class CraftEnderCrystal extends CraftEntity implements EnderCrystal {
         }
     }
 
+    // KioCG start
+    @Override
+    public int getTurretCooldown() {
+        return this.getHandle().getTurretCooldown();
+    }
+
+    @Override
+    public void setTurretCooldown(@javax.annotation.Nullable Integer cooldown) {
+        this.getHandle().setTurretCooldown(cooldown);
+    }
+
+    @Override
+    public double getTurretRange() {
+        return this.getHandle().getTurretRange();
+    }
+
+    @Override
+    public void setTurretRange(@javax.annotation.Nullable Double range) {
+        this.getHandle().setTurretRange(range);
+    }
+
+    @Override
+    public double getTurretDamage() {
+        return this.getHandle().getTurretDamage();
+    }
+
+    @Override
+    public void setTurretDamage(@javax.annotation.Nullable Double damage) {
+        this.getHandle().setTurretDamage(damage);
+    }
+    // KioCG end
+
     @Override
     public EndCrystal getHandle() {
         return (EndCrystal) this.entity;
