From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 5 Apr 2024 19:22:58 +0800
Subject: [PATCH] Real big chest


diff --git a/src/main/java/net/minecraft/world/level/block/ChestBlock.java b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
index 8fbfd18b3caeed769396b3ffb1b1778b2f38edc0..303f50597d205f7ccf7f9ec316fdd2767dc125d9 100644
--- a/src/main/java/net/minecraft/world/level/block/ChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
@@ -225,6 +225,7 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
             }
         }
 
+        blockpropertychesttype = ChestType.SINGLE; // KioCG - 强制单箱子
         return (BlockState) ((BlockState) ((BlockState) this.defaultBlockState().setValue(ChestBlock.FACING, enumdirection)).setValue(ChestBlock.TYPE, blockpropertychesttype)).setValue(ChestBlock.WATERLOGGED, fluid.getType() == Fluids.WATER);
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/TrappedChestBlock.java b/src/main/java/net/minecraft/world/level/block/TrappedChestBlock.java
index 17f0f170d68bb55efc61771e9a32ce62924bab2b..035ecad765ad8c0149f335c84f7ab09d8aeb3765 100644
--- a/src/main/java/net/minecraft/world/level/block/TrappedChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TrappedChestBlock.java
@@ -37,6 +37,8 @@ public class TrappedChestBlock extends ChestBlock {
         return Stats.CUSTOM.get(Stats.TRIGGER_TRAPPED_CHEST);
     }
 
+    // KioCG start
+    /*
     @Override
     protected boolean isSignalSource(BlockState state) {
         return true;
@@ -51,4 +53,6 @@ public class TrappedChestBlock extends ChestBlock {
     protected int getDirectSignal(BlockState state, BlockGetter world, BlockPos pos, Direction direction) {
         return direction == Direction.UP ? state.getSignal(world, pos, direction) : 0;
     }
+     */
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
index b88aa184cd06a0485146f58a5b61a56a50911209..4f3d36721fbc021df3d21476f8019ab05da9fe44 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
@@ -68,7 +68,7 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
 
     protected ChestBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state) {
         super(type, pos, state);
-        this.items = NonNullList.withSize(27, ItemStack.EMPTY);
+        this.items = NonNullList.withSize(type == BlockEntityType.TRAPPED_CHEST ? 54 : 27, ItemStack.EMPTY); // KioCG
         this.openersCounter = new ContainerOpenersCounter() {
             @Override
             protected void onOpen(Level world, BlockPos pos, BlockState state) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java
index 74ad37e2b6ebb7a54497767337df98525cf55f22..46ca82cc4f29f8f404ac8049268d1488a9a959ba 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TrappedChestBlockEntity.java
@@ -10,6 +10,8 @@ public class TrappedChestBlockEntity extends ChestBlockEntity {
         super(BlockEntityType.TRAPPED_CHEST, pos, state);
     }
 
+    // KioCG start
+    /*
     @Override
     protected void signalOpenCount(Level world, BlockPos pos, BlockState state, int oldViewerCount, int newViewerCount) {
         super.signalOpenCount(world, pos, state, oldViewerCount, newViewerCount);
@@ -19,4 +21,21 @@ public class TrappedChestBlockEntity extends ChestBlockEntity {
             world.updateNeighborsAt(pos.below(), block);
         }
     }
+     */
+
+    @Override
+    protected net.minecraft.world.inventory.AbstractContainerMenu createMenu(int syncId, net.minecraft.world.entity.player.Inventory playerInventory) {
+        return net.minecraft.world.inventory.ChestMenu.sixRows(syncId, playerInventory, this);
+    }
+
+    @Override
+    protected net.minecraft.network.chat.Component getDefaultName() {
+        return net.minecraft.network.chat.Component.translatable("container.chestDouble");
+    }
+
+    @Override
+    public int getContainerSize() {
+        return 54;
+    }
+    // KioCG end
 }
