From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 18 Apr 2024 04:53:43 +0800
Subject: [PATCH] Anti auto fishing


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 1223c5d23d0ea6aed068bdf0f5725e2ad49fc82c..59af9a55600e419c568fa434045055ca622cfa05 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -330,7 +330,10 @@ public class FishingHook extends Projectile {
             if (this.nibble <= 0) {
                 this.timeUntilLured = 0;
                 this.timeUntilHooked = 0;
-                this.getEntityData().set(FishingHook.DATA_BITING, false);
+                // KioCG start - 防止自动钓鱼
+                // this.getEntityData().set(FishingHook.DATA_BITING, false);
+                this.biting = false;
+                // KioCG end - 防止自动钓鱼
                 // CraftBukkit start
                 PlayerFishEvent playerFishEvent = new PlayerFishEvent((Player) this.getPlayerOwner().getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.FAILED_ATTEMPT);
                 this.level().getCraftServer().getPluginManager().callEvent(playerFishEvent);
@@ -381,7 +384,11 @@ public class FishingHook extends Projectile {
                     worldserver.sendParticles(ParticleTypes.BUBBLE, this.getX(), d3, this.getZ(), (int) (1.0F + this.getBbWidth() * 20.0F), (double) this.getBbWidth(), 0.0D, (double) this.getBbWidth(), 0.20000000298023224D);
                     worldserver.sendParticles(ParticleTypes.FISHING, this.getX(), d3, this.getZ(), (int) (1.0F + this.getBbWidth() * 20.0F), (double) this.getBbWidth(), 0.0D, (double) this.getBbWidth(), 0.20000000298023224D);
                     this.nibble = Mth.nextInt(this.random, 20, 40);
-                    this.getEntityData().set(FishingHook.DATA_BITING, true);
+                    // KioCG start - 防止自动钓鱼
+                    // this.getEntityData().set(FishingHook.DATA_BITING, true);
+                    this.biting = true;
+                    this.setDeltaMovement(this.getDeltaMovement().x, (double) (-0.4F * Mth.nextFloat(this.syncronizedRandom, 0.6F, 1.0F)), this.getDeltaMovement().z);
+                    // KioCG end - 防止自动钓鱼
                 }
             } else if (this.timeUntilLured > 0) {
                 this.timeUntilLured -= i;
