From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 3 Jun 2024 17:38:22 +0800
Subject: [PATCH] Chunk hot


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystem.java b/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystem.java
index 0abba00741b39b69a7f167e5d2670f2565c9a752..4f5506313085ea192faee48ed40998a20a752866 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystem.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystem.java
@@ -115,6 +115,7 @@ public final class ChunkSystem {
     }
 
     public static void onChunkNotTicking(final LevelChunk chunk, final ChunkHolder holder) {
+        chunk.getChunkHot().clear(); // KioCG
         ((ChunkSystemServerLevel)((ServerLevel)chunk.getLevel())).moonrise$getTickingChunks().remove(
                 ((ChunkSystemLevelChunk)chunk).moonrise$getChunkAndHolder()
         );
diff --git a/src/main/java/com/kiocg/ChunkHot.java b/src/main/java/com/kiocg/ChunkHot.java
new file mode 100644
index 0000000000000000000000000000000000000000..bdb263f682c4bb047cdb8ed52d8df7ee0dc875ad
--- /dev/null
+++ b/src/main/java/com/kiocg/ChunkHot.java
@@ -0,0 +1,86 @@
+package com.kiocg;
+
+import java.util.Arrays;
+
+public class ChunkHot {
+    // 热度统计总区间数量
+    private static final int TIMES_LENGTH = 10;
+    // 当前统计区间下标
+    private static int index = -1;
+
+    // 热度统计区间
+    private final long[] times = new long[TIMES_LENGTH];
+    // 存放临时的区间数值
+    // 用于修正正在统计的当前区间热度没有计入总值的问题
+    private long temp;
+    // 所有区间的热度总值
+    private long total;
+
+    // 用于每个具体统计的计算
+    private long nanos;
+    // 当前统计是否进行中
+    private boolean started = false;
+
+    /**
+     * 更新区间下标
+     */
+    public static void nextTick() {
+        index = ++index % TIMES_LENGTH;
+    }
+
+    /**
+     * 开始统计一个新区间
+     */
+    public void start() {
+        started = true;
+        temp = times[index];
+        times[index] = 0L;
+    }
+
+    /**
+     * 结束当前区间的统计
+     * 将统计值更新入热度总值
+     */
+    public void stop() {
+        started = false;
+        total -= temp;
+        total += times[index];
+    }
+
+    /**
+     * 开始一个具体统计
+     */
+    public void startTicking() {
+        if (!started) return;
+        nanos = System.nanoTime();
+    }
+
+    /**
+     * 结束一个具体统计
+     * 将统计值计入当前热度区间
+     */
+    public void stopTickingAndCount() {
+        if (!started) return;
+        // 定义一个具体统计的最大值为 1,000,000
+        // 有时候某个具体统计的计算值会在某1刻飙升，可能是由于保存数据到磁盘？
+        times[index] += Math.min(System.nanoTime() - nanos, 1_000_000L);
+    }
+
+    /**
+     * 清空统计 (当区块卸载时)
+     */
+    public void clear() {
+        started = false;
+        Arrays.fill(times, 0L);
+        temp = 0L;
+        total = 0L;
+        nanos = 0L;
+    }
+
+    /**
+     * @return 获取区块热度平均值
+     */
+    public long getAverage() {
+        return total / ((long) TIMES_LENGTH * 20L);
+    }
+}
diff --git a/src/main/java/com/kiocg/command/ChunkHotCommand.java b/src/main/java/com/kiocg/command/ChunkHotCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..9e2da03a2c841e04f191dbcbd3792cd726a3e720
--- /dev/null
+++ b/src/main/java/com/kiocg/command/ChunkHotCommand.java
@@ -0,0 +1,43 @@
+package com.kiocg.command;
+
+import com.kiocg.task.TPSBarTask;
+import com.mojang.brigadier.CommandDispatcher;
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.EntityArgument;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.level.chunk.LevelChunk;
+
+import java.util.Collection;
+import java.util.Collections;
+
+public class ChunkHotCommand {
+    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        dispatcher.register(Commands.literal("chunkhot")
+                                    .requires(listener -> listener.hasPermission(2, "bukkit.command.chunkhot"))
+                                    .executes(context -> execute(context.getSource(), Collections.singleton(context.getSource().getPlayerOrException())))
+                                    .then(Commands.argument("targets", EntityArgument.players())
+                                                  .requires(listener -> listener.hasPermission(2, "bukkit.command.chunkhot.other"))
+                                                  .executes(context -> execute(context.getSource(), EntityArgument.getPlayers(context, "targets")))
+                                         )
+                           );
+    }
+
+    private static int execute(CommandSourceStack sender, Collection<ServerPlayer> targets) {
+        for (ServerPlayer player : targets) {
+            LevelChunk chunk = player.level().getChunkIfLoaded(player.blockPosition());
+            long chunkhot = chunk != null ? chunk.getChunkHot().getAverage() : 0L;
+
+            Component component = MiniMessage.miniMessage().deserialize("<green>[<aqua>豆渣子<green>] <gold>玩家<target>的区域热度: <totalhot>/<gray><chunkhot>",
+                                                                        Placeholder.parsed("target", player.getGameProfile().getName()),
+                                                                        Placeholder.component("totalhot", TPSBarTask.instance().getChunkHotColor(player)),
+                                                                        Placeholder.parsed("chunkhot", String.valueOf(chunkhot)));
+            sender.sendSuccess(() -> PaperAdventure.asVanilla(component), false);
+        }
+        return targets.size();
+    }
+}
diff --git a/src/main/java/com/kiocg/task/TPSBarTask.java b/src/main/java/com/kiocg/task/TPSBarTask.java
index 3c44d94ba77a24023e77ef4c778b181736ff5f98..ae3ebd432a89f63b5f6277ba66df898d1e1ecc40 100644
--- a/src/main/java/com/kiocg/task/TPSBarTask.java
+++ b/src/main/java/com/kiocg/task/TPSBarTask.java
@@ -28,9 +28,15 @@ public class TPSBarTask extends BossBarTask {
 
     @Override
     void updateBossBar(BossBar bossbar, Player player) {
-        bossbar.progress(getBossBarProgress((float) mspt, 50.0F));
-        bossbar.color(getBossBarColor(FillMode.MSPT));
+        net.minecraft.server.level.ServerPlayer serverPlayer = ((org.bukkit.craftbukkit.entity.CraftPlayer) player).getHandle();
+        net.minecraft.world.level.chunk.LevelChunk chunk = serverPlayer.level().getChunkIfLoaded(serverPlayer.blockPosition());
+        long chunkhot = chunk != null ? chunk.getChunkHot().getAverage() : 0L;
+
+        bossbar.progress(getBossBarProgress((float) serverPlayer.getNearbyChunkHot(), (float) serverPlayer.level().paperConfig().kiocgConfig.chunkHot.phaseOne));
+        bossbar.color(getBossBarColor(serverPlayer));
         bossbar.name(MiniMessage.miniMessage().deserialize(GlobalConfiguration.get().kiocgConfig.commandTpsBarTitle,
+                                                           Placeholder.component("totalhot", getChunkHotColor(serverPlayer)),
+                                                           Placeholder.parsed("chunkhot", String.valueOf(chunkhot)),
                                                            Placeholder.component("tps", getTPSColor()),
                                                            Placeholder.component("mspt", getMSPTColor()),
                                                            Placeholder.component("ping", getPingColor(player.getPing()))
@@ -54,10 +60,11 @@ public class TPSBarTask extends BossBarTask {
         return Math.max(Math.min(dividend / divisor, 1.0F), 0.0F);
     }
 
-    private BossBar.Color getBossBarColor(FillMode fillMode) {
-        if (isGood(fillMode)) {
+    private BossBar.Color getBossBarColor(net.minecraft.server.level.ServerPlayer player) {
+        long hot = player.getNearbyChunkHot();
+        if (hot < player.level().paperConfig().kiocgConfig.chunkHot.phaseOne) {
             return GlobalConfiguration.get().kiocgConfig.commandTpsBarProgressColorGood;
-        } else if (isMedium(fillMode)) {
+        } else if (hot < player.level().paperConfig().kiocgConfig.chunkHot.phaseTwo) {
             return GlobalConfiguration.get().kiocgConfig.commandTpsBarProgressColorMedium;
         } else {
             return GlobalConfiguration.get().kiocgConfig.commandTpsBarProgressColorLow;
@@ -132,6 +139,19 @@ public class TPSBarTask extends BossBarTask {
         return MiniMessage.miniMessage().deserialize(color, Placeholder.parsed("text", String.format("%s", ping)));
     }
 
+    public Component getChunkHotColor(net.minecraft.server.level.ServerPlayer player) {
+        long hot = player.getNearbyChunkHot();
+        String color;
+        if (hot < player.level().paperConfig().kiocgConfig.chunkHot.phaseOne) {
+            color = GlobalConfiguration.get().kiocgConfig.commandTpsBarTextColorGood;
+        } else if (hot < player.level().paperConfig().kiocgConfig.chunkHot.phaseTwo) {
+            color = GlobalConfiguration.get().kiocgConfig.commandTpsBarTextColorMedium;
+        } else {
+            color = GlobalConfiguration.get().kiocgConfig.commandTpsBarTextColorLow;
+        }
+        return MiniMessage.miniMessage().deserialize(color, Placeholder.parsed("text", String.format("%s", hot)));
+    }
+
     public enum FillMode {
         TPS, MSPT, PING
     }
diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 525c8d31ec3097bc975252c6ff1e07ade12cb74e..bc847741e9ab752ef8b7c3724a3a257ee01a7a0b 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -430,7 +430,7 @@ public class GlobalConfiguration extends ConfigurationPart {
         public double petTeleportDistanceSqr = 256.0;
         public boolean petFollowKeepChunkLoaded = true;
 
-        public String commandTpsBarTitle = "<gray>TPS<yellow>:</yellow> <tps> MSPT<yellow>:</yellow> <mspt> Ping<yellow>:</yellow> <ping>ms";
+        public String commandTpsBarTitle = "<gray>TPS<yellow>:</yellow> <tps> MSPT<yellow>:</yellow> <mspt> Ping<yellow>:</yellow> <ping>ms SAN<yellow>:</yellow> <totalhot>/<gray><chunkhot>";
         public net.kyori.adventure.bossbar.BossBar.Overlay commandTpsBarProgressOverlay = net.kyori.adventure.bossbar.BossBar.Overlay.NOTCHED_20;
         public net.kyori.adventure.bossbar.BossBar.Color commandTpsBarProgressColorGood = net.kyori.adventure.bossbar.BossBar.Color.GREEN;
         public net.kyori.adventure.bossbar.BossBar.Color commandTpsBarProgressColorMedium = net.kyori.adventure.bossbar.BossBar.Color.YELLOW;
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 4d3d41c31985506ace170cdf8e78ce1f6f207406..1d3203953d48432c9239ca9eac0c123c7e4ba64f 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -625,6 +625,14 @@ public class WorldConfiguration extends ConfigurationPart {
         public double sleepHealingRegenRate = 20.0;
         public double saplingsWitheredChance = 0.025;
         public boolean flowersAndGrassesSpreadNaturally = true;
+
+        public ChunkHot chunkHot;
+        public class ChunkHot extends ConfigurationPart {
+            public long phaseOne = 300000L;
+            public long phaseTwo = 500000L;
+            public String phaseOneMessage = "<red><i>这片区域令你感到不适";
+            public String phaseTwoMessage = "<dark_red><i><b>古神在你耳边窃窃私语";
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index d26aecbfbd1930e2c6938d049e45b700b3197230..71c4291b90e5bbe2fb60f2a637e9e727e332b670 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -251,6 +251,7 @@ public class Commands {
             TransferCommand.register(this.dispatcher);
             WhitelistCommand.register(this.dispatcher);
             com.kiocg.command.TPSBarCommand.register(this.dispatcher); // KioCG
+            com.kiocg.command.ChunkHotCommand.register(this.dispatcher); // KioCG
         }
 
         if (environment.includeIntegrated) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 8206de9c96c19ce3c32293beebdd75d778cd4974..030a7b816c638de6bd01f93621cca4a3dce19d06 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1594,6 +1594,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
 
         ++this.tickCount;
+        if (this.tickCount % 20 == 1) { com.kiocg.ChunkHot.nextTick(); this.startChunkHotTick(); } // KioCG
         this.tickRateManager.tick();
         this.tickChildren(shouldKeepTicking);
         if (i - this.lastServerStatus >= MinecraftServer.STATUS_EXPIRE_TIME_NANOS) {
@@ -1635,6 +1636,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         new com.destroystokyo.paper.event.server.ServerTickEndEvent(this.tickCount, ((double)(endTime - lastTick) / 1000000D), remaining).callEvent();
         this.server.spark.tickEnd(((double)(endTime - lastTick) / 1000000D)); // Paper - spark
         // Paper end - Server Tick Events
+        if (this.tickCount % 20 == 0) this.stopChunkHotTick(); // KioCG
         this.profiler.push("tallying");
         long j = Util.getNanos() - i;
         int k = this.tickCount % 100;
@@ -1654,6 +1656,26 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         co.aikar.timings.TimingsManager.FULL_SERVER_TICK.stopTiming(); // Paper
     }
 
+    // KioCG start
+    private void startChunkHotTick() {
+        this.getAllLevels().forEach(level -> {
+            final Iterator<ServerChunkCache.ChunkAndHolder> iterator = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel) level).moonrise$getTickingChunks().iterator();
+            while (iterator.hasNext()) {
+                iterator.next().chunk().getChunkHot().start();
+            }
+        });
+    }
+
+    private void stopChunkHotTick() {
+        this.getAllLevels().forEach(level -> {
+            final Iterator<ServerChunkCache.ChunkAndHolder> iterator = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel) level).moonrise$getTickingChunks().iterator();
+            while (iterator.hasNext()) {
+                iterator.next().chunk().getChunkHot().stop();
+            }
+        });
+    }
+    // KioCG end
+
     private void logTickMethodTime(long tickStartTime) {
         if (this.isTickTimeLoggingEnabled()) {
             this.getTickTimeLogger().logPartialSample(Util.getNanos() - tickStartTime, TpsDebugDimensions.TICK_SERVER_METHOD.ordinal());
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index c1dc46ea9c35950073f323db5f9f724b19de5110..ba95a98ba7b06fbb0b07df6c2b51f03dca7b0d21 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1323,6 +1323,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         // Paper start- timings
         final boolean isActive = org.spigotmc.ActivationRange.checkIfActive(entity);
         timer = isActive ? entity.getType().tickTimer.startTiming() : entity.getType().inactiveTickTimer.startTiming(); // Paper
+        LevelChunk levelChunk = entity.shouldTickHot() ? this.getChunkIfLoaded(entity.blockPosition()) : null; // KioCG
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
         try {
         // Paper end - timings
         entity.setOldPosAndRot();
@@ -1340,6 +1342,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         } else { entity.inactiveTick(); } // Paper - EAR 2
         this.getProfiler().pop();
         } finally { timer.stopTiming(); } // Paper - timings
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
         Iterator iterator = entity.getPassengers().iterator();
 
         while (iterator.hasNext()) {
@@ -1363,6 +1366,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
                 // Paper - EAR 2
                 final boolean isActive = org.spigotmc.ActivationRange.checkIfActive(passenger);
                 co.aikar.timings.Timing timer = isActive ? passenger.getType().passengerTickTimer.startTiming() : passenger.getType().passengerInactiveTickTimer.startTiming(); // Paper
+                LevelChunk levelChunk = !(passenger instanceof Player) ? this.getChunkIfLoaded(passenger.blockPosition()) : null; // KioCG
+                if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
                 try {
                 // Paper end
                 passenger.setOldPosAndRot();
@@ -1394,6 +1399,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
                 }
 
             } finally { timer.stopTiming(); }// Paper - EAR2 timings
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
             }
         } else {
             passenger.stopRiding();
@@ -1922,7 +1928,15 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     private boolean doBlockEvent(BlockEventData event) {
         BlockState iblockdata = this.getBlockState(event.pos());
 
-        return iblockdata.is(event.block()) ? iblockdata.triggerEvent(this, event.pos(), event.paramA(), event.paramB()) : false;
+        // KioCG start
+        final boolean triggerEvent;
+        LevelChunk levelChunk = this.getChunkIfLoaded(event.pos());
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try {
+            triggerEvent = iblockdata.triggerEvent(this, event.pos(), event.paramA(), event.paramB());
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); }
+
+        return iblockdata.is(event.block()) ? triggerEvent : false;
+        // KioCG end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 7287a23cff32e3458a626197eb8ef0cf369a98a7..9991900c9c428aea357deb26797a27555f2ed022 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -303,6 +303,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     public @Nullable String clientBrandName = null; // Paper - Brand support
     public boolean inStructure; // KioCG
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
+    private long nearbyChunkHot = 0; public long getNearbyChunkHot() { return this.nearbyChunkHot; } // KioCG
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
@@ -805,6 +806,34 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.trackEnteredOrExitedLavaOnVehicle();
         this.updatePlayerAttributes();
         this.advancements.flushDirty(this);
+
+        // KioCG start
+        if (this.server.getTickCount() % 20 == 1) {
+            this.nearbyChunkHot = this.refreshNearbyChunkHot();
+            if (this.nearbyChunkHot >= this.level().paperConfig().kiocgConfig.chunkHot.phaseTwo) {
+                this.hurt(this.damageSources().generic(), 2.0F);
+                this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(this.level().paperConfig().kiocgConfig.chunkHot.phaseTwoMessage)), true);
+            } else if (this.nearbyChunkHot >= this.level().paperConfig().kiocgConfig.chunkHot.phaseOne) {
+                this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(this.level().paperConfig().kiocgConfig.chunkHot.phaseOneMessage)), true);
+            }
+        }
+    }
+
+    private long refreshNearbyChunkHot() {
+        int x = SectionPos.blockToSectionCoord(this.getBlockX());
+        int z = SectionPos.blockToSectionCoord(this.getBlockZ());
+        net.minecraft.world.level.Level world = this.level();
+        long total = 0L;
+        for (int i = x - 4; i <= x + 4; ++i) {
+            for (int j = z - 4; j <= z + 4; ++j) {
+                net.minecraft.world.level.chunk.LevelChunk chunk0 = world.getChunkIfLoaded(i, j);
+                if (chunk0 != null) {
+                    total += chunk0.getChunkHot().getAverage();
+                }
+            }
+        }
+        return total;
+        // KioCG end
     }
 
     private void updatePlayerAttributes() {
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7796e191747be545e744564a2b0b65790f69114d..048c324ba344d168ced8569771b67627ee57530d 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -400,6 +400,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
 
         if (this.player.getLastActionTime() > 0L && this.server.getPlayerIdleTimeout() > 0 && Util.getMillis() - this.player.getLastActionTime() > (long) this.server.getPlayerIdleTimeout() * 1000L * 60L && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits
             this.player.resetLastActionTime(); // CraftBukkit - SPIGOT-854
+            if (this.player.getNearbyChunkHot() >= this.player.level().paperConfig().kiocgConfig.chunkHot.phaseTwo || this.player.getNearbyChunkHot() >= this.player.level().paperConfig().kiocgConfig.chunkHot.phaseOne && MinecraftServer.getServer().recentTps[1] < 18.0) // KioCG
             this.disconnect((Component) Component.translatable("multiplayer.disconnect.idling"), org.bukkit.event.player.PlayerKickEvent.Cause.IDLING); // Paper - kick event cause
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java b/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
index 1859477e96709368683fe5707327e92f56fbfc8e..4c4814094552fa4382b1cbe649f24f473cdbf22e 100644
--- a/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
@@ -449,4 +449,11 @@ public class AreaEffectCloud extends Entity implements TraceableEntity {
     public EntityDimensions getDimensions(Pose pose) {
         return EntityDimensions.scalable(this.getRadius() * 2.0F, 0.5F);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 7414082fb9cb2b57fb4f1893e85d76bb5eb8903e..a983d4a1789236aba8aa80e8723a6c445de801fa 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -5017,4 +5017,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return ((net.minecraft.server.level.ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
     }
     // Paper end - Expose entity id counter
+
+    // KioCG start
+    public boolean shouldTickHot() {
+        return this.tickCount > 20 * 10 && this.isAlive();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/LightningBolt.java b/src/main/java/net/minecraft/world/entity/LightningBolt.java
index f3b78df6bdd2f3fd833f63948594d0eebc9f717d..aa76cdbd5bd0574b7ec430bbbdc62daf82252c20 100644
--- a/src/main/java/net/minecraft/world/entity/LightningBolt.java
+++ b/src/main/java/net/minecraft/world/entity/LightningBolt.java
@@ -315,4 +315,11 @@ public class LightningBolt extends Entity {
     public Stream<Entity> getHitEntities() {
         return this.hitEntities.stream().filter(Entity::isAlive);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index ddb21438ccb7c2a19d211f28d76d0b38558ebabc..59cc026fd79a343e7b35a2ae79e642c8fb6be202 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1856,4 +1856,11 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
     // KioCG end
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && (!this.removeWhenFarAway(0.0) || this.isPersistenceRequired() || this.requiresCustomPersistence());
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java b/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
index bbfc94237bbd546361cc4a7bde773c810e8c5d49..d9f48fe2bcf1a2e317e97046618c056b05643268 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
@@ -166,4 +166,11 @@ public class TraderLlama extends Llama {
             super.start();
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && !this.canDespawn();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java b/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
index 9323e2c92f4fe1268352121c960e2507de8c7aed..c7694429755160edcf8ef293d29e6dcda02e6952 100644
--- a/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
+++ b/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
@@ -330,4 +330,11 @@ public class WanderingTrader extends net.minecraft.world.entity.npc.AbstractVill
             return !pos.closerToCenterThan(this.trader.position(), proximityDistance);
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 8b84a5f2c3419ec3a1de653d80937cc52e232fb8..e011d5b281b49b0b384ec08a9227f3f6ae021678 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2521,4 +2521,11 @@ public abstract class Player extends LivingEntity {
             return this.message;
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
index 3a71b9c79fffe3a575bbf66f5b8b0ab28100d1a5..991a95c838965bb1cb8cf2eb8133e8ce8a1cd8ac 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
@@ -378,4 +378,11 @@ public abstract class Projectile extends Entity implements TraceableEntity {
 
         return DoubleDoubleImmutablePair.of(d0, d1);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index bfe1d4a232745b8e1fab51e30b6454d2ed68e4e2..d948b0da3c33987c6573866cc4126a7b7373a908 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -85,6 +85,7 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
     private final Int2ObjectMap<GameEventListenerRegistry> gameEventListenerRegistrySections;
     private final LevelChunkTicks<Block> blockTicks;
     private final LevelChunkTicks<Fluid> fluidTicks;
+    private final com.kiocg.ChunkHot chunkHot = new com.kiocg.ChunkHot(); public com.kiocg.ChunkHot getChunkHot() { return this.chunkHot; } // KioCG
 
     public LevelChunk(Level world, ChunkPos pos) {
         this(world, pos, UpgradeData.EMPTY, new LevelChunkTicks<>(), new LevelChunkTicks<>(), 0L, 0, (LevelChunkSection[]) null, (LevelChunk.PostLoadProcessor) null, (BlendingData) null); // KioCG - CropDays
@@ -1022,6 +1023,7 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
                         gameprofilerfiller.push(this::getType);
                         this.blockEntity.tickTimer.startTiming(); // Spigot
+                        LevelChunk.this.chunkHot.startTicking(); // KioCG
                         BlockState iblockdata = LevelChunk.this.getBlockState(blockposition);
 
                         if (this.blockEntity.getType().isValid(iblockdata)) {
@@ -1049,6 +1051,7 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
                         // Spigot start
                     } finally {
                         this.blockEntity.tickTimer.stopTiming();
+                        LevelChunk.this.chunkHot.stopTickingAndCount(); // KioCG
                         // Spigot end
                     }
                 }
