From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 15:25:43 +0800
Subject: [PATCH] Configurable can drop mobs skull


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index f5d568d965d661bc781dd01cee5225d440484cbd..dd084c971bd268e879914fec64d14773de91bee6 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -598,6 +598,15 @@ public class WorldConfiguration extends ConfigurationPart {
                 public int darkEnoughToSprint = 7;
                 public boolean babyZombieJockey = true;
                 public boolean zombiesSkinnedIntoSkeletons = true;
+
+                public Monster.Creeper creeper;
+                public class Creeper extends ConfigurationPart {
+                    public boolean canDropCreeperSkull = true;
+                    public boolean canDropSkeletonSkull = true;
+                    public boolean canDropWitherSkeletonSkull = false;
+                    public boolean canDropZombieSkull = true;
+                    public boolean canDropPiglinSkull = true;
+                }
             }
 
             public KiocgConfig.Entity.Animal animal;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Creeper.java b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
index cb1b19e2e0d8f0744b2355b8f4da0206b196b19c..5c32b2c473e5e56d6bd30218d3fea64ce5809b12 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Creeper.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
@@ -191,7 +191,7 @@ public class Creeper extends Monster implements PowerableMob {
         Entity entity = source.getEntity();
 
         if (entity != this && entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropCreeperSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.CREEPER_HEAD);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index cee42ae2b75c29c89e7fc5b1c77d3b45ce40e9ba..2edba3d770fd577f331174b6fe75e8e7410f8e95 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -134,7 +134,7 @@ public class Skeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
index bb2e7cee612dc1fafa042674a0b0d07d7165b54c..98b60652732fcaf3b13a3c093b90a6fa2dfa32d9 100644
--- a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -65,7 +65,7 @@ public class WitherSkeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropWitherSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.WITHER_SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index cc936e61602eff9595c3f2d605adddbed7a18e48..2a0fd38912c280cf12195a185e2d7447e8fd3641 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -653,7 +653,7 @@ public class Zombie extends Monster {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropZombieSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = this.getSkull();
 
                 if (!itemstack.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index 16f12a0af49e3b116a589e6089b06a065f5c0641..fdd065dc27b65dceb8989b950824b12d1614b4f5 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -139,7 +139,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropPiglinSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = new ItemStack(Items.PIGLIN_HEAD);
 
                 entitycreeper.increaseDroppedSkulls();
