From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 15:25:43 +0800
Subject: [PATCH] Configurable can drop mobs skull


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 2520cbcf01958c282010e8da8b6d21a4b56d250f..0061012c227ca33e58b0ced05996240769e9fd14 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -593,6 +593,15 @@ public class WorldConfiguration extends ConfigurationPart {
                 public int darkEnoughToSprint = 7;
                 public boolean babyZombieJockey = true;
                 public boolean zombiesSkinnedIntoSkeletons = true;
+
+                public Monster.Creeper creeper;
+                public class Creeper extends ConfigurationPart {
+                    public boolean canDropCreeperSkull = true;
+                    public boolean canDropSkeletonSkull = true;
+                    public boolean canDropWitherSkeletonSkull = true;
+                    public boolean canDropZombieSkull = true;
+                    public boolean canDropPiglinSkull = true;
+                }
             }
 
             public KiocgConfig.Entity.Animal animal;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Creeper.java b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
index 0ae4ba060b2ce2c79e1235c939f3c1926eb6e33e..ad9758b5e7dee6a99e713da235a427a69036ccca 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Creeper.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
@@ -191,7 +191,7 @@ public class Creeper extends Monster implements PowerableMob {
         Entity entity = source.getEntity();
 
         if (entity != this && entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropCreeperSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.CREEPER_HEAD);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index 5642bddc8268d70e5bb5446b65be1d8ce34feb9b..f0e33976fc386b54f67b7c5506f0d133829f7b4e 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -133,7 +133,7 @@ public class Skeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
index 3f1191795e58f31b7e2fe34ef2774df13b9a789f..99d561086e6c1397a38f3a9b8fe96016c1c7653c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -64,7 +64,7 @@ public class WitherSkeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropWitherSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.WITHER_SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 0a5fbaa96332026b4be7444776a991b4d191ccd0..d288aa2eb47006c97f20e758b4af65a4db56c8f3 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -639,7 +639,7 @@ public class Zombie extends Monster {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropZombieSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = this.getSkull();
 
                 if (!itemstack.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index 156c4f7f85b61e0c274315425c58286e4337237b..dd242fbbe09c1e677a88ccac263644ed20bac57f 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -140,7 +140,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropPiglinSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = new ItemStack(Items.PIGLIN_HEAD);
 
                 entitycreeper.increaseDroppedSkulls();
