From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 15:58:27 +0800
Subject: [PATCH] Configurable persistent when picked up items


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 39703f1cf2bd79ba79674e3f4ccb8d4855797116..812b29f2d6035b9e9064eb3bd711073f470092c8 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -603,6 +603,8 @@ public class WorldConfiguration extends ConfigurationPart {
                     public boolean canDropZombieSkull = true;
                     public boolean canDropPiglinSkull = true;
                 }
+
+                public boolean persistentWhenPickedUpItems = false;
             }
 
             public KiocgConfig.Entity.Animal animal;
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 41f36499774589982977dda68fcb8bfb3578efd2..22117ef536082bb1090ad686109ce183299fe1ed 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -775,9 +775,32 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot slot, ItemStack stack) {
         this.setItemSlot(slot, stack);
         this.setGuaranteedDrop(slot);
+        if (!this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) // KioCG
         this.persistenceRequired = true;
     }
 
+    // KioCG start
+    private void dropPickedUpItems() {
+        if (this.canPickUpLoot()) {
+            EquipmentSlot[] aenumitemslot = EquipmentSlot.values();
+            int j = aenumitemslot.length;
+            for (int k = 0; k < j; ++k) {
+                EquipmentSlot enumitemslot = aenumitemslot[k];
+                if (this.getEquipmentDropChance(enumitemslot) != 2.0f) {
+                    continue;
+                }
+
+                ItemStack itemstack = this.getItemBySlot(enumitemslot);
+                if (!itemstack.isEmpty() && !EnchantmentHelper.hasVanishingCurse(itemstack)) {
+                    this.forceDrops = true;
+                    this.spawnAtLocation(itemstack);
+                    this.forceDrops = false;
+                }
+            }
+        }
+    }
+    // KioCG end
+
     public void setGuaranteedDrop(EquipmentSlot slot) {
         switch (slot.getType()) {
             case HAND:
@@ -890,6 +913,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+            if (this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) this.dropPickedUpItems(); // KioCG
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Player entityhuman = this.level().findNearbyPlayer(this, -1.0D, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
 
@@ -900,6 +924,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
 
                 if (d0 > (double) j && this.removeWhenFarAway(d0)) {
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    if (this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) this.dropPickedUpItems(); // KioCG
                 }
 
                 int k = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).soft(); // Paper - Configurable despawn distances
@@ -907,6 +932,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
 
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double) l && this.removeWhenFarAway(d0)) {
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    if (this.level().paperConfig().kiocgConfig.entity.monster.persistentWhenPickedUpItems) this.dropPickedUpItems(); // KioCG
                 } else if (d0 < (double) l) {
                     this.noActionTime = 0;
                 }
