From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 15:25:43 +0800
Subject: [PATCH] Configurable can drop mobs skull


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 0e26cf13ed55281c1fd35f86a5ede4b6a4911419..119342180d84af79cd4db9b2f94f8fa141e367dc 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -606,6 +606,15 @@ public class WorldConfiguration extends ConfigurationPart {
                 public int darkEnoughToSprint = 7;
                 public double zombiesSkinnedChance = 0.25;
                 public boolean babyZombieJockey = true;
+
+                public Monster.Creeper creeper;
+                public class Creeper extends ConfigurationPart {
+                    public boolean canDropCreeperSkull = true;
+                    public boolean canDropSkeletonSkull = true;
+                    public boolean canDropWitherSkeletonSkull = false;
+                    public boolean canDropZombieSkull = true;
+                    public boolean canDropPiglinSkull = true;
+                }
             }
 
             public KiocgConfig.Entity.Animal animal;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Creeper.java b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
index d1041f2a4f1b3ea29ad532006e18cdc30c5019fa..aeb03a9648cbb149cc555177301cbe96e24e0486 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Creeper.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
@@ -191,7 +191,7 @@ public class Creeper extends Monster implements PowerableMob {
         Entity entity = source.getEntity();
 
         if (entity != this && entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropCreeperSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.CREEPER_HEAD);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index cee42ae2b75c29c89e7fc5b1c77d3b45ce40e9ba..2edba3d770fd577f331174b6fe75e8e7410f8e95 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -134,7 +134,7 @@ public class Skeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
index bb2e7cee612dc1fafa042674a0b0d07d7165b54c..98b60652732fcaf3b13a3c093b90a6fa2dfa32d9 100644
--- a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -65,7 +65,7 @@ public class WitherSkeleton extends AbstractSkeleton {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropWitherSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.WITHER_SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 8de73725f9121cd9ba79ed7b2020c3ccff684d41..e5eab2c80b86d70ef0783180f3923ed0f74a41a0 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -663,7 +663,7 @@ public class Zombie extends Monster {
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropZombieSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = this.getSkull();
 
                 if (!itemstack.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index e55b3c44cfb7d71646a5d0b5102f4c4c593b2a52..e1137ed1af882abce0b6c9d1bbbd487ccbeb566c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -139,7 +139,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
         Entity entity = source.getEntity();
 
         if (entity instanceof Creeper entitycreeper) {
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entity.monster.creeper.canDropPiglinSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = new ItemStack(Items.PIGLIN_HEAD);
 
                 entitycreeper.increaseDroppedSkulls();
