From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 16:03:11 +0800
Subject: [PATCH] Transform spawned via mob spawner


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 73c41171173864ed4ca44bca2461c5e4ed17e4ce..413d430f0eb68b34d5d17eb54d79f8cd8f23c0f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -614,6 +614,7 @@ public class CraftEventFactory {
         List<org.bukkit.entity.Entity> list = new ArrayList<>();
         for (net.minecraft.world.entity.LivingEntity entityLiving : convertedList) {
             entityLiving.transformScaleSimulate(original); // KioCG
+            entityLiving.spawnedViaMobSpawner = original.spawnedViaMobSpawner; // KioCG
             list.add(entityLiving.getBukkitEntity());
         }
 
@@ -1346,6 +1347,7 @@ public class CraftEventFactory {
 
     public static PigZapEvent callPigZapEvent(Entity pig, Entity lightning, Entity pigzombie) {
         ((net.minecraft.world.entity.LivingEntity) pigzombie).transformScaleSimulate((net.minecraft.world.entity.LivingEntity) pig); // KioCG
+        pigzombie.spawnedViaMobSpawner = pig.spawnedViaMobSpawner; // KioCG
         PigZapEvent event = new PigZapEvent((Pig) pig.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), (PigZombie) pigzombie.getBukkitEntity());
         pig.getBukkitEntity().getServer().getPluginManager().callEvent(event);
         return event;
@@ -1359,6 +1361,7 @@ public class CraftEventFactory {
 
     // Paper start
     public static com.destroystokyo.paper.event.entity.EntityZapEvent callEntityZapEvent(Entity entity, Entity lightning, Entity changedEntity) {
+        changedEntity.spawnedViaMobSpawner = entity.spawnedViaMobSpawner; // KioCG
         com.destroystokyo.paper.event.entity.EntityZapEvent event = new com.destroystokyo.paper.event.entity.EntityZapEvent(entity.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), changedEntity.getBukkitEntity());
         entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
         return event;
