From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 5 Jun 2024 16:03:11 +0800
Subject: [PATCH] Transform spawned via mob spawner


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index b6045bf89f3b4d4e04664ab4f0e0ca592e6d63c3..ae09b77db91184553ee1a7d40fad461671512896 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -620,6 +620,7 @@ public class CraftEventFactory {
         List<org.bukkit.entity.Entity> list = new ArrayList<>();
         for (net.minecraft.world.entity.LivingEntity entityLiving : convertedList) {
             entityLiving.transformScaleSimulate(original); // KioCG
+            entityLiving.spawnedViaMobSpawner = original.spawnedViaMobSpawner; // KioCG
             list.add(entityLiving.getBukkitEntity());
         }
 
@@ -1352,6 +1353,7 @@ public class CraftEventFactory {
 
     public static PigZapEvent callPigZapEvent(Entity pig, Entity lightning, Entity pigzombie) {
         ((net.minecraft.world.entity.LivingEntity) pigzombie).transformScaleSimulate((net.minecraft.world.entity.LivingEntity) pig); // KioCG
+        pigzombie.spawnedViaMobSpawner = pig.spawnedViaMobSpawner; // KioCG
         PigZapEvent event = new PigZapEvent((Pig) pig.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), (PigZombie) pigzombie.getBukkitEntity());
         pig.getBukkitEntity().getServer().getPluginManager().callEvent(event);
         return event;
@@ -1365,6 +1367,7 @@ public class CraftEventFactory {
 
     // Paper start
     public static com.destroystokyo.paper.event.entity.EntityZapEvent callEntityZapEvent(Entity entity, Entity lightning, Entity changedEntity) {
+        changedEntity.spawnedViaMobSpawner = entity.spawnedViaMobSpawner; // KioCG
         com.destroystokyo.paper.event.entity.EntityZapEvent event = new com.destroystokyo.paper.event.entity.EntityZapEvent(entity.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), changedEntity.getBukkitEntity());
         entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
         return event;
