From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KioProject <1105919949@qq.com>
Date: Sun, 6 Apr 2025 23:22:52 +0800
Subject: [PATCH] Custom block entity


diff --git a/net/minecraft/core/flowerpot/FlowerPotInteraction.java b/net/minecraft/core/flowerpot/FlowerPotInteraction.java
new file mode 100644
index 0000000000000000000000000000000000000000..4df32bf9db75aa28ef82417dc907906ab3a472db
--- /dev/null
+++ b/net/minecraft/core/flowerpot/FlowerPotInteraction.java
@@ -0,0 +1,296 @@
+package net.minecraft.core.flowerpot;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.entity.Display;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.*;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
+import net.minecraft.world.level.block.state.properties.Property;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.shapes.VoxelShape;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+// KioCG start - custom block entity: all class
+public interface FlowerPotInteraction {
+    RandomSource RANDOM = RandomSource.create();
+    double pixel1 = 1.0 / 16.0, pixel2 = 2.0 / 16.0, pixel3 = 3.0 / 16.0, pixel4 = 4.0 / 16.0;
+    double pixel5 = 5.0 / 16.0, pixel6 = 6.0 / 16.0, pixel7 = 7.0 / 16.0, pixel8 = 8.0 / 16.0;
+    double pixel9 = 9.0 / 16.0, pixel10 = 10.0 / 16.0, pixel11 = 11.0 / 16.0, pixel12 = 12.0 / 16.0;
+    double pixel13 = 13.0 / 16.0, pixel14 = 14.0 / 16.0, pixel15 = 15.0 / 16.0, pixel16 = 16.0 / 16.0;
+
+    @Nullable Consumer<Display.BlockDisplay> interact(Level level, BlockPos pos, BlockState blockState);
+
+    Object2ObjectOpenHashMap<Block, Data> map = new Object2ObjectOpenHashMap<>();
+
+    static Optional<BlockState> getBlockState(Level level, BlockPos pos, Block block) {
+        Data data = FlowerPotInteraction.map.get(block);
+        BlockState blockState = data.blockStateGenerator().apply(block);
+        return Optional.ofNullable(data.interaction().interact(level, pos, blockState)).map(consumer -> blockState);
+    }
+
+    static Optional<Consumer<Display.BlockDisplay>> getInteract(Level level, BlockPos pos, Block potted, BlockState pottedState) {
+        Data data = FlowerPotInteraction.map.get(potted);
+        return Optional.ofNullable(data.interaction().interact(level, pos, pottedState));
+    }
+
+    static void bootStrap() {
+        map.defaultReturnValue(Data.get(FlowerPotInteraction::defaultInteraction));
+        map.put(Blocks.PALE_HANGING_MOSS, Data.get((level, pos, state) -> {
+            if (state.getValue(HangingMossBlock.TIP)) {
+                return FlowerPotInteraction.hangingInteraction(level, pos, state);
+            } else {
+                return FlowerPotInteraction.normalInteraction(level, pos, state);
+            }
+        }).randomProperty(HangingMossBlock.TIP));
+        map.put(Blocks.SMALL_AMETHYST_BUD, Data.get((level, pos, state) -> consumer(state, 0.65)));
+        map.put(Blocks.MEDIUM_AMETHYST_BUD, Data.get((level, pos, state) -> consumer(state, 0.65)));
+        map.put(Blocks.LARGE_AMETHYST_BUD, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.AMETHYST_CLUSTER, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.SHORT_GRASS, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.SHORT_DRY_GRASS, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.BUSH, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.PINK_PETALS, Data.get(FlowerPotInteraction::quarterInteraction));
+        map.put(Blocks.WILDFLOWERS, Data.get(FlowerPotInteraction::quarterInteraction));
+        map.put(Blocks.LEAF_LITTER, Data.get((level, pos, state) -> consumer(state, pixel6 / pixel16).andThen(consumerFlat()), state -> {
+            return state.setValue(LeafLitterBlock.AMOUNT, LeafLitterBlock.MAX_SEGMENT);
+        }));
+        map.put(Blocks.SPORE_BLOSSOM, Data.get((level, pos, state) -> consumer(state, pixel6 / pixel12).andThen(consumerInvert()).andThen(consumerFlat())));
+        map.put(Blocks.FIREFLY_BUSH, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.SUGAR_CANE, Data.get((level, pos, state) -> consumer(state, 0.45, pixel12, 0.45)));
+        map.put(Blocks.NETHER_SPROUTS, Data.get((level, pos, state) -> consumer(state, 0.45)));
+        map.put(Blocks.WEEPING_VINES, Data.get(FlowerPotInteraction::hangingInteraction, Blocks.WEEPING_VINES_PLANT));
+        map.put(Blocks.TWISTING_VINES, Data.get(FlowerPotInteraction::normalInteraction, Blocks.TWISTING_VINES_PLANT));
+        map.put(Blocks.VINE, Data.get(FlowerPotInteraction::sphericalInteraction));
+
+        Data doublePlantData = Data.get(FlowerPotInteraction::normalInteraction).randomProperty(DoublePlantBlock.HALF);
+        Data doublePlantLowerInvertData = Data.get((level, pos, state) -> {
+            if (state.getValue(DoublePlantBlock.HALF) == DoubleBlockHalf.UPPER) {
+                return FlowerPotInteraction.normalInteraction(level, pos, state);
+            } else {
+                return FlowerPotInteraction.hangingInteraction(level, pos, state);
+            }
+        }).randomProperty(DoublePlantBlock.HALF);
+        map.put(Blocks.TALL_GRASS, doublePlantData);
+        map.put(Blocks.LARGE_FERN, doublePlantLowerInvertData);
+        map.put(Blocks.TALL_DRY_GRASS, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.SUNFLOWER, doublePlantData);
+        map.put(Blocks.LILAC, doublePlantData);
+        map.put(Blocks.ROSE_BUSH, doublePlantData);
+        map.put(Blocks.PEONY, doublePlantLowerInvertData);
+        map.put(Blocks.PITCHER_PLANT, Data.get((level, pos, state) -> {
+            double size = pixel12;
+            if (state.getValue(DoublePlantBlock.HALF) == DoubleBlockHalf.UPPER) {
+                return consumer(state, size, size, size, pixel4 + pixel5 * size);
+            } else {
+                return consumer(state, size, size, size, pixel4 - pixel5 * size).andThen(consumerInvert());
+            }
+        }).randomProperty(DoublePlantBlock.HALF));
+        map.put(Blocks.BIG_DRIPLEAF, Data.get((level, pos, state) -> {
+            double size = pixel12;
+            return consumer(state, size)
+                    .andThen(blockDisplay -> {
+                        Direction facing = state.getValue(BigDripleafBlock.FACING);
+                        blockDisplay.setTranslation(blockDisplay.getTranslation().add((float) (facing.getStepX() * (pixel4 * size)), 0.0F, (float) (facing.getStepZ() * (pixel4 * size))));
+                    });
+        }).randomProperty(BlockStateProperties.TILT));
+        map.put(Blocks.SMALL_DRIPLEAF, doublePlantData);
+        map.put(Blocks.CHORUS_PLANT, Data.get((level, pos, state) -> {
+            // 每次加载都随机样式
+            state = state.setValue(ChorusPlantBlock.NORTH, RANDOM.nextBoolean())
+                         .setValue(ChorusPlantBlock.EAST, RANDOM.nextBoolean())
+                         .setValue(ChorusPlantBlock.SOUTH, RANDOM.nextBoolean())
+                         .setValue(ChorusPlantBlock.WEST, RANDOM.nextBoolean())
+                         .setValue(ChorusPlantBlock.UP, RANDOM.nextBoolean());
+            return consumer(state, pixel4 / pixel8, pixel12, pixel4 / pixel8);
+        }, state -> state.setValue(ChorusPlantBlock.UP, true).setValue(ChorusPlantBlock.DOWN, true)));
+        map.put(Blocks.GLOW_LICHEN, Data.get(FlowerPotInteraction::sphericalInteraction));
+        map.put(Blocks.HANGING_ROOTS, Data.get(FlowerPotInteraction::hangingInteraction));
+        map.put(Blocks.WHEAT, Data.get(FlowerPotInteraction::cropCircularInteraction).randomProperty(CropBlock.AGE));
+        map.put(Blocks.COCOA, Data.get((level, pos, state) -> {
+            double size = pixel4 / pixel8 + pixel1;
+            return consumer(state, size, size, size, pixel4 - pixel3 * size)
+                    .andThen(blockDisplay -> {
+                        Direction facing = state.getValue(BigDripleafBlock.FACING);
+                        blockDisplay.setTranslation(blockDisplay.getTranslation().add((float) -(facing.getStepX() * (pixel3 * size)), 0.0F, (float) -(facing.getStepZ() * (pixel3 * size))));
+                    });
+        }, state -> state.setValue(CocoaBlock.AGE, CocoaBlock.MAX_AGE)));
+        map.put(Blocks.PUMPKIN_STEM, Data.get(FlowerPotInteraction::normalInteraction, state -> state.setValue(StemBlock.AGE, RANDOM.nextInt(1, StemBlock.MAX_AGE))));
+        map.put(Blocks.MELON_STEM, Data.get(FlowerPotInteraction::normalInteraction, state -> state.setValue(StemBlock.AGE, RANDOM.nextInt(1, StemBlock.MAX_AGE))));
+        map.put(Blocks.BEETROOTS, Data.get(FlowerPotInteraction::cropCircularInteraction).randomProperty(BeetrootBlock.AGE));
+        map.put(Blocks.TORCHFLOWER_CROP, Data.get(FlowerPotInteraction::normalInteraction).randomProperty(TorchflowerCropBlock.AGE));
+        map.put(Blocks.CACTUS_FLOWER, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.PITCHER_CROP, Data.get((level, pos, state) -> consumer(state, pixel4 / pixel10 + pixel1).andThen(consumerFlat()), state -> state.setValue(PitcherCropBlock.AGE, RANDOM.nextInt(1, 3))));
+        map.put(Blocks.CAVE_VINES, Data.get(FlowerPotInteraction::hangingInteraction).randomProperty(CaveVines.BERRIES));
+        map.put(Blocks.SWEET_BERRY_BUSH, Data.get(FlowerPotInteraction::normalInteraction).randomProperty(SweetBerryBushBlock.AGE));
+        map.put(Blocks.NETHER_WART, Data.get(FlowerPotInteraction::cropCircularInteraction).randomProperty(NetherWartBlock.AGE));
+        map.put(Blocks.LILY_PAD, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.SEAGRASS, Data.get((level, pos, state) -> consumer(state, pixel8 - 1.0E-4, pixel10, pixel8 - 1.0E-4)));
+        map.put(Blocks.SEA_PICKLE, Data.get(FlowerPotInteraction::originalInteraction).randomProperty(SeaPickleBlock.WATERLOGGED));
+        map.put(Blocks.KELP, Data.get((level, pos, state) -> consumer(state, pixel12, pixel16, pixel12)));
+        map.put(Blocks.TUBE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.BRAIN_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.BUBBLE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.FIRE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.HORN_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.DEAD_TUBE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.DEAD_BRAIN_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.DEAD_BUBBLE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.DEAD_FIRE_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.DEAD_HORN_CORAL, Data.get(FlowerPotInteraction::normalInteraction));
+        map.put(Blocks.TUBE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.BRAIN_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.BUBBLE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.FIRE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.HORN_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.DEAD_TUBE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.DEAD_BRAIN_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.DEAD_BUBBLE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.DEAD_FIRE_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.DEAD_HORN_CORAL_FAN, Data.get(FlowerPotInteraction::flatInteraction));
+        map.put(Blocks.SCULK_VEIN, Data.get(FlowerPotInteraction::sphericalInteraction));
+        map.put(Blocks.CARROTS, Data.get(FlowerPotInteraction::cropCircularInteraction).randomProperty(CarrotBlock.AGE));
+        map.put(Blocks.POTATOES, Data.get(FlowerPotInteraction::cropCircularInteraction).randomProperty(PotatoBlock.AGE));
+        map.put(Blocks.CAKE, Data.get((level, pos, state) -> {
+            double size = pixel6 / pixel14 - pixel1;
+            return consumer(state, size, size, size, pixel6 - pixel8 * size + (pixel1 / 2));
+        }));
+    }
+
+    private static Consumer<Display.BlockDisplay> normalInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return consumer(blockState);
+    }
+
+    private static Consumer<Display.BlockDisplay> originalInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return consumer(blockState, 1.0);
+    }
+
+    private static Consumer<Display.BlockDisplay> hangingInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return normalInteraction(level, pos, blockState).andThen(consumerInvert());
+    }
+
+    private static Consumer<Display.BlockDisplay> flatInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return normalInteraction(level, pos, blockState).andThen(consumerFlat());
+    }
+
+    private static Consumer<Display.BlockDisplay> quarterInteraction(Level level, BlockPos pos, BlockState blockState) {
+        double size = 0.7;
+        return consumer(blockState, size, 1.0, size).andThen(blockDisplay -> {
+            blockDisplay.setTranslation(blockDisplay.getTranslation().add((float) (pixel4 * size), 0.0F, (float) (pixel4 * size)));
+        });
+    }
+
+    private static Consumer<Display.BlockDisplay> sphericalInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return consumer(blockState, pixel8);
+    }
+
+    private static Consumer<Display.BlockDisplay> cropInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return consumer(blockState, pixel12);
+    }
+
+    private static Consumer<Display.BlockDisplay> cropCircularInteraction(Level level, BlockPos pos, BlockState blockState) {
+        return consumer(blockState, pixel8 - 1.0E-4, pixel8 + pixel1, pixel8 - 1.0E-4).andThen(consumerFlat());
+    }
+
+    static @Nullable Consumer<Display.BlockDisplay> defaultInteraction(Level level, BlockPos pos, BlockState blockState) {
+        if (blockState.is(BlockTags.WALLS)) {
+            return consumer(blockState, pixel4 / pixel8, pixel12, pixel4 / pixel8);
+        }
+
+        if (blockState.hasProperty(HorizontalDirectionalBlock.FACING)) {
+            return null;
+        }
+
+        VoxelShape voxelShape = blockState.getShape(level, pos);
+        if (voxelShape.isEmpty()) {
+            return null;
+        }
+
+        AABB bounds = voxelShape.bounds();
+        if (bounds.minY != 0) {
+            return null;
+        }
+
+        double xsize = bounds.getXsize();
+        double ysize = bounds.getYsize();
+        double zsize = bounds.getZsize();
+        if (xsize <= pixel4 && zsize <= pixel4 && ysize >= pixel4 && ysize <= pixel12) {
+            return FlowerPotInteraction.originalInteraction(level, pos, blockState);
+        } else if (xsize <= pixel6 && zsize <= pixel6 && ysize >= pixel6) {
+            return FlowerPotInteraction.normalInteraction(level, pos, blockState);
+        } else {
+            return null;
+        }
+    }
+
+    private static Consumer<Display.BlockDisplay> consumer(BlockState blockState) {
+        return consumer(blockState, pixel12);
+    }
+
+    private static Consumer<Display.BlockDisplay> consumer(BlockState blockState, double size) {
+        return consumer(blockState, size, size, size);
+    }
+
+    private static Consumer<Display.BlockDisplay> consumer(BlockState blockState, double xsize, double ysize, double zsize) {
+        return consumer(blockState, xsize, ysize, zsize, pixel4);
+    }
+
+    private static Consumer<Display.BlockDisplay> consumer(BlockState blockState, double xsize, double ysize, double zsize, double floorY) {
+        return blockDisplay -> {
+            blockDisplay.setScaleCenterFloor(xsize, ysize, zsize, floorY);
+            blockDisplay.setBlockState(blockState);
+        };
+    }
+
+    private static Consumer<Display.BlockDisplay> consumerInvert() {
+        return Display::setInvert;
+    }
+
+    private static Consumer<Display.BlockDisplay> consumerFlat() {
+        return display -> display.setTranslation(display.getTranslation().add(0.0F, (float) pixel2, 0.0F));
+    }
+
+    record Data(FlowerPotInteraction interaction, Function<Block, BlockState> blockStateGenerator) {
+        private static Data get(FlowerPotInteraction interaction) {
+            return new Data(interaction, Block::defaultBlockState);
+        }
+
+        private static Data get(FlowerPotInteraction interaction, Block fixed) {
+            return new Data(interaction, block -> fixed.defaultBlockState());
+        }
+
+        private static Data get(FlowerPotInteraction interaction, Function<BlockState, BlockState> defaultStateFunction) {
+            return new Data(interaction, block -> defaultStateFunction.apply(block.defaultBlockState()));
+        }
+
+        private static Data get(FlowerPotInteraction interaction, Block fixed, Function<BlockState, BlockState> defaultStateFunction) {
+            return new Data(interaction, block -> defaultStateFunction.apply(fixed.defaultBlockState()));
+        }
+
+        private <T extends Comparable<T>, V extends T> Data randomProperty(Property<?>... properties) {
+            return new Data(this.interaction, this.blockStateGenerator.andThen(blockState -> {
+                for (Property<?> property : properties) {
+                    if (property instanceof BooleanProperty booleanProperty) {
+                        blockState = blockState.setValue(booleanProperty, RANDOM.nextBoolean());
+                    } else {
+                        List<?> possibleValues = property.getPossibleValues();
+                        blockState = blockState.setValue((Property<T>) property, (V) possibleValues.get(RANDOM.nextInt(possibleValues.size())));
+                    }
+                }
+                return blockState;
+            }));
+        }
+    }
+}
+// KioCG end - custom block entity: all class
diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 9f6d7c5dc0e591488a8a3763d8a1f1b3671d5299..add904be7002fea4cbdbba012eadb9051a6c1384 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -62,6 +62,7 @@ public class ClientboundLevelChunkPacketData {
         int totalTileEntities = 0; // Paper - Handle oversized block entities in chunks
 
         for (Entry<BlockPos, BlockEntity> entry : levelChunk.getBlockEntities().entrySet()) {
+            if (BlockEntityType.UNABLE_SEND_TO_CLIENT.contains(entry.getValue().getType())) continue; // KioCG - custom block entity
             // Paper start - Handle oversized block entities in chunks
             if (++totalTileEntities > BLOCK_ENTITY_LIMIT) {
                 net.minecraft.network.protocol.Packet<ClientGamePacketListener> packet = entry.getValue().getUpdatePacket();
diff --git a/net/minecraft/server/Bootstrap.java b/net/minecraft/server/Bootstrap.java
index 83bc2f6b7774c8753a15dbeb00f0c9103713fd1b..f371683beb6c797deca109e86986dc6241ed5a64 100644
--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -55,6 +55,7 @@ public class Bootstrap {
                     EntitySelectorOptions.bootStrap();
                     DispenseItemBehavior.bootStrap();
                     CauldronInteraction.bootStrap();
+                    net.minecraft.core.flowerpot.FlowerPotInteraction.bootStrap(); // KioCG - custom block entity
                     // Paper start
                     BuiltInRegistries.bootStrap(() -> {
                         io.papermc.paper.world.worldgen.OptionallyFlatBedrockConditionSource.bootstrap(); // Paper - Flat bedrock generator settings
diff --git a/net/minecraft/world/entity/Display.java b/net/minecraft/world/entity/Display.java
index 91e9ed6a11c97d1b0482f870078fa355c5ce01d8..698965c983de4c7ab05d0c1807e604e5994e2a58 100644
--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -181,6 +181,39 @@ public abstract class Display extends Entity {
         }
     }
 
+    // KioCG start - custom block entity
+    @Nullable
+    public static <T extends Display> T getCustomDisplay(EntityType<T> entityType, ServerLevel level, net.minecraft.core.BlockPos pos) {
+        List<T> list = com.google.common.collect.Lists.newArrayList();
+        double x = pos.getX() + 0.5, y = pos.getY() + 0.5, z = pos.getZ() + 0.5;
+        level.getEntities(entityType, new AABB(x, y, z, x, y, z).inflate(1.0E-5), display -> {
+            return !display.persist;
+        }, list, 1);
+        return !list.isEmpty() ? list.getFirst() : null;
+    }
+
+    public static <T extends Display> void createCustomDisplay(EntityType<T> entityType, ServerLevel level, net.minecraft.core.BlockPos pos, java.util.function.Consumer<T> consumer) {
+        T display = entityType.create(level, EntitySpawnReason.COMMAND);
+        if (display != null) {
+            display.snapTo(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, 0.0F, 0.0F);
+            // 将方块展示实体居中
+            if (display instanceof BlockDisplay) {
+                display.setTranslation(new Vector3f(-0.5F, -0.5F, -0.5F));
+            }
+            display.persist = false;
+            consumer.accept(display);
+            level.addFreshEntity(display, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CUSTOM);
+        }
+    }
+
+    public static <T extends Display> void discardCustomDisplay(EntityType<T> entityType, ServerLevel level, net.minecraft.core.BlockPos pos) {
+        T customDisplay = getCustomDisplay(entityType, level, pos);
+        if (customDisplay != null) {
+            customDisplay.discard();
+        }
+    }
+    // KioCG end - custom block entity
+
     @Override
     public InterpolationHandler getInterpolation() {
         return this.interpolation;
diff --git a/net/minecraft/world/level/block/AbstractCauldronBlock.java b/net/minecraft/world/level/block/AbstractCauldronBlock.java
index 1a927b08f86c0585e62d8c4207954ab36d3316ee..bac5504e668c9c49bdea394d677636d5c68167af 100644
--- a/net/minecraft/world/level/block/AbstractCauldronBlock.java
+++ b/net/minecraft/world/level/block/AbstractCauldronBlock.java
@@ -23,7 +23,7 @@ import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public abstract class AbstractCauldronBlock extends Block {
+public abstract class AbstractCauldronBlock extends BaseEntityBlock { // KioCG - custom block entity
     protected static final int FLOOR_LEVEL = 4;
     private static final VoxelShape SHAPE_INSIDE = Block.column(12.0, 4.0, 16.0);
     protected static final VoxelShape SHAPE = Util.make(
@@ -99,4 +99,17 @@ public abstract class AbstractCauldronBlock extends Block {
 
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
     }
+
+    // KioCG start - custom block entity
+    @Override
+    public net.minecraft.world.level.block.entity.BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
+        return new net.minecraft.world.level.block.entity.CauldronBlockEntity(pos, state);
+    }
+
+    @javax.annotation.Nullable
+    @Override
+    public <T extends net.minecraft.world.level.block.entity.BlockEntity> net.minecraft.world.level.block.entity.BlockEntityTicker<T> getTicker(Level level, BlockState state, net.minecraft.world.level.block.entity.BlockEntityType<T> blockEntityType) {
+        return null;
+    }
+    // KioCG end - custom block entity
 }
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index be6f37f91569c659c609e5e8d38671ca86f8cd95..8a1f42a77800c2c105e7a101a4d914fb7f91151c 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -595,6 +595,12 @@ public class Block extends BlockBehaviour implements ItemLike {
         return this.builtInRegistryHolder;
     }
 
+    // KioCG start - custom block entity
+    public boolean hasBlockEntity(BlockState blockState) {
+        return this instanceof EntityBlock;
+    }
+    // KioCG end - custom block entity
+
     protected int tryDropExperience(ServerLevel level, BlockPos pos, ItemStack heldItem, IntProvider amount) { // CraftBukkit
         int i = EnchantmentHelper.processBlockExperience(level, heldItem, amount.sample(level.getRandom()));
         if (i > 0) {
diff --git a/net/minecraft/world/level/block/FlowerPotBlock.java b/net/minecraft/world/level/block/FlowerPotBlock.java
index a972c5b4141c3f267070b65bbac5937967fdb3c8..7ace060aeb7751977eac512345c41cb9580c580a 100644
--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -28,7 +28,7 @@ import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class FlowerPotBlock extends Block {
+public class FlowerPotBlock extends BaseEntityBlock { // KioCG - custom block entity
     public static final MapCodec<FlowerPotBlock> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(
                 BuiltInRegistries.BLOCK.byNameCodec().fieldOf("potted").forGetter(flowerPotBlock -> flowerPotBlock.potted), propertiesCodec()
@@ -59,13 +59,75 @@ public class FlowerPotBlock extends Block {
     protected InteractionResult useItemOn(
         ItemStack stack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
+        if (!this.isEmpty()) return InteractionResult.TRY_WITH_EMPTY_HAND; // KioCG - custom block entity: 优化原版花盆交互行为
+
         BlockState blockState = (stack.getItem() instanceof BlockItem blockItem ? POTTED_BY_CONTENT.getOrDefault(blockItem.getBlock(), Blocks.AIR) : Blocks.AIR)
             .defaultBlockState();
+
+        // KioCG start - custom block entity
+        if (level.getBlockEntity(pos) instanceof net.minecraft.world.level.block.entity.FlowerPotBlockEntity flowerPotBlockEntity) {
+            if (!flowerPotBlockEntity.isEmpty()) {
+                BlockState pottedState = flowerPotBlockEntity.getPottedState();
+                if (pottedState.is(net.minecraft.tags.BlockTags.CANDLES) || pottedState.is(net.minecraft.tags.BlockTags.CANDLE_CAKES)) {
+                    if (pottedState.hasProperty(AbstractCandleBlock.LIT) && !pottedState.getValue(AbstractCandleBlock.LIT) && stack.is(net.minecraft.world.item.Items.FLINT_AND_STEEL)) {
+                        // CraftBukkit start - Store the clicked block
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, pos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+                            stack.hurtAndBreak(1, player, net.minecraft.world.entity.LivingEntity.getSlotForHand(hand));
+                            return InteractionResult.CONSUME;
+                        }
+                        // CraftBukkit end
+                        Player.sendSoundEffect(player, pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, net.minecraft.sounds.SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
+                        flowerPotBlockEntity.updatePottedState(pottedState.setValue(AbstractCandleBlock.LIT, true));
+                        level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
+                        stack.hurtAndBreak(1, player, net.minecraft.world.entity.LivingEntity.getSlotForHand(hand));
+                        return InteractionResult.SUCCESS_SERVER;
+                    }
+                } else if (pottedState.is(Blocks.CAKE)) {
+                    net.minecraft.world.item.Item item = stack.getItem();
+                    if (pottedState.getValue(CakeBlock.BITES) == 0 && stack.is(net.minecraft.tags.ItemTags.CANDLES) && Block.byItem(item) instanceof CandleBlock candleBlock) {
+                        InteractionResult interactionResult = onPut(stack, level, pos, player, hand);
+                        if (interactionResult == InteractionResult.SUCCESS) {
+                            level.playSound(null, pos, net.minecraft.sounds.SoundEvents.CAKE_ADD_CANDLE, SoundSource.BLOCKS, 1.0F, 1.0F);
+                            flowerPotBlockEntity.updatePottedState(CandleCakeBlock.byCandle(candleBlock));
+                            player.awardStat(Stats.ITEM_USED.get(item));
+                            return InteractionResult.SUCCESS_SERVER;
+                        }
+                        return InteractionResult.CONSUME;
+                    }
+                }
+                return InteractionResult.TRY_WITH_EMPTY_HAND;
+            }
+
+            if (blockState.isAir() && stack.getItem() instanceof BlockItem blockItem) {
+                Block block = blockItem.getBlock();
+                java.util.Optional<BlockState> optional = net.minecraft.core.flowerpot.FlowerPotInteraction.getBlockState(level, pos, block);
+                return optional.map(pottedState -> {
+                    InteractionResult interactionResult = onPut(stack, level, pos, player, hand);
+                    if (interactionResult == InteractionResult.SUCCESS) {
+                        flowerPotBlockEntity.setPotted(block, pottedState, player.getDirection().getOpposite());
+                        return InteractionResult.SUCCESS_SERVER;
+                    }
+                    return interactionResult;
+                }).orElse(InteractionResult.CONSUME);
+            }
+        }
+        // KioCG end - custom block entity
+
         if (blockState.isAir()) {
-            return InteractionResult.TRY_WITH_EMPTY_HAND;
-        } else if (!this.isEmpty()) {
+            // KioCG start - custom block entity
             return InteractionResult.CONSUME;
         } else {
+            InteractionResult interactionResult = onPut(stack, level, pos, player, hand);
+            if (interactionResult == InteractionResult.SUCCESS) {
+                level.setBlock(pos, blockState, 3);
+            }
+            return interactionResult;
+        }
+    }
+
+    private InteractionResult onPut(ItemStack stack, Level level, BlockPos pos, Player player, InteractionHand hand) {
+        {
+            // KioCG end - custom block entity
             // Paper start - Add PlayerFlowerPotManipulateEvent
             org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
             org.bukkit.inventory.ItemStack placedStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack);
@@ -78,7 +140,7 @@ public class FlowerPotBlock extends Block {
                 return InteractionResult.CONSUME;
             }
             // Paper end - Add PlayerFlowerPotManipulateEvent
-            level.setBlock(pos, blockState, 3);
+            // level.setBlock(pos, blockState, 3); // KioCG - custom block entity
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
             player.awardStat(Stats.POT_FLOWER);
             stack.consume(1, player);
@@ -89,12 +151,44 @@ public class FlowerPotBlock extends Block {
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         if (this.isEmpty()) {
+            // KioCG start - custom block entity
+            if (level.getBlockEntity(pos) instanceof net.minecraft.world.level.block.entity.FlowerPotBlockEntity flowerPotBlockEntity) {
+                if (!flowerPotBlockEntity.isEmpty()) {
+                    player.containerMenu.sendAllDataToRemote();
+                    InteractionResult interactionResult = onPick(level, pos, player, flowerPotBlockEntity.getPotted());
+                    if (interactionResult == InteractionResult.SUCCESS) {
+                        if (flowerPotBlockEntity.getPottedState().getBlock() instanceof CandleCakeBlock candleCakeBlock) {
+                            ItemStack itemStack = new ItemStack(candleCakeBlock.candleBlock);
+                            if (!player.addItem(itemStack)) {
+                                player.drop(itemStack, false);
+                            }
+                        }
+                        flowerPotBlockEntity.getPottedAndRemove();
+                        return InteractionResult.SUCCESS_SERVER;
+                    }
+                    return interactionResult;
+                }
+            }
+            // KioCG end - custom block entity
             return InteractionResult.CONSUME;
         } else {
-            ItemStack itemStack = new ItemStack(this.potted);
+            // KioCG start - custom block entity
+            InteractionResult interactionResult = onPick(level, pos, player, this.potted);
+            if (interactionResult == InteractionResult.SUCCESS) {
+                level.setBlock(pos, Blocks.FLOWER_POT.defaultBlockState(), 3);
+                player.swing(InteractionHand.MAIN_HAND, true); // 优化原版花盆交互行为
+            }
+            return interactionResult;
+        }
+    }
+
+    private InteractionResult onPick(Level level, BlockPos pos, Player player, Block potted) {
+        {
+            // KioCG end - custom block entity
+            ItemStack itemStack = new ItemStack(potted); // KioCG - custom block entity
             // Paper start - Add PlayerFlowerPotManipulateEvent
             org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
-            org.bukkit.inventory.ItemStack pottedStack = new org.bukkit.inventory.ItemStack(org.bukkit.craftbukkit.block.CraftBlockType.minecraftToBukkit(this.potted));
+            org.bukkit.inventory.ItemStack pottedStack = itemStack.asBukkitCopy(); // KioCG - custom block entity
 
             io.papermc.paper.event.player.PlayerFlowerPotManipulateEvent event = new io.papermc.paper.event.player.PlayerFlowerPotManipulateEvent((org.bukkit.entity.Player) player.getBukkitEntity(), block, pottedStack, false);
             if (!event.callEvent()) {
@@ -108,12 +202,25 @@ public class FlowerPotBlock extends Block {
                 player.drop(itemStack, false);
             }
 
-            level.setBlock(pos, Blocks.FLOWER_POT.defaultBlockState(), 3);
+            // level.setBlock(pos, Blocks.FLOWER_POT.defaultBlockState(), 3); // KioCG - custom block entity
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
             return InteractionResult.SUCCESS;
         }
     }
 
+    // KioCG start - custom block entity: for test
+    private void printAllSupportBlocks(Level level, BlockPos pos) {
+        for (net.minecraft.world.item.Item item : BuiltInRegistries.ITEM) {
+            if (item instanceof BlockItem blockItem) {
+                Block block = blockItem.getBlock();
+                if (!POTTED_BY_CONTENT.containsKey(block) && net.minecraft.core.flowerpot.FlowerPotInteraction.defaultInteraction(level, pos, block.defaultBlockState()) != null) {
+                    org.bukkit.Bukkit.getLogger().info(BuiltInRegistries.BLOCK.wrapAsHolder(block).getRegisteredName());
+                }
+            }
+        }
+    }
+    // KioCG end - custom block entity: for test
+
     @Override
     protected ItemStack getCloneItemStack(LevelReader level, BlockPos pos, BlockState state, boolean includeData) {
         return this.isEmpty() ? super.getCloneItemStack(level, pos, state, includeData) : new ItemStack(this.potted);
@@ -176,4 +283,23 @@ public class FlowerPotBlock extends Block {
             return state.is(Blocks.POTTED_CLOSED_EYEBLOSSOM) ? Blocks.POTTED_OPEN_EYEBLOSSOM.defaultBlockState() : state;
         }
     }
+
+    // KioCG start - custom block entity
+    @Override
+    public boolean hasBlockEntity(BlockState state) {
+        return state.is(Blocks.FLOWER_POT);
+    }
+
+    @Override
+    public net.minecraft.world.level.block.entity.BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
+        return new net.minecraft.world.level.block.entity.FlowerPotBlockEntity(pos, state);
+    }
+
+    @javax.annotation.Nullable
+    @Override
+    public <T extends net.minecraft.world.level.block.entity.BlockEntity> net.minecraft.world.level.block.entity.BlockEntityTicker<T> getTicker(Level level, BlockState state, net.minecraft.world.level.block.entity.BlockEntityType<T> blockEntityType) {
+        return createTickerHelper(blockEntityType, net.minecraft.world.level.block.entity.BlockEntityType.FLOWER_POT,
+                                  (level1, pos, state1, blockEntity) -> net.minecraft.world.level.block.entity.FlowerPotBlockEntity.serverTick((ServerLevel) level1, pos, state1, blockEntity));
+    }
+    // KioCG end - custom block entity
 }
diff --git a/net/minecraft/world/level/block/LayeredCauldronBlock.java b/net/minecraft/world/level/block/LayeredCauldronBlock.java
index 34495f08c3ff3d298a5f5ca9121928cf461caf82..8e63b50bb0b0b7069af18cd95bf841295e3577cf 100644
--- a/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -135,6 +135,17 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
             return false;
         }
 
+        // KioCG start - custom block entity: campfire boiling water
+        BlockState oldState = level.getBlockState(pos);
+        if (oldState.is(newBlock.getBlock())) {
+            if (level.getBlockEntity(pos) instanceof net.minecraft.world.level.block.entity.CauldronBlockEntity cauldronBlockEntity) {
+                int oldLevel = oldState.getValue(LayeredCauldronBlock.LEVEL);
+                int newLevel = newBlock.getValue(LayeredCauldronBlock.LEVEL);
+                cauldronBlockEntity.levelChanged(oldLevel, newLevel);
+            }
+        }
+        // KioCG end - custom block entity: campfire boiling water
+
         newState.place(3);
         if (sendGameEvent) level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newBlock)); // Paper - Call CauldronLevelChangeEvent
         return true;
@@ -171,4 +182,18 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
             level.levelEvent(1047, pos, 0);
         }
     }
+
+    // KioCG start - custom block entity
+    @Override
+    protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @org.jetbrains.annotations.Nullable net.minecraft.world.level.redstone.Orientation orientation, boolean movedByPiston) {
+        if (level.getBlockEntity(pos) instanceof net.minecraft.world.level.block.entity.CauldronBlockEntity cauldron) cauldron.updateBlockEntityTicker(); // KioCG - optimize block entity
+    }
+
+    @javax.annotation.Nullable
+    @Override
+    public <T extends net.minecraft.world.level.block.entity.BlockEntity> net.minecraft.world.level.block.entity.BlockEntityTicker<T> getTicker(Level level, BlockState state, net.minecraft.world.level.block.entity.BlockEntityType<T> blockEntityType) {
+        return createTickerHelper(blockEntityType, net.minecraft.world.level.block.entity.BlockEntityType.CAULDRON,
+                                  (level1, pos, state1, blockEntity) -> net.minecraft.world.level.block.entity.CauldronBlockEntity.serverTickLayered((ServerLevel) level1, pos, state1, blockEntity));
+    }
+    // KioCG end - custom block entity
 }
diff --git a/net/minecraft/world/level/block/entity/BlockEntityType.java b/net/minecraft/world/level/block/entity/BlockEntityType.java
index 386e6a48701b4c9256e33174123381a93d61e292..cd186c9d962fa19400e30fbd6a051327d9bcaa97 100644
--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -285,4 +285,22 @@ public class BlockEntityType<T extends BlockEntity> {
     interface BlockEntitySupplier<T extends BlockEntity> {
         T create(BlockPos pos, BlockState state);
     }
+
+    // KioCG start - custom block entity: 确保最后一个注册以免影响数据包
+    public static final BlockEntityType<CauldronBlockEntity> CAULDRON = register(
+            "cauldron",
+            CauldronBlockEntity::new,
+            Blocks.CAULDRON,
+            Blocks.WATER_CAULDRON,
+            Blocks.LAVA_CAULDRON,
+            Blocks.POWDER_SNOW_CAULDRON
+    );
+    public static final BlockEntityType<FlowerPotBlockEntity> FLOWER_POT = register(
+            "flower_pot",
+            FlowerPotBlockEntity::new,
+            Blocks.FLOWER_POT
+    );
+
+    public static final Set<BlockEntityType> UNABLE_SEND_TO_CLIENT = Set.of(CAULDRON, FLOWER_POT);
+    // KioCG end - custom block entity: 确保最后一个注册以免影响数据包
 }
diff --git a/net/minecraft/world/level/block/entity/CauldronBlockEntity.java b/net/minecraft/world/level/block/entity/CauldronBlockEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..d16c3fb08d8c2e6b326a0ccbbd92ecaf819da919
--- /dev/null
+++ b/net/minecraft/world/level/block/entity/CauldronBlockEntity.java
@@ -0,0 +1,119 @@
+package net.minecraft.world.level.block.entity;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.CampfireBlock;
+import net.minecraft.world.level.block.LayeredCauldronBlock;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.storage.ValueInput;
+import net.minecraft.world.level.storage.ValueOutput;
+
+// KioCG start - custom block entity: all class
+public class CauldronBlockEntity extends BlockEntity {
+    private boolean heating = false;
+    private int temperature = 0;
+    public boolean cleanWater = false;
+
+    public CauldronBlockEntity(BlockPos pos, BlockState blockState) {
+        super(BlockEntityType.CAULDRON, pos, blockState);
+    }
+
+    public static void serverTickLayered(ServerLevel level, BlockPos pos, BlockState state, CauldronBlockEntity cauldron) {
+        boolean flag = false;
+        boolean hasVisualEffect = false;
+
+        if (cauldron.heating) {
+            if (!cauldron.isBoiling()) {
+                int per = LayeredCauldronBlock.MAX_FILL_LEVEL - state.getValue(LayeredCauldronBlock.LEVEL) + 1;
+                cauldron.temperature = Math.min(cauldron.temperature + per, level.paperConfig().kiocgConfig.theLongDark.campfireBoilingWaterTime);
+
+                if (cauldron.isBoiling() && state.is(Blocks.POWDER_SNOW_CAULDRON)) {
+                    if (LayeredCauldronBlock.changeLevel(level, pos, Blocks.WATER_CAULDRON.defaultBlockState(), null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.EVAPORATE, false)) { // avoid duplicate game event
+                        level.setBlockAndUpdate(pos, Blocks.WATER_CAULDRON.defaultBlockState());
+                        level.gameEvent(null, GameEvent.BLOCK_CHANGE, pos);
+                        return;
+                    }
+                }
+
+                flag = true;
+            } else {
+                if (!cauldron.cleanWater) {
+                    cauldron.cleanWater = true;
+                    flag = true;
+                }
+
+                if (level.getGameTime() % 4 == 0 && level.getRandom().nextFloat() < 0.25F) {
+                    level.playSound(null, pos, SoundEvents.POINTED_DRIPSTONE_DRIP_WATER_INTO_CAULDRON, SoundSource.BLOCKS, 1.0F, 0.1F);
+                }
+
+                double x = (double) pos.getX() + 0.5;
+                double y = (double) pos.getY() + 1.0;
+                double z = (double) pos.getZ() + 0.5;
+                level.sendParticles(ParticleTypes.FALLING_WATER, x, y, z, 10, 0.2, 0.1, 0.2, 1.0);
+                hasVisualEffect = true;
+            }
+        } else if (cauldron.temperature > 0) {
+            int per = LayeredCauldronBlock.MAX_FILL_LEVEL - state.getValue(LayeredCauldronBlock.LEVEL) + 1;
+            cauldron.temperature = Math.max(cauldron.temperature - per, 0);
+            flag = true;
+        }
+
+        if (flag) {
+            setChanged(level, pos, state);
+        } else if (!hasVisualEffect) {
+            cauldron.updateBlockEntityTicker();
+        }
+    }
+
+    // KioCG start - optimize block entity
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        if (state.getBlock() instanceof LayeredCauldronBlock) {
+            this.heating = false;
+            BlockState belowState = level.getBlockState(pos.below());
+            if (belowState.is(BlockTags.CAMPFIRES, blockStateBase -> blockStateBase.getValueOrElse(CampfireBlock.LIT, false)) || belowState.is(Blocks.FIRE) || belowState.is(Blocks.LAVA)) {
+                this.heating = true;
+                return true;
+            } else {
+                return this.temperature > 0;
+            }
+        }
+        return false;
+    }
+
+    public boolean isBoiling() {
+        return this.temperature >= this.level.paperConfig().kiocgConfig.theLongDark.campfireBoilingWaterTime;
+    }
+
+    public void setTemperature(int temperature) {
+        this.temperature = temperature;
+        this.setChanged();
+    }
+
+    public void levelChanged(int oldLevel, int newLevel) {
+        if (newLevel > oldLevel) {
+            setTemperature(this.temperature * oldLevel / newLevel);
+            this.cleanWater = false;
+        }
+    }
+
+    @Override
+    protected void loadAdditional(ValueInput input) {
+        this.temperature = input.getIntOr("KioCG.Temperature", 0);
+        this.cleanWater = input.getBooleanOr("KioCG.CleanWater", false);
+    }
+
+    @Override
+    protected void saveAdditional(ValueOutput output) {
+        output.putInt("KioCG.Temperature", this.temperature);
+        output.putBoolean("KioCG.CleanWater", this.cleanWater);
+    }
+}
+// KioCG end - custom block entity: all class
diff --git a/net/minecraft/world/level/block/entity/FlowerPotBlockEntity.java b/net/minecraft/world/level/block/entity/FlowerPotBlockEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..bfcf84db441b886fe59944c1a594db005db1002d
--- /dev/null
+++ b/net/minecraft/world/level/block/entity/FlowerPotBlockEntity.java
@@ -0,0 +1,206 @@
+package net.minecraft.world.level.block.entity;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.core.flowerpot.FlowerPotInteraction;
+import net.minecraft.core.particles.DustParticleOptions;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.core.particles.SimpleParticleType;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.world.entity.Display;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.AbstractCandleBlock;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.RedstoneTorchBlock;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.storage.ValueInput;
+import net.minecraft.world.level.storage.ValueOutput;
+import net.minecraft.world.phys.Vec3;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+// KioCG start - custom block entity: all class
+public class FlowerPotBlockEntity extends BlockEntity {
+    private Block potted = Blocks.AIR;
+    private BlockState pottedState = potted.defaultBlockState();
+    private PottedTick pottedTick = PottedTick.NONE;
+    private Direction facing = Direction.SOUTH;
+
+    public FlowerPotBlockEntity(BlockPos pos, BlockState blockState) {
+        super(BlockEntityType.FLOWER_POT, pos, blockState);
+    }
+
+    public static void serverTick(ServerLevel level, BlockPos pos, BlockState state, FlowerPotBlockEntity flowerPot) {
+        flowerPot.pottedTick.tick(level, pos, flowerPot.pottedState);
+    }
+
+    @Override
+    public boolean needTicking(Level level, BlockPos pos, BlockState state) {
+        return this.pottedTick != PottedTick.NONE;
+    }
+
+    public Block getPotted() {
+        return this.potted;
+    }
+
+    public BlockState getPottedState() {
+        return this.pottedState;
+    }
+
+    public Block getPottedAndRemove() {
+        Block temp = this.potted;
+        this.setPotted(Blocks.AIR, Blocks.AIR.defaultBlockState(), Direction.SOUTH);
+        return temp;
+    }
+
+    public boolean isEmpty() {
+        return this.potted == Blocks.AIR;
+    }
+
+    public void updatePottedState(BlockState pottedState) {
+        setPotted(this.potted, pottedState, this.facing);
+    }
+
+    public void setPotted(Block potted, BlockState pottedState, Direction facing) {
+        this.potted = potted;
+        this.pottedState = pottedState;
+        this.pottedTick = PottedTick.pottedTick(this);
+        this.facing = facing;
+        if (this.hasLevel()) {
+            this.refreshDisplayBlock((ServerLevel) this.getLevel(), this.getBlockPos());
+            this.updateBlockEntityTicker();
+        }
+        this.setChanged();
+    }
+
+    private void refreshDisplayBlock(ServerLevel level, BlockPos pos) {
+        if (this.isEmpty()) {
+            Display.BlockDisplay.discardCustomDisplay(EntityType.BLOCK_DISPLAY, level, pos);
+            return;
+        }
+
+        Optional<Consumer<Display.BlockDisplay>> optional = FlowerPotInteraction.getInteract(level, pos, this.getPotted(), this.getPottedState());
+        optional.ifPresent(consumer -> {
+            consumer = consumer.andThen(display -> display.setYRot(facing.toYRot()));
+            Display.BlockDisplay displayBlock = Display.BlockDisplay.getCustomDisplay(EntityType.BLOCK_DISPLAY, level, pos);
+            if (displayBlock != null) {
+                consumer.accept(displayBlock);
+            } else {
+                Display.BlockDisplay.createCustomDisplay(EntityType.BLOCK_DISPLAY, level, pos, consumer);
+            }
+        });
+    }
+
+    @Override
+    public void preRemoveSideEffects(BlockPos pos, BlockState state) {
+        if (this.level != null && !this.isEmpty()) {
+            Block.popResource(this.level, pos, new net.minecraft.world.item.ItemStack(this.getPottedAndRemove()));
+        }
+    }
+
+    @Override
+    protected void loadAdditional(ValueInput input) {
+        Block potted = input.read("KioCG.Potted", BuiltInRegistries.BLOCK.byNameCodec()).orElse(Blocks.AIR);
+        BlockState pottedState = input.read("KioCG.PottedState", BlockState.CODEC).orElse(Blocks.AIR.defaultBlockState());
+        Direction facing = input.read("KioCG.Facing", Direction.CODEC).orElse(net.minecraft.core.Direction.SOUTH);
+        this.setPotted(potted, pottedState, facing);
+    }
+
+    @Override
+    protected void saveAdditional(ValueOutput output) {
+        if (!this.isEmpty()) {
+            output.store("KioCG.Potted", BuiltInRegistries.BLOCK.byNameCodec(), this.getPotted());
+            output.store("KioCG.PottedState", BlockState.CODEC, this.getPottedState());
+            output.store("KioCG.Facing", Direction.CODEC, this.facing);
+        }
+    }
+
+    @Override
+    public void setLevel(Level level) {
+        super.setLevel(level);
+        this.refreshDisplayBlock((ServerLevel) level, this.getBlockPos());
+    }
+
+    private interface PottedTick {
+        PottedTick NONE = (level, pos, pottedState) -> {
+        };
+
+        void tick(ServerLevel level, BlockPos pos, BlockState pottedState);
+
+        private static PottedTick pottedTick(FlowerPotBlockEntity flowerPot) {
+            if (!flowerPot.isEmpty()) {
+                BlockState pottedState = flowerPot.getPottedState();
+                if (pottedState.is(Blocks.TORCH)) {
+                    return (level, pos, state) -> {
+                        PottedTick.torchTick(level, pos, state, ParticleTypes.FLAME);
+                    };
+                } else if (pottedState.is(Blocks.SOUL_TORCH)) {
+                    return (level, pos, state) -> {
+                        PottedTick.torchTick(level, pos, state, ParticleTypes.SOUL_FIRE_FLAME);
+                    };
+                } else if (pottedState.is(Blocks.REDSTONE_TORCH)) {
+                    if (pottedState.getValue(RedstoneTorchBlock.LIT)) {
+                        return PottedTick::redstoneTorchTick;
+                    }
+                } else if (pottedState.is(Blocks.SPORE_BLOSSOM)) {
+                    return PottedTick::sporeBlossomTick;
+                } else if (pottedState.is(Blocks.FIREFLY_BUSH)) {
+                    return PottedTick::fireflyBushTick;
+                } else {
+                    boolean isCandle = pottedState.is(BlockTags.CANDLES);
+                    if (isCandle || pottedState.is(BlockTags.CANDLE_CAKES)) {
+                        if (pottedState.getValueOrElse(AbstractCandleBlock.LIT, false)) {
+                            return (level, pos, state) -> {
+                                PottedTick.candleTick(level, pos, state, isCandle ? 0.5 + 0.25 : 0.65);
+                            };
+                        }
+                    }
+                }
+            }
+            return NONE;
+        }
+
+        private static void torchTick(ServerLevel level, BlockPos pos, BlockState pottedState, SimpleParticleType flameParticle) {
+            if (level.random.nextFloat() < (float) (1.0 / 8.0)) {
+                double d = pos.getX() + 0.5;
+                double d1 = pos.getY() + 0.7 + 0.25;
+                double d2 = pos.getZ() + 0.5;
+                level.sendParticles(ParticleTypes.SMOKE, d, d1, d2, 1, 0.0, 0.0, 0.0, 0.0);
+                level.sendParticles(flameParticle, d, d1, d2, 1, 0.0, 0.0, 0.0, 0.0);
+            }
+        }
+
+        private static void redstoneTorchTick(ServerLevel level, BlockPos pos, BlockState pottedState) {
+            if (level.random.nextFloat() < (float) (1.0 / 8.0)) {
+                double d = pos.getX() + 0.5 + (level.random.nextDouble() - 0.5) * 0.2;
+                double d1 = pos.getY() + 0.7 + 0.25 + (level.random.nextDouble() - 0.5) * 0.2;
+                double d2 = pos.getZ() + 0.5 + (level.random.nextDouble() - 0.5) * 0.2;
+                level.sendParticles(DustParticleOptions.REDSTONE, d, d1, d2, 1, 0.0, 0.0, 0.0, 0.0);
+            }
+        }
+
+        private static void candleTick(ServerLevel level, BlockPos pos, BlockState pottedState, double yoffset) {
+            if (level.random.nextFloat() < (float) (1.0 / 8.0)) {
+                AbstractCandleBlock.addParticlesAndSound(level, new Vec3(pos.getX() + 0.5, pos.getY() + yoffset, pos.getZ() + 0.5), level.random);
+            }
+        }
+
+        private static void sporeBlossomTick(ServerLevel level, BlockPos pos, BlockState pottedState) {
+            if (level.random.nextFloat() < (float) (1.0 / 20.0)) {
+                level.sendParticles(ParticleTypes.SPORE_BLOSSOM_AIR, pos.getX() + 0.5, pos.getY() + 1.0, pos.getZ() + 0.5, 2, 0.5, 0.5, 0.5, 0.0);
+            }
+        }
+
+        private static void fireflyBushTick(ServerLevel level, BlockPos pos, BlockState pottedState) {
+            if (level.random.nextFloat() < (float) (1.0 / 8.0)) {
+                pottedState.getBlock().animateTick(pottedState, level, pos, level.random);
+            }
+        }
+    }
+}
+// KioCG end - custom block entity: all class
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 627f260b9ae640a4baae1a33e60fd98da1e60c0e..7c5bfc6738e8248bb8f6cf196b6a813eb7414b95 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1047,12 +1047,12 @@ public abstract class BlockBehaviour implements FeatureElement {
         }
 
         public boolean hasBlockEntity() {
-            return this.getBlock() instanceof EntityBlock;
+            return this.getBlock().hasBlockEntity(this.asState()); // KioCG - custom block entity
         }
 
         @Nullable
         public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level level, BlockEntityType<T> blockEntityType) {
-            return this.getBlock() instanceof EntityBlock ? ((EntityBlock)this.getBlock()).getTicker(level, this.asState(), blockEntityType) : null;
+            return hasBlockEntity() ? ((EntityBlock)this.getBlock()).getTicker(level, this.asState(), blockEntityType) : null; // KioCG - custom block entity
         }
 
         public boolean is(Block block) {
