From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KioProject <1105919949@qq.com>
Date: Sun, 27 Apr 2025 19:44:51 +0800
Subject: [PATCH] YUNGs package


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 9c9b601a3f903bebb0dd1bda0e24745587229727..b0d70914386226c852c75a211fa6b5ae04944f7a 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -314,6 +314,8 @@ public class Main {
                 return;
             }
 
+            com.yungnickyoung.minecraft.yungsextras.module.BiomeModificationModuleFabric.init(worldStem.registries().compositeAccess()); // KioCG
+
             /*
             RegistryAccess.Frozen frozen = worldStem.registries().compositeAccess();
             WorldData worldData = worldStem.worldData();
diff --git a/net/minecraft/world/entity/monster/ElderGuardian.java b/net/minecraft/world/entity/monster/ElderGuardian.java
index fd33a8b59f40299ab644a4c52921b66a9b6552ca..8ea80e1cbe252509815328d246f81134024bdc93 100644
--- a/net/minecraft/world/entity/monster/ElderGuardian.java
+++ b/net/minecraft/world/entity/monster/ElderGuardian.java
@@ -63,7 +63,7 @@ public class ElderGuardian extends Guardian {
     @Override
     protected void customServerAiStep(ServerLevel level) {
         super.customServerAiStep(level);
-        if ((this.tickCount + this.getId()) % 1200 == 0) {
+        if ((this.tickCount + this.getId()) % 100 == 0) { // KioCG
             MobEffectInstance mobEffectInstance = new MobEffectInstance(MobEffects.MINING_FATIGUE, 6000, 2);
             List<ServerPlayer> list = MobEffectUtil.addEffectToPlayersAround(level, this, this.position(), 50.0, mobEffectInstance, 1200, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK, (player) -> new io.papermc.paper.event.entity.ElderGuardianAppearanceEvent((org.bukkit.entity.ElderGuardian) this.getBukkitEntity(), player.getBukkitEntity()).callEvent()); // CraftBukkit // Paper - Add ElderGuardianAppearanceEvent
             list.forEach(
diff --git a/net/minecraft/world/entity/monster/Husk.java b/net/minecraft/world/entity/monster/Husk.java
index 8d6def85583a111841b28f20f58ddb8b8cbd7bc1..885a75f1d83e895dd9c9e22a549eb0833323916f 100644
--- a/net/minecraft/world/entity/monster/Husk.java
+++ b/net/minecraft/world/entity/monster/Husk.java
@@ -52,6 +52,43 @@ public class Husk extends Zombie {
         return SoundEvents.HUSK_STEP;
     }
 
+    // KioCG start
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        effectNearbyPlayers(this.serverLevel());
+    }
+
+    @Override
+    protected void customServerAiStep(ServerLevel level) {
+        super.customServerAiStep(level);
+        effectNearbyPlayers(level);
+    }
+
+    @org.jetbrains.annotations.Nullable Boolean isPharaoh = null;
+    private void effectNearbyPlayers(ServerLevel level) {
+        if (isPharaoh == null) {
+            isPharaoh = com.yungnickyoung.minecraft.betterdeserttemples.util.PharaohUtil.isPharaoh(this);
+        }
+        if (isPharaoh && com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.CONFIG.general.applyMiningFatigue) {
+            if ((this.tickCount + this.getId()) % 100 == 0) {
+                java.util.function.Predicate<net.minecraft.server.level.ServerPlayer> playerPredicate = serverPlayer -> {
+                    ServerLevel playerLevel = serverPlayer.serverLevel();
+                    BlockPos playerPos = serverPlayer.blockPosition();
+                    // Do not apply mining fatigue if player is not in temple
+                    return playerLevel.isLoaded(playerPos) && playerLevel.structureManager().getStructureWithPieceAt(playerPos, com.yungnickyoung.minecraft.betterdeserttemples.module.TagModule.APPLIES_MINING_FATIGUE).isValid();
+                };
+                MobEffectInstance mobEffectInstance = new MobEffectInstance(MobEffects.MINING_FATIGUE, 6000, 2);
+                java.util.List<net.minecraft.server.level.ServerPlayer> list = net.minecraft.world.effect.MobEffectUtil.addEffectToPlayersAround(level, this, this.position(), 100.0, mobEffectInstance, 1200, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK, playerPredicate);
+                list.forEach(
+                        serverPlayer -> serverPlayer.connection
+                                .send(new net.minecraft.network.protocol.game.ClientboundSoundPacket(net.minecraft.core.registries.BuiltInRegistries.SOUND_EVENT.wrapAsHolder(net.minecraft.sounds.SoundEvents.ELDER_GUARDIAN_CURSE), net.minecraft.sounds.SoundSource.HOSTILE, serverPlayer.getX(), serverPlayer.getY(), serverPlayer.getZ(), this.isSilent() ? 0.0F : 1.0F, 1.0F, this.getRandom().nextLong()))
+                            );
+            }
+        }
+    }
+    // KioCG end
+
     @Override
     public boolean doHurtTarget(ServerLevel level, Entity source) {
         boolean flag = super.doHurtTarget(level, source);
diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 2e8d70c7db9876d59e73d718c06682a83e1c4ccd..fca22e24047e3bdec6c444aa160e3c5e096fdb3b 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -128,7 +128,7 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     protected void setOwnerThroughUUID(@Nullable UUID uuid) {
         if (!Objects.equals(this.ownerUUID, uuid)) {
             this.ownerUUID = uuid;
-            this.cachedOwner = uuid != null ? this.findOwner(uuid) : null;
+            this.cachedOwner = null; // KioCG
         }
     }
 
diff --git a/net/minecraft/world/entity/projectile/ThrownTrident.java b/net/minecraft/world/entity/projectile/ThrownTrident.java
index afa8ce32b3d950f30dd1f60933a7119645e70198..5721d7c8e827e2d32bbd1322622e38bd95eb3fa6 100644
--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -301,12 +301,28 @@ public class ThrownTrident extends AbstractArrow {
 
     @Override
     public void tickDespawn() {
+        if (moderOwner()) return; // KioCG
         int i = this.entityData.get(ID_LOYALTY);
         if (this.pickup != AbstractArrow.Pickup.ALLOWED || i <= 0) {
             super.tickDespawn();
         }
     }
 
+    // KioCG start
+    private static final java.util.UUID moderUUID = java.util.UUID.fromString("e624cdc1-c238-4dde-9f22-1f76b5123ce8");
+
+    /**
+     * Prevents trident from despawning if it's within a Better Ocean Monument and
+     * its owner UUID is my own, meaning this behavior should never be invoked for tridents
+     * thrown by ordinary players.
+     * This is done so that any thrown tridents placed as part of the structure never despawn,
+     * but they will despawn as normal once picked up and thrown again by a player.
+     */
+    private boolean moderOwner() {
+        return moderUUID.equals(this.ownerUUID);
+    }
+    // KioCG end
+
     @Override
     protected float getWaterInertia() {
         return 0.99F;
diff --git a/net/minecraft/world/level/biome/Biome.java b/net/minecraft/world/level/biome/Biome.java
index bad1a03167f7586e5279592adcb43350c9b528cd..9759922b95d7885f65b1ec41a7c0617ac5632e84 100644
--- a/net/minecraft/world/level/biome/Biome.java
+++ b/net/minecraft/world/level/biome/Biome.java
@@ -65,7 +65,7 @@ public final class Biome {
     public static final PerlinSimplexNoise BIOME_INFO_NOISE = new PerlinSimplexNoise(new WorldgenRandom(new LegacyRandomSource(2345L)), ImmutableList.of(0));
     private static final int TEMPERATURE_CACHE_SIZE = 1024;
     public final Biome.ClimateSettings climateSettings;
-    private final BiomeGenerationSettings generationSettings;
+    public BiomeGenerationSettings generationSettings; // KioCG
     private final MobSpawnSettings mobSettings;
     private final BiomeSpecialEffects specialEffects;
     private final ThreadLocal<Long2FloatLinkedOpenHashMap> temperatureCache = ThreadLocal.withInitial(() -> Util.make(() -> {
diff --git a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
index e2abd838252a50618789123477a96545454621b4..9501ab81873eb16e515d1bd098ecdf54a3845bf0 100644
--- a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -155,6 +155,8 @@ public class JukeboxBlockEntity extends BlockEntity implements ContainerSingleIt
     public void setTheItem(ItemStack item) {
         this.item = item;
         boolean flag = !this.item.isEmpty();
+
+        if (this.level == null) return; // KioCG
         Optional<Holder<JukeboxSong>> optional = JukeboxSong.fromStack(this.level.registryAccess(), this.item);
         this.notifyItemChangedInJukebox(flag);
         if (flag && optional.isPresent()) {
diff --git a/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java b/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java
index 619b98e42e254c0c260c171a26a2472ddf59b885..c8dceaa0fd8e3e31d40d71268c745744db511d88 100644
--- a/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java
+++ b/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java
@@ -34,7 +34,7 @@ import org.slf4j.Logger;
 public class ChunkGeneratorStructureState {
     private static final Logger LOGGER = LogUtils.getLogger();
     private final RandomState randomState;
-    private final BiomeSource biomeSource;
+    public final BiomeSource biomeSource; // KioCG - private -> public
     private final long levelSeed;
     private final long concentricRingsSeed;
     private final Map<Structure, List<StructurePlacement>> placementsForStructure = new Object2ObjectOpenHashMap<>();
diff --git a/net/minecraft/world/level/levelgen/Beardifier.java b/net/minecraft/world/level/levelgen/Beardifier.java
index 74d8202b5c9bb2a3ee832be70f95c0b5cbecb460..eb1137a9790ec7ee377dc766752de0a47cb18934 100644
--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -27,8 +27,38 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
             }
         }
     });
-    private final ObjectListIterator<Beardifier.Rigid> pieceIterator;
-    private final ObjectListIterator<JigsawJunction> junctionIterator;
+    public final ObjectListIterator<Beardifier.Rigid> pieceIterator; // KioCG - private -> public
+    public final ObjectListIterator<JigsawJunction> junctionIterator; // KioCG - private -> public
+
+    // KioCG start
+    public ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedJigsawJunction> enhancedJunctionIterator;
+    public ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedBeardifierRigid> enhancedPieceIterator;
+    public NoiseChunk noiseChunk;
+
+    public ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedBeardifierRigid> getEnhancedPieceIterator() {
+        return this.enhancedPieceIterator;
+    }
+
+    public void setEnhancedPieceIterator(ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedBeardifierRigid> enhancedPieceIterator) {
+        this.enhancedPieceIterator = enhancedPieceIterator;
+    }
+
+    public ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedJigsawJunction> getEnhancedJunctionIterator() {
+        return enhancedJunctionIterator;
+    }
+
+    public void setEnhancedJunctionIterator(ObjectListIterator<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedJigsawJunction> enhancedJunctionIterator) {
+        this.enhancedJunctionIterator = enhancedJunctionIterator;
+    }
+
+    public NoiseChunk getNoiseChunk() {
+        return this.noiseChunk;
+    }
+
+    public void setNoiseChunk(NoiseChunk noiseChunk) {
+        this.noiseChunk = noiseChunk;
+    }
+    // KioCG end
 
     public static Beardifier forStructuresInChunk(StructureManager structureManager, ChunkPos chunkPos) {
         int minBlockX = chunkPos.getMinBlockX();
@@ -66,7 +96,7 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
                         }
                     }
         } // Paper - Perf: Remove streams from hot code
-        return new Beardifier(list.iterator(), list1.iterator());
+        return com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedBeardifierHelper.forStructuresInChunk(structureManager, chunkPos, new Beardifier(list.iterator(), list1.iterator())); // KioCG
     }
 
     @VisibleForTesting
@@ -117,7 +147,7 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
         }
 
         this.junctionIterator.back(Integer.MAX_VALUE);
-        return d;
+        return com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.beardifier.EnhancedBeardifierHelper.computeDensity(context, d, this); // KioCG
     }
 
     @Override
diff --git a/net/minecraft/world/level/levelgen/NoiseChunk.java b/net/minecraft/world/level/levelgen/NoiseChunk.java
index f861f9e087182470a3bbb22678dbdacb8a73e943..f60e3d7598b003dc0d2fe43dc69e0307b589ad02 100644
--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -21,7 +21,7 @@ import net.minecraft.world.level.levelgen.blending.Blender;
 import net.minecraft.world.level.levelgen.material.MaterialRuleList;
 
 public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunction.FunctionContext {
-    private final NoiseSettings noiseSettings;
+    public final NoiseSettings noiseSettings; // KioCG - private -> public
     final int cellCountXZ;
     final int cellCountY;
     final int cellNoiseMinY;
@@ -78,6 +78,11 @@ public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunct
         }
     };
 
+    // KioCG start
+    private ThreadLocal<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.AquiferOverrideMask> aquiferOverrideMask = new ThreadLocal<>();
+    private BlockState defaultBlockState;
+    // KioCG end
+
     public static NoiseChunk forChunk(
         ChunkAccess chunk,
         RandomState state,
@@ -158,6 +163,13 @@ public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunct
 
         this.blockStateRule = new MaterialRuleList(list.toArray(new NoiseChunk.BlockStateFiller[0]));
         this.initialDensityNoJaggedness = noiseRouter1.initialDensityWithoutJaggedness();
+
+        // KioCG start
+        if (this.beardifier instanceof Beardifier beardifier1 && beardifier1.getNoiseChunk() == null) {
+            beardifier1.setNoiseChunk((NoiseChunk) (Object) this);
+        }
+        this.defaultBlockState = noiseGeneratorSettings.defaultBlock();
+        // KioCG end
     }
 
     protected Climate.Sampler cachedClimateSampler(NoiseRouter noiseRouter, List<Climate.ParameterPoint> points) {
@@ -174,7 +186,28 @@ public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunct
 
     @Nullable
     protected BlockState getInterpolatedState() {
-        return this.blockStateRule.calculate(this);
+        // KioCG start
+        final BlockState calculate = this.blockStateRule.calculate(this);
+        if (calculate != null && (calculate.is(net.minecraft.world.level.block.Blocks.WATER) || calculate.is(net.minecraft.world.level.block.Blocks.LAVA))) {
+            com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.AquiferOverrideMask mask = this.getOrCreateAquiferOverrideMask(() -> new com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.AquiferOverrideMask(this.noiseSettings.height(), this.noiseSettings.minY()));
+
+            // Special handling for SolidifyAquiferOverride
+            if (mask.getAquiferOverride() instanceof com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.SolidifyAquiferOverride solidifyAquiferOverride) {
+                solidifyAquiferOverride.setSolidBlockState(this.defaultBlockState);
+            }
+
+            BlockState blockState = mask.getBlockStateForPos(this.blockX(), this.blockY(), this.blockZ(), calculate);
+            return blockState;
+        }
+        return calculate;
+    }
+
+    public com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.AquiferOverrideMask getOrCreateAquiferOverrideMask(java.util.function.Supplier<com.yungnickyoung.minecraft.yungsapi.world.structure.terrainadaptation.aquiferoverride.AquiferOverrideMask> aquiferOverrideMaskSupplier) {
+        if (this.aquiferOverrideMask.get() == null) {
+            this.aquiferOverrideMask.set(aquiferOverrideMaskSupplier.get());
+        }
+        return this.aquiferOverrideMask.get();
+        // KioCG end
     }
 
     @Override
diff --git a/net/minecraft/world/level/levelgen/feature/BasaltColumnsFeature.java b/net/minecraft/world/level/levelgen/feature/BasaltColumnsFeature.java
index e5eb87a49042ac4ea960d86eea1b3f5a962dbc44..b8534f5d8fd5a45c628e13e206d9877358708ba4 100644
--- a/net/minecraft/world/level/levelgen/feature/BasaltColumnsFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BasaltColumnsFeature.java
@@ -109,6 +109,14 @@ public class BasaltColumnsFeature extends Feature<ColumnFeatureConfiguration> {
     }
 
     private static boolean canPlaceAt(LevelAccessor level, int seaLevel, BlockPos.MutableBlockPos pos) {
+        // KioCG start
+        if (level instanceof net.minecraft.server.level.WorldGenRegion worldGenRegion) {
+            if (com.yungnickyoung.minecraft.yungsapi.util.MixinUtils.isPositionInTaggedStructure(worldGenRegion, pos, com.yungnickyoung.minecraft.yungsapi.module.TagModule.NO_BASALT)) {
+                return false;
+            }
+        }
+        // KioCG end
+
         if (!isAirOrLavaOcean(level, seaLevel, pos)) {
             return false;
         } else {
diff --git a/net/minecraft/world/level/levelgen/feature/DeltaFeature.java b/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
index b57460d2eb3e8cb312d0c24a23bd5c2a6baf18f5..d9b94378f63398c5acaf588f44f2b1c84851a657 100644
--- a/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
@@ -25,6 +25,14 @@ public class DeltaFeature extends Feature<DeltaFeatureConfiguration> {
 
     @Override
     public boolean place(FeaturePlaceContext<DeltaFeatureConfiguration> context) {
+        // KioCG start
+        if (context.level() instanceof net.minecraft.server.level.WorldGenRegion worldGenRegion) {
+            if (com.yungnickyoung.minecraft.yungsapi.util.MixinUtils.isPositionInTaggedStructure(worldGenRegion, context.origin(), com.yungnickyoung.minecraft.yungsapi.module.TagModule.NO_DELTA)) {
+                return false;
+            }
+        }
+        // KioCG end
+
         boolean flag = false;
         RandomSource randomSource = context.random();
         WorldGenLevel worldGenLevel = context.level();
diff --git a/net/minecraft/world/level/levelgen/feature/Feature.java b/net/minecraft/world/level/levelgen/feature/Feature.java
index c5d3d476ef3704061681337475d6e000bb1ee66e..406a2312782dffa21095be0d02881cf6361842cb 100644
--- a/net/minecraft/world/level/levelgen/feature/Feature.java
+++ b/net/minecraft/world/level/levelgen/feature/Feature.java
@@ -158,6 +158,23 @@ public abstract class Feature<FC extends FeatureConfiguration> {
         "pointed_dripstone", new PointedDripstoneFeature(PointedDripstoneConfiguration.CODEC)
     );
     public static final Feature<SculkPatchConfiguration> SCULK_PATCH = register("sculk_patch", new SculkPatchFeature(SculkPatchConfiguration.CODEC));
+
+    // KioCG start
+    /* Desert Features */
+    public static Feature<?> YEX_DESERT_WELL = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "desert_well", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.DESERT_WELL);
+    public static Feature<?> YEX_DESERT_OBELISK = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "desert_obelisk", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.DESERT_OBELISK);
+    public static Feature<?> YEX_DESERT_GIANT_TORCH = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "desert_giant_torch", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.DESERT_GIANT_TORCH);
+    public static Feature<?> YEX_DESERT_RUINS_0 = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "desert_ruins_0", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.DESERT_RUINS_0);
+    public static Feature<?> YEX_DESERT_CHILLZONE = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "desert_chillzone", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.DESERT_CHILLZONE);
+    /* Swamp Features */
+    public static Feature<?> YEX_SWAMP_PILLAR = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_pillar", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_PILLAR);
+    public static Feature<?> YEX_SWAMP_OGRE = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_ogre", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_OGRE);
+    public static Feature<?> YEX_SWAMP_CUBBY = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_cubby", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_CUBBY);
+    public static Feature<?> YEX_SWAMP_ARCH = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_arch", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_ARCH);
+    public static Feature<?> YEX_SWAMP_DOUBLE_ARCH = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_double_arch", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_DOUBLE_ARCH);
+    public static Feature<?> YEX_SWAMP_CHURCH = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "swamp_church", com.yungnickyoung.minecraft.yungsextras.module.FeatureModule.SWAMP_CHURCH);
+    // KioCG end
+
     private final MapCodec<ConfiguredFeature<FC, Feature<FC>>> configuredCodec;
 
     private static <C extends FeatureConfiguration, F extends Feature<C>> F register(String key, F value) {
diff --git a/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java b/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
index 5934fb6a6d4ddb7d8bd5d697ce63d655bc07aed6..d65600f362b01c24749b53cc76f9b6e10ba9999d 100644
--- a/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
@@ -23,6 +23,14 @@ public class UnderwaterMagmaFeature extends Feature<UnderwaterMagmaConfiguration
 
     @Override
     public boolean place(FeaturePlaceContext<UnderwaterMagmaConfiguration> context) {
+        // KioCG start
+        if (context.level() instanceof net.minecraft.server.level.WorldGenRegion worldGenRegion) {
+            if (com.yungnickyoung.minecraft.yungsapi.util.MixinUtils.isPositionInTaggedStructure(worldGenRegion, context.origin(), com.yungnickyoung.minecraft.yungsapi.module.TagModule.NO_MAGMA)) {
+                return false;
+            }
+        }
+        // KioCG end
+
         WorldGenLevel worldGenLevel = context.level();
         BlockPos blockPos = context.origin();
         UnderwaterMagmaConfiguration underwaterMagmaConfiguration = context.config();
diff --git a/net/minecraft/world/level/levelgen/feature/VinesFeature.java b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
index 231f91a96cb2ab2e9d246bb57704d6364979e2aa..237d5300c6340eb23df7fc491a22be5536b6926f 100644
--- a/net/minecraft/world/level/levelgen/feature/VinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
@@ -15,6 +15,18 @@ public class VinesFeature extends Feature<NoneFeatureConfiguration> {
 
     @Override
     public boolean place(FeaturePlaceContext<NoneFeatureConfiguration> context) {
+        // KioCG start
+        if (context.level() instanceof net.minecraft.server.level.WorldGenRegion worldGenRegion) {
+            BlockPos.MutableBlockPos mutable = new BlockPos.MutableBlockPos();
+            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                mutable.setWithOffset(context.origin(), direction);
+                if (com.yungnickyoung.minecraft.yungsapi.util.MixinUtils.isPositionInTaggedStructure(worldGenRegion, mutable, com.yungnickyoung.minecraft.yungsapi.module.TagModule.NO_VINES)) {
+                    return false;
+                }
+            }
+        }
+        // KioCG end
+
         WorldGenLevel worldGenLevel = context.level();
         BlockPos blockPos = context.origin();
         context.config();
diff --git a/net/minecraft/world/level/levelgen/placement/PlacementModifierType.java b/net/minecraft/world/level/levelgen/placement/PlacementModifierType.java
index bc2076ddc8ad65d414cfd1b6733a1a94c5f2d6b6..92027062a627f124a88f54d066f7a63d144431b7 100644
--- a/net/minecraft/world/level/levelgen/placement/PlacementModifierType.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacementModifierType.java
@@ -23,6 +23,14 @@ public interface PlacementModifierType<P extends PlacementModifier> {
     PlacementModifierType<RandomOffsetPlacement> RANDOM_OFFSET = register("random_offset", RandomOffsetPlacement.CODEC);
     PlacementModifierType<FixedPlacement> FIXED_PLACEMENT = register("fixed_placement", FixedPlacement.CODEC);
 
+    // KioCG start
+    PlacementModifierType<?> YEX_RNG_INITIALIZER = register(com.yungnickyoung.minecraft.yungsextras.YungsExtrasCommon.MOD_ID_ + "rng_initializer", com.yungnickyoung.minecraft.yungsextras.module.PlacementModifierTypeModule.RNG_INITIALIZER);
+
+    private static <P extends PlacementModifier> PlacementModifierType<P> register(String name, PlacementModifierType<P> codec) {
+        return Registry.register(BuiltInRegistries.PLACEMENT_MODIFIER_TYPE, name, codec);
+    }
+    // KioCG end
+
     MapCodec<P> codec();
 
     private static <P extends PlacementModifier> PlacementModifierType<P> register(String name, MapCodec<P> codec) {
diff --git a/net/minecraft/world/level/levelgen/structure/BoundingBox.java b/net/minecraft/world/level/levelgen/structure/BoundingBox.java
index dae141a072a03ef7075a389bad630694f08fbc40..6856684d2003ef70ce796519571a6759552a5cf3 100644
--- a/net/minecraft/world/level/levelgen/structure/BoundingBox.java
+++ b/net/minecraft/world/level/levelgen/structure/BoundingBox.java
@@ -303,4 +303,30 @@ public class BoundingBox {
     public int maxZ() {
         return this.maxZ;
     }
+
+    // KioCG start
+    public void setMinX(int value) {
+        this.minX = value;
+    }
+
+    public void setMinY(int value) {
+        this.minY = value;
+    }
+
+    public void setMinZ(int value) {
+        this.minZ = value;
+    }
+
+    public void setMaxX(int value) {
+        this.maxX = value;
+    }
+
+    public void setMaxY(int value) {
+        this.maxY = value;
+    }
+
+    public void setMaxZ(int value) {
+        this.maxZ = value;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/level/levelgen/structure/StructureType.java b/net/minecraft/world/level/levelgen/structure/StructureType.java
index eaabab8a7f822d0abed72dbb27363307e5889179..de3c152c6aa21cd14a35f9dad0b6100b388f8f11 100644
--- a/net/minecraft/world/level/levelgen/structure/StructureType.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureType.java
@@ -38,6 +38,19 @@ public interface StructureType<S extends Structure> {
     StructureType<SwampHutStructure> SWAMP_HUT = register("swamp_hut", SwampHutStructure.CODEC);
     StructureType<WoodlandMansionStructure> WOODLAND_MANSION = register("woodland_mansion", WoodlandMansionStructure.CODEC);
 
+    // KioCG start
+    StructureType<?> API_YUNG_JIGSAW = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "yung_jigsaw", com.yungnickyoung.minecraft.yungsapi.module.StructureTypeModule.YUNG_JIGSAW);
+
+    StructureType<?> BDS_SPIDER_DUNGEON = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "spider_dungeon", com.yungnickyoung.minecraft.betterdungeons.module.StructureTypeModule.SPIDER_DUNGEON);
+    StructureType<?> BDS_SMALL_NETHER_DUNGEON = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon", com.yungnickyoung.minecraft.betterdungeons.module.StructureTypeModule.SMALL_NETHER_DUNGEON);
+
+    StructureType<?> BMS_BETTER_MINESHAFT = register(com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "mineshaft", com.yungnickyoung.minecraft.bettermineshafts.module.StructureTypeModule.BETTER_MINESHAFT);
+
+    private static <S extends Structure> StructureType<S> register(String name, StructureType<S> codec) {
+        return Registry.register(BuiltInRegistries.STRUCTURE_TYPE, name, codec);
+    }
+    // KioCG end
+
     MapCodec<S> codec();
 
     private static <S extends Structure> StructureType<S> register(String name, MapCodec<S> codec) {
diff --git a/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceType.java b/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceType.java
index b8ae40cdc0ffcc05a66e42825f3ba147480bf38d..a29fd6de64db062bf71cda5b3f2d3e8e6ebbe50d 100644
--- a/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceType.java
+++ b/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceType.java
@@ -81,6 +81,25 @@ public interface StructurePieceType {
     StructurePieceType NETHER_FOSSIL = setTemplatePieceId(NetherFossilPieces.NetherFossilPiece::new, "NeFos");
     StructurePieceType JIGSAW = setFullContextPieceId(PoolElementStructurePiece::new, "jigsaw");
 
+    // KioCG start
+    StructurePieceType BDS_BIG_TUNNEL = setFullContextPieceId(com.yungnickyoung.minecraft.betterdungeons.module.StructurePieceTypeModule.BIG_TUNNEL, com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "spider_dungeon_big_tunnel_piece");
+    StructurePieceType BDS_NEST = setFullContextPieceId(com.yungnickyoung.minecraft.betterdungeons.module.StructurePieceTypeModule.NEST, com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "spider_dungeon_nest_piece");
+    StructurePieceType BDS_SMALL_TUNNEL = setFullContextPieceId(com.yungnickyoung.minecraft.betterdungeons.module.StructurePieceTypeModule.SMALL_TUNNEL, com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "spider_dungeon_small_tunnel_piece");
+    StructurePieceType BDS_EGG_ROOM = setFullContextPieceId(com.yungnickyoung.minecraft.betterdungeons.module.StructurePieceTypeModule.EGG_ROOM, com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "spider_dungeon_egg_room_piece");
+
+    StructurePieceType BMS_VERTICAL_ENTRANCE = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.VERTICAL_ENTRANCE, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmsverticalentrance");
+    StructurePieceType BMS_BIG_TUNNEL = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.BIG_TUNNEL, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmsbigtunnel");
+    StructurePieceType BMS_SMALL_TUNNEL = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.SMALL_TUNNEL, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmssmalltunnel");
+    StructurePieceType BMS_SMALL_TUNNEL_TURN = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.SMALL_TUNNEL_TURN, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmssmalltunnelturn");
+    StructurePieceType BMS_SMALL_TUNNEL_STAIRS = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.SMALL_TUNNEL_STAIRS, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmssmalltunnelstairs");
+    StructurePieceType BMS_SIDE_ROOM = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.SIDE_ROOM, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmssideroom");
+    StructurePieceType BMS_SIDE_ROOM_DUNGEON = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.SIDE_ROOM_DUNGEON, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmssideroomdungeon");
+    StructurePieceType BMS_ORE_DEPOSIT = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.ORE_DEPOSIT, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmsoredeposit");
+    StructurePieceType BMS_LAYERED_INTERSECTION_4 = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.LAYERED_INTERSECTION_4, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmslayeredintersection4");
+    StructurePieceType BMS_LAYERED_INTERSECTION_5 = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.LAYERED_INTERSECTION_5, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmslayeredintersection5");
+    StructurePieceType BMS_ZOMBIE_VILLAGER_ROOM = setFullContextPieceId(com.yungnickyoung.minecraft.bettermineshafts.module.StructurePieceTypeModule.ZOMBIE_VILLAGER_ROOM, com.yungnickyoung.minecraft.bettermineshafts.BetterMineshaftsCommon.MOD_ID_ + "bmszombievillagerroom");
+    // KioCG end
+
     StructurePiece load(StructurePieceSerializationContext context, CompoundTag tag);
 
     private static StructurePieceType setFullContextPieceId(StructurePieceType pieceType, String pieceId) {
diff --git a/net/minecraft/world/level/levelgen/structure/placement/StructurePlacementType.java b/net/minecraft/world/level/levelgen/structure/placement/StructurePlacementType.java
index 1d5e72c2255acec0f1b12e5751d1a1976a9afc0a..e2ecfb8e8c1a464c2092186d69af2d56fced8d16 100644
--- a/net/minecraft/world/level/levelgen/structure/placement/StructurePlacementType.java
+++ b/net/minecraft/world/level/levelgen/structure/placement/StructurePlacementType.java
@@ -8,6 +8,20 @@ public interface StructurePlacementType<SP extends StructurePlacement> {
     StructurePlacementType<RandomSpreadStructurePlacement> RANDOM_SPREAD = register("random_spread", RandomSpreadStructurePlacement.CODEC);
     StructurePlacementType<ConcentricRingsStructurePlacement> CONCENTRIC_RINGS = register("concentric_rings", ConcentricRingsStructurePlacement.CODEC);
 
+    // KioCG start
+    StructurePlacementType<?> API_ENHANCED_RANDOM_SPREAD = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "enhanced_random_spread", com.yungnickyoung.minecraft.yungsapi.module.StructurePlacementTypeModule.ENHANCED_RANDOM_SPREAD);
+
+    StructurePlacementType<?> BDT_BETTER_DESERT_TEMPLE_PLACEMENT = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "desert_temple", com.yungnickyoung.minecraft.betterdeserttemples.module.StructurePlacementTypeModule.BETTER_DESERT_TEMPLE_PLACEMENT);
+
+    StructurePlacementType<?> BJT_BETTER_JUNGLE_TEMPLE_PLACEMENT = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "jungle_temple", com.yungnickyoung.minecraft.betterjungletemples.module.StructurePlacementTypeModule.BETTER_JUNGLE_TEMPLE_PLACEMENT);
+
+    StructurePlacementType<?> BSH_BETTER_STRONGHOLD_PLACEMENT = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "stronghold", com.yungnickyoung.minecraft.betterstrongholds.module.StructurePlacementTypeModule.BETTER_STRONGHOLD_PLACEMENT);
+
+    private static <SP extends StructurePlacement> StructurePlacementType<SP> register(String name, StructurePlacementType<SP> codec) {
+        return Registry.register(BuiltInRegistries.STRUCTURE_PLACEMENT, name, codec);
+    }
+    // KioCG end
+
     MapCodec<SP> codec();
 
     private static <SP extends StructurePlacement> StructurePlacementType<SP> register(String name, MapCodec<SP> codec) {
diff --git a/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java b/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java
index f24ed59d75188a14f7d42787adb6fa1c9d7d56b6..d5e7cb74ef7af4471438cf9c13687fb1e6366b47 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java
@@ -31,7 +31,7 @@ public class FeaturePoolElement extends StructurePoolElement {
             .apply(instance, FeaturePoolElement::new)
     );
     private static final ResourceLocation DEFAULT_JIGSAW_NAME = ResourceLocation.withDefaultNamespace("bottom");
-    private final Holder<PlacedFeature> feature;
+    public final Holder<PlacedFeature> feature; // KioCG - private -> public
     private final CompoundTag defaultJigsawNBT;
 
     protected FeaturePoolElement(Holder<PlacedFeature> feature, StructureTemplatePool.Projection projection) {
diff --git a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
index 5c081a5b3d10f713e4e82fe1a43758f553fe50e0..8271f0393ce6f8025d1ab5c88e94e7aad80023af 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
@@ -88,7 +88,7 @@ public class SinglePoolElement extends StructurePoolElement {
         return template.getSize(rotation);
     }
 
-    private StructureTemplate getTemplate(StructureTemplateManager structureTemplateManager) {
+    public StructureTemplate getTemplate(StructureTemplateManager structureTemplateManager) { // KioCG - private -> public
         return this.template.map(structureTemplateManager::getOrCreate, Function.identity());
     }
 
diff --git a/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElementType.java b/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElementType.java
index 4c87c071c0229e000503676813653e007c2ccaba..a86b303d21573438280343d47b811966c2028be5 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElementType.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElementType.java
@@ -11,6 +11,19 @@ public interface StructurePoolElementType<P extends StructurePoolElement> {
     StructurePoolElementType<EmptyPoolElement> EMPTY = register("empty_pool_element", EmptyPoolElement.CODEC);
     StructurePoolElementType<LegacySinglePoolElement> LEGACY = register("legacy_single_pool_element", LegacySinglePoolElement.CODEC);
 
+    // KioCG start
+    StructurePoolElementType<?> API_MAX_COUNT_SINGLE_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "max_count_single_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.MAX_COUNT_SINGLE_ELEMENT);
+    StructurePoolElementType<?> API_MAX_COUNT_LEGACY_SINGLE_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "max_count_legacy_single_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.MAX_COUNT_LEGACY_SINGLE_ELEMENT);
+    StructurePoolElementType<?> API_MAX_COUNT_FEATURE_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "max_count_feature_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.MAX_COUNT_FEATURE_ELEMENT);
+    StructurePoolElementType<?> API_MAX_COUNT_LIST_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "max_count_list_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.MAX_COUNT_LIST_ELEMENT);
+    StructurePoolElementType<?> API_YUNG_SINGLE_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "yung_single_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.YUNG_SINGLE_ELEMENT);
+    StructurePoolElementType<?> API_YUNG_FEATURE_ELEMENT = register(com.yungnickyoung.minecraft.yungsapi.YungsApiCommon.MOD_ID_ + "yung_feature_element", com.yungnickyoung.minecraft.yungsapi.module.StructurePoolElementTypeModule.YUNG_FEATURE_ELEMENT);
+
+    static <P extends StructurePoolElement> StructurePoolElementType<P> register(String name, StructurePoolElementType<P> codec) {
+        return Registry.register(BuiltInRegistries.STRUCTURE_POOL_ELEMENT, name, codec);
+    }
+    // KioCG end
+
     MapCodec<P> codec();
 
     static <P extends StructurePoolElement> StructurePoolElementType<P> register(String name, MapCodec<P> codec) {
diff --git a/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java b/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
index 03edf021df78da367f3185b378810636fd89c5b1..03772af3eb2665795a547195e73d6bf4142b7b1d 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
@@ -29,7 +29,7 @@ public class StructureTemplatePool {
     public static final Codec<StructureTemplatePool> DIRECT_CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 Codec.lazyInitialized(CODEC_REFERENCE::getValue).fieldOf("fallback").forGetter(StructureTemplatePool::getFallback),
-                Codec.mapPair(StructurePoolElement.CODEC.fieldOf("element"), Codec.intRange(1, 150).fieldOf("weight"))
+                Codec.mapPair(StructurePoolElement.CODEC.fieldOf("element"), Codec.intRange(1, 5000).fieldOf("weight")) // KioCG
                     .codec()
                     .listOf()
                     .fieldOf("elements")
@@ -40,7 +40,7 @@ public class StructureTemplatePool {
     public static final Codec<Holder<StructureTemplatePool>> CODEC = Util.make(
         RegistryFileCodec.create(Registries.TEMPLATE_POOL, DIRECT_CODEC), CODEC_REFERENCE::setValue
     );
-    private final List<Pair<StructurePoolElement, Integer>> rawTemplates;
+    public final List<Pair<StructurePoolElement, Integer>> rawTemplates; // KioCG - private -> public
     private final ObjectArrayList<StructurePoolElement> templates;
     private final Holder<StructureTemplatePool> fallback;
     private int maxSize = Integer.MIN_VALUE;
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java
index fbddece382f84a1795ebe84bbbff9b20a248b57d..515303e12d71260ac33abc1c557933b4ef421b0e 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java
@@ -27,6 +27,117 @@ public interface StructureProcessorType<P extends StructureProcessor> {
     StructureProcessorType<ProtectedBlockProcessor> PROTECTED_BLOCKS = register("protected_blocks", ProtectedBlockProcessor.CODEC);
     StructureProcessorType<CappedProcessor> CAPPED = register("capped", CappedProcessor.CODEC);
 
+    // KioCG start
+    StructureProcessorType<?> BDT_POLISHED_DIORITE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "polished_diorite_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.POLISHED_DIORITE_PROCESSOR);
+    StructureProcessorType<?> BDT_DIORITE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "diorite_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.DIORITE_PROCESSOR);
+    StructureProcessorType<?> BDT_SPONGE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "sponge_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.SPONGE_PROCESSOR);
+    StructureProcessorType<?> BDT_END_STONE_BRICK_WALL_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "end_stone_brick_wall_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.END_STONE_BRICK_WALL_PROCESSOR);
+    StructureProcessorType<?> BDT_WHITE_STAINED_GLASS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "white_stained_glass_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.WHITE_STAINED_GLASS_PROCESSOR);
+    StructureProcessorType<?> BDT_LIME_BANNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "lime_banner_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.LIME_BANNER_PROCESSOR);
+    StructureProcessorType<?> BDT_RED_BANNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "red_banner_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.RED_BANNER_PROCESSOR);
+    StructureProcessorType<?> BDT_PURPUR_PILLAR_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "purpur_pillar_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.PURPUR_PILLAR_PROCESSOR);
+    StructureProcessorType<?> BDT_QUARTZ_PILLAR_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "quartz_pillar_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.QUARTZ_PILLAR_PROCESSOR);
+    StructureProcessorType<?> BDT_ACACIA_WOOD_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "acacia_wood_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.ACACIA_WOOD_PROCESSOR);
+    StructureProcessorType<?> BDT_INFESTED_CRACKED_STONE_BRICKS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "infested_cracked_stone_bricks_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.INFESTED_CRACKED_STONE_BRICKS_PROCESSOR);
+    StructureProcessorType<?> BDT_BONE_BLOCK_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "bone_block_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.BONE_BLOCK_PROCESSOR);
+    StructureProcessorType<?> BDT_YELLOW_WOOL_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "yellow_wool_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.YELLOW_WOOL_PROCESSOR);
+    StructureProcessorType<?> BDT_YELLOW_CONCRETE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "yellow_concrete_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.YELLOW_CONCRETE_PROCESSOR);
+    StructureProcessorType<?> BDT_BLUE_CONCRETE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "blue_concrete_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.BLUE_CONCRETE_PROCESSOR);
+    StructureProcessorType<?> BDT_GRAVEL_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "gravel_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.GRAVEL_PROCESSOR);
+    StructureProcessorType<?> BDT_TORCH_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "torch_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.TORCH_PROCESSOR);
+    StructureProcessorType<?> BDT_LIT_CAMPFIRE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "lit_campfire_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.LIT_CAMPFIRE_PROCESSOR);
+    StructureProcessorType<?> BDT_YELLOW_STAINED_GLASS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "yellow_stained_glass_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.YELLOW_STAINED_GLASS_PROCESSOR);
+    StructureProcessorType<?> BDT_ORANGE_STAINED_GLASS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "orange_stained_glass_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.ORANGE_STAINED_GLASS_PROCESSOR);
+    StructureProcessorType<?> BDT_RED_STAINED_GLASS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "red_stained_glass_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.RED_STAINED_GLASS_PROCESSOR);
+    StructureProcessorType<?> BDT_RED_WOOL_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "red_wool_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.RED_WOOL_PROCESSOR);
+    StructureProcessorType<?> BDT_POT_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "pot_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.POT_PROCESSOR);
+    StructureProcessorType<?> BDT_ARMOR_STAND_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "armor_stand_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.ARMOR_STAND_PROCESSOR);
+    StructureProcessorType<?> BDT_ITEM_FRAME_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "item_frame_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.ITEM_FRAME_PROCESSOR);
+    StructureProcessorType<?> BDT_PHARAOH_PROCESSOR = register(com.yungnickyoung.minecraft.betterdeserttemples.BetterDesertTemplesCommon.MOD_ID_ + "pharaoh_processor", com.yungnickyoung.minecraft.betterdeserttemples.module.StructureProcessorModule.PHARAOH_PROCESSOR);
+
+    StructureProcessorType<?> BDS_MOB_SPAWNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "mob_spawner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.MOB_SPAWNER_PROCESSOR);
+    StructureProcessorType<?> BDS_HEAD_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "head_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.HEAD_PROCESSOR);
+    StructureProcessorType<?> BDS_NETHER_BLOCK_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "nether_block_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.NETHER_BLOCK_PROCESSOR);
+    StructureProcessorType<?> BDS_CANDLE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "candle_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.CANDLE_PROCESSOR);
+    /* Small dungeons */
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_CEILING_PROP_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_ceiling_prop_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_CEILING_PROP_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_CEILING_LAMP_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_ceiling_lamp_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_CEILING_LAMP_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_BANNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_banner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_BANNER_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_CHEST_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_chest_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_CHEST_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_COBBLE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_cobblestone_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_COBBLE_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_leg_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_LEG_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_CEILING_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_ceiling_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_CEILING_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_DUNGEON_ORE_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_dungeon_ore_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_DUNGEON_ORE_PROCESSOR);
+    /* Small Nether Dungeons */
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_BANNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_banner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_BANNER_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_HEAD_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_head_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_HEAD_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_LAVA_BLOCK_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_lava_block_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_LAVA_BLOCK_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_ENTRANCE_STAIRS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_entrance_stairs_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_ENTRANCE_STAIRS_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_leg_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_LEG_PROCESSOR);
+    StructureProcessorType<?> BDS_SMALL_NETHER_DUNGEON_MOB_SPAWNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "small_nether_dungeon_mob_spawner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SMALL_NETHER_DUNGEON_MOB_SPAWNER_PROCESSOR);
+    /* Skeleton dungeons */
+    StructureProcessorType<?> BDS_SKELETON_DUNGEON_RUINED_STONE_BRICKS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "skeleton_dungeon_ruined_stone_bricks_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SKELETON_DUNGEON_RUINED_STONE_BRICKS_PROCESSOR);
+    StructureProcessorType<?> BDS_SKELETON_MOB_SPAWNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "skeleton_mob_spawner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SKELETON_MOB_SPAWNER_PROCESSOR);
+    StructureProcessorType<?> BDS_SKELETON_DUNGEON_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "skeleton_dungeon_leg_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.SKELETON_DUNGEON_LEG_PROCESSOR);
+    /* Zombie dungeons */
+    StructureProcessorType<?> BDS_ZOMBIE_DUNGEON_CUBBY_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_dungeon_cubby_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_DUNGEON_CUBBY_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_DUNGEON_STAIR_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_dungeon_stair_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_DUNGEON_STAIR_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_MOB_SPAWNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_mob_spawner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_MOB_SPAWNER_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_TOMBSTONE_SPAWNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_tombstone_spawner_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_TOMBSTONE_SPAWNER_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_MAIN_STAIRS_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_main_stairs_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_MAIN_STAIRS_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_ROT_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_rot_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_ROT_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_DUNGEON_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_dungeon_leg_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_DUNGEON_LEG_PROCESSOR);
+    StructureProcessorType<?> BDS_ZOMBIE_DUNGEON_FLOWER_POT_PROCESSOR = register(com.yungnickyoung.minecraft.betterdungeons.BetterDungeonsCommon.MOD_ID_ + "zombie_dungeon_flower_pot_processor", com.yungnickyoung.minecraft.betterdungeons.module.StructureProcessorTypeModule.ZOMBIE_DUNGEON_FLOWER_POT_PROCESSOR);
+
+    StructureProcessorType<?> BFS_PILLAR_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "pillar_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.PILLAR_PROCESSOR);
+    StructureProcessorType<?> BFS_STAIR_PILLAR_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "stair_pillar_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.STAIR_PILLAR_PROCESSOR);
+    StructureProcessorType<?> BFS_RED_SANDSTONE_STAIRS_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "red_sandstone_stairs_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.RED_SANDSTONE_STAIRS_PROCESSOR);
+    StructureProcessorType<?> BFS_BRIDGE_ARCH_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "bridge_arch_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.BRIDGE_ARCH_PROCESSOR);
+    StructureProcessorType<?> BFS_LIQUID_BLOCK_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "liquid_block_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.LIQUID_BLOCK_PROCESSOR);
+    StructureProcessorType<?> BFS_NETHER_WART_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "nether_wart_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.NETHER_WART_PROCESSOR);
+    StructureProcessorType<?> BFS_ITEM_FRAME_PROCESSOR = register(com.yungnickyoung.minecraft.betterfortresses.BetterFortressesCommon.MOD_ID_ + "item_frame_processor", com.yungnickyoung.minecraft.betterfortresses.module.StructureProcessorTypeModule.ITEM_FRAME_PROCESSOR);
+
+    StructureProcessorType<?> BJT_PILLAR_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "pillar_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.PILLAR_PROCESSOR);
+    StructureProcessorType<?> BJT_BLOCK_REPLACE_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "block_replace_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.BLOCK_REPLACE_PROCESSOR);
+    StructureProcessorType<?> BJT_EMPTY_DISPENSER_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "empty_dispenser_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.EMPTY_DISPENSER_PROCESSOR);
+    StructureProcessorType<?> BJT_FIREBALL_DISPENSER_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "fireball_dispenser_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.FIREBALL_DISPENSER_PROCESSOR);
+    StructureProcessorType<?> BJT_CAVE_VINE_DECORATION_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "cave_vine_decoration_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.CAVE_VINE_DECORATION_PROCESSOR);
+    StructureProcessorType<?> BJT_TORCH_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "torch_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.TORCH_PROCESSOR);
+    StructureProcessorType<?> BJT_BLAST_FURNACE_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "blast_furnace_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.BLAST_FURNACE_PROCESSOR);
+    StructureProcessorType<?> BJT_ITEM_FRAME_PROCESSOR = register(com.yungnickyoung.minecraft.betterjungletemples.BetterJungleTemplesCommon.MOD_ID_ + "item_frame_processor", com.yungnickyoung.minecraft.betterjungletemples.module.StructureProcessorTypeModule.ITEM_FRAME_PROCESSOR);
+
+    StructureProcessorType<?> BOM_AIR_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "air_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.AIR_PROCESSOR);
+    StructureProcessorType<?> BOM_WATERLOG_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "waterlog_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.WATERLOG_PROCESSOR);
+    StructureProcessorType<?> BOM_RANDOM_PRISMARINE_SLAB_DECORATION_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "random_prismarine_slab_decoration_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.RANDOM_PRISMARINE_SLAB_DECORATION_PROCESSOR);
+    StructureProcessorType<?> BOM_RANDOM_DARK_PRISMARINE_SLAB_DECORATION_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "random_dark_prismarine_slab_decoration_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.RANDOM_DARK_PRISMARINE_SLAB_DECORATION_PROCESSOR);
+    StructureProcessorType<?> BOM_STRUCTURE_VOID_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "structure_void_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.STRUCTURE_VOID_PROCESSOR);
+    StructureProcessorType<?> BOM_SAND_GRAVEL_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "sand_gravel_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.SAND_GRAVEL_PROCESSOR);
+    StructureProcessorType<?> BOM_RANDOM_OXIDIZATION_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "random_oxidization_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.RANDOM_OXIDIZATION_PROCESSOR);
+    StructureProcessorType<?> BOM_SEAGRASS_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "seagrass_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.SEAGRASS_PROCESSOR);
+    StructureProcessorType<?> BOM_SPONGE_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "random_sponge_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.SPONGE_PROCESSOR);
+    StructureProcessorType<?> BOM_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betteroceanmonuments.BetterOceanMonumentsCommon.MOD_ID_ + "leg_processor", com.yungnickyoung.minecraft.betteroceanmonuments.module.StructureProcessorTypeModule.LEG_PROCESSOR);
+
+    StructureProcessorType<?> BSH_RUIN_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "ruin_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.RUIN_PROCESSOR);
+    StructureProcessorType<?> BSH_BANNER_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "banner_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.BANNER_PROCESSOR);
+    StructureProcessorType<?> BSH_ORE_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "ore_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.ORE_PROCESSOR);
+    StructureProcessorType<?> BSH_RARE_BLOCK_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "rare_block_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.RARE_BLOCK_PROCESSOR);
+    StructureProcessorType<?> BSH_REDSTONE_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "redstone_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.REDSTONE_PROCESSOR);
+    StructureProcessorType<?> BSH_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "leg_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.LEG_PROCESSOR);
+    StructureProcessorType<?> BSH_END_PORTAL_FRAME_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "end_portal_frame_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.END_PORTAL_FRAME_PROCESSOR);
+    StructureProcessorType<?> BSH_ARMORSTAND_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "armorstand_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.ARMORSTAND_PROCESSOR);
+    StructureProcessorType<?> BSH_ITEMFRAME_PROCESSOR = register(com.yungnickyoung.minecraft.betterstrongholds.BetterStrongholdsCommon.MOD_ID_ + "itemframe_processor", com.yungnickyoung.minecraft.betterstrongholds.module.StructureProcessorTypeModule.ITEMFRAME_PROCESSOR);
+
+    StructureProcessorType<?> BWH_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterwitchhuts.BetterWitchHutsCommon.MOD_ID_ + "leg_processor", com.yungnickyoung.minecraft.betterwitchhuts.module.StructureProcessorTypeModule.LEG_PROCESSOR);
+    StructureProcessorType<?> BWH_FENCE_LEG_PROCESSOR = register(com.yungnickyoung.minecraft.betterwitchhuts.BetterWitchHutsCommon.MOD_ID_ + "fence_leg_processor", com.yungnickyoung.minecraft.betterwitchhuts.module.StructureProcessorTypeModule.FENCE_LEG_PROCESSOR);
+    StructureProcessorType<?> BWH_WITCH_CIRCLE_PROCESSOR = register(com.yungnickyoung.minecraft.betterwitchhuts.BetterWitchHutsCommon.MOD_ID_ + "witch_circle_processor", com.yungnickyoung.minecraft.betterwitchhuts.module.StructureProcessorTypeModule.WITCH_CIRCLE_PROCESSOR);
+    StructureProcessorType<?> BWH_BREWING_STAND_PROCESSOR = register(com.yungnickyoung.minecraft.betterwitchhuts.BetterWitchHutsCommon.MOD_ID_ + "brewing_stand_processor", com.yungnickyoung.minecraft.betterwitchhuts.module.StructureProcessorTypeModule.BREWING_STAND_PROCESSOR);
+    StructureProcessorType<?> BWH_POTTED_MUSHROOM_PROCESSOR = register(com.yungnickyoung.minecraft.betterwitchhuts.BetterWitchHutsCommon.MOD_ID_ + "potted_mushroom_processor", com.yungnickyoung.minecraft.betterwitchhuts.module.StructureProcessorTypeModule.POTTED_MUSHROOM_PROCESSOR);
+
+    static <P extends StructureProcessor> StructureProcessorType<P> register(String name, StructureProcessorType<P> codec) {
+        return Registry.register(BuiltInRegistries.STRUCTURE_PROCESSOR, name, codec);
+    }
+    // KioCG end
+
     MapCodec<P> codec();
 
     static <P extends StructureProcessor> StructureProcessorType<P> register(String name, MapCodec<P> codec) {
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
index bb8b5ca9044ecb33687e7f20cfb0acbf55f887c7..862203e3f9433ce876d45d55da65332d142c9108 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -78,6 +78,8 @@ public class StructureTemplate {
     public org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(StructureTemplate.DATA_TYPE_REGISTRY);
     // CraftBukkit end
 
+    private static final ThreadLocal<com.yungnickyoung.minecraft.yungsapi.world.processor.StructureProcessingContext> context = new ThreadLocal<>(); // KioCG
+
     public Vec3i getSize() {
         return this.size;
     }
@@ -248,6 +250,7 @@ public class StructureTemplate {
     }
 
     public boolean placeInWorld(ServerLevelAccessor serverLevel, BlockPos offset, BlockPos pos, StructurePlaceSettings settings, RandomSource random, int flags) {
+        com.yungnickyoung.minecraft.betterdungeons.world.DungeonContext.initialize(); // KioCG
         if (this.palettes.isEmpty()) {
             return false;
         } else {
@@ -409,6 +412,15 @@ public class StructureTemplate {
                 }
 
                 if (!settings.isIgnoreEntities()) {
+                    // KioCG start
+                    context.set(new com.yungnickyoung.minecraft.yungsapi.world.processor.StructureProcessingContext(
+                            serverLevel,
+                            settings,
+                            offset,
+                            pos,
+                            entityInfoList
+                    ));
+                    // KioCG end
                     this.placeEntities(
                         wrappedAccess, // CraftBukkit
                         offset,
@@ -418,6 +430,7 @@ public class StructureTemplate {
                         boundingBox,
                         settings.shouldFinalizeEntities()
                     );
+                    context.remove(); // KioCG
                 }
 
                 return true;
@@ -494,6 +507,47 @@ public class StructureTemplate {
         @Nullable BoundingBox boundingBox,
         boolean withEntities
     ) {
+        // KioCG start
+        com.yungnickyoung.minecraft.yungsapi.world.processor.StructureProcessingContext ctx = context.get();
+
+        // If structure is not using YUNG's API entity processors, we don't need to do anything
+        if (!ctx.structurePlaceSettings().getProcessors().stream().noneMatch(p -> p instanceof com.yungnickyoung.minecraft.yungsapi.world.processor.StructureEntityProcessor)) {
+
+            // Apply processors to entities
+            List<StructureTemplate.StructureEntityInfo> processedEntities = processEntityInfoList(ctx);
+
+            // Place the processed entities
+            for (StructureTemplate.StructureEntityInfo entityInfo : processedEntities) {
+                BlockPos entityBlockPos = entityInfo.blockPos;
+                if (ctx.structurePlaceSettings().getBoundingBox() == null || ctx.structurePlaceSettings().getBoundingBox().isInside(entityBlockPos)) {
+                    CompoundTag entityNbt = entityInfo.nbt.copy();
+                    Vec3 entityPos = entityInfo.pos;
+                    ListTag listTag = new ListTag();
+                    listTag.add(DoubleTag.valueOf(entityPos.x));
+                    listTag.add(DoubleTag.valueOf(entityPos.y));
+                    listTag.add(DoubleTag.valueOf(entityPos.z));
+                    entityNbt.put("Pos", listTag);
+                    entityNbt.remove("UUID");
+                    tryCreateEntity(serverLevel, entityNbt).ifPresent((entity) -> {
+                        float f = entity.mirror(ctx.structurePlaceSettings().getMirror());
+                        f += entity.getYRot() - entity.rotate(ctx.structurePlaceSettings().getRotation());
+                        entity.snapTo(entityPos.x, entityPos.y, entityPos.z, f, entity.getXRot());
+                        if (ctx.structurePlaceSettings().shouldFinalizeEntities() && entity instanceof Mob) {
+                            ((Mob) entity).finalizeSpawn(serverLevel, serverLevel.getCurrentDifficultyAt(BlockPos.containing(entityPos)), EntitySpawnReason.STRUCTURE, null);
+                        }
+
+                        serverLevel.addFreshEntityWithPassengers(entity);
+                    });
+                }
+            }
+
+            // Cancel the original entity placement to prevent double-spawning entities.
+            // Note that other mods like Porting Lib will still be able to run their own entity processors,
+            // as long as the structure being processed is not also using YUNG's API entity processors.
+            return;
+        }
+        // KioCG end
+
         for (StructureTemplate.StructureEntityInfo structureEntityInfo : this.entityInfoList) {
             BlockPos blockPos = transform(structureEntityInfo.blockPos, mirror, rotation, offset).offset(pos);
             if (boundingBox == null || boundingBox.isInside(blockPos)) {
@@ -527,6 +581,62 @@ public class StructureTemplate {
 
     }
 
+    // KioCG start
+    private List<StructureTemplate.StructureEntityInfo> processEntityInfoList(com.yungnickyoung.minecraft.yungsapi.world.processor.StructureProcessingContext ctx) {
+        List<StructureTemplate.StructureEntityInfo> processedEntities = new ArrayList<>();
+
+        // Extract context data
+        ServerLevelAccessor serverLevelAccessor = ctx.serverLevelAccessor();
+        BlockPos structurePiecePos = ctx.structurePiecePos();
+        BlockPos structurePiecePivotPos = ctx.structurePiecePivotPos();
+        StructurePlaceSettings structurePlaceSettings = ctx.structurePlaceSettings();
+        List<StructureTemplate.StructureEntityInfo> rawEntityInfos = ctx.rawEntityInfos();
+
+        for (StructureTemplate.StructureEntityInfo rawEntityInfo : rawEntityInfos) {
+            // Calculate transformed position so processors have access to the actual global world coordinates of the entity
+            Vec3 globalPos = StructureTemplate
+                    .transform(rawEntityInfo.pos,
+                               structurePlaceSettings.getMirror(),
+                               structurePlaceSettings.getRotation(),
+                               structurePlaceSettings.getRotationPivot())
+                    .add(Vec3.atLowerCornerOf(structurePiecePos));
+            BlockPos globalBlockPos = StructureTemplate
+                    .transform(rawEntityInfo.blockPos,
+                               structurePlaceSettings.getMirror(),
+                               structurePlaceSettings.getRotation(),
+                               structurePlaceSettings.getRotationPivot())
+                    .offset(structurePiecePos);
+            StructureTemplate.StructureEntityInfo globalEntityInfo = new StructureTemplate.StructureEntityInfo(globalPos, globalBlockPos, rawEntityInfo.nbt);
+
+            // Apply processors
+            for (StructureProcessor processor : structurePlaceSettings.getProcessors()) {
+                if (processor instanceof com.yungnickyoung.minecraft.yungsapi.world.processor.StructureEntityProcessor) {
+                    globalEntityInfo = ((com.yungnickyoung.minecraft.yungsapi.world.processor.StructureEntityProcessor) processor).processEntity(serverLevelAccessor, structurePiecePos, structurePiecePivotPos, rawEntityInfo, globalEntityInfo, structurePlaceSettings);
+                    if (globalEntityInfo == null) break;
+                }
+            }
+
+            if (globalEntityInfo != null) { // null value from processor indicates the entity should not be spawned
+                processedEntities.add(globalEntityInfo);
+            }
+        }
+
+        return processedEntities;
+    }
+
+    /**
+     * Attempts to create an entity from a CompoundTag.
+     * If the entity cannot be created, returns an empty Optional.
+     */
+    private static Optional<Entity> tryCreateEntity(ServerLevelAccessor serverLevelAccessor, CompoundTag compoundTag) {
+        try {
+            return EntityType.create(compoundTag, serverLevelAccessor.getLevel(), EntitySpawnReason.STRUCTURE);
+        } catch (Exception exception) {
+            return Optional.empty();
+        }
+    }
+    // KioCG end
+
     private static Optional<Entity> createEntityIgnoreException(ServerLevelAccessor level, CompoundTag tag) {
         // CraftBukkit start
         // try {
