--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
@@ -217,6 +_,13 @@
         return this.getSnapshot().persistentDataContainer;
     }
 
+    // KioCG start - temporary meta
+    @Override
+    public java.util.Map<String, Object> getTemporaryMeta() {
+        return this.getSnapshot().temporaryMeta;
+    }
+    // KioCG end - temporary meta
+
     @Nullable
     public Packet<ClientGamePacketListener> getUpdatePacket(@NotNull Location location) {
         return new ClientboundBlockEntityDataPacket(CraftLocation.toBlockPosition(location), this.snapshot.getType(), this.getUpdateNBT());
@@ -234,4 +_,41 @@
         return !this.snapshotDisabled;
     }
     // Paper end
+
+    // KioCG start - data component API: 从物品复制
+    @Override
+    public <T> T getData(final io.papermc.paper.datacomponent.DataComponentType.Valued<T> type) {
+        return io.papermc.paper.datacomponent.PaperDataComponentType.convertDataComponentValue(this.snapshot.components(), (io.papermc.paper.datacomponent.PaperDataComponentType.ValuedImpl<T, ?>) type);
+    }
+
+    @Override
+    public boolean hasData(final io.papermc.paper.datacomponent.DataComponentType type) {
+        return this.snapshot.components().has(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type));
+    }
+
+    @Override
+    public java.util.Set<io.papermc.paper.datacomponent.DataComponentType> getDataTypes() {
+        return io.papermc.paper.datacomponent.PaperDataComponentType.minecraftToBukkit(this.snapshot.components().keySet());
+    }
+
+    @Override
+    public <T> void setData(final io.papermc.paper.datacomponent.DataComponentType.Valued<T> type, final T value) {
+        com.google.common.base.Preconditions.checkArgument(value != null, "value cannot be null");
+        this.setDataInternal((io.papermc.paper.datacomponent.PaperDataComponentType.ValuedImpl<T, ?>) type, value);
+    }
+
+    @Override
+    public void setData(final io.papermc.paper.datacomponent.DataComponentType.NonValued type) {
+        this.setDataInternal((io.papermc.paper.datacomponent.PaperDataComponentType.NonValuedImpl<?, ?>) type, null);
+    }
+
+    private <A, V> void setDataInternal(final io.papermc.paper.datacomponent.PaperDataComponentType<A, V> type, final A value) {
+        this.snapshot.setComponent(type.getHandle(), type.getAdapter().toVanilla(value, type.getHolder()));
+    }
+
+    @Override
+    public void unsetData(final io.papermc.paper.datacomponent.DataComponentType type) {
+        this.snapshot.setComponent(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type), null);
+    }
+    // KioCG end - data component API: 从物品复制
 }
