--- a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
@@ -31,6 +_,7 @@
     @Override
     public void setItem(int index, ItemStack item) {
         this.getSnapshot().getItems().set(index, CraftItemStack.asNMSCopy(item));
+        this.getSnapshot().shouldTick[index] = true; // KioCG - optimize block entity
     }
 
     @Override
@@ -98,4 +_,11 @@
         return getSnapshot().stopCooking[index];
     }
     // Paper end
+
+    // KioCG start - campfire boiling water
+    @Override
+    public boolean isCauldronBoiling() {
+        return getSnapshot().isCauldronBoiling();
+    }
+    // KioCG end - campfire boiling water
 }
