--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -133,6 +_,7 @@
             }
         }
 
+        if (true) return false; // KioCG - fix redstone quirks: remove bud
         if (signalGetter.hasSignal(pos, Direction.DOWN)) {
             return true;
         } else {
@@ -150,6 +_,11 @@
 
     @Override
     protected boolean triggerEvent(BlockState state, Level level, BlockPos pos, int id, int param) {
+        // KioCG start - chunk hot
+        net.minecraft.world.level.chunk.LevelChunk levelChunk = level.getChunkIfLoaded(pos);
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking();
+        try {
+        // KioCG end - chunk hot
         Direction direction = state.getValue(FACING);
         // Paper start - Protect Bedrock and End Portal/Frames from being destroyed; prevent retracting when we're facing the wrong way (we were replaced before retraction could occur)
         Direction directionQueuedAs = Direction.from3DDataValue(param & 7); // Paper - copied from below
@@ -249,6 +_,7 @@
             level.gameEvent(GameEvent.BLOCK_DEACTIVATE, pos, GameEvent.Context.of(blockState1));
         }
 
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().endTickingAndCount(); } // KioCG - chunk hot
         return true;
     }
 
