--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -785,7 +_,7 @@
                 return this.reconstructPath(node1, node);
             }
 
-            if (node3.distanceTo(node) < node2.distanceTo(node)) {
+            if (node3.distanceToSqr(node) < node2.distanceToSqr(node)) { // KioCG - optimize distanceTo
                 node2 = node3;
             }
 
