--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -390,6 +_,13 @@
             }
         }
 
+        // KioCG start - optimize entity behavior
+        @Override
+        protected int canUseDelay() {
+            return 20;
+        }
+        // KioCG end - optimize entity behavior
+
         @Override
         public boolean canContinueToUse() {
             return this.pursuedBannerItemEntity != null
@@ -400,7 +_,7 @@
         }
 
         private boolean cannotPickUpBanner() {
-            if (!getServerLevel(this.mob).getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING) || !this.mob.canPickUpLoot()) return false; // Paper - respect game and entity rules for picking up items
+            if (false || !this.mob.canPickUpLoot()) return false; // Paper - respect game and entity rules for picking up items // KioCG - better mob griefing setting
             if (!this.mob.hasActiveRaid()) {
                 return true;
             } else if (this.mob.getCurrentRaid().isOver()) {
