--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -125,7 +_,7 @@
     public final int numGroups;
     private Raid.RaidStatus status;
     private int celebrationTicks;
-    private Optional<BlockPos> waveSpawnPos = Optional.empty();
+    private Set<BlockPos> waveSpawnPos = new java.util.HashSet<>(); // KioCG - anti mob farm
 
     public Raid(BlockPos center, Difficulty difficulty) {
         this.active = true;
@@ -157,7 +_,7 @@
         this.ticksActive = ticksActive;
         this.raidOmenLevel = raidOmenLevel;
         this.groupsSpawned = groupsSpawned;
-        this.raidCooldownTicks = raidCooldownTicks;
+        this.raidCooldownTicks = raidCooldownTicks > 0 ? DEFAULT_PRE_RAID_TICKS : 0; // KioCG - anti mob farm: 重新计算袭击可用位置
         this.postRaidTicks = postRaidTicks;
         this.totalHealth = totalHealth;
         this.center = center;
@@ -326,15 +_,17 @@
                             return;
                         }
                     } else {
-                        boolean isPresent = this.waveSpawnPos.isPresent();
-                        boolean flag1 = !isPresent && this.raidCooldownTicks % 5 == 0;
-                        if (isPresent && !level.isPositionEntityTicking(this.waveSpawnPos.get())) {
-                            flag1 = true;
-                        }
+                        // KioCG start - anti mob farm
+                        // boolean isPresent = this.waveSpawnPos.isPresent();
+                        // boolean flag1 = !isPresent && this.raidCooldownTicks % 5 == 0;
+                        // if (isPresent && !level.isPositionEntityTicking(this.waveSpawnPos.get())) {
+                        //     flag1 = true;
+                        // }
 
-                        if (flag1) {
-                            this.waveSpawnPos = this.getValidSpawnPos(level);
+                        if (this.raidCooldownTicks % 5 == 0) {
+                            this.findRandomSpawnPos(level, 8, this.waveSpawnPos);
                         }
+                        // KioCG end - anti mob farm
 
                         if (this.raidCooldownTicks == 300 || this.raidCooldownTicks % 20 == 0) {
                             this.updatePlayers(level);
@@ -365,10 +_,27 @@
                 }
 
                 boolean isPresentx = false;
-                int i = 0;
+                int i = NUM_SPAWN_ATTEMPTS; // KioCG - anti mob farm: 只尝试一次
 
                 while (this.shouldSpawnGroup()) {
-                    BlockPos blockPos = this.waveSpawnPos.orElseGet(() -> this.findRandomSpawnPos(level, 20));
+                    // KioCG start - anti mob farm
+                    List<BlockPos> list = new java.util.ArrayList<>(this.waveSpawnPos);
+                    this.waveSpawnPos.clear();
+                    if (list.size() < 100) {
+                        list.clear();
+                    }
+
+                    BlockPos blockPos = null;
+                    while (!list.isEmpty()) {
+                        blockPos = list.remove(this.random.nextInt(list.size()));
+                        if (!level.isPositionEntityTicking(blockPos)) {
+                            blockPos = null;
+                            continue;
+                        }
+                        break;
+                    }
+                    // KioCG end - anti mob farm
+
                     if (blockPos != null) {
                         this.started = true;
                         this.spawnGroup(level, blockPos);
@@ -437,10 +_,12 @@
         stream.filter(level::isVillage).map(SectionPos::center).min(Comparator.comparingDouble(pos -> pos.distSqr(this.center))).ifPresent(this::setCenter);
     }
 
-    private Optional<BlockPos> getValidSpawnPos(ServerLevel level) {
-        BlockPos blockPos = this.findRandomSpawnPos(level, 8);
-        return blockPos != null ? Optional.of(blockPos) : Optional.empty();
-    }
+    // KioCG start - anti mob farm
+    // private Optional<BlockPos> getValidSpawnPos(ServerLevel level) {
+    //     BlockPos blockPos = this.findRandomSpawnPos(level, 8);
+    //     return blockPos != null ? Optional.of(blockPos) : Optional.empty();
+    // }
+    // KioCG end - anti mob farm
 
     private boolean hasMoreWaves() {
         return this.hasBonusWave() ? !this.hasSpawnedBonusWave() : !this.isFinalWave();
@@ -563,7 +_,7 @@
             }
         }
 
-        this.waveSpawnPos = Optional.empty();
+        this.waveSpawnPos.clear(); // KioCG - anti mob farm
         this.groupsSpawned++;
         this.updateBossbar();
         this.setDirty(level);
@@ -655,8 +_,13 @@
         return this.groupToLeaderMap.get(wave);
     }
 
-    @Nullable
-    private BlockPos findRandomSpawnPos(ServerLevel level, int attempts) {
+    // KioCG start - anti mob farm
+    // @Nullable
+    private Collection<BlockPos> findRandomSpawnPos(ServerLevel level, int attempts, Collection<BlockPos> list) {
+        if (list.size() >= 100) {
+            return list;
+        }
+    // KioCG end - anti mob farm
         int i = this.raidCooldownTicks / 20;
         float f = 0.22F * i - 0.24F;
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
@@ -666,7 +_,7 @@
             float f2 = f1 + (float) Math.PI * i1 / 8.0F;
             int i2 = this.center.getX() + Mth.floor(Mth.cos(f2) * 32.0F * f) + level.random.nextInt(3) * Mth.floor(f);
             int i3 = this.center.getZ() + Mth.floor(Mth.sin(f2) * 32.0F * f) + level.random.nextInt(3) * Mth.floor(f);
-            int height = level.getHeight(Heightmap.Types.WORLD_SURFACE, i2, i3);
+            int height = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, i2, i3); // KioCG - anti mob farm
             if (Mth.abs(height - this.center.getY()) <= 96) {
                 mutableBlockPos.set(i2, height, i3);
                 if (!level.isVillage(mutableBlockPos) || i <= 7) {
@@ -677,13 +_,18 @@
                             RAVAGER_SPAWN_PLACEMENT_TYPE.isSpawnPositionOk(level, mutableBlockPos, EntityType.RAVAGER)
                                 || level.getBlockState(mutableBlockPos.below()).is(Blocks.SNOW) && level.getBlockState(mutableBlockPos).isAir()
                         )) {
-                        return mutableBlockPos;
+                        // KioCG start - anti mob farm
+                        list.add(mutableBlockPos.immutable());
+                        if (list.size() >= 100) {
+                            return list;
+                        }
+                        // KioCG end - anti mob farm
                     }
                 }
             }
         }
 
-        return null;
+        return list; // KioCG - anti mob farm
     }
 
     private boolean addWaveMob(ServerLevel level, int wave, Raider raider) {
