--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -56,6 +_,9 @@
     public boolean canMobPickup = true; // Paper - Item#canEntityPickup
     private int despawnRate = -1; // Paper - Alternative item-despawn-rate
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
+    // KioCG start - optimize block entity: item search hopper
+    public boolean canDecoratedPotSuck = true;
+    // KioCG end - optimize block entity: item search hopper
 
     public ItemEntity(EntityType<? extends ItemEntity> entityType, Level level) {
         super(entityType, level);
@@ -183,16 +_,29 @@
                 }
             }
 
-            if (!this.onGround() || this.getDeltaMovement().horizontalDistanceSqr() > 1.0E-5F || (this.tickCount + this.getId()) % 4 == 0) { // Paper - Diff on change; ActivationRange immunity
+            // KioCG start - optimize block entity: item search hopper // 提取变量
+            Level level = this.level();
+            BlockPos blockPosBelowThatAffectsMyMovement = null;
+            net.minecraft.world.level.block.state.BlockState belowState = null;
+            // KioCG end - optimize block entity: item search hopper // 提取变量
+
+            // KioCG start - item entity collision
+            boolean checkPush = true;
+            if (!this.onGround() || (checkPush = this.getDeltaMovement().horizontalDistanceSqr() > 1.0E-5F) || (this.tickCount + this.getId()) % 4 == 0) { // Paper - Diff on change; ActivationRange immunity
+            // KioCG end - item entity collision
                 this.move(MoverType.SELF, this.getDeltaMovement());
                 this.applyEffectsFromBlocks();
                 float f = 0.98F;
+                // KioCG start - optimize block entity: item search hopper // 提取变量
+                blockPosBelowThatAffectsMyMovement = this.getBlockPosBelowThatAffectsMyMovement();
+                belowState = level.getBlockState(blockPosBelowThatAffectsMyMovement);
+                // KioCG end - optimize block entity: item search hopper // 提取变量
                 // Paper start - Friction API
                 if (this.frictionState == net.kyori.adventure.util.TriState.FALSE) {
                     f = 1F;
                 } else if (this.onGround()) {
                     // Paper end - Friction API
-                    f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                    f = belowState.getBlock().getFriction() * 0.98F; // KioCG - optimize block entity: item search hopper // 提取变量
                 }
 
                 this.setDeltaMovement(this.getDeltaMovement().multiply(f, 0.98, f));
@@ -202,7 +_,17 @@
                         this.setDeltaMovement(deltaMovement1.multiply(1.0, -0.5, 1.0));
                     }
                 }
-            }
+
+                if (checkPush) this.pushEntities(); // KioCG - item entity collision
+            }
+
+            // KioCG start - optimize block entity: item search hopper
+            if (blockPosBelowThatAffectsMyMovement == null) {
+                blockPosBelowThatAffectsMyMovement = this.getBlockPosBelowThatAffectsMyMovement();
+                belowState = level.getBlockState(blockPosBelowThatAffectsMyMovement);
+            }
+            belowState.itemOnTop(level, blockPosBelowThatAffectsMyMovement, this, net.minecraft.world.entity.InsideBlockEffectApplier.NOOP);
+            // KioCG end - optimize block entity: item search hopper
 
             boolean flag = Mth.floor(this.xo) != Mth.floor(this.getX())
                 || Mth.floor(this.yo) != Mth.floor(this.getY())
@@ -236,6 +_,37 @@
         }
     }
 
+    // KioCG start - item entity collision
+    protected void pushEntities() {
+        if (this.level().paperConfig().collisions.maxEntityCollisions <= 0) {
+            return;
+        }
+
+        this.numCollisions = Math.max(0, this.numCollisions - this.level().paperConfig().collisions.maxEntityCollisions);
+        if (this.numCollisions >= this.level().paperConfig().collisions.maxEntityCollisions) {
+            return;
+        }
+
+        java.util.List<ItemEntity> entities = this.level().getEntitiesOfClass(ItemEntity.class, this.getBoundingBox(), itemEntity -> itemEntity.isAlive() && itemEntity != this);
+        // Paper start - Cap entity collisions
+        for (Entity entity1 : entities) {
+            if (this.numCollisions >= this.level().paperConfig().collisions.maxEntityCollisions) {
+                break;
+            }
+
+            entity1.numCollisions++;
+            this.numCollisions++;
+            // Paper end - Cap entity collisions
+            entity1.push(this);
+        }
+    }
+
+    @Override
+    public boolean isCollidable(boolean ignoreClimbing) {
+        return this.isAlive();
+    }
+    // KioCG end - item entity collision
+
     @Override
     public BlockPos getBlockPosBelowThatAffectsMyMovement() {
         return this.getOnPos(0.999999F);
@@ -344,7 +_,7 @@
     public final boolean hurtServer(ServerLevel level, DamageSource damageSource, float amount) {
         if (this.isInvulnerableToBase(damageSource)) {
             return false;
-        } else if (!level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && damageSource.getEntity() instanceof Mob) {
+        } else if (!level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING1) && damageSource.getEntity() instanceof Mob) { // KioCG - better mob griefing setting
             return false;
         } else if (!this.getItem().canBeHurtBy(damageSource)) {
             return false;
@@ -368,6 +_,7 @@
 
     @Override
     public boolean ignoreExplosion(Explosion explosion) {
+        if (this.level().paperConfig().kiocgConfig.entity.itemEntityExplosionKnockback) return false; // KioCG - item entity explosion knockback
         return !explosion.shouldAffectBlocklikeEntities() || super.ignoreExplosion(explosion);
     }
 
@@ -436,11 +_,12 @@
                 }
             }
 
+            final boolean canPickupOwner = this.target == null || this.target.equals(entity.getUUID()); // KioCG - fix pickup item event cancelled: 提取变量
             if (this.pickupDelay <= 0 && canHold > 0) {
                 item.setCount(canHold);
                 // Call legacy event
                 org.bukkit.event.player.PlayerPickupItemEvent playerEvent = new org.bukkit.event.player.PlayerPickupItemEvent((org.bukkit.entity.Player) entity.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
-                playerEvent.setCancelled(!playerEvent.getPlayer().getCanPickupItems());
+                playerEvent.setCancelled(!playerEvent.getPlayer().getCanPickupItems() || !canPickupOwner); // KioCG - fix pickup item event cancelled
                 this.level().getCraftServer().getPluginManager().callEvent(playerEvent);
                 flyAtPlayer = playerEvent.getFlyAtPlayer(); // Paper
                 if (playerEvent.isCancelled()) {
@@ -455,7 +_,7 @@
 
                 // Call newer event afterwards
                 org.bukkit.event.entity.EntityPickupItemEvent entityEvent = new org.bukkit.event.entity.EntityPickupItemEvent(entity.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
-                entityEvent.setCancelled(!entityEvent.getEntity().getCanPickupItems());
+                entityEvent.setCancelled(!entityEvent.getEntity().getCanPickupItems() || !canPickupOwner); // KioCG - fix pickup item event cancelled
                 this.level().getCraftServer().getPluginManager().callEvent(entityEvent);
                 if (entityEvent.isCancelled()) {
                     item.setCount(count); // SPIGOT-5294 - restore count
@@ -478,7 +_,17 @@
             }
             // CraftBukkit end
             // Paper end - PlayerAttemptPickupItemEvent
-            if (this.pickupDelay == 0 && (this.target == null || this.target.equals(entity.getUUID())) && entity.getInventory().add(item)) {
+            // KioCG start - allow overstacked items: fix pickup animation
+            if (this.pickupDelay == 0 && canPickupOwner) { // KioCG - fix pickup item event cancelled: 提取变量
+                if (!entity.getInventory().add(item) && flyAtPlayer) {
+                    org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(this.position(), this.level().getWorld(), this.getYRot(), this.getXRot());
+                    org.bukkit.entity.Item fakeItem = this.level().getWorld().createEntity(location, org.bukkit.entity.Item.class);
+                    fakeItem.setItemStack(item.asBukkitMirror());
+                    fakeItem.getItemStack().setAmount(canHold);
+                    fakeItem.sendSpawn(new java.util.ArrayList<>(this.getBukkitEntity().getTrackedBy()), location);
+                    entity.take(((org.bukkit.craftbukkit.entity.CraftItem) fakeItem).getHandle(), count);
+                } else {
+            // KioCG end - allow overstacked items: fix pickup animation
                 if (flyAtPlayer) // Paper - PlayerPickupItemEvent
                 entity.take(this, count);
                 if (item.isEmpty()) {
@@ -489,6 +_,7 @@
                 entity.awardStat(Stats.ITEM_PICKED_UP.get(item1), count);
                 entity.onItemPickup(this);
             }
+            } // KioCG - allow overstacked items: fix pickup animation
         }
     }
 
@@ -562,6 +_,12 @@
     public boolean hasPickUpDelay() {
         return this.pickupDelay > 0;
     }
+
+    // KioCG start - change container behavior
+    public boolean hasCanPickUpDelay() {
+        return hasPickUpDelay() && this.pickupDelay != INFINITE_PICKUP_DELAY;
+    }
+    // KioCG end - change container behavior
 
     public void setUnlimitedLifetime() {
         this.age = -32768;
