--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -115,6 +_,15 @@
             presetBlockStateBitsEndStoneGlobal = null;
         }
 
+        // KioCG start - optimize anti xray: 模式1隐藏所有不可见方块来节约宽带
+        if (engineMode == EngineMode.HIDE) {
+            for (BlockState blockState : Block.BLOCK_STATE_REGISTRY) {
+                if (!blockState.isAir()) {
+                    obfuscateGlobal[GLOBAL_BLOCKSTATE_PALETTE.idFor(blockState)] = true;
+                }
+            }
+        } else
+        // KioCG end - optimize anti xray: 模式1隐藏所有不可见方块来节约宽带
         for (Block block : toObfuscate) {
 
             // Don't obfuscate air because air causes unnecessary block updates and causes block updates to fail in the void
@@ -609,18 +_,18 @@
     @Override
     public void onBlockChange(Level level, BlockPos blockPos, BlockState newBlockState, BlockState oldBlockState, int flags, int maxUpdateDepth) {
         if (oldBlockState != null && solidGlobal[GLOBAL_BLOCKSTATE_PALETTE.idFor(oldBlockState)] && !solidGlobal[GLOBAL_BLOCKSTATE_PALETTE.idFor(newBlockState)] && blockPos.getY() <= maxBlockHeightUpdatePosition) {
-            updateNearbyBlocks(level, blockPos);
+            updateNearbyBlocks(level, blockPos, Math.min(updateRadius, 1)); // KioCG - optimize anti xray
         }
     }
 
     @Override
     public void onPlayerLeftClickBlock(ServerPlayerGameMode serverPlayerGameMode, BlockPos blockPos, ServerboundPlayerActionPacket.Action action, Direction direction, int worldHeight, int sequence) {
         if (blockPos.getY() <= maxBlockHeightUpdatePosition) {
-            updateNearbyBlocks(serverPlayerGameMode.level, blockPos);
+            updateNearbyBlocks(serverPlayerGameMode.level, blockPos, updateRadius); // KioCG - optimize anti xray
         }
     }
 
-    private void updateNearbyBlocks(Level level, BlockPos blockPos) {
+    private void updateNearbyBlocks(Level level, BlockPos blockPos, int updateRadius) { // KioCG - optimize anti xray
         if (updateRadius >= 2) {
             BlockPos temp = blockPos.west();
             updateBlock(level, temp);
