--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -802,6 +_,12 @@
 
         // Don't fire on respawn
         if (fromWorld != level) {
+            // KioCG start - portal only pass through player
+            if (level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER) {
+                player.enteredNetherDimension = fromWorld.dimension();
+            }
+            // KioCG end - portal only pass through player
+
             org.bukkit.event.player.PlayerChangedWorldEvent event = new org.bukkit.event.player.PlayerChangedWorldEvent(player.getBukkitEntity(), fromWorld.getWorld());
             this.server.server.getPluginManager().callEvent(event);
         }
@@ -1254,7 +_,7 @@
     public void broadcastChatMessage(PlayerChatMessage message, Predicate<ServerPlayer> shouldFilterMessageTo, @Nullable ServerPlayer sender, ChatType.Bound boundChatType, @Nullable Function<net.kyori.adventure.audience.Audience, Component> unsignedFunction) {
         // Paper end
         boolean flag = this.verifyChatTrusted(message);
-        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), boundChatType, flag ? null : "Not Secure"); // Paper
+        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), boundChatType, null); // Paper // KioCG - remove "Not Secure"
         OutgoingChatMessage outgoingChatMessage = OutgoingChatMessage.create(message);
         boolean flag1 = false;
 
